Index: twisted/names/hosts.py
===================================================================
--- twisted/names/hosts.py	(revision 29567)
+++ twisted/names/hosts.py	(working copy)
@@ -14,21 +14,23 @@
         fp = open(file)
     except:
         return None
+    try:
 
-    lines = fp.readlines()
-    for line in lines:
-        idx = line.find('#')
-        if idx != -1:
-            line = line[:idx]
-        if not line:
-            continue
-        parts = line.split()
-        if name.lower() in [s.lower() for s in parts[1:]]:
-            return parts[0]
-    return None
+        lines = fp.readlines()
+        for line in lines:
+            idx = line.find('#')
+            if idx != -1:
+                line = line[:idx]
+            if not line:
+                continue
+            parts = line.split()
+            if name.lower() in [s.lower() for s in parts[1:]]:
+                return parts[0]
+        return None
+    finally:
+        fp.close()
 
 
-
 class Resolver(common.ResolverBase, styles.Versioned):
     """A resolver that services hosts(5) format files."""
     #TODO: IPv6 support
