Index: doc/core/development/policy/coding-standard.xhtml
===================================================================
--- doc/core/development/policy/coding-standard.xhtml	(revision 32735)
+++ doc/core/development/policy/coding-standard.xhtml	(working copy)
@@ -281,18 +281,9 @@
     <a href="http://epydoc.sourceforge.net/manual-epytext.html"
     >Epytext Markup Language documentation</a>.</p>
 
-    <p>Additionally, to accommodate emacs users:</p>
-
-    <ul>
-      <li>Single quotes of the type of the docstring's triple-quote
-      should be escaped. This will prevent font-lock from
-      accidentally fontifying large portions of the file as a
-      string.</li>
-
-      <li>Code examples in docstrings should be prefixed by the |
-      character. This will prevent IM-Python from regarding sample
-      code as real functions, methods, and classes.</li>
-    </ul>
+    <p>Additionally, to accommodate emacs users, single quotes of the type of
+    the docstring's triple-quote should be escaped. This will prevent font-lock from
+    accidentally fontifying large portions of the file as a string.</p>
 
     <p>For example,</p>
 <pre class="python">
@@ -309,13 +300,13 @@
 
     For example::
 
-      |  import wombat
-      |  def sample(something):
-      |      f = something.getFoo()
-      |      f.doFooThing()
-      |      b = wombat.foo2bar(f)
-      |      b.doBarThing()
-      |      return b
+        import wombat
+        def sample(something):
+            f = something.getFoo()
+            f.doFooThing()
+            b = wombat.foo2bar(f)
+            b.doBarThing()
+            return b
 
     """
     # Optionally, actual code can go here.
