diff --git a/twisted/test/test_defer.py b/twisted/test/test_defer.py
index c8b966b..03cec12 100644
--- a/twisted/test/test_defer.py
+++ b/twisted/test/test_defer.py
@@ -34,7 +34,7 @@ class DeferredTestCase(unittest.TestCase):
     def _callback(self, *args, **kw):
         """
         A callback handler that records its given arguments in
-        C{self.callback_results}
+        C{self.callbackResults}
         """
         self.callbackResults = args, kw
         return args[0]
@@ -42,14 +42,14 @@ class DeferredTestCase(unittest.TestCase):
     def _callback2(self, *args, **kw):
         """
         A callback handler that records its given arguments in
-        C{self.callback2_results}, and converts the result to None.
+        C{self.callback2Results}, and converts the result to None.
         """
         self.callback2Results = args, kw
 
     def _errback(self, *args, **kw):
         """
         An errback handler that records its given arguments in
-        C{self.errback_results}, and converts the result to None.
+        C{self.errbackResults}, and converts the result to None.
         """
         self.errbackResults = args, kw
 
@@ -145,7 +145,7 @@ class DeferredTestCase(unittest.TestCase):
                       defer.Deferred.addBoth]:
             try:
                 cback(deferred)
-            except TypeError, e:
+            except TypeError:
                 pass
             else:
                 self.assert_(False, "addCallback did not raise TypeError")
