Index: doc/core/howto/constants.xhtml
===================================================================
--- doc/core/howto/constants.xhtml	(revision 38455)
+++ doc/core/howto/constants.xhtml	(working copy)
@@ -236,7 +236,7 @@
         Return True if the given status is associated with a response body,
         False otherwise.
         """
-        return status in (cls.NO_CONTENT, cls.NOT_MODIFIED)
+        return status not in (cls.NO_CONTENT, cls.NOT_MODIFIED)
   </pre>
 
   <p>This functionality can be used as any class methods are used:</p>
