--- twisted/words/protocols/irc.py
+++ twisted/words/protocols/irc.py
@@ -929,7 +929,8 @@ class IRCClient(basic.LineReceiver):
         Send a message to a channel
 
         @type channel: C{str}
-        @param channel: The channel to say the message on.
+        @param channel: The channel to say the message on. If it has no
+            prefix, C{'#'} will to prepended to it.
         @type message: C{str}
         @param message: The message to say.
         @type length: C{int}
