diff --git a/twisted/internet/test/_posixifaces.py b/twisted/internet/test/_posixifaces.py
index 1238233..7b9e2ad 100644
--- a/twisted/internet/test/_posixifaces.py
+++ b/twisted/internet/test/_posixifaces.py
@@ -30,7 +30,7 @@ if _PY3:
 
 libc = CDLL(find_library("c"))
 
-if sys.platform == 'darwin':
+if sys.platform.startswith('freebsd') or sys.platform == 'darwin':
     _sockaddrCommon = [
         ("sin_len", c_uint8),
         ("sin_family", c_uint8),
@@ -144,5 +144,12 @@ def posixGetLinkLocalIPv6Addresses():
         interface = nativeString(interface)
         address = nativeString(address)
         if family == socket.AF_INET6 and address.startswith('fe80:'):
+            if sys.platform.startswith('freebsd'):
+                ## kill embedded interface indices in link-local scoped addresses
+                ## https://twistedmatrix.com/trac/ticket/6843
+                ## http://www.freebsd.org/doc/en/books/developers-handbook/ipv6.html#ipv6-scope-index
+                l = address.split(':')
+                l[1] = ''
+                address = ':'.join(l)
             retList.append('%s%%%s' % (address, interface))
     return retList
