diff --git a/twisted/internet/test/_posixifaces.py b/twisted/internet/test/_posixifaces.py
index 1238233..e205e6a 100644
--- a/twisted/internet/test/_posixifaces.py
+++ b/twisted/internet/test/_posixifaces.py
@@ -30,7 +30,10 @@ if _PY3:
 
 libc = CDLL(find_library("c"))
 
-if sys.platform == 'darwin':
+if sys.platform.startswith('freebsd') or \
+   sys.platform.startswith('netbsd') or \
+   sys.platform.startswith('openbsd') or \
+   sys.platform == 'darwin':
     _sockaddrCommon = [
         ("sin_len", c_uint8),
         ("sin_family", c_uint8),
