Index: twisted/web/test/test_agent.py
===================================================================
--- twisted/web/test/test_agent.py	(revision 46644)
+++ twisted/web/test/test_agent.py	(working copy)
@@ -2770,7 +2770,7 @@
         is used by L{DummyResponse.protocol} to make a connection.
     """
 
-    code = b"200"
+    code = 200
     phrase = b"OK"
 
     def __init__(self, headers=None, transportFactory=AbortableStringTransport):
