Index: twisted/web/test/test_xml.py
===================================================================
--- twisted/web/test/test_xml.py	(revision 19805)
+++ twisted/web/test/test_xml.py	(working copy)
@@ -683,7 +683,20 @@
         self.assertEquals(parent.toxml(),
                           '<div xmlns="http://www.w3.org/1999/xhtml"><ol></ol></div>')
 
+    def testNodeEquality(self):
+        """
+        Check that nodes that are different return false for isEqualToNode
+        """
 
+        a = '<a href="hello">bob</a>'
+        b = '<p>Hoity Toity Text</p>'
+        an = microdom.parseString(a).documentElement
+        bn = microdom.parseString(b).documentElement
+
+        self.failIf(an.isEqualToNode(bn), "%s is equal to %s" % (a,
+                                                                 b))
+
+
 class TestBrokenHTML(TestCase):
     """
     Tests for when microdom encounters very bad HTML and C{beExtremelyLenient}
