Index: win32/pyx.x-foo.iss.template
===================================================================
--- win32/pyx.x-foo.iss.template	(revision 23189)
+++ win32/pyx.x-foo.iss.template	(working copy)
@@ -24,6 +24,7 @@
 ; *.pyc keeps this directory around
 Type: filesandordirs; Name: "{app}\twisted"
 Type: filesandordirs; Name: "{app}\TwistedDocs"
+Type: files; Name: "{app}\zope\interface\*.pyc"
 
 [Icons]
 Name: "{group}\Manual"; Filename: "{app}\TwistedDocs\howto\index.xhtml"
