Index: win32/pyx.x-foo.iss.template
===================================================================
--- win32/pyx.x-foo.iss.template	(revision 23235)
+++ win32/pyx.x-foo.iss.template	(working copy)
@@ -25,6 +25,10 @@
 ; *.pyc keeps this directory around
 Type: filesandordirs; Name: "{app}\twisted"
 Type: filesandordirs; Name: "{app}\TwistedDocs"
+Type: files; Name: "{app}\zope\interface\*.pyc"
+Type: files; Name: "{app}\zope\interface\*.pyo"
+Type: files; Name: "{app}\zope\__init__.pyc"
+Type: files; Name: "{app}\zope\__init__.pyo"
 
 [Icons]
 Name: "{group}\Manual"; Filename: "{app}\TwistedDocs\howto\index.xhtml"
