Index: twisted/test/test_application.py
===================================================================
--- twisted/test/test_application.py	(revision 28774)
+++ twisted/test/test_application.py	(working copy)
@@ -569,6 +569,15 @@
                       'deprecated in Twisted 10.1.0: See IReactorFDSet.')]
 
 
+    def test_importAll(self):
+        """
+        L{twisted.application.internet} dynamically defines Service
+        subclasses. This test ensures that the subclasses exposed by __all__
+        are valid attributes of the module.
+        """
+        for cls in internet.__all__:
+            self.assert_(hasattr(internet, cls), 
+                '%s not importable from twisted.application.internet' % (cls,))
 
 
     def test_reactorParametrizationInServer(self):
Index: twisted/application/internet.py
===================================================================
--- twisted/application/internet.py	(revision 28774)
+++ twisted/application/internet.py	(working copy)
@@ -308,7 +308,7 @@
 
 
 
-__all__ = (['TimerService', 'CooperatorService'] +
+__all__ = (['TimerService', 'CooperatorService', 'MulticastServer'] +
            [tran+side
-         for tran in 'Generic TCP UNIX SSL UDP UNIXDatagram Multicast'.split()
+         for tran in 'Generic TCP UNIX SSL UDP UNIXDatagram'.split()
          for side in 'Server Client'.split()])
