Index: twisted/test/test_application.py
===================================================================
--- twisted/test/test_application.py	(revision 28765)
+++ twisted/test/test_application.py	(working copy)
@@ -571,6 +571,16 @@
 
 
 
+    def testImportAll(self):
+        """
+        L{twisted.application.internet} dynamically defines Service
+        subclasses. This test ensures that the subclasses exposed by __all__
+        are valid attributes of the module.
+        """
+        for cls in internet.__all__:
+            self.assert_(hasattr(internet, cls), 
+                '%s not importable from twisted.application.internet' % (cls,))
+        
     def test_reactorParametrizationInServer(self):
         """
         L{internet._AbstractServer} supports a C{reactor} keyword argument
