diff --git a/twisted/web2/topfiles/4338.misc b/twisted/web2/topfiles/4338.misc
new file mode 100644
index 0000000..e69de29
diff --git a/twisted/web2/test/test_stream.py b/twisted/web2/test/test_stream.py
index 1e2c10b..1918e46 100644
--- a/twisted/web2/test/test_stream.py
+++ b/twisted/web2/test/test_stream.py
@@ -434,8 +434,8 @@ class ProcessStreamerTest(unittest.TestCase):
         d = stream.readStream(p.errStream, l.append)
         def verify(_):
             self.assertEquals("".join(l), ("x" * 1000) * 100)
-        p.run()
-        return d.addCallback(verify)
+        d2 = p.run()
+        return d.addCallback(verify).addCallback(lambda _: d2)
 
     def test_input(self):
         p = self.runCode("import sys\nsys.stdout.write(sys.stdin.read())",
