diff --git a/twisted/web/client.py b/twisted/web/client.py
index ae6b904..4b9a56e 100644
--- a/twisted/web/client.py
+++ b/twisted/web/client.py
@@ -246,7 +246,7 @@ class HTTPPageGetter(http.HTTPClient):
 
     def timeout(self):
         self.quietLoss = True
-        self.transport.loseConnection()
+        self.transport.abortConnection()
         self.factory.noPage(defer.TimeoutError("Getting %s took longer than %s seconds." % (self.factory.url, self.factory.timeout)))
 
 
diff --git a/twisted/web/topfiles/8318.bugfix b/twisted/web/topfiles/8318.bugfix
new file mode 100644
index 0000000..6895867
--- /dev/null
+++ b/twisted/web/topfiles/8318.bugfix
@@ -0,0 +1 @@
+Resolved bug where timeouts as part of getPage, downloadPage, and HTTPClientFactory could fail to fire on TLS connections where the remote party is not responding on the connection.
\ No newline at end of file
