diff --git a/twisted/conch/client/knownhosts.py b/twisted/conch/client/knownhosts.py
index 4605f21..29b04a2 100644
--- a/twisted/conch/client/knownhosts.py
+++ b/twisted/conch/client/knownhosts.py
@@ -15,11 +15,7 @@ import sys
 from zope.interface import implements
 
 from twisted.python.randbytes import secureRandom
-if sys.version_info >= (2, 5):
-    from twisted.python.hashlib import sha1
-else:
-    # We need to have an object with a method named 'new'.
-    import sha as sha1
+from twisted.python.hashlib import sha1
 
 from twisted.internet import defer
 
