diff --git a/twisted/python/dist.py b/twisted/python/dist.py
index ee55533..e640a83 100644
--- a/twisted/python/dist.py
+++ b/twisted/python/dist.py
@@ -73,7 +73,8 @@ _EXTRA_OPTIONS = dict(
          'sphinx >= 1.2.2',
          'pydoctor >= 0.5'],
     tls=['pyopenssl >= 0.11',
-         'service_identity'],
+         'service_identity',
+         'idna'],
     conch=['gmpy',
            'pyasn1',
            'pycrypto'],
diff --git a/twisted/python/test/test_dist.py b/twisted/python/test/test_dist.py
index 190bb06..52b2612 100644
--- a/twisted/python/test/test_dist.py
+++ b/twisted/python/test/test_dist.py
@@ -118,6 +118,7 @@ class OptionalDependenciesTests(TestCase):
         deps = _EXTRAS_REQUIRE['tls']
         self.assertIn('pyopenssl >= 0.11', deps)
         self.assertIn('service_identity', deps)
+        self.assertIn('idna', deps)
 
 
     def test_extrasRequiresConchDeps(self):
@@ -163,6 +164,7 @@ class OptionalDependenciesTests(TestCase):
         deps = _EXTRAS_REQUIRE['all_non_platform']
         self.assertIn('pyopenssl >= 0.11', deps)
         self.assertIn('service_identity', deps)
+        self.assertIn('idna', deps)
         self.assertIn('gmpy', deps)
         self.assertIn('pyasn1', deps)
         self.assertIn('pycrypto', deps)
@@ -179,6 +181,7 @@ class OptionalDependenciesTests(TestCase):
         deps = _EXTRAS_REQUIRE['osx_platform']
         self.assertIn('pyopenssl >= 0.11', deps)
         self.assertIn('service_identity', deps)
+        self.assertIn('idna', deps)
         self.assertIn('gmpy', deps)
         self.assertIn('pyasn1', deps)
         self.assertIn('pycrypto', deps)
@@ -196,6 +199,7 @@ class OptionalDependenciesTests(TestCase):
         deps = _EXTRAS_REQUIRE['windows_platform']
         self.assertIn('pyopenssl >= 0.11', deps)
         self.assertIn('service_identity', deps)
+        self.assertIn('idna', deps)
         self.assertIn('gmpy', deps)
         self.assertIn('pyasn1', deps)
         self.assertIn('pycrypto', deps)
