Index: bin/tap2deb
===================================================================
--- bin/tap2deb	(revision 31541)
+++ bin/tap2deb	(working copy)
@@ -11,8 +11,8 @@
 ### Twisted Preamble
 # This makes sure that users don't have to set up their environment
 # specially in order to run these programs from bin/.
-import sys, os, string
-if string.find(os.path.abspath(sys.argv[0]), os.sep+'Twisted') != -1:
+import sys, os
+if os.path.abspath(sys.argv[0]).find(os.sep+'Twisted') != -1:
     sys.path.insert(0, os.path.normpath(os.path.join(os.path.abspath(sys.argv[0]), os.pardir, os.pardir)))
 ### end of preamble
 
Index: bin/manhole
===================================================================
--- bin/manhole	(revision 31541)
+++ bin/manhole	(working copy)
@@ -11,8 +11,8 @@
 ### Twisted Preamble
 # This makes sure that users don't have to set up their environment
 # specially in order to run these programs from bin/.
-import sys, os, string
-if string.find(os.path.abspath(sys.argv[0]), os.sep+'Twisted') != -1:
+import sys, os
+if os.path.abspath(sys.argv[0]).find(os.sep+'Twisted') != -1:
     sys.path.insert(0, os.path.normpath(os.path.join(os.path.abspath(sys.argv[0]), os.pardir, os.pardir)))
 ### end of preamble
 
Index: bin/tap2rpm
===================================================================
--- bin/tap2rpm	(revision 31541)
+++ bin/tap2rpm	(working copy)
@@ -13,8 +13,8 @@
 ### Twisted Preamble
 # This makes sure that users don't have to set up their environment
 # specially in order to run these programs from bin/.
-import sys, os, string
-if string.find(os.path.abspath(sys.argv[0]), os.sep+'Twisted') != -1:
+import sys, os
+if os.path.abspath(sys.argv[0]).find(os.sep+'Twisted') != -1:
     sys.path.insert(0, os.path.normpath(os.path.join(os.path.abspath(sys.argv[0]), os.pardir, os.pardir)))
 ### end of preamble
 
Index: bin/tapconvert
===================================================================
--- bin/tapconvert	(revision 31541)
+++ bin/tapconvert	(working copy)
@@ -7,8 +7,8 @@
 ### Twisted Preamble
 # This makes sure that users don't have to set up their environment
 # specially in order to run these programs from bin/.
-import sys, os, string
-if string.find(os.path.abspath(sys.argv[0]), os.sep+'Twisted') != -1:
+import sys, os
+if os.path.abspath(sys.argv[0]).find(os.sep+'Twisted') != -1:
     sys.path.insert(0, os.path.normpath(os.path.join(os.path.abspath(sys.argv[0]), os.pardir, os.pardir)))
 if not hasattr(os, "getuid") or os.getuid() != 0:
     sys.path.insert(0, os.getcwd())
Index: bin/pyhtmlizer
===================================================================
--- bin/pyhtmlizer	(revision 31541)
+++ bin/pyhtmlizer	(working copy)
@@ -7,8 +7,8 @@
 ### Twisted Preamble
 # This makes sure that users don't have to set up their environment
 # specially in order to run these programs from bin/.
-import sys, os, string
-if string.find(os.path.abspath(sys.argv[0]), os.sep+'Twisted') != -1:
+import sys, os
+if os.path.abspath(sys.argv[0]).find(os.sep+'Twisted') != -1:
     sys.path.insert(0, os.path.normpath(os.path.join(os.path.abspath(sys.argv[0]), os.pardir, os.pardir)))
 sys.path.insert(0, os.curdir)
 ### end of preamble
Index: bin/trial
===================================================================
--- bin/trial	(revision 31541)
+++ bin/trial	(working copy)
@@ -7,8 +7,8 @@
 ### Twisted Preamble
 # This makes sure that users don't have to set up their environment
 # specially in order to run these programs from bin/.
-import sys, os, string
-if string.find(os.path.abspath(sys.argv[0]), os.sep+'Twisted') != -1:
+import sys, os
+if os.path.abspath(sys.argv[0]).find(os.sep+'Twisted') != -1:
     sys.path.insert(0, os.path.normpath(os.path.join(os.path.abspath(sys.argv[0]), os.pardir, os.pardir)))
 if hasattr(os, "getuid") and os.getuid() != 0:
     sys.path.insert(0, os.curdir)
Index: bin/mktap
===================================================================
--- bin/mktap	(revision 31541)
+++ bin/mktap	(working copy)
@@ -6,8 +6,8 @@
 ### Twisted Preamble
 # This makes sure that users don't have to set up their environment
 # specially in order to run these programs from bin/.
-import sys, os, string
-if string.find(os.path.abspath(sys.argv[0]), os.sep+'Twisted') != -1:
+import sys, os
+if os.path.abspath(sys.argv[0]).find(os.sep+'Twisted') != -1:
     sys.path.insert(0, os.path.normpath(os.path.join(os.path.abspath(sys.argv[0]), os.pardir, os.pardir)))
 if not hasattr(os, "getuid") or os.getuid() != 0:
     sys.path.insert(0, os.getcwd())
Index: bin/twistd
===================================================================
--- bin/twistd	(revision 31541)
+++ bin/twistd	(working copy)
@@ -7,8 +7,8 @@
 ### Twisted Preamble
 # This makes sure that users don't have to set up their environment
 # specially in order to run these programs from bin/.
-import sys, os, string
-if string.find(os.path.abspath(sys.argv[0]), os.sep+'Twisted') != -1:
+import sys, os
+if os.path.abspath(sys.argv[0]).find(os.sep+'Twisted') != -1:
     sys.path.insert(0, os.path.normpath(os.path.join(os.path.abspath(sys.argv[0]), os.pardir, os.pardir)))
 if hasattr(os, "getuid") and os.getuid() != 0:
     sys.path.insert(0, os.path.abspath(os.getcwd()))
Index: doc/core/development/policy/coding-standard.xhtml
===================================================================
--- doc/core/development/policy/coding-standard.xhtml	(revision 31541)
+++ doc/core/development/policy/coding-standard.xhtml	(working copy)
@@ -364,8 +364,8 @@
 ### Twisted Preamble
 # This makes sure that users don't have to set up their environment
 # specially in order to run these programs from bin/.
-import sys, os, string
-if string.find(os.path.abspath(sys.argv[0]), os.sep+'Twisted') != -1:
+import sys, os
+if os.path.abspath(sys.argv[0]).find(os.sep+'Twisted') != -1:
     sys.path.insert(0, os.path.normpath(os.path.join(os.path.abspath(sys.argv[0]), os.pardir, os.pardir)))
 if not hasattr(os, "getuid") or os.getuid() != 0:
     sys.path.insert(0, os.getcwd())
