Index: twisted/web/topfiles/4855.misc
===================================================================
Index: twisted/web/xmlrpc.py
===================================================================
--- twisted/web/xmlrpc.py	(revision 30862)
+++ twisted/web/xmlrpc.py	(working copy)
@@ -34,7 +34,7 @@
     DateTime.decode = lambda self, value: _decode(self, value.encode('ascii'))
 
 
-def withRequest(f, *args, **kwargs):
+def withRequest(f):
     """
     Decorator to cause the request to be passed as the first argument
     to the method.
@@ -46,6 +46,8 @@
         @withRequest
         def xmlrpc_echo(self, request, s):
             return s
+
+    @since: 10.2
     """
     f.withRequest = True
     return f
