Index: twisted/python/components.py
===================================================================
--- twisted/python/components.py	(revision 23176)
+++ twisted/python/components.py	(working copy)
@@ -74,7 +74,7 @@
     An adapter class is expected to implement the given interface, by
     adapting instances implementing 'origInterface'. An adapter class's
     __init__ method should accept one parameter, an instance implementing
-    'origInterface'.
+    'origInterface'.When adapterFactory is None adapter will be unregistered.
     """
     self = globalRegistry
     assert interfaceClasses, "You need to pass an Interface"
@@ -86,7 +86,7 @@
 
     for interfaceClass in interfaceClasses:
         factory = _registered(self, origInterface, interfaceClass)
-        if factory is not None and not ALLOW_DUPLICATES:
+        if factory is not None and not ALLOW_DUPLICATES and adapterFactory is not None:
             raise ValueError("an adapter (%s) was already registered." % (factory, ))
     for interfaceClass in interfaceClasses:
         self.register([origInterface], interfaceClass, '', adapterFactory)
