Index: twisted/conch/test/test_channel.py
===================================================================
--- twisted/conch/test/test_channel.py	(revision 38850)
+++ twisted/conch/test/test_channel.py	(working copy)
@@ -173,6 +173,20 @@
         self.channel.closeReceived()
         self.assertTrue(self.channel.closing)
 
+    def test_closeReceievedAfterLoseConnection(self):
+        """
+        Test that a closeReceieved after explicitly calling loseConnection does
+        not try to close it again.
+        """
+        self.assertFalse(self.channel.closing)
+        self.channel.loseConnection()
+        self.assertTrue(self.channel.closing)
+        self.assertIn(self.channel, self.conn.closes)
+
+        self.conn.closes.clear()
+        self.channel.closeReceived()
+        self.assertNotIn(self.channel, self.conn.closes)
+
     def test_write(self):
         """
         Test that write handles data correctly.  Send data up to the size
Index: twisted/conch/ssh/channel.py
===================================================================
--- twisted/conch/ssh/channel.py	(revision 38850)
+++ twisted/conch/ssh/channel.py	(working copy)
@@ -166,7 +166,8 @@
         Called when the other side has closed the channel.
         """
         log.msg('remote close')
-        self.loseConnection()
+        if not self.closing:
+            self.loseConnection()
 
     def closed(self):
         """
