Index: test_imap.py
===================================================================
--- test_imap.py	(révision 15767)
+++ test_imap.py	(copie de travail)
@@ -319,6 +319,7 @@
             '''foo NIL "baz buz" bar''',
             '''foo "NIL" "baz buz" bar''',
             '''"NIL" bar "baz buz" foo''',
+            '''"NIL" bar "baz \\"buz" foo''',
         ]
 
         answers = [
@@ -331,6 +332,7 @@
             ['foo', None, 'baz buz', 'bar'],
             ['foo', 'NIL', 'baz buz', 'bar'],
             ['NIL', 'bar', 'baz buz', 'foo'],
+            ['NIL', 'bar', 'baz \\"buz', 'foo'],
         ]
 
         errors = [
