Index: imap4.py
===================================================================
--- imap4.py	(révision 15767)
+++ imap4.py	(copie de travail)
@@ -3773,9 +3773,10 @@
             inQuote = 1
             start = i + 1
         elif c == '"' and inQuote:
-            inQuote = 0
-            result.append(s[start:i])
-            start = i + 1
+            if s[i-1] != '\\':
+                inQuote = 0
+                result.append(s[start:i])
+                start = i + 1
         elif not inWord and not inQuote and c not in ('"' + string.whitespace):
             inWord = 1
             start = i
