Index: twisted/topfiles/3108.removal
===================================================================
--- twisted/topfiles/3108.removal	(revision 0)
+++ twisted/topfiles/3108.removal	(revision 0)
@@ -0,0 +1 @@
+twisted.trial.util.findObject has been deprecated.
\ No newline at end of file
Index: twisted/trial/test/test_util.py
===================================================================
--- twisted/trial/test/test_util.py	(revision 29269)
+++ twisted/trial/test/test_util.py	(working copy)
@@ -1,3 +1,11 @@
+# Copyright (c) 2004-2010 Twisted Matrix Laboratories.
+# See LICENSE for details.
+#
+
+"""
+A collection of tests for L{twisted.trial.util}
+"""
+
 import os
 
 from zope.interface import implements
@@ -46,6 +54,24 @@
 
 
 class TestFindObject(packages.SysPathManglingTest):
+    """
+    Tests for L{twisted.trial.util.findObject}
+    """
+
+    def test_deprecation(self):
+        """
+        Test that calling L{findObject} results in a deprecation warning
+        """
+        util.findObject('')
+        warningsShown = self.flushWarnings()
+        self.assertEquals(len(warningsShown), 1)
+        self.assertIdentical(warningsShown[0]['category'], DeprecationWarning)
+        self.assertEquals(warningsShown[0]['message'], 
+                          "twisted.trial.util.findObject was deprecated "
+                          "in Twisted 10.1.0: Please use "
+                          "twisted.python.reflect.namedAny instead.")
+
+
     def test_importPackage(self):
         package1 = util.findObject('package')
         import package as package2
Index: twisted/trial/util.py
===================================================================
--- twisted/trial/util.py	(revision 29269)
+++ twisted/trial/util.py	(working copy)
@@ -1,5 +1,5 @@
 # -*- test-case-name: twisted.trial.test.test_util -*-
-# Copyright (c) 2001-2009 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2010 Twisted Matrix Laboratories.
 # See LICENSE for details.
 #
 
@@ -19,6 +19,7 @@
 
 from twisted.internet import defer, utils, interfaces
 from twisted.python.failure import Failure
+from twisted.python import deprecate, versions
 
 
 DEFAULT_TIMEOUT = object()
@@ -344,6 +345,13 @@
 
 
 
+deprecate.deprecatedModuleAttribute(
+    versions.Version("Twisted", 10, 1, 0),
+    "Please use twisted.python.reflect.namedAny instead.",
+    __name__, "findObject")
+
+
+
 def _runSequentially(callables, stopOnFirstError=False):
     """
     Run the given callables one after the other. If a callable returns a
