Index: doc/web/howto/xmlrpc.xhtml
===================================================================
--- doc/web/howto/xmlrpc.xhtml	(revision 27387)
+++ doc/web/howto/xmlrpc.xhtml	(working copy)
@@ -35,7 +35,7 @@
 class="API">Deferred</code> instances whose results are one of the
 above. This allows you to return results that can't be calculated
 immediately, such as database queries. See the <a
-href="defer.xhtml">Deferred documentation</a> for more details.</p>
+href="../../core/howto/defer.xhtml">Deferred documentation</a> for more details.</p>
 
 <p><code base="twisted.web.xmlrpc" class="API">XMLRPC</code> instances
 are Resource objects, and they can thus be published using a Site. The
