Index: __init__.py
===================================================================
--- __init__.py	(revision 24199)
+++ __init__.py	(working copy)
@@ -2,9 +2,6 @@
 # Copyright (c) 2001-2004 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
-#
-
-
 """
 Twisted.Conch: The Twisted Shell. Terminal emulation, SSHv2 and telnet.
 
Index: ssh/agent.py
===================================================================
--- ssh/agent.py	(revision 24199)
+++ ssh/agent.py	(working copy)
@@ -1,7 +1,6 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
-#
 """
 Implements the old SSHv1 key agent protocol.
 
@@ -9,7 +8,7 @@
 """
 
 import struct
-from common import NS, getNS
+from twisted.conch.ssh.common import NS, getNS
 from twisted.conch.error import ConchError
 from twisted.internet import defer, protocol
 
Index: ssh/userauth.py
===================================================================
--- ssh/userauth.py	(revision 24199)
+++ ssh/userauth.py	(working copy)
@@ -1,4 +1,4 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 # 
@@ -15,8 +15,8 @@
 from twisted.cred import credentials
 from twisted.internet import defer, reactor
 from twisted.python import failure, log
-from common import NS, getNS, MP
-import keys, transport, service
+from twisted.conch.ssh.common import NS, getNS, MP
+from twisted.conch.ssh import keys, transport, service
 
 class SSHUserAuthServer(service.SSHService):
     name = 'ssh-userauth'
Index: ssh/forwarding.py
===================================================================
--- ssh/forwarding.py	(revision 24199)
+++ ssh/forwarding.py	(working copy)
@@ -1,8 +1,6 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
-# 
-
 """
 This module contains the implementation of the TCP forwarding, which allows
 clients and servers to forward arbitrary TCP data across the connection.
@@ -14,9 +12,8 @@
 
 from twisted.internet import protocol, reactor
 from twisted.python import log
+from twisted.conch.ssh import common, channel
 
-import common, channel
-
 class SSHListenForwardingFactory(protocol.Factory):
     def __init__(self, connection, hostport, klass):
         self.conn = connection
Index: ssh/factory.py
===================================================================
--- ssh/factory.py	(revision 24199)
+++ ssh/factory.py	(working copy)
@@ -18,8 +18,7 @@
 from twisted.python.reflect import qual
 
 from twisted.conch import error
-from twisted.conch.ssh import keys
-import transport, userauth, connection
+from twisted.conch.ssh import keys, transport, userauth, connection
 
 import random
 import warnings
Index: ssh/filetransfer.py
===================================================================
--- ssh/filetransfer.py	(revision 24199)
+++ ssh/filetransfer.py	(working copy)
@@ -9,13 +9,11 @@
 from twisted.internet import defer, protocol
 from twisted.python import failure, log
 
-from common import NS, getNS
+from twisted.conch.ssh.common import NS, getNS
 from twisted.conch.interfaces import ISFTPServer, ISFTPFile
 
 from zope import interface
 
-
-
 class FileTransferBase(protocol.Protocol):
 
     versions = (3, )
Index: avatar.py
===================================================================
--- avatar.py	(revision 24199)
+++ avatar.py	(working copy)
@@ -1,7 +1,11 @@
 # -*- test-case-name: twisted.conch.test.test_conch -*-
-from interfaces import IConchUser
-from error import ConchError
-from ssh.connection import OPEN_UNKNOWN_CHANNEL_TYPE
+#
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
+# See LICENSE for details.
+
+from twisted.conch.interfaces import IConchUser
+from twisted.conch.error import ConchError
+from twisted.ssh.connection import OPEN_UNKNOWN_CHANNEL_TYPE
 from twisted.python import log
 from zope import interface
 
Index: unix.py
===================================================================
--- unix.py	(revision 24199)
+++ unix.py	(working copy)
@@ -1,23 +1,23 @@
-# Copyright (c) 2001-2007 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 from twisted.cred import portal
 from twisted.python import components, log
 from twisted.internet.error import ProcessExitedAlready
 from zope import interface
-from ssh import session, forwarding, filetransfer
-from ssh.filetransfer import FXF_READ, FXF_WRITE, FXF_APPEND, FXF_CREAT, FXF_TRUNC, FXF_EXCL
+
+from twisted.conch.ssh import session, forwarding, filetransfer
+from twisted.conch.ssh.filetransfer import FXF_READ, FXF_WRITE, FXF_APPEND, FXF_CREAT, FXF_TRUNC, FXF_EXCL
 from twisted.conch.ls import lsLine
+from twisted.conch.avatar import ConchUser
+from twisted.conch.error import ConchError
+from twisted.conch.interfaces import ISession, ISFTPServer, ISFTPFile
 
-from avatar import ConchUser
-from error import ConchError
-from interfaces import ISession, ISFTPServer, ISFTPFile
-
 import struct, os, time, socket
 import fcntl, tty
 import pwd, grp
 import pty
-import ttymodes
+from twisted.conch import ttymodes
 
 try:
     import utmp
Index: ui/tkvt100.py
===================================================================
--- ui/tkvt100.py	(revision 24199)
+++ ui/tkvt100.py	(working copy)
@@ -1,17 +1,19 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 #
 
-"""Module to emulate a VT100 terminal in Tkinter.
+"""
+Module to emulate a VT100 terminal in Tkinter.
 
 Maintainer: U{Paul Swartz <mailto:z3p@twistedmatrix.com>}
 """
 
 import Tkinter, tkFont
-import ansi
 import string
 
+from twisted.conch.ui import ansi
+
 ttyFont = None#tkFont.Font(family = 'Courier', size = 10)
 fontWidth, fontHeight = None,None#max(map(ttyFont.measure, string.letters+string.digits)), int(ttyFont.metrics()['linespace'])
 
Index: client/connect.py
===================================================================
--- client/connect.py	(revision 24199)
+++ client/connect.py	(working copy)
@@ -1,8 +1,7 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
-#
-import direct, unix
+from twisted.conch.client import direct, unix
 
 connectTypes = {"direct" : direct.connect,
                 "unix" : unix.connect}
Index: client/direct.py
===================================================================
--- client/direct.py	(revision 24199)
+++ client/direct.py	(working copy)
@@ -11,8 +11,6 @@
 
 from twisted.conch.client import unix
 
-
-
 class SSHClientFactory(protocol.ClientFactory):
 
     def __init__(self, d, options, verifyHostKey, userAuthObject):
Index: client/options.py
===================================================================
--- client/options.py	(revision 24199)
+++ client/options.py	(working copy)
@@ -3,10 +3,9 @@
 
 #
 from twisted.conch.ssh.transport import SSHClientTransport, SSHCiphers
+from twisted.conch.client import connect
 from twisted.python import usage
 
-import connect
-
 import sys
 
 class ConchOptions(usage.Options):
Index: client/default.py
===================================================================
--- client/default.py	(revision 24199)
+++ client/default.py	(working copy)
@@ -1,14 +1,13 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 #
 from twisted.conch.error import ConchError
 from twisted.conch.ssh import common, keys, userauth, agent
+from twisted.conch.client import agent
 from twisted.internet import defer, protocol, reactor
 from twisted.python import log
 
-import agent
-
 import os, sys, base64, getpass
 
 def verifyHostKey(transport, host, pubKey, fingerprint):
