Index: vhost.py
===================================================================
RCS file: /cvs/Twisted/twisted/web/vhost.py,v
retrieving revision 1.22
diff -u -r1.22 vhost.py
--- vhost.py	29 Apr 2003 20:27:16 -0000	1.22
+++ vhost.py	14 Sep 2003 01:31:30 -0000
@@ -110,7 +110,11 @@
         """Implementation of resource.Resource's getChild method.
         """
         resrc = self._getResourceForRequest(request)
-        return resrc.getChildWithDefault(path, request)
+        if resrc.isLeaf:
+            request.postpath.insert(0,request.prepath.pop(-1))
+            return resrc
+        else:
+            return resrc.getChildWithDefault(path, request)
 
 class _HostResource(resource.Resource):
 
