Index: pb-cred.xhtml
===================================================================
--- pb-cred.xhtml	(revision 38186)
+++ pb-cred.xhtml	(working copy)
@@ -253,7 +253,7 @@
 will give them anything else. The dict goes away when the connection is
 dropped, further limiting the scope of those references.</p>
 
-<p>Futhermore, it is not possible for Bob to send <em>his</em> 
+<p>Furthermore, it is not possible for Bob to send <em>his</em> 
  <code>User</code> reference to Alice (perhaps over some other PB channel
 just between the two of them). Outside the context of Bob's connection to
 the server, that reference is just a meaningless number. To prevent
