=== modified file 'doc/core/howto/endpoints.xhtml'
--- doc/core/howto/endpoints.xhtml	2012-08-20 15:25:12 +0000
+++ doc/core/howto/endpoints.xhtml	2013-01-22 22:58:48 +0000
@@ -241,5 +241,157 @@
   </li>
 </ul>
 
+<h2>Using <code>systemd</code> Server Endpoints</h2>
+<p>The following example demonstrates how to launch a Twisted web
+  server using <code>systemd</code> socket activation.</p>
+<ol>
+  <li>
+    <p>Create
+    a <a href="http://0pointer.de/public/systemd-man/systemd.socket.html"><code>systemd.socket</code></a>
+    file.</p>
+
+    <a href="listings/systemd/twistedweb.socket"
+    class="listing">/etc/systemd/system/twistedweb.socket</a>
+
+    <p>This file configures <code>systemd</code> to listen on an IPv4 address on
+    port 8080.</p>
+  </li>
+
+  <li>
+    <p>Create
+    a <a href="http://0pointer.de/public/systemd-man/systemd.service.html"><code>systemd.service</code></a>
+    file.</p>
+
+    <a href="listings/systemd/twistedweb.service"
+    class="listing">/etc/systemd/system/twistedweb.service</a>
+
+    <p>This configuration file contains the following important
+    sections:</p>
+
+    <dl>
+      <dt>ExecStart</dt>
+      <dd>Include the full path to <code>twistd</code> (in this case the path to a
+        symlink maintained
+        by <a href="http://twistedmatrix.com/trac/wiki/Combinator">Combinator</a>)</dd>
+      <dd>The <code>--nodaemon</code> flag makes <code>twistd</code> run in the
+        foreground. <code>systemd</code> expects child processes to remain in the
+        foreground.</dd>
+      <dd>The <code>domain=INET</code> endpoint argument makes <code>twistd</code>
+        treat the inherited file descriptor as an IPv4 socket.</dd>
+      <dd>The <code>index=0</code> endpoint argument
+        makes <code>twistd</code> adopt the first file descriptor
+        inherited from <code>systemd</code>.</dd>
+
+      <dt>Environment</dt>
+      <dd>In this example, Twisted has been installed
+        with <a href="http://twistedmatrix.com/trac/wiki/Combinator">Combinator</a>
+        so the Combinator path is added
+        to <code>PYTHONPATH</code>.</dd>
+
+      <dt>NonBlocking</dt>
+      <dd>This must be set to <code>true</code> to ensure that the
+      sockets have the O_NONBLOCK flag set and hence are in
+      non-blocking mode before they are passed to Twisted.</dd>
+    </dl>
+
+    <p>There are many more service options which are documented in the
+    <a href="http://0pointer.de/public/systemd-man/systemd.service.html"><code>systemd.service</code>
+    man page</a>.</p>
+  </li>
+  <li>
+    <p>Reload <code>systemd</code>.</p>
+    <pre class="shell">
+      $ sudo systemctl --system daemon-reload
+    </pre>
+    <p>This forces <code>systemd</code> to read the new configuration files.</p>
+  </li>
+  <li>
+    <p>Start <code>systemd</code> listening on the new port.</p>
+    <pre class="shell">
+      $ sudo systemctl start twistedweb.socket
+    </pre>
+
+    <p><code>systemd</code> should now be listening on port 8080. You can verify
+     this using the <code>netstat</code> command. eg</p>
+    <pre class="shell">
+      $ sudo netstat -lntp | grep 8080
+      ...
+      tcp 0 0 127.0.0.1:8080 0.0.0.0:* LISTEN 1/init
+    </pre>
+
+    <p><strong>Note</strong>: <code>twistd</code> will not be started
+    until the first client connects. You can verify this by searching
+    the current process list. eg</p>
+    <pre class="shell">
+      $ ps -ef | grep twistd
+      ...
+      richard   5815  4237  0 01:20 pts/5    00:00:00 grep --color=auto twistd
+    </pre>
+  </li>
+  <li>
+    <p>Connect to port 8080 using <a href="http://curl.haxx.se/">curl</a>.</p>
+    <pre class="shell">
+      $ curl http://localhost:8080
+    </pre>
+
+    <p><code>systemd</code> will accept the connection and
+    launch <code>twistd</code>, passing it the listening socket. You
+    can verify this by searching the process list again. eg</p>
+    <pre class="shell">
+      $ ps -ef | grep twistd
+      ...
+      richard   5831     1  1 01:29 ?        00:00:00 python /home/richard/projects/Twisted/trunk/bin/twistd --nodaemon web --port systemd:domain=INET:index=0 --path .
+      richard   5836  4237  0 01:29 pts/5    00:00:00 grep --color=auto twistd
+    </pre>
+
+    <p>The <code>twistd</code> subprocess will log messages
+    to <code>stderr</code> and <code>systemd</code> will in turn log these messages
+    to syslog. You can verify this by monitoring the syslog messages
+    using the <code>tail</code> command. eg</p>
+    <pre class="shell">
+      $ sudo tail -f /var/log/messages
+      ...
+      Log opened.
+      twistd 12.0.0+r34099 (/usr/bin/python 2.7.2) starting up.
+      reactor class: twisted.internet.pollreactor.PollReactor.
+      Site starting on 8080
+      "GET / HTTP/1.1" 200 - "-" "curl/7.21.7 (x86_64-redhat-linux-gnu) libcurl/7.21.7 NSS/3.13.3.0 zlib/1.2.5 libidn/1.22 libssh2/1.2.7"
+    </pre>
+  </li>
+</ol>
+
+<h3>Process Monitoring and Supervision With <code>systemd</code></h3>
+<p>In addition to launching Twisted on-demand, <code>systemd</code> also provides
+  process monitoring and supervision.</p>
+<p>Here is a short demonstration.</p>
+<ol>
+  <li>
+    <p>Kill the <code>twistd</code> process.</p>
+
+    <pre class="shell">
+      $ kill -9 5831
+    </pre>
+
+    <p><code>systemd</code> will log a warning when
+    the <code>twistd</code> process ends unexpectedly, but it will
+    continue listening on port 8080.</p>
+    <pre class="shell">
+      $ sudo tail -f /var/log/messages
+      ...
+      twistedweb.service: main process exited, code=killed, status=9
+      Unit twistedweb.service entered failed state.
+    </pre>
+
+    <p>It will immediately re-launch <code>twistd</code> when another client
+    connects to that port.</p>
+  </li>
+</ol>
+
+
+<h3>Further Reading</h3>
+<ul>
+  <li><a href="http://0pointer.de/blog/projects/systemd-docs.html"><code>systemd</code>
+  Documentation</a></li>
+</ul>
 </body>
 </html>

=== added directory 'doc/core/howto/listings/systemd'
=== added file 'doc/core/howto/listings/systemd/twistedweb.service'
--- doc/core/howto/listings/systemd/twistedweb.service	1970-01-01 00:00:00 +0000
+++ doc/core/howto/listings/systemd/twistedweb.service	2013-01-22 23:02:18 +0000
@@ -0,0 +1,12 @@
+[Service]
+ExecStart=/home/richard/projects/combinator_paths/bincache/twistd --nodaemon \
+    web --port systemd:domain=INET:index=0 --path /home/richard/www
+
+Environment=\
+    PYTHONPATH=/home/richard/projects/Divmod/trunk/Combinator
+
+WorkingDirectory=/home/richard/www
+
+User=richard
+
+NonBlocking=true

=== added file 'doc/core/howto/listings/systemd/twistedweb.socket'
--- doc/core/howto/listings/systemd/twistedweb.socket	1970-01-01 00:00:00 +0000
+++ doc/core/howto/listings/systemd/twistedweb.socket	2013-01-22 18:55:32 +0000
@@ -0,0 +1,5 @@
+[Socket]
+ListenStream=0.0.0.0:8080
+
+[Install]
+WantedBy=sockets.target

=== added file 'twisted/topfiles/5601.misc'
