Index: root.py
===================================================================
--- root.py	(revision 26541)
+++ root.py	(working copy)
@@ -194,15 +194,3 @@
     d = defer.DeferredList(L)
     d.addCallback(lambda r: Resolver([e[1] for e in r if e[0]]))
     return DeferredResolver(d)
-
-if __name__ == '__main__':
-    if len(sys.argv) < 2:
-        print 'Specify a domain'
-    else:
-        log.startLogging(sys.stdout)
-        from twisted.names.client import ThreadedResolver
-        r = bootstrap(ThreadedResolver())
-        d = r.lookupAddress(sys.argv[1])
-        d.addCallbacks(log.msg, log.err).addBoth(lambda _: reactor.stop())
-        from twisted.internet import reactor
-        reactor.run()
