--- mail/relaymanager-orig.py	2010-08-03 19:29:54.860609389 +0200
+++ mail/relaymanager.py	2010-08-03 19:32:34.850610710 +0200
@@ -39,6 +39,8 @@
 from twisted.mail import smtp
 from twisted.application import internet
 
+from zope.interface import Interface, implements
+
 class ManagedRelayerMixin:
     """SMTP Relayer which notifies a manager
 
@@ -131,9 +133,54 @@
         protocol.factory = self
         return protocol
 
+class IQueue(Interface):
+    """A queue of ougoing emails."""
+
+    def createNewMessage():
+        """Create a new message in the queue.
+
+        Return a tuple - file-like object for headers, and ISMTPMessage.
+        """
+
+    def hasWaiting():
+        """Returns a boolean"""
+
+    def getWaiting():
+        """Get a list of message IDs waiting to be relayed"""
+
+    def setWaiting(message):
+        """Change the status of the message to waiting"""
+
+    def setRelaying(message):
+        """Change the status of the message to relayed"""
+
+    def getRelayed():
+        """Get a list of message IDs that are relayed"""
+
+    def done(message):
+        """Remove message to from queue."""
+
+    def getEnvelope(message):
+        """Get a message object for the givven ID"""
+
+    #XXX this thing is sooo file-oriented
+    def readDirectory():
+        """Read the messages directory.
+
+        look for new messages.
+        """
+
+    def getEnvelopeFile(message):
+        """Get the message file object"""
+
+    def getPath(msg):
+        """Get the path in the filesystem of a message."""
+
 class Queue:
     """A queue of ougoing emails."""
 
+    implements(IQueue)
+
     noisy = True
 
     def __init__(self, directory):
