Index: twisted/python/log.py
===================================================================
--- twisted/python/log.py	(revision 36616)
+++ twisted/python/log.py	(working copy)
@@ -381,7 +381,7 @@
         @rtype: C{str}
         """
         if self.timeFormat is not None:
-            return time.strftime(self.timeFormat, time.localtime(when))
+            return datetime.fromtimestamp(when).strftime(self.timeFormat)
 
         tzOffset = -self.getTimezoneOffset(when)
         when = datetime.utcfromtimestamp(when + tzOffset)
 
