--- /usr/lib/python2.4/site-packages/twisted/cred/credentials.py	2006-06-20 14:47:13.000000000 +0200
+++ credentials.py	2006-06-22 16:17:12.000000000 +0200
@@ -4,14 +4,13 @@
 # See LICENSE for details.
 
 
-from twisted.python import components
 from zope import interface
 
 import hmac
 import time
 import random
 
-class ICredentials(components.Interface):
+class ICredentials(interface.Interface):
     """I check credentials.
 
     Implementors _must_ specify which sub-interfaces of ICredentials
