Index: twisted/topfiles/6407.misc
===================================================================
Index: doc/core/howto/process.xhtml
===================================================================
--- doc/core/howto/process.xhtml	(revision 38130)
+++ doc/core/howto/process.xhtml	(working copy)
@@ -130,7 +130,7 @@
         self.transport.closeStdin()
 </pre>
 
-<p>At this point, the process has receieved the data, and it's time for us
+<p>At this point, the process has received the data, and it's time for us
 to read the results. Instead of being received in <code>dataReceived</code>,
 data from standard output is received in <code>outReceived</code>. This is
 to distinguish it from data on standard error.</p>
