Index: twisted/words/test/test_irc.py
===================================================================
--- twisted/words/test/test_irc.py	(revision 37845)
+++ twisted/words/test/test_irc.py	(working copy)
@@ -1686,7 +1686,7 @@
 
     def test_away(self):
         """
-        L{IRCCLient.away} sends an AWAY command with the specified message.
+        L{IRCClient.away} sends an AWAY command with the specified message.
         """
         message = "Sorry, I'm not here."
         self.protocol.away(message)
