Index: twisted/internet/selectreactor.py
===================================================================
--- twisted/internet/selectreactor.py	(revision 31396)
+++ twisted/internet/selectreactor.py	(working copy)
@@ -124,6 +124,11 @@
                     return
                 elif se.args[0] == EBADF:
                     self._preenDescriptors()
+                elif se.args[0] == 10038:
+                    # WSAENOTSOCK (10038)
+                    # Windows-only; socket operation on non-socket FD.
+                    # Probably means that we need to preen FDs.
+                    self._preenDescriptors()
                 else:
                     # OK, I really don't know what's going on.  Blow up.
                     raise
