Index: trunk/doc/core/howto/tutorial/listings/finger/finger20.tac
===================================================================
--- trunk/doc/core/howto/tutorial/listings/finger/finger20.tac	(revision 30844)
+++ trunk/doc/core/howto/tutorial/listings/finger/finger20.tac	(working copy)
@@ -274,6 +274,7 @@
 application = service.Application('finger', uid=1, gid=1)
 f = FingerService('/etc/users')
 serviceCollection = service.IServiceCollection(application)
+f.setServiceParent(serviceCollection)
 internet.TCPServer(79, IFingerFactory(f)
                    ).setServiceParent(serviceCollection)
 internet.TCPServer(8000, server.Site(resource.IResource(f))
Index: trunk/doc/core/howto/tutorial/listings/finger/finger21.tac
===================================================================
--- trunk/doc/core/howto/tutorial/listings/finger/finger21.tac	(revision 30844)
+++ trunk/doc/core/howto/tutorial/listings/finger/finger21.tac	(working copy)
@@ -306,6 +306,7 @@
 application = service.Application('finger', uid=1, gid=1)
 f = FingerService('/etc/users')
 serviceCollection = service.IServiceCollection(application)
+f.setServiceParent(serviceCollection)
 internet.TCPServer(79, IFingerFactory(f)
                    ).setServiceParent(serviceCollection)
 internet.TCPServer(8000, server.Site(resource.IResource(f))
Index: trunk/doc/core/howto/tutorial/listings/finger/finger15.tac
===================================================================
--- trunk/doc/core/howto/tutorial/listings/finger/finger15.tac	(revision 30844)
+++ trunk/doc/core/howto/tutorial/listings/finger/finger15.tac	(working copy)
@@ -80,6 +80,7 @@
 application = service.Application('finger', uid=1, gid=1)
 f = FingerService('/etc/users')
 serviceCollection = service.IServiceCollection(application)
+f.setServiceParent(serviceCollection)
 internet.TCPServer(79, f.getFingerFactory()
                    ).setServiceParent(serviceCollection)
 internet.TCPServer(8000, server.Site(f.getResource())
Index: trunk/doc/core/howto/tutorial/listings/finger/finger16.tac
===================================================================
--- trunk/doc/core/howto/tutorial/listings/finger/finger16.tac	(revision 30844)
+++ trunk/doc/core/howto/tutorial/listings/finger/finger16.tac	(working copy)
@@ -92,6 +92,7 @@
 application = service.Application('finger', uid=1, gid=1)
 f = FingerService('/etc/users')
 serviceCollection = service.IServiceCollection(application)
+f.setServiceParent(serviceCollection)
 internet.TCPServer(79, f.getFingerFactory()
                    ).setServiceParent(serviceCollection)
 internet.TCPServer(8000, server.Site(f.getResource())
Index: trunk/doc/core/howto/tutorial/listings/finger/finger17.tac
===================================================================
--- trunk/doc/core/howto/tutorial/listings/finger/finger17.tac	(revision 30844)
+++ trunk/doc/core/howto/tutorial/listings/finger/finger17.tac	(working copy)
@@ -93,6 +93,7 @@
 application = service.Application('finger', uid=1, gid=1)
 f = FingerService('/etc/users')
 serviceCollection = service.IServiceCollection(application)
+f.setServiceParent(serviceCollection)
 internet.TCPServer(79, f.getFingerFactory()
                    ).setServiceParent(serviceCollection)
 internet.TCPServer(8000, server.Site(f.getResource())
Index: trunk/doc/core/howto/tutorial/listings/finger/finger18.tac
===================================================================
--- trunk/doc/core/howto/tutorial/listings/finger/finger18.tac	(revision 30844)
+++ trunk/doc/core/howto/tutorial/listings/finger/finger18.tac	(working copy)
@@ -138,6 +138,7 @@
 application = service.Application('finger', uid=1, gid=1)
 f = FingerService('/etc/users')
 serviceCollection = service.IServiceCollection(application)
+f.setServiceParent(serviceCollection)
 internet.TCPServer(79, f.getFingerFactory()
                    ).setServiceParent(serviceCollection)
 internet.TCPServer(8000, server.Site(f.getResource())
Index: trunk/doc/core/howto/tutorial/listings/finger/finger19.tac
===================================================================
--- trunk/doc/core/howto/tutorial/listings/finger/finger19.tac	(revision 30844)
+++ trunk/doc/core/howto/tutorial/listings/finger/finger19.tac	(working copy)
@@ -259,6 +259,7 @@
 application = service.Application('finger', uid=1, gid=1)
 f = FingerService('/etc/users')
 serviceCollection = service.IServiceCollection(application)
+f.setServiceParent(serviceCollection)
 internet.TCPServer(79, IFingerFactory(f)
                    ).setServiceParent(serviceCollection)
 internet.TCPServer(8000, server.Site(resource.IResource(f))
Index: trunk/doc/core/howto/tutorial/listings/finger/finger22.py
===================================================================
--- trunk/doc/core/howto/tutorial/listings/finger/finger22.py	(revision 30844)
+++ trunk/doc/core/howto/tutorial/listings/finger/finger22.py	(working copy)
@@ -321,6 +321,7 @@
 application = service.Application('finger', uid=1, gid=1)
 f = FingerService('/etc/users')
 serviceCollection = service.IServiceCollection(application)
+f.setServiceParent(serviceCollection)
 internet.TCPServer(79, IFingerFactory(f)
                    ).setServiceParent(serviceCollection)
 site = server.Site(resource.IResource(f))
