--- twisted/web/http.py	2005-04-12 17:08:52.000000000 +0200
+++ /usr/lib/python2.3/site-packages/twisted/web/http.py	2005-04-12 17:26:18.246796008 +0200
@@ -796,7 +796,7 @@
 
         modified_since = self.getHeader('if-modified-since')
         if modified_since:
-            modified_since = stringToDatetime(modified_since)
+            modified_since = stringToDatetime(modified_since.split(';', 1)[0])
             if modified_since >= when:
                 self.setResponseCode(NOT_MODIFIED)
                 return CACHED
