Index: doc/howto/defer.xhtml
===================================================================
RCS file: /cvs/Twisted/doc/howto/defer.xhtml,v
retrieving revision 1.5
diff -u -d -r1.5 defer.xhtml
--- doc/howto/defer.xhtml	2 Dec 2003 05:07:20 -0000	1.5
+++ doc/howto/defer.xhtml	9 Dec 2003 16:30:17 -0000
@@ -224,7 +224,7 @@
 class="python">processResult</code>'s return value will be passed to <code
 class="python">printResult</code>, instead of the value initially passed into
 the callback. This gives you a flexible way to chain results together, possibly
-modifying values along the way, (for example, you may wish to pre-processed
+modifying values along the way (for example, you may wish to pre-process
 database query results).</p>
 
 <h3>More about errbacks</h3>
