=== modified file 'doc/core/howto/process.xhtml'
--- doc/core/howto/process.xhtml	2011-05-12 14:39:58 +0000
+++ doc/core/howto/process.xhtml	2012-10-02 15:40:09 +0000
@@ -477,11 +477,13 @@
         if childFD == 1: self.plaintext += data
         if childFD == 4: self.status += data
     def processEnded(self, status):
-        rc = status.value.exitCode
-        if rc == 0:
-            self.deferred.callback(self)
-        else:
-            self.deferred.errback(rc)
+	if self.deferred is not None:
+            d, self.deferred = self.deferred, None
+            rc = status.value.exitCode
+            if rc == 0:
+                d.callback(self)
+            else:
+                d.errback(rc)
 
 def decrypt(crypttext):
     gp = GPGProtocol(crypttext)

=== modified file 'doc/web/howto/client.xhtml'
--- doc/web/howto/client.xhtml	2012-03-18 01:04:00 +0000
+++ doc/web/howto/client.xhtml	2012-10-02 15:39:59 +0000
@@ -347,7 +347,9 @@
         pass
 
     def connectionLost(self, reason):
-        self.deferred.callback(None)
+	if self.deferred is not None:
+            d, self.deferred = self.deferred, None
+            d.callback(None)
 
 
 def cbRequest(response):

