Index: twisted/topfiles/4687.misc
===================================================================
Index: twisted/conch/test/test_ssh.py
===================================================================
--- twisted/conch/test/test_ssh.py	(revision 38110)
+++ twisted/conch/test/test_ssh.py	(working copy)
@@ -6,6 +6,8 @@
 """
 
 import struct
+from zope.interface import implementer
+from zope.interface.verify import verifyClass
 
 try:
     import Crypto.Cipher.DES3
@@ -124,7 +126,7 @@
             listener.stopListening()
 
 
-
+@implementer(session.ISession)
 class ConchSessionForTestAvatar(object):
     """
     An ISession adapter for ConchTestAvatar.
@@ -182,7 +184,13 @@
             raise error.ConchError('bad exec')
         self.avatar.conn.transport.expectedLoseConnection = 1
 
+    def windowChanged(self, newWindowSize):
+        """
+        Called when the size of the remote screen has changed.
+        """
+        pass
 
+
     def eofReceived(self):
         self.eof = 1
 
@@ -525,7 +533,20 @@
             self.onClose.callback(None)
 
 
+class ConchSessionForTestAvatarTest(unittest.TestCase):
+    """
+    Test for L{ConchSessionForTestAvatar}
+    """
 
+    def test_conchSessionForTestAvatarProvidesISession(self):
+        """
+        L{ConchSessionForTestAvatar} provides the L{session.ISession}
+        interface.
+        """
+        verifyClass(session.ISession, ConchSessionForTestAvatar)
+
+
+
 class SSHProtocolTestCase(unittest.TestCase):
     """
     Tests for communication between L{SSHServerTransport} and
