=== modified file 'twisted/conch/test/test_ssh.py'
--- twisted/conch/test/test_ssh.py	2010-06-26 17:40:59 +0000
+++ twisted/conch/test/test_ssh.py	2010-10-13 12:08:22 +0000
@@ -29,15 +29,17 @@
 
 from twisted.conch.test.test_recvline import LoopbackRelay
 
+from zope.interface import implements
 
 
 class ConchTestRealm:
-
     def requestAvatar(self, avatarID, mind, *interfaces):
         unittest.assertEquals(avatarID, 'testuser')
         a = ConchTestAvatar()
         return interfaces[0], a, a.logout
 
+
+
 class ConchTestAvatar(avatar.ConchUser):
     loggedOut = False
 
@@ -86,7 +88,10 @@
             log.msg('stopListening %s' % listener)
             listener.stopListening()
 
+
+
 class ConchSessionForTestAvatar:
+    implements(session.ISession)
 
     def __init__(self, avatar):
         unittest.assert_(isinstance(avatar, ConchTestAvatar))
@@ -155,9 +160,11 @@
         elif self.cmd == 'shell':
             unittest.assert_(self.eof)
 
+
 from twisted.python import components
 components.registerAdapter(ConchSessionForTestAvatar, ConchTestAvatar, session.ISession)
 
+
 class CrazySubsystem(protocol.Protocol):
 
     def __init__(self, *args, **kw):

