Index: test_amp.py
===================================================================
--- test_amp.py	(revision 26541)
+++ test_amp.py	(working copy)
@@ -1164,7 +1164,7 @@
         p.flush()
         self.failIf(tl.isKey)
         self.failUnless(tl.isLocal)
-        self.failUnlessIdentical(tl.keyName, 'hello')
+        self.assertEquals(tl.keyName, 'hello')
         self.failUnlessIdentical(tl.value, x)
         self.failUnless(str(len(x)) in repr(tl))
         self.failUnless("value" in repr(tl))
