--- twisted/protocols/sip.py.orig    2015-10-02 15:39:11.180713603 +0100
+++ twisted/protocols/sip.py    2015-10-02 15:39:37.216713601 +0100
@@ -509,7 +509,7 @@
         for l in paramstring.split(";"):
             if not l:
                 continue
-            k, v = l.split("=")
+            k, v = l.split("=") if "=" in l else (l, 'TRUE')
             params[k] = v
     if clean:
         # rfc 2543 6.21
