Index: twisted/test/test_threads.py
===================================================================
--- twisted/test/test_threads.py	(revision 17329)
+++ twisted/test/test_threads.py	(working copy)
@@ -21,7 +21,18 @@
         reactor.suggestThreadPoolSize(34)
         reactor.suggestThreadPoolSize(4)
 
+    def setUp(self):
+        # We can let the reactor create the threadpool itself, just copy its creation
+        from twisted.python import threadpool
+        reactor.threadpool = threadpool.ThreadPool(0, 10, 'twisted.internet.reactor')
+        # Here, we start it immediately to be sure it'd be up during tests
+        reactor.threadpool.start()
 
+    def tearDown(self):
+        # Just the counter part of setUp
+        reactor.threadpool.stop()
+        reactor.threadpool = None
+
     def testCallInThread(self):
         waiter = threading.Event()
         result = []
