diff --git a/twisted/python/util.py b/twisted/python/util.py
index a937ad6..bb9e01b 100644
--- a/twisted/python/util.py
+++ b/twisted/python/util.py
@@ -564,10 +564,17 @@ class FancyEqMixin:
 
 
 
-def dsu(list, key):
-    L2 = [(key(e), i, e) for (i, e) in zip(range(len(list)), list)]
-    L2.sort()
-    return [e[2] for e in L2]
+def dsu(L, key):
+    """
+    This function passes through to the built-in C{sorted}. Just use
+    C{sorted} instead.
+
+    This function used to have an implementation of decorate-sort-undecorate
+    (aka "Schwartzian transform") to achieve a stable sort. Python 2.4+ now
+    guarantees sort stability, so the implementation became unnecessary.
+    """
+    return sorted(L, key=key)
+
 
 try:
     from twisted.python._initgroups import initgroups as _c_initgroups
diff --git a/twisted/topfiles/4339.misc b/twisted/topfiles/4339.misc
new file mode 100644
index 0000000..e69de29
