Index: twisted/names/client.py
===================================================================
--- twisted/names/client.py	(revision 30567)
+++ twisted/names/client.py	(working copy)
@@ -520,10 +520,11 @@
     Create and return a Resolver.
 
     @type servers: C{list} of C{(str, int)} or C{None}
-    @param servers: If not C{None}, interpreted as a list of addresses of
-    domain name servers to attempt to use.  Addresses should be in dotted-quad
-    form.
 
+    @param servers: If not C{None}, interpreted as a list of domain
+    name servers to attempt to use. Each server is a tuple of address
+    in C{str} dotted-quad form and C{int} port number.
+
     @type resolvconf: C{str} or C{None}
     @param resolvconf: If not C{None}, on posix systems will be interpreted as
     an alternate resolv.conf to use. Will do nothing on windows systems. If
