Index: Twisted/twisted/words/test/test_irc_service.py
===================================================================
--- Twisted/twisted/words/test/test_irc_service.py	(revision 30844)
+++ Twisted/twisted/words/test/test_irc_service.py	(working copy)
@@ -7,7 +7,7 @@
 
 from twisted.trial import unittest
 from twisted.test import proto_helpers
-from twisted.words.service import InMemoryWordsRealm, IRCFactory
+from twisted.words.service import InMemoryWordsRealm, IRCFactory, IRCUser
 from twisted.words.protocols import irc
 from twisted.cred import checkers, portal
 
@@ -108,3 +108,60 @@
              ('example.com', '004',
               ['john', 'example.com', self.factory._serverInfo["serviceVersion"],
                'w', 'n'])])
+
+
+
+class MocksyIRCUser(IRCUser):
+    def __init__(self):
+        self._mocked_codes = []
+    
+    def sendMessage(self, code, *_, **__):
+        self._mocked_codes.append(code)
+
+BADTEXT = '\xff'
+
+class IRCUserBadEncodingTestCase(unittest.TestCase):
+    # TODO: irc_NICK -- but NICKSERV is used for that, so it isn't as easy.
+    
+    def setUp(self):
+        self.ircuser = MocksyIRCUser()
+    
+    def assert_chokesOnBadBytes(self, irc_x, error):
+        getattr(self.ircuser, 'irc_%s' % irc_x)(None, [BADTEXT])
+        self.assertEqual(self.ircuser._mocked_codes, [error])
+    
+    # no such channel
+    
+    def test_JOIN(self):
+        self.assert_chokesOnBadBytes('JOIN', irc.ERR_NOSUCHCHANNEL)
+    
+    def test_NAMES(self):
+        self.assert_chokesOnBadBytes('NAMES', irc.ERR_NOSUCHCHANNEL)
+    
+    def test_TOPIC(self):
+        self.assert_chokesOnBadBytes('TOPIC', irc.ERR_NOSUCHCHANNEL)
+    
+    def test_LIST(self):
+        self.assert_chokesOnBadBytes('LIST', irc.ERR_NOSUCHCHANNEL)
+    
+    # no such nick
+    
+    def test_MODE(self):
+        self.assert_chokesOnBadBytes('MODE', irc.ERR_NOSUCHNICK)
+    
+    def test_PRIVMSG(self):
+        self.assert_chokesOnBadBytes('PRIVMSG', irc.ERR_NOSUCHNICK)
+    
+    def test_WHOIS(self):
+        self.assert_chokesOnBadBytes('WHOIS', irc.ERR_NOSUCHNICK)
+    
+    # not on channel
+    
+    def test_PART(self):
+        self.assert_chokesOnBadBytes('PART', irc.ERR_NOTONCHANNEL)
+    
+    # probably nothing
+    
+    def test_WHO(self):
+        self.assert_chokesOnBadBytes('WHO', irc.RPL_ENDOFWHO)
+    
