Index: twisted/web/_auth/wrapper.py
===================================================================
--- twisted/web/_auth/wrapper.py	(revision 40183)
+++ twisted/web/_auth/wrapper.py	(working copy)
@@ -132,7 +132,7 @@
 
     def getChildWithDefault(self, path, request):
         """
-        Inspect the Authorization HTTP header, and return a deferred which,
+        Inspect the Authorization HTTP header, and return a C{DeferredResource} which,
         when fired after successful authentication, will return an authorized
         C{Avatar}. On authentication failure, an C{UnauthorizedResource} will
         be returned, essentially halting further dispatch on the wrapped
