Index: twisted/mail/test/test_imap.py
===================================================================
--- twisted/mail/test/test_imap.py	(revision 28756)
+++ twisted/mail/test/test_imap.py	(working copy)
@@ -3602,6 +3602,8 @@
             FakeyMessage({}, (), '', '', 999, None),
             FakeyMessage({}, (), '', '', 10101, None),
             FakeyMessage({}, (), '', '', 12345, None),
+            FakeyMessage({}, (), '', '', 20001, None),
+            FakeyMessage({}, (), '', '', 20002, None),
         ]
 
 
@@ -3648,7 +3650,7 @@
         If the search filter ends with a star, all the message from the
         starting point are returned.
         """
-        return self._messageSetSearchTest('2:*', [2, 3])
+        return self._messageSetSearchTest('2:*', [2, 3, 4, 5])
 
 
     def test_searchMessageSetWithList(self):
@@ -3656,10 +3658,10 @@
         If the search filter contains nesting terms, one of which includes a
         message sequence set with a wildcard, IT ALL WORKS GOOD.
         """
-        # 5 is bigger than the biggest message sequence number, but that's
+        # 6 is bigger than the biggest message sequence number, but that's
         # okay, because N:* includes the biggest message sequence number even
         # if N is bigger than that (read the rfc nub).
-        return self._messageSetSearchTest('(5:*)', [3])
+        return self._messageSetSearchTest('(6:*)', [5])
 
 
     def test_searchOr(self):
@@ -3677,7 +3679,7 @@
         all messages in that set are considered for inclusion in the
         results.
         """
-        return self._messageSetSearchTest('OR 2:* 2:*', [2, 3])
+        return self._messageSetSearchTest('OR 2:* 2:*', [2, 3, 4, 5])
 
 
     def test_searchNot(self):
@@ -3685,7 +3687,7 @@
         If the search filter contains a I{NOT} term, all messages
         which do not match the subexpression are returned.
         """
-        return self._messageSetSearchTest('NOT 3', [1, 2])
+        return self._messageSetSearchTest('NOT 3', [1, 2, 4, 5])
 
 
     def test_searchNotMessageSet(self):
Index: twisted/mail/imap4.py
===================================================================
--- twisted/mail/imap4.py	(revision 28756)
+++ twisted/mail/imap4.py	(working copy)
@@ -1435,7 +1435,7 @@
             searchResults = []
         i = 0
 
-        lastSequenceId = result[-1][0]
+        lastSequenceId = result and result[-1][0]
 
         for (i, (id, msg)) in zip(range(5), result):
             # searchFilter and singleSearchStep will mutate the query.  Dang.
@@ -1449,7 +1449,7 @@
         if i == 4:
             from twisted.internet import reactor
             reactor.callLater(
-                0, self.__cbManualSearch, result, tag, mbox, query, uid,
+                0, self.__cbManualSearch, result[5:], tag, mbox, query, uid,
                 searchResults)
         else:
             if searchResults:
