Index: twisted/application/app.py
===================================================================
--- twisted/application/app.py	(revision 38203)
+++ twisted/application/app.py	(working copy)
@@ -621,7 +621,7 @@
             try:
                 self['logger'] = namedAny(self['logger'])
             except Exception, e:
-                raise usage.UsageError("Logger '%s' could not be imported: %s" 
+                raise usage.UsageError("Logger '%s' could not be imported: %s"
                                        % (self['logger'], e))
 
 
@@ -646,8 +646,13 @@
     try:
         config.parseOptions()
     except usage.error, ue:
+        commandChain = [sys.argv[0]]
+        commandConfig = config
+        while commandConfig.subCommand:
+            commandChain.append(commandConfig.subCommand)
+            commandConfig = commandConfig.subOptions
         print config
-        print "%s: %s" % (sys.argv[0], ue)
+        print "%s: %s" % (' '.join(commandChain), ue)
     else:
         runApp(config)
 
@@ -671,4 +676,3 @@
          reactor.addSystemEventTrigger('after', 'shutdown', p.save, 'shutdown')
     reactor.addSystemEventTrigger('before', 'shutdown',
                                   service.IService(application).stopService)
-
