Index: twisted/internet/protocol.py
===================================================================
--- twisted/internet/protocol.py	(revision 30554)
+++ twisted/internet/protocol.py	(working copy)
@@ -306,6 +306,9 @@
         be parametrized in tests. If the factory is serialized, this attribute
         will not be serialized, and the default value (the reactor) will be
         restored when deserialized.
+    @ivar maxRetries: Maximum number of consecutive unsuccessful connection
+        attempts, after which no further connection attempts will be made. If
+        this is not explicitly set, no maximum is applied.
     """
     maxDelay = 3600
     initialDelay = 1.0
