Index: util.py
===================================================================
--- util.py	(revision 24890)
+++ util.py	(working copy)
@@ -157,8 +157,8 @@
 
     def __init__(self, collector, fd, callback=None, *args, **kw):
         self.chunks = []
+        self.startProducing(fd)
         Pager.__init__(self, collector, callback, *args, **kw)
-        self.startProducing(fd)
 
     def startProducing(self, fd):
         self.deferred = basic.FileSender().beginFileTransfer(fd, self)
