=== modified file 'twisted/conch/checkers.py'
--- twisted/conch/checkers.py	2008-04-02 02:10:22 +0000
+++ twisted/conch/checkers.py	2008-04-03 06:53:32 +0000
@@ -84,9 +84,8 @@
             return failure.Failure(error.ValidPublicKey())
         else:
             try:
-                pubKey = keys.getPublicKeyObject(data = credentials.blob)
-                if keys.verifySignature(pubKey, credentials.signature,
-                                        credentials.sigData):
+                pubKey = keys.Key.fromString(credentials.blob)
+                if pubKey.verify(credentials.signature, credentials.sigData):
                     return credentials.username
             except: # any error should be treated as a failed login
                 f = failure.Failure()

=== modified file 'twisted/conch/ssh/userauth.py'
--- twisted/conch/ssh/userauth.py	2008-04-02 02:10:22 +0000
+++ twisted/conch/ssh/userauth.py	2008-04-03 06:49:26 +0000
@@ -130,7 +130,7 @@
     def auth_publickey(self, packet):
         hasSig = ord(packet[0])
         algName, blob, rest = getNS(packet[1:], 2)
-        pubKey = keys.getPublicKeyObject(data = blob)
+        pubKey = keys.Key.fromString(blob).keyObject
         signature = hasSig and getNS(rest)[0] or None
         if hasSig:
             b = NS(self.transport.sessionID) + chr(MSG_USERAUTH_REQUEST) + \

