--- client.py.old	Fri Sep 26 09:41:42 2003
+++ client.py	Fri Sep 26 09:38:43 2003
@@ -193,12 +193,11 @@
         self.response_headers = headers
         if headers.has_key('set-cookie'):
             for cookie in headers['set-cookie']:
-                for cookie in cookie.split(','):
-                    cookparts = cookie.split(';')
-                    cook = cookparts[0]
-                    cook.lstrip()
-                    k, v = cook.split('=', 1)
-                    self.cookies[k.lstrip()] = v.lstrip()
+                cookparts = cookie.split(';')
+                cook = cookparts[0]
+                cook.lstrip()
+                k, v = cook.split('=', 1)
+                self.cookies[k.lstrip()] = v.lstrip()
 
     def gotStatus(self, version, status, message):
         self.version, self.status, self.message = version, status, message
