diff --git twisted/pair/raw.py twisted/pair/raw.py
index 0d3875b..f205f66 100644
--- twisted/pair/raw.py
+++ twisted/pair/raw.py
@@ -1,13 +1,10 @@
 # Copyright (c) Twisted Matrix Laboratories.
 # See LICENSE for details.
-
-#
-
 """Interface definitions for working with raw packets"""
 
-from twisted.internet import protocol
 from zope.interface import Interface
 
+
 class IRawDatagramProtocol(Interface):
     """An interface for protocols such as UDP, ICMP and TCP."""
 
@@ -21,6 +18,7 @@ class IRawDatagramProtocol(Interface):
         An IP datagram has been received. Parse and process it.
         """
 
+
 class IRawPacketProtocol(Interface):
     """An interface for low-level protocols such as IP and ARP."""
 
diff --git twisted/pair/test/test_ethernet.py twisted/pair/test/test_ethernet.py
index 2b675fe..253efb5 100644
--- twisted/pair/test/test_ethernet.py
+++ twisted/pair/test/test_ethernet.py
@@ -1,17 +1,15 @@
 # Copyright (c) Twisted Matrix Laboratories.
 # See LICENSE for details.
-
-#
 from twisted.trial import unittest
 
-from twisted.internet import protocol, reactor, error
-from twisted.python import failure, components
+from twisted.python import components
 from twisted.pair import ethernet, raw
 from zope.interface import implements
 
+
 class MyProtocol:
     implements(raw.IRawPacketProtocol)
-    
+
     def __init__(self, expecting):
         self.expecting = list(expecting)
 
diff --git twisted/pair/test/test_ip.py twisted/pair/test/test_ip.py
index ed1623b..95216f8 100644
--- twisted/pair/test/test_ip.py
+++ twisted/pair/test/test_ip.py
@@ -1,17 +1,15 @@
 # Copyright (c) Twisted Matrix Laboratories.
 # See LICENSE for details.
-
-#
 from twisted.trial import unittest
 
-from twisted.internet import protocol, reactor, error
-from twisted.python import failure, components
+from twisted.python import components
 from twisted.pair import ip, raw
 from zope import interface
 
+
 class MyProtocol:
     interface.implements(raw.IRawDatagramProtocol)
-    
+
     def __init__(self, expecting):
         self.expecting = list(expecting)
 
diff --git twisted/pair/test/test_rawudp.py twisted/pair/test/test_rawudp.py
index f53f078..fe9f12f 100644
--- twisted/pair/test/test_rawudp.py
+++ twisted/pair/test/test_rawudp.py
@@ -4,8 +4,7 @@
 #
 from twisted.trial import unittest
 
-from twisted.internet import protocol, reactor, error
-from twisted.python import failure
+from twisted.internet import protocol
 from twisted.pair import rawudp
 
 class MyProtocol(protocol.DatagramProtocol):
diff --git twisted/pair/topfiles/setup.py twisted/pair/topfiles/setup.py
index c42754f..e9eb5e6 100644
--- twisted/pair/topfiles/setup.py
+++ twisted/pair/topfiles/setup.py
@@ -1,8 +1,5 @@
 # Copyright (c) Twisted Matrix Laboratories.
 # See LICENSE for details.
-
-import sys
-
 try:
     from twisted.python import dist
 except ImportError:
