--- manhole.py_FCS	2006-03-13 13:51:51.000000000 +0100
+++ manhole.py	2006-03-13 15:08:18.000000000 +0100
@@ -245,8 +245,11 @@
                 self.written.append(self.normalColor)
 
     def __str__(self):
-        s = ''.join(self.written)
-        return s.strip('\n').splitlines()[-1]
+        s = ''.join(self.written).strip("\n")
+        if s:
+            return s.splitlines()[-1]
+        else:
+            return ''
 
 def lastColorizedLine(source):
     """Tokenize and colorize the given Python source.
