diff --git twisted/conch/test/test_cftp.py twisted/conch/test/test_cftp.py
index e8f6c10..41aa2c7 100644
--- twisted/conch/test/test_cftp.py
+++ twisted/conch/test/test_cftp.py
@@ -1393,6 +1393,7 @@ class OurServerSftpClientTests(CFTPClientTestBase):
         cmds = ('-o', 'IdentityFile=dsa_test',
                 '-o', 'UserKnownHostsFile=kh_test',
                 '-o', 'HostKeyAlgorithms=ssh-rsa',
+                '-o', 'PubkeyAcceptedKeyTypes=ssh-dss',
                 '-o', 'Port=%i' % (port,), '-b', fn, 'testuser@127.0.0.1')
         d = getProcessOutputAndValue("sftp", cmds)
         def check(result):
diff --git twisted/conch/test/test_conch.py twisted/conch/test/test_conch.py
index 1dc88f0..2d249ea 100644
--- twisted/conch/test/test_conch.py
+++ twisted/conch/test/test_conch.py
@@ -508,6 +508,7 @@ class OpenSSHClientMixin:
         cmdline = ('ssh -2 -l testuser -p %i '
                    '-oUserKnownHostsFile=kh_test '
                    '-oPasswordAuthentication=no '
+                   '-oPubkeyAcceptedKeyTypes=ssh-dss '
                    # Always use the RSA key, since that's the one in kh_test.
                    '-oHostKeyAlgorithms=ssh-rsa '
                    '-a '
diff --git twisted/topfiles/8180.misc twisted/topfiles/8180.misc
new file mode 100644
index 0000000..e69de29
