Index: twisted/lore/tree.py
===================================================================
--- twisted/lore/tree.py	(revision 20920)
+++ twisted/lore/tree.py	(working copy)
@@ -148,7 +148,7 @@
     htmlizer.filter(oldio, newio, writer=htmlizer.SmallerHTMLWriter)
     newio.seek(0)
     newel = microdom.parse(newio).documentElement
-    newel.setAttribute("class", "python")
+    newel.attributes = node.attributes
     node.parentNode.replaceChild(newel, node)
 
 
