Index: twisted/names/client.py
===================================================================
--- twisted/names/client.py	(revision 18863)
+++ twisted/names/client.py	(working copy)
@@ -215,10 +215,12 @@
             timeout = self.timeout
 
         addresses = self.servers + list(self.dynServers)
+        
         if not addresses:
             return defer.fail(IOError("No domain name servers available"))
 
-        used = addresses.pop()
+        addresses.reverse(); # put servers to correct ascending order according to "man resolv.conf"
+        used = addresses.pop()        
         return self.protocol.query(used, queries, timeout[0]
             ).addErrback(self._reissue, addresses, [used], queries, timeout
             )
