=== modified file 'twisted/internet/utils.py'
--- twisted/internet/utils.py	2012-10-01 13:37:18 +0000
+++ twisted/internet/utils.py	2013-02-20 12:00:18 +0000
@@ -103,7 +103,7 @@
                        full path should be used.
 
     @param args: the command line arguments to pass to the process; a
-                 sequence of strings. The first string should *NOT* be the
+                 sequence of strings. The first string should B{NOT} be the
                  executable's name.
 
     @param env: the environment variables to pass to the processs; a

