Index: twisted/application/internet.py
===================================================================
--- twisted/application/internet.py	(revision 43542)
+++ twisted/application/internet.py	(working copy)
@@ -38,8 +38,6 @@
 """
 
 from twisted.python import log
-from twisted.python.deprecate import deprecatedModuleAttribute
-from twisted.python.versions import Version
 from twisted.application import service
 from twisted.internet import task
 
@@ -209,6 +207,8 @@
     for side in 'Server Client'.split():
         if tran == "Multicast" and side == "Client":
             continue
+        if tran == "UDP" and side == "Client":
+            continue
         base = globals()['_Abstract'+side]
         doc = _doc[side] % vars()
         klass = types.ClassType(tran+side, (base,),
@@ -217,15 +217,6 @@
 
 
 
-deprecatedModuleAttribute(
-        Version("Twisted", 13, 1, 0),
-        "It relies upon IReactorUDP.connectUDP "
-        "which was removed in Twisted 10. "
-        "Use twisted.application.internet.UDPServer instead.",
-        "twisted.application.internet", "UDPClient")
-
-
-
 class TimerService(_VolatileDataService):
     """
     Service to periodically call a function
@@ -399,7 +390,7 @@
 
 
 __all__ = (['TimerService', 'CooperatorService', 'MulticastServer',
-            'StreamServerEndpointService'] +
+            'StreamServerEndpointService', 'UDPServer'] +
            [tran+side
-            for tran in 'TCP UNIX SSL UDP UNIXDatagram'.split()
+            for tran in 'TCP UNIX SSL UNIXDatagram'.split()
             for side in 'Server Client'.split()])
Index: twisted/test/test_application.py
===================================================================
--- twisted/test/test_application.py	(revision 43542)
+++ twisted/test/test_application.py	(working copy)
@@ -358,21 +358,6 @@
         return defer.maybeDeferred(t.stopService).addCallback(onStop)
 
 
-    def test_deprecatedUDPClient(self):
-        """
-        L{internet.UDPClient} is deprecated since Twisted-13.1.
-        """
-        internet.UDPClient
-        warningsShown = self.flushWarnings([self.test_deprecatedUDPClient])
-        self.assertEqual(1, len(warningsShown))
-        self.assertEqual(
-                "twisted.application.internet.UDPClient was deprecated in "
-                "Twisted 13.1.0: It relies upon IReactorUDP.connectUDP "
-                "which was removed in Twisted 10. "
-                "Use twisted.application.internet.UDPServer instead.",
-                warningsShown[0]['message'])
-
-
     def testPrivileged(self):
         factory = protocol.ServerFactory()
         factory.protocol = TestEcho
@@ -520,6 +505,8 @@
             for side in 'Server Client'.split():
                 if tran == "Multicast" and side == "Client":
                     continue
+                if tran == "UDP" and side == "Client":
+                    continue
                 self.assertTrue(hasattr(internet, tran + side))
                 method = getattr(internet, tran + side).method
                 prefix = {'Server': 'listen', 'Client': 'connect'}[side]
Index: twisted/topfiles/7702.removal
===================================================================
--- twisted/topfiles/7702.removal	(revision 0)
+++ twisted/topfiles/7702.removal	(revision 0)
@@ -0,0 +1 @@
+twisted.application.internet.UDPClient, deprecated since Twisted 13.1.0, has been removed.
