Index: twisted/application/internet.py
===================================================================
--- twisted/application/internet.py	(revision 43542)
+++ twisted/application/internet.py	(working copy)
@@ -217,15 +217,6 @@
 
 
 
-deprecatedModuleAttribute(
-        Version("Twisted", 13, 1, 0),
-        "It relies upon IReactorUDP.connectUDP "
-        "which was removed in Twisted 10. "
-        "Use twisted.application.internet.UDPServer instead.",
-        "twisted.application.internet", "UDPClient")
-
-
-
 class TimerService(_VolatileDataService):
     """
     Service to periodically call a function
Index: twisted/test/test_application.py
===================================================================
--- twisted/test/test_application.py	(revision 43542)
+++ twisted/test/test_application.py	(working copy)
@@ -358,21 +358,6 @@
         return defer.maybeDeferred(t.stopService).addCallback(onStop)
 
 
-    def test_deprecatedUDPClient(self):
-        """
-        L{internet.UDPClient} is deprecated since Twisted-13.1.
-        """
-        internet.UDPClient
-        warningsShown = self.flushWarnings([self.test_deprecatedUDPClient])
-        self.assertEqual(1, len(warningsShown))
-        self.assertEqual(
-                "twisted.application.internet.UDPClient was deprecated in "
-                "Twisted 13.1.0: It relies upon IReactorUDP.connectUDP "
-                "which was removed in Twisted 10. "
-                "Use twisted.application.internet.UDPServer instead.",
-                warningsShown[0]['message'])
-
-
     def testPrivileged(self):
         factory = protocol.ServerFactory()
         factory.protocol = TestEcho
Index: twisted/topfiles/7702.removal
===================================================================
--- twisted/topfiles/7702.removal	(revision 0)
+++ twisted/topfiles/7702.removal	(revision 0)
@@ -0,0 +1 @@
+twisted.application.internet.UDPClient, deprecated since Twisted 13.1.0, has been removed.
