Index: doc/core/development/policy/test-standard.xhtml
===================================================================
--- doc/core/development/policy/test-standard.xhtml	(revision 27948)
+++ doc/core/development/policy/test-standard.xhtml	(working copy)
@@ -109,10 +109,10 @@
     <p>PyUnit provides a large number of assertion methods to be used when
     writing tests.  Many of these are redundant.  For consistency, Twisted
     unit tests should use the <code>assert</code> forms rather than the
-    <code>fail</code> forms and the singular names rather than the plural
-    names.  For example, don't use <code>failUnlessEqual</code>, use <code>
-    assertEqual</code> and don't use <code>assertNotEquals</code>, use
-    <code>assertNotEqual</code>.  <code>assertTrue</code> is also preferred
+    <code>fail</code> forms and the plural names rather than the singular
+    names.  For example, don't use <code>failUnlessEquals</code>, use <code>
+    assertEquals</code> and don't use <code>assertNotEqual</code>, use
+    <code>assertNotEquals</code>.  <code>assertTrue</code> is also preferred
     over <code>assert_</code>.  You may notice this convention is not
     followed everywhere in the Twisted codebase.  If you are changing some
     test code and notice the wrong method being used in nearby code, feel
