--- twisted/internet/abstract.py	2008-03-04 22:36:22.000000000 +0100
+++ /usr/lib/python2.5/site-packages/twisted/internet/abstract.py	2008-11-16 20:25:27.000000000 +0100
@@ -144,6 +144,8 @@
         Whatever this returns is then returned by doWrite.
         """
         # default implementation, telling reactor we're finished
+        if isinstance(self.disconnecting, Exception):
+            return self.disconnecting
         return main.CONNECTION_DONE
 
     def _closeWriteConnection(self):
@@ -234,7 +236,7 @@
             else:
                 self.stopReading()
                 self.startWriting()
-                self.disconnecting = 1
+                self.disconnecting = _connDone
 
     def loseWriteConnection(self):
         self._writeDisconnecting = True
--- twisted/protocols/basic.py	2008-03-04 22:36:22.000000000 +0100
+++ /usr/lib/python2.5/site-packages/twisted/protocols/basic.py	2008-11-16 20:23:09.000000000 +0100
@@ -284,7 +284,7 @@
         be more than one line, or may be only the initial portion of the
         line.
         """
-        return self.transport.loseConnection()
+        return self.transport.loseConnection(error.ConnectionLost('Line length exceeded'))
 
 
 class StringTooLongError(AssertionError):
