Index: doc/web/howto/web-in-60/dynamic-dispatch.xhtml
===================================================================
--- doc/web/howto/web-in-60/dynamic-dispatch.xhtml	(revision 35394)
+++ doc/web/howto/web-in-60/dynamic-dispatch.xhtml	(working copy)
@@ -78,7 +78,7 @@
 </pre>
 
 <p>And that's all. Any resource-based dynamic URL handling is going to look
-basically like <code>Calendar.getPage</code>. Here's the full example code:</p>
+basically like <code>Calendar.getChild</code>. Here's the full example code:</p>
 
 <pre class="python">
 from twisted.web.server import Site
