Index: trunk/twisted/web/monitor.py
===================================================================
--- trunk/twisted/web/monitor.py	(revision 24190)
+++ trunk/twisted/web/monitor.py	(working copy)
@@ -5,9 +5,9 @@
 from twisted.web import client
 from twisted.internet import reactor
 import md5
-from zope.interface import implements
+from zope.interface import implements, Interface
 
-class IChangeNotified:
+class IChangeNotified(Interface):
     pass
 
 class BaseChangeNotified:
