Index: twisted/python/urlpath.py
===================================================================
--- twisted/python/urlpath.py	(revision 30568)
+++ twisted/python/urlpath.py	(working copy)
@@ -33,6 +33,12 @@
             return result
 
     def fromString(klass, st):
+        """
+        Creates a URLPath instance from the given string.
+
+        @type st: C{str}
+        @param st: String for creating the object from.
+        """
         t = urlparse.urlsplit(st)
         u = klass(*t)
         return u
@@ -40,6 +46,9 @@
     fromString = classmethod(fromString)
 
     def fromRequest(klass, request):
+        """
+        Creates a URLPath instance from a given request.
+        """
         return klass.fromString(request.prePathURL())
 
     fromRequest = classmethod(fromRequest)
@@ -56,6 +65,12 @@
 
 
     def sibling(self, path, keepQuery=False):
+        """
+        Return a sibling to a parent.
+
+        @type path: C{str}
+        @param path: The parent path for finding a sibling to.
+        """
         l = self.pathList()
         l[-1] = path
         return self._pathMod(l, keepQuery)
@@ -127,9 +142,10 @@
                         query,
                         fragment)
 
-
-    
-    def __str__(self):
+    def toString(self):
+        """
+        Returns a string on the called part of URLPath
+        """
         x = urlparse.urlunsplit((
             self.scheme, self.netloc, self.path,
             self.query, self.fragment))
