Index: twisted/python/urlpath.py
===================================================================
--- twisted/python/urlpath.py	(revision 27248)
+++ twisted/python/urlpath.py	(working copy)
@@ -60,11 +60,15 @@
         return self._pathMod(l, keepQuery)
 
     def child(self, path, keepQuery=0):
+        #Clean up path
+        # 1) replace all / with urlescaped version
+        # 2) escape the rest
+        fixedPath=urllib.quote(path).replace('/','%2f')
         l = self.pathList()
         if l[-1] == '':
-            l[-1] = path
+            l[-1] = fixedPath
         else:
-            l.append(path)
+            l.append(fixedPath)
         return self._pathMod(l, keepQuery)
 
     def parent(self, keepQuery=0):
Index: twisted/test/test_paths.py
===================================================================
--- twisted/test/test_paths.py	(revision 27248)
+++ twisted/test/test_paths.py	(working copy)
@@ -831,4 +831,8 @@
         # here should be equivalent to '.'
         self.assertEquals(str(self.path.here()), 'http://example.com/foo/')
         self.assertEquals(str(self.path.child('').here()), 'http://example.com/foo/bar/')
+        
+    def test_childQuotes(self):
+        #child() with quotes (",') and slashes  should be properly escaped
+        self.assertEquals(str(self.path.child("hello/ there")),'http://example.com/foo/bar/hello%2f%20there')
 
