Index: twisted/conch/ssh/channel.py
===================================================================
--- twisted/conch/ssh/channel.py	(revision 27230)
+++ twisted/conch/ssh/channel.py	(working copy)
@@ -254,7 +254,9 @@
 
     def getPeer(self):
         """
-        Return a tuple describing the other side of the connection.
+        Return a tuple describing the side of the connection that didn't open
+        the channel. The tuple contains 'SSH' prepended to what was returned by
+        the transport's C{getPeer} method.
 
         @rtype: C{tuple}
         """
@@ -262,7 +264,9 @@
 
     def getHost(self):
         """
-        Return a tuple describing our side of the connection.
+        Return a tuple describing the side of the connection that opened the
+        channel. The tuple contains 'SSH' prepended to what was returned by the
+        transport's C{getHost} method.
 
         @rtype: C{tuple}
         """
