diff --git a/twisted/web/xmlrpc.py b/twisted/web/xmlrpc.py
index 86415d1..83d7e0c 100644
--- a/twisted/web/xmlrpc.py
+++ b/twisted/web/xmlrpc.py
@@ -79,7 +79,9 @@ class XMLRPC(resource.Resource):
     Methods published can return XML-RPC serializable results, Faults,
     Binary, Boolean, DateTime, Deferreds, or Handler instances.
 
-    By default methods beginning with 'xmlrpc_' are published.
+    By default, methods beginning with 'xmlrpc_' are published.  If such
+    a method has an attribute called usesRequest, it will be passed the
+    request as its first argument.
 
     Sub-handlers for prefixed methods (e.g., system.listMethods)
     can be added with putSubHandler. By default, prefixes are
@@ -123,6 +125,8 @@ class XMLRPC(resource.Resource):
             except Fault, f:
                 self._cbRender(f, request)
             else:
+                if hasattr(function, "usesRequest"):
+                    args = (request,) + args
                 defer.maybeDeferred(function, *args).addErrback(
                     self._ebRender
                 ).addCallback(
@@ -190,6 +194,16 @@ class XMLRPC(resource.Resource):
         return reflect.prefixedMethodNames(self.__class__, 'xmlrpc_')
 
 
+def usesRequest(method):
+    """Decorater which sets the usesRequest attribute for the given method.
+
+    If an XML RPC method is decorated with usesRequest, then the request
+    will be passed as the first argument after self.
+    """
+    method.usesRequest = True
+    return method
+
+
 class XMLRPCIntrospection(XMLRPC):
     """
     Implement the XML-RPC Introspection API.
