Index: twisted/internet/pollreactor.py
===================================================================
--- twisted/internet/pollreactor.py     (revision 13035)
+++ twisted/internet/pollreactor.py     (working copy)
@@ -53,6 +53,8 @@
         try:
             # the easy way
             fd = selectable.fileno()
+            # make sure the fd is actually real
+            mdict[fd]
         except:
             # the hard way: necessary because fileno() may disappear at any
             # moment, thanks to python's underlying sockets impl
