Index: twisted/internet/pollreactor.py
===================================================================
--- twisted/internet/pollreactor.py     (revision 13035)
+++ twisted/internet/pollreactor.py     (working copy)
@@ -53,6 +53,9 @@
         try:
             # the easy way
             fd = selectable.fileno()
+            # abstract.FileDescriptor.fileno returns -1; that's never valid
+            if fd == -1:
+                raise ValueError
         except:
             # the hard way: necessary because fileno() may disappear at any
             # moment, thanks to python's underlying sockets impl
