Index: twisted/internet/tcp.py
===================================================================
--- twisted/internet/tcp.py     (revision 13035)
+++ twisted/internet/tcp.py     (working copy)
@@ -459,12 +459,13 @@
             not hasattr(self, "connector")):
             return

+        self.fileno = lambda fileno=self.fileno(): fileno
         try:
             self._closeSocket()
         except AttributeError:
             pass
         else:
-            del self.socket, self.fileno
+            del self.socket

         self.connector.connectionFailed(failure.Failure(err))
         if hasattr(self, "reactor"):
