Index: twisted/words/test/test_jabberjid.py
===================================================================
--- twisted/words/test/test_jabberjid.py	(revision 13065)
+++ twisted/words/test/test_jabberjid.py	(working copy)
@@ -59,13 +59,13 @@
 
 class JIDClassTest(unittest.TestCase):
     def testBasic(self):
-        j = jid.intern("user@host")
+        j = jid.internJID("user@host")
         self.assertEquals(j.userhost(), "user@host")
         self.assertEquals(j.user, "user")
         self.assertEquals(j.host, "host")
         self.assertEquals(j.resource, None)
 
-        j2 = jid.intern("user@host")
+        j2 = jid.internJID("user@host")
         self.assertEquals(id(j), id(j2))
 
         j_uhj = j.userhostJID()
Index: twisted/words/protocols/jabber/jid.py
===================================================================
--- twisted/words/protocols/jabber/jid.py	(revision 13065)
+++ twisted/words/protocols/jabber/jid.py	(working copy)
@@ -63,7 +63,7 @@
 
 __internJIDs = {}
 
-def intern(str):
+def internJID(str):
     # XXX: Ensure that stringprep'd jids map to same JID
     if str in __internJIDs:
         return __internJIDs[str]
@@ -94,7 +94,7 @@
     def userhostJID(self):
         if self.resource:
             if "_uhjid" not in self.__dict__:
-                self._uhjid = jid.intern(self.userhost())
+                self._uhjid = internJID(self.userhost())
             return self._uhjid
         else:
             return self
