Index: doc/words/examples/ircLogBot.py
===================================================================
--- doc/words/examples/ircLogBot.py	(revision 34188)
+++ doc/words/examples/ircLogBot.py	(working copy)
@@ -1,5 +1,6 @@
 # Copyright (c) Twisted Matrix Laboratories.
 # See LICENSE for details.
+from __future__ import with_statement
 
 
 """
@@ -25,11 +26,12 @@
 
 # twisted imports
 from twisted.words.protocols import irc
-from twisted.internet import reactor, protocol
+from twisted.internet import reactor, protocol, task
 from twisted.python import log
 
 # system imports
 import time, sys
+import datetime
 
 
 class MessageLogger:
@@ -143,16 +145,36 @@
         print "connection failed:", reason
         reactor.stop()
 
+def logDate(filename):
+    with open(filename, "a") as log:
+        today = datetime.date.today()
+        text = time.asctime(today.timetuple()) \
+            .replace("00:00:00 ", "")
+        log.write('[%s]\n' % (text))
 
 if __name__ == '__main__':
+    channel = sys.argv[1]
+    filename = sys.argv[2]
+
     # initialize logging
     log.startLogging(sys.stdout)
     
     # create factory protocol and application
-    f = LogBotFactory(sys.argv[1], sys.argv[2])
+    f = LogBotFactory(channel, filename)
 
     # connect factory to this host and port
     reactor.connectTCP("irc.freenode.net", 6667, f)
 
+    # set up date logging
+    def startDateLoop(self, filename):
+        task.LoopingCall(logDate, filename).start(60*60*24)
+    # schedule it at one second past midnight
+    now = datetime.datetime.utcnow()
+    now_ts = time.mktime(now.timetuple())
+    midnight = (now + datetime.timedelta(days=1)) \
+            .replace(hour=0, minute=0, second=1, microsecond=0)
+    midnight_ts = time.mktime(midnight.timetuple())
+    reactor.callLater(midnight_ts - now_ts, startDateLoop, filename)
+
     # run bot
     reactor.run()
