Index: twisted/xish/test/test_domish.py
===================================================================
--- twisted/xish/test/test_domish.py	(revision 12945)
+++ twisted/xish/test/test_domish.py	(working copy)
@@ -122,6 +122,8 @@
         self.match_list = matches
     
     def testSuxStream(self):
+        if domish.SuxElementStream is None:
+            raise unittest.SkipTest, "Skipping SuxElementStream test, since twisted.web is not available."
         # Setup the stream
         self.setupStream(domish.SuxElementStream(),
                          [query1_root, query1_elem1])
