Index: doc/core/development/policy/coding-standard.xhtml
===================================================================
--- doc/core/development/policy/coding-standard.xhtml	(revision 31665)
+++ doc/core/development/policy/coding-standard.xhtml	(working copy)
@@ -272,7 +272,7 @@
 
     <p>Docstrings should be written in epytext format; more
     documentation is available in the
-    <a href="http://epydoc.sourceforge.net/epytext.html"
+    <a href="http://epydoc.sourceforge.net/manual-epytext.html"
     >Epytext Markup Language documentation</a>.</p>
 
     <p>Additionally, to accommodate emacs users:</p>
