Index: twisted/web/template.py
===================================================================
--- twisted/web/template.py	(revision 32810)
+++ twisted/web/template.py	(working copy)
@@ -375,19 +375,26 @@
         return self._loadedTemplate
 
 
-
+# Last updated October 2011, using W3Schools as a reference. Link:
+# http://www.w3schools.com/html5/html5_reference.asp
+# Note that <xmp> is explicitly omitted; its semantics do not work with
+# t.w.template and it is officially deprecated.
 VALID_HTML_TAG_NAMES = set([
-    'a', 'abbr', 'acronym', 'address', 'applet', 'area', 'b', 'base',
-    'basefont', 'bdo', 'big', 'blockquote', 'body', 'br', 'button', 'caption',
-    'center', 'cite', 'code', 'col', 'colgroup', 'dd', 'del', 'dfn', 'dir',
-    'div', 'dl', 'dt', 'em', 'fieldset', 'font', 'form', 'frame', 'frameset',
-    'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'head', 'hr', 'html', 'i', 'iframe',
-    'img', 'input', 'ins', 'isindex', 'kbd', 'label', 'legend', 'li', 'link',
-    'map', 'menu', 'meta', 'noframes', 'noscript', 'object', 'ol', 'optgroup',
-    'option', 'p', 'param', 'pre', 'q', 's', 'samp', 'script', 'select',
-    'small', 'span', 'strike', 'strong', 'style', 'sub', 'sup', 'table',
-    'tbody', 'td', 'textarea', 'tfoot', 'th', 'thead', 'title', 'tr', 'tt', 'u',
-    'ul', 'var'
+    'a', 'abbr', 'acronym', 'address', 'applet', 'area', 'article', 'aside',
+    'audio', 'b', 'base', 'basefont', 'bdi', 'bdo', 'big', 'blockquote',
+    'body', 'br', 'button', 'canvas', 'caption', 'center', 'cite', 'code',
+    'col', 'colgroup', 'command', 'datalist', 'dd', 'del', 'details', 'dfn',
+    'dir', 'div', 'dl', 'dt', 'em', 'embed', 'fieldset', 'figcaption',
+    'figure', 'font', 'footer', 'form', 'frame', 'frameset', 'h1', 'h2', 'h3',
+    'h4', 'h5', 'h6', 'head', 'header', 'hgroup', 'hr', 'html', 'i', 'iframe',
+    'img', 'input', 'ins', 'isindex', 'keygen', 'kbd', 'label', 'legend',
+    'li', 'link', 'map', 'mark', 'menu', 'meta', 'meter', 'nav', 'noframes',
+    'noscript', 'object', 'ol', 'optgroup', 'option', 'output', 'p', 'param',
+    'pre', 'progress', 'q', 'rp', 'rt', 'ruby', 's', 'samp', 'script',
+    'section', 'select', 'small', 'source', 'span', 'strike', 'strong',
+    'style', 'sub', 'summary', 'sup', 'table', 'tbody', 'td', 'textarea',
+    'tfoot', 'th', 'thead', 'time', 'title', 'tr', 'tt', 'u', 'ul', 'var',
+    'video', 'wbr',
 ])
 
 
Index: twisted/web/test/test_template.py
===================================================================
--- twisted/web/test/test_template.py	(revision 32810)
+++ twisted/web/test/test_template.py	(working copy)
@@ -12,13 +12,60 @@
 from twisted.internet.defer import succeed, gatherResults
 from twisted.trial.unittest import TestCase
 from twisted.web.template import (
-    Element, renderer, XMLFile, XMLString)
+    Element, renderer, tags, XMLFile, XMLString)
 
 from twisted.web.error import MissingTemplateLoader, MissingRenderMethod
 
 from twisted.web._element import UnexposedMethodError
 from twisted.web.test._util import FlattenTestCase
 
+class TagFactoryTests(TestCase):
+    """
+    Tests for L{_TagFactory} through the publicly-exposed L{tags} object.
+    """
+
+    def test_lookupTag(self):
+        """
+        HTML tags can be retrieved through C{tags}.
+        """
+
+        tag = tags.a
+        self.assertEqual(tag.tagName, "a")
+
+    def test_lookupHTML5Tag(self):
+        """
+        Twisted supports the latest and greatest HTML tags from the HTML5
+        specification.
+        """
+
+        tag = tags.video
+        self.assertEqual(tag.tagName, "video")
+
+    def test_lookupTransparentTag(self):
+        """
+        To support transparent inclusion in templates, there is a special tag,
+        the transparent tag, which has no name of its own but is accessed
+        through the "transparent" attribute.
+        """
+
+        tag = tags.transparent
+        self.assertEqual(tag.tagName, "")
+
+    def test_lookupInvalidTag(self):
+        """
+        Invalid tags which are not part of HTML cause AttributeErrors when
+        accessed through C{tags}.
+        """
+
+        self.assertRaises(AttributeError, getattr, tags, "invalid")
+
+    def test_lookupXMP(self):
+        """
+        As a special case, the <xmp> tag is simply not available through
+        C{tags} or any other part of the templating machinery.
+        """
+
+        self.assertRaises(AttributeError, getattr, tags, "xmp")
 
 class ElementTests(TestCase):
     """
Index: twisted/web/topfiles/5306.feature
===================================================================
--- twisted/web/topfiles/5306.feature	(revision 0)
+++ twisted/web/topfiles/5306.feature	(revision 0)
@@ -0,0 +1 @@
+twisted.web.template now supports HTML tags from the HTML5 standard, including <canvas> and <video>.
