Index: nevow/scripts/xmlgettext.py
===================================================================
--- nevow/scripts/xmlgettext.py	(revision 1805)
+++ nevow/scripts/xmlgettext.py	(working copy)
@@ -1,4 +1,5 @@
 from xml.dom import pulldom
+from cStringIO import StringIO
 from twisted.python import usage
 from twisted.application import app
 import nevow
@@ -36,7 +37,19 @@
             data, self.buffer = self.buffer, ''
         return data
 
-def process(filename):
+def getMsgID(node):
+    out = StringIO()
+    print >>out, 'msgid ""'
+    for child in node.childNodes:
+        s = child.toxml('utf-8')
+        s = s.replace('\\', '\\\\')
+        s = s.replace('"', '\\"')
+        s = s.replace('\n', '\\n')
+        print >>out, '"%s"' % s
+    print >>out, 'msgstr ""'
+    return out.getvalue()
+
+def process(filename, messages):
     f = open(filename, 'rU')
     stream = LineBasedStream(f)
     events = pulldom.parse(stream)
@@ -49,17 +62,17 @@
                 if value == 'i18n':
                     events.expandNode(node)
 
-                    print '#: %s:%d' % (filename, events.parser.getLineNumber())
-                    print 'msgid ""'
-                    for child in node.childNodes:
-                        s = child.toxml('utf-8')
-                        s = s.replace('\\', '\\\\')
-                        s = s.replace('"', '\\"')
-                        s = s.replace('\n', '\\n')
-                        print '"%s"' % s
-                    print 'msgstr ""'
-                    print
+                    msgid = getMsgID(node)
+                    l = messages.setdefault(msgid, [])
+                    l.append('#: %s:%d' % (filename, events.parser.getLineNumber()))
 
+
+def report(messages):
+    for msgid, locations in messages.items():
+        for line in locations:
+            print line
+        print msgid
+
 class GettextOptions(usage.Options):
     def opt_version(self):
         print 'Nevow version:', nevow.__version__
@@ -69,9 +82,13 @@
         self['files'] = files
 
 def runApp(config):
+    messages = {}
+
     for filename in config['files']:
-        process(filename)
+        process(filename, messages)
 
+    report(messages)
+
 def run():
     from twisted.application import app
     app.run(runApp, GettextOptions)
Index: examples/i18n/hello.html
===================================================================
--- examples/i18n/hello.html	(revision 1756)
+++ examples/i18n/hello.html	(working copy)
@@ -11,5 +11,12 @@
       <span nevow:render="langs"/>
     </p>
     <p nevow:render="i18n">Good morning</p>
+
+    <p>
+      We'll repeat a translatable string, just to
+      trigger bugs:
+      <span nevow:render="i18n">Greeting</span>
+    </p>
+
   </body>
 </html>
