diff -r eb14b0918a10 python/log.py
--- a/python/log.py	Sat Aug 05 04:02:16 2006 +0200
+++ b/python/log.py	Tue Aug 08 19:07:10 2006 +0200
@@ -310,6 +310,7 @@ class StdioOnnaStick:
     def __init__(self, isError=0):
         self.isError = isError
         self.buf = ''
+        self.encoding = sys.stdout.encoding
 
     def close(self):
         pass
@@ -329,6 +330,7 @@ class StdioOnnaStick:
     tell = read
 
     def write(self, data):
+        data = data.encode(self.encoding)
         d = (self.buf + data).split('\n')
         self.buf = d[-1]
         messages = d[0:-1]
@@ -337,6 +339,7 @@ class StdioOnnaStick:
 
     def writelines(self, lines):
         for line in lines:
+            line = line.encode(self.encoding)
             msg(line, printed=1, isError=self.isError)
 
 
