Index: twisted/web/_auth/digest.py
===================================================================
--- twisted/web/_auth/digest.py	(révision 32261)
+++ twisted/web/_auth/digest.py	(copie de travail)
@@ -11,6 +11,7 @@
 from zope.interface import implements
 from twisted.cred import credentials
 from twisted.web.iweb import ICredentialFactory
+from twisted.web import http
 
 class DigestCredentialFactory(object):
     """
@@ -28,6 +29,14 @@
         self.digest = credentials.DigestCredentialFactory(algorithm,
                                                           authenticationRealm)
 
+    def checkURI(self,request):
+        """
+        Check that the resource designated by the "uri" directive is the 
+        same as the resource specified in the Request-Line, as asked by
+        RFC 2617
+        """
+        if request.getHeader('digest-uri') != request.uri :
+            request.setResponseCode(http.BAD_REQUEST)
 
     def getChallenge(self, request):
         """
@@ -39,6 +48,7 @@
         @return: The C{dict} that can be used to generate a WWW-Authenticate
             header.
         """
+        self.checkURI(request)
         return self.digest.getChallenge(request.getClientIP())
 
 
@@ -49,6 +59,7 @@
 
         @see: L{ICredentialFactory.decode}
         """
+        self.checkURI(request)
         return self.digest.decode(response,
                                   request.method,
                                   request.getClientIP())
