diff --git twisted/conch/client/knownhosts.py twisted/conch/client/knownhosts.py
index 614cdaa..edb222b 100644
--- twisted/conch/client/knownhosts.py
+++ twisted/conch/client/knownhosts.py
@@ -11,11 +11,11 @@ An implementation of the OpenSSH known_hosts database.
 from binascii import Error as DecodeError, b2a_base64
 import hmac
 import sys
+from hashlib import sha1
 
 from zope.interface import implements
 
 from twisted.python.randbytes import secureRandom
-from twisted.python.hashlib import sha1
 from twisted.internet import defer
 from twisted.python import log
 from twisted.python.util import FancyEqMixin
diff --git twisted/conch/ssh/keys.py twisted/conch/ssh/keys.py
index 1ddb6cf..6a288fa 100644
--- twisted/conch/ssh/keys.py
+++ twisted/conch/ssh/keys.py
@@ -11,6 +11,7 @@ Maintainer: U{Paul Swartz}
 # base library imports
 import base64
 import itertools
+from hashlib import md5, sha1
 
 # external library imports
 from Crypto.Cipher import DES3, AES
@@ -22,7 +23,6 @@ from pyasn1.codec.ber import encoder as berEncoder
 
 # twisted
 from twisted.python import randbytes
-from twisted.python.hashlib import md5, sha1
 
 # sibling imports
 from twisted.conch.ssh import common, sexpy
diff --git twisted/conch/ssh/transport.py twisted/conch/ssh/transport.py
index 9e0c753..11ce003 100644
--- twisted/conch/ssh/transport.py
+++ twisted/conch/ssh/transport.py
@@ -14,6 +14,7 @@ Maintainer: Paul Swartz
 import struct
 import zlib
 import array
+from hashlib import md5, sha1
 
 # external library imports
 from Crypto import Util
@@ -24,7 +25,6 @@ from twisted.internet import protocol, defer
 
 from twisted.conch import error
 from twisted.python import log, randbytes
-from twisted.python.hashlib import md5, sha1
 
 
 # sibling imports
diff --git twisted/conch/test/test_keys.py twisted/conch/test/test_keys.py
index 8403e1e..6d5cba3 100644
--- twisted/conch/test/test_keys.py
+++ twisted/conch/test/test_keys.py
@@ -20,9 +20,9 @@ if Crypto and pyasn1:
     from twisted.conch.ssh import keys, common, sexpy
 
 import os, base64
+from hashlib import sha1
 from twisted.conch.test import keydata
 from twisted.python import randbytes
-from twisted.python.hashlib import sha1
 from twisted.trial import unittest
 
 
diff --git twisted/conch/test/test_transport.py twisted/conch/test/test_transport.py
index 8b801b0..88baa97 100644
--- twisted/conch/test/test_transport.py
+++ twisted/conch/test/test_transport.py
@@ -33,12 +33,13 @@ else:
         class SSHFactory:
             pass
 
+from hashlib import md5, sha1
+
 from twisted.trial import unittest
 from twisted.internet import defer
 from twisted.protocols import loopback
 from twisted.python import randbytes
 from twisted.python.reflect import qual, getClass
-from twisted.python.hashlib import md5, sha1
 from twisted.conch.ssh import address, service, common
 from twisted.test import proto_helpers
 
diff --git twisted/cred/_digest.py twisted/cred/_digest.py
index 4640a1d..8257234 100644
--- twisted/cred/_digest.py
+++ twisted/cred/_digest.py
@@ -8,7 +8,7 @@ Calculations for HTTP Digest authentication.
 @see: U{http://www.faqs.org/rfcs/rfc2617.html}
 """
 
-from twisted.python.hashlib import md5, sha1
+from hashlib import md5, sha1
 
 
 
diff --git twisted/cred/credentials.py twisted/cred/credentials.py
index 63fb44e..b87cc4d 100644
--- twisted/cred/credentials.py
+++ twisted/cred/credentials.py
@@ -7,7 +7,7 @@
 from zope.interface import implements, Interface
 
 import hmac, time, random
-from twisted.python.hashlib import md5
+from hashlib import md5
 from twisted.python.randbytes import secureRandom
 from twisted.cred._digest import calcResponse, calcHA1, calcHA2
 from twisted.cred import error
diff --git twisted/internet/test/test_unix.py twisted/internet/test/test_unix.py
index ecfad0b..ff4dd78 100644
--- twisted/internet/test/test_unix.py
+++ twisted/internet/test/test_unix.py
@@ -10,6 +10,7 @@ from os import stat, close
 from tempfile import mktemp
 from socket import AF_INET, SOCK_STREAM, socket
 from pprint import pformat
+from hashlib import md5
 
 try:
     from socket import AF_UNIX
@@ -20,7 +21,6 @@ from zope.interface import implements
 
 from twisted.python.log import addObserver, removeObserver, err
 from twisted.python.failure import Failure
-from twisted.python.hashlib import md5
 from twisted.python.runtime import platform
 from twisted.internet.interfaces import IFileDescriptorReceiver, IReactorUNIX
 from twisted.internet.error import ConnectionClosed, FileDescriptorOverrun
diff --git twisted/mail/maildir.py twisted/mail/maildir.py
index 31e9e0d..fa72125 100644
--- twisted/mail/maildir.py
+++ twisted/mail/maildir.py
@@ -10,6 +10,7 @@ Maildir-style mailbox support
 import os
 import stat
 import socket
+from hashlib import md5
 
 from zope.interface import implementer
 
@@ -23,7 +24,6 @@ from twisted.mail import smtp
 from twisted.protocols import basic
 from twisted.persisted import dirdbm
 from twisted.python import log, failure
-from twisted.python.hashlib import md5
 from twisted.mail import mail
 from twisted.internet import interfaces, defer, reactor
 from twisted.cred import portal, credentials, checkers
diff --git twisted/mail/pop3.py twisted/mail/pop3.py
index 3b65242..54b9ee6 100644
--- twisted/mail/pop3.py
+++ twisted/mail/pop3.py
@@ -14,6 +14,7 @@ Post-office Protocol version 3
 import base64
 import binascii
 import warnings
+from hashlib import md5
 
 from zope.interface import implements, Interface
 
@@ -24,7 +25,6 @@ from twisted.internet import task
 from twisted.internet import defer
 from twisted.internet import interfaces
 from twisted.python import log
-from twisted.python.hashlib import md5
 
 from twisted import cred
 import twisted.cred.error
diff --git twisted/mail/pop3client.py twisted/mail/pop3client.py
index fe8f497..d650532 100644
--- twisted/mail/pop3client.py
+++ twisted/mail/pop3client.py
@@ -12,9 +12,9 @@ Don't use this module directly.  Use twisted.mail.pop3 instead.
 """
 
 import re
+from hashlib import md5
 
 from twisted.python import log
-from twisted.python.hashlib import md5
 from twisted.internet import defer
 from twisted.protocols import basic
 from twisted.protocols import policies
diff --git twisted/mail/test/test_mail.py twisted/mail/test/test_mail.py
index 72ed2d2..87b1e79 100644
--- twisted/mail/test/test_mail.py
+++ twisted/mail/test/test_mail.py
@@ -13,6 +13,7 @@ import StringIO
 import rfc822
 import tempfile
 import signal
+from hashlib import md5
 
 from zope.interface.verify import verifyClass
 from zope.interface import Interface, implements
@@ -32,7 +33,6 @@ from twisted.internet.error import ProcessDone, ProcessTerminated
 from twisted.internet import address
 from twisted.python import failure
 from twisted.python.filepath import FilePath
-from twisted.python.hashlib import md5
 
 from twisted import mail
 import twisted.mail.mail
diff --git twisted/news/database.py twisted/news/database.py
index 137736a..dda5fbe 100644
--- twisted/news/database.py
+++ twisted/news/database.py
@@ -9,6 +9,7 @@ News server backend implementations.
 import getpass, pickle, time, socket
 import os
 import StringIO
+from hashlib import md5
 from email.Message import Message
 from email.Generator import Generator
 from zope.interface import implements, Interface
@@ -18,7 +19,6 @@ from twisted.mail import smtp
 from twisted.internet import defer
 from twisted.enterprise import adbapi
 from twisted.persisted import dirdbm
-from twisted.python.hashlib import md5
 
 
 
diff --git twisted/persisted/sob.py twisted/persisted/sob.py
index 2ba2e49..89d9caa 100644
--- twisted/persisted/sob.py
+++ twisted/persisted/sob.py
@@ -18,8 +18,8 @@ try:
     import cStringIO as StringIO
 except ImportError:
     import StringIO
+from hashlib import md5
 from twisted.python import log, runtime
-from twisted.python.hashlib import md5
 from twisted.persisted import styles
 from zope.interface import implements, Interface
 
diff --git twisted/protocols/sip.py twisted/protocols/sip.py
index 8a3f05c..8b71721 100644
--- twisted/protocols/sip.py
+++ twisted/protocols/sip.py
@@ -16,13 +16,13 @@ See L{twisted.cred.credentials} and L{twisted.cred._digest} for its new home.
 
 # system imports
 import socket, time, sys, random, warnings
+from hashlib import md5
 from zope.interface import implements, Interface
 
 # twisted imports
 from twisted.python import log, util
 from twisted.python.deprecate import deprecated
 from twisted.python.versions import Version
-from twisted.python.hashlib import md5
 from twisted.internet import protocol, defer, reactor
 
 from twisted import cred
diff --git twisted/python/__init__.py twisted/python/__init__.py
index ae78c7b..199e496 100644
--- twisted/python/__init__.py
+++ twisted/python/__init__.py
@@ -9,5 +9,3 @@ Twisted Python: Utilities and Enhancements for Python.
 
 """
 
-    
-
diff --git twisted/python/hashlib.py twisted/python/hashlib.py
index f3ee0fe..4b358e8 100644
--- twisted/python/hashlib.py
+++ twisted/python/hashlib.py
@@ -3,22 +3,21 @@
 # See LICENSE for details.
 
 """
+Deprecated in Twisted 12.3.1; please use hashlib from stdlib instead.
+
 L{twisted.python.hashlib} presents a subset of the interface provided by
 U{hashlib<http://docs.python.org/library/hashlib.html>}.  The subset is the
 interface required by various parts of Twisted.  This allows application code
 to transparently use APIs which existed before C{hashlib} was introduced or to
 use C{hashlib} if it is available.
 """
-
-
-try:
-    _hashlib = __import__("hashlib")
-except ImportError:
-    from md5 import md5
-    from sha import sha as sha1
-else:
-    md5  = _hashlib.md5
-    sha1 = _hashlib.sha1
-
+from __future__ import absolute_import
+from hashlib import md5, sha1
+import warnings
 
 __all__ = ["md5", "sha1"]
+
+warnings.warn(
+    "twisted.python.hashlib was deprecated in "
+    "Twisted 12.3.1: Please use hashlib from stdlib.",
+    DeprecationWarning, stacklevel=2)
diff --git twisted/python/test/test_hashlib.py twisted/python/test/test_hashlib.py
index b50997c..2c0814b 100644
--- twisted/python/test/test_hashlib.py
+++ twisted/python/test/test_hashlib.py
@@ -4,21 +4,36 @@
 """
 Tests for L{twisted.python.hashlib}
 """
-
 from twisted.trial.unittest import TestCase
+from twisted.trial import util
 
-from twisted.python.hashlib import md5, sha1
 
 
 class HashObjectTests(TestCase):
     """
     Tests for the hash object APIs presented by L{hashlib}, C{md5} and C{sha1}.
     """
+    def test_deprecation(self):
+        """
+        Test to ensure deprecation of L{twisted.python.hashlib}
+        """
+        from twisted.python import hashlib
+        warnings = self.flushWarnings(
+                offendingFunctions=[self.test_deprecation])
+        self.assertIdentical(warnings[0]['category'], DeprecationWarning)
+        self.assertEqual(len(warnings), 1)
+        self.assertEqual(warnings[0]['message'],
+                "twisted.python.hashlib was deprecated in "
+                "Twisted 12.3.1: Please use hashlib from stdlib.")
+
+
     def test_md5(self):
         """
         L{hashlib.md5} returns an object which can be used to compute an MD5
         hash as defined by U{RFC 1321<http://www.ietf.org/rfc/rfc1321.txt>}.
         """
+        from twisted.python.hashlib import md5
+
         # Test the result using values from section A.5 of the RFC.
         self.assertEqual(
             md5().hexdigest(), "d41d8cd98f00b204e9800998ecf8427e")
@@ -53,6 +68,8 @@ class HashObjectTests(TestCase):
 
         # Instances of it should have a digest_size attribute
         self.assertEqual(md5().digest_size, 16)
+    test_md5.suppress = [util.suppress(message="twisted.python.hashlib"
+          "was deprecated in Twisted 12.3.1: Please use hashlib from stdlib.")]
 
 
     def test_sha1(self):
@@ -60,6 +77,9 @@ class HashObjectTests(TestCase):
         L{hashlib.sha1} returns an object which can be used to compute a SHA1
         hash as defined by U{RFC 3174<http://tools.ietf.org/rfc/rfc3174.txt>}.
         """
+
+        from twisted.python.hashlib import sha1
+
         def format(s):
             return ''.join(s.split()).lower()
         # Test the result using values from section 7.3 of the RFC.
@@ -88,3 +108,5 @@ class HashObjectTests(TestCase):
         # Instances of it should have a digest_size attribute.
         self.assertEqual(
             sha1().digest_size, 20)
+    test_sha1.suppress = [util.suppress(message="twisted.python.hashlib"
+          "was deprecated in Twisted 12.3.1: Please use hashlib from stdlib.")]
diff --git twisted/python/test/test_zipstream.py twisted/python/test/test_zipstream.py
index 8871563..ff8ca88 100644
--- twisted/python/test/test_zipstream.py
+++ twisted/python/test/test_zipstream.py
@@ -8,9 +8,9 @@ Tests for L{twisted.python.zipstream}
 import sys
 import random
 import zipfile
+from hashlib import md5
 
 from twisted.python import zipstream, filepath
-from twisted.python.hashlib import md5
 from twisted.trial import unittest
 
 
diff --git twisted/spread/pb.py twisted/spread/pb.py
index 7e9a5b6..99b5316 100644
--- twisted/spread/pb.py
+++ twisted/spread/pb.py
@@ -29,12 +29,12 @@ To get started, begin with L{PBClientFactory} and L{PBServerFactory}.
 
 import random
 import types
+from hashlib import md5
 
 from zope.interface import implements, Interface
 
 # Twisted Imports
 from twisted.python import log, failure, reflect
-from twisted.python.hashlib import md5
 from twisted.internet import defer, protocol
 from twisted.cred.portal import Portal
 from twisted.cred.credentials import IAnonymous, ICredentials
diff --git twisted/test/test_digestauth.py twisted/test/test_digestauth.py
index 41368a0..363927d 100644
--- twisted/test/test_digestauth.py
+++ twisted/test/test_digestauth.py
@@ -5,10 +5,10 @@
 Tests for L{twisted.cred._digest} and the associated bits in
 L{twisted.cred.credentials}.
 """
+from hashlib import md5, sha1
 
 from zope.interface.verify import verifyObject
 from twisted.trial.unittest import TestCase
-from twisted.python.hashlib import md5, sha1
 from twisted.internet.address import IPv4Address
 from twisted.cred.error import LoginFailed
 from twisted.cred.credentials import calcHA1, calcHA2, IUsernameDigestHash
diff --git twisted/topfiles/6342.removal twisted/topfiles/6342.removal
new file mode 100644
index 0000000..057d71f
--- /dev/null
+++ twisted/topfiles/6342.removal
@@ -0,0 +1 @@
+twisted.python.hashlib is now deprecated in favor of hashlib from stdlib.
diff --git twisted/trial/test/test_loader.py twisted/trial/test/test_loader.py
index a49fa6e..937fea2 100644
--- twisted/trial/test/test_loader.py
+++ twisted/trial/test/test_loader.py
@@ -9,9 +9,9 @@ import os
 import sys
 
 import unittest as pyunit
+from hashlib import md5
 
 from twisted.python import util, filepath
-from twisted.python.hashlib import md5
 from twisted.trial.test import packages
 from twisted.trial import runner, reporter, unittest
 from twisted.trial.itrial import ITestCase
diff --git twisted/web/server.py twisted/web/server.py
index bc3f07c..7a5f59d 100644
--- twisted/web/server.py
+++ twisted/web/server.py
@@ -660,7 +660,7 @@ class Site(http.HTTPFactory):
         """
         (internal) Generate an opaque, unique ID for a user's session.
         """
-        from twisted.python.hashlib import md5
+        from hashlib import md5
         import random
         self.counter = self.counter + 1
         return md5(networkString(
diff --git twisted/words/protocols/jabber/sasl_mechanisms.py twisted/words/protocols/jabber/sasl_mechanisms.py
index 5d51be2..186eca5 100644
--- twisted/words/protocols/jabber/sasl_mechanisms.py
+++ twisted/words/protocols/jabber/sasl_mechanisms.py
@@ -8,10 +8,10 @@ Protocol agnostic implementations of SASL authentication mechanisms.
 """
 
 import binascii, random, time, os
+from hashlib import md5
 
 from zope.interface import Interface, Attribute, implements
 
-from twisted.python.hashlib import md5
 
 class ISASLMechanism(Interface):
     name = Attribute("""Common name for the SASL Mechanism.""")
diff --git twisted/words/protocols/jabber/xmlstream.py twisted/words/protocols/jabber/xmlstream.py
index cc2745b..e30508b 100644
--- twisted/words/protocols/jabber/xmlstream.py
+++ twisted/words/protocols/jabber/xmlstream.py
@@ -11,12 +11,12 @@ doing authentication on either client or server side, and working with XML
 Stanzas.
 """
 
+from hashlib import sha1
 from zope.interface import directlyProvides, implements
 
 from twisted.internet import defer, protocol
 from twisted.internet.error import ConnectionLost
 from twisted.python import failure, log, randbytes
-from twisted.python.hashlib import sha1
 from twisted.words.protocols.jabber import error, ijabber, jid
 from twisted.words.xish import domish, xmlstream
 from twisted.words.xish.xmlstream import STREAM_CONNECTED_EVENT
diff --git twisted/words/protocols/msn.py twisted/words/protocols/msn.py
index 79c0fa1..f9dc854 100644
--- twisted/words/protocols/msn.py
+++ twisted/words/protocols/msn.py
@@ -74,9 +74,9 @@ TODO
 import types, operator, os
 from random import randint
 from urllib import quote, unquote
+from hashlib import md5
 
 from twisted.python import failure, log
-from twisted.python.hashlib import md5
 from twisted.internet import reactor
 from twisted.internet.defer import Deferred, execute
 from twisted.internet.protocol import ClientFactory
diff --git twisted/words/protocols/oscar.py twisted/words/protocols/oscar.py
index 81571d4..14908c6 100644
--- twisted/words/protocols/oscar.py
+++ twisted/words/protocols/oscar.py
@@ -15,10 +15,10 @@ import socket
 import random
 import types
 import re
+from hashlib import md5
 
 from twisted.internet import reactor, defer, protocol
 from twisted.python import log
-from twisted.python.hashlib import md5
 
 def logPacketData(data):
     lines = len(data)/16
diff --git twisted/words/test/test_jabberclient.py twisted/words/test/test_jabberclient.py
index 87af883..dd536ef 100644
--- twisted/words/test/test_jabberclient.py
+++ twisted/words/test/test_jabberclient.py
@@ -4,9 +4,9 @@
 """
 Tests for L{twisted.words.protocols.jabber.client}
 """
+from hashlib import sha1
 
 from twisted.internet import defer
-from twisted.python.hashlib import sha1
 from twisted.trial import unittest
 from twisted.words.protocols.jabber import client, error, jid, xmlstream
 from twisted.words.protocols.jabber.sasl import SASLInitiatingInitializer
diff --git twisted/words/test/test_jabbercomponent.py twisted/words/test/test_jabbercomponent.py
index d8bb108..91e107b 100644
--- twisted/words/test/test_jabbercomponent.py
+++ twisted/words/test/test_jabbercomponent.py
@@ -4,9 +4,9 @@
 """
 Tests for L{twisted.words.protocols.jabber.component}
 """
+from hashlib import sha1
 
 from twisted.python import failure
-from twisted.python.hashlib import sha1
 from twisted.trial import unittest
 from twisted.words.protocols.jabber import component, xmlstream
 from twisted.words.protocols.jabber.jid import JID
diff --git twisted/words/test/test_msn.py twisted/words/test/test_msn.py
index ece580f..669c9aa 100644
--- twisted/words/test/test_msn.py
+++ twisted/words/test/test_msn.py
@@ -7,6 +7,7 @@ Test cases for L{twisted.words.protocols.msn}.
 
 # System imports
 import StringIO
+from hashlib import md5
 
 # Twisted imports
 
@@ -24,7 +25,6 @@ else:
     from twisted.words.protocols import msn
 
 
-from twisted.python.hashlib import md5
 from twisted.protocols import loopback
 from twisted.internet.defer import Deferred
 from twisted.trial import unittest
