--- Twisted-10.0.0/twisted/python/logfile.py	2010-02-21 15:57:49.000000000 -0700
+++ logfile_datestamp.py	2010-03-17 19:02:42.000000000 -0600
@@ -198,15 +198,9 @@
         """
         if not (os.access(self.directory, os.W_OK) and os.access(self.path, os.W_OK)):
             return
-        logs = self.listLogs()
-        logs.reverse()
-        for i in logs:
-            if self.maxRotatedFiles is not None and i >= self.maxRotatedFiles:
-                os.remove("%s.%d" % (self.path, i))
-            else:
-                os.rename("%s.%d" % (self.path, i), "%s.%d" % (self.path, i + 1))
+        
         self._file.close()
-        os.rename(self.path, "%s.1" % self.path)
+        os.rename(self.path, "%s.%d.%d" % (self.path,os.getpid(),time.time()))
         self._openFile()
 
     def listLogs(self):
