Index: twisted/web2/http_headers.py
===================================================================
--- twisted/web2/http_headers.py	(revision 28048)
+++ twisted/web2/http_headers.py	(working copy)
@@ -71,7 +71,7 @@
         """
         parser = self.HTTPParsers.get(name, None)
         if parser is None:
-            raise ValueError("No header parser for header '%s', either add one or use getHeaderRaw." % (name,))
+            raise ValueError("No header parser for header '%s', either add one or use getRawHeaders." % (name,))
 
         try:
             for p in parser:
@@ -101,7 +101,7 @@
 
         if generator is None:
             # print self.generators
-            raise ValueError("No header generator for header '%s', either add one or use setHeaderRaw." % (name,))
+            raise ValueError("No header generator for header '%s', either add one or use setRawHeaders." % (name,))
 
         for g in generator:
             header = g(header)
