diff --git twisted/conch/ssh/factory.py twisted/conch/ssh/factory.py
index 6f8f104..934db68 100644
--- twisted/conch/ssh/factory.py
+++ twisted/conch/ssh/factory.py
@@ -59,7 +59,7 @@ class SSHFactory(protocol.Factory):
                     'because we cannot find moduli file')
             t.supportedKeyExchanges = [
                 kexAlgorithm for kexAlgorithm in t.supportedKeyExchanges
-                if _kex.isFixedGroup(kexAlgorithm)]
+                if _kex.isFixedGroup(kexAlgorithm) or (hasattr(_kex, "isEllipticCurve") and _kex.isEllipticCurve(kexAlgorithm))]
         return t
 
 
