Index: doc/core/examples/cursesclient.py
===================================================================
--- doc/core/examples/cursesclient.py	(revision 11659)
+++ doc/core/examples/cursesclient.py	(working copy)
@@ -45,8 +45,6 @@
     """ A protocol object for IRC """
 
     def __init__(self, screenObj):
-        self.nick = 'testcurses'
-
         # screenObj should be 'stdscr' or a curses window/pad object
         self.screenObj = screenObj
         # for testing (hacky way around initial bad design for this example) :)
@@ -57,6 +55,8 @@
         self.screenObj.addLine(line)
 
     def connectionMade(self):
+        self.nick = "testcurses"
+        IRCClient.connectionMade(self)
         self.screenObj.addLine("* CONNECTED")
 
     def clientConnectionLost(self, connection, reason):
Index: doc/words/examples/ircLogBot.py
===================================================================
--- doc/words/examples/ircLogBot.py	(revision 11659)
+++ doc/words/examples/ircLogBot.py	(working copy)
@@ -52,11 +52,9 @@
 class LogBot(irc.IRCClient):
     """A logging IRC bot."""
     
-    def __init__(self):
-        self.nickname = "twistedbot"
-
     def connectionMade(self):
         irc.IRCClient.connectionMade(self)
+        self.nickname = "twistedbot"
         self.logger = MessageLogger(open(self.factory.filename, "a"))
         self.logger.log("[connected at %s]" % 
                         time.asctime(time.localtime(time.time())))
