Only in .: callFromThread.diff
diff -ru ./internet/base.py ..\twisted/internet/base.py
--- ./internet/base.py	2006-01-06 18:18:14.000000000 -0800
+++ ..\twisted/internet/base.py	2006-05-22 14:22:46.761580400 -0700
@@ -513,13 +513,20 @@
             # making them, in case another call is added to the queue
             # while we're in this loop.
             count = 0
+            total = len(self.threadCallQueue)
             for (f, a, kw) in self.threadCallQueue:
                 try:
                     f(*a, **kw)
                 except:
                     log.err()
                 count += 1
+                if count == total:
+                    break
             del self.threadCallQueue[:count]
+            if len(self.threadCallQueue) > 0:
+                if self.waker:
+                    self.waker.wakeUp()
+
 
         # insert new delayed calls now
         self._insertNewDelayedCalls()
