Index: twisted/python/filepath.py
===================================================================
--- twisted/python/filepath.py	(revision 28871)
+++ twisted/python/filepath.py	(working copy)
@@ -663,8 +663,11 @@
     def parent(self):
         return self.clonePath(self.dirname())
 
-    def setContent(self, content, ext='.new'):
-        sib = self.siblingExtension(ext)
+    def setContent(self, content, ext=None):
+        if ext is None:
+            sib = self.temporarySibling()
+        else:
+            sib = self.siblingExtension(ext)
         f = sib.open('w')
         try:
             f.write(content)
