Index: docs/projects/historic/2003/pycon/deferex/deferex-listing0.py
===================================================================
--- docs/projects/historic/2003/pycon/deferex/deferex-listing0.py	(revision 41316)
+++ docs/projects/historic/2003/pycon/deferex/deferex-listing0.py	(working copy)
@@ -5,7 +5,8 @@
 
     def loadDocuments(callback, server, keyword):
         "Retrieve a set of documents!"
-        ...
+        # ...
+        pass
 
     def callback(self, documents):
         try:
@@ -15,4 +16,5 @@
             self.cleanup()
 
     def cleanup(self):
-        ...
+        # ...
+        pass
Index: docs/projects/historic/2003/pycon/deferex/deferex-bad-adding.py
===================================================================
--- docs/projects/historic/2003/pycon/deferex/deferex-bad-adding.py	(revision 41316)
+++ docs/projects/historic/2003/pycon/deferex/deferex-bad-adding.py	(working copy)
@@ -3,6 +3,6 @@
     print myResult
     return myResult
 
-...
+# ...
 
 adder.callRemote("add", 1, 1).addCallback(successCallback)
Index: docs/projects/historic/2003/pycon/deferex/deferex-listing2.py
===================================================================
--- docs/projects/historic/2003/pycon/deferex/deferex-listing2.py	(revision 41316)
+++ docs/projects/historic/2003/pycon/deferex/deferex-listing2.py	(working copy)
@@ -3,6 +3,6 @@
     print myResult
     return myResult
 
-...
+# ...
 
 adder.callRemote("add", 1, 1).addCallback(successCallback)
