Index: twisted/topfiles/6483.misc
===================================================================
--- twisted/topfiles/6483.misc	(revision 0)
+++ twisted/topfiles/6483.misc	(working copy)
@@ -0,0 +1 @@
+
Index: twisted/internet/iocpreactor/reactor.py
===================================================================
--- twisted/internet/iocpreactor/reactor.py	(revision 38280)
+++ twisted/internet/iocpreactor/reactor.py	(working copy)
@@ -149,18 +149,12 @@
 
 
     def listenTCP(self, port, factory, backlog=50, interface=''):
-        """
-        @see: twisted.internet.interfaces.IReactorTCP.listenTCP
-        """
         p = tcp.Port(port, factory, backlog, interface, self)
         p.startListening()
         return p
 
 
     def connectTCP(self, host, port, factory, timeout=30, bindAddress=None):
-        """
-        @see: twisted.internet.interfaces.IReactorTCP.connectTCP
-        """
         c = tcp.Connector(host, port, factory, timeout, bindAddress, self)
         c.connect()
         return c
@@ -168,9 +162,6 @@
 
     if TLSMemoryBIOFactory is not None:
         def listenSSL(self, port, factory, contextFactory, backlog=50, interface=''):
-            """
-            @see: twisted.internet.interfaces.IReactorSSL.listenSSL
-            """
             port = self.listenTCP(
                 port,
                 TLSMemoryBIOFactory(contextFactory, False, factory),
@@ -180,9 +171,6 @@
 
 
         def connectSSL(self, host, port, factory, contextFactory, timeout=30, bindAddress=None):
-            """
-            @see: twisted.internet.interfaces.IReactorSSL.connectSSL
-            """
             return self.connectTCP(
                 host, port,
                 TLSMemoryBIOFactory(contextFactory, True, factory),
