Index: doc/web/howto/client.xhtml
===================================================================
--- doc/web/howto/client.xhtml	(revision 32147)
+++ doc/web/howto/client.xhtml	(working copy)
@@ -289,14 +289,14 @@
     <h3>Using a HTTP proxy</h3>
 
     <p>
-      To be able to use HTTP proxies with an agent, you can use <code
+      To be able to use HTTP proxies with an agent, you can use the <code
       class="API">twisted.web.client.ProxyAgent</code> class. It supports the
       same interface as <code>Agent</code>, but takes the endpoint of the proxy
-      as argument.
+      as initializer argument.
     </p>
 
     <p>
-      Here's an example to use a HTTP proxy, this proxy running on
+      Here's an example demonstrating the use of an HTTP proxy running on
       localhost:8000.
     </p>
 
@@ -320,12 +320,14 @@
 
 if __name__ == "__main__":
     main()
-    </pre>
+</pre>
 
     <p>
       Please refer to the <a
-      href="../../core/howto/endpoints.xhtml">Endpoints</a> documentation for
-      more information.
+      href="../../core/howto/endpoints.xhtml">endpoints documentation</a> for
+      more information about how they work and the <code
+      class="API">twisted.internet.endpoints</code> API documentation to learn
+      what other kinds of endpoints exist.
     </p>
 
     <h2>
