=== modified file 'twisted/internet/threads.py'
--- twisted/internet/threads.py	2009-01-22 13:01:46 +0000
+++ twisted/internet/threads.py	2010-02-09 15:24:45 +0000
@@ -113,5 +113,6 @@
     return result
 
 
-__all__ = ["deferToThread", "callMultipleInThread", "blockingCallFromThread"]
+__all__ = ["deferToThread", "deferToThreadPool", "callMultipleInThread",
+           "blockingCallFromThread"]
 

