Index: twisted/test/test_ftp.py
===================================================================
--- twisted/test/test_ftp.py	(revision 30600)
+++ twisted/test/test_ftp.py	(working copy)
@@ -267,6 +267,14 @@
             chainDeferred=d)
         return d
 
+    def testAnonymousWriteDenied(self):
+        # Anonymous users should not be allowed to edit server-side filesystem
+        d = self._anonymousLogin()
+        return self.assertCommandFailed(
+              'MKD newdir',
+              ['550 Anonymous users are forbidden to change the filesystem'],
+              chainDeferred=d)
+
     def testUnknownCommand(self):
         d = self._anonymousLogin()
         return self.assertCommandFailed(
Index: twisted/protocols/ftp.py
===================================================================
--- twisted/protocols/ftp.py	(revision 30600)
+++ twisted/protocols/ftp.py	(working copy)
@@ -254,7 +254,7 @@
     """
     def __init__(self):
         # No message
-        FTPCmdError.__init__(self, None)
+        FTPCmdError.__init__(self)
 
     errorCode = ANON_USER_DENIED
 
