diff --git twisted/news/test/test_news.py twisted/news/test/test_news.py
index 35ac7d7..8d1ca6c 100644
--- twisted/news/test/test_news.py
+++ twisted/news/test/test_news.py
@@ -1,12 +1,10 @@
 # Copyright (c) Twisted Matrix Laboratories.
 # See LICENSE for details.
 
-import sys, types
 from pprint import pformat
 
 from twisted.trial import unittest
 from twisted.news import database
-from twisted.internet import reactor
 
 MESSAGE_ID = "f83ba57450ed0fd8ac9a472b847e830e"
 
