Index: doc/web/howto/using-twistedweb.xhtml
===================================================================
--- doc/web/howto/using-twistedweb.xhtml	(revision 35287)
+++ doc/web/howto/using-twistedweb.xhtml	(working copy)
@@ -649,7 +649,7 @@
 
 <h3>Serving WSGI Applications</h3>
 
-<p><a href="http://wsgi.org/wsgi">WSGI</a> is the Web Server Gateway
+<p><a href="http://wsgi.readthedocs.org/en/latest/index.html">WSGI</a> is the Web Server Gateway
 Interface. It is a specification for web servers and application servers to
 communicate with Python web applications. All modern Python web frameworks
 support the WSGI interface.</p>
Index: doc/web/howto/web-in-60/wsgi.xhtml
===================================================================
--- doc/web/howto/web-in-60/wsgi.xhtml	(revision 35287)
+++ doc/web/howto/web-in-60/wsgi.xhtml	(working copy)
@@ -45,7 +45,7 @@
 </pre>
 
 <p>If this doesn't make sense to you, take a look at one of
-these <a href="http://wsgi.org/wsgi/Learn_WSGI">fine tutorials</a>. Otherwise,
+these <a href="http://wsgi.readthedocs.org/en/latest/learn.html">fine tutorials</a>. Otherwise,
 or once you're done with that, the next step is to create
 a <code>WSGIResource</code> instance, as this is going to be
 another <a href="rpy-scripts.xhtml">rpy script</a> example:</p>
