diff --git twisted/python/util.py twisted/python/util.py
index ee4cc8a..2c7529f 100644
--- twisted/python/util.py
+++ twisted/python/util.py
@@ -561,11 +561,13 @@ class FancyEqMixin:
         return not result
 
 
-
 try:
-    from twisted.python._initgroups import initgroups as _c_initgroups
-except ImportError:
-    _c_initgroups = None
+    from os import initgroups as _c_initgroups
+except ImportError: # Python 2.6 or lower
+    try:
+        from twisted.python._initgroups import initgroups as _c_initgroups
+    except ImportError:
+        _c_initgroups = None
 
 
 
