Index: twisted/conch/test/test_ckeygen.py
===================================================================
--- twisted/conch/test/test_ckeygen.py	(revision 40352)
+++ twisted/conch/test/test_ckeygen.py	(working copy)
@@ -68,8 +68,24 @@
         self.assertEqual(
             self.stdout.getvalue(),
             '768 3d:13:5f:cb:c9:79:8a:93:06:27:65:bc:3d:0b:8f:af temp\n')
+    
 
+    def test_printFingerprintDefultKeyfile(self):
+        """
+        L{printFingerprint} should provide a default keyfile if nothing is
+        specified.
+        """
+        def fake_input(str):
+            return ''
+        filename = self.mktemp()
+        FilePath(filename).setContent(publicRSA_openssh)
+        printFingerprint(
+                {'filename':''}, default_file=filename, _raw_input=fake_input)
+        self.assertEqual(
+            self.stdout.getvalue(),
+            '768 3d:13:5f:cb:c9:79:8a:93:06:27:65:bc:3d:0b:8f:af temp\n')
 
+
     def test_saveKey(self):
         """
         L{_saveKey} writes the private and public parts of a key to two
@@ -131,6 +147,24 @@
             pubKey.toString('openssh'))
 
 
+    def test_displayPublicKeyDefaultKeyfile(self):
+        """
+        L{displayPublicKey} should provide a default keyfile if nothing is 
+        specified.
+        """
+        def fake_input(str):
+            return ''
+        filename = self.mktemp()
+        pubKey = Key.fromString(publicRSA_openssh)
+        FilePath(filename).setContent(privateRSA_openssh)
+        displayPublicKey(
+                    {'filename': ''}, default_file=filename, 
+                    _raw_input=fake_input)
+        self.assertEqual(
+            self.stdout.getvalue().strip('\n'),
+            pubKey.toString('openssh'))
+
+
     def test_displayPublicKeyEncrypted(self):
         """
         L{displayPublicKey} prints out the public key associated with a given
@@ -338,3 +372,24 @@
         self.assertEqual(
             'Could not change passphrase: key not encrypted', str(error))
         self.assertEqual(publicRSA_openssh, FilePath(filename).getContent())
+
+
+    def test_changePassphraseDefaultkeyfile(self):
+        """
+        L{changePassPhrase} should provide a default for keyfile if nothing
+        is specified.
+        """
+        def fake_input(str):
+            return ''
+
+        oldNewConfirm = makeGetpass('encrypted', 'newpass', 'newpass')
+        self.patch(getpass, 'getpass', oldNewConfirm)
+
+        filename = self.mktemp()
+        FilePath(filename).setContent(privateRSA_openssh_encrypted)
+
+        try:
+            changePassPhrase({'filename': ''}, default_file=filename, 
+                    _raw_input=fake_input)
+        except IOError:
+            self.fail()
Index: twisted/conch/topfiles/6607.feature
===================================================================
--- twisted/conch/topfiles/6607.feature	(revision 0)
+++ twisted/conch/topfiles/6607.feature	(revision 0)
@@ -0,0 +1 @@
+ckeygen now provides default keyfile for --fingerprint,--showpub and --changepass.
Index: twisted/conch/scripts/ckeygen.py
===================================================================
--- twisted/conch/scripts/ckeygen.py	(revision 40352)
+++ twisted/conch/scripts/ckeygen.py	(working copy)
@@ -99,10 +99,18 @@
 
 
 
-def printFingerprint(options):
+def _getFilename(default_file, _raw_input):
+    filename = os.path.expanduser(default_file)
+    filename = _raw_input(
+                'Enter file in which the key is (%s): ' % filename
+                ).strip() or filename
+    return filename
+
+
+
+def printFingerprint(options, default_file='~/.ssh/id_rsa', _raw_input=raw_input):
     if not options['filename']:
-        filename = os.path.expanduser('~/.ssh/id_rsa')
-        options['filename'] = raw_input('Enter file in which the key is (%s): ' % filename)
+        options['filename'] = _getFilename(default_file, _raw_input)
     if os.path.exists(options['filename']+'.pub'):
         options['filename'] += '.pub'
     try:
@@ -118,11 +126,9 @@
 
 
 
-def changePassPhrase(options):
+def changePassPhrase(options, default_file='~/.ssh/id_rsa', _raw_input=raw_input):
     if not options['filename']:
-        filename = os.path.expanduser('~/.ssh/id_rsa')
-        options['filename'] = raw_input(
-            'Enter file in which the key is (%s): ' % filename)
+        options['filename'] = _getFilename(default_file, _raw_input)
     try:
         key = keys.Key.fromFile(options['filename']).keyObject
     except keys.EncryptedKeyError as e:
@@ -168,10 +174,9 @@
 
 
 
-def displayPublicKey(options):
+def displayPublicKey(options, default_file='~/.ssh/id_rsa', _raw_input=raw_input):
     if not options['filename']:
-        filename = os.path.expanduser('~/.ssh/id_rsa')
-        options['filename'] = raw_input('Enter file in which the key is (%s): ' % filename)
+        options['filename'] = _getFilename(default_file, _raw_input)
     try:
         key = keys.Key.fromFile(options['filename']).keyObject
     except keys.EncryptedKeyError:
