Index: twisted/conch/scripts/ckeygen.py
===================================================================
--- twisted/conch/scripts/ckeygen.py	(revision 40183)
+++ twisted/conch/scripts/ckeygen.py	(working copy)
@@ -102,7 +102,7 @@
 def printFingerprint(options):
     if not options['filename']:
         filename = os.path.expanduser('~/.ssh/id_rsa')
-        options['filename'] = raw_input('Enter file in which the key is (%s): ' % filename)
+        options['filename'] = raw_input('Enter file in which the key is (%s): ' % filename).strip() or filename
     if os.path.exists(options['filename']+'.pub'):
         options['filename'] += '.pub'
     try:
@@ -122,7 +122,7 @@
     if not options['filename']:
         filename = os.path.expanduser('~/.ssh/id_rsa')
         options['filename'] = raw_input(
-            'Enter file in which the key is (%s): ' % filename)
+            'Enter file in which the key is (%s): ' % filename).strip() or filename
     try:
         key = keys.Key.fromFile(options['filename']).keyObject
     except keys.EncryptedKeyError as e:
@@ -171,7 +171,7 @@
 def displayPublicKey(options):
     if not options['filename']:
         filename = os.path.expanduser('~/.ssh/id_rsa')
-        options['filename'] = raw_input('Enter file in which the key is (%s): ' % filename)
+        options['filename'] = raw_input('Enter file in which the key is (%s): ' % filename).strip() or filename
     try:
         key = keys.Key.fromFile(options['filename']).keyObject
     except keys.EncryptedKeyError:
