--- client-xmpp.py	2005-05-24 11:39:35.414188824 -0400
+++ client.py	2005-05-24 14:25:54.283170152 -0400
@@ -424,6 +424,8 @@
     def _saslSuccess(self, s):
         self.success = 1
         self.xmlstream._reset()
+        self.connectionMade()
+
 
     # TLS stuff  - maybe put this in its own class?
     def _proceed(self, p):
@@ -432,6 +434,9 @@
         ctx = ssl.ClientContextFactory()
         self.xmlstream.transport.startTLS(ctx)
         self.xmlstream._reset()
+        sh = "<stream:stream xmlns='%s' version='1.0' xmlns:stream='http://etherx.jabber.org/streams' to='%s'>" % \
+             (self.namespace, self.streamHost.encode('utf-8'))
+        self.xmlstream.send(sh)
         self.tls = 1
 
     def _tlsError(self, e):
