Index: twisted/words/__init__.py
===================================================================
--- twisted/words/__init__.py	(revision 38280)
+++ twisted/words/__init__.py	(working copy)
@@ -3,8 +3,9 @@
 # See LICENSE for details.
 
 
-"""Twisted Words: a Twisted Chat service.
 """
+Twisted Words: a Twisted Chat service.
+"""
 
 from twisted.words._version import version
 __version__ = version.short()
