Index: _dumbwin32proc.py
===================================================================
--- _dumbwin32proc.py	(revision 26940)
+++ _dumbwin32proc.py	(working copy)
@@ -74,7 +74,7 @@
 
     @return: a str representing another filename.
     """
-    f = file(filename, 'ru')
+    f = file(filename, 'rU')
     if f.read(2) == '#!':
         exe = f.readline(1024).strip('\n')
         return exe
