Index: twisted/test/test_iutils.py
===================================================================
--- twisted/test/test_iutils.py	(revision 27203)
+++ twisted/test/test_iutils.py	(working copy)
@@ -259,7 +259,28 @@
             utils.getProcessOutputAndValue, check)
 
 
+    def test_shebang(self):
+        """
+        Running a script that starts with a shebang (#!) works
+        """
+        SHEBANG_OUTPUT = 'this is the shebang output'
 
+        scriptFile = self.makeSourceFile([
+                "#!%s" % (self.exe,),
+                "import sys",
+                "sys.stdout.write('%s')" % (SHEBANG_OUTPUT,),
+                "sys.stdout.flush()"])
+        os.chmod(scriptFile, 0700)
+
+        def gotOutputAndValue((out, err, code)):
+            self.assertEquals(out, SHEBANG_OUTPUT)
+            self.assertEquals(err, "")
+            self.assertEquals(code, 0)
+        d = utils.getProcessOutputAndValue(scriptFile)
+        return d.addCallback(gotOutputAndValue)
+
+
+
 class WarningSuppression(unittest.TestCase):
     def setUp(self):
         self.warnings = []
Index: twisted/internet/_dumbwin32proc.py
===================================================================
--- twisted/internet/_dumbwin32proc.py	(revision 27203)
+++ twisted/internet/_dumbwin32proc.py	(working copy)
@@ -74,7 +74,7 @@
 
     @return: a str representing another filename.
     """
-    f = file(filename, 'ru')
+    f = file(filename, 'rU')
     if f.read(2) == '#!':
         exe = f.readline(1024).strip('\n')
         return exe
