=== modified file 'twisted/news/nntp.py'
--- old/twisted/news/nntp.py	2011-02-14 04:45:15 +0000
+++ new/twisted/news/nntp.py	2012-03-13 05:32:43 +0000
@@ -47,7 +47,7 @@
         try:
             a = int(articles[0])
             return a, a
-        except ValueError, e:
+        except ValueError:
             return None, None
     elif len(articles) == 2:
         try:
@@ -59,7 +59,7 @@
                 h = int(articles[1])
             else:
                 h = None
-        except ValueError, e:
+        except ValueError:
             return None, None
     return l, h
 
@@ -700,24 +700,6 @@
         self.sendLine('502 no permission')
 
 
-    def do_XROVER(self, header, range = None):
-        d = self.xhdrWork(header, range)
-        if d:
-            d.addCallbacks(self._gotXROVER, self._errXROVER)
-    
-    
-    def _gotXROVER(self, parts):
-        self.sendLine('224 Overview information follows')
-        for i in parts:
-            self.sendLine('%d %s' % i)
-        self.sendLine('.')
-
-
-    def _errXROVER(self, failure):
-        print 'XROVER failed: ',
-        self._errXHDR(failure)
-
-
     def do_POST(self):
         self.inputHandler = self._doingPost
         self.message = ''
@@ -819,7 +801,7 @@
                     try:
                         article = int(article)
                         return func(self.currentGroup, article) 
-                    except ValueError, e:
+                    except ValueError:
                         self.sendLine('501 command syntax error')
 
 

