Index: twisted/internet/_sslverify.py
===================================================================
--- twisted/internet/_sslverify.py	(revision 34864)
+++ twisted/internet/_sslverify.py	(working copy)
@@ -102,7 +102,7 @@
         l = []
         lablen = 0
         def uniqueValues(mapping):
-            return dict.fromkeys(mapping.itervalues()).keys()
+            return set(mapping.itervalues())
         for k in sorted(uniqueValues(_x509names)):
             label = util.nameToLabel(k)
             lablen = max(len(label), lablen)
