Index: enterprise/adbapi.py
===================================================================
--- enterprise/adbapi.py	(revision 13857)
+++ enterprise/adbapi.py	(working copy)
@@ -371,13 +371,15 @@
         conn = Connection(self)
         trans = Transaction(self, conn)
         try:
-            result = interaction(trans, *args, **kw)
+            try:
+                result = interaction(trans, *args, **kw)
+                conn.commit()
+                return result
+            except:
+                conn.rollback()
+                raise
+        finally:
             trans.close()
-            conn.commit()
-            return result
-        except:
-            conn.rollback()
-            raise
 
     def _runQuery(self, trans, *args, **kw):
         trans.execute(*args, **kw)
