*** adbapi.py~  Sat Mar 12 05:25:19 2005
--- adbapi.py   Thu May 26 12:54:34 2005
***************
*** 299,312 ****
      def _runInteraction(self, interaction, *args, **kw):
          trans = Transaction(self)
          try:
!             result = interaction(trans, *args, **kw)
              trans.close()
!             trans._connection.commit()
!             return result
!         except:
!             self._rollback(trans)
!             raise
!
      def _runQuery(self, trans, *args, **kw):
          trans.execute(*args, **kw)
          return trans.fetchall()
--- 299,314 ----
      def _runInteraction(self, interaction, *args, **kw):
          trans = Transaction(self)
          try:
!             try:
!                 result = interaction(trans, *args, **kw)
!                 trans._connection.commit()
!             except:
!                 self._rollback(trans)
!                 raise
!         finally:
              trans.close()
!         return result
!
      def _runQuery(self, trans, *args, **kw):
          trans.execute(*args, **kw)
          return trans.fetchall()
