Index: twisted/web/guard.py
===================================================================
--- twisted/web/guard.py	(revision 27096)
+++ twisted/web/guard.py	(working copy)
@@ -12,6 +12,6 @@
 from twisted.web._auth.digest import DigestCredentialFactory
 
 __all__ = [
-    HTTPAuthSessionWrapper.__name__,
+    "HTTPAuthSessionWrapper",
 
-    BasicCredentialFactory.__name__, DigestCredentialFactory.__name__]
+    "BasicCredentialFactory", "DigestCredentialFactory"]
