Index: twisted/web/client.py
===================================================================
--- twisted/web/client.py	(revision 23332)
+++ twisted/web/client.py	(working copy)
@@ -203,7 +203,7 @@
     def __init__(self, url, method='GET', postdata=None, headers=None,
                  agent="Twisted PageGetter", timeout=0, cookies=None,
                  followRedirect=1):
-        self.protocol.followRedirect = followRedirect
+        self.followRedirect = followRedirect
         self.timeout = timeout
         self.agent = agent
 
@@ -241,6 +241,7 @@
 
     def buildProtocol(self, addr):
         p = protocol.ClientFactory.buildProtocol(self, addr)
+        p.followRedirect = self.followRedirect
         if self.timeout:
             timeoutCall = reactor.callLater(self.timeout, p.timeout)
             self.deferred.addBoth(self._cancelTimeout, timeoutCall)
