Index: twisted/web/woven/widgets.py
===================================================================
RCS file: /cvs/Twisted/twisted/web/woven/widgets.py,v
retrieving revision 1.97
diff -u -r1.97 widgets.py
--- twisted/web/woven/widgets.py	26 Jul 2003 09:27:17 -0000	1.97
+++ twisted/web/woven/widgets.py	30 Jul 2003 00:40:18 -0000
@@ -638,6 +638,14 @@
 class Select(Input):
     tagName = 'select'
 
+    def setUp(self, request, node, data):
+        if node.hasChildNodes():
+            for childNode in node.childNodes:
+                if childNode.attributes.has_key("value"):
+                    if childNode.getAttribute("value") == data:
+                        childNode.setAttribute("selected", "selected")
+                    else:
+                        childNode.removeAttribute("selected")
 
 class Option(Input):
     tagName = 'option'
