Index: twisted/conch/ssh/filetransfer.py
===================================================================
--- twisted/conch/ssh/filetransfer.py	(revision 28947)
+++ twisted/conch/ssh/filetransfer.py	(working copy)
@@ -708,8 +708,11 @@
         d, data = self._parseRequest(data)
         code, = struct.unpack('!L', data[:4])
         data = data[4:]
-        msg, data = getNS(data)
-        lang = getNS(data)
+        if data:
+            msg, data = getNS(data)
+            lang = getNS(data)
+        else:
+            msg, lang = None, None
         if code == FX_OK:
             d.callback((msg, lang))
         elif code == FX_EOF:
