diff --git twisted/python/reflect.py twisted/python/reflect.py
index 7740867..c84caf9 100644
--- twisted/python/reflect.py
+++ twisted/python/reflect.py
@@ -103,12 +103,9 @@ def getClass(obj):
 if not _PY3:
     # The following functions aren't documented, nor tested, have much simpler
     # builtin implementations and are not used within Twisted or "known"
-    # projects. There are to be deprecated in
-    # https://twistedmatrix.com/trac/ticket/6859
-    # Refer to this ticket, as well as
-    # https://twistedmatrix.com/trac/ticket/5929?replyto=27#comment:28 and
-    # following comments for more information.
+    # projects.
 
+    @deprecated(Version("Twisted", 14, 0, 0))
     def getcurrent(clazz):
         assert type(clazz) == types.ClassType, 'must be a class...'
         module = namedModule(clazz.__module__)
@@ -120,6 +117,7 @@ if not _PY3:
 
     # Class graph nonsense
     # I should really have a better name for this...
+    @deprecated(Version("Twisted", 14, 0, 0), "isinstance")
     def isinst(inst,clazz):
         if type(inst) != compat.InstanceType or type(clazz)!= types.ClassType:
             return isinstance(inst,clazz)
diff --git twisted/test/test_reflect.py twisted/test/test_reflect.py
index 75a39a9..bdce996 100644
--- twisted/test/test_reflect.py
+++ twisted/test/test_reflect.py
@@ -188,3 +188,26 @@ if not _PY3:
             self.callDeprecated(
                 (Version("Twisted", 11, 0, 0), "inspect.getmro"),
                 reflect.accumulateBases, DeprecationTestCase, l, None)
+
+
+        def test_getcurrent(self):
+            """
+            Test deprecation of L{reflect.getcurrent}.
+            """
+
+            class C:
+                pass
+
+            self.callDeprecated(
+                Version("Twisted", 14, 0, 0),
+                reflect.getcurrent, C)
+
+
+        def test_isinst(self):
+            """
+            Test deprecation of L{reflect.isinst}.
+            """
+
+            self.callDeprecated(
+                (Version("Twisted", 14, 0, 0), "isinstance"),
+                reflect.isinst, object(), object)
diff --git twisted/topfiles/6859.removal twisted/topfiles/6859.removal
new file mode 100644
index 0000000..ddde377
--- /dev/null
+++ twisted/topfiles/6859.removal
@@ -0,0 +1,2 @@
+twisted.python.reflect.getcurrent is now deprecated and must not be used.
+twisted.python.reflect.isinst is now deprecated in favor of the built-in isinstance.
