diff --git twisted/web/http.py twisted/web/http.py
index f4bccfb..fddb56a 100644
--- twisted/web/http.py
+++ twisted/web/http.py
@@ -1050,12 +1050,16 @@ class Request:
         Set the HTTP response code.
 
         @type code: C{int}
-        @type message: C{bytes}
+        @type message: C{unicode}
+        @param message: Message describing the response code. Only ascii
+            characters are allowed.
         """
         if not isinstance(code, _intTypes):
             raise TypeError("HTTP response code must be int or long")
         self.code = code
         if message:
+            if isinstance(message, bytes):
+                message = message.decode('ascii')
             self.code_message = message
         else:
             self.code_message = RESPONSES.get(code, "Unknown Status")
