Index: twisted/names/client.py
===================================================================
--- twisted/names/client.py (revision 26335)
+++ twisted/names/client.py (working copy)
@@ -164,9 +164,10 @@
                 self._lastResolvTime = mtime
                 self.parseConfig(resolvConf)
 
-        # Check again in a little while
-        self._parseCall = self._reactor.callLater(
-            self._resolvReadInterval, self.maybeParseConfig)
+        if self._resolvReadInterval:
+            # Check again in a little while
+            self._parseCall = self._reactor.callLater(
+                self._resolvReadInterval, self.maybeParseConfig)
 
 
     def parseConfig(self, resolvConf):
