Index: twisted/topfiles/6125.misc
===================================================================
Index: twisted/web/test/test_http.py
===================================================================
--- twisted/web/test/test_http.py	(revision 39028)
+++ twisted/web/test/test_http.py	(working copy)
@@ -1455,7 +1455,27 @@
               b"Last-Modified: Thu, 01 Jan 1970 00:00:00 GMT",
               b"Hello")])
 
+    def test_addCookie(self):
+        """
+        L{http.Request.addCookie} sets an outgoing HTTP cookie.
+        """
 
+        req = http.Request(DummyChannel(), None)
+        req.addCookie(b"Hector", b"Berlioz")
+        self.assertEqual(req.cookies[-1], "Hector=Berlioz")
+        req.addCookie(b"Richard", b"Wagner",
+            expires = b"Never",
+            domain = b"Opera",
+            path = b"Germany",
+            max_age = b"70",
+            comment = b"Great!",
+            secure = True)
+        self.assertEqual(req.cookies[-1], 
+            "Richard=Wagner; Expires=Never; Domain=Opera; Path=Germany; "
+            "Max-Age=70; Comment=Great!; Secure")
+
+
+
     def test_parseCookies(self):
         """
         L{http.Request.parseCookies} extracts cookies from C{requestHeaders}
