Index: twisted/web/xmlrpc.py
===================================================================
--- twisted/web/xmlrpc.py	(revision 38002)
+++ twisted/web/xmlrpc.py	(working copy)
@@ -9,7 +9,7 @@
 """
 
 # System Imports
-import sys, xmlrpclib, urlparse
+import base64, sys, xmlrpclib, urlparse
 
 
 # Sibling Imports
@@ -340,7 +340,7 @@
         self.sendHeader('Content-length', str(len(self.factory.payload)))
         if self.factory.user:
             auth = '%s:%s' % (self.factory.user, self.factory.password)
-            auth = auth.encode('base64').strip()
+            auth = base64.b64encode(auth)
             self.sendHeader('Authorization', 'Basic %s' % (auth,))
         self.endHeaders()
         self.transport.write(self.factory.payload)
Index: twisted/web/test/test_xmlrpc.py
===================================================================
--- twisted/web/test/test_xmlrpc.py	(revision 38002)
+++ twisted/web/test/test_xmlrpc.py	(working copy)
@@ -659,6 +659,15 @@
         return d
 
 
+    def test_longPassword(self):
+        long_password = self.password*40
+        p = xmlrpc.Proxy("http://127.0.0.1:%d/" % (
+            self.port,), self.user, long_password)
+        d = p.callRemote("authinfo")
+        d.addCallback(self.assertEqual, [self.user, long_password])
+        return d
+
+
     def test_explicitAuthInfoOverride(self):
         p = xmlrpc.Proxy("http://wrong:info@127.0.0.1:%d/" % (
                 self.port,), self.user, self.password)
