diff --git a/twisted/internet/iocpreactor/tcp.py b/twisted/internet/iocpreactor/tcp.py
index d34f698..5363a52 100644
--- a/twisted/internet/iocpreactor/tcp.py
+++ b/twisted/internet/iocpreactor/tcp.py
@@ -77,7 +77,7 @@ class Connection(abstract.FileHandle, _SocketCloser, _AbortingMixin):
 
     def _closeWriteConnection(self):
         try:
-            getattr(self.socket, self._socketShutdownMethod)(1)
+            self.socket.shutdown(1)
         except socket.error:
             pass
         p = interfaces.IHalfCloseableProtocol(self.protocol, None)
diff --git a/twisted/internet/tcp.py b/twisted/internet/tcp.py
index 988bb7b..c30fd37 100644
--- a/twisted/internet/tcp.py
+++ b/twisted/internet/tcp.py
@@ -108,7 +108,7 @@ else:
 
 
 class _SocketCloser(object):
-    _socketShutdownMethod = 'shutdown'
+    _shouldShutdown = True
 
     def _closeSocket(self, orderly):
         # The call to shutdown() before close() isn't really necessary, because
@@ -118,8 +118,8 @@ class _SocketCloser(object):
         skt = self.socket
         try:
             if orderly:
-                if self._socketShutdownMethod is not None:
-                    getattr(skt, self._socketShutdownMethod)(2)
+                if not self._shouldShutdown:
+                    skt.shutdown(2)
             else:
                 # Set SO_LINGER to 1,0 which, by convention, causes a
                 # connection reset to be sent when close is called,
@@ -248,7 +248,7 @@ class Connection(_TLSConnectionMixin, abstract.FileDescriptor, _SocketCloser,
 
     def _closeWriteConnection(self):
         try:
-            getattr(self.socket, self._socketShutdownMethod)(1)
+            self.socket.shutdown(1)
         except socket.error:
             pass
         p = interfaces.IHalfCloseableProtocol(self.protocol, None)
@@ -978,7 +978,7 @@ class Port(base.BasePort, _SocketCloser):
             skt = self._preexistingSocket
             self._preexistingSocket = None
             # Avoid shutting it down at the end.
-            self._socketShutdownMethod = None
+            self._shouldShutdown = False
 
         # Make sure that if we listened on port 0, we update that to
         # reflect what the OS actually assigned us.
diff --git a/twisted/topfiles/7051.misc b/twisted/topfiles/7051.misc
new file mode 100644
index 0000000..e69de29
