Index: twisted/test/test_ftp.py
===================================================================
--- twisted/test/test_ftp.py    (revision 30184)
+++ twisted/test/test_ftp.py    (working copy)
@@ -2855,3 +2855,18 @@
         self.assertTrue(stor_done)

         return d # just in case an errback occurred
+
+class FTPResponseCodeTest(unittest.TestCase):
+    """Test to check for duplicate use of response codes """
+
+    def test_responseocdes(self):
+        ftp_responsecode_list = ftp.RESPONSE.keys()
+        ftp_responsecode_list_complete = ftp.RESPONSE.keys()
+        for key, value in ftp.__dict__.items():
+            if type(value) == str and key.isupper():
+                if value not in ftp_responsecode_list:
+                    self.assertFalse(value in ftp_responsecode_list_complete,
+                                    "Duplicate responscodes for %s" % value)
+                else:
+                    ftp_responsecode_list.remove(value)
+
Index: twisted/protocols/ftp.py
===================================================================
--- twisted/protocols/ftp.py    (revision 30184)
+++ twisted/protocols/ftp.py    (working copy)
@@ -56,11 +56,11 @@
 NAME_SYS_TYPE                           = "215"
 SVC_READY_FOR_NEW_USER                  = "220.1"
 WELCOME_MSG                             = "220.2"
-SVC_CLOSING_CTRL_CNX                    = "221"
-GOODBYE_MSG                             = "221"
+SVC_CLOSING_CTRL_CNX                    = "221.1"
+GOODBYE_MSG                             = "221.2"
 DATA_CNX_OPEN_NO_XFR_IN_PROGRESS        = "225"
-CLOSING_DATA_CNX                        = "226"
-TXFR_COMPLETE_OK                        = "226"
+CLOSING_DATA_CNX                        = "226.1"
+TXFR_COMPLETE_OK                        = "226.2"
 ENTERING_PASV_MODE                      = "227"
 ENTERING_EPSV_MODE                      = "229"
 USR_LOGGED_IN_PROCEED                   = "230.1"     # v1 of code 230