--- twisted\protocols\ftp.py
+++ twisted\protocols\ftp.py
@@ -59,8 +59,8 @@
 SVC_CLOSING_CTRL_CNX                    = "221"
 GOODBYE_MSG                             = "221"
 DATA_CNX_OPEN_NO_XFR_IN_PROGRESS        = "225"
-CLOSING_DATA_CNX                        = "226"
-TXFR_COMPLETE_OK                        = "226"
+CLOSING_DATA_CNX                        = "226.1"
+TXFR_COMPLETE_OK                        = "226.2"
 ENTERING_PASV_MODE                      = "227"
 ENTERING_EPSV_MODE                      = "229"
 USR_LOGGED_IN_PROCEED                   = "230.1"     # v1 of code 230
