diff --git twisted/trial/runner.py twisted/trial/runner.py
index 4275c24..12f93c6 100644
--- twisted/trial/runner.py
+++ twisted/trial/runner.py
@@ -8,6 +8,9 @@ A miscellany of code used to run Trial tests.
 Maintainer: Jonathan Lange
 """
 
+from __future__ import print_function,division,absolute_import
+from twisted.python.compat import _PY3
+
 __all__ = [
     'TestSuite',
 
@@ -33,7 +36,7 @@ from twisted.trial.reporter import _ExitWrapper, UncleanWarningsReporterWrapper
 # These are imported so that they remain in the public API for t.trial.runner
 from twisted.trial.unittest import TestSuite
 
-from zope.interface import implements
+from zope.interface import implementer
 
 pyunit = __import__('unittest')
 
@@ -252,14 +255,13 @@ def isTestCase(obj):
 
 
 
+@implementer(ITestCase)
 class TestHolder(object):
     """
     Placeholder for a L{TestCase} inside a reporter. As far as a L{TestResult}
     is concerned, this looks exactly like a unit test.
     """
 
-    implements(ITestCase)
-
     failureException = None
 
     def __init__(self, description):
@@ -652,7 +654,7 @@ class TrialRunner(object):
     def _setUpTestdir(self):
         self._tearDownLogFile()
         currentDir = os.getcwd()
-        base = filepath.FilePath(self.workingDirectory)
+        base = filepath.FilePath(self.workingDirectory.encode('utf-8'))
         testdir, self._testDirLock = util._unusedTestDirectory(base)
         os.chdir(testdir.path)
         return currentDir
@@ -715,7 +717,7 @@ class TrialRunner(object):
         if self.logfile == '-':
             logFile = sys.stdout
         else:
-            logFile = file(self.logfile, 'a')
+            logFile = open(self.logfile, 'a')
         self._logFileObject = logFile
         self._logFileObserver = log.FileLogObserver(logFile)
         log.startLoggingWithObserver(self._logFileObserver.emit, 0)
