Index: twisted/web2/channel/fastcgi.py
===================================================================
--- twisted/web2/channel/fastcgi.py	(revision 19737)
+++ twisted/web2/channel/fastcgi.py	(working copy)
@@ -247,6 +247,7 @@
                 if not d:
                     break
                 self.write(d)
+                n += 1
             return
         
         self.writePacket(Record(FCGI_STDOUT, self.reqId, data))
