Index: twisted/protocols/basic.py
===================================================================
--- twisted/protocols/basic.py	(revision 46098)
+++ twisted/protocols/basic.py	(working copy)
@@ -749,8 +749,8 @@
             messageStart = currentOffset + prefixLength
             length, = unpack(fmt, alldata[currentOffset:messageStart])
             if length > self.MAX_LENGTH:
-                self._unprocessed = alldata
-                self._compatibilityOffset = currentOffset
+                self._unprocessed = alldata[currentOffset:]
+                self._compatibilityOffset = 0
                 self.lengthLimitExceeded(length)
                 return
             messageEnd = messageStart + length
Index: twisted/protocols/test/test_basic.py
===================================================================
--- twisted/protocols/test/test_basic.py	(revision 46098)
+++ twisted/protocols/test/test_basic.py	(working copy)
@@ -847,6 +847,20 @@
         self.assertEqual(r.received, [])
 
 
+    def test_noDoubleParseAfterLengthLimitExceeded(self):
+        r = self.getProtocol()
+        r.MAX_LENGTH = 10
+        r.dataReceived(''.join([
+            struct.pack(r.structFormat, 3) + b'w' * 3,
+            struct.pack(r.structFormat, 6) + b'x' * 6,
+            struct.pack(r.structFormat, 9) + b'y' * 9,
+            struct.pack(r.structFormat, 12) + b'z' * 12,
+        ]))
+        self.assertEqual(r.received, [b'w' * 3, b'x' * 6, b'y' * 9])
+        r.dataReceived(b'')
+        self.assertEqual(r.received, [b'w' * 3, b'x' * 6, b'y' * 9])
+
+
     def test_stringReceivedNotImplemented(self):
         """
         When L{IntNStringReceiver.stringReceived} is not overridden in a
