Index: twisted/test/test_amp.py
===================================================================
--- twisted/test/test_amp.py	(revision 17782)
+++ twisted/test/test_amp.py	(working copy)
@@ -78,8 +78,16 @@
 
     fatalErrors = {DeathThreat: 'DEAD'}
 
-class NoAnswerHello(Hello):
-    commandName = Hello.commandName
+class NoAnswerHello(amp.Command):
+    commandName = 'hello'
+
+    arguments = [('hello', amp.String()),
+                 ('optional', amp.Boolean(optional=True)),
+                 ('print', amp.Unicode(optional=True)),
+                 ('from', TransportPeer(optional=True))]
+    response = []
+    errors = {UnfriendlyGreeting: 'UNFRIENDLY'}
+    fatalErrors = {DeathThreat: 'DEAD'}
     requiresAnswer = False
 
 class FutureHello(amp.Command):
@@ -206,6 +214,10 @@
         return result
     Hello.responder(cmdHello)
 
+    def cmdNoAnswerHello(self, hello, From, optional=None, Print=None):
+        self.greeted = True
+    NoAnswerHello.responder(cmdNoAnswerHello)
+
     def cmdGetlist(self, length):
         return {'body': [dict(x=1)] * length}
     GetList.responder(cmdGetlist)
