diff -uNr twisted.orig//web2/http_headers.py twisted/web2/http_headers.py
--- twisted.orig//web2/http_headers.py	2011-02-06 22:43:08.000000000 +0100
+++ twisted/web2/http_headers.py	2011-02-06 22:44:40.000000000 +0100
@@ -471,7 +471,10 @@
                 self.params == other.params)
 
     def __ne__(self, other):
-        return not self.__eq__(other)
+        if not isinstance(other, MimeType): return NotImplemented
+        return not (self.mediaType == other.mediaType and
+                self.mediaSubtype == other.mediaSubtype and
+                self.params == other.params)
 
     def __repr__(self):
         return "MimeType(%r, %r, %r)" % (self.mediaType, self.mediaSubtype, self.params)
