Index: shoutcast.py
===================================================================
--- shoutcast.py	(revision 26541)
+++ shoutcast.py	(working copy)
@@ -107,14 +107,3 @@
     def gotMP3Data(self, data):
         """Called with chunk of MP3 data."""
         raise NotImplementedError, "implement in subclass"
-
-
-if __name__ == '__main__':
-    class Test(ShoutcastClient):
-        def gotMetaData(self, data): print "meta:", data
-        def gotMP3Data(self, data): pass
-    
-    from twisted.internet import protocol, reactor
-    import sys
-    protocol.ClientCreator(reactor, Test).connectTCP(sys.argv[1], int(sys.argv[2]))
-    reactor.run()
