Index: twisted/python/failure.py
===================================================================
--- twisted/python/failure.py	(revision 30970)
+++ twisted/python/failure.py	(working copy)
@@ -17,6 +17,7 @@
 import inspect
 import opcode
 from cStringIO import StringIO
+from inspect import getmro
 
 from twisted.python import reflect
 
@@ -280,7 +281,7 @@
                 ])
             tb = tb.tb_next
         if inspect.isclass(self.type) and issubclass(self.type, Exception):
-            parentCs = reflect.allYourBase(self.type)
+            parentCs = getmro(self.type)
             self.parents = map(reflect.qual, parentCs)
             self.parents.append(reflect.qual(self.type))
         else:
Index: twisted/trial/test/test_reporter.py
===================================================================
--- twisted/trial/test/test_reporter.py	(revision 30970)
+++ twisted/trial/test/test_reporter.py	(working copy)
@@ -9,11 +9,11 @@
 
 
 import errno, sys, os, re, StringIO
+from inspect import getmro
 
 from twisted.internet.utils import suppressWarnings
 from twisted.python import log
 from twisted.python.failure import Failure
-from twisted.python.reflect import allYourBase
 from twisted.trial import itrial, unittest, runner, reporter, util
 from twisted.trial.reporter import UncleanWarningsReporterWrapper
 from twisted.trial.test import erroneous
@@ -1219,7 +1219,7 @@
         method = getattr(klass, methodName, None)
         if method is None:
             return
-        for base in [klass] + allYourBase(klass):
+        for base in [klass] + getmro(klass):
             try:
                 delattr(base, methodName)
             except (AttributeError, TypeError):
