Index: twisted/mail/test/test_pb.py
===================================================================
--- twisted/mail/test/test_pb.py	(revision 0)
+++ twisted/mail/test/test_pb.py	(revision 0)
@@ -0,0 +1,39 @@
+# Copyright (c) Twisted Matrix Laboratories.
+# See LICENSE for details.
+
+from twisted.trial import unittest
+
+from twisted import mail
+import twisted.mail.pb
+
+class MaildirBrokerTests(unittest.TestCase):
+    """
+    Tests that L{twistd.mail.pb.MaildirBroker} is deprecated and the reason
+    for the deprecation which is that invoking either of the two methods 
+    defined in the class will fail with an attribute error.
+    """
+
+    def test_deprecation(self):
+        """
+        Tests both deprecation warning and AttributeError for both methods.
+        """
+        mb = mail.pb.MaildirBroker()
+        warningsShown = self.flushWarnings([self.test_deprecation])
+        self.assertEqual(len(warningsShown), 1)
+        self.assertIdentical(warningsShown[0]['category'], DeprecationWarning)
+        self.assertEqual(
+            warningsShown[0]['message'],
+            "twisted.mail.pb.MaildirBroker was deprecated in Twisted 13.0.0: There is no suggested replacement for this class.")
+        self.assertRaises(AttributeError,
+                          mail.pb.MaildirBroker.proto_getCollection,
+                          mb,
+                          None,
+                          None,
+                          None,
+                          None)
+        self.assertRaises(AttributeError,
+                          mail.pb.MaildirBroker.getCollection,
+                          mb,
+                          None,
+                          None,
+                          None)
Index: twisted/mail/topfiles/6547.removal
===================================================================
--- twisted/mail/topfiles/6547.removal	(revision 0)
+++ twisted/mail/topfiles/6547.removal	(revision 0)
@@ -0,0 +1 @@
+twisted.mail.MaildirBroker deprecated because both its methods are obsolete and fail with an AttributeError if invoked.
Index: twisted/mail/pb.py
===================================================================
--- twisted/mail/pb.py	(revision 38628)
+++ twisted/mail/pb.py	(working copy)
@@ -2,6 +2,8 @@
 # See LICENSE for details.
 
 
+from twisted.python.deprecate import deprecatedModuleAttribute
+from twisted.python.versions import Version
 from twisted.spread import pb
 from twisted.spread import banana
 
@@ -91,6 +93,12 @@
 
 class MaildirBroker(pb.Broker):
 
+    deprecatedModuleAttribute(
+        Version("Twisted", 13, 0, 0),
+        "There is no suggested replacement for this class.",
+        "twisted.mail.pb",
+        "MaildirBroker")
+
     def proto_getCollection(self, requestID, name, domain, password):
         collection = self._getCollection()
         if collection is None:
