diff --git a/twisted/cred/checkers.py b/twisted/cred/checkers.py
index 92cdeee..68086bc 100644
--- a/twisted/cred/checkers.py
+++ b/twisted/cred/checkers.py
@@ -197,24 +197,28 @@ class FilePasswordDB:
 
 
     def _loadCredentials(self):
+        """
+        Reads lines from the credentials file and return them as username,
+        password couples
+
+        @return: an iterable of username, password couples
+        @rtype: iterable
+        """
         try:
-            f = open(self.filename, "rb")
-        except:
-            log.err()
+            with open(self.filename, "rb") as f:
+                for line in f:
+                    line = line.rstrip()
+                    parts = line.split(self.delim)
+
+                    if self.ufield >= len(parts) or self.pfield >= len(parts):
+                        continue
+                    if self.caseSensitive:
+                        yield parts[self.ufield], parts[self.pfield]
+                    else:
+                        yield parts[self.ufield].lower(), parts[self.pfield]
+        except IOError as e:
+            log.err("Unable to load credentials db: {}".format(e))
             raise error.UnauthorizedLogin()
-        else:
-            for line in f:
-                line = line.rstrip()
-                parts = line.split(self.delim)
-
-                if self.ufield >= len(parts) or self.pfield >= len(parts):
-                    continue
-                if self.caseSensitive:
-                    yield parts[self.ufield], parts[self.pfield]
-                else:
-                    yield parts[self.ufield].lower(), parts[self.pfield]
-        finally:
-            f.close()
 
 
     def getUser(self, username):
diff --git a/twisted/cred/test/test_cred.py b/twisted/cred/test/test_cred.py
index f2dafb8..254548b 100644
--- a/twisted/cred/test/test_cred.py
+++ b/twisted/cred/test/test_cred.py
@@ -206,6 +206,15 @@ class OnDiskDatabaseTests(unittest.TestCase):
                 f.write(u + b":" + p + b"\n")
 
 
+    def test_getUserNonexistentDatabase(self):
+        """
+        A missing db file should cause a permanent rejection of authorization
+        attempts.
+        """
+        self.db = checkers.FilePasswordDB('test_thisbetternoteverexist.db')
+        self.failUnlessRaises(error.UnauthorizedLogin, self.db.getUser, 'user')
+
+
     def testUserLookup(self):
         self.db = checkers.FilePasswordDB(self.dbfile)
         for (u, p) in self.users:
diff --git a/twisted/topfiles/8028.bugfix b/twisted/topfiles/8028.bugfix
new file mode 100644
index 0000000..ab4ac32
--- /dev/null
+++ b/twisted/topfiles/8028.bugfix
@@ -0,0 +1 @@
+twisted.cred.checkers.FilePasswordDB now logs an error if the credentials db file does not exist, no longer raises an unhandled error.
