Index: twisted/mail/pop3client.py
===================================================================
--- twisted/mail/pop3client.py	(revision 37075)
+++ twisted/mail/pop3client.py	(working copy)
@@ -697,10 +697,4 @@
         """
         return self.sendShort('QUIT', None)
 
-__all__ = [
-    # Exceptions
-    'InsecureAuthenticationDisallowed', 'LineTooLong', 'POP3ClientError',
-    'ServerErrorResponse', 'TLSError', 'TLSNotSupportedError',
-
-    # Protocol classes
-    'POP3Client']
+__all__ = []
