Index: client.py
===================================================================
RCS file: /cvs/Twisted/twisted/protocols/jabber/client.py,v
retrieving revision 1.2
diff -u -r1.2 client.py
--- client.py	18 Oct 2003 20:30:55 -0000	1.2
+++ client.py	14 Dec 2003 22:26:29 -0000
@@ -149,6 +149,9 @@
             self.xmlstream.dispatch(self.xmlstream, xmlstream.STREAM_AUTHD_EVENT)
         else:
             self.xmlstream.streamError(None)
+        if iq["type"] == "error":
+            if iq.error["code"] == "401":
+                self.xmlstream.dispatch(iq, self.AUTH_FAILED_EVENT)
 
     def registerAccount(self, username = None, password = None):
         if username:
