Index: twisted/words/test/test_irc.py
===================================================================
--- twisted/words/test/test_irc.py	(revision 38290)
+++ twisted/words/test/test_irc.py	(working copy)
@@ -1024,7 +1024,8 @@
 
 
     def test_ERRMSG(self):
-        """Testing CTCP query ERRMSG.
+        """
+        Testing CTCP query ERRMSG.
 
         Not because this is this is an especially important case in the
         field, but it does go through the entire dispatch/decode/encode
@@ -2179,9 +2180,11 @@
         action = 'waves'
         self.protocol.describe(target, action)
         self.protocol.describe(channel, action)
+        self.protocol.describe(channel, r'\o/')
         expected = [
             'PRIVMSG %s :\01ACTION %s\01' % (target, action),
             'PRIVMSG %s :\01ACTION %s\01' % (channel, action),
+            'PRIVMSG %s :\01ACTION %s\01' % (channel, r'\o/'),
             '']
         self.assertEqual(self.transport.value().split('\r\n'), expected)
 
Index: twisted/words/protocols/irc.py
===================================================================
--- twisted/words/protocols/irc.py	(revision 38290)
+++ twisted/words/protocols/irc.py	(working copy)
@@ -3400,7 +3400,7 @@
 
 xQuoteTable = {
     X_DELIM: X_QUOTE + 'a',
-    X_QUOTE: X_QUOTE + X_QUOTE
+    # X_QUOTE: X_QUOTE + X_QUOTE
     }
 
 xDequoteTable = {}
@@ -3411,7 +3411,7 @@
 xEscape_re = re.compile('%s.' % (re.escape(X_QUOTE),), re.DOTALL)
 
 def ctcpQuote(s):
-    for c in (X_QUOTE, X_DELIM):
+    for c in xQuoteTable:
         s = string.replace(s, c, xQuoteTable[c])
     return s
 
