Index: twisted/topfiles/2618.misc
===================================================================
--- twisted/topfiles/2618.misc	(revision 0)
+++ twisted/topfiles/2618.misc	(revision 0)
@@ -0,0 +1 @@
+twisted.words.xish.xmlstream.XmlStream.connectionLost now passes the reason, wrapped in an exception, when dispatching the STREAM_END_EVENT
Index: twisted/words/test/test_xmlstream.py
===================================================================
--- twisted/words/test/test_xmlstream.py	(revision 30761)
+++ twisted/words/test/test_xmlstream.py	(working copy)
@@ -51,7 +51,9 @@
 
     def test_receiveBadXML(self):
         """
-        Receiving malformed XML should result in in error.
+        Receiving malformed XML should result in in error, firing a
+        L{STREAM_ERROR_EVENT} and a consequent L{STREAM_END_EVENT} via
+        inherited L{EventDispatcher.dispatch}
         """
         streamError = []
         streamEnd = []
@@ -59,8 +61,8 @@
         def streamErrorEvent(reason):
             streamError.append(reason)
 
-        def streamEndEvent(_):
-            streamEnd.append(None)
+        def streamEndEvent(reason):
+            streamEnd.append(reason)
 
         self.xmlstream.addObserver(xmlstream.STREAM_ERROR_EVENT,
                                    streamErrorEvent)
@@ -76,6 +78,7 @@
         self.assertEquals(1, len(streamError))
         self.assertTrue(streamError[0].check(domish.ParserError))
         self.assertEquals(1, len(streamEnd))
+        self.assertTrue(streamEnd[0].check(xmlstream.StreamEndException))
 
 
 
Index: twisted/words/xish/xmlstream.py
===================================================================
--- twisted/words/xish/xmlstream.py	(revision 30761)
+++ twisted/words/xish/xmlstream.py	(working copy)
@@ -82,7 +82,8 @@
 
         Dispatches the L{STREAM_END_EVENT}.
         """
-        self.dispatch(self, STREAM_END_EVENT)
+        self.dispatch(failure.Failure(StreamEndException(reason)),
+                STREAM_END_EVENT)
         self.stream = None
 
     ### --------------------------------------------------------------
@@ -259,3 +260,11 @@
         """
         self.resetDelay()
         return XmlStreamFactoryMixin.buildProtocol(self, addr)
+
+
+
+class StreamEndException(Exception):
+    """
+    L{XmlStream.connectionLost} has been called and a STREAM_END event is being
+    fired.  This wraps the reason to be passed to L{failure.Failure}
+    """
