Index: win32/pyx.x-foo.iss.template
===================================================================
--- win32/pyx.x-foo.iss.template	(revision 23166)
+++ win32/pyx.x-foo.iss.template	(working copy)
@@ -13,6 +13,7 @@
 DisableProgramGroupPage=yes
 PrivilegesRequired=admin
 UninstallFilesDir={code:pythonRegKey}
+SetupIconFile=win32\twisted-icon.ico
 
 [Files]
 Source: "%(twhome)s\win32dist\PYTHON%(pyversion10)s\Lib\site-packages\*.*"; DestDir: "{app}"; Flags: ignoreversion recursesubdirs
