diff --git twisted/persisted/sob.py twisted/persisted/sob.py
index 14b7a9a..2ba2e49 100644
--- twisted/persisted/sob.py
+++ twisted/persisted/sob.py
@@ -21,7 +21,7 @@ except ImportError:
 from twisted.python import log, runtime
 from twisted.python.hashlib import md5
 from twisted.persisted import styles
-from zope.interface import implementer, Interface
+from zope.interface import implements, Interface
 
 # Note:
 # These encrypt/decrypt functions only work for data formats
@@ -58,9 +58,10 @@ class IPersistable(Interface):
         """
 
 
-@implementer(IPersistable)
 class Persistent:
 
+    implements(IPersistable)
+
     style = "pickle"
 
     def __init__(self, original, name):
