diff --git a/twisted/internet/task.py b/twisted/internet/task.py
index f784f14..a7d07d4 100644
--- a/twisted/internet/task.py
+++ b/twisted/internet/task.py
@@ -813,10 +813,15 @@ def react(main, argv=(), _reactor=None):
       - Exit the application when done, with exit code 0 in case of success and
         1 in case of failure. If C{main} fails with a C{SystemExit} error, the
         code returned is used.
+      - The following is a simple example of a C{main} function::
+          def main(reactor, username, password):
+              pass
 
-    @param main: A callable which returns a L{Deferred}.  It should take as
-        many arguments as there are elements in the list C{argv}. Called with
-        _reactor as first parameter.
+          task.react(main, ('alice', 'secret'))
+
+    @param main: A callable which returns a L{Deferred}. It should
+        take the reactor as its first parameter, followed by the elements of
+        C{argv}.
 
     @param argv: A list of arguments to pass to C{main}. If omitted the
         callable will be invoked with no additional arguments.
