diff --git task_docstring.patch task_docstring.patch
new file mode 100644
index 0000000..e69de29
diff --git twisted/internet/task.py twisted/internet/task.py
index 5d8ff4b..d5fa5ad 100644
--- twisted/internet/task.py
+++ twisted/internet/task.py
@@ -815,7 +815,8 @@ def react(main, argv=(), _reactor=None):
         code returned is used.
 
     @param main: A callable which returns a L{Deferred}.  It should take as
-        many arguments as there are elements in the list C{argv}.
+        many arguments as there are elements in the list C{argv}. Called with 
+        _reactor as first parameter.
 
     @param argv: A list of arguments to pass to C{main}. If omitted the
         callable will be invoked with no additional arguments.
