=== modified file 'twisted/web/http.py'
--- twisted/web/http.py	2013-01-02 04:21:46 +0000
+++ twisted/web/http.py	2013-04-05 08:22:02 +0000
@@ -830,11 +830,6 @@
             self.transport.unregisterProducer()
         self.producer = None
 
-    # private http response methods
-
-    def _sendError(self, code, resp=''):
-        self.transport.write('%s %s %s\r\n\r\n' % (self.clientproto, code, resp))
-
 
     # The following is the public interface that people should be
     # writing to.

=== added file 'twisted/web/topfiles/6122.removal'
--- twisted/web/topfiles/6122.removal	1970-01-01 00:00:00 +0000
+++ twisted/web/topfiles/6122.removal	2013-04-05 08:28:47 +0000
@@ -0,0 +1,1 @@
+twisted.web.http.Request._sendError has been removed.

