Index: twisted/internet/interfaces.py
===================================================================
--- twisted/internet/interfaces.py	(revision 16629)
+++ twisted/internet/interfaces.py	(working copy)
@@ -531,18 +531,9 @@
         """
 
     def iterate(self, delay=0):
-        """Run the main loop's I/O polling function for a period of time.
-
-        This is most useful in applications where the UI is being drawn "as
-        fast as possible", such as games. All pending L{IDelayedCall}s will
-        be called.
-
-        The reactor must have been started (via the run() method) prior to
-        any invocations of this method.  It must also be stopped manually
-        after the last call to this method (via the stop() method).  This
-        method is not re-entrant: you must not call it recursively; in
-        particular, you must not call it while the reactor is running.
         """
+        Not a public method.  Do not call this method.
+        """
 
     def fireSystemEvent(self, eventType):
         """Fire a system-wide event.
