Index: twisted/protocols/ftp.py
===================================================================
--- twisted/protocols/ftp.py	(revision 34534)
+++ twisted/protocols/ftp.py	(working copy)
@@ -36,7 +36,6 @@
 # constants
 # response codes
 
-RESTART_MARKER_REPLY                    = "100"
 SERVICE_READY_IN_N_MINUTES              = "120"
 DATA_CNX_ALREADY_OPEN_START_XFR         = "125"
 FILE_STATUS_OK_OPEN_DATA_CNX            = "150"
@@ -100,7 +99,6 @@
 
 RESPONSE = {
     # -- 100's --
-    RESTART_MARKER_REPLY:               '110 MARK yyyy-mmmm', # TODO: this must be fixed
     SERVICE_READY_IN_N_MINUTES:         '120 service ready in %s minutes',
     DATA_CNX_ALREADY_OPEN_START_XFR:    '125 Data connection already open, starting transfer',
     FILE_STATUS_OK_OPEN_DATA_CNX:       '150 File status okay; about to open data connection.',
