diff --git a/doc/core/man/trial.1 b/doc/core/man/trial.1
index 00c491c..bfb6f85 100644
--- a/doc/core/man/trial.1
+++ b/doc/core/man/trial.1
@@ -203,7 +203,8 @@ it look like the modules are not present in the system, causing tests to check
 the behavior for that configuration.
 .TP
 \fB-z\fR, \fB--random\fR \fIseed\fR
-Run the tests in random order using the specified seed.
+Run the tests in random order using the specified seed. Don't pass this option
+if you also are passing --order.
 .PP
 .SH SEE ALSO
 The latest version of the trial documentation can be found at
diff --git a/twisted/scripts/trial.py b/twisted/scripts/trial.py
index 7cc7ebc..5e02e30 100644
--- a/twisted/scripts/trial.py
+++ b/twisted/scripts/trial.py
@@ -466,9 +466,6 @@ class Options(_BasicOptions, usage.Options, app.ReactorSelectionMixin):
                 raise usage.UsageError("You must specify --debug when using "
                                        "--nopm ")
             failure.DO_POST_MORTEM = False
-        if self['order'] and self['random']:
-            raise usage.UsageError(
-                "You can't specify --random when using --order")
 
 
 
diff --git a/twisted/trial/test/test_script.py b/twisted/trial/test/test_script.py
index fa7b788..524f376 100644
--- a/twisted/trial/test/test_script.py
+++ b/twisted/trial/test/test_script.py
@@ -480,17 +480,6 @@ class OptionsTestCase(unittest.TestCase):
         self.assertEqual(["--recursionlimit", "2000"], args)
 
 
-    def test_randomConflictsWithOrder(self):
-        """
-        C{parseOptions} raises a C{UsageError} when C{--random} is passed along
-        C{--order}.
-        """
-        opts = ["--random", "4", "--order", "alphabetical"]
-        error = self.assertRaises(UsageError, self.options.parseOptions, opts)
-        self.assertEqual("You can't specify --random when using --order",
-                         str(error))
-
-
     def test_jobsConflictWithDebug(self):
         """
         C{parseOptions} raises a C{UsageError} when C{--debug} is passed along
