diff --git a/twisted/scripts/trial.py b/twisted/scripts/trial.py
index 5e02e30..81e868a 100644
--- a/twisted/scripts/trial.py
+++ b/twisted/scripts/trial.py
@@ -142,7 +142,7 @@ _runOrders = {
 }
 
 
-def _coerceOrder(order):
+def _checkKnownRunOrder(order):
     """
     Check that the given order is a known test running order.
 
@@ -185,9 +185,9 @@ class _BasicOptions(object):
                 ]
 
     optParameters = [
-        ["order", "o", "alphabetical",
+        ["order", "o", None,
          "Specify what order to run test cases and methods. "
-         "See --help-orders for more info.", _coerceOrder],
+         "See --help-orders for more info.", _checkKnownRunOrder],
         ["random", "z", None,
          "Run tests in random order using the specified seed"],
         ['temp-directory', None, '_trial_temp',
@@ -225,6 +225,14 @@ class _BasicOptions(object):
         return result
 
 
+    # TODO: Some of the opt_* methods on this class have docstrings and some do
+    #       not. This is mostly because usage.Options's currently will replace
+    #       any intended output in optFlags and optParameters with the
+    #       docstring. See #6427. When that is fixed, all methods should be
+    #       given docstrings (and it should be verified that those with
+    #       docstrings already have content suitable for printing as usage
+    #       information).
+
     def opt_coverage(self):
         """
         Generate coverage information in the coverage file in the
@@ -371,6 +379,9 @@ class _BasicOptions(object):
         self['reporter'] = self._loadReporterByName(self['reporter'])
         if 'tbformat' not in self:
             self['tbformat'] = 'default'
+        if self['order'] is not None and self['random'] is not None:
+            raise usage.UsageError(
+                "You can't specify --random when using --order")
 
 
 
diff --git a/twisted/trial/test/test_script.py b/twisted/trial/test/test_script.py
index 524f376..028c54a 100644
--- a/twisted/trial/test/test_script.py
+++ b/twisted/trial/test/test_script.py
@@ -522,6 +522,19 @@ class OptionsTestCase(unittest.TestCase):
             str(error))
 
 
+    def test_orderConflictWithRandom(self):
+        """
+        C{parseOptions} raises a C{UsageError} when C{--order} is passed along
+        with C{--random}.
+        """
+        error = self.assertRaises(
+            UsageError,
+            self.options.parseOptions,
+            ["--order", "alphabetical", "--random", "1234"])
+        self.assertEqual("You can't specify --random when using --order",
+                         str(error))
+
+
 
 class MakeRunnerTestCase(unittest.TestCase):
     """
