Index: twisted/topfiles/6444.misc
===================================================================
Index: twisted/test/test_usage.py
===================================================================
--- twisted/test/test_usage.py	(revision 38130)
+++ twisted/test/test_usage.py	(working copy)
@@ -36,7 +36,7 @@
 
 class ParseCorrectnessTest(unittest.TestCase):
     """
-    Test Options.parseArgs for correct values under good conditions.
+    Test Options.parseOptions for correct values under good conditions.
     """
     def setUp(self):
         """
@@ -96,7 +96,7 @@
 
 class TypedTestCase(unittest.TestCase):
     """
-    Test Options.parseArgs for options with forced types.
+    Test Options.parseOptions for options with forced types.
     """
     def setUp(self):
         self.usage = TypedOptions()
@@ -177,7 +177,7 @@
 
 class WrongTypedTestCase(unittest.TestCase):
     """
-    Test Options.parseArgs for wrong coerce options.
+    Test Options.parseOptions for wrong coerce options.
     """
     def test_nonCallable(self):
         """
