diff --git a/twisted/words/test/test_service.py b/twisted/words/test/test_service.py
--- a/twisted/words/test/test_service.py
+++ b/twisted/words/test/test_service.py
@@ -416,7 +416,10 @@ class IRCProtocolTestCase(unittest.TestC
 
         self.assertEqual(response[1][0], 'realmname')
         self.assertEqual(response[1][1], '353')
-        self.assertEqual(response[1][2], ['otheruser', '=', '#somechannel', 'firstuser otheruser'])
+        self.assertIn(response[1][2], [
+                      ['otheruser', '=', '#somechannel', 'firstuser otheruser'],
+                      ['otheruser', '=', '#somechannel', 'otheruser firstuser'],
+                      ])
     testJoin = dG(testJoin)
 
 
