Index: adbapi.py
===================================================================
--- adbapi.py	(revision 17118)
+++ adbapi.py	(working copy)
@@ -123,7 +123,7 @@
 
     CP_ARGS = "min max name noisy openfun reconnect good_sql".split()
 
-    noisy = True # if true, generate informational log messages
+    noisy = False # if true, generate informational log messages
     min = 3 # minimum number of connections in pool
     max = 5 # maximum number of connections in pool
     name = None # Name to assign to thread pool for debugging
