--- irc.py	2008-07-30 23:50:22.000000000 +0200
+++ irc-fixd.py	2008-07-31 00:31:00.000000000 +0200
@@ -576,6 +576,7 @@
     lineRate = None
     _queue = None
     _queueEmptying = None
+    _registered = False
 
     delimiter = '\n' # '\r\n' will also work (see dataReceived)
 
@@ -752,7 +753,7 @@
     def signedOn(self):
         """Called after sucessfully signing on to the server.
         """
-        pass
+        self._registered = True
 
     def kickedFrom(self, channel, kicker, message):
         """Called when I am kicked from a channel.
@@ -927,7 +928,7 @@
         self.sendLine("USER %s %s %s :%s" % (self.username, hostname, servername, self.realname))
 
     def setNick(self, nickname):
-        self.nickname = nickname
+        self.attemptednick = nickname
         self.sendLine("NICK %s" % nickname)
 
     def quit(self, message = ''):
@@ -1008,7 +1009,8 @@
     ### but it is safe to leave them alone.
 
     def irc_ERR_NICKNAMEINUSE(self, prefix, params):
-        self.register(self.nickname+'_')
+        if not self._registered:
+            self.setNick(self.attemptednick+'_')
 
     def irc_ERR_PASSWDMISMATCH(self, prefix, params):
         raise IRCPasswordMismatch("Password Incorrect.")
