diff --git a/twisted/test/test_reflect.py b/twisted/test/test_reflect.py
--- a/twisted/test/test_reflect.py
+++ b/twisted/test/test_reflect.py
@@ -731,15 +731,23 @@ class FullyQualifiedNameTests(unittest.T
             "twisted.python.reflect.fullyQualifiedName")
 
 
-    def test_method(self):
+    def test_bound_method(self):
         """
-        L{reflect.fullyQualifiedName} returns the name of a method inside its
-        class and its module.
+        L{reflect.fullyQualifiedName} returns the name of a bound method
+        inside its class and its module.
+        """
+        self._checkFullyQualifiedName(reflect.PropertyAccessor().reallyDel,
+            "twisted.python.reflect.PropertyAccessor.reallyDel")
+
+
+    def test_unbound_method(self):
+        """
+        L{reflect.fullyQualifiedName} returns the name of an unbound method
+        inside its class and its module.
         """
         self._checkFullyQualifiedName(reflect.PropertyAccessor.reallyDel,
             "twisted.python.reflect.PropertyAccessor.reallyDel")
-        self._checkFullyQualifiedName(reflect.PropertyAccessor().reallyDel,
-            "twisted.python.reflect.PropertyAccessor.reallyDel")
+
 
 
 class DeprecationTestCase(unittest.TestCase):
