diff --git twisted/web/test/test_agent.py twisted/web/test/test_agent.py
index ca1d556..8b1bf61 100644
--- twisted/web/test/test_agent.py
+++ twisted/web/test/test_agent.py
@@ -1955,10 +1955,10 @@ class ContentDecoderAgentTests(TestCase, FakeReactorAndConnectMixin,
         self.assertEqual(len(protocol.requests), 1)
         req, res = protocol.requests.pop()
         self.assertEqual(
-            list(req.headers.getAllRawHeaders()),
-            [('Host', ['example.com']),
+            list(sorted(req.headers.getAllRawHeaders())),
+            [('Accept-Encoding', ['fizz', 'decoder1,decoder2']),
              ('Foo', ['bar']),
-             ('Accept-Encoding', ['fizz', 'decoder1,decoder2'])])
+             ('Host', ['example.com'])])
 
 
     def test_plainEncodingResponse(self):
diff --git twisted/web/topfiles/7680.misc twisted/web/topfiles/7680.misc
new file mode 100644
index 0000000..e69de29
