diff --git twisted/application/service.py twisted/application/service.py
index 66fef85..786169a 100644
--- twisted/application/service.py
+++ twisted/application/service.py
@@ -12,8 +12,9 @@ a sibling).
 
 Maintainer: Moshe Zadka
 """
+from __future__ import print_function,division,absolute_import
 
-from zope.interface import implements, Interface, Attribute
+from zope.interface import implementer, Interface, Attribute
 
 from twisted.python.reflect import namedAny
 from twisted.python import components
@@ -54,13 +55,11 @@ class IServiceMaker(Interface):
         options to desired configuration values.
         """
 
-
-
+@implementer(IPlugin, IServiceMaker)
 class ServiceMaker(object):
     """
     Utility class to simplify the definition of L{IServiceMaker} plugins.
     """
-    implements(IPlugin, IServiceMaker)
 
     def __init__(self, name, module, description, tapname):
         self.name = name
@@ -152,6 +151,7 @@ class IService(Interface):
         """
 
 
+@implementer(IService)
 class Service:
     """
     Base class for services.
@@ -161,8 +161,6 @@ class Service:
     as not serializing this book-keeping information.
     """
 
-    implements(IService)
-
     running = 0
     name = None
     parent = None
@@ -253,7 +251,7 @@ class IServiceCollection(Interface):
         """
 
 
-
+@implementer(IServiceCollection)
 class MultiService(Service):
     """
     Straightforward Service Container.
@@ -264,8 +262,6 @@ class MultiService(Service):
     will finish.
     """
 
-    implements(IServiceCollection)
-
     def __init__(self):
         self.services = []
         self.namedServices = {}
@@ -346,7 +342,7 @@ class IProcess(Interface):
         """)
 
 
-
+@implementer(IProcess)
 class Process:
     """
     Process running parameters.
@@ -354,7 +350,6 @@ class Process:
     Sets up uid/gid in the constructor, and has a default
     of C{None} as C{processName}.
     """
-    implements(IProcess)
     processName = None
 
     def __init__(self, uid=None, gid=None):
