Index: twisted/names/client.py
===================================================================
--- twisted/names/client.py	(revision 33677)
+++ twisted/names/client.py	(working copy)
@@ -543,8 +543,8 @@
 
     @type hosts: C{str} or C{None}
     @param hosts: If not C{None}, an alternate hosts file to use. If C{None}
-    on posix systems, /etc/hosts will be used. On windows, C:\windows\hosts
-    will be used.
+    on posix systems, /etc/hosts will be used. On windows,
+    %SYSTEMROOT%/system32/drivers/etc/hosts will be used.
 
     @rtype: C{IResolver}
     """
@@ -556,9 +556,10 @@
             hosts = '/etc/hosts'
         theResolver = Resolver(resolvconf, servers)
         hostResolver = hostsModule.Resolver(hosts)
-    else:
+    elif platform.isWindows():
         if hosts is None:
-            hosts = r'c:\windows\hosts'
+            systemroot = os.environ["SYSTEMROOT"]
+            hosts = systemroot + r"\system32\drivers\etc\hosts"
         from twisted.internet import reactor
         bootstrap = _ThreadedResolverImpl(reactor)
         hostResolver = hostsModule.Resolver(hosts)
