Index: twisted/words/test/test_jabberxmppstringprep.py
===================================================================
--- twisted/words/test/test_jabberxmppstringprep.py	(revision 30373)
+++ twisted/words/test/test_jabberxmppstringprep.py	(working copy)
@@ -1,4 +1,4 @@
-# Copyright (c) 2005 Twisted Matrix Laboratories.
+# Copyright (c) 2005, 2011 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 from twisted.trial import unittest
@@ -70,6 +70,14 @@
         self.assertEquals(nodeprep.prepare(u'User'), u'user')
         self.assertRaises(UnicodeError, nodeprep.prepare, u'us&er')
 
+
+    def test_nodeprepUnassignedInUnicode32(self):
+        """
+        Make sure unassigned code points from Unicode 3.2 are rejected.
+        """
+        self.assertRaises(UnicodeError, nodeprep.prepare, u'\u1d39')
+
+
     def testNamePrep(self):
         self.assertEquals(nameprep.prepare(u'example.com'), u'example.com')
         self.assertEquals(nameprep.prepare(u'Example.com'), u'example.com')
Index: twisted/words/protocols/jabber/xmpp_stringprep.py
===================================================================
--- twisted/words/protocols/jabber/xmpp_stringprep.py	(revision 30373)
+++ twisted/words/protocols/jabber/xmpp_stringprep.py	(working copy)
@@ -1,6 +1,6 @@
 # -*- test-case-name: twisted.words.test.test_jabberxmppstringprep -*-
 #
-# Copyright (c) 2001-2005 Twisted Matrix Laboratories.
+# Copyright (c) 2005, 2011 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 import sys, warnings
@@ -26,7 +26,12 @@
 
 else:
     import stringprep
-    import unicodedata
+    # We require Unicode version 3.2. Python 2.5 and later provide this as
+    # a separate object. Before that the unicodedata module uses 3.2. 
+    try:
+        from unicodedata import ucd_3_2_0 as unicodedata
+    except:
+        import unicodedata
     from encodings import idna
 
     crippled = False
Index: twisted/words/topfiles/4850.bugfix
===================================================================
--- twisted/words/topfiles/4850.bugfix	(revision 0)
+++ twisted/words/topfiles/4850.bugfix	(revision 0)
@@ -0,0 +1 @@
+twisted.words.protocols.jabber.xmpp_stringprep now always uses Unicode version 3.2 for stringprep normalization.
