Index: twisted/copyright.py
===================================================================
--- twisted/copyright.py	(revision 25128)
+++ twisted/copyright.py	(working copy)
@@ -1,8 +1,6 @@
-# Copyright (c) 2001-2006 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
-
-
 """
 Copyright information for Twisted.
 """
@@ -12,7 +10,7 @@
 longversion = str(longversion)
 
 copyright="""\
-Copyright (c) 2001-2006 Twisted Matrix Laboratories.
+Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 See LICENSE for details."""
 
 disclaimer='''
