diff --git a/twisted/names/server.py b/twisted/names/server.py
index d2b895f..4b990ae 100644
--- a/twisted/names/server.py
+++ b/twisted/names/server.py
@@ -16,6 +16,7 @@ for resolvers to deal with.  Fix it.
 
 @author: Jp Calderone
 """
+from __future__ import division, absolute_import
 
 import time
 
@@ -260,7 +261,7 @@ class DNSServerFactory(protocol.ServerFactory):
         return response
 
 
-    def gotResolverResponse(self, (ans, auth, add), protocol, message, address):
+    def gotResolverResponse(self, ans_auth_add, protocol, message, address):
         """
         A callback used by L{DNSServerFactory.handleQuery} for handling the
         deferred response from C{self.resolver.query}.
@@ -274,14 +275,8 @@ class DNSServerFactory(protocol.ServerFactory):
         The resolved answers count will be logged if C{DNSServerFactory.verbose}
         is C{>1}.
 
-        @param ans: A list of answer records
-        @type ans: L{list} of L{dns.RRHeader} instances
-
-        @param auth: A list of authority records
-        @type auth: L{list} of L{dns.RRHeader} instances
-
-        @param add: A list of additional records
-        @type add: L{list} of L{dns.RRHeader} instances
+        @param ans_auth_add: Answer records, authority records and additional records
+        @type ans_auth_add: L{tuple} of L{list} of L{dns.RRHeader} instances
 
         @param protocol: The DNS protocol instance to which to send a response
             message.
@@ -295,6 +290,7 @@ class DNSServerFactory(protocol.ServerFactory):
             or L{None} if C{protocol} is a stream protocol.
         @type address: L{tuple} or L{None}
         """
+        ans, auth, add = ans_auth_add
         response = self._responseFromMessage(
             message=message, rCode=dns.OK,
             answers=ans, authority=auth, additional=add)
diff --git a/twisted/names/test/test_server.py b/twisted/names/test/test_server.py
index 1be6d1e..5fae655 100644
--- a/twisted/names/test/test_server.py
+++ b/twisted/names/test/test_server.py
@@ -4,6 +4,7 @@
 """
 Test cases for L{twisted.names.server}.
 """
+from __future__ import division, absolute_import
 
 from zope.interface.verify import verifyClass
 
diff --git a/twisted/names/topfiles/8195.feature b/twisted/names/topfiles/8195.feature
new file mode 100644
index 0000000..10d7d80
--- /dev/null
+++ b/twisted/names/topfiles/8195.feature
@@ -0,0 +1 @@
+twisted.names.server is now ported to Python 3
diff --git a/twisted/python/dist3.py b/twisted/python/dist3.py
index 2f33ed5..39ff621 100644
--- a/twisted/python/dist3.py
+++ b/twisted/python/dist3.py
@@ -130,6 +130,7 @@ modules = [
     "twisted.names.dns",
     "twisted.names.error",
     "twisted.names.hosts",
+    "twisted.names.server",
     "twisted.names.resolve",
     "twisted.names.test.__init__",
     "twisted.persisted.__init__",
@@ -315,6 +316,7 @@ testModules = [
     "twisted.names.test.test_hosts",
     "twisted.names.test.test_rfc1982",
     "twisted.names.test.test_util",
+    "twisted.names.test_server",
     "twisted.persisted.test.test_styles",
     "twisted.positioning.test.test_base",
     "twisted.positioning.test.test_nmea",
