Index: twisted/names/client.py
===================================================================
--- twisted/names/client.py	(revision 36613)
+++ twisted/names/client.py	(working copy)
@@ -149,6 +149,7 @@
                 log.msg('%s changed, reparsing' % (self.resolv,))
                 self._lastResolvTime = mtime
                 self.parseConfig(resolvConf)
+            resolvConf.close()
 
         # Check again in a little while
         self._parseCall = self._reactor.callLater(
