Index: twisted/internet/defer.py
===================================================================
--- twisted/internet/defer.py	(revision 45131)
+++ twisted/internet/defer.py	(working copy)
@@ -1267,10 +1267,10 @@
             raise TypeError(
                 "inlineCallbacks requires %r to produce a generator; instead"
                 "caught returnValue being used in a non-generator" % (f,))
-        if not isinstance(gen, types.GeneratorType):
+        if not gen:
             raise TypeError(
-                "inlineCallbacks requires %r to produce a generator; "
-                "instead got %r" % (f, gen))
+                "inlineCallbacks requires %r to produce a generator; instead"
+                "got %r" % (f, gen))
         return _inlineCallbacks(None, gen, Deferred())
     return unwindGenerator
 
