Index: application/service.py
===================================================================
--- application/service.py	(revision 11003)
+++ application/service.py	(working copy)
@@ -102,7 +102,7 @@
     as not serializing this book-keeping information.
     """
 
-    __implements__ = IService,
+    __implements__ = IService
 
     running = 0
     name = None
Index: application/compat.py
===================================================================
--- application/compat.py	(revision 11003)
+++ application/compat.py	(working copy)
@@ -382,7 +382,7 @@
         if hasattr(s, 'privileged'):
             s.privileged = 1
     for s in oldApp.services.values():
-        if not components.implements(s, service.IService):
+        if not service.IService.providedBy(s):
             s.serviceParent = None
             s = _NewService(s)
             s.setServiceParent(IOldApplication(c))
