--- test_runtime.py	Tue Aug 21 20:03:30 2012
+++ test_runtime_test_added.py	Tue Aug 21 20:02:56 2012
@@ -65,6 +65,16 @@
 
 
 
+    def test_isWinNT(self):
+        """
+        L{Platform.isWinNT}'s return value depends on the running system, but
+        it should not raise an exception in any case.
+        """
+        platform = Platform()
+        _ = platform.isWinNT()
+
+
+
 class ForeignPlatformTests(TestCase):
     """
     Tests for L{Platform} based overridden initializer values.
