Index: doc/index.xhtml
===================================================================
--- doc/index.xhtml	(revision 34664)
+++ doc/index.xhtml	(working copy)
@@ -103,10 +103,10 @@
         <a href="historic/index.xhtml">Historical Documents</a>
         <ul>
             <li>
-                <a href="historic/index.xhtml#id1">2003</a>
+                <a href="historic/index.xhtml#auto0">2003</a>
             </li>
             <li>
-                <a href="historic/index.xhtml#previously">Previously</a>
+                <a href="historic/index.xhtml#auto2">Previously</a>
             </li>
         </ul>
     </li>
