diff --git a/twisted/names/test/test_names.py b/twisted/names/test/test_names.py
index b3bc7eb..669dead 100644
--- a/twisted/names/test/test_names.py
+++ b/twisted/names/test/test_names.py
@@ -145,10 +145,6 @@ my_domain_com = NoFileAuthority(
 SOA_com = dns.Record_SOA(
     mname='a.gtld.example.net',
     rname='hostmaster.example.net',
-    serial=2013061801,
-    refresh=1000,
-    retry=2000,
-    expire=3000,
     minimum=4000,
     )
 
@@ -185,10 +181,6 @@ ZONE_com = NoFileAuthority(
 SOA_example_com = dns.Record_SOA(
     mname='ns1.example.com',
     rname='hostmaster.example.com',
-    serial=2013061801,
-    refresh=1000,
-    retry=2000,
-    expire=3000,
     minimum=4000,
     )
 
