--- /home/hestern/pybuild/Twisted-12.1.0/twisted/scripts/tap2rpm.py	2012-08-17 16:38:36.213992031 +0000
+++ /apps/lib/python2.7/site-packages/Twisted-12.1.0-py2.7-linux-x86_64.egg/twisted/scripts/tap2rpm.py	2012-08-17 17:12:02.741992134 +0000
@@ -35,9 +35,9 @@
 [ -r /etc/default/%(rpm_file)s ] && . /etc/default/%(rpm_file)s
 
 #  check for required files
-if [ ! -x %(twistd) ]
+if [ ! -x %(twistd)s ]
 then
-        echo "$0: Aborting, no %(twistd) found"
+        echo "$0: Aborting, no %(twistd)s found"
         exit 0
 fi
 if [ ! -r "$file" ]
@@ -56,7 +56,7 @@
 case "$1" in
         start)
                 echo -n "Starting %(rpm_file)s: twistd"
-                daemon twistd  \\
+                daemon %(twistd)s  \\
                                 --pidfile=$pidfile \\
                                 --rundir=$rundir \\
                                 --%(twistd_option)s=$file \\
@@ -95,7 +95,7 @@
 Group:      Networking/Daemons
 Source:     %(tarfile_basename)s
 BuildRoot:  %%{_tmppath}/%%{name}-%%{version}-root
-Requires:   %(twistd)
+Requires:   %(twistd)s
 BuildArch:  noarch
 
 %%description
