--- tap2rpm.py.orig	2012-08-17 16:34:32.381232942 +0000
+++ tap2rpm.py	2012-08-17 16:38:36.213992031 +0000
@@ -35,9 +35,9 @@
 [ -r /etc/default/%(rpm_file)s ] && . /etc/default/%(rpm_file)s
 
 #  check for required files
-if [ ! -x /usr/bin/twistd ]
+if [ ! -x %(twistd) ]
 then
-        echo "$0: Aborting, no /usr/bin/twistd found"
+        echo "$0: Aborting, no %(twistd) found"
         exit 0
 fi
 if [ ! -r "$file" ]
@@ -95,7 +95,7 @@
 Group:      Networking/Daemons
 Source:     %(tarfile_basename)s
 BuildRoot:  %%{_tmppath}/%%{name}-%%{version}-root
-Requires:   /usr/bin/twistd
+Requires:   %(twistd)
 BuildArch:  noarch
 
 %%description
@@ -237,6 +237,8 @@
 
     tapFileBase = os.path.basename(config['tapfile'])
 
+    twistdPath = os.path.join(os.path.dirname(sys.argv[0]), 'twistd')
+
     initFileInfo = tarfile.TarInfo(
             os.path.join(
                 config['release-name'],
@@ -248,6 +250,7 @@
     initFileRealData = initFileData % {
             'tap_file': tapFileBase,
             'rpm_file': config['release-name'],
+            'twistd': twistdPath,
             'twistd_option': config['twistd_option'],
         }
     initFileInfo.size = len(initFileRealData)
@@ -279,6 +282,7 @@
             'date': time.strftime('%a %b %d %Y', time.localtime(time.time())),
             'maintainer': config['maintainer'],
             'long_description': config['long_description'],
+            'twistd': twistdPath,
         }
     specHandle.write(specFileRealData)
     specHandle.close()
