diff -r e27822cf8aee twisted/internet/test/reactormixins.py
--- a/twisted/internet/test/reactormixins.py	Sun Jun 26 13:08:35 2011 +0000
+++ b/twisted/internet/test/reactormixins.py	Mon Jul 04 16:33:41 2011 +0200
@@ -7,6 +7,7 @@
 
 __metaclass__ = type
 
+import time
 import signal
 
 from twisted.internet.defer import TimeoutError
@@ -196,9 +197,11 @@
             timedOut.append(None)
             reactor.stop()
 
-        reactor.callLater(timeout, stop)
+        reactor.callLater(timeout * 1.1, stop)
+        start = time.time()
         reactor.run()
-        if timedOut:
+        end = time.time()
+        if timedOut or end - start > timeout:
             raise TimeoutError(
                 "reactor still running after %s seconds" % (timeout,))
 
diff -r e27822cf8aee twisted/internet/test/test_tcp.py
--- a/twisted/internet/test/test_tcp.py	Sun Jun 26 13:08:35 2011 +0000
+++ b/twisted/internet/test/test_tcp.py	Mon Jul 04 16:33:41 2011 +0200
@@ -483,7 +483,8 @@
 
         clientFactory = ClientFactory()
         clientFactory.protocol = ClientProtocol
-        reactor.connectTCP('127.0.0.1', port.getHost().port, clientFactory)
+        reactor.connectTCP('127.0.0.1', port.getHost().port, clientFactory,
+                           self.getTimeout() * 2)
         self.runReactor(reactor)
         self.assertFalse(
             writing[0], "Transport was writing after unregisterProducer.")
