=== modified file 'doc/names/examples/gethostbyname.py'
--- doc/names/examples/gethostbyname.py	2012-04-02 15:26:04 +0000
+++ doc/names/examples/gethostbyname.py	2012-09-09 11:12:06 +0000
@@ -5,24 +5,46 @@
 
 """
 Returns the IP address for a given hostname.
+
 To run this script:
 $ python gethostbyname.py <hostname>
 e.g.:
 $ python gethostbyname.py www.google.com
+
+This script does a host lookup using the default Twisted Names
+resolver.
+
+The default Twisted Names resolver is a chained resolver, which
+attempts to lookup a name using:
+ * local hosts file
+ * memory cache of previous lookup results
+ * system wide recursive DNS servers
+
+The hosts file lookup is synchronous, which means that the deferred
+response can run its callbacks before the reactor is started.
+
+To ensure that the reactor is started and stopped in the correct
+order, we use reactor.callWhenRunning to postpone the host lookup
+until after the reactor has started.
 """
 import sys
 from twisted.names import client
 from twisted.internet import reactor
 
+
 def gotResult(result):
     print result
-    reactor.stop()
+
 
 def gotFailure(failure):
     failure.printTraceback()
-    reactor.stop()
-
-d = client.getHostByName(sys.argv[1])
-d.addCallbacks(gotResult, gotFailure)
-
+
+
+def startLookup():
+    d = client.getHostByName(sys.argv[1])
+    d.addCallbacks(gotResult, gotFailure)
+    d.addBoth(lambda ign: reactor.stop())
+
+
+reactor.callWhenRunning(startLookup)
 reactor.run()

=== added file 'twisted/names/topfiles/5596.misc'
