Index: pydoctor/driver.py
===================================================================
--- pydoctor/driver.py	(revisão 39044)
+++ pydoctor/driver.py	(cópia de trabalho)
@@ -54,6 +54,10 @@
         '--project-url', dest='projecturl',
         help=("The project url, appears in the html if given."))
     parser.add_option(
+        '--public', dest='public', action='store_true',
+        help=("Only output public API, skips classes and methods with a "
+              "leading underscore"))
+    parser.add_option(
         '--testing', dest='testing', action='store_true',
         help=("Don't complain if the run doesn't have any effects."))
     parser.add_option(
Index: pydoctor/astbuilder.py
===================================================================
--- pydoctor/astbuilder.py	(revisão 39044)
+++ pydoctor/astbuilder.py	(cópia de trabalho)
@@ -36,6 +36,8 @@
             self.builder.popModule()
 
     def visitClass(self, node):
+        if node.name.startswith('_') and self.system.options.public:
+            return
         rawbases = []
         bases = []
         baseobjects = []
@@ -131,6 +133,8 @@
                 name2fullname[asname] = fullname
 
     def visitFunction(self, node):
+        if node.name.startswith('_') and self.system.options.public:
+            return
         func = self.builder.pushFunction(node.name, node.doc)
         if node.lineno is not None:
             func.linenumber = node.lineno
