diff --git twisted/python/constants.py twisted/python/constants.py
index 416c433..2d4630e 100644
--- twisted/python/constants.py
+++ twisted/python/constants.py
@@ -104,9 +104,9 @@ class _ConstantsContainerType(type):
             descriptor._realize(cls, enumerant, value)
             enumerants[enumerant] = descriptor
 
-        # Save the dictionary which contains *only* constants (distinct from any
-        # other attributes the application may have given the container) where
-        # the class can use it later (eg for lookupByName).
+        # Save the dictionary which contains *only* constants (distinct from
+        # any other attributes the application may have given the container)
+        # where the class can use it later (eg for lookupByName).
         cls._enumerants = enumerants
 
         return cls
@@ -368,11 +368,9 @@ class FlagConstant(_Constant):
 
     def __contains__(self, flag):
         """
-        @param flag: The flag to test for membership in this instance
-            set.
+        @param flag: The flag to test for membership in this instance set.
 
-        @return: C{True} if C{flag} is in this instance set, else
-            C{False}.
+        @return: C{True} if C{flag} is in this instance set, else C{False}.
         """
         # Optimization for testing membership without iteration.
         return bool(flag & self)
@@ -405,13 +403,12 @@ class Flags(Values):
 
         @param name: The name of the constant to create.
 
-        @param descriptor: An instance of a L{FlagConstant} which is
-            assigned to C{name}.
+        @param descriptor: An instance of a L{FlagConstant} which is assigned
+            to C{name}.
 
-        @return: Either the value passed to the C{descriptor}
-            constructor, or the next power of 2 value which will be
-            assigned to C{descriptor}, relative to the value of the
-            last defined L{FlagConstant}.
+        @return: Either the value passed to the C{descriptor} constructor, or
+            the next power of 2 value which will be assigned to C{descriptor},
+            relative to the value of the last defined L{FlagConstant}.
         """
         if descriptor.value is _unspecified:
             value = cls._value
