=== modified file 'twisted/python/modules.py'
--- twisted/python/modules.py	2009-07-17 16:55:51 +0000
+++ twisted/python/modules.py	2009-09-07 20:38:55 +0000
@@ -527,7 +527,7 @@
     I represent the very top of the Python object-space, the module list in
     sys.path and the modules list in sys.modules.
 
-    @ivar _sysPath: a sequence of strings like sys.path.  This attribute is
+    @ivar sysPath: a sequence of strings like sys.path.  This attribute is
     read-only.
 
     @ivar moduleDict: a dictionary mapping string module names to module

