--- server.py	Tue Sep 16 13:09:55 2003
+++ server.fix.py	Tue Sep 16 13:30:48 2003
@@ -340,6 +340,19 @@
             hostport,
             string.join(self.prepath, '/')), "/:")
 
+    def currentPathURL(self):
+        inet, addr, port = self.getHost()
+        if port == 80:
+            hostport = ''
+        else:
+            hostport = ':%d' % port
+        return quote('http%s://%s%s/%s' % (
+            self.isSecure() and 's' or '',
+            self.getRequestHostname(),
+            hostport,
+            string.join(self.prepath[:-1], '/')), "/:")
+
+
     def URLPath(self):
         from twisted.python import urlpath
         return urlpath.URLPath.fromRequest(self)
@@ -349,7 +362,7 @@
         Remember the currently-processed part of the URL for later
         recalling.
         """
-        self.appRootURL = self.prePathURL()
+        self.appRootURL = self.currentPathURL()
 
     def getRootURL(self):
         """
