Index: doc/core/upgrades/2.0/split.xhtml
===================================================================
--- doc/core/upgrades/2.0/split.xhtml	(revision 28115)
+++ doc/core/upgrades/2.0/split.xhtml	(working copy)
@@ -68,9 +68,6 @@
       <li><a href="http://twistedmatrix.com/trac/wiki/TwistedConch">Twisted
       Conch</a> - This contains twisted.conch.</li>
 
-      <li><a href="http://twistedmatrix.com/trac/wiki/TwistedFlow">Twisted
-      Flow</a> - This contains twisted.flow (deprecated).</li>
-
       <li><a href="http://twistedmatrix.com/trac/wiki/TwistedLore">Twisted
       Lore</a> - This contains twisted.lore</li>
 
