diff --git a/twisted/python/_release.py b/twisted/python/_release.py
index b96762d..e9dada6 100644
--- a/twisted/python/_release.py
+++ b/twisted/python/_release.py
@@ -16,9 +16,6 @@ import textwrap
 from datetime import date
 import sys
 import os
-import tarfile
-import errno
-import shutil
 from tempfile import mkdtemp
 
 from subprocess import PIPE, STDOUT, Popen
@@ -27,8 +24,6 @@ from twisted.python.versions import Version
 from twisted.python.filepath import FilePath
 from twisted.python._dist import LoreBuilderMixin, DistributionBuilder
 from twisted.python._dist import makeAPIBaseURL, twisted_subprojects
-from twisted.python._dist import ManBuilder, DocBuilder, NoDocumentsFound
-from twisted.python._dist import isDistributable
 
 # The offset between a year and the corresponding major version number.
 VERSION_OFFSET = 2000
diff --git a/twisted/python/test/test_dist.py b/twisted/python/test/test_dist.py
index 6d887e4..00b819d 100644
--- a/twisted/python/test/test_dist.py
+++ b/twisted/python/test/test_dist.py
@@ -264,8 +264,8 @@ class GetDataFilesTests(TestCase):
             results,
             [(os.path.join(rawBaseDir, "foo"),
               [os.path.join(rawBaseDir, "foo", "bar.txt")]),
-             (os.path.join(rawBaseDir, "foo/baz"),
-              [os.path.join(rawBaseDir, "foo/baz", "qux.txt")])])
+             (os.path.join(rawBaseDir, "foo", "baz"),
+              [os.path.join(rawBaseDir, "foo", "baz", "qux.txt")])])
 
 
     def test_ignoreVCSMetadata(self):
diff --git a/twisted/python/test/test_release.py b/twisted/python/test/test_release.py
index c6b1e96..2d0ce1d 100644
--- a/twisted/python/test/test_release.py
+++ b/twisted/python/test/test_release.py
@@ -58,13 +58,6 @@ else:
 
 
 try:
-    from popen2 import Popen4
-except ImportError:
-    popen4Skip = "popen2.Popen4 is not available."
-else:
-    popen4Skip = skip
-
-try:
     import pydoctor.driver
     # it might not be installed, or it might use syntax not available in
     # this version of Python.
