Index: twisted/words/test/test_jabbersaslmechanisms.py
===================================================================
--- twisted/words/test/test_jabbersaslmechanisms.py	(revision 35245)
+++ twisted/words/test/test_jabbersaslmechanisms.py	(working copy)
@@ -44,6 +44,11 @@
         """
         self.assertIdentical(self.mechanism.getInitialResponse(), None)
 
+    def test_getResponseUnicode(self):
+        self.mechanism = sasl_mechanisms.DigestMD5(u'xmpp', u'example.org', None,
+                                                   u'test', u'\u0418secret')
+        self.test_getResponseNoRealm()
+
     def test_getResponse(self):
         """
         Partially test challenge response.
Index: twisted/words/protocols/jabber/sasl_mechanisms.py
===================================================================
--- twisted/words/protocols/jabber/sasl_mechanisms.py	(revision 35245)
+++ twisted/words/protocols/jabber/sasl_mechanisms.py	(working copy)
@@ -205,6 +205,8 @@
         try:
             username = self.username.encode(charset)
             password = self.password.encode(charset)
+            realm = realm.encode(charset)
+            digest_uri = self.digest_uri.encode(charset)
         except UnicodeError:
             # TODO - add error checking
             raise
@@ -217,7 +219,7 @@
         a1 = "%s:%s:%s" % (H("%s:%s:%s" % (username, realm, password)),
                            nonce,
                            cnonce)
-        a2 = "AUTHENTICATE:%s" % self.digest_uri
+        a2 = "AUTHENTICATE:%s" % digest_uri
 
         response = HEX( KD ( HEX(H(a1)),
                              "%s:%s:%s:%s:%s" % (nonce, nc,
@@ -229,7 +231,7 @@
                       'cnonce' : cnonce,
                       'nc' : nc,
                       'qop' : qop,
-                      'digest-uri': self.digest_uri,
+                      'digest-uri': digest_uri,
                       'response': response,
                       'charset': charset}
 
