*** TwistedWords-0.4.0/twisted/words/test/test_msn.py	2006-03-11 14:58:35.000000000 -0800
--- /usr/lib/python2.4/site-packages/twisted/words/test/test_msn.py	2006-08-24 15:42:33.000000000 -0700
***************
*** 243,248 ****
--- 243,257 ----
          self.client.lineReceived("RNG 1234 192.168.1.1:1863 CKI 123.456 foo@foo.com Screen%20Name")
          self.failUnless(self.client.state == "SBINVITED")
  
+     def testErrback(self):
+         id, d = self.client._createIDMapping()
+         self.x = 'fail'
+         def error(failure):
+             self.x = 'pass'
+         d.addErrback(error)
+         self.client.lineReceived("201 %s" % id)
+         self.failUnless(self.x == 'pass', "Failed to call errback")
+ 
  class MessageHandlingTests(unittest.TestCase):
      """ testing various message handling methods from SwichboardClient """
  
