*** TwistedWords-0.4.0/twisted/words/protocols/msn.py	2006-03-11 14:58:35.000000000 -0800
--- /usr/lib/python2.4/site-packages/twisted/words/protocols/msn.py	2006-08-24 03:04:36.000000000 -0700
***************
*** 626,633 ****
          if len(cmd) != 3:
              raise MSNProtocolError, "Invalid Command, %s" % repr(cmd)
          if cmd.isdigit():
!             if self.ids.has_key(params.split()[0]):
!                 self.ids[id].errback(int(cmd))
                  del self.ids[id]
                  return
              else:       # we received an error which doesn't map to a sent command
--- 626,634 ----
          if len(cmd) != 3:
              raise MSNProtocolError, "Invalid Command, %s" % repr(cmd)
          if cmd.isdigit():
!             id = int(params.split()[0])
!             if self.ids.has_key(id):
!                 self.ids[id][0].errback(int(cmd))
                  del self.ids[id]
                  return
              else:       # we received an error which doesn't map to a sent command
