diff --git twisted/web/iweb.py twisted/web/iweb.py
index 944b7a3..076ed18 100644
--- twisted/web/iweb.py
+++ twisted/web/iweb.py
@@ -12,11 +12,8 @@ Interface definitions for L{twisted.web}.
 
 from zope.interface import Interface, Attribute
 
-from twisted.python.compat import _PY3
 from twisted.internet.interfaces import IPushProducer
-if not _PY3:
-    # Re-enable when cred is ported to Python 3. Fix as part of #6176:
-    from twisted.cred.credentials import IUsernameDigestHash
+from twisted.cred.credentials import IUsernameDigestHash
 
 
 class IRequest(Interface):
diff --git twisted/web/topfiles/7975.misc twisted/web/topfiles/7975.misc
new file mode 100644
index 0000000..e69de29
