Index: twisted/internet/interfaces.py
===================================================================
--- twisted/internet/interfaces.py	(revision 27231)
+++ twisted/internet/interfaces.py	(working copy)
@@ -964,7 +964,11 @@
 
     def doRead():
         """
-        Some data is available for reading on your descriptor.
+        Some data is available for reading on your descriptor. If the
+        connection is closed, return an Exception indicating
+        why. Otherwise, return None. For connections closed cleanly,
+        return L{main.CONNECTION_DONE}. For failed connections return
+        L{main.CONNECTION_LOST}.
         """
 
 
@@ -977,7 +981,11 @@
 
     def doWrite():
         """
-        Some data can be written to your descriptor.
+        Some data can be written to your descriptor. If the connection
+        is closed, return an Exception indicating why. Otherwise,
+        return None. For connections closed cleanly, return
+        L{main.CONNECTION_DONE}. For failed connections return
+        L{main.CONNECTION_LOST}.
         """
 
 
