Index: twisted/internet/interfaces.py
===================================================================
--- twisted/internet/interfaces.py	(revision 27231)
+++ twisted/internet/interfaces.py	(working copy)
@@ -823,7 +823,8 @@
 
         @param reader: An L{IReadDescriptor} provider that will be checked for
                        read events until it is removed from the reactor with
-                       L{removeReader}.
+                       L{removeReader}. Should also implement
+                       L{ILoggingContext}.
 
         @return: C{None}.
         """
@@ -834,7 +835,8 @@
 
         @param writer: An L{IWriteDescriptor} provider that will be checked for
                        read events until it is removed from the reactor with
-                       L{removeWriter}.
+                       L{removeWriter}. Should also implement
+                       L{ILoggingContext}.
 
         @return: C{None}.
         """
@@ -964,7 +966,11 @@
 
     def doRead():
         """
-        Some data is available for reading on your descriptor.
+        Some data is available for reading on your descriptor. If the
+        connection is closed, return an Exception indicating
+        why. Otherwise, return None. For connections closed cleanly,
+        return L{main.CONNECTION_DONE}. For failed connections return
+        L{main.CONNECTION_LOST}.
         """
 
 
@@ -977,7 +983,11 @@
 
     def doWrite():
         """
-        Some data can be written to your descriptor.
+        Some data can be written to your descriptor. If the connection
+        is closed, return an Exception indicating why. Otherwise,
+        return None. For connections closed cleanly, return
+        L{main.CONNECTION_DONE}. For failed connections return
+        L{main.CONNECTION_LOST}.
         """
 
 
