Index: twisted/python/usage.py
===================================================================
--- twisted/python/usage.py	(revision 31837)
+++ twisted/python/usage.py	(working copy)
@@ -307,13 +307,6 @@
 
         parameters = []
 
-        reflect.accumulateClassList(self.__class__, 'optStrings',
-                                    parameters)
-        if parameters:
-            import warnings
-            warnings.warn("Options.optStrings is deprecated, "
-                          "please use optParameters instead.", stacklevel=2)
-
         reflect.accumulateClassList(self.__class__, 'optParameters',
                                     parameters)
 
Index: twisted/topfiles/4552.removal
===================================================================
--- twisted/topfiles/4552.removal	(revision 0)
+++ twisted/topfiles/4552.removal	(revision 0)
@@ -0,0 +1 @@
+Options.optStrings is now removed, after being deprecated for 7 years.
