Index: twisted/trial/unittest.py
===================================================================
--- twisted/trial/unittest.py	(revision 22025)
+++ twisted/trial/unittest.py	(working copy)
@@ -138,7 +138,7 @@
 
     def failUnlessRaises(self, exception, f, *args, **kwargs):
         """fails the test unless calling the function C{f} with the given C{args}
-        and C{kwargs} does not raise C{exception}. The failure will report the
+        and C{kwargs} raises C{exception}. The failure will report the
         traceback and call stack of the unexpected exception.
 
         @param exception: exception type that is to be expected
