diff --git twisted/internet/iocpreactor/udp.py twisted/internet/iocpreactor/udp.py
index 4dec51f..cf30450 100644
--- twisted/internet/iocpreactor/udp.py
+++ twisted/internet/iocpreactor/udp.py
@@ -275,6 +275,28 @@ class Port(abstract.FileHandle):
         return address.IPv4Address('UDP', *self.socket.getsockname())
 
 
+    def setBroadcastAllowed(self, enabled):
+        """
+        Set whether this UDP socket may broadcast. This is disabled by default.
+
+        @param enabled: Whether the socket may broadcast.
+        @type enabled: C{bool}
+        """
+        self.socket.setsockopt(
+            socket.SOL_SOCKET, socket.SO_BROADCAST, enabled)
+
+
+    def getBroadcastAllowed(self):
+        """
+        Checks if broadcast is currently allowed.
+
+        @return: Whether this UDP socket may broadcast.
+        @rtype: C{bool}
+        """
+        return operator.truth(
+            self.socket.getsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST))
+
+
 
 class MulticastMixin:
     """
diff --git twisted/internet/udp.py twisted/internet/udp.py
index b68f63d..7378b52 100644
--- twisted/internet/udp.py
+++ twisted/internet/udp.py
@@ -361,15 +361,16 @@ class Port(base.BasePort):
         @param enabled: Whether the socket may broadcast.
         @type enabled: C{bool}
         """
-        if enabled == True: enabled_int = 1
-        else: enabled_int = 0
         self.socket.setsockopt(
-            socket.SOL_SOCKET, socket.SO_BROADCAST, enabled_int)
+            socket.SOL_SOCKET, socket.SO_BROADCAST, enabled)
 
 
     def getBroadcastAllowed(self):
         """
-        Return whether this UDP socket may broadcast.
+        Checks if broadcast is currently allowed.
+
+        @return: Whether this UDP socket may broadcast.
+        @rtype: C{bool}
         """
         return operator.truth(
             self.socket.getsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST))
