Index: twisted/scripts/twistd.py
===================================================================
--- twisted/scripts/twistd.py	(revision 14674)
+++ twisted/scripts/twistd.py	(working copy)
@@ -144,6 +144,19 @@
 
 def shedPrivileges(euid, uid, gid):
     if uid is not None or gid is not None:
+        import threading
+        # On linux with < kernel 2.6 or < glibc 2.3.5, and perhaps other OSes
+        # too, shedding privs on one thread doesn't shed privs on the other
+        # threads, which is a nasty security hole if it happens.
+        # 
+        # This check is not 100% reliable, as it only catches threads made
+        # with the threading module, not those created by a C library, or
+        # even those created with thread.create_new. But it'll get any
+        # created by twisted itself, and nearly all 3rd party ones as well.
+        if len(threading.enumerate()) > 1:
+            sys.stderr.write("ERROR: Cannot reliably shed privileges when multiple threads are running.\n")
+            os._exit(1)
+            
         switchUID(uid, gid, euid)
         extra = euid and 'e' or ''
         log.msg('set %suid/%sgid %s/%s' % (extra, extra, uid, gid))
