Index: twisted/web/proxy.py
===================================================================
--- twisted/web/proxy.py	(revision 21936)
+++ twisted/web/proxy.py	(working copy)
@@ -106,6 +106,7 @@
         self.father.transport.write("Content-Type: text/html\r\n")
         self.father.transport.write("\r\n")
         self.father.transport.write('''<H1>Could not connect</H1>''')
+        self.father.transport.loseConnection()
 
 
 
Index: twisted/web/test/test_proxy.py
===================================================================
--- twisted/web/test/test_proxy.py	(revision 21936)
+++ twisted/web/test/test_proxy.py	(working copy)
@@ -250,6 +250,7 @@
                 "HTTP/1.0 501 Gateway error\r\n"
                 "Content-Type: text/html\r\n\r\n"
                 "<H1>Could not connect</H1>")
+        self.assertIsInstance(channel.lostReason, ConnectionDone)
 
 
     def test_buildProtocol(self):
