Index: doc/web/howto/using-twistedweb.xhtml
===================================================================
--- doc/web/howto/using-twistedweb.xhtml	(revision 32149)
+++ doc/web/howto/using-twistedweb.xhtml	(working copy)
@@ -498,8 +498,8 @@
 <h3>Web UIs</h3>
 
 <p>
-The <a href="http://www.divmod.org/projects/nevow">Nevow</a> framework, available as
-part of the <a href="http://www.divmod.org/projects/quotient">Quotient</a> project,
+The <a href="https://launchpad.net/nevow">Nevow</a> framework, available as
+part of the <a href="https://launchpad.net/quotient">Quotient</a> project,
 is an advanced system for giving Web UIs to your application. Nevow uses Twisted Web but is
 not itself part of Twisted.
 </p>
@@ -560,7 +560,7 @@
 
 <h3>Serving WSGI Applications</h3>
 
-<p><a href="http://wsgi.org/wsgi/">WSGI</a> is the Web Server Gateway
+<p><a href="http://wsgi.org">WSGI</a> is the Web Server Gateway
 Interface. It is a specification for web servers and application servers to
 communicate with Python web applications. All modern Python web frameworks
 support the WSGI interface.</p>
Index: doc/web/howto/xmlrpc.xhtml
===================================================================
--- doc/web/howto/xmlrpc.xhtml	(revision 32149)
+++ doc/web/howto/xmlrpc.xhtml	(working copy)
@@ -46,7 +46,9 @@
 from twisted.web import xmlrpc, server
 
 class Example(xmlrpc.XMLRPC):
-    """An example object to be published."""
+    """
+    An example object to be published.
+    """
 
     def xmlrpc_echo(self, x):
         """
@@ -130,21 +132,31 @@
 from twisted.web import xmlrpc, server
 
 class Example(xmlrpc.XMLRPC):
-    """An example object to be published."""
+    """
+    An example object to be published.
+    """
 
     def xmlrpc_echo(self, x):
-        """Return all passed args."""
+        """
+        Return all passed args.
+        """
         return x
 
     def xmlrpc_add(self, a, b):
-        """Return sum of arguments."""
+        """
+        Return sum of arguments.
+        """
         return a + b
 
 class Date(xmlrpc.XMLRPC):
-    """Serve the XML-RPC 'time' method."""
+    """
+    Serve the XML-RPC 'time' method.
+    """
 
     def xmlrpc_time(self):
-        """Return UNIX time."""
+        """
+        Return UNIX time.
+        """
         return time.time()
 
 if __name__ == '__main__':
@@ -176,7 +188,7 @@
 <h3>Adding XML-RPC Introspection support</h3>
 
 <p>XML-RPC has an
-informal <a href="http://ldp.kernelnotes.de/HOWTO/XML-RPC-HOWTO/xmlrpc-howto-interfaces.html">Introspection
+informal <a href="http://tldp.org/HOWTO/XML-RPC-HOWTO/xmlrpc-howto-interfaces.html">Introspection
 API</a> that specifies three methods in a <code>system</code>
 sub-handler which allow a client to query a server about the server's
 API. Adding Introspection support to
@@ -288,9 +300,9 @@
 
 <p><code class="API">twisted.web.xmlrpc.XMLRPC</code> and <code
 class="API">twisted.web.soap.SOAPPublisher</code> are both <code class="API"
-base="twisted.web.resource">Resources</code>.  So, to serve both XML-RPC and
+base="twisted.web.resource">Resource</code>s.  So, to serve both XML-RPC and
 SOAP in the one web server, you can use the <code class="API"
-base="twisted.web.resource.IResource">putChild</code> method of Resources.</p>
+base="twisted.web.resource.IResource">putChild</code> method of Resource.</p>
 
 <p>The following example uses an empty <code class="API"
 base="twisted.web">resource.Resource</code> as the root resource for
Index: doc/web/howto/web-overview.xhtml
===================================================================
--- doc/web/howto/web-overview.xhtml	(revision 32149)
+++ doc/web/howto/web-overview.xhtml	(working copy)
@@ -55,7 +55,7 @@
 
     <p>
 	Web programmers seeking a higher level abstraction than the Resource system
-	should look at <a href="http://divmod.org/trac/wiki/DivmodNevow">Nevow</a>.
+	should look at <a href="https://launchpad.net/nevow">Nevow</a>.
         Nevow is based on ideas previously developed in Twisted, but is now maintained
         outside of Twisted to easy development and release cycle pressures.
     </p>
Index: doc/web/howto/resource-templates.xhtml
===================================================================
--- doc/web/howto/resource-templates.xhtml	(revision 32149)
+++ doc/web/howto/resource-templates.xhtml	(working copy)
@@ -13,7 +13,7 @@
 <h2>Overview</h2>
 
 <p>While high-level templating systems can be used with Twisted (for
-example, <a href="http://divmod.org/trac/wiki/DivmodNevow">Divmod
+example, <a href="https://launchpad.net/nevow">Divmod
 Nevow</a>, sometimes one needs a less file-heavy system which lets one
 directly write HTML. While 
 <code class="API" base="twisted.web.script">ResourceScript</code> is
