Index: base.py
===================================================================
--- base.py	(revision 16114)
+++ base.py	(working copy)
@@ -11,7 +11,7 @@
 """
 
 import socket # needed only for sync-dns
-from zope.interface import implements
+from zope.interface import implements, classImplements
 
 import imp
 import sys
@@ -240,11 +240,7 @@
     """Default base class for Reactors.
     """
 
-    _implList = [IReactorCore, IReactorTime, IReactorPluggableResolver]
-    if platform.supportsThreads():
-        _implList.append(IReactorThreads)
-    implements(*_implList)
-    del _implList
+    implements(IReactorCore, IReactorTime, IReactorPluggableResolver)
 
     installed = 0
     usingThreads = 0
@@ -605,6 +601,9 @@
             # See comment in the other callFromThread implementation.
             self.threadCallQueue.append((f, args, kw))
 
+if platform.supportsThreads():
+    classImplements(ReactorBase, IReactorThreads)
+
 components.backwardsCompatImplements(ReactorBase)
 
 
