Index: TwistedCore-2.5.0/twisted/internet/base.py
===================================================================
--- TwistedCore-2.5.0/twisted/internet/base.py	(revision 5259)
+++ TwistedCore-2.5.0/twisted/internet/base.py	(working copy)
@@ -407,16 +407,15 @@
 
 
     def _continueSystemEvent(self, eventType):
-        sysEvtTriggers = self._eventTriggers.get(eventType)
+        # no need to store references to these callbacks anymore, in
+        # fact this can cause problems.
+        sysEvtTriggers = self._eventTriggers.pop(eventType)
         for callList in sysEvtTriggers[1], sysEvtTriggers[2]:
             for callable, args, kw in callList:
                 try:
                     callable(*args, **kw)
                 except:
                     log.deferr()
-        # now that we've called all callbacks, no need to store
-        # references to them anymore, in fact this can cause problems.
-        del self._eventTriggers[eventType]
 
     def addSystemEventTrigger(self, _phase, _eventType, _f, *args, **kw):
         """See twisted.internet.interfaces.IReactorCore.addSystemEventTrigger.
