Index: twisted/test/test_htb.py
===================================================================
--- twisted/test/test_htb.py	(revision 42666)
+++ twisted/test/test_htb.py	(working copy)
@@ -27,6 +27,15 @@
         htb.time = self._realTimeFunc
 
 class TestBucket(TestBucketBase):
+    def test_bucketAddReturnsLong(self):
+        """
+        L{htb.Bucket.add} returns C{long}, no matter what.
+        """
+        b = SomeBucket()
+        b.content = 1.0
+        fit = b.add(1000)
+        self.assertIsInstance(fit, long)
+
     def testBucketSize(self):
         """Testing the size of the bucket."""
         b = SomeBucket()
