Index: twisted/protocols/htb.py
===================================================================
--- twisted/protocols/htb.py	(revision 42640)
+++ twisted/protocols/htb.py	(working copy)
@@ -104,7 +104,7 @@
             now = time()
             deltaTime = now - self.lastDrip
             deltaTokens = deltaTime * self.rate
-            self.content = max(0, self.content - deltaTokens)
+            self.content = long(max(0, self.content - deltaTokens))
             self.lastDrip = now
         return self.content == 0
 
