Index: twisted/test/test_process.py
===================================================================
--- twisted/test/test_process.py	(revision 19551)
+++ twisted/test/test_process.py	(working copy)
@@ -643,13 +643,12 @@
 
 class TestTwoProcessesPosix(TestTwoProcessesBase, SignalMixin, unittest.TestCase):
     def tearDown(self):
-        for i in (0,1):
-            pp, process = self.pp[i], self.processes[i]
+        for pp, process in zip(self.pp, self.processes):
             if not pp.finished:
                 try:
                     os.kill(process.pid, signal.SIGTERM)
                 except OSError:
-                    print "OSError"
+                    pass
         return self._onClose()
 
     def kill(self, num):
