Index: twisted/python/zipstream.py
===================================================================
--- twisted/python/zipstream.py	(revision 30992)
+++ twisted/python/zipstream.py	(working copy)
@@ -249,6 +249,10 @@
     still get an error if you try to create a directory over a file
     with the same name or vice-versa.
     """
+    warnings.warn("zipstream.unzip is deprecated since Twisted 11.0.0 for " +
+                  "security reasons.  Use Python's zipfile instead.",
+                  category=DeprecationWarning, stacklevel=2)
+
     for i in unzipIter(filename, directory, overwrite):
         pass
 
@@ -261,6 +265,10 @@
 
     The value it yields is the number of files left to unzip.
     """
+    warnings.warn("zipstream.unzipIter is deprecated since Twisted 11.0.0 " +
+                  "for security reasons. Use Python's zipfile instead.",
+                  category=DeprecationWarning, stacklevel=2)
+
     zf = zipfile.ZipFile(filename, 'r')
     names = zf.namelist()
     if not os.path.exists(directory):
Index: twisted/topfiles/3666.removal
===================================================================
--- twisted/topfiles/3666.removal	(revision 0)
+++ twisted/topfiles/3666.removal	(revision 0)
@@ -0,0 +1 @@
+twisted.python.zipstream.{unzip,unzipIter} are now deprecated.
