--- http.py.orig	2008-04-11 23:05:22.000000000 +0900
+++ http.py	2008-04-11 23:06:20.000000000 +0900
@@ -574,7 +574,11 @@
 
         This method is not intended for users.
         """
-        self.content.write(data)
+        try:
+            self.content.write(data)
+        except:
+            self.content.close()
+            raise
 
     def requestReceived(self, command, path, version):
         """Called by channel when all data has been received.
