diff --git twisted/test/test_reflect.py twisted/test/test_reflect.py
index b049ce4..40bd765 100644
--- twisted/test/test_reflect.py
+++ twisted/test/test_reflect.py
@@ -358,17 +358,6 @@ class LookupsTestCase(TestCase):
             self.assertIs(result, default)
 
 
-    def test_requireModuleOtherError(self):
-        """
-        When module import fails with error, other than ImportError, it
-        raises that error.
-        """
-        self.assertRaises(
-            ValueError,
-            reflect.requireModule, 'twisted..python',
-            )
-
-
     def test_requireModuleDefaultNone(self):
         """
         When module import fails it returns C{None} by default.
@@ -378,7 +367,6 @@ class LookupsTestCase(TestCase):
         self.assertIs(None, result)
 
 
-
     def test_requireModuleRequestedImport(self):
         """
         When module import succeed it returns the module and not the default
