Index: twisted/mail/test/test_mail.py
===================================================================
--- twisted/mail/test/test_mail.py	(revision 11659)
+++ twisted/mail/test/test_mail.py	(working copy)
@@ -186,6 +186,22 @@
         self.failUnless(self.service.lookupPortal('domain') is o1)
         self.failUnless(self.service.defaultPortal() is o2)
 
+class MaildirAppendTestCase(unittest.TestCase):
+    def setUp(self):
+        self.d = self.mktemp()
+        mail.maildir.initializeMaildir(self.d)
+
+    def tearDown(self):
+        shutil.rmtree(self.d)
+
+    def testAppend(self):
+        mbox = mail.maildir.MaildirMailbox(self.d)
+        for i in xrange(1, 11):
+            mbox.appendMessage("X" * i)
+        self.assertEquals(len(mbox.listMessages()),
+                          10)
+        self.assertEquals(len(mbox.getMessage(5).read()), 6)
+
 class MaildirTestCase(unittest.TestCase):
     def setUp(self):
         self.d = self.mktemp()
Index: twisted/mail/maildir.py
===================================================================
--- twisted/mail/maildir.py	(revision 11659)
+++ twisted/mail/maildir.py	(working copy)
@@ -13,6 +13,7 @@
 import socket
 import time
 import md5
+import tempfile
 
 from zope.interface import implements
 
@@ -239,6 +240,25 @@
                     self.list.append(real)
         self.deleted.clear()
 
+    def appendMessage(self, txt):
+        """Appends a message into the mailbox."""
+        txt = str(txt)
+        fh, tmpname = tempfile.mkstemp(dir=os.path.join(self.path, "tmp"))
+        os.write(fh, txt)
+        os.close(fh)
+        done = False
+        while not done:
+            newname = tempfile.mktemp(dir=os.path.join(self.path, "new"))
+            try:
+                os.rename(tmpname, newname)
+                done = True
+            except OSError, (err, estr):
+                import errno
+                # if the new exists, retry with a new newname.
+                if err != errno.EEXIST: 
+                    raise
+        self.list.append(newname)
+
 class StringListMailbox:
     implements(pop3.IMailbox)
     
