--- /usr/share/pyshared/twisted/enterprise/adbapi.py.orig	2010-04-16 18:52:16.000000000 +0000
+++ /usr/share/pyshared/twisted/enterprise/adbapi.py	2010-04-16 18:58:14.000000000 +0000
@@ -273,8 +273,10 @@
             excType, excValue, excTraceback = sys.exc_info()
             try:
                 conn.rollback()
-            except:
+            except Exception, ex:
                 log.err(None, "Rollback failed")
+                if self.reconnect and isinstance(ex, ConnectionLost):
+                    raise ex
             raise excType, excValue, excTraceback
 
 
@@ -434,8 +436,10 @@
             excType, excValue, excTraceback = sys.exc_info()
             try:
                 conn.rollback()
-            except:
+            except Exception, ex:
                 log.err(None, "Rollback failed")
+                if self.reconnect and isinstance(ex, ConnectionLost):
+                    raise ex
             raise excType, excValue, excTraceback
 
 
