Index: twisted/mail/test/test_pop3.py
===================================================================
--- twisted/mail/test/test_pop3.py	(revision 12695)
+++ twisted/mail/test/test_pop3.py	(working copy)
@@ -409,7 +409,7 @@
         s.truncate(0)
         
         p.lineReceived("LIST 1")
-        self.assertEquals(s.getvalue(), "+OK 44\r\n")
+        self.assertEquals(s.getvalue(), "+OK 1 44\r\n")
         s.truncate(0)
         
         p.lineReceived("LIST")
Index: twisted/mail/pop3.py
===================================================================
--- twisted/mail/pop3.py	(revision 12695)
+++ twisted/mail/pop3.py	(working copy)
@@ -369,7 +369,7 @@
             self.sendLine('.')
         else:
             msg = self.mbox.listMessages(int(i) - 1)
-            self.successResponse(str(msg))
+            self.successResponse('%s %d'%(i,msg))
 
     def do_UIDL(self, i=None):
         if i is None:
