Index: twisted/web2/log.py
===================================================================
--- twisted/web2/log.py	(revision 19642)
+++ twisted/web2/log.py	(working copy)
@@ -153,7 +153,7 @@
             '.'.join([str(x) for x in request.clientproto]))
         
         self.logMessage(
-            '%s - %s [%s] "%s" %s %d "%s" "%s"' %(
+            '%s - %s [%s] "%s" %s %d "%s" "%s" [%.1f ms]' %(
                 request.remoteAddr.host,
                 # XXX: Where to get user from?
                 "-",
@@ -163,7 +163,8 @@
                 response.code,
                 loginfo.bytesSent,
                 request.headers.getHeader('referer', '-'),
-                request.headers.getHeader('user-agent', '-')
+                request.headers.getHeader('user-agent', '-'),
+                (time.time() - request.initTime) * 1000,
                 )
             )
 
