--- cfreactor.py (saved version)
+++ (current document)
@@ -238,9 +238,9 @@
         if NSApp() is None and self.nsRunLoop.currentMode() is None:
             # Most of the time the NSRunLoop will have already started,
             # but in this case it wasn't.
-            runLoop.run()
             self.crashing = False
             self.didStartRunLoop = True
+            runLoop.run()
 
     def callLater(self, howlong, *args, **kwargs):
         rval = posixbase.PosixReactorBase.callLater(self, howlong, *args, **kwargs)
