Index: imap4.py
===================================================================
--- imap4.py	(revision 25540)
+++ imap4.py	(working copy)
@@ -4573,8 +4573,9 @@
         @param negate: If True, indicates that the headers listed in C{names}
         should be omitted from the return value, rather than included.
 
-        @rtype: C{dict}
-        @return: A mapping of header field names to header field values
+        @rtype: C{dict} or C{email.message.Message}
+        @return: A mapping of header field names to header field
+        values or an email.message.Message object
         """
 
     def getBodyFile():
@@ -4889,8 +4890,10 @@
         """
 
 def _formatHeaders(headers):
+    # Specifically use headers.items() here so that it'll work with
+    # email.message.Message
     hdrs = [': '.join((k.title(), '\r\n'.join(v.splitlines()))) for (k, v)
-            in headers.iteritems()]
+            in headers.items()]
     hdrs = '\r\n'.join(hdrs) + '\r\n'
     return hdrs
 
