diff --git twisted/topfiles/6165.doc twisted/topfiles/6165.doc
new file mode 100644
index 0000000..e96a518
--- /dev/null
+++ twisted/topfiles/6165.doc
@@ -0,0 +1 @@
+twisted.trial.unittest.SynchronousTestCase.mktemp now makes clear that it creates a path that is unique within the current working directory.
diff --git twisted/trial/_synctest.py twisted/trial/_synctest.py
index 14c4c2c..f7549ca 100644
--- twisted/trial/_synctest.py
+++ twisted/trial/_synctest.py
@@ -1020,11 +1020,16 @@ class SynchronousTestCase(_Assertions):
 
     def mktemp(self):
         """
-        Returns a unique name that may be used as either a temporary directory
-        or filename.
-
-        @note: you must call os.mkdir on the value returned from this method if
-            you wish to use it as a directory!
+        Return a relative path that is guaranteed to be unique within the
+        current working directory. Create every directory between the current 
+        working directory and the last one if necessary. Do not create the 
+        last directory/file.
+        
+        For a temporary directory call os.mkdir on the path, for a temporary
+        file, just create the file (e.g. by opening the path for writing and
+        then closing it).
+
+        Trial should delete the temporary file automatically.
 
         @return: C{str}
         """
