Index: doc/development/policy/coding-standard.xhtml
===================================================================
--- doc/development/policy/coding-standard.xhtml	(revision 24967)
+++ doc/development/policy/coding-standard.xhtml	(working copy)
@@ -37,12 +37,12 @@
 
     <p>Unit tests are written using the <code
     class="API">twisted.trial</code> framework. Many examples are in the
-    <code class="API">twisted.test</code> package. Test modules should start
-    with 'test_' in their name. Source files should have <a
-    href="test-standard.xhtml"><code>test-case-name</code></a> tags that
-    point to their related tests.</p>
+    <code class="API">twisted.test</code> package. Test modules should 
+    start with 'test_' in their name. Source files should have 
+    <code>test-case-name</code> tags that point to their related 
+    tests, as per the <a href="test-standard.xhtml">Test Standard</a>.</p>
 
-    <p>Run the unit tests tests before you check anything in.</p>
+    <p>Run the unit tests before you check anything in.</p>
 
     <p>Let me repeat that, for emphasis: <em>run the unit tests before you
     check <strong>anything</strong> in</em>.  Code which breaks functionality
@@ -57,6 +57,21 @@
     frequently.  Support for other editors is unavailable at this time
     but we would love to provide it.</p>
 
+    <p>Alternatively, to perform the tests, run the following from the root 
+    of the Twisted source tree:</p>
+
+    <pre class="shell">
+$ bin/trial twisted
+    </pre>
+
+    <p>To run an individual test module, such as 
+    <code>twisted/mail/test/test_pop3.py</code>, specify the module name 
+    accordingly:</p>
+
+    <pre class="shell">
+$ bin/trial twisted.mail.test.test_pop3
+    </pre>
+
     <p>All unit test methods should have docstrings specifying at a high 
     level the intent of the test. That is, a description that users of the 
     method would understand.</p>
