Index: microdom.py
===================================================================
--- microdom.py	(revision 25409)
+++ microdom.py	(working copy)
@@ -178,9 +178,7 @@
         assert isinstance(newChild, Node)
         #if newChild.parentNode:
         #    newChild.parentNode.removeChild(newChild)
-        assert (oldChild.parentNode is self,
-                ('oldChild (%s): oldChild.parentNode (%s) != self (%s)'
-                 % (oldChild, oldChild.parentNode, self)))
+        assert oldChild.parentNode is self
         self.childNodes[self.childNodes.index(oldChild)] = newChild
         oldChild.parentNode = None
         newChild.parentNode = self
