--- twisted/web/websockets.py	(revision 36466)
+++ twisted/web/websockets.py	(working copy)
@@ -378,6 +378,12 @@
 
             ProtocolWrapper.loseConnection(self)
 
+    def onRequest(self, request):
+        request_hanler = getattr(self.wrappedProtocol,
+                                 'onRequest',
+                                 None)
+        if request_hanler:
+            request_hanler(request)
 
 
 class _WebSocketsFactory(WrappingFactory):
