*** websocket.py	2010-07-12 11:11:29.239039046 +0200
--- websocket_update.py	2010-07-15 11:03:41.753587508 +0200
***************
*** 195,200 ****
--- 195,201 ----
          if not handlerFactory:
              return finish()
          transport = WebSocketTransport(self)
+         transport.cookie = self.requestHeaders.getRawHeaders("Cookie", [])
          handler = handlerFactory(transport)
          transport._attachHandler(handler)
  
***************
*** 253,258 ****
--- 254,260 ----
              # XXX we probably don't want to set _transferDecoder
              self.channel._transferDecoder = WebSocketFrameDecoder(
                  self, handler)
+             handler.connectionMade()
              return
  
  
***************
*** 374,380 ****
          connection is closed.
          """
  
! 
  
  class WebSocketFrameDecoder(object):
      """
--- 376,385 ----
          connection is closed.
          """
  
!     def connectionMade(self):
!         """
!         Callback called when the underlying transport is ready
!         """
  
  class WebSocketFrameDecoder(object):
      """
