diff a/twisted/python/dist3.py b/twisted/python/dist3.py
index 340c3f2..c7a8e61 100644
--- a/twisted/python/dist3.py
+++ b/twisted/python/dist3.py
@@ -160,6 +160,7 @@ modules = [
     "twisted.python.procutils",
     "twisted.python.randbytes",
     "twisted.python.reflect",
+    "twisted.python.roots",
     "twisted.python.runtime",
     "twisted.python.sendmsg",
     "twisted.python.systemd",
@@ -337,6 +338,7 @@ testModules = [
     "twisted.test.test_process",
     "twisted.test.test_randbytes",
     "twisted.test.test_reflect",
+    "twisted.test.test_roots",
     "twisted.test.test_sob",
     "twisted.test.test_ssl",
     "twisted.test.test_sslverify",
diff a/twisted/python/roots.py b/twisted/python/roots.py
index ee3c8a3..5991ece 100644
--- a/twisted/python/roots.py
+++ b/twisted/python/roots.py
@@ -9,7 +9,6 @@ Maintainer: Glyph Lefkowitz
 """
 
 # System imports
-import types
 from twisted.python import reflect
 
 class NotSupportedError(NotImplementedError):
@@ -232,7 +231,7 @@ class Homogenous(Constrained):
     specified by my 'entityType' attribute.
     """
 
-    entityType = types.InstanceType
+    entityType = object
 
     def entityConstraint(self, entity):
         if isinstance(entity, self.entityType):
diff a/twisted/test/test_roots.py b/twisted/test/test_roots.py
index a73d86a..9061986 100644
--- a/twisted/test/test_roots.py
+++ b/twisted/test/test_roots.py
@@ -3,7 +3,6 @@
 
 from twisted.trial import unittest
 from twisted.python import roots
-import types
 
 class RootsTests(unittest.TestCase):
 
@@ -55,7 +54,7 @@ class RootsTests(unittest.TestCase):
 
     def testHomogenous(self):
         h = roots.Homogenous()
-        h.entityType = types.IntType
+        h.entityType = int
         h.putEntity('a', 1)
         self.assertEqual(h.getStaticEntity('a'),1 )
         self.failUnlessRaises(roots.ConstraintViolation,
diff a/twisted/topfiles/8131.feature b/twisted/topfiles/8131.feature
new file mode 100644
index 0000000..bc2c6f2
--- /dev/null
+++ b/twisted/topfiles/8131.feature
@@ -0,0 +1 @@
+twisted.python.roots now works with Python 3.
