Index: smtp.py
===================================================================
--- smtp.py	(revision 12877)
+++ smtp.py	(working copy)
@@ -566,17 +566,22 @@
         return getattr(self, 'state_' + self.mode)(line)
 
     def state_COMMAND(self, line):
-        words = line.split(None, 1)
-        try:
-            command = words[0]
-        except IndexError:
+        if (len(line) < 4):
             self.sendSyntaxError()
+
+        spacePos = line.find(" ")
+        if spacePos != -1:
+            command = line[:spacePos]
+            rest = line[spacePos+1:]
         else:
-            method = self.lookupMethod(command)
-            if method is None:
-                method = self.do_UNKNOWN
-            method(line[len(command):].strip())
+            command = line
+            rest = ''
 
+        method = self.lookupMethod(command)
+        if method is None:
+            method = self.do_UNKNOWN
+        method(rest)
+
     def sendSyntaxError(self):
         self.sendCode(500, 'Error: bad syntax')
 
