diff --git a/twisted/internet/iocpreactor/iocpsupport/acceptex.pxi b/twisted/internet/iocpreactor/iocpsupport/acceptex.pxi
--- a/twisted/internet/iocpreactor/iocpsupport/acceptex.pxi
+++ b/twisted/internet/iocpreactor/iocpsupport/acceptex.pxi
@@ -22,6 +22,7 @@ def accept(long listening, long acceptin
     if not rc:
         rc = WSAGetLastError()
         if rc != ERROR_IO_PENDING:
+            PyMem_Free(ov)
             return rc
 
     # operation is in progress
diff --git a/twisted/internet/iocpreactor/iocpsupport/connectex.pxi b/twisted/internet/iocpreactor/iocpsupport/connectex.pxi
--- a/twisted/internet/iocpreactor/iocpsupport/connectex.pxi
+++ b/twisted/internet/iocpreactor/iocpsupport/connectex.pxi
@@ -29,6 +29,7 @@ def connect(long s, object addr, object 
     if not rc:
         rc = WSAGetLastError()
         if rc != ERROR_IO_PENDING:
+            PyMem_Free(ov)
             return rc
 
     # operation is in progress
diff --git a/twisted/internet/iocpreactor/iocpsupport/iocpsupport.pyx b/twisted/internet/iocpreactor/iocpsupport/iocpsupport.pyx
--- a/twisted/internet/iocpreactor/iocpsupport/iocpsupport.pyx
+++ b/twisted/internet/iocpreactor/iocpsupport/iocpsupport.pyx
@@ -186,6 +186,9 @@ cdef class CompletionPort:
 
         rc = PostQueuedCompletionStatus(self.port, bytes, key, <OVERLAPPED *>ov)
         if not rc:
+            if ov:
+                Py_DECREF(obj)
+                PyMem_Free(ov)
             raise_error(0, 'PostQueuedCompletionStatus')
 
     def __del__(self):
diff --git a/twisted/internet/iocpreactor/iocpsupport/wsarecv.pxi b/twisted/internet/iocpreactor/iocpsupport/wsarecv.pxi
--- a/twisted/internet/iocpreactor/iocpsupport/wsarecv.pxi
+++ b/twisted/internet/iocpreactor/iocpsupport/wsarecv.pxi
@@ -28,6 +28,7 @@ def recv(long s, object bufflist, object
         if rc == SOCKET_ERROR:
             rc = WSAGetLastError()
             if rc != ERROR_IO_PENDING:
+                PyMem_Free(ov)
                 return rc, 0
 
         Py_XINCREF(obj)
@@ -61,6 +62,7 @@ def recvfrom(long s, object buff, object
     if rc == SOCKET_ERROR:
         rc = WSAGetLastError()
         if rc != ERROR_IO_PENDING:
+            PyMem_Free(ov)
             return rc, 0
 
     Py_XINCREF(obj)
diff --git a/twisted/internet/iocpreactor/iocpsupport/wsasend.pxi b/twisted/internet/iocpreactor/iocpsupport/wsasend.pxi
--- a/twisted/internet/iocpreactor/iocpsupport/wsasend.pxi
+++ b/twisted/internet/iocpreactor/iocpsupport/wsasend.pxi
@@ -19,6 +19,7 @@ def send(long s, object buff, object obj
     if rc == SOCKET_ERROR:
         rc = WSAGetLastError()
         if rc != ERROR_IO_PENDING:
+            PyMem_Free(ov)
             return rc, bytes
 
     Py_XINCREF(obj)
