Index: twisted/trial/test/test_runner.py
===================================================================
--- twisted/trial/test/test_runner.py	(revision 26936)
+++ twisted/trial/test/test_runner.py	(working copy)
@@ -477,7 +477,24 @@
         self.assertEqual(['runcall'], debugger._calls)
 
 
+    def test_noMarker(self):
+        """
+        Specifying a temp directory that does not have a trial marker results
+        in an exception and does not remove the specified directory.
+        """
+        tempPath = self.mktemp()
+        # Create our working directory outside of trial.
+        os.mkdir(tempPath)
+        self.parseOptions(['--temp-directory', tempPath,
+                           'twisted.trial.test.sample'])
+        my_runner = self.getRunner()
+        loader = runner.TestLoader()
+        suite = loader.loadByName('twisted.trial.test.sample', True)
+        self.assertRaises(runner.NoTrialMarker, my_runner.run, suite)
+        self.assertTrue(os.path.exists(tempPath))
 
+
+
 class TestTrialSuite(unittest.TestCase):
 
     def test_imports(self):
Index: twisted/trial/runner.py
===================================================================
--- twisted/trial/runner.py	(revision 26936)
+++ twisted/trial/runner.py	(working copy)
@@ -40,6 +40,13 @@
 
 
 
+class NoTrialMarker(Exception):
+    """
+    No trial marker file could be found.
+    """
+
+
+
 def isPackage(module):
     """Given an object return True if the object looks like a package"""
     if not isinstance(module, types.ModuleType):
@@ -697,6 +704,17 @@
 
 
     def _removeSafely(self, path):
+        """
+        Safely remove a path, recursively.
+
+        If C{path} does not contain a node named C{_trial_marker}, a
+        C{NoTrialmarker} exception is raised and the path is not removed.
+        """
+        if not os.path.exists(os.path.join(path, '_trial_marker')):
+            raise NoTrialMarker(
+                '%r is not a trial temporary path, refusing to remove it'
+                % (path,))
+
         try:
             shutil.rmtree(path)
         except OSError, e:
@@ -739,6 +757,7 @@
 
         os.mkdir(testdir)
         os.chdir(testdir)
+        file('_trial_marker', 'w').close()
         return currentDir
 
 
