Index: twisted/mail/topfiles/3751.misc
===================================================================
Index: twisted/mail/imap4.py
===================================================================
--- twisted/mail/imap4.py	(revision 31982)
+++ twisted/mail/imap4.py	(working copy)
@@ -2678,8 +2678,8 @@
             chal = base64.decodestring(rest + '\n')
         except binascii.Error:
             self.sendLine('*')
+            self.transport.loseConnection()
             raise IllegalServerResponse(rest)
-            self.transport.loseConnection()
         else:
             auth = self.authenticators[scheme]
             chal = auth.challengeResponse(secret, chal)
