Index: twisted/conch/__init__.py
===================================================================
--- twisted/conch/__init__.py	(revision 24265)
+++ twisted/conch/__init__.py	(working copy)
@@ -2,9 +2,6 @@
 # Copyright (c) 2001-2004 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
-#
-
-
 """
 Twisted.Conch: The Twisted Shell. Terminal emulation, SSHv2 and telnet.
 
Index: twisted/conch/ssh/agent.py
===================================================================
--- twisted/conch/ssh/agent.py	(revision 24265)
+++ twisted/conch/ssh/agent.py	(working copy)
@@ -1,7 +1,6 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
-#
 """
 Implements the old SSHv1 key agent protocol.
 
@@ -9,7 +8,7 @@
 """
 
 import struct
-from common import NS, getNS
+from twisted.conch.ssh.common import NS, getNS
 from twisted.conch.error import ConchError
 from twisted.internet import defer, protocol
 
Index: twisted/conch/ssh/userauth.py
===================================================================
--- twisted/conch/ssh/userauth.py	(revision 24265)
+++ twisted/conch/ssh/userauth.py	(working copy)
@@ -1,4 +1,4 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 # 
@@ -15,8 +15,8 @@
 from twisted.cred import credentials
 from twisted.internet import defer, reactor
 from twisted.python import failure, log
-from common import NS, getNS, MP
-import keys, transport, service
+from twisted.conch.ssh.common import NS, getNS, MP
+from twisted.conch.ssh import keys, transport, service
 
 class SSHUserAuthServer(service.SSHService):
     name = 'ssh-userauth'
Index: twisted/conch/ssh/forwarding.py
===================================================================
--- twisted/conch/ssh/forwarding.py	(revision 24265)
+++ twisted/conch/ssh/forwarding.py	(working copy)
@@ -1,8 +1,6 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
-# 
-
 """
 This module contains the implementation of the TCP forwarding, which allows
 clients and servers to forward arbitrary TCP data across the connection.
@@ -14,9 +12,8 @@
 
 from twisted.internet import protocol, reactor
 from twisted.python import log
+from twisted.conch.ssh import common, channel
 
-import common, channel
-
 class SSHListenForwardingFactory(protocol.Factory):
     def __init__(self, connection, hostport, klass):
         self.conn = connection
Index: twisted/conch/ssh/filetransfer.py
===================================================================
--- twisted/conch/ssh/filetransfer.py	(revision 24265)
+++ twisted/conch/ssh/filetransfer.py	(working copy)
@@ -9,13 +9,11 @@
 from twisted.internet import defer, protocol
 from twisted.python import failure, log
 
-from common import NS, getNS
+from twisted.conch.ssh.common import NS, getNS
 from twisted.conch.interfaces import ISFTPServer, ISFTPFile
 
 from zope import interface
 
-
-
 class FileTransferBase(protocol.Protocol):
 
     versions = (3, )
Index: twisted/conch/ui/tkvt100.py
===================================================================
--- twisted/conch/ui/tkvt100.py	(revision 24265)
+++ twisted/conch/ui/tkvt100.py	(working copy)
@@ -1,17 +1,19 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 #
 
-"""Module to emulate a VT100 terminal in Tkinter.
+"""
+Module to emulate a VT100 terminal in Tkinter.
 
 Maintainer: U{Paul Swartz <mailto:z3p@twistedmatrix.com>}
 """
 
 import Tkinter, tkFont
-import ansi
 import string
 
+from twisted.conch.ui import ansi
+
 ttyFont = None#tkFont.Font(family = 'Courier', size = 10)
 fontWidth, fontHeight = None,None#max(map(ttyFont.measure, string.letters+string.digits)), int(ttyFont.metrics()['linespace'])
 
Index: twisted/conch/client/connect.py
===================================================================
--- twisted/conch/client/connect.py	(revision 24265)
+++ twisted/conch/client/connect.py	(working copy)
@@ -1,8 +1,7 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
-#
-import direct, unix
+from twisted.conch.client import direct, unix
 
 connectTypes = {"direct" : direct.connect,
                 "unix" : unix.connect}
Index: twisted/conch/client/direct.py
===================================================================
--- twisted/conch/client/direct.py	(revision 24265)
+++ twisted/conch/client/direct.py	(working copy)
@@ -11,8 +11,6 @@
 
 from twisted.conch.client import unix
 
-
-
 class SSHClientFactory(protocol.ClientFactory):
 
     def __init__(self, d, options, verifyHostKey, userAuthObject):
Index: twisted/conch/client/options.py
===================================================================
--- twisted/conch/client/options.py	(revision 24265)
+++ twisted/conch/client/options.py	(working copy)
@@ -3,10 +3,9 @@
 
 #
 from twisted.conch.ssh.transport import SSHClientTransport, SSHCiphers
+from twisted.conch.client import connect
 from twisted.python import usage
 
-import connect
-
 import sys
 
 class ConchOptions(usage.Options):
Index: twisted/conch/client/default.py
===================================================================
--- twisted/conch/client/default.py	(revision 24265)
+++ twisted/conch/client/default.py	(working copy)
@@ -1,14 +1,13 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 #
 from twisted.conch.error import ConchError
 from twisted.conch.ssh import common, keys, userauth, agent
+from twisted.conch.client import agent
 from twisted.internet import defer, protocol, reactor
 from twisted.python import log
 
-import agent
-
 import os, sys, base64, getpass
 
 def verifyHostKey(transport, host, pubKey, fingerprint):
