Index: twisted/words/protocols/irc.py
===================================================================
--- twisted/words/protocols/irc.py	(revision 24327)
+++ twisted/words/protocols/irc.py	(working copy)
@@ -918,6 +918,10 @@
     def away(self, message=''):
         self.sendLine("AWAY :%s" % message)
 
+    def back(self):
+        """Clear the away status"""
+        self.away(self) # An empty away marks us as back
+
     def register(self, nickname, hostname='foo', servername='bar'):
         if self.password is not None:
             self.sendLine("PASS %s" % self.password)
