Index: twisted/web/sux.py
===================================================================
--- twisted/web/sux.py	(revision 31837)
+++ twisted/web/sux.py	(working copy)
@@ -635,23 +635,3 @@
 
         Default behaviour is to print.'''
         print 'end', name
-
-if __name__ == '__main__':
-    from cStringIO import StringIO
-    testDocument = '''
-
-    <!DOCTYPE ignore all this shit, hah its malformed!!!!@$>
-    <?xml version="suck it"?>
-    <foo>
-    &#65;
-    <bar />
-    <baz boz="buz">boz &zop;</baz>
-    <![CDATA[ foo bar baz ]]>
-    </foo>
-    '''
-    x = XMLParser()
-    x.makeConnection(FileWrapper(StringIO()))
-    # fn = "/home/glyph/Projects/Twisted/doc/howto/ipc10paper.html"
-    fn = "/home/glyph/gruesome.xml"
-    # testDocument = open(fn).read()
-    x.dataReceived(testDocument)
