diff --git a/twisted/internet/defer.py b/twisted/internet/defer.py
index 6a58bbb..f32c50a 100644
--- a/twisted/internet/defer.py
+++ b/twisted/internet/defer.py
@@ -1196,7 +1196,7 @@ def inlineCallbacks(f):
     inlineCallbacks helps you write L{Deferred}-using code that looks like a
     regular sequential function. For example::
 
-        @inlineCallBacks
+        @inlineCallbacks
         def thingummy():
             thing = yield makeSomeRequestResultingInDeferred()
             print(thing)  # the result! hoorj!
