Index: twisted/python/filepath.py
===================================================================
--- twisted/python/filepath.py	(revision 38209)
+++ twisted/python/filepath.py	(working copy)
@@ -1234,7 +1234,13 @@
 
         @return: C{None}
         """
-        return os.makedirs(self.path)
+        try:
+            os.makedirs(self.path)
+        except OSError as e:
+            if e.errno == errno.EISDIR or e.errno == errno.EEXIST:
+                pass
+            else:
+                raise OSError(e.errno, e.strerror)
 
 
     def globChildren(self, pattern):
