diff --git twisted/python/filepath.py twisted/python/filepath.py
index 2211f6a..3e5790a 100644
--- twisted/python/filepath.py
+++ twisted/python/filepath.py
@@ -1425,7 +1425,8 @@ class FilePath(AbstractFilePath):
         try:
             return os.makedirs(self.path)
         except OSError as e:
-            if not (e.errno == errno.EEXIST and ignoreAlreadyExist):
+            if not (e.errno == errno.EEXIST and self.isdir() and
+                    ignoreAlreadyExist):
                 raise
 
 
diff --git twisted/test/test_paths.py twisted/test/test_paths.py
index 00460e9..1562a9f 100644
--- twisted/test/test_paths.py
+++ twisted/test/test_paths.py
@@ -1400,6 +1400,24 @@ class FilePathTests(AbstractFilePathTests):
         self.assertEqual(exception.errno, errno.EEXIST)
 
 
+    def test_makedirsRaisesNonEexistErrorsIgnoreAlreadyExist(self):
+        """
+        When C{FilePath.makedirs} is called with ignoreAlreadyExist set to
+        C{True} it raises an C{OSError} exception if exception errno is not
+        EEXIST.
+        """
+        def faultyMakedirs(path):
+            raise OSError(errno.EACCES, b'Permission Denied')
+
+        self.patch(os, 'makedirs', faultyMakedirs)
+        fp = filepath.FilePath(self.mktemp())
+
+        exception = self.assertRaises(
+            OSError, fp.makedirs, ignoreAlreadyExist=True)
+
+        self.assertEqual(exception.errno, errno.EACCES)
+
+
     def test_changed(self):
         """
         L{FilePath.changed} indicates that the L{FilePath} has changed, but does
