Index: twisted/python/test/test_filepath.py
===================================================================
--- twisted/python/test/test_filepath.py    (revision 0)
+++ twisted/python/test/test_filepath.py    (working copy)
@@ -0,0 +1,26 @@
+# Copyright (c) Twisted Matrix Laboratories.
+# See LICENSE for details.
+
+"""
+Tests for L{twisted.python.filepath}
+"""
+
+import os
+from twisted.trial.unittest import TestCase
+from twisted.python.filepath import FilePath
+
+
+class FilePathTests(TestCase):
+    def setUp(self):
+        os.mkdir('dir1')
+    
+    def test_makedirsForced(self):
+        dir1 = FilePath('dir1')
+        dir1.makedirsForced()
+        
+    def tearDown(self):
+        os.removedirs('dir1')
+
Index: twisted/python/filepath.py
===================================================================
--- twisted/python/filepath.py  (revision 38232)
+++ twisted/python/filepath.py  (working copy)
@@ -1234,8 +1234,21 @@
 
         @return: C{None}
         """
-        return os.makedirs(self.path)
+        os.makedirs(self.path)
+        
 
+    def makedirsForced(self):
+        try:
+            os.makedirs(self.path, exists=True)
+        except TypeError:
+            try:
+                os.makedirs(self.path)
+            except OSError as e:
+                if e.errno == errno.EISDIR or e.errno == errno.EEXIST:
+                    pass
+                else:
+                    raise
+        
 
     def globChildren(self, pattern):
         """
