Index: doc/core/howto/quotes.xhtml
===================================================================
--- doc/core/howto/quotes.xhtml	(revision 32209)
+++ doc/core/howto/quotes.xhtml	(working copy)
@@ -31,9 +31,9 @@
     <li><a href="listings/TwistedQuotes/__init__.py" class="py-listing">__init__.py</a> (this
         file marks it as a package, see <a
         href="http://docs.python.org/tutorial/modules.html#packages"
-        >this section</a> of the Python tutorial for more on packages);</li>
-    <li><a href="listings/TwistedQuotes/quoters.py" class="py-listing">quoters.py</a>;</li>
-    <li><a href="listings/TwistedQuotes/quoteproto.py" class="py-listing">quoteproto.py</a>;</li>
+        >this section</a> of the Python tutorial for more on packages)</li>
+    <li><a href="listings/TwistedQuotes/quoters.py" class="py-listing">quoters.py</a></li>
+    <li><a href="listings/TwistedQuotes/quoteproto.py" class="py-listing">quoteproto.py</a></li>
     </ul>
 </li>
 <li>Add the <code>TwistedQuotes</code> directory's <em>parent</em> to your Python
