Index: twisted/lore/test/test_lore.py
===================================================================
--- twisted/lore/test/test_lore.py	(revision 43547)
+++ twisted/lore/test/test_lore.py	(working copy)
@@ -87,7 +87,7 @@
 
 
 
-class TestFactory(unittest.TestCase, XMLAssertionMixin):
+class FactoryTests(unittest.TestCase, XMLAssertionMixin):
 
     file = sp('simple.html')
     linkrel = ""
@@ -1187,7 +1187,7 @@
 
 
 
-class LatexSpitterTestCase(unittest.TestCase):
+class LatexSpitterTests(unittest.TestCase):
     """
     Tests for the Latex output plugin.
     """
Index: twisted/lore/test/test_lint.py
===================================================================
--- twisted/lore/test/test_lint.py	(revision 43547)
+++ twisted/lore/test/test_lint.py	(working copy)
@@ -15,7 +15,7 @@
 
 
 
-class DeprecationTestCase(TestCase):
+class DeprecationTests(TestCase):
     """
     Deprecations in L{twisted.lore.lint}.
     """
Index: twisted/lore/test/test_man2lore.py
===================================================================
--- twisted/lore/test/test_man2lore.py	(revision 43547)
+++ twisted/lore/test/test_man2lore.py	(working copy)
@@ -20,7 +20,7 @@
 """)
 
 
-class ManConverterTestCase(TestCase):
+class ManConverterTests(TestCase):
     """
     Tests for L{ManConverter}.
     """
Index: twisted/lore/test/test_scripts.py
===================================================================
--- twisted/lore/test/test_scripts.py	(revision 43547)
+++ twisted/lore/test/test_scripts.py	(working copy)
@@ -20,7 +20,7 @@
 
 
 
-class ZshIntegrationTestCase(TestCase, ZshScriptTestMixin):
+class ZshIntegrationTests(TestCase, ZshScriptTestMixin):
     """
     Test that zsh completion functions are generated without error
     """
Index: twisted/lore/topfiles/7729.misc
===================================================================
