=== modified file 'doc/core/howto/tutorial/configuration.xhtml'
--- doc/core/howto/tutorial/configuration.xhtml	2010-10-02 00:52:46 +0000
+++ doc/core/howto/tutorial/configuration.xhtml	2010-10-10 11:40:45 +0000
@@ -16,10 +16,9 @@
 <p>In this part, we make it easier for non-programmers to configure a finger
 server and show how to package it in the .deb and RPM package formats. Plugins
 are discussed further in the <a href="../plugin.xhtml">Twisted Plugin System</a>
-howto. .tap files are covered in <a href="../tap.xhtml">Writing a twistd
-Plugin</a>, and .tac applications are covered
-in <a href="../application.xhtml">Using the Twisted Application
-Framework</a>.</p>
+howto. Writing twistd plugins is covered in <a href="../tap.xhtml">Writing a
+twistd Plugin</a>, and .tac applications are covered in <a
+href="../application.xhtml">Using the Twisted Application Framework</a>.</p>
 
 <h2>Plugins</h2>
 
@@ -39,6 +38,9 @@
 twisted/plugins/finger_tutorial.py
 </a>
 
+<p>Note that the second argument to ServiceMaker, 'finger.tap', is a
+  reference to a module (finger/tap.py), not to a filename.</p>
+
 <p>And now, the following works</p>
 
 <pre class="shell">

