diff -rc ../Twisted-8.1.0/twisted/test/generator_failure_tests.py ./twisted/test/generator_failure_tests.py
*** ../Twisted-8.1.0/twisted/test/generator_failure_tests.py    2007-11-24 03:27:03.000000000 +0900
--- ./twisted/test/generator_failure_tests.py   2008-12-15 13:55:53.000000000 +0900
***************
*** 63,69 ****
          stuff = []
          def generator():
              try:
!                 yield
              except:
                  stuff.append(sys.exc_info())
              else:
--- 63,69 ----
          stuff = []
          def generator():
              try:
!                 yield()
              except:
                  stuff.append(sys.exc_info())
              else:
***************
*** 91,97 ****
          foundFailures = []
          def generator():
              try:
!                 yield
              except:
                  foundFailures.append(Failure._findFailure())
              else:
--- 91,97 ----
          foundFailures = []
          def generator():
              try:
!                 yield()
              except:
                  foundFailures.append(Failure._findFailure())
              else:
***************
*** 118,124 ****

          def generator():
              try:
!                 yield
              except:
                  newFailures.append(Failure())
              else:
--- 118,124 ----

          def generator():
              try:
!                 yield()
              except:
                  newFailures.append(Failure())
              else:
***************
*** 139,145 ****
          def generator():
              try:
                  try:
!                     yield
                  except:
                      [][1]
              except:
--- 139,145 ----
          def generator():
              try:
                  try:
!                     yield()
                  except:
                      [][1]
              except:
***************
*** 157,163 ****
          """
          def generator():
              try:
!                 yield
              except:
                  [][1]
          g = generator()
--- 157,163 ----
          """
          def generator():
              try:
!                 yield()
              except:
                  [][1]
          g = generator()
