diff --git twisted/runner/inetdtap.py twisted/runner/inetdtap.py
index 3e62877..4f9f8f9 100644
--- twisted/runner/inetdtap.py
+++ twisted/runner/inetdtap.py
@@ -1,8 +1,5 @@
 # Copyright (c) Twisted Matrix Laboratories.
 # See LICENSE for details.
-
-# 
-
 """
 Twisted inetd TAP support
 
@@ -22,6 +19,7 @@ try:
     import portmap
     rpcOk = 1
 except ImportError:
+    portmap  # Silence the linter.
     rpcOk = 0
 
 
@@ -49,17 +47,18 @@ class RPCServer(internet.TCPServer):
         self.rpcConf = rpcConf
         self.proto = proto
         self.service = service
+        self.rpcVersions = rpcVersions
 
     def startService(self):
         internet.TCPServer.startService(self)
         import portmap
         portNo = self._port.getHost()[2]
         service = self.service
-        for version in rpcVersions:
+        for version in self.rpcVersions:
             portmap.set(self.rpcConf.services[name], version, self.proto,
                         portNo)
             inetd.forkPassingFD(service.program, service.programArgs,
-                                os.environ, service.user, service.group, p)
+                                os.environ, service.user, service.group)
 
 def makeService(config):
     s = appservice.MultiService()
diff --git twisted/runner/procmon.py twisted/runner/procmon.py
index 3515995..28be909 100644
--- twisted/runner/procmon.py
+++ twisted/runner/procmon.py
@@ -5,8 +5,6 @@
 """
 Support for starting, monitoring, and restarting child process.
 """
-import warnings
-
 from twisted.python import log
 from twisted.internet import error, protocol, reactor as _reactor
 from twisted.application import service
