Index: twisted/web/util.py
===================================================================
--- twisted/web/util.py	(révision 27231)
+++ twisted/web/util.py	(copie de travail)
@@ -97,13 +97,7 @@
         return NOT_DONE_YET
 
     def _cbChild(self, child, request):
-        result = resource.getChildForRequest(child, request).render(request)
-        from twisted.web.server import NOT_DONE_YET
-        if result == NOT_DONE_YET:
-            return
-        else:
-            request.write(result)
-            request.finish()
+        request.render(resource.getChildForRequest(child, request))
 
     def _ebChild(self, reason, request):
         request.processingFailed(reason)
