--- imap.py	Fri Oct  1 15:58:29 2004
+++ imap.new	Fri Oct  1 15:59:26 2004
@@ -3930,7 +3930,14 @@
     def challengeResponse(self, secret, chal):
         if chal == 'User Name\0':
             return self.user
+
+        elif chal == 'Username:':
+            return self.user
+
         elif chal == 'Password\0':
+            return secret
+
+        elif chal == 'Password:':
             return secret
 
 components.backwardsCompatImplements(LOGINAuthenticator)
