Index: twisted/internet/test/test_tls.py
===================================================================
--- twisted/internet/test/test_tls.py	(revision 31475)
+++ twisted/internet/test/test_tls.py	(working copy)
@@ -18,10 +18,14 @@
 try:
     from OpenSSL.crypto import FILETYPE_PEM
 except ImportError:
-    FILETYPE_PEM = None
+    tls = FILETYPE_PEM = None
 else:
     from twisted.internet.ssl import PrivateCertificate, KeyPair
     from twisted.internet.ssl import ClientContextFactory
+    try:
+        from twisted.protocols import tls
+    except ImportError:
+        tls = None
 
 
 
@@ -32,7 +36,7 @@
     if FILETYPE_PEM is None:
         skip = "OpenSSL is unavailable"
 
-    if platform.isWindows():
+    if platform.isWindows() and tls is None:
         msg = (
             "For some reason, these reactors don't deal with SSL "
             "disconnection correctly on Windows.  See #3371.")
