Index: twisted/protocols/ftp.py
===================================================================
--- twisted/protocols/ftp.py	(revision 14850)
+++ twisted/protocols/ftp.py	(working copy)
@@ -1117,6 +1117,7 @@
 
 
     def ftp_QUIT(self):
+        self.reply(GOODBYE_MSG)
         self.transport.loseConnection()
         self.disconnected = True
 
Index: twisted/test/test_ftp.py
===================================================================
--- twisted/test/test_ftp.py	(revision 14850)
+++ twisted/test/test_ftp.py	(working copy)
@@ -167,6 +167,15 @@
     def testAnonymousLogin(self):
         self._anonymousLogin()
 
+    def testQuit(self):
+        """Issuing QUIT should return a 221 message."""
+        self._anonymousLogin()
+        responseLines = wait(self.client.queueStringCommand('QUIT'))
+        self.assertEquals(
+            ['221 Goodbye.'],
+            responseLines
+        )
+
     def testAnonymousLoginDenied(self):
         # Reconfigure the server to disallow anonymous access, and to have an
         # IUsernamePassword checker that always rejects.
