Index: twisted/web/xmlrpc.py
===================================================================
--- twisted/web/xmlrpc.py	(revision 17374)
+++ twisted/web/xmlrpc.py	(working copy)
@@ -92,9 +92,10 @@
     isLeaf = 1
     separator = '.'
 
-    def __init__(self):
+    def __init__(self, allowNone=False):
         resource.Resource.__init__(self)
         self.subHandlers = {}
+        self.allowNone = allowNone
 
     def putSubHandler(self, prefix, handler):
         self.subHandlers[prefix] = handler
@@ -127,10 +128,12 @@
         if not isinstance(result, Fault):
             result = (result,)
         try:
-            s = xmlrpclib.dumps(result, methodresponse=1)
+            s = xmlrpclib.dumps(result, methodresponse=True,
+                                allow_none=self.allowNone)
         except:
             f = Fault(self.FAILURE, "can't serialize output")
-            s = xmlrpclib.dumps(f, methodresponse=1)
+            s = xmlrpclib.dumps(f, methodresponse=True,
+                                allow_none=self.allowNone)
         request.setHeader("content-length", str(len(s)))
         request.write(s)
         request.finish()
