--- doc/examples/pbsimple.py	2001-12-04 15:47:00.000000000 -0600
+++ ../Twisted.wip/doc/examples/pbsimple.py	2003-09-24 12:25:37.000000000 -0500
@@ -15,13 +15,13 @@
 # License along with this library; if not, write to the Free Software
 # Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 
+from twisted.internet import reactor
 from twisted.spread import pb
-from twisted.internet import app
+
 class Echoer(pb.Root):
     def remote_echo(self, st):
         print 'echoing:', st
         return st
-if __name__ == '__main__':
-    appl = app.Application("pbsimple")
-    appl.listenTCP(8789, pb.BrokerFactory(Echoer()))
-    appl.run()
+
+reactor.listenTCP(8789, pb.PBServerFactory(Echoer()))
+reactor.run()
