=== modified file 'twisted/internet/base.py'
--- twisted/internet/base.py	2009-06-01 18:07:14 +0000
+++ twisted/internet/base.py	2009-10-29 08:45:58 +0000
@@ -17,6 +17,7 @@
 import traceback
 
 from twisted.python.compat import set
+from twisted.python.compat3 import reduce
 from twisted.python.util import unsignedID
 from twisted.internet.interfaces import IReactorCore, IReactorTime, IReactorThreads
 from twisted.internet.interfaces import IResolverSimple, IReactorPluggableResolver

=== modified file 'twisted/names/srvconnect.py'
--- twisted/names/srvconnect.py	2007-08-24 08:44:59 +0000
+++ twisted/names/srvconnect.py	2009-10-29 08:46:33 +0000
@@ -10,6 +10,7 @@
 
 from twisted.names import client, dns
 from twisted.names.error import DNSNameError
+from twisted.python.compat3 import reduce
 
 class _SRVConnector_ClientFactoryWrapper:
     def __init__(self, connector, wrappedFactory):

=== modified file 'twisted/names/test/test_names.py'
--- twisted/names/test/test_names.py	2008-09-12 20:48:33 +0000
+++ twisted/names/test/test_names.py	2009-10-29 08:46:59 +0000
@@ -23,6 +23,7 @@
 from twisted.names.client import Resolver
 
 from twisted.names.test.test_client import StubPort
+from twisted.python.compat3 import reduce
 
 def justPayload(results):
     return [r.payload for r in results[0]]

=== modified file 'twisted/protocols/ftp.py'
--- twisted/protocols/ftp.py	2009-01-09 18:17:14 +0000
+++ twisted/protocols/ftp.py	2009-10-29 08:47:41 +0000
@@ -33,6 +33,7 @@
 from twisted.protocols import basic, policies
 
 from twisted.python import log, failure, filepath
+from twisted.python.compat3 import reduce
 
 from twisted.cred import error as cred_error, portal, credentials, checkers
 

=== modified file 'twisted/protocols/gps/nmea.py'
--- twisted/protocols/gps/nmea.py	2008-07-29 20:13:54 +0000
+++ twisted/protocols/gps/nmea.py	2009-10-29 08:48:09 +0000
@@ -21,6 +21,7 @@
 
 import operator
 from twisted.protocols import basic
+from twisted.python.compat3 import reduce
 
 POSFIX_INVALID, POSFIX_SPS, POSFIX_DGPS, POSFIX_PPS = 0, 1, 2, 3
 MODE_AUTO, MODE_FORCED = 'A', 'M'

=== modified file 'twisted/protocols/pcp.py'
--- twisted/protocols/pcp.py	2006-07-01 16:08:17 +0000
+++ twisted/protocols/pcp.py	2009-10-29 08:48:30 +0000
@@ -13,6 +13,7 @@
 from zope.interface import implements
 
 from twisted.internet import interfaces
+from twisted.python.compat3 import reduce
 
 
 class BasicProducerConsumerProxy:

=== added file 'twisted/python/compat3.py'
--- twisted/python/compat3.py	1970-01-01 00:00:00 +0000
+++ twisted/python/compat3.py	2009-10-29 09:02:14 +0000
@@ -0,0 +1,17 @@
+# -*- test-case-name: twisted.test.test_compat3 -*-
+#
+# Copyright (c) 2009 Twisted Matrix Laboratories.
+# See LICENSE for details.
+
+
+"""
+Compatibility module to provide backwards compatibility for Python features
+missing from Python 3.x.
+
+This is mainly for use of internal Twisted code.
+"""
+
+try:
+    from functools import reduce
+except ImportError:
+    reduce = reduce

=== modified file 'twisted/spread/jelly.py'
--- twisted/spread/jelly.py	2009-04-23 16:25:50 +0000
+++ twisted/spread/jelly.py	2009-10-29 08:49:06 +0000
@@ -114,6 +114,7 @@
 from twisted.persisted.crefutil import _DictKeyAndValue, _Dereference
 from twisted.persisted.crefutil import _Container
 from twisted.python import runtime
+from twisted.python.compat3 import reduce
 
 from twisted.spread.interfaces import IJellyable, IUnjellyable
 

=== added file 'twisted/test/test_compat3.py'
--- twisted/test/test_compat3.py	1970-01-01 00:00:00 +0000
+++ twisted/test/test_compat3.py	2009-10-29 09:35:26 +0000
@@ -0,0 +1,15 @@
+# Copyright (c) 2009 Twisted Matrix Laboratories.
+# See LICENSE for details.
+
+"""
+Tests for L{twisted.python.compat3}.
+"""
+
+from twisted.python.compat3 import reduce
+from twisted.trial import unittest
+
+
+class CompatTestCase(unittest.TestCase):
+    def test_reduce(self):
+        self.assertEquals(15, reduce(lambda x, y: x + y, [1, 2, 3, 4, 5]))
+        self.assertEquals(16, reduce(lambda x, y: x + y, [1, 2, 3, 4, 5], 1))

=== modified file 'twisted/words/test/test_toc.py'
--- twisted/words/test/test_toc.py	2009-02-14 21:04:19 +0000
+++ twisted/words/test/test_toc.py	2009-10-29 08:49:24 +0000
@@ -7,6 +7,7 @@
 from twisted.words.protocols import toc
 from twisted.internet import protocol, main
 from twisted.python import failure
+from twisted.python.compat3 import reduce
 from twisted.test.proto_helpers import StringIOWithoutClosing
 
 from struct import pack,unpack

