Index: twisted/topfiles/6498.misc
===================================================================
Index: twisted/internet/_sslverify.py
===================================================================
--- twisted/internet/_sslverify.py	(revision 38313)
+++ twisted/internet/_sslverify.py	(working copy)
@@ -601,7 +601,6 @@
         Sign a CertificateRequest instance, returning a Certificate instance.
         """
         req = requestObject.original
-        dn = requestObject.getSubject()
         cert = crypto.X509()
         issuerDistinguishedName._copyInto(cert.get_issuer())
         cert.set_subject(req.get_subject())
