Index: twisted/python/components.py
===================================================================
--- twisted/python/components.py	(revision 37642)
+++ twisted/python/components.py	(working copy)
@@ -245,8 +245,6 @@
         Otherwise, an 'appropriate' interface is one for which its class has
         been registered as an adapter for my class according to the rules of
         getComponent.
-
-        @return: the list of appropriate interfaces
         """
         for iface in declarations.providedBy(component):
             if (ignoreClass or
