Index: filepath.py
===================================================================
--- filepath.py	(révision 32383)
+++ filepath.py	(copie de travail)
@@ -874,18 +874,27 @@
             else:
                 return False
 
-
-    def isdir(self):
+    def commonIs(self):
         """
-        @return: C{True} if this L{FilePath} refers to a directory, C{False}
-            otherwise.
+        Common code to is* methods
         """
         st = self.statinfo
         if not st:
             self.restat(False)
             st = self.statinfo
-            if not st:
+            if not st :
                 return False
+        return st
+
+
+    def isdir(self):
+        """
+        @return: C{True} if this L{FilePath} refers to a directory, C{False}
+            otherwise.
+        """
+        st = self.commonIs()
+        if st == False :
+            return False
         return S_ISDIR(st.st_mode)
 
 
@@ -894,12 +903,9 @@
         @return: C{True} if this L{FilePath} points to a regular file (not a
             directory, socket, named pipe, etc), C{False} otherwise.
         """
-        st = self.statinfo
-        if not st:
-            self.restat(False)
-            st = self.statinfo
-            if not st:
-                return False
+        st = self.commonIs()
+        if st == False :
+            return False
         return S_ISREG(st.st_mode)
 
 
@@ -911,12 +917,9 @@
         @rtype: C{bool}
         @since: 11.1
         """
-        st = self.statinfo
-        if not st:
-            self.restat(False)
-            st = self.statinfo
-            if not st:
-                return False
+        st = self.commonIs()
+        if st == False :
+            return False
         return S_ISBLK(st.st_mode)
 
 
@@ -928,12 +931,9 @@
         @rtype: C{bool}
         @since: 11.1
         """
-        st = self.statinfo
-        if not st:
-            self.restat(False)
-            st = self.statinfo
-            if not st:
-                return False
+        st = self.commonIs()
+        if st == False :
+            return False
         return S_ISSOCK(st.st_mode)
 
 
