Index: twisted/web/microdom.py
===================================================================
RCS file: /cvs/Twisted/twisted/web/microdom.py,v
retrieving revision 1.101
diff -u -r1.101 microdom.py
--- twisted/web/microdom.py	13 Sep 2003 03:13:33 -0000	1.101
+++ twisted/web/microdom.py	19 Oct 2003 09:52:49 -0000
@@ -342,15 +342,6 @@
 class _Attr(CharacterData):
     "Support class for getAttributeNode."
 
-def _selectDict(dct, caseInsensitive, preserveCase):
-    """Return a copy of the dct with case sensitivity/preserving 
-    turned on as appropriate.
-    """
-    if caseInsensitive:
-        return InsensitiveDict(dct, preserve=preserveCase)
-    else:
-        return dct
-
 class Element(Node):
 
     preserveCase = 0
@@ -374,8 +365,9 @@
             for k, v in self.attributes.items():
                 self.attributes[k] = unescape(v)
 
-        self.attributes=_selectDict(self.attributes, caseInsensitive, 
-                                    preserveCase)
+        if caseInsensitive:
+            self.attributes = InsensitiveDict(self.attributes, 
+                                              preserve=preserveCase)
 
         self.endTagName = self.nodeName = self.tagName = tagName
         self._filename = filename
