Index: imap4.py
===================================================================
--- imap4.py	(Revision 15571)
+++ imap4.py	(Arbeitskopie)
@@ -2056,10 +2056,14 @@
             lastPart = line.rfind(' ')
             if lastPart != -1:
                 lastPart = line[lastPart + 1:]
-                if lastPart.startswith('{') and lastPart.endswith('}'):
+                if lastPart.endswith('}'):
                     # It's a literal a-comin' in
+                    s = lastPart.rfind("{")
+                    if s == -1:
+                        # no matching '{' found
+                        raise IllegalServerResponse(line)
                     try:
-                        octets = int(lastPart[1:-1])
+                        octets = int(lastPart[s + 1:-1])
                     except ValueError:
                         raise IllegalServerResponse(line)
                     self._tag, parts = line.split(None, 1)
