Index: twisted/trial/reporter.py
===================================================================
--- twisted/trial/reporter.py	(revision 22758)
+++ twisted/trial/reporter.py	(working copy)
@@ -743,6 +743,8 @@
         # isatty() returns False when SSHd into Win32 machine
         if 'CYGWIN' in os.environ:
             return True
+        if not hasattr(sys.stderr, 'isatty'):
+            return False # isatty may not be available. (wrapped output)
         if not sys.stderr.isatty():
             return False # auto color only on TTYs
         try:
