Index: twisted/python/log.py
===================================================================
--- twisted/python/log.py	(revision 22758)
+++ twisted/python/log.py	(working copy)
@@ -487,8 +487,10 @@
     def writelines(self, lines):
         for line in lines:
             msg(line, printed=1, isError=self.isError)
+            
+    def isatty(self):
+        return False
 
-
 try:
     _oldshowwarning
 except NameError:
