Index: doc/lore/howto/lore.xhtml
===================================================================
--- doc/lore/howto/lore.xhtml	(revision 24967)
+++ doc/lore/howto/lore.xhtml	(working copy)
@@ -252,7 +252,7 @@
 checkout, do:</p>
 
 <pre class="shell">
-% lore --config template=doc/howto/template.tpl doc/howto/*.xhtml
+% lore --config template=doc/core/howto/template.tpl doc/core/howto/*.xhtml
 </pre>
 
 <p>
