Index: twisted/test/test_plugin.py
===================================================================
--- twisted/test/test_plugin.py	(revision 30995)
+++ twisted/test/test_plugin.py	(working copy)
@@ -322,8 +322,13 @@
         # Windows report OSError, others IOError
         errors[0].trap(OSError, IOError)
 
+        warning_list = self.flushWarnings()
+        self.assertEqual(len(warning_list), 1)
+        self.assertRegexpMatches(str(warning_list),
+                          r'Unable to write to plugin cache')
 
 
+
 # This is something like the Twisted plugins file.
 pluginInitFile = """
 from twisted.plugin import pluginPackagePaths
@@ -537,8 +542,13 @@
         # Sanity check to make sure we're only flushing the error logged
         # below...
         self.assertEqual(len(self.flushLoggedErrors()), 0)
+        self.assertEqual(len(self.flushWarnings()), 0)
         self.assertIn('one', self.getAllPlugins())
         self.assertEqual(len(self.flushLoggedErrors()), 1)
+        warning_list = self.flushWarnings()
+        self.assertEqual(len(warning_list), 1)
+        self.assertRegexpMatches(str(warning_list),
+                          r'Unable to write to plugin cache')
 
 
 
Index: twisted/plugin.py
===================================================================
--- twisted/plugin.py	(revision 30995)
+++ twisted/plugin.py	(working copy)
@@ -12,6 +12,7 @@
 
 import os
 import sys
+import warnings
 
 from zope.interface import Interface, providedBy
 
@@ -178,6 +179,8 @@
                 dropinPath.setContent(pickle.dumps(dropinDotCache))
             except:
                 log.err()
+                warnings.warn("Unable to write to plugin cache " + dropinPath.path,
+                  category=RuntimeWarning, stacklevel=1)
         allCachesCombined.update(dropinDotCache)
     return allCachesCombined
 
Index: twisted/topfiles/2409.feature
===================================================================
--- twisted/topfiles/2409.feature	(revision 0)
+++ twisted/topfiles/2409.feature	(revision 0)
@@ -0,0 +1 @@
+twisted.plugin will now issue a warning when it is unable to write to the plugin cache.
