Index: twisted/web/client.py
===================================================================
--- twisted/web/client.py	(revision 47126)
+++ twisted/web/client.py	(working copy)
@@ -2020,7 +2020,9 @@
             err = error.RedirectWithNoLocation(
                 response.code, b'No location header field', uri)
             raise ResponseFailed([Failure(err)], response)
-        location = self._resolveLocation(uri, locationHeaders[0])
+        location = self._resolveLocation(
+            response.request.absoluteURI,
+            locationHeaders[0])
         deferred = self._agent.request(method, location, headers)
         def _chainResponse(newResponse):
             newResponse.setPreviousResponse(response)
