diff --git "a/C:\\Users\\MATTHO~1.000\\AppData\\Local\\Temp\\TortoiseGit\\makE37C.tmp\\make-2fc481e-left.bat" "b/C:\\Users\\matt.HOME.000\\twistedworkspace\\twisted\\docs\\make.bat"
index 7264bd8..07fb2fd 100644
--- "a/C:\\Users\\MATTHO~1.000\\AppData\\Local\\Temp\\TortoiseGit\\makE37C.tmp\\make-2fc481e-left.bat"
+++ "b/C:\\Users\\matt.HOME.000\\twistedworkspace\\twisted\\docs\\make.bat"
@@ -4,7 +4,7 @@ REM Command file for Sphinx documentation
 
 set SPHINXBUILD=sphinx-build
 set BUILDDIR=build
-set ALLSPHINXOPTS=-d %BUILDDIR%/doctrees %SPHINXOPTS% source
+set ALLSPHINXOPTS=-d %BUILDDIR%/doctrees %SPHINXOPTS% .
 if NOT "%PAPER%" == "" (
 	set ALLSPHINXOPTS=-D latex_paper_size=%PAPER% %ALLSPHINXOPTS%
 )
