Index: docs/core/howto/tutorial/listings/finger/finger/finger.py
===================================================================
--- docs/core/howto/tutorial/listings/finger/finger/finger.py	(revision 46984)
+++ docs/core/howto/tutorial/listings/finger/finger/finger.py	(working copy)
@@ -290,7 +290,7 @@
 
     def _read(self):
         self.users = {}
-        for line in file(self.filename):
+        for line in open(self.filename):
             user, status = line.split(':', 1)
             user = user.strip()
             status = status.strip()
Index: docs/core/howto/tutorial/listings/finger/finger14.tac
===================================================================
--- docs/core/howto/tutorial/listings/finger/finger14.tac	(revision 46984)
+++ docs/core/howto/tutorial/listings/finger/finger14.tac	(working copy)
@@ -23,7 +23,7 @@
         self.filename = filename
 
     def _read(self):
-        for line in file(self.filename):
+        for line in open(self.filename):
             user, status = line.split(':', 1)
             user = user.strip()
             status = status.strip()
Index: docs/core/howto/tutorial/listings/finger/finger15.tac
===================================================================
--- docs/core/howto/tutorial/listings/finger/finger15.tac	(revision 46984)
+++ docs/core/howto/tutorial/listings/finger/finger15.tac	(working copy)
@@ -48,7 +48,7 @@
 
     def _read(self):
         self.users.clear()
-        for line in file(self.filename):
+        for line in open(self.filename):
             user, status = line.split(':', 1)
             user = user.strip()
             status = status.strip()
Index: docs/core/howto/tutorial/listings/finger/finger16.tac
===================================================================
--- docs/core/howto/tutorial/listings/finger/finger16.tac	(revision 46984)
+++ docs/core/howto/tutorial/listings/finger/finger16.tac	(working copy)
@@ -47,7 +47,7 @@
 
     def _read(self):
         self.users.clear()
-        for line in file(self.filename):
+        for line in open(self.filename):
             user, status = line.split(':', 1)
             user = user.strip()
             status = status.strip()
Index: docs/core/howto/tutorial/listings/finger/finger17.tac
===================================================================
--- docs/core/howto/tutorial/listings/finger/finger17.tac	(revision 46984)
+++ docs/core/howto/tutorial/listings/finger/finger17.tac	(working copy)
@@ -46,7 +46,7 @@
 
     def _read(self):
         self.users.clear()
-        for line in file(self.filename):
+        for line in open(self.filename):
             user, status = line.split(':', 1)
             user = user.strip()
             status = status.strip()
Index: docs/core/howto/tutorial/listings/finger/finger18.tac
===================================================================
--- docs/core/howto/tutorial/listings/finger/finger18.tac	(revision 46984)
+++ docs/core/howto/tutorial/listings/finger/finger18.tac	(working copy)
@@ -94,7 +94,7 @@
 
     def _read(self):
         self.users.clear()
-        for line in file(self.filename):
+        for line in open(self.filename):
             user, status = line.split(':', 1)
             user = user.strip()
             status = status.strip()
Index: docs/core/howto/tutorial/listings/finger/finger19.tac
===================================================================
--- docs/core/howto/tutorial/listings/finger/finger19.tac	(revision 46984)
+++ docs/core/howto/tutorial/listings/finger/finger19.tac	(working copy)
@@ -234,7 +234,7 @@
 
     def _read(self):
         self.users.clear()
-        for line in file(self.filename):
+        for line in open(self.filename):
             user, status = line.split(':', 1)
             user = user.strip()
             status = status.strip()
Index: docs/core/howto/tutorial/listings/finger/finger19b.tac
===================================================================
--- docs/core/howto/tutorial/listings/finger/finger19b.tac	(revision 46984)
+++ docs/core/howto/tutorial/listings/finger/finger19b.tac	(working copy)
@@ -243,7 +243,7 @@
 
     def _read(self):
         self.users.clear()
-        for line in file(self.filename):
+        for line in open(self.filename):
             user, status = line.split(':', 1)
             user = user.strip()
             status = status.strip()
Index: docs/core/howto/tutorial/listings/finger/finger19c.tac
===================================================================
--- docs/core/howto/tutorial/listings/finger/finger19c.tac	(revision 46984)
+++ docs/core/howto/tutorial/listings/finger/finger19c.tac	(working copy)
@@ -245,7 +245,7 @@
 
     def _read(self):
         self.users.clear()
-        for line in file(self.filename):
+        for line in open(self.filename):
             user, status = line.split(':', 1)
             user = user.strip()
             status = status.strip()
Index: docs/core/howto/tutorial/listings/finger/finger20.tac
===================================================================
--- docs/core/howto/tutorial/listings/finger/finger20.tac	(revision 46984)
+++ docs/core/howto/tutorial/listings/finger/finger20.tac	(working copy)
@@ -249,7 +249,7 @@
 
     def _read(self):
         self.users.clear()
-        for line in file(self.filename):
+        for line in open(self.filename):
             user, status = line.split(':', 1)
             user = user.strip()
             status = status.strip()
Index: docs/core/howto/tutorial/listings/finger/finger21.tac
===================================================================
--- docs/core/howto/tutorial/listings/finger/finger21.tac	(revision 46984)
+++ docs/core/howto/tutorial/listings/finger/finger21.tac	(working copy)
@@ -281,7 +281,7 @@
 
     def _read(self):
         self.users.clear()
-        for line in file(self.filename):
+        for line in open(self.filename):
             user, status = line.split(':', 1)
             user = user.strip()
             status = status.strip()
Index: docs/core/howto/tutorial/listings/finger/finger22.py
===================================================================
--- docs/core/howto/tutorial/listings/finger/finger22.py	(revision 46984)
+++ docs/core/howto/tutorial/listings/finger/finger22.py	(working copy)
@@ -281,7 +281,7 @@
 
     def _read(self):
         self.users.clear()
-        for line in file(self.filename):
+        for line in open(self.filename):
             user, status = line.split(':', 1)
             user = user.strip()
             status = status.strip()
