--- msn.py	Thu Jun 30 18:16:26 2005
+++ msn.py.mine	Thu Jun 30 18:16:22 2005
@@ -1800,6 +1800,9 @@
                     continue
             if self._checkFileInvitation(message, info) or self._checkFileInfo(message, info) or self._checkFileResponse(message, info):
                 return 0
+        elif 'text/x-clientcaps' in cTypes:
+            # do something with capabilities
+            return 0
         return 1
 
     # negotiation
