--- test_msn.py	Thu Jun 30 18:14:49 2005
+++ test_msn.py.mine	Thu Jun 30 18:14:44 2005
@@ -255,6 +255,12 @@
     def tearDown(self):
         self.client = None
 
+    def testClientCapabilitiesCheck(self):
+        m = msn.MSNMessage()
+        m.setHeader('Content-Type', 'text/x-clientcaps')
+        self.client.checkMessage(m)
+        self.failUnless((self.client.checkMessage(m) == 0), msg='Failed to detect client capability message')
+        
     def testTypingCheck(self):
         m = msn.MSNMessage()
         m.setHeader('Content-Type', 'text/x-msmsgscontrol')
