Index: doc/conch/index.xhtml
===================================================================
--- doc/conch/index.xhtml	(revision 27366)
+++ doc/conch/index.xhtml	(working copy)
@@ -1,5 +1,5 @@
 <html xmlns="http://www.w3.org/1999/xhtml"><head><title>Twisted Conch Documentation</title></head><body>
-<h1>Twisted Conch</h1>
+<h1>Twisted Conch Documentation</h1>
 
 <ul>
 <li><a href="howto/index.xhtml">Developer guides</a>: documentation on using
Index: doc/core/howto/application.xhtml
===================================================================
--- doc/core/howto/application.xhtml	(revision 27366)
+++ doc/core/howto/application.xhtml	(working copy)
@@ -14,11 +14,11 @@
 
 <h3>Audience</h3>
 
-The target audience of this document is a Twisted user who wants to deploy a
+<p>The target audience of this document is a Twisted user who wants to deploy a
 significant amount of Twisted code in a re-usable, standard and easily
 configurable fashion.  A Twisted user who wishes to use the Application
 framework needs to be familiar with developing Twisted <a
-href="servers.xhtml">servers</a> and/or <a href="clients.xhtml">clients</a>.
+href="servers.xhtml">servers</a> and/or <a href="clients.xhtml">clients</a></p>.
 
 <h3>Goals</h3>
 
Index: doc/core/howto/ssl.xhtml
===================================================================
--- doc/core/howto/ssl.xhtml	(revision 27366)
+++ doc/core/howto/ssl.xhtml	(working copy)
@@ -6,7 +6,8 @@
     <title>Using SSL in Twisted</title>
   </head>
   <body>
-
+    <h1>Using SSL in Twisted</h1>
+    
     <h2>Overview</h2>
 
     <p>This document describes how to use SSL in Twisted servers and clients. It
Index: doc/core/index.xhtml
===================================================================
--- doc/core/index.xhtml	(revision 27366)
+++ doc/core/index.xhtml	(working copy)
@@ -1,5 +1,5 @@
 <html xmlns="http://www.w3.org/1999/xhtml"><head><title>Twisted Core Documentation</title></head><body>
-<h1>Start</h1>
+<h1>Twisted Core Documentation</h1>
 
 <ul>
 <li><a href="howto/index.xhtml">Developer guides</a>: documentation on using
Index: doc/core/upgrades/2.0/components.xhtml
===================================================================
--- doc/core/upgrades/2.0/components.xhtml	(revision 27366)
+++ doc/core/upgrades/2.0/components.xhtml	(working copy)
@@ -1,4 +1,4 @@
-<html>
+<html xmlns="http://www.w3.org/1999/xhtml">
 
 <head>
 <title>The Twisted Zope Interfaces FAQ</title>
Index: doc/core/upgrades/2.0/split.xhtml
===================================================================
--- doc/core/upgrades/2.0/split.xhtml	(revision 27366)
+++ doc/core/upgrades/2.0/split.xhtml	(working copy)
@@ -1,4 +1,4 @@
-<html>
+<html xmlns="http://www.w3.org/1999/xhtml">
 <head>
 <title>The Twisted Split FAQ</title>
 </head>
Index: doc/lore/index.xhtml
===================================================================
--- doc/lore/index.xhtml	(revision 27366)
+++ doc/lore/index.xhtml	(working copy)
@@ -1,5 +1,5 @@
 <html xmlns="http://www.w3.org/1999/xhtml"><head><title>Twisted Lore Documentation</title></head><body>
-<h1>Twisted Lore</h1>
+<h1>Twisted Lore Documentation</h1>
 
 <ul>
 <li><a href="howto/index.xhtml">Developer guides</a>: documentation on using
Index: doc/mail/index.xhtml
===================================================================
--- doc/mail/index.xhtml	(revision 27366)
+++ doc/mail/index.xhtml	(working copy)
@@ -1,5 +1,5 @@
 <html xmlns="http://www.w3.org/1999/xhtml"><head><title>Twisted Mail Documentation</title></head><body>
-<h1>Twisted Mail</h1>
+<h1>Twisted Mail Documentation</h1>
 
 <ul>
 <li><a href="examples/index.xhtml">Examples</a>: short code examples using
Index: doc/names/index.xhtml
===================================================================
--- doc/names/index.xhtml	(revision 27366)
+++ doc/names/index.xhtml	(working copy)
@@ -1,5 +1,5 @@
 <html xmlns="http://www.w3.org/1999/xhtml"><head><title>Twisted Names Documentation</title></head><body>
-<h1>Twisted Names</h1>
+<h1>Twisted Names Documentation</h1>
 
 <ul>
 <li><a href="howto/index.xhtml">Developer guides</a>: documentation on using
Index: doc/pair/index.xhtml
===================================================================
--- doc/pair/index.xhtml	(revision 27366)
+++ doc/pair/index.xhtml	(working copy)
@@ -1,5 +1,5 @@
 <html xmlns="http://www.w3.org/1999/xhtml"><head><title>Twisted Pair Documentation</title></head><body>
-<h1>Twisted Pair</h1>
+<h1>Twisted Pair Documentation</h1>
 
 <ul>
 <li><a href="howto/index.xhtml">Developer guides</a>: documentation on using
Index: doc/vfs/index.xhtml
===================================================================
--- doc/vfs/index.xhtml	(revision 27366)
+++ doc/vfs/index.xhtml	(working copy)
@@ -1,5 +1,5 @@
 <html xmlns="http://www.w3.org/1999/xhtml"><head><title>Twisted VFS Documentation</title></head><body>
-<h1>Twisted VFS</h1>
+<h1>Twisted VFS Documentation</h1>
 
 <ul>
 <li><a href="examples/index.xhtml">Examples</a>: short code examples using
Index: doc/web2/index.xhtml
===================================================================
--- doc/web2/index.xhtml	(revision 27366)
+++ doc/web2/index.xhtml	(working copy)
@@ -1,5 +1,5 @@
 <html xmlns="http://www.w3.org/1999/xhtml"><head><title>Twisted Web2 Documentation</title></head><body>
-<h1>Twisted Web2</h1>
+<h1>Twisted Web2 Documentation</h1>
 
 <ul>
 <li><a href="howto/index.xhtml">Developer guides</a>: documentation on using
Index: doc/web/index.xhtml
===================================================================
--- doc/web/index.xhtml	(revision 27366)
+++ doc/web/index.xhtml	(working copy)
@@ -1,5 +1,5 @@
 <html xmlns="http://www.w3.org/1999/xhtml"><head><title>Twisted Web Documentation</title></head><body>
-<h1>Twisted Web</h1>
+<h1>Twisted Web Documentation</h1>
 
 <ul>
 <li><a href="howto/index.xhtml">Developer guides</a>: documentation on using
Index: doc/words/index.xhtml
===================================================================
--- doc/words/index.xhtml	(revision 27366)
+++ doc/words/index.xhtml	(working copy)
@@ -1,5 +1,5 @@
 <html xmlns="http://www.w3.org/1999/xhtml"><head><title>Twisted Words Documentation</title></head><body>
-<h1>Twisted Words</h1>
+<h1>Twisted Words Documentation</h1>
 
 <ul>
 <li><a href="howto/index.xhtml">Developer guides</a>: documentation on using
