diff --git twisted/pair/testing.py twisted/pair/testing.py
index daefd3e..615520a 100644
--- twisted/pair/testing.py
+++ twisted/pair/testing.py
@@ -507,7 +507,7 @@ class MemoryIOSystem(object):
             received.
         @type port: L{int}
 
-        @return: A L{socket.socket}-like object which can be used to receive
+        @return: A C{socket.socket}-like object which can be used to receive
             the specified datagram.
         """
         return _FakePort(self, fileno)
@@ -529,11 +529,11 @@ class _FakePort(object):
         Receive a datagram sent to this port using the L{MemoryIOSystem} which
         created this object.
 
-        This behaves like L{socket.socket.recv} but the data being I{sent} and
+        This behaves like C{socket.socket.recv} but the data being I{sent} and
         I{received} only passes through various memory buffers managed by this
         object and L{MemoryIOSystem}.
 
-        @see: L{socket.socket.recv}
+        @see: C{socket.socket.recv}
         """
         data = self._system._openFiles[self._fileno].writeBuffer.popleft()
 
diff --git twisted/pair/tuntap.py twisted/pair/tuntap.py
index afbddca..2102dc7 100644
--- twisted/pair/tuntap.py
+++ twisted/pair/tuntap.py
@@ -195,7 +195,7 @@ class _IInputOutputSystem(Interface):
             received.
         @type port: L{int}
 
-        @return: A L{socket.socket} which can be used to receive the specified
+        @return: A C{socket.socket} which can be used to receive the specified
             datagram.
         """
 
