Index: twisted/conch/test/test_recvline.py                                                                                                                              
===================================================================
--- twisted/conch/test/test_recvline.py (revision 12962)                                                                                                                
+++ twisted/conch/test/test_recvline.py (working copy)                                                                                                                  
@@ -478,8 +478,22 @@
         exe = sys.executable                                                                                                                                           
         module = stdio.__file__                                                                                                                                        
         args = ["python2.3", module, reflect.qual(self.serverProtocol)]                                                                                                
-        env = {"PYTHONPATH": os.pathsep.join(("..", os.environ.get("PYTHONPATH", "")))}                                                                                
                                                                                                                                                                        
+        pythonpath = os.environ.get("PYTHONPATH", "")                                                                                                                  
+        if pythonpath:                                                                                                                                                 
+            paths = pythonpath.split(os.pathsep)                                                                                                                       
+            abspaths = []                                                                                                                                              
+            for p in paths:                                                                                                                                            
+                if os.path.isabs(p):                                                                                                                                   
+                    abspaths.append(p)                                                                                                                                 
+                else:                                                                                                                                                  
+                    abspaths.append(os.path.join("..", p))                                                                                                             
+            abspaths.append(os.path.abspath(".."))                                                                                                                     
+            pythonpath = os.pathsep.join(abspaths)                                                                                                                     
+        else:                                                                                                                                                          
+            pythonpath = os.path.abspath("..")                                                                                                                         
+        env = {"PYTHONPATH": pythonpath}                                                                                                                               
+                                                                                                                                                                       
         from twisted.internet import reactor                                                                                                                           
         clientTransport = reactor.spawnProcess(processClient, exe, args,                                                                                               
                                                env=env, usePTY=True)                                                                                                   
