Index: twisted/internet/qtreactor.py
===================================================================
--- twisted/internet/qtreactor.py	(revision 27325)
+++ twisted/internet/qtreactor.py	(working copy)
@@ -1,3 +1,6 @@
+# Copyright (c) 2001-2009 Twisted Matrix Laboratories.
+# See LICENSE for details.
+
 try:
     # 'import qtreactor' would have imported this file instead of the
     # top-level qtreactor. __import__ does the right thing
@@ -4,7 +7,7 @@
     # (kids, don't repeat this at home)
     install = __import__('qtreactor').install
 except ImportError:
-    from twisted.plugins.qtreactor_stub import errorMessage
+    from twisted.plugins.twisted_qtstub import errorMessage
     raise ImportError(errorMessage)
 else:
     import warnings
Index: twisted/test/test_application.py
===================================================================
--- twisted/test/test_application.py	(revision 27325)
+++ twisted/test/test_application.py	(working copy)
@@ -869,7 +869,21 @@
     test_qtStub.skip = skipWindowsNopywin32
 
 
+    def test_importQtreactor(self):
+        """
+        Attempting to import L{twisted.internet.qtreactor} should raise an
+        C{ImportError} indicating that C{qtreactor} is no longer a part of
+        Twisted.
+        """
+        sys.modules["qtreactor"] = None
+        from twisted.plugins.twisted_qtstub import errorMessage
+        try:
+            import twisted.internet.qtreactor
+        except ImportError, e:
+            self.assertEquals(e.message, errorMessage)
 
+
+
 class ReportProfileTestCase(unittest.TestCase):
     """
     Tests for L{app.reportProfile}.
