Index: twisted/test/test_paths.py
===================================================================
--- twisted/test/test_paths.py	(revision 37995)
+++ twisted/test/test_paths.py	(working copy)
@@ -1431,7 +1431,10 @@
         self.assertEqual(self.path.child(b"sub1").getPermissions().shorthand(),
                           "r-xr-xr-x")
 
+        # Change permission after test so file can be deleted
+        self.addCleanup(self.path.child(b"sub1").chmod, 0o777)
 
+
     def test_whetherBlockOrSocket(self):
         """
         Ensure that a file is not a block or socket
Index: twisted/topfiles/5228.bugfix
===================================================================
--- twisted/topfiles/5228.bugfix	(revision 0)
+++ twisted/topfiles/5228.bugfix	(revision 0)
@@ -0,0 +1 @@
+twisted.test.test_paths.FilePathTestCase.test_getPermissions_Windows now creates deleteable files
\ No newline at end of file
