Index: twisted/test/test_paths.py
===================================================================
--- twisted/test/test_paths.py	(revision 37995)
+++ twisted/test/test_paths.py	(working copy)
@@ -1426,10 +1426,10 @@
             self.path.child(b"sub1").chmod(mode)
             self.assertEqual(self.path.child(b"sub1").getPermissions(),
                               filepath.Permissions(mode))
-        self.path.child(b"sub1").chmod(0o511) #sanity check to make sure that
+        self.path.child(b"sub1").chmod(0o711) #sanity check to make sure that
         # user=group=other permissions
         self.assertEqual(self.path.child(b"sub1").getPermissions().shorthand(),
-                          "r-xr-xr-x")
+                          "rwxrwxrwx")
 
 
     def test_whetherBlockOrSocket(self):
