--- test_paths.py	2013-03-31 02:23:49.798448233 +0530
+++ /home/dhanush/Desktop/test_paths.py	2013-03-31 17:16:31.614801079 +0530
@@ -1429,7 +1429,7 @@
         self.path.child(b"sub1").chmod(0o511) #sanity check to make sure that
         # user=group=other permissions
         self.assertEqual(self.path.child(b"sub1").getPermissions().shorthand(),
-                          "r-xr-xr-x")
+                          "rwxrwxrwx")
 
 
     def test_whetherBlockOrSocket(self):
