Index: twisted/web/tap.py
===================================================================
--- twisted/web/tap.py	(revision 47176)
+++ twisted/web/tap.py	(working copy)
@@ -10,7 +10,7 @@
 
 import os
 
-from twisted.web import server, static, script, demo
+from twisted.web import server, static, script, demo, wsgi
 from twisted.internet import interfaces, reactor
 from twisted.python import usage, reflect, threadpool
 from twisted.python.compat import _PY3
@@ -20,9 +20,6 @@
     # FIXME: https://twistedmatrix.com/trac/ticket/8009
     from twisted.web import twcgi
 
-    # FIXME: https://twistedmatrix.com/trac/ticket/7993
-    from twisted.web import wsgi
-
     # FIXME: https://twistedmatrix.com/trac/ticket/8010
     # FIXME: https://twistedmatrix.com/trac/ticket/7598
     from twisted.web import distrib
Index: twisted/web/test/test_tap.py
===================================================================
--- twisted/web/test/test_tap.py	(revision 47176)
+++ twisted/web/test/test_tap.py	(working copy)
@@ -24,14 +24,12 @@
 from twisted.web.static import Data, File
 from twisted.web.tap import Options, makeService
 from twisted.web.script import PythonScript
+from twisted.web.wsgi import WSGIResource
 
 if not _PY3:
     # FIXME: https://twistedmatrix.com/trac/ticket/8009
     from twisted.web.twcgi import CGIScript
 
-    # FIXME: https://twistedmatrix.com/trac/ticket/7993
-    from twisted.web.wsgi import WSGIResource
-
     # FIXME: https://twistedmatrix.com/trac/ticket/8010
     # FIXME: https://twistedmatrix.com/trac/ticket/7598
     from twisted.web.distrib import ResourcePublisher, UserDirectory
@@ -230,11 +228,6 @@
         reactor.fireSystemEvent('shutdown')
         self.assertTrue(root._threadpool.joined)
 
-    if _PY3:
-        test_wsgi.skip = (
-            "Will be ported in https://twistedmatrix.com/trac/ticket/7993")
-
-
     def test_invalidApplication(self):
         """
         If I{--wsgi} is given an invalid name, L{Options.parseOptions}
