Index: twisted/mail/protocols.py
===================================================================
RCS file: /cvs/Twisted/twisted/mail/protocols.py,v
retrieving revision 1.16
diff -a -u -r1.16 protocols.py
--- twisted/mail/protocols.py	15 Jul 2003 00:07:35 -0000	1.16
+++ twisted/mail/protocols.py	6 Aug 2003 19:01:53 -0000
@@ -168,12 +168,17 @@
             )
 
     def authenticateUserPASS(self, user, password):
-        portal = self.service.defaultPortal()
-        return portal.login(
-            cred.credentials.UsernamePassword(user, password),
-            None,
-            pop3.IMailbox
-        )
+        user, domain = self.lookupDomain(user)
+        try:
+            portal = self.service.lookupPortal(domain)
+        except KeyError:
+            return defer.fail(cred.error.UnauthorizedLogin())
+        else:
+            return portal.login(
+                cred.credentials.UsernamePassword(user, password),
+                None,
+                pop3.IMailbox,
+                )
 
     def lookupDomain(self, user):
         try:
