Index: log.py
===================================================================
--- log.py	(revision 16877)
+++ log.py	(working copy)
@@ -20,6 +20,8 @@
 except NameError:
     UnicodeEncodeError = UnicodeError
 
+#for ILogObserver
+from zope.interface import Interface
 
 class ILogContext:
     """Actually, this interface is just a synoym for the dictionary interface,
@@ -237,6 +239,25 @@
     msg = theLogPublisher.msg
 
 
+class ILogObserver(Interface):
+
+    def setFile(self,f):
+        """ set the file like object used for writing """
+    
+    def emit(self, eventDict):
+        """
+        emit a log message.
+        """
+        
+    def start(self):
+        """Start observing log events."""
+
+    def stop(self):
+        """Stop observing log events."""
+
+    def __repr__(self):
+        """ print out what this log observer is all about """
+
 class FileLogObserver:
     """Log observer that writes to a file-like object.
 
