Index: web2/howto/intro.xhtml
===================================================================
--- web2/howto/intro.xhtml	(revision 24190)
+++ web2/howto/intro.xhtml	(working copy)
@@ -73,7 +73,7 @@
       <a
          href="http://localhost:8080/">http://localhost:8080/</a>&quot;. 
       For more deployment options, see
-         the <a href="deployment.html">deployment</a> chapter.</p>
+         the <a href="deployment.xhtml">deployment</a> chapter.</p>
 
     <h3>What this is doing</h3>
 
@@ -125,7 +125,7 @@
         also be methods, and, if even that is not powerful enough for
         you, it is possible to override the method lookup machinery at
         an even lower
-        level. See <a class="object-traversal.html">Object
+        level. See <a href="object-traversal.xhtml">Object
         Traversal</a> for more details.
       </li>
       <li><code class="API">twisted.web2.static.File</code> lets you serve
@@ -139,7 +139,7 @@
       <li><code class="API">twisted.web2.http.Response</code> takes three
         arguments: the response code, the output headers, and the output
         stream. For more information,
-        see <a href="resource-apis.html">Resource APIs</a>.
+        see <a href="resource-apis.xhtml">Resource APIs</a>.
       </li>
     </ul>
     
