--- twisted/protocols/gps/nmea.py	2007-12-04 16:16:33.000000000 +0100
+++ twisted/protocols/gps/nmea_new.py	2007-12-04 16:02:03.000000000 +0100
@@ -176,8 +176,14 @@
         utc = self._decode_utc(utc)
         posfix = int(posfix)
         satellites = int(satellites)
-        hdop = float(hdop)
-        altitude = (float(altitude), altitude_units)
+        if hdop != '':
+            hdop = float(hdop)
+        else:
+            hdop = None
+        if altitude != '':
+            altitude = (float(altitude), altitude_units)
+        else:
+            altitude = None
         if geoid_separation != '':
             geoid = (float(geoid_separation), geoid_separation_units)
         else:
