diff --git a/twisted/test/test_adbapi.py b/twisted/test/test_adbapi.py
index 80ed7f2..85afdb6 100644
--- a/twisted/test/test_adbapi.py
+++ b/twisted/test/test_adbapi.py
@@ -843,7 +843,7 @@ class ArraySizeTestCase(unittest.TestCase):
         """
 
         def _interaction(cursor, expectedArraysize):
-            self.failUnlessEqual(cursor.arraysize, expectedArraysize)
+            self.assertEquals(cursor.arraysize, expectedArraysize)
 
         dl = []
         for arraysize in [50, 100, 150, 200]:
