diff --git twisted/internet/serialport.py twisted/internet/serialport.py
index 500d8ba..65c295f 100644
--- twisted/internet/serialport.py
+++ twisted/internet/serialport.py
@@ -24,6 +24,9 @@ from serial import PARITY_NONE, PARITY_EVEN, PARITY_ODD
 from serial import STOPBITS_ONE, STOPBITS_TWO
 from serial import FIVEBITS, SIXBITS, SEVENBITS, EIGHTBITS
 
+from twisted.python.runtime import platform
+
+
 
 
 class BaseSerialPort:
@@ -81,7 +84,9 @@ class SerialPort(BaseSerialPort):
     pass
 
 # replace SerialPort with appropriate serial port
-if os.name == 'posix':
+if platform.getType() == 'posix':
     from twisted.internet._posixserialport import SerialPort
-elif sys.platform == 'win32':
+elif platform.isWindows():
     from twisted.internet._win32serialport import SerialPort
+else:
+    raise AssertionError('OS not supported.')
diff --git twisted/topfiles/7176.misc twisted/topfiles/7176.misc
new file mode 100644
index 0000000..e69de29
