=== modified file 'twisted/python/filepath.py'
--- twisted/python/filepath.py	2011-02-14 04:45:15 +0000
+++ twisted/python/filepath.py	2011-03-25 00:27:50 +0000
@@ -876,6 +876,16 @@
             f.write(content)
         finally:
             f.close()
+
+        # Attempt to get permissions of existing file - if it exists - and apply
+        # those to the new file.
+        try:
+            self.restat()
+        except OSError:
+            pass
+        else:
+            sib.chmod(self.statinfo.st_mode)
+
         if platform.isWindows() and exists(self.path):
             os.unlink(self.path)
         os.rename(sib.path, self.path)

=== modified file 'twisted/test/test_paths.py'
--- twisted/test/test_paths.py	2011-02-14 04:45:15 +0000
+++ twisted/test/test_paths.py	2011-03-25 00:34:05 +0000
@@ -677,6 +677,23 @@
                             opened[1].basename()))
 
 
+    def test_setContentPreservePermissions(self):
+        """
+        L{FilePath.setContent} should attempt to preserve the permissions of an
+        existing file.
+        """
+        fp = TrackingFilePath(self.mktemp())
+        fp.create()
+        fp.chmod(0400)
+        fp.restat()
+        originalMode = fp.statinfo.st_mode
+        fp.setContent("bar")
+        fp.restat()
+        newMode = fp.statinfo.st_mode
+
+        self.assertEquals(originalMode, newMode)
+
+
     def test_symbolicLink(self):
         """
         Verify the behavior of the C{isLink} method against links and

