Index: twisted/internet/posixbase.py
===================================================================
--- twisted/internet/posixbase.py	(revision 27019)
+++ twisted/internet/posixbase.py	(working copy)
@@ -53,7 +53,7 @@
         win32process = None
 
 
-class _Win32Waker(log.Logger, styles.Ephemeral):
+class _SocketWaker(log.Logger, styles.Ephemeral):
     """I am a workaround for the lack of pipes on win32.
 
     I am a pair of connected sockets which can wake up the main loop
@@ -101,7 +101,7 @@
         self.w.close()
 
 
-class _UnixWaker(log.Logger, styles.Ephemeral):
+class _PipeWaker(log.Logger, styles.Ephemeral):
     """This class provides a simple interface to wake up the event loop.
 
     This is used by threads or signals to wake up the event loop.
@@ -152,10 +152,10 @@
         del self.i, self.o
 
 
-if platformType == 'posix':
-    _Waker = _UnixWaker
-elif platformType == 'win32':
-    _Waker = _Win32Waker
+if hasattr(os, 'pipe'):
+    _Waker = _PipeWaker
+else:
+    _Waker = _SocketWaker
 
 
 class PosixReactorBase(_SignalReactorMixin, ReactorBase):
