--- twisted-svn/trunk/twisted/web/http.py	2007-10-02 15:18:50.000000000 +0200
+++ /usr/share/pycentral/python-twisted-web/site-packages/twisted/web/http.py	2007-10-02 16:01:15.000000000 +0200
@@ -413,8 +413,11 @@
             self.handleStatus(version, status, message)
             return
         if line:
-            key, val = line.split(':', 1)
-            val = val.lstrip()
+            if ":" in line[1:-1]:
+                key, val = line.split(':', 1)
+                val = val.lstrip()
+            else:
+                key = val = ""
             self.handleHeader(key, val)
             if key.lower() == 'content-length':
                 self.length = int(val)
