Index: twisted/conch/client/knownhosts.py
===================================================================
--- twisted/conch/client/knownhosts.py	(revision 38074)
+++ twisted/conch/client/knownhosts.py	(working copy)
@@ -369,6 +369,9 @@
         self._savePath = savePath
         self._clobber = True
 
+    @property
+    def savePath(self):
+        return self._savePath
 
     def iterentries(self):
         """
