Index: twisted/conch/client/knownhosts.py
===================================================================
--- twisted/conch/client/knownhosts.py	(revision 38074)
+++ twisted/conch/client/knownhosts.py	(working copy)
@@ -369,6 +369,9 @@
         self._savePath = savePath
         self._clobber = True
 
+    @property
+    def savePath(self):
+        return self._savePath
 
     def iterentries(self):
         """
Index: twisted/conch/test/test_knownhosts.py
===================================================================
--- twisted/conch/test/test_knownhosts.py	(revision 38074)
+++ twisted/conch/test/test_knownhosts.py	(working copy)
@@ -699,7 +699,7 @@
         key = Key.fromString(sampleKey)
         entry = PlainEntry(["brandnew.example.com"], key.sshType(), key, "")
         hostsFile = self.loadSampleHostsFile()
-        with hostsFile._savePath.open("a") as hostsFileObj:
+        with hostsFile.savePath.open("a") as hostsFileObj:
             hostsFileObj.write(entry.toString() + "\n")
         self.assertEqual(
             True, hostsFile.hasHostKey("brandnew.example.com", key))
@@ -719,7 +719,7 @@
             "www.twistedmatrix.com", Key.fromString(otherSampleKey))
         self.assertEqual(exception.offendingEntry, entries[0])
         self.assertEqual(exception.lineno, 1)
-        self.assertEqual(exception.path, hostsFile._savePath)
+        self.assertEqual(exception.path, hostsFile.savePath)
 
 
     def test_savedEntryAfterAddHasKeyMismatch(self):
@@ -737,7 +737,7 @@
             HostKeyChanged, hostsFile.hasHostKey,
             "www.twistedmatrix.com", Key.fromString(otherSampleKey))
         self.assertEqual(exception.lineno, 1)
-        self.assertEqual(exception.path, hostsFile._savePath)
+        self.assertEqual(exception.path, hostsFile.savePath)
 
 
     def test_unsavedEntryHasKeyMismatch(self):
@@ -861,7 +861,7 @@
         ui, l, knownHostsFile = self.verifyNonPresentKey()
         ui.promptDeferred.callback(True)
         self.assertEqual([True], l)
-        reloaded = KnownHostsFile.fromPath(knownHostsFile._savePath)
+        reloaded = KnownHostsFile.fromPath(knownHostsFile.savePath)
         self.assertEqual(
             True,
             reloaded.hasHostKey("4.3.2.1", Key.fromString(thirdSampleKey)))
@@ -909,7 +909,7 @@
         hostsFile.verifyHostKey(
             ui, "www.twistedmatrix.com", "5.4.3.2", expectedKey)
         self.assertEqual(
-            True, KnownHostsFile.fromPath(hostsFile._savePath).hasHostKey(
+            True, KnownHostsFile.fromPath(hostsFile.savePath).hasHostKey(
                 "5.4.3.2", expectedKey))
         self.assertEqual(
             ["Warning: Permanently added the RSA host key for IP address "
