diff --git a/twisted/conch/interfaces.py b/twisted/conch/interfaces.py
index d42811a..3eb0217 100644
--- a/twisted/conch/interfaces.py
+++ b/twisted/conch/interfaces.py
@@ -96,12 +96,18 @@ class ISession(Interface):
 
 class ISFTPServer(Interface):
     """
-    The only attribute of this class is "avatar".  It is the avatar
-    returned by the Realm that we are authenticated with, and
-    represents the logged-in user.  Each method should check to verify
-    that the user has permission for their actions.
+    SFTP subsystem for server-side communication.
+
+    Each method should check to verify that the user has permission for
+    their actions.
     """
 
+    avatar = Attribute(
+        """
+        The avatar returned by the Realm that we are authenticated with,
+        and represents the logged-in user.
+        """)
+
     def gotVersion(otherVersion, extData):
         """
         Called when the client sends their version info.
diff --git a/twisted/conch/test/test_filetransfer.py b/twisted/conch/test/test_filetransfer.py
index 15c0daf..8216602 100644
--- a/twisted/conch/test/test_filetransfer.py
+++ b/twisted/conch/test/test_filetransfer.py
@@ -148,6 +148,15 @@ class TestOurServerOurClient(SFTPTestBase):
         self.assertEqual(self._extData, {'conchTest' : 'ext data'})
 
 
+    def test_interface_implementation(self):
+        """
+        It implements the ISFTPServer interface.
+        """
+        self.assertTrue(
+            filetransfer.ISFTPServer.providedBy(self.server.client),
+            "ISFTPServer not provided by %r" % (self.server.client,))
+
+
     def test_openedFileClosedWithConnection(self):
         """
         A file opened with C{openFile} is close when the connection is lost.
diff --git a/twisted/topfiles/6892.misc b/twisted/topfiles/6892.misc
new file mode 100644
index 0000000..32e72f2
--- /dev/null
+++ b/twisted/topfiles/6892.misc
@@ -0,0 +1 @@
+twisted.conch.interfaces.ISFTPServer now defines the `avatar` attribute.
