Index: twisted/test/test_sslverify.py
===================================================================
--- twisted/test/test_sslverify.py	(revision 38118)
+++ twisted/test/test_sslverify.py	(working copy)
@@ -19,11 +19,12 @@
     pass
 
 from twisted.python.compat import nativeString
-from twisted.trial import unittest
+from twisted.trial import unittest, util
 from twisted.internet import protocol, defer, reactor
 
 from twisted.internet.error import CertificateError, ConnectionLost
 from twisted.internet import interfaces
+from twisted.python.versions import Version
 
 
 # A couple of static PEM-format certificates to be used by various tests.
@@ -373,7 +374,10 @@
         self.assertEqual(opts.fixBrokenPeers, True)
         self.assertEqual(opts.enableSessionTickets, True)
 
+    test_certificateOptionsSerialization.suppress = [util.suppress(
+        category = DeprecationWarning)]
 
+
     def test_certificateOptionsSessionTickets(self):
         """
         Enabling session tickets should not set the OP_NO_TICKET option.
@@ -549,8 +553,32 @@
         ctx = opts.getContext()
         self.assertEqual(SSL.OP_NO_SSLv2, ctx.set_options(0) & SSL.OP_NO_SSLv2)
 
+    def test_getstateDeprecation(self):
+        """
+        Test deprecation of L{_sslverify.OpenSSLCertificateOptions.__getstate__}.
+        See #6166 for removal.
+        """
+        self.callDeprecated(
+            (Version("Twisted", 13, 1, 0), "a real persistence system"),
+            sslverify.OpenSSLCertificateOptions().__getstate__)
 
+    def test_setstateDeprecation(self):
+        """
+        Test deprecation of L{_sslverify.OpenSSLCertificateOptions.__setstate__}.
+        See #6166 for removal.
+        """
+        state = sslverify.OpenSSLCertificateOptions().__getstate__()
+        self.callDeprecated(
+            (Version("Twisted", 13, 1, 0), "a real persistence system"),
+            sslverify.OpenSSLCertificateOptions().__setstate__, state)
 
+    test_setstateDeprecation.suppress = [util.suppress(
+        message = "twisted.internet._sslverify.OpenSSLCertificateOptions.__getstate__ " 
+        "was deprecated in Twisted 13.1.0; please use a real persistence system instead",
+        category = DeprecationWarning)]
+
+
+
 if interfaces.IReactorSSL(reactor, None) is None:
     OpenSSLOptions.skip = "Reactor does not support SSL, cannot run SSL tests"
 
@@ -648,3 +676,32 @@
 
 if interfaces.IReactorSSL(reactor, None) is None:
     Constructors.skip = "Reactor does not support SSL, cannot run SSL tests"
+
+class KeyPair(unittest.TestCase):
+    sKey = makeCertificate(
+        O=b"Server Test Certificate",
+        CN=b"server")[0]
+
+    def test_getstateDeprecation(self):
+        """
+        Test deprecation of L{_sslverify.KeyPair.__getstate__}.
+        See #6166 for removal.
+        """
+        self.callDeprecated(
+            (Version("Twisted", 13, 1, 0), "a real persistence system"),
+            sslverify.KeyPair(self.sKey).__getstate__)
+
+    def test_setstateDeprecation(self):
+        """
+        Test deprecation of L{_sslverify.KeyPair.__setstate__}.
+        See #6166 for removal.
+        """
+        state = sslverify.KeyPair(self.sKey).__getstate__()
+        self.callDeprecated(
+            (Version("Twisted", 13, 1, 0), "a real persistence system"),
+            sslverify.KeyPair(self.sKey).__setstate__, state)
+
+    test_setstateDeprecation.suppress = [util.suppress(
+        message = "twisted.internet._sslverify.KeyPair.__getstate__ "
+        "was deprecated in Twisted 13.1.0; please use a real persistence system instead",
+        category = DeprecationWarning)]
Index: twisted/topfiles/6166.removal
===================================================================
--- twisted/topfiles/6166.removal	(revision 0)
+++ twisted/topfiles/6166.removal	(revision 0)
@@ -0,0 +1,2 @@
+Pickling twisted.internet._sslverify.OptionSSLCertificationOptions and twisted.internet._sslverify.Keypair is no longer supported.
+__getstate__ and __setstate__ methods of these classes have been deprecated.
Index: twisted/internet/_sslverify.py
===================================================================
--- twisted/internet/_sslverify.py	(revision 38118)
+++ twisted/internet/_sslverify.py	(working copy)
@@ -14,6 +14,8 @@
 from twisted.python import _reflectpy3 as reflect, util
 from twisted.internet.defer import Deferred
 from twisted.internet.error import VerifyError, CertificateError
+from twisted.python.deprecate import deprecated
+from twisted.python.versions import Version
 
 def _sessionCounter(counter=itertools.count()):
     """
@@ -619,8 +621,13 @@
             self.signRequestObject(dn, self.requestObject(dn), serialNumber),
             self)
 
+KeyPair.__getstate__ = deprecated(Version("Twisted", 13, 1, 0),
+    "a real persistence system")(KeyPair.__getstate__)
+KeyPair.__setstate__ = deprecated(Version("Twisted", 13, 1, 0),
+    "a real persistence system")(KeyPair.__setstate__)
 
 
+
 class OpenSSLCertificateOptions(object):
     """
     A factory for SSL context objects for both SSL servers and clients.
@@ -789,3 +796,8 @@
             ctx.set_options(self._OP_NO_TICKET)
 
         return ctx
+
+OpenSSLCertificateOptions.__getstate__ = deprecated(Version("Twisted", 13, 1, 0),
+    "a real persistence system")(OpenSSLCertificateOptions.__getstate__)
+OpenSSLCertificateOptions.__setstate__ = deprecated(Version("Twisted", 13, 1, 0),
+    "a real persistence system")(OpenSSLCertificateOptions.__setstate__)
