=== modified file 'twisted/internet/defer.py'
--- twisted/internet/defer.py	2011-08-01 19:06:32 +0000
+++ twisted/internet/defer.py	2011-09-30 21:02:34 +0000
@@ -1136,6 +1136,10 @@
     available. The generator will be sent the result of the L{Deferred} with the
     'send' method on generators, or if the result was a failure, 'throw'.
 
+    Things that are not L{Deferred}s may also be yielded, and your generator
+    will be resumed with the same object sent back. This means C{yield}
+    performs an operation roughly equivalent to L{maybeDeferred}.
+
     Your inlineCallbacks-enabled generator will return a L{Deferred} object, which
     will result in the return value of the generator (or will fail with a
     failure object if your generator raises an unhandled exception). Note that
@@ -1143,6 +1147,13 @@
     instead. Falling off the end of the generator, or simply using C{return}
     will cause the L{Deferred} to have a result of C{None}.
 
+    Be aware that L{returnValue} will not accept a L{Deferred} as a parameter.
+    If you believe the thing you'd like to return could be a L{Deferred}, do
+    this::
+
+        result = yield result
+        returnValue(result)
+
     The L{Deferred} returned from your deferred generator may errback if your
     generator raised an exception::
 

