Index: core/howto/pb-usage.xhtml
===================================================================
--- core/howto/pb-usage.xhtml	(revision 24520)
+++ core/howto/pb-usage.xhtml	(working copy)
@@ -19,9 +19,9 @@
 >pbsimpleclient.py</a>
 
 <p>First we look at the server. This defines an Echoer class (derived from
-<code class="API" base="twisted.spread">pb.Root</code>), with a method called
+<code class="API" base="twisted.spread.flavors">Root</code>), with a method called
 <code>remote_echo()</code>. 
-<code class="API" base="twisted.spread">pb.Root</code> objects (because of
+<code class="API" base="twisted.spread.flavors">Root</code> objects (because of
 their inheritance of 
 <code class="API" base="twisted.spread">pb.Referenceable</code>, described
 later) can define methods with names of the form <code>remote_*</code>; a
