Index: twisted/test/test_spread.py
===================================================================
RCS file: /cvs/Twisted/twisted/test/test_spread.py,v
retrieving revision 1.3
diff -u -u -r1.3 test_spread.py
--- twisted/test/test_spread.py	11 Feb 2003 23:45:52 -0000	1.3
+++ twisted/test/test_spread.py	2 Oct 2003 12:42:00 -0000
@@ -22,6 +22,7 @@
 from twisted.trial import unittest
 
 from twisted.spread.util import LocalAsyncForwarder
+from twisted.internet import defer
 from twisted.python.components import Interface
 
 class IForwarded:
@@ -39,8 +40,9 @@
 
     def dontForwardMe(self):
         self.unforwarded = 1
-
-        
+    
+    def forwardDeferred(self):
+        return defer.succeed(1)
 
 class SpreadUtilTest(unittest.TestCase):
     def testLocalAsyncForwarder(self):
@@ -50,3 +52,6 @@
         assert f.forwarded
         lf.callRemote("dontForwardMe")
         assert not f.unforwarded
+        rr = lf.callRemote("forwardDeferred")        
+        assert rr == 1
+        
\ No newline at end of file
