Index: doc/core/howto/trial.xhtml
===================================================================
--- doc/core/howto/trial.xhtml	(revision 38111)
+++ doc/core/howto/trial.xhtml	(working copy)
@@ -309,7 +309,7 @@
 
 <p>If you run the above, you'll see that not all tests fail. In Python it's
 often valid to add and multiply objects of different and even differing
-types, so the code in the add and mutiply tests does not raise an exception
+types, so the code in the add and multiply tests does not raise an exception
 and those tests therefore fail. So let's add explicit type conversion to
 our API class. This brings us to <code
 class="py-filename">calculus/base_3.py</code>:</p>
@@ -439,7 +439,7 @@
 
 <h3>Creating and testing the client</h3>
 
-<p>Of course, what we build is not particulary useful for now : we'll now build
+<p>Of course, what we build is not particularly useful for now : we'll now build
 a client to our server, to be able to use it inside a Python program. And it
 will serve our next purpose.</p>
 
@@ -707,7 +707,7 @@
 have complete coverage :).</p>
 
 <p>Again, this is only another useful pointer, but it doesn't mean your
-code is perfect: your tests should consider every possibile input and
+code is perfect: your tests should consider every possible input and
 output, to get <strong>full</strong> coverage (condition, path, etc.) as well
 .</p>
 
