diff --git a/twisted/internet/defer.py b/twisted/internet/defer.py
index 9b85d37..a90ee81 100644
--- a/twisted/internet/defer.py
+++ b/twisted/internet/defer.py
@@ -1430,7 +1430,7 @@ class QueueUnderflow(Exception):
 
 class DeferredQueue(object):
     """
-    An event driven queue.
+    An event driven first-in-first-out queue.
 
     Objects may be added as usual to this queue.  When an attempt is
     made to retrieve an object when the queue is empty, a L{Deferred} is
