Index: twisted/mail/maildir.py
===================================================================
--- twisted/mail/maildir.py	(revision 30075)
+++ twisted/mail/maildir.py	(working copy)
@@ -27,12 +27,7 @@
 from twisted.python.hashlib import md5
 from twisted.mail import mail
 from twisted.internet import interfaces, defer, reactor
-
 from twisted import cred
-import twisted.cred.portal
-import twisted.cred.credentials
-import twisted.cred.checkers
-import twisted.cred.error
 
 INTERNAL_ERROR = '''\
 From: Twisted.mail Internals
