Index: twisted/web/proxy.py
===================================================================
--- twisted/web/proxy.py	(revision 22180)
+++ twisted/web/proxy.py	(working copy)
@@ -253,7 +253,10 @@
 
 
     def render(self, request):
-        request.received_headers['host'] = self.host
+        if self.port == 80:
+            request.received_headers['host'] = self.host
+        else:
+            request.received_headers['host'] = "%s:%d" % (self.host, self.port)
         request.content.seek(0, 0)
         qs = urlparse.urlparse(request.uri)[4]
         if qs:
Index: twisted/web/test/test_proxy.py
===================================================================
--- twisted/web/test/test_proxy.py	(revision 22180)
+++ twisted/web/test/test_proxy.py	(working copy)
@@ -74,6 +74,7 @@
         factory = reactor.connect[0][2]
         self.assertIsInstance(factory, ProxyClientFactory)
         self.assertEquals(factory.rest, expectedURI)
+        self.assertEquals(factory.headers["host"], "127.0.0.1:1234")
 
 
     def test_render(self):
