Index: twisted/application/app.py
===================================================================
--- twisted/application/app.py	(revision 28956)
+++ twisted/application/app.py	(working copy)
@@ -717,14 +717,3 @@
     reactor.addSystemEventTrigger('before', 'shutdown',
                                   service.IService(application).stopService)
 
-def getLogFile(logfilename):
-    """
-    Build a log file from the full path.
-    """
-    warnings.warn(
-        "app.getLogFile is deprecated. Use "
-        "twisted.python.logfile.LogFile.fromFullPath instead",
-        DeprecationWarning, stacklevel=2)
-
-    return logfile.LogFile.fromFullPath(logfilename)
-
