Index: twisted/topfiles/7724.misc
===================================================================
Index: twisted/positioning/test/test_nmea.py
===================================================================
--- twisted/positioning/test/test_nmea.py	(revision 43542)
+++ twisted/positioning/test/test_nmea.py	(working copy)
@@ -125,7 +125,7 @@
 
 
 
-class SplitTest(TestCase):
+class SplitTests(TestCase):
     """
     Checks splitting of NMEA sentences.
     """
@@ -1008,7 +1008,7 @@
 
 
 
-class NMEAReceiverTest(TestCase):
+class NMEAReceiverTests(TestCase):
     """
     Tests for the NMEA receiver.
     """
