Index: doc/development/policy/coding-standard.xhtml
===================================================================
--- doc/development/policy/coding-standard.xhtml	(revision 21321)
+++ doc/development/policy/coding-standard.xhtml	(working copy)
@@ -527,6 +527,16 @@
     </li>
     </ul>
 
+    <h2>Fallback</h2>
+
+    <p>In case of conventions not enforced in this document, the reference
+    documents to use in fallback is
+    <a href="http://www.python.org/dev/peps/pep-0008/">PEP 8</a> for Python
+    code and <a href="http://www.python.org/dev/peps/pep-0007/">PEP 7</a> for
+    C code. For example, the paragraph <strong>Whitespace in Expressions and
+    Statements</strong> in PEP 8 describes what should be done in Twisted
+    code.</p>
+
     <h2>Recommendations</h2>
 
     <p>These things aren't necessarily standardizeable (in that
