diff --git a/twisted/topfiles/5220.bugfix b/twisted/topfiles/5220.bugfix
deleted file mode 100644
index 3a56142..0000000
--- a/twisted/topfiles/5220.bugfix
+++ /dev/null
@@ -1 +0,0 @@
-trial now deterministically runs tests in the order in which they were specified on the command line, instead of quasi-randomly according to dictionary key ordering.
\ No newline at end of file
diff --git a/twisted/topfiles/5520.feature b/twisted/topfiles/5520.feature
index e69de29..3a56142 100644
--- a/twisted/topfiles/5520.feature
+++ b/twisted/topfiles/5520.feature
@@ -0,0 +1 @@
+trial now deterministically runs tests in the order in which they were specified on the command line, instead of quasi-randomly according to dictionary key ordering.
\ No newline at end of file
diff --git a/twisted/trial/test/test_script.py b/twisted/trial/test/test_script.py
index 1eb8e0b..ec80e64 100644
--- a/twisted/trial/test/test_script.py
+++ b/twisted/trial/test/test_script.py
@@ -598,6 +598,11 @@ class TestRun(unittest.TestCase):
 
 
 class TestArgumentOrderTests(unittest.TestCase):
+    """
+    Tests for the order-preserving behavior on provided command-line tests.
+
+    """
+
     def setUp(self):
         self.config = trial.Options()
         self.loader = TestLoader()
@@ -605,14 +610,11 @@ class TestArgumentOrderTests(unittest.TestCase):
     def test_preserveArgumentOrder(self):
         """
         Multiple tests passed on the command line are not reordered.
-
         """
-
         tests = [
-            "twisted.manhole.test.test_explorer",
-            "twisted.conch.test.test_conch",
-            "twisted.internet.test.test_default",
-            "twisted.test.test_abstract",
+            "twisted.trial.test.test_tests",
+            "twisted.trial.test.test_assertions",
+            "twisted.trial.test.test_deferreds",
         ]
         self.config.parseOptions(tests)
 
