Index: ftp.py
===================================================================
RCS file: /cvs/Twisted/twisted/protocols/ftp.py,v
retrieving revision 1.98
diff -u -d -r1.98 ftp.py
--- ftp.py	24 Nov 2003 15:18:09 -0000	1.98
+++ ftp.py	12 Dec 2003 17:22:19 -0000
@@ -41,7 +41,7 @@
 from twisted import application, internet, python
 from twisted.python import failure, log, components
 
-from twisted.cred import error, portal, checkers, credentials
+from twisted.cred import error as cred_error, portal, checkers, credentials
 
 # constants
 
@@ -823,7 +823,7 @@
         self.reply(USR_LOGGED_IN_PROCEED)
 
     def _ebLogin(self, failure):
-        r = failure.trap(error.UnauthorizedLogin, TLDNotSetInRealmError)
+        r = failure.trap(cred_error.UnauthorizedLogin, TLDNotSetInRealmError)
         if r == TLDNotSetInRealmError:
             log.debug(failure.getErrorMessage())
             self.reply(REQ_ACTN_NOT_TAKEN, 'internal server error')
