Index: twisted/topfiles/5326.bugfix
===================================================================
--- twisted/topfiles/5326.bugfix	(revision 0)
+++ twisted/topfiles/5326.bugfix	(revision 0)
@@ -0,0 +1 @@
+twisted.plugins.twisted_lore's MathProcessor plugin now correctly calls twisted.lore.lmath instead of twsited.lore.lmath.
\ No newline at end of file
Index: twisted/plugins/twisted_lore.py
===================================================================
--- twisted/plugins/twisted_lore.py	(revision 33557)
+++ twisted/plugins/twisted_lore.py	(working copy)
@@ -19,7 +19,7 @@
 
 MathProcessor = _LorePlugin(
     "mlore",
-    "twsited.lore.lmath",
+    "twisted.lore.lmath",
     "Lore format with LaTeX formula")
 
 SlideProcessor = _LorePlugin(
