Index: twisted/plugins/twisted_lore.py
===================================================================
--- twisted/plugins/twisted_lore.py	(revision 33453)
+++ twisted/plugins/twisted_lore.py	(working copy)
@@ -19,7 +19,7 @@
 
 MathProcessor = _LorePlugin(
     "mlore",
-    "twsited.lore.lmath",
+    "twisted.lore.lmath",
     "Lore format with LaTeX formula")
 
 SlideProcessor = _LorePlugin(
