--- smtp.py.orig	2005-10-20 22:11:18.000000000 +0000
+++ smtp.py	2006-08-28 22:13:30.000000000 +0000
@@ -1530,7 +1530,7 @@
         if code not in SUCCESS:
             errlog = []
             for addr, acode, aresp in addresses:
-                if code not in SUCCESS:
+                if acode not in SUCCESS:
                     errlog.append("%s: %03d %s" % (addr, acode, aresp))
 
             errlog.append(log.str())
