Index: twisted/topfiles/4520.bugfix
===================================================================
--- twisted/topfiles/4520.bugfix	(revision 31785)
+++ twisted/topfiles/4520.bugfix	(working copy)
@@ -1,2 +1 @@
-Stopped CopiedFailure.throwExceptionIntoGenerator from throwing string
-exceptions as that isn't permitted in Python 2.6 or greater.
\ No newline at end of file
+twisted.spread.pb.CopiedFailure will no longer be thrown into a generator as a (deprecated) string exception but as a twisted.spread.pb.RemoteException.
\ No newline at end of file
Index: twisted/spread/pb.py
===================================================================
--- twisted/spread/pb.py	(revision 31785)
+++ twisted/spread/pb.py	(working copy)
@@ -107,19 +107,18 @@
 
 class RemoteError(Exception):
     """
-    When the remote side of the connection returns an exception back, it comes
-    in as a string. Raising string exceptions doesn't work in Python 2.6 or
-    greater, so this exception is used instead.
+    This class is used to wrap a string-ified exception from the remote side
+    to be able to reraise it. (Raising string exceptions is no longer possible 
+    in Python 2.6+)
 
-    The value of this exception will be a L{CopiedFailure} if the value on the
-    remote side was a L{twisted.python.failure.Failure}. Else it will be the
-    str() representation of the remote value.
+    The value of this exception will be a str() representation of the remote 
+    value.
 
-    @param remoteType: The full import path of the exception class which was
+    @ivar remoteType: The full import path of the exception class which was
         raised on the remote end.
     @type remoteType: C{str}
 
-    @param remoteTraceback: The remote traceback.
+    @ivar remoteTraceback: The remote traceback.
     @type remoteTraceback: C{str}
 
     @note: It's not possible to include the remoteTraceback if this exception is
@@ -457,16 +456,14 @@
 
 class CopiedFailure(RemoteCopy, failure.Failure):
     """
-    A L{CopiedFailure} is a L{flavors.RemoteCopy} of a L{failure.Failure}
+    A L{CopiedFailure} is a L{pb.RemoteCopy} of a L{failure.Failure}
     transfered via PB.
 
     @ivar type: The full import path of the exception class which was raised on
         the remote end.
     @type type: C{str}
 
-    @ivar value: A L{CopiedFailure} if the value on the remote side was a
-        L{twisted.python.failure.Failure}. Else a str() representation of the
-        remote value.
+    @ivar value: A str() representation of the remote value.
     @type value: L{CopiedFailure} or C{str}
 
     @ivar remoteTraceback: The remote traceback.
