Index: doc/web/howto/twisted-templates.xhtml
===================================================================
--- doc/web/howto/twisted-templates.xhtml	(revision 35287)
+++ doc/web/howto/twisted-templates.xhtml	(working copy)
@@ -20,7 +20,7 @@
 as <a href="web-in-60/wsgi.xhtml">WSGI applications</a> or by calling your
 preferred templating system's APIs to produce their output as strings, and then
 writing those strings to <code class="API"
-base="twisted.web.request">Request.write</code>.
+base="twisted.web.http">Request.write</code>.
 </p>
 <p>Before we begin explaining how to use it, I'd like to stress that you
 don't <i>need</i> to use Twisted's templating system if you prefer some other
