Index: twisted/persisted/test/test_styles.py
===================================================================
--- twisted/persisted/test/test_styles.py	(revision 43542)
+++ twisted/persisted/test/test_styles.py	(working copy)
@@ -27,7 +27,7 @@
 
 
 
-class UnpickleMethodTestCase(unittest.TestCase):
+class UnpickleMethodTests(unittest.TestCase):
     """
     Tests for the unpickleMethod function.
     """
Index: twisted/topfiles/7723.misc
===================================================================
