diff --git a/enterprise/adbapi.py b/enterprise/adbapi.py
index 54e87c4..58663fd 100644
--- a/enterprise/adbapi.py
+++ b/enterprise/adbapi.py
@@ -330,6 +330,8 @@ class ConnectionPool:
         @return: a Deferred which will fire the return value of a DB-API
         cursor's 'fetchall' method, or a Failure.
         """
+        print '-'*80
+        print args[0]
         return self.runInteraction(self._runQuery, *args, **kw)
 
 
@@ -384,7 +386,6 @@ class ConnectionPool:
 
         @return: a database connection from the pool.
         """
-
         tid = self.threadID()
         conn = self.connections.get(tid)
         if conn is None:
@@ -431,6 +432,17 @@ class ConnectionPool:
             conn.commit()
             return result
         except:
+            print 'TRYING AGAIN!', args[0]
+            curs = conn._connection.cursor()
+            try:
+                curs.execute(args[0])
+                conn._connection.commit()
+                a = curs.fetchall()
+                print 'no error'
+                return
+            except:
+                print 'still error'
+                pass
             excType, excValue, excTraceback = sys.exc_info()
             try:
                 conn.rollback()
