Index: twisted/internet/inotify.py
===================================================================
--- twisted/internet/inotify.py	(revision 28865)
+++ twisted/internet/inotify.py	(working copy)
@@ -8,14 +8,15 @@
 In order to use this support, simply do the following::
 
     from twisted.internet import inotify
+    from twisted.python import filepath
 
-    def notify(self, iwp, filepath, mask, extraArg):
-        print "event %s on %s %s" % (
-            ', '.join(inotify.humanReadableMask(mask)), iwp.path, filepath)
+    def notify(self, filepath, mask, extraArg=None):
+        print "event %s on %s" % (
+            ', '.join(inotify.humanReadableMask(mask)), filepath)
 
     notifier = inotify.INotify()
     notifier.startReading()
-    notifier.watch("/some/directory", callbacks=[notify])
+    notifier.watch(filepath.FilePath("/some/directory"), callbacks=[notify])
 
 @since: 10.1
 """
