--- /usr/lib/python2.3/site-packages/twisted/test/test_domish.py	2003-09-19 07:48:43.000000000 +1000
+++ twisted/test/test_domish.py	2004-09-05 11:44:58.155987285 +1000
@@ -92,7 +100,7 @@
         self.assertEquals(e[("testns2", "attrib2")], "value2")
 
 xml1 = """<stream:stream xmlns:stream='etherx' xmlns='jabber'>
-             <message to='bar'><x xmlns='xdelay'>some&amp;data&gt;</x></message>
+             <message to='bar' xml:lang='en'><x xmlns='xdelay'>some&amp;data&gt;</x></message>
           </stream:stream>"""
 query1_root = xpath.intern("/stream[@xmlns='etherx']")    
 query1_elem1 = xpath.intern("/message[@to='bar']/x[@xmlns='xdelay'][text()='some&data>']")
