Index: doc/web2/examples/auth/credsetup.py
===================================================================
--- doc/web2/examples/auth/credsetup.py	(revision 16974)
+++ doc/web2/examples/auth/credsetup.py	(working copy)
@@ -1,4 +1,5 @@
 from zope.interface import Interface, implements
+from twisted.cred import portal
 
 class IHTTPUser(Interface):
     pass
Index: doc/web2/examples/auth/httpauth.tac
===================================================================
--- doc/web2/examples/auth/httpauth.tac	(revision 16974)
+++ doc/web2/examples/auth/httpauth.tac	(working copy)
@@ -20,8 +20,9 @@
 portal.registerChecker(checker)
 
 root = wrapper.HTTPAuthResource(ProtectedResource(),
-                                (digest.DigestCredentialFactory('md5', 
-                                                               'My Realm'),),
+                                (basic.BasicCredentialFactory('My Realm'),
+                                 digest.DigestCredentialFactory('md5', 
+                                                               'My Realm')),
                                 portal, (credsetup.IHTTPUser,))
 
 site = server.Site(root)
