Index: web2/tap.py
===================================================================
--- web2/tap.py	(revision 24262)
+++ web2/tap.py	(working copy)
@@ -138,7 +138,7 @@
         """
         
         if self['root']:
-            if not isintance(self['root'], vhost.NameVirtualHost):
+            if not isinstance(self['root'], vhost.NameVirtualHost):
                 raise usage.UsageError("You may only have one root resource")
         else:
             self['root'] = vhost.NameVirtualHost()
