Index: twisted/conch/recvline.py
===================================================================
--- twisted/conch/recvline.py	(revision 22299)
+++ twisted/conch/recvline.py	(working copy)
@@ -147,9 +147,6 @@
         self.initializeScreen()
 
     def initializeScreen(self):
-        # Hmm, state sucks.  Oh well.
-        # For now we will just take over the whole terminal.
-        self.terminal.reset()
         self.terminal.write(self.ps[self.pn])
         # XXX Note: I would prefer to default to starting in insert
         # mode, however this does not seem to actually work!  I do not
Index: twisted/conch/insults/insults.py
===================================================================
--- twisted/conch/insults/insults.py	(revision 22299)
+++ twisted/conch/insults/insults.py	(working copy)
@@ -610,7 +610,7 @@
     def nextLine(self):
         self.cursorPos.x = 0
         self.cursorPos.y = min(self.cursorPos.y + 1, self.termSize.y - 1)
-        self.write('\x1bE')
+        self.write("\n")
 
     def saveCursor(self):
         self._savedCursorPos = Vector(self.cursorPos.x, self.cursorPos.y)
@@ -767,7 +767,6 @@
         self.write(''.join(bytes))
 
     def loseConnection(self):
-        self.reset()
         self.transport.loseConnection()
 
     def connectionLost(self, reason):
Index: twisted/conch/stdio.py
===================================================================
--- twisted/conch/stdio.py	(revision 22299)
+++ twisted/conch/stdio.py	(working copy)
@@ -76,7 +76,7 @@
         reactor.run()
     finally:
         termios.tcsetattr(fd, termios.TCSANOW, oldSettings)
-        os.write(fd, "\r\x1bc\r")
+        os.write(fd, '\n')
 
 
 
