diff -ru tmp1/twisted-twisted-18.9.0/src/twisted/web/http.py tmp2/twisted-twisted-18.9.0/src/twisted/web/http.py
--- twisted-twisted-18.9.0.orig/src/twisted/web/http.py	2018-10-10 12:25:13.000000000 +0100
+++ twisted-twisted-18.9.0/src/twisted/web/http.py	2018-10-23 16:01:08.462172767 +0100
@@ -2047,14 +2047,14 @@
         self._receivedHeaderSize += len(line)
         if (self._receivedHeaderSize > self.totalHeadersSize):
             self._respondToBadRequestAndDisconnect()
-            return
+            return True # stop processing more lines
 
         if self.__first_line:
             # if this connection is not persistent, drop any data which
             # the client (illegally) sent after the last request.
             if not self.persistent:
                 self.dataReceived = self.lineReceived = lambda *args: None
-                return
+                return True # stop processing more lines
 
             # IE sends an extraneous empty line (\r\n) after a POST request;
             # eat up such a line, but only ONCE
@@ -2074,13 +2074,13 @@
             parts = line.split()
             if len(parts) != 3:
                 self._respondToBadRequestAndDisconnect()
-                return
+                return True # stop processing more lines
             command, request, version = parts
             try:
                 command.decode("ascii")
             except UnicodeDecodeError:
                 self._respondToBadRequestAndDisconnect()
-                return
+                return True # stop processing more lines
 
             self._command = command
             self._path = request
