Index: doc/core/howto/quotes.xhtml
===================================================================
--- doc/core/howto/quotes.xhtml	(revision 19828)
+++ doc/core/howto/quotes.xhtml	(working copy)
@@ -28,13 +28,13 @@
 <li>Make a <code>TwistedQuotes</code> directory on your system</li>
 <li>Place the following files in the <code>TwistedQuotes</code> directory:
     <ul>
-    <li><a href="listings/TwistedQuotes/__init__.py">__init__.py</a> (this
+    <li><a href="listings/TwistedQuotes/__init__.py" class="py-listing">__init__.py</a> (this
         file marks it as a package, see <a
         href="http://www.python.org/doc/current/tut/node8.html#SECTION008400000000000000000"
         >this section</a> of the Python tutorial for more on packages);</li>
-    <li><a href="listings/TwistedQuotes/quoters.py">quoters.py</a>;</li>
-    <li><a href="listings/TwistedQuotes/quoteproto.py">quoteproto.py</a>; and</li>
-    <li><a href="listings/TwistedQuotes/plugins.tml">plugins.tml</a>.</li>
+    <li><a href="listings/TwistedQuotes/quoters.py" class="py-listing">quoters.py</a>;</li>
+    <li><a href="listings/TwistedQuotes/quoteproto.py" class="py-listing">quoteproto.py</a>; and</li>
+    <li><a href="listings/TwistedQuotes/plugins.tml" class="py-listing">plugins.tml</a>.</li>
     </ul>
 </li>
 <li>Add the <code>TwistedQuotes</code> directory's <em>parent</em> to your Python
