Index: utility.py
===================================================================
--- utility.py	(revision 33207)
+++ utility.py	(working copy)
@@ -157,7 +157,7 @@
     observe an event, the observer is removed from the list of observers after
     the first observed event.
 
-    Obsevers can also prioritized, by providing an optional C{priority}
+    Observers can also be prioritized, by providing an optional C{priority}
     parameter to the L{addObserver} and L{addOnetimeObserver} methods. Higher
     priority observers are then called before lower priority observers.
 
