Index: twisted/internet/test/test_iocp.py
===================================================================
--- twisted/internet/test/test_iocp.py	(revision 33849)
+++ twisted/internet/test/test_iocp.py	(working copy)
@@ -115,15 +115,15 @@
         """
         Verify that IOCP socket-representing classes implement IReadWriteHandle
         """
-        verifyClass(IReadWriteHandle, tcp.Connection)
-        verifyClass(IReadWriteHandle, udp.Port)
+        self.assertTrue(verifyClass(IReadWriteHandle, tcp.Connection))
+        self.assertTrue(verifyClass(IReadWriteHandle, udp.Port))
 
 
     def test_fileHandleInterfaces(self):
         """
         Verify that L{Filehandle} implements L{IPushProducer}.
         """
-        verifyClass(IPushProducer, FileHandle)
+        self.assertTrue(verifyClass(IPushProducer, FileHandle))
 
 
     def test_maxEventsPerIteration(self):
