Index: twisted/internet/abstract.py
===================================================================
--- twisted/internet/abstract.py	(revision 33559)
+++ twisted/internet/abstract.py	(working copy)
@@ -155,7 +155,7 @@
 
     SEND_LIMIT = 128*1024
 
-    implements(interfaces.IProducer, interfaces.IReadWriteDescriptor,
+    implements(interfaces.IPushProducer, interfaces.IReadWriteDescriptor,
                interfaces.IConsumer, interfaces.ITransport, interfaces.IHalfCloseableDescriptor)
 
     def __init__(self, reactor=None):
Index: twisted/internet/test/test_tcp.py
===================================================================
--- twisted/internet/test/test_tcp.py	(revision 33559)
+++ twisted/internet/test/test_tcp.py	(working copy)
@@ -361,8 +361,14 @@
         self.server.TLS = True
         self.test_writeSequenceAfterDisconnect()
 
+    def test_serverImplementsIPushProducer(self):
+        """
+        L{Server} implements L{IPushProducer}
+        """
+        self.assertEqual(IPushProducer.implementedBy(Server), True)
 
 
+
 class TCPConnectionTests(TestCase):
     """
     Whitebox tests for L{twisted.internet.tcp.Connection}.
