Index: options.xhtml
===================================================================
--- options.xhtml	(revision 37995)
+++ options.xhtml	(working copy)
@@ -174,7 +174,7 @@
 
     <h2>Option Subcommands</h2>
 
-    <p>It is useful, on occassion, to group a set of options together based
+    <p>It is useful, on occasion, to group a set of options together based
     on the logical <q>action</q> to which they belong.  For this, the
     <code>usage.Options</code> class allows you to define a set of
     <q>subcommands</q>, each of which can provide its own
