=== modified file 'doc/core/howto/plugin.xhtml'
--- doc/core/howto/plugin.xhtml	2011-03-14 01:10:16 +0000
+++ doc/core/howto/plugin.xhtml	2011-10-04 15:19:52 +0000
@@ -178,10 +178,12 @@
     need to pass this argument.  However, if you have developed an
     interface of your own, you may want to mandate that plugins for it
     are installed in your own plugins package, rather than in
-    Twisted's.  In this case, you probably also want to support <code
+    Twisted's.</p>
+    
+    <p>For example, you way want to support <code
     class="shell">yourproject/plugins/</code> directories for ease of
-    development.  To do so, you should make the <code
-    class="shell">__init__.py</code> for that package contain at least
+    development.  To do so, you should make <code
+    class="shell">yourproject/plugins/__init__.py</code> contain at least
     the following lines.</p>
 
     <pre class="python">

