diff --git bin/admin/build-docs bin/admin/build-docs
index c9aff67..cc2b8b1 100755
--- bin/admin/build-docs
+++ bin/admin/build-docs
@@ -5,6 +5,15 @@
 
 # This script is not meant to be distributed to users of Twisted.
 # It is only for use in making upstream Twisted releases.
+#
+# You can call it as:
+# ./bin/admin/build-docs . ./website-template.tpl
+#
+# Results are in 'doc'.
+#
+# More info at:
+# https://twistedmatrix.com/trac/wiki/ReleaseProcess
+
 
 import sys, os
 extra = os.path.dirname(os.path.dirname(sys.argv[0]))
diff --git twisted/python/_release.py twisted/python/_release.py
index 8d80d45..3027d10 100644
--- twisted/python/_release.py
+++ twisted/python/_release.py
@@ -75,6 +75,9 @@ class CommandFailed(Exception):
         self.exitSignal = exitSignal
         self.output = output
 
+    def __str__(self):
+        return '\nExit code: %s\nExit signal: %s\n%s' % (
+            self.exitStatus, self.exitSignal, self.output)
 
 
 def _changeVersionInFile(old, new, filename):
@@ -739,9 +742,13 @@ class SphinxBuilder(object):
 
         doctreeDir = buildDir.child('doctrees')
 
-        runCommand(['sphinx-build', '-b', 'html',
-                    '-d', doctreeDir.path, docDir.path,
-                    buildDir.path])
+        # Turn all warning into errors, complain about broken references and
+        # rebuild all files.
+        # It runs in quite mode to reduce noises when reporting errors.
+        runCommand(['sphinx-build', '-W', '-n', '-a', '-E', '-q',
+                   '-b', 'html',
+                   '-d', doctreeDir.path, docDir.path,
+                   buildDir.path])
 
         for path in docDir.walk():
             if path.basename() == "man":
diff --git twisted/topfiles/6980.feature twisted/topfiles/6980.feature
new file mode 100644
index 0000000..c68e184
--- /dev/null
+++ twisted/topfiles/6980.feature
@@ -0,0 +1 @@
+bin/admin/build-docs now fails if narrative documentation contains warnings.
\ No newline at end of file
