Index: pb.py
===================================================================
--- pb.py	(revisione 23919)
+++ pb.py	(copia locale)
@@ -1268,6 +1268,7 @@
     """
     Root Referenceable object, used to login to portal.
     """
+    implements(IUsernameHashedPassword, IUsernameMD5Password)
 
     def __init__(self, portal, broker):
         self.portal = portal
@@ -1278,8 +1279,9 @@
         """
         Start of username/password login.
         """
-        c = challenge()
-        return c, _PortalAuthChallenger(self.portal, self.broker, username, c)
+        self.challenge = challenge()
+        self.username = username
+        return self.challenge, self
 
 
     def remote_loginAnonymous(self, mind):
@@ -1298,20 +1300,6 @@
         return d
 
 
-
-class _PortalAuthChallenger(Referenceable, _JellyableAvatarMixin):
-    """
-    Called with response to password challenge.
-    """
-    implements(IUsernameHashedPassword, IUsernameMD5Password)
-
-    def __init__(self, portal, broker, username, challenge):
-        self.portal = portal
-        self.broker = broker
-        self.username = username
-        self.challenge = challenge
-
-
     def remote_respond(self, response, mind):
         self.response = response
         d = self.portal.login(self, mind, IPerspective)
