Index: twisted/words/topfiles/setup.py
===================================================================
--- twisted/words/topfiles/setup.py	(revision 38111)
+++ twisted/words/topfiles/setup.py	(working copy)
@@ -42,7 +42,7 @@
         license="MIT",
         long_description="""\
 Twisted Words contains implementations of many Instant Messaging
-protocols, including IRC, Jabber, MSN, OSCAR (AIM & ICQ), TOC (AOL),
+protocols, including IRC, Jabber, OSCAR (AIM & ICQ), 
 and some functionality for creating bots, inter-protocol gateways, and
 a client application for many of the protocols.
 
