Index: twisted/words/topfiles/setup.py
===================================================================
--- twisted/words/topfiles/setup.py	(revision 38111)
+++ twisted/words/topfiles/setup.py	(working copy)
@@ -41,10 +41,10 @@
         url="http://twistedmatrix.com/trac/wiki/TwistedWords",
         license="MIT",
         long_description="""\
-Twisted Words contains implementations of many Instant Messaging
-protocols, including IRC, Jabber, MSN, OSCAR (AIM & ICQ), TOC (AOL),
-and some functionality for creating bots, inter-protocol gateways, and
-a client application for many of the protocols.
+Twisted Words contains implementations of many Instant Messaging protocols, 
+including IRC, Jabber, OSCAR (AIM & ICQ), and some functionality for creating 
+bots, inter-protocol gateways, and a client application for many of the 
+protocols.
 
 In support of Jabber, Twisted Words also contains X-ish, a library for
 processing XML with Twisted and Python, with support for a Pythonic DOM and
