Index: twisted/python/failure.py
===================================================================
--- twisted/python/failure.py	(revision 2066)
+++ twisted/python/failure.py	(revision 2067)
@@ -335,6 +335,9 @@
         @raise StopIteration: If there are no more values in the generator.
         @raise anything else: Anything that the generator raises.
         """
+        if isinstance(self.type, basestring):
+             module, type_name = self.type.rsplit(".", 1)
+             self.type = getattr(sys.modules[module], type_name)
         return g.throw(self.type, self.value, self.tb)
 
 
