Index: twisted/mail/pop3.py
===================================================================
--- twisted/mail/pop3.py	(revision 11428)
+++ twisted/mail/pop3.py	(working copy)
@@ -148,7 +148,8 @@
             self.magic = self.generateMagic()
         self.successResponse(self.magic)
         self.setTimeout(self.timeOut)
-        log.msg("New connection from " + str(self.transport.getPeer()))
+	if self.factory.noisy:
+            log.msg("New connection from " + str(self.transport.getPeer()))
 
     def connectionLost(self, reason):
         if self._onLogout is not None:
@@ -330,7 +331,8 @@
         self.mbox = avatar
         self._onLogout = logout
         self.successResponse('Authentication succeeded')
-        log.msg("Authenticated login for " + user)
+	if self.factory.noisy:
+            log.msg("Authenticated login for " + user)
 
     def _ebMailbox(self, failure):
         failure = failure.trap(cred.error.LoginDenied, cred.error.LoginFailed)
@@ -338,7 +340,8 @@
             self.failResponse("Access denied: " + str(failure))
         elif issubclass(failure, cred.error.LoginFailed):
             self.failResponse('Authentication failed')
-        log.msg("Denied login attempt from " + str(self.transport.getPeer()))
+	if self.factory.noisy:
+            log.msg("Denied login attempt from " + str(self.transport.getPeer()))
 
     def _ebUnexpected(self, failure):
         self.failResponse('Server error: ' + failure.getErrorMessage())
