diff --git a/twisted/internet/abstract.py b/twisted/internet/abstract.py
index 6e69f4a..c4e4dcc 100644
--- a/twisted/internet/abstract.py
+++ b/twisted/internet/abstract.py
@@ -366,14 +366,15 @@ class FileDescriptor(_ConsumerMixin, _LogOwner):
         streaming producer is registered, it will be paused until the buffered
         data is written to the underlying file descriptor.
         """
+        totalIovecLength = 0
         for i in iovec:
             if isinstance(i, unicode): # no, really, I mean it
                 raise TypeError("Data must not be unicode")
+            totalIovecLength += len(i)
         if not self.connected or not iovec or self._writeDisconnected:
             return
         self._tempDataBuffer.extend(iovec)
-        for i in iovec:
-            self._tempDataLen += len(i)
+        self._tempDataLen += totalIovecLength
         self._maybePauseProducer()
         self.startWriting()
 
