Index: common.py
===================================================================
--- common.py	(revision 21361)
+++ common.py	(working copy)
@@ -192,7 +192,12 @@
     for r in answers:
         if r.type == dns.NS:
             from twisted.names import client
-            r = client.Resolver(servers=[(str(r.payload.name), dns.PORT)])
+            server = str(r.payload.name)
+            for s in answers:
+                if s.type == dns.A and str(s.name) == nsServerName:
+                    server = socket.inet_ntop(socket.AF_INET, s.payload.address)
+                    break
+            r = client.Resolver(servers=[(server, dns.PORT)])
             return r.lookupAddress(str(name)
                 ).addCallback(lambda (ans, auth, add): extractRecord(r, name, ans + auth + add, level - 1)
                 ).addBoth(lambda passthrough: (r.protocol.transport.stopListening(), passthrough)[1])
