Index: imap4client.py
===================================================================
--- imap4client.py	(revision 14241)
+++ imap4client.py	(working copy)
@@ -15,7 +15,7 @@
 from twisted.internet import ssl
 from twisted.internet import defer
 from twisted.internet import stdio
-from twisted.protocols import imap4
+from twisted.mail import imap4
 from twisted.protocols import basic
 from twisted.python import util
 from twisted.python import log
Index: emailserver.tac
===================================================================
--- emailserver.tac	(revision 14241)
+++ emailserver.tac	(working copy)
@@ -11,7 +11,7 @@
 """
 
 from twisted.internet import defer
-from twisted.protocols import smtp
+from twisted.mail import smtp
 
 class ConsoleMessageDelivery:
     __implements__ = (smtp.IMessageDelivery,)
