Index: twisted/topfiles/5598.removal
===================================================================
--- twisted/topfiles/5598.removal	(revision 0)
+++ twisted/topfiles/5598.removal	(revision 0)
@@ -0,0 +1,2 @@
+Twisted.trunk.twisted.words.im.__init__.error was deleted, deprecated since Twiseted 12.0.0
+

Index: twisted/words/im/__init__.py
===================================================================
--- twisted/words/im/__init__.py	(revision 34062)
+++ twisted/words/im/__init__.py	(working copy)
@@ -5,4 +5,4 @@
 """Instance Messenger, Pan-protocol chat client."""
 
 import warnings
-warnings.warn("twisted.im will be undergoing a rewrite at some point in the future.")
+warnings.warn("")
