Index: twisted/trial/runner.py
===================================================================
--- twisted/trial/runner.py	(revision 30599)
+++ twisted/trial/runner.py	(working copy)
@@ -604,7 +604,7 @@
                 if remove is not None:
                     try:
                         remove()
-                    except:
+                    except OSError:
                         pass
                 continue
             if pycOnly and not self.forceOrphan:
