--- /home/drew/arc/erlang-amp-1961/amp.erl	2008-01-06 16:16:12.000000000 -0500
+++ amp.erl	2008-01-07 10:10:27.000000000 -0500
@@ -184,12 +184,13 @@
             List;
         <<Prefix:16, RemainData/binary>> ->
             BinaryLength = length(binary_to_list(RemainData)),
-            case BinaryLength of
-                Prefix ->
+            P = BinaryLength < Prefix,
+            case P of
+                false ->
                     {ReadData, Remain} = split_binary(RemainData, Prefix),
                     NewList = lists:append(List, [binary_to_list(ReadData)]),
                     parse_int_prefixed(Remain, NewList);
-                _ -> 
+                true -> 
                     error
             end;
         _ ->
