Index: twisted/scripts/_twistd_unix.py
===================================================================
--- twisted/scripts/_twistd_unix.py	(revision 27268)
+++ twisted/scripts/_twistd_unix.py	(working copy)
@@ -224,11 +224,9 @@
             if e.errno == errno.EACCES or e.errno == errno.EPERM:
                 log.msg("Warning: No permission to delete pid file")
             else:
-                log.msg("Failed to unlink PID file:")
-                log.deferr()
+                log.err(e, "Failed to unlink PID file")
         except:
-            log.msg("Failed to unlink PID file:")
-            log.deferr()
+            log.err(None, "Failed to unlink PID file")
 
 
     def setupEnvironment(self, chroot, rundir, nodaemon, umask, pidfile):
