Index: twisted/internet/task.py
===================================================================
--- twisted/internet/task.py    (revision 21024)
+++ twisted/internet/task.py    (working copy)
@@ -246,6 +246,7 @@
                     self.iterators.remove(taskObj)
                     def cbContinue(result, taskObj=taskObj):
                         self.coiterate(*taskObj)
+                        return result
                     result.addCallbacks(cbContinue, doneDeferred.errback)
         self._reschedule()
