Index: twisted/python/compat.py
===================================================================
--- twisted/python/compat.py	(revision 40965)
+++ twisted/python/compat.py	(working copy)
@@ -253,8 +253,10 @@
 else:
     unicode = unicode
 
+#raw_input no longer exists in py3k; it's just called "input"
+if _PY3:
+    raw_input = input
 
-
 def nativeString(s):
     """
     Convert C{bytes} or C{unicode} to the native C{str} type, using ASCII
