diff --git twisted/conch/test/test_conch.py twisted/conch/test/test_conch.py
index f08a096..1dc88f0 100644
--- twisted/conch/test/test_conch.py
+++ twisted/conch/test/test_conch.py
@@ -570,6 +570,15 @@ class OpenSSHKeyExchangeTestCase(ConchServerSetupMixin, OpenSSHClientMixin,
             'diffie-hellman-group-exchange-sha1')
 
 
+    def test_DH_GROUP_EXCHANGE_SHA256(self):
+        """
+        The diffie-hellman-group-exchange-sha256 key exchange algorithm is
+        compatible with OpenSSH.
+        """
+        return self.assertExecuteWithKexAlgorithm(
+            'diffie-hellman-group-exchange-sha256')
+
+
 
 class OpenSSHClientForwardingTests(ForwardingMixin, OpenSSHClientMixin,
                                       unittest.TestCase):
