Index: adbapi.py
===================================================================
--- adbapi.py	(revision 23668)
+++ adbapi.py	(working copy)
@@ -7,6 +7,8 @@
 An asynchronous mapping to U{DB-API 2.0<http://www.python.org/topics/database/DatabaseAPI-2.0.html>}.
 """
 
+import sys
+
 from twisted.internet import defer, threads
 from twisted.python import reflect, log
 from twisted.python.deprecate import deprecated
@@ -55,7 +57,7 @@
             self._connection.commit()
             return
         except:
-            pass
+            log.err()
 
         self._pool.disconnect(self._connection)
 
@@ -103,7 +105,9 @@
         except:
             if not self._pool.reconnect:
                 raise
-
+            else:
+                log.err()
+        
         if self._pool.noisy:
             log.msg('Connection lost, reconnecting')
 
@@ -229,8 +233,12 @@
             conn.commit()
             return result
         except:
-            conn.rollback()
-            raise
+            e, typ, tb = sys.exc_info()
+            try:
+                conn.rollback()
+            except:
+                log.err()
+            raise e, typ, tb
         
     def runInteraction(self, interaction, *args, **kw):
         """Interact with the database and return the result.
@@ -366,7 +374,7 @@
         try:
             conn.close()
         except:
-            pass
+            log.err()
 
     def _runInteraction(self, interaction, *args, **kw):
         conn = Connection(self)
@@ -377,8 +385,12 @@
             conn.commit()
             return result
         except:
-            conn.rollback()
-            raise
+            e, typ, tb = sys.exc_info()
+            try:
+                conn.rollback()
+            except:
+                log.err()
+            raise e, typ, tb
 
     def _runQuery(self, trans, *args, **kw):
         trans.execute(*args, **kw)
