From 9b988fe9775453987f6b80845299cb8172752ea5 Mon Sep 17 00:00:00 2001
From: Vadim Markovtsev <v.markovtsev@samsung.com>
Date: Mon, 26 May 2014 11:14:10 +0400
Subject: [PATCH] Port twisted.web.client.downloadPage to Python 3

---
 twisted/web/client.py | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/twisted/web/client.py b/twisted/web/client.py
index 3bee9e5..4f91410 100644
--- a/twisted/web/client.py
+++ b/twisted/web/client.py
@@ -328,7 +328,7 @@ class HTTPClientFactory(protocol.ClientFactory):
     port = None
     path = None
 
-    def __init__(self, url, method='GET', postdata=None, headers=None,
+    def __init__(self, url, method=b'GET', postdata=None, headers=None,
                  agent=b"Twisted PageGetter", timeout=0, cookies=None,
                  followRedirect=True, redirectLimit=20,
                  afterFoundGet=False):
@@ -445,12 +445,12 @@ class HTTPDownloader(HTTPClientFactory):
     value = None
 
     def __init__(self, url, fileOrName,
-                 method='GET', postdata=None, headers=None,
-                 agent="Twisted client", supportPartial=0,
+                 method=b'GET', postdata=None, headers=None,
+                 agent=b"Twisted client", supportPartial=0,
                  timeout=0, cookies=None, followRedirect=1,
                  redirectLimit=20, afterFoundGet=False):
         self.requestedPartial = 0
-        if isinstance(fileOrName, types.StringTypes):
+        if isinstance(fileOrName, types.StringTypes if not _PY3 else str):
             self.fileName = fileOrName
             self.file = None
             if supportPartial and os.path.exists(self.fileName):
-- 
1.9.3
