Index: doc/words/examples/minchat.py
===================================================================
--- doc/words/examples/minchat.py	(revision 34050)
+++ doc/words/examples/minchat.py	(working copy)
@@ -6,13 +6,15 @@
 
 """
 A very simple twisted.im-based logbot.
+To run the script:
+$ python minchat.py
 """
 
-from twisted.im import basechat, baseaccount
+from twisted.words.im import basechat, baseaccount, ircsupport
 
 # A list of account objects. We might as well create them at runtime, this is
 # supposed to be a Minimalist Implementation, after all.
-from twisted.im import ircsupport 
+
 accounts = [
     ircsupport.IRCAccount("IRC", 1,
         "Tooty",            # nickname
