Index: twisted/python/_release.py
===================================================================
--- twisted/python/_release.py	(revision 38280)
+++ twisted/python/_release.py	(working copy)
@@ -258,7 +258,16 @@
     else:
         prerelease = ""
     data = '''\
+# Copyright (c) Twisted Matrix Laboratories.
+# See LICENSE for details.
+
+
 # This is an auto-generated file. Do not edit it.
+
+"""
+Gives version of Twisted: major.minor.micro
+"""
+
 from twisted.python import versions
 version = versions.Version(%r, %s, %s, %s%s)
 ''' % (version.package, version.major, version.minor, version.micro,
