Index: twisted/python/log.py
===================================================================
--- twisted/python/log.py	(revision 27020)
+++ twisted/python/log.py	(working copy)
@@ -629,9 +629,8 @@
         removeObserver(self._emit)
 
 
-# Some more sibling imports, at the bottom and unqualified to avoid
-# unresolvable circularity
-import threadable, failure
+# These imports are at bottom to avoid unresolvable circularity
+from twisted.python import threadable, failure
 threadable.synchronize(LogPublisher)
 
 
