=== modified file 'doc/core/howto/tutorial/protocol.xhtml'
--- doc/core/howto/tutorial/protocol.xhtml	2010-10-02 00:52:46 +0000
+++ doc/core/howto/tutorial/protocol.xhtml	2010-10-11 20:45:41 +0000
@@ -86,16 +86,16 @@
 none of our protocol classes had to be changed, and neither will have to
 change until the end of the tutorial.</p>
 
-<p>As an application <code class="API"
-base="twisted.application.service">service</code> , this new finger
-service implements the <code class="API"
-base="twisted.application.service">IService</code> interface and can
-be started and stopped in a standardized manner.  We'll make use of
-this in the next example.</p>
+<p>As an application service, this new finger service implements the <code
+class="API" base="twisted.application.service">IService</code> interface and
+can be started and stopped in a standardized manner. We'll make use of this in
+the next example.</p>
 
 <a href="listings/finger/finger13.tac" class="py-listing">finger13.tac</a>
 
-
+<p>Most application services will want to use the <code class="API"
+base="twisted.application.service">Service</code> base class, which implements
+all the generic IService behavior.</p>
 
 <h2>Read Status File</h2>
 

