--- /home/htj/src/twisted-trunk/twisted/web2/http.py	2007-07-30 15:21:33.000000000 +0200
+++ web2/http.py	2008-03-13 13:29:53.000000000 +0100
@@ -330,9 +330,9 @@
         self.doStartReading = None
         stream.ProducerStream.write(self, data)
 
-    def finish(self):
+    def finish(self, failure=None):
         self.doStartReading = None
-        stream.ProducerStream.finish(self)
+        stream.ProducerStream.finish(self, failure)
 
 
 # response codes that must have empty bodies
@@ -399,7 +399,8 @@
         
     def connectionLost(self, reason):
         """connection was lost"""
-        pass
+        self.stream.unregisterProducer()
+        self.stream.finish(reason)
 
     def __repr__(self):
         return '<%s %s %s>'% (self.method, self.uri, self.clientproto)
