Index: twisted/test/test_strcred.py
===================================================================
--- twisted/test/test_strcred.py	(revision 29007)
+++ twisted/test/test_strcred.py	(working copy)
@@ -244,10 +244,15 @@
 
 
     if None in (pwd, spwd, crypt):
+        availability = []
+        for module, name in ((pwd, "pwd"), (spwd, "swpd"), (crypt, "crypt")):
+            if module is None:
+                availability += [name]
         for method in (test_unixCheckerSucceeds,
                        test_unixCheckerFailsUsername,
                        test_unixCheckerFailsPassword):
-            method.skip = 'pwd and spwd are both unavailable'
+            method.skip = ("Required module(s) are unavailable: " + 
+                           ", ".join(availability))
 
 
 
