Index: test_strcred.py
===================================================================
--- test_strcred.py	(revision 28994)
+++ test_strcred.py	(working copy)
@@ -244,10 +244,21 @@
 
 
     if None in (pwd, spwd, crypt):
+        availability = []
+        if pwd is None:
+            availability += ["pwd"]
+        if spwd is None:
+            availability += ["spwd"]
+        if crypt is None:
+            availability += ["crypt"]
+        if len(availability) == 1:
+            availability = availability[0] + " is unavailable"
+        else:
+            availability = " and ".join(availability) + " are unavailable"
         for method in (test_unixCheckerSucceeds,
                        test_unixCheckerFailsUsername,
                        test_unixCheckerFailsPassword):
-            method.skip = 'pwd and spwd are both unavailable'
+            method.skip = availability
 
 
 
