diff --git a/twisted/web/http.py b/twisted/web/http.py
index 18dfec2..df4281b 100644
--- a/twisted/web/http.py
+++ b/twisted/web/http.py
@@ -1031,27 +1031,60 @@ class Request:
             else:
                 self.transport.write(data)
 
-    def addCookie(self, k, v, expires=None, domain=None, path=None, max_age=None, comment=None, secure=None):
+    def addCookie(self, k, v, expires=None, domain=None, path=None,
+                  max_age=None, comment=None, secure=None, httpOnly=False):
         """
         Set an outgoing HTTP cookie.
 
         In general, you should consider using sessions instead of cookies, see
         L{twisted.web.server.Request.getSession} and the
         L{twisted.web.server.Session} class for details.
+
+        @param k: cookie name
+        @type k: L{str}
+
+        @param v: cookie value
+        @type v: L{str}
+
+        @param expires: cookie expire attribute value in
+        "Wdy, DD Mon YYYY HH:MM:SS GMT" format
+        @type expires: L{str}
+
+        @param domain: cookie domain
+        @type domain: L{str}
+
+        @param path: cookie path
+        @type path: L{str}
+
+        @param max_age: cookie expiration in seconds from reception
+        @type max_age: L{str}
+
+        @param comment: cookie comment
+        @type comment: L{str}
+
+        @param secure: direct browser to send the cookie on encrypted
+        connections only
+        @type secure: L{bool}
+
+        @param httpOnly: direct browser not to expose cookies through channels
+        other than HTTP (and HTTPS) requests
+        @type httpOnly: L{bool}
         """
         cookie = '%s=%s' % (k, v)
         if expires is not None:
-            cookie = cookie +"; Expires=%s" % expires
+            cookie = cookie + "; Expires=%s" % (expires, )
         if domain is not None:
-            cookie = cookie +"; Domain=%s" % domain
+            cookie = cookie + "; Domain=%s" % (domain, )
         if path is not None:
-            cookie = cookie +"; Path=%s" % path
+            cookie = cookie + "; Path=%s" % (path, )
         if max_age is not None:
-            cookie = cookie +"; Max-Age=%s" % max_age
+            cookie = cookie + "; Max-Age=%s" % (max_age, )
         if comment is not None:
-            cookie = cookie +"; Comment=%s" % comment
+            cookie = cookie + "; Comment=%s" % (comment, )
         if secure:
-            cookie = cookie +"; Secure"
+            cookie = cookie + "; Secure"
+        if httpOnly:
+            cookie = cookie + "; HttpOnly"
         self.cookies.append(cookie)
 
     def setResponseCode(self, code, message=None):
diff --git a/twisted/web/test/test_http.py b/twisted/web/test/test_http.py
index 0d6b5c3..da83d37 100644
--- a/twisted/web/test/test_http.py
+++ b/twisted/web/test/test_http.py
@@ -1529,6 +1529,31 @@ class RequestTests(unittest.TestCase, ResponseTestMixin):
         self.assertEqual(req.responseHeaders.getRawHeaders(b"test"), [b"lemur"])
 
 
+    def test_addCookieWithMinimumArguments(self):
+        """
+        Add a Set-Cookie header with just name and value to the response.
+        """
+        req = http.Request(DummyChannel(), False)
+        req.addCookie("foo", "bar")
+        self.assertEqual(req.cookies[0], "foo=bar")
+
+
+    def test_addCookieWithAllArguments(self):
+        """
+        Add a Set-Cookie header with name and value and all the supported
+        options to the response.
+        """
+        req = http.Request(DummyChannel(), False)
+        req.addCookie(
+            "foo", "bar", expires="Fri, 31 Dec 9999 23:59:59 GMT",
+            domain=".example.com", path="/", max_age="31536000",
+            comment="test", secure=True, httpOnly=True)
+        self.assertEqual(req.cookies[0],
+                         "foo=bar; Expires=Fri, 31 Dec 9999 23:59:59 GMT; "
+                         "Domain=.example.com; Path=/; Max-Age=31536000; "
+                         "Comment=test; Secure; HttpOnly")
+
+
     def test_firstWrite(self):
         """
         For an HTTP 1.0 request, L{http.Request.write} sends an HTTP 1.0
diff --git a/twisted/web/topfiles/5911.feature b/twisted/web/topfiles/5911.feature
new file mode 100644
index 0000000..6560891
--- /dev/null
+++ b/twisted/web/topfiles/5911.feature
@@ -0,0 +1 @@
+twisted.web.http.Request.addCookie now supports httpOnly attribute.
