diff --git a/twisted/web/http.py b/twisted/web/http.py
index c9bd313..6b6d2bb 100644
--- a/twisted/web/http.py
+++ b/twisted/web/http.py
@@ -1022,27 +1022,53 @@ class Request:
             else:
                 self.transport.write(data)
 
-    def addCookie(self, k, v, expires=None, domain=None, path=None, max_age=None, comment=None, secure=None):
+    def addCookie(self, k, v, expires=None, domain=None, path=None,
+                  max_age=None, comment=None, secure=None, httpOnly=False):
         """
         Set an outgoing HTTP cookie.
 
         In general, you should consider using sessions instead of cookies, see
         L{twisted.web.server.Request.getSession} and the
         L{twisted.web.server.Session} class for details.
+
+        @param k: cookie name
+
+        @param v: cookie value
+
+        @param expires: cookie expire attribute value in
+        "Wdy, DD Mon YYYY HH:MM:SS GMT" format
+
+        @param domain: cookie domain
+
+        @param path: cookie path
+
+        @param max_age: cookie expiration in seconds from reception
+
+        @param comment: cookie comment
+
+        @param secure: direct browser to send the cookie on encrypted
+        connections only
+
+        @param httpOnly: direct browser not to expose cookies through channels
+        other than HTTP (and HTTPS) requests
+
+
         """
         cookie = '%s=%s' % (k, v)
         if expires is not None:
-            cookie = cookie +"; Expires=%s" % expires
+            cookie = cookie + "; Expires=%s" % (expires, )
         if domain is not None:
-            cookie = cookie +"; Domain=%s" % domain
+            cookie = cookie + "; Domain=%s" % (domain, )
         if path is not None:
-            cookie = cookie +"; Path=%s" % path
+            cookie = cookie + "; Path=%s" % (path, )
         if max_age is not None:
-            cookie = cookie +"; Max-Age=%s" % max_age
+            cookie = cookie + "; Max-Age=%s" % (max_age, )
         if comment is not None:
-            cookie = cookie +"; Comment=%s" % comment
+            cookie = cookie + "; Comment=%s" % (comment, )
         if secure:
-            cookie = cookie +"; Secure"
+            cookie = cookie + "; Secure"
+        if httpOnly:
+            cookie = cookie + "; HttpOnly"
         self.cookies.append(cookie)
 
     def setResponseCode(self, code, message=None):
diff --git a/twisted/web/test/test_http.py b/twisted/web/test/test_http.py
index 81ab830..2b36bba 100644
--- a/twisted/web/test/test_http.py
+++ b/twisted/web/test/test_http.py
@@ -1463,6 +1463,34 @@ class RequestTests(unittest.TestCase, ResponseTestMixin):
         self.assertEqual(req.responseHeaders.getRawHeaders(b"test"), [b"lemur"])
 
 
+    def test_addCookie(self):
+        """
+        L{http.Request.addCookie} adds a Set-Cookie header to the response.
+        """
+        req = http.Request(DummyChannel(), False)
+        req.addCookie("foo", "bar")
+        self.assertEqual("foo=bar", req.cookies[0])
+
+
+    def test_addCookieWithAttibutes(self):
+        """
+        L{http.Request.addCookie} adds a Set-Cookie header with cookie
+        attributes to the response.
+        """
+        req = http.Request(DummyChannel(), False)
+        req.addCookie(
+            "foo", "bar", expires="Fri, 31 Dec 9999 23:59:59 GMT",
+            domain=".example.com", path="/", max_age="31536000",
+            comment="test", secure=True, httpOnly=True)
+        self.assertIn("Expires=Fri, 31 Dec 9999 23:59:59 GMT", req.cookies[0])
+        self.assertIn("Domain=.example.com", req.cookies[0])
+        self.assertIn("Path=/", req.cookies[0])
+        self.assertIn("Max-Age=31536000", req.cookies[0])
+        self.assertIn("Comment=test", req.cookies[0])
+        self.assertIn("Secure", req.cookies[0])
+        self.assertIn("HttpOnly", req.cookies[0])
+
+
     def test_firstWrite(self):
         """
         For an HTTP 1.0 request, L{http.Request.write} sends an HTTP 1.0
diff --git a/twisted/web/topfiles/5911.feature b/twisted/web/topfiles/5911.feature
new file mode 100644
index 0000000..6560891
--- /dev/null
+++ b/twisted/web/topfiles/5911.feature
@@ -0,0 +1 @@
+twisted.web.http.Request.addCookie now supports httpOnly attribute.
