Index: twisted/web/http.py
===================================================================
--- twisted/web/http.py	(revision 35367)
+++ twisted/web/http.py	(working copy)
@@ -952,7 +952,7 @@
             else:
                 self.transport.write(data)
 
-    def addCookie(self, k, v, expires=None, domain=None, path=None, max_age=None, comment=None, secure=None):
+    def addCookie(self, k, v, expires=None, domain=None, path=None, max_age=None, comment=None, secure=None, httponly=None):
         """
         Set an outgoing HTTP cookie.
 
@@ -973,6 +973,8 @@
             cookie = cookie +"; Comment=%s" % comment
         if secure:
             cookie = cookie +"; Secure"
+        if httponly:
+            cookie = cookie +"; HttpOnly"
         self.cookies.append(cookie)
 
     def setResponseCode(self, code, message=None):
Index: twisted/web/iweb.py
===================================================================
--- twisted/web/iweb.py	(revision 35367)
+++ twisted/web/iweb.py	(working copy)
@@ -218,7 +218,7 @@
         """
 
 
-    def addCookie(k, v, expires=None, domain=None, path=None, max_age=None, comment=None, secure=None):
+    def addCookie(k, v, expires=None, domain=None, path=None, max_age=None, comment=None, secure=None, httponly=None):
         """
         Set an outgoing HTTP cookie.
 
