diff --git a/twisted/web/http.py b/twisted/web/http.py
index 84a6ab7..0bb325d 100644
--- a/twisted/web/http.py
+++ b/twisted/web/http.py
@@ -428,7 +428,10 @@ class HTTPClient(basic.LineReceiver):
             'field-name: value'.
         @type header: C{bytes}
         """
-        key, val = header.split(b':', 1)
+        try:
+            key, val = header.split(b':', 1)
+        except ValueError:
+            return
         val = val.lstrip()
         self.handleHeader(key, val)
         if key.lower() == b'content-length':
diff --git a/twisted/web/test/test_http.py b/twisted/web/test/test_http.py
index 28729ee..f67e5c6 100644
--- a/twisted/web/test/test_http.py
+++ b/twisted/web/test/test_http.py
@@ -2521,6 +2521,18 @@ class MultilineHeadersTests(unittest.TestCase):
 
         self.assertEqual(c.length, 10)
 
+    def test_extractIncorrectHeader(self):
+        c = ClientDriver()
+        # def check_headers()
+        c.handleHeader = self.ourHandleHeader
+        c.handleEndHeaders = self.ourHandleEndHeaders
+        c.lineReceived(b'HTTP/1.0 201')
+        c.lineReceived(b'this is not a correct header')
+        self.assertFalse(self.handleHeaderCalled)
+
+        c.lineReceived(b'')
+        self.assertFalse(self.handleHeaderCalled)
+
 
     def test_noHeaders(self):
         """
