Index: twisted/lore/lint.py
===================================================================
--- twisted/lore/lint.py	(revision 27446)
+++ twisted/lore/lint.py	(working copy)
@@ -100,7 +100,7 @@
                     self._reportError(filename, node, 
                                       'text wider than 80 columns in pre')
         for node in domhelpers.findNodesNamed(dom, 'a'):
-            if node.getAttribute('class', '').endswith('listing'):
+            if node.getAttribute('class').endswith('listing'):
                 try:
                     fn = os.path.dirname(filename) 
                     fn = os.path.join(fn, node.getAttribute('href'))
Index: twisted/lore/test/test_lint.py
===================================================================
--- twisted/lore/test/test_lint.py	(revision 27446)
+++ twisted/lore/test/test_lint.py	(working copy)
@@ -61,3 +61,24 @@
         patch.restore()
 
         self.assertEqual(output.getvalue(), "")
+        
+    def test_aNode(self):
+        '''
+        If there is an <a> tag in the document, the checker returned by
+        L{getDefaultChecker} does not report an error.
+        '''
+        documentSource = (
+            '<html>'
+            '<head><title>foo</title></head>'
+            '<body><h1>foo</h1><a>A link.</a></body>'
+            '</html>')
+        document = minidom.parseString(documentSource)
+        filename = self.mktemp()
+        checker = getDefaultChecker()
+
+        output = StringIO()
+        patch = self.patch(sys, 'stdout', output)
+        checker.check(document, filename)
+        patch.restore()
+
+        self.assertEqual(output.getvalue(), "")
