Index: twisted/lore/lint.py
===================================================================
--- twisted/lore/lint.py	(revision 27366)
+++ twisted/lore/lint.py	(working copy)
@@ -100,7 +100,7 @@
                     self._reportError(filename, node, 
                                       'text wider than 80 columns in pre')
         for node in domhelpers.findNodesNamed(dom, 'a'):
-            if node.getAttribute('class', '').endswith('listing'):
+            if node.getAttribute('class').endswith('listing'):
                 try:
                     fn = os.path.dirname(filename) 
                     fn = os.path.join(fn, node.getAttribute('href'))
