Index: twisted/internet/interfaces.py
===================================================================
--- twisted/internet/interfaces.py	(revision 29255)
+++ twisted/internet/interfaces.py	(working copy)
@@ -1540,7 +1540,7 @@
         Send a signal to the process.
 
         @param signalID: can be
-          - one of C{\"HUP\"}, C{\"KILL\"}, C{\"STOP\"}, or C{\"INT\"}.
+          - one of C{\"KILL\"}, C{\"TERM\"}, or C{\"INT\"}.
               These will be implemented in a
               cross-platform manner, and so should be used
               if possible.
Index: twisted/internet/process.py
===================================================================
--- twisted/internet/process.py	(revision 29255)
+++ twisted/internet/process.py	(working copy)
@@ -23,6 +23,9 @@
 except ImportError:
     fcntl = None
 
+from zope.interface import implements
+from twisted.internet import interfaces
+
 from twisted.python import log, failure
 from twisted.python.util import switchUID
 from twisted.internet import fdesc, abstract, error
@@ -476,6 +479,7 @@
     code is not cross-platform. (also, windows can only select
     on sockets...)
     """
+    implements(interfaces.IProcessTransport)
 
     debug = False
     debug_child = False
