--- twisted/web/xmlrpc.py.orig	2008-08-27 14:55:17.000000000 -0400
+++ twisted/web/xmlrpc.py	2008-08-27 14:59:27.000000000 -0400
@@ -313,12 +313,13 @@
             return
         try:
             response = xmlrpclib.loads(contents)
+            response = response[0][0]
         except:
             deferred, self.deferred = self.deferred, None
             deferred.errback(failure.Failure())
         else:
             deferred, self.deferred = self.deferred, None
-            deferred.callback(response[0][0])
+            deferred.callback(response)
 
     def clientConnectionLost(self, _, reason):
         if self.deferred is not None:
