diff --git twisted/manhole/_inspectro.py twisted/manhole/_inspectro.py
index 430ae7b..fc9a305 100644
--- twisted/manhole/_inspectro.py
+++ twisted/manhole/_inspectro.py
@@ -315,7 +315,7 @@ class LogViewer:
             vals.append(p.outLog[0][0])
         self.startTime = min(vals)
         p.logViewer = self
-        self.xml = x = gtk.glade.XML(sibpath(__file__, "logview.glade"))
+        self.xml = gtk.glade.XML(sibpath(__file__, "logview.glade"))
         self.xml.signal_autoconnect(self)
         self.loglist = self.xml.get_widget("loglist")
         # setup model, connect it to my treeview
diff --git twisted/manhole/gladereactor.py twisted/manhole/gladereactor.py
index 148fc5e..6fc7cfd 100644
--- twisted/manhole/gladereactor.py
+++ twisted/manhole/gladereactor.py
@@ -10,7 +10,7 @@ suspend, resume and inspect transports interactively.
 __all__ = ['install']
 
 # Twisted Imports
-from twisted.python import log, threadable, runtime, failure, util, reflect
+from twisted.python import util, reflect
 from twisted.internet.gtk2reactor import Gtk2Reactor as sup
 
 import gtk
diff --git twisted/manhole/service.py twisted/manhole/service.py
index c9d4679..84fb589 100644
--- twisted/manhole/service.py
+++ twisted/manhole/service.py
@@ -17,13 +17,9 @@ from zope.interface import implements, Interface
 # sibling imports
 import explorer
 
-# system imports
-from cStringIO import StringIO
-
 import string
 import sys
 import traceback
-import types
 
 
 class FakeStdIO:
diff --git twisted/manhole/telnet.py twisted/manhole/telnet.py
index d63b3a6..c689a16 100644
--- twisted/manhole/telnet.py
+++ twisted/manhole/telnet.py
@@ -11,7 +11,6 @@ from twisted.python import log, failure
 
 # system imports
 import string, copy, sys
-from cStringIO import StringIO
 
 
 class Shell(telnet.Telnet):
@@ -74,7 +73,6 @@ class Shell(telnet.Telnet):
                         self.write('\r\n>>> ')
                         return
                 except:
-                    io = StringIO()
                     failure.Failure().printTraceback(file=self)
                     log.deferr()
                     self.write('\r\n>>> ')
diff --git twisted/manhole/ui/gtk2manhole.py twisted/manhole/ui/gtk2manhole.py
index 2c6a532..9985980 100644
--- twisted/manhole/ui/gtk2manhole.py
+++ twisted/manhole/ui/gtk2manhole.py
@@ -218,7 +218,6 @@ class ConsoleInput:
             self.rkeymap[getattr(gtk.keysyms, keysymName)] = keysymName
 
     def _on_key_press_event(self, entry, event):
-        stopSignal = False
         ksym = self.rkeymap.get(event.keyval, None)
 
         mods = []
diff --git twisted/tap/manhole.py twisted/tap/manhole.py
index 8d727fa..d3d4f32 100644
--- twisted/tap/manhole.py
+++ twisted/tap/manhole.py
@@ -12,7 +12,7 @@ from twisted.spread import pb
 from twisted.python import usage, util
 from twisted.cred import portal, checkers
 from twisted.application import strports
-import os, sys
+
 
 class Options(usage.Options):
     synopsis = "[options]"
diff --git twisted/tap/socks.py twisted/tap/socks.py
index e0780ad..a418f0c 100644
--- twisted/tap/socks.py
+++ twisted/tap/socks.py
@@ -10,7 +10,7 @@ I am a support module for making SOCKSv4 servers with twistd.
 from twisted.protocols import socks
 from twisted.python import usage
 from twisted.application import internet
-import sys
+
 
 class Options(usage.Options):
     synopsis = "[-i <interface>] [-p <port>] [-l <file>]"
