Index: src/twisted/python/components.py
===================================================================
--- src/twisted/python/components.py	(revision 13659)
+++ src/twisted/python/components.py	(working copy)
@@ -69,6 +69,8 @@
 except ImportError:
     raise ImportError, "you need zope.interface installed (http://zope.org/Products/ZopeInterface/)"
 
+_marker = object()
+
 class ComponentsDeprecationWarning(DeprecationWarning):
     """So you can filter new-components related deprecations easier."""
     pass
@@ -350,15 +352,22 @@
     def __getattr__(self, attr):
         if attr.startswith("_v_"):
             raise AttributeError # z.i internal thing,
-        if attr != "__instadapt__": # __instadapt__ is part of our own backwards compat layer
-            warnings.warn("Don't get attributes (in this case, %r) off Interface, use "
-                          ".queryDescriptionFor() etc. instead" % (attr,),
-                          ComponentsDeprecationWarning, stacklevel=3)
+
+        value = _marker
         if self.__attrs.has_key(attr):
-            return self.__attrs[attr]
-        result = self.queryDescriptionFor(attr)
-        if result != None:
-            return result
+            value = self.__attrs[attr]
+        else:
+            value = self.queryDescriptionFor(attr, _marker)
+
+        if value != _marker:
+            # __instadapt__ is part of our own backwards compat layer
+            if attr != "__instadapt__": 
+                warnings.warn(
+                    "Don't get attributes (in this case, %r) off Interface, "
+                    "use .queryDescriptionFor() etc. instead" % (attr,),
+                    ComponentsDeprecationWarning, stacklevel=3)
+            return value
+
         raise AttributeError, attr
 
 
