Index: twisted/conch/scripts/cftp.py
===================================================================
--- twisted/conch/scripts/cftp.py	(revision 40413)
+++ twisted/conch/scripts/cftp.py	(working copy)
@@ -809,7 +809,7 @@
 
     def eofReceived(self):
         log.msg('got eof')
-        self.stdio.closeStdin()
+        self.stdio.loseWriteConnection()
 
     def closeReceived(self):
         log.msg('remote side closed %s' % self)
Index: twisted/internet/_posixstdio.py
===================================================================
--- twisted/internet/_posixstdio.py	(revision 40413)
+++ twisted/internet/_posixstdio.py	(working copy)
@@ -11,7 +11,6 @@
 Maintainer: James Y Knight
 """
 
-import warnings
 from zope.interface import implements
 
 from twisted.internet import process, error, interfaces
@@ -159,13 +158,6 @@
         if self._reader is not None:
             self._reader.resumeProducing()
 
-    # Stupid compatibility:
-    def closeStdin(self):
-        """Compatibility only, don't use. Same as loseWriteConnection."""
-        warnings.warn("This function is deprecated, use loseWriteConnection instead.",
-                      category=DeprecationWarning, stacklevel=2)
-        self.loseWriteConnection()
-
     def stopReading(self):
         """Compatibility only, don't use. Call pauseProducing."""
         self.pauseProducing()
Index: twisted/topfiles/6785.removal
===================================================================
--- twisted/topfiles/6785.removal	(revision 0)
+++ twisted/topfiles/6785.removal	(working copy)
@@ -0,0 +1 @@
+twisted.internet._posixstdio.closeStdin, deprecated since Twisted 2.1, has been removed.
