Index: twisted/internet/_posixstdio.py
===================================================================
--- twisted/internet/_posixstdio.py	(revision 40413)
+++ twisted/internet/_posixstdio.py	(working copy)
@@ -159,13 +159,6 @@
         if self._reader is not None:
             self._reader.resumeProducing()
 
-    # Stupid compatibility:
-    def closeStdin(self):
-        """Compatibility only, don't use. Same as loseWriteConnection."""
-        warnings.warn("This function is deprecated, use loseWriteConnection instead.",
-                      category=DeprecationWarning, stacklevel=2)
-        self.loseWriteConnection()
-
     def stopReading(self):
         """Compatibility only, don't use. Call pauseProducing."""
         self.pauseProducing()
Index: twisted/topfiles/6785.removal
===================================================================
--- twisted/topfiles/6785.removal	(revision 0)
+++ twisted/topfiles/6785.removal	(working copy)
@@ -0,0 +1 @@
+twisted.internet._posixstdio.closeStdin, deprecated, has been removed.
