diff --git twisted/internet/test/test_protocol.py twisted/internet/test/test_protocol.py
index 4cb1f23..c88cc18 100644
--- twisted/internet/test/test_protocol.py
+++ twisted/internet/test/test_protocol.py
@@ -19,7 +19,7 @@ from twisted.internet.protocol import (
     ConsumerToProtocolAdapter)
 from twisted.trial.unittest import TestCase
 from twisted.test.proto_helpers import MemoryReactorClock, StringTransport
-from twisted.logger import Logger, formatEvent, LogLevel
+from twisted.logger import Logger, LogLevel
 
 
 
