diff --git twisted/internet/protocol.py twisted/internet/protocol.py
index 73ab587..4c5b7da 100644
--- twisted/internet/protocol.py
+++ twisted/internet/protocol.py
@@ -34,7 +34,7 @@ class Factory:
     numPorts = 0
     noisy = True
 
-    log = Logger()
+    _log = Logger()
 
     @classmethod
     def forProtocol(cls, protocol, *args, **kwargs):
@@ -71,7 +71,7 @@ class Factory:
         """
         if not self.numPorts:
             if self.noisy:
-                self.log.info("Starting factory {factory!r}", factory=self)
+                self._log.info("Starting factory {factory!r}", factory=self)
             self.startFactory()
         self.numPorts = self.numPorts + 1
 
@@ -87,7 +87,7 @@ class Factory:
         self.numPorts = self.numPorts - 1
         if not self.numPorts:
             if self.noisy:
-                self.log.info("Stopping factory {factory!r}", factory=self)
+                self._log.info("Stopping factory {factory!r}", factory=self)
             self.stopFactory()
 
     def startFactory(self):
diff --git twisted/internet/test/test_protocol.py twisted/internet/test/test_protocol.py
index 3d47031..4cb1f23 100644
--- twisted/internet/test/test_protocol.py
+++ twisted/internet/test/test_protocol.py
@@ -401,7 +401,7 @@ class FactoryTests(TestCase):
         """
         events = []
         class MyFactory(Factory):
-            log = Logger(observer=events.append)
+            _log = Logger(observer=events.append)
         f = MyFactory()
 
         f.doStart()
@@ -410,10 +410,6 @@ class FactoryTests(TestCase):
         self.assertEqual(events[0]['log_level'], LogLevel.info)
         self.assertEqual(events[0]['log_format'],
                          'Starting factory {factory!r}')
-        self.assertEqual(
-            formatEvent(events[0]),
-            u'Starting factory {factory!r}'.format(factory=f)
-        )
 
 
     def test_doStopLoggingStatement(self):
@@ -423,7 +419,7 @@ class FactoryTests(TestCase):
         """
         events = []
         class MyFactory(Factory):
-            log = Logger(observer=events.append)
+            _log = Logger(observer=events.append)
             numPorts = 1
         f = MyFactory()
 
@@ -433,10 +429,6 @@ class FactoryTests(TestCase):
         self.assertEqual(events[0]['log_level'], LogLevel.info)
         self.assertEqual(events[0]['log_format'],
                          'Stopping factory {factory!r}')
-        self.assertEqual(
-            formatEvent(events[0]),
-            u'Stopping factory {factory!r}'.format(factory=f)
-        )
 
 
 
diff --git twisted/topfiles/7897.feature twisted/topfiles/7897.feature
index 14e6b5e..4972629 100644
--- twisted/topfiles/7897.feature
+++ twisted/topfiles/7897.feature
@@ -1 +1 @@
-twisted.internet.protocol.Factory now uses the new logging system (twisted.logger) for all it's logging statements.
+twisted.internet.protocol.Factory now uses the new logging system (twisted.logger) for all its logging statements.
