Index: doc/core/howto/tutorial/listings/finger/finger15.tac
===================================================================
--- doc/core/howto/tutorial/listings/finger/finger15.tac	(revision 26633)
+++ doc/core/howto/tutorial/listings/finger/finger15.tac	(working copy)
@@ -60,7 +60,6 @@
         f = protocol.ServerFactory()
         f.protocol = FingerProtocol
         f.getUser = self.getUser
-        f.startService = self.startService
         return f
 
     def getResource(self):
