Index: doc/names/examples/gethostbyname.py
===================================================================
--- doc/names/examples/gethostbyname.py	(revision 34050)
+++ doc/names/examples/gethostbyname.py	(working copy)
@@ -3,6 +3,13 @@
 # Copyright (c) Twisted Matrix Laboratories.
 # See LICENSE for details.
 
+"""
+Returns the IP address for a given hostname.
+To run this script:
+$ python gethostbyname.py <hostname>
+e.g.: 
+$ python gethostbyname.py www.google.com
+"""
 import sys
 from twisted.names import client
 from twisted.internet import reactor
Index: doc/names/examples/index.xhtml
===================================================================
--- doc/names/examples/index.xhtml	(revision 34050)
+++ doc/names/examples/index.xhtml	(working copy)
@@ -12,9 +12,9 @@
 
     <h2>DNS (Twisted Names)</h2>
     <ul>
-        <li><a href="testdns.py">testdns.py</a></li>
-        <li><a href="dns-service.py">dns-service.py</a></li>
-        <li><a href="gethostbyname.py">gethostbyname.py</a></li>
+        <li><a href="testdns.py">testdns.py</a> - Prints the results of an Address record lookup, Mail-Exchanger record lookup, and Nameserver record lookup for the given hostname for a given hostname.</li>
+        <li><a href="dns-service.py">dns-service.py</a> - Searches for SRV records in DNS.</li>
+        <li><a href="gethostbyname.py">gethostbyname.py</a> - Returns the IP address for a given hostname.</li>
     </ul>
 </body>
 </html>
Index: doc/names/examples/dns-service.py
===================================================================
--- doc/names/examples/dns-service.py	(revision 34050)
+++ doc/names/examples/dns-service.py	(working copy)
@@ -5,6 +5,15 @@
 
 """
 Sample app to lookup SRV records in DNS.
+To run this script:
+$ python dns-service.py <service> <proto> <domain>
+where,
+service = the symbolic name of the desired service.
+proto = the transport protocol of the desired service; this is usually either TCP or UDP.
+domain =  the domain name for which this record is valid.
+e.g.: 
+$ python dns-service.py sip udp yahoo.com
+$ python dns-service.py xmpp-client tcp gmail.com
 """
 
 from twisted.names import client
Index: doc/names/examples/testdns.py
===================================================================
--- doc/names/examples/testdns.py	(revision 34050)
+++ doc/names/examples/testdns.py	(working copy)
@@ -3,6 +3,14 @@
 # Copyright (c) Twisted Matrix Laboratories.
 # See LICENSE for details.
 
+"""
+Prints the results of an Address record lookup, Mail-Exchanger record lookup, and Nameserver record lookup for the given hostname for a given hostname.
+To run this script:
+$ python testdns.py <hostname>
+e.g.: 
+$ python testdns.py www.google.com
+"""
+
 import sys
 from twisted.names import client
 from twisted.internet import reactor
