Index: words/im/interfaces.py
===================================================================
--- words/im/interfaces.py	(revision 16185)
+++ words/im/interfaces.py	(working copy)
@@ -169,7 +169,7 @@
         @param metadata: Valid keys for this dictionary include:
         
             style: associated with one of:
-                emote: indicates this is an action
+            emote: indicates this is an action
         """
 
     def join(self):
Index: words/xish/xmlstream.py
===================================================================
--- words/xish/xmlstream.py	(revision 16185)
+++ words/xish/xmlstream.py	(working copy)
@@ -12,7 +12,7 @@
 The most prominent use of XML Streams is Jabber, but this module is generically
 usable. See Twisted Words for Jabber specific protocol support.
 
-Maintainer: U{Ralph Meijer<mailto:twisted@ralphm.ik.nu>)
+Maintainer: U{Ralph Meijer<mailto:twisted@ralphm.ik.nu>}
 """
 
 from twisted.internet import protocol
@@ -97,7 +97,7 @@
 
     def onElement(self, element):
         """ Called whenever a direct child element of the root element has
-            been received.
+        been received.
 
         Dispatches the received element.
         """
Index: web2/http_headers.py
===================================================================
--- web2/http_headers.py	(revision 16185)
+++ web2/http_headers.py	(working copy)
@@ -164,12 +164,12 @@
     """Tokenize a string according to normal HTTP header parsing rules.
 
     In particular:
-    - Whitespace is irrelevant and eaten next to special separator tokens.
-      Its existance (but not amount) is important between character strings.
-    - Quoted string support including embedded backslashes.
-    - Case is insignificant (and thus lowercased), except in quoted strings.
-       (unless foldCase=False)
-    - Multiple headers are concatenated with ','
+     - Whitespace is irrelevant and eaten next to special separator tokens.
+       Its existance (but not amount) is important between character strings.
+     - Quoted string support including embedded backslashes.
+     - Case is insignificant (and thus lowercased), except in quoted strings.
+        (unless foldCase=False)
+     - Multiple headers are concatenated with ','
 
     NOTE: not all headers can be parsed with this function.
 
Index: spread/pb.py
===================================================================
--- spread/pb.py	(revision 16185)
+++ spread/pb.py	(working copy)
@@ -1569,7 +1569,7 @@
     def getRootObject(self):
         """Get root object of remote PB server.
 
-        @return Deferred of the root object.
+        @return: Deferred of the root object.
         """
         if self._broker and not self._broker.disconnected:
            return defer.succeed(self._root)
@@ -1583,7 +1583,7 @@
 
         New systems should use login() instead.
 
-        @return Deferred of RemoteReference to the perspective.
+        @return: Deferred of RemoteReference to the perspective.
         """
         warnings.warn("Update your backend to use PBServerFactory, and then use login().",
                       DeprecationWarning, 2)
