diff --git twisted/python/.compat.py.swp twisted/python/.compat.py.swp
new file mode 100644
index 0000000..a7e3155
Binary files /dev/null and twisted/python/.compat.py.swp differ
diff --git twisted/python/_release.py twisted/python/_release.py
index 8d80d45..af4a071 100644
--- twisted/python/_release.py
+++ twisted/python/_release.py
@@ -359,7 +359,7 @@ class NewsBuilder(object):
     """
     Generate the new section of a NEWS file.
 
-    The C{_FEATURE}, C{_BUGFIX}, C{_DOC}, C{_REMOVAL}, and C{_MISC}
+    The C{_FEATURE}, C{_BUGFIX}, C{_DOC}, C{_REMOVAL}, C{_PORT} and C{_MISC}
     attributes of this class are symbolic names for the news entry types
     which are supported.  Conveniently, they each also take on the value of
     the file name extension which indicates a news entry of that type.
@@ -380,6 +380,7 @@ class NewsBuilder(object):
     _BUGFIX = ".bugfix"
     _DOC = ".doc"
     _REMOVAL = ".removal"
+    _PORT = ".port"
     _MISC = ".misc"
 
     _headings = {
@@ -387,6 +388,7 @@ class NewsBuilder(object):
         _BUGFIX: "Bugfixes",
         _DOC: "Improved Documentation",
         _REMOVAL: "Deprecations and Removals",
+        _PORT: "Porting efforts to new environments like Python 3",
         _MISC: "Other"}
 
     _NO_CHANGES = "No significant changes have been made for this release.\n"
@@ -412,7 +414,8 @@ class NewsBuilder(object):
 
         @param ticketType: The type of news entries to search for.  One of
             L{NewsBuilder._FEATURE}, L{NewsBuilder._BUGFIX},
-            L{NewsBuilder._REMOVAL}, or L{NewsBuilder._MISC}.
+            L{NewsBuilder._REMOVAL}, L{NewsBuilder._PORT}
+            or L{NewsBuilder._MISC}.
 
         @return: A C{list} of two-tuples.  The first element is the ticket
             number as an C{int}.  The second element of each tuple is the
@@ -528,7 +531,8 @@ class NewsBuilder(object):
         @raise NotWorkingDirectory: If the C{path} is not an SVN checkout.
         """
         changes = []
-        for part in (self._FEATURE, self._BUGFIX, self._DOC, self._REMOVAL):
+        for part in (self._FEATURE, self._BUGFIX, self._DOC, self._REMOVAL,
+                     self._PORT):
             tickets = self._findChanges(path, part)
             if tickets:
                 changes.append((part, tickets))
diff --git twisted/python/topfiles/7628.feature twisted/python/topfiles/7628.feature
new file mode 100644
index 0000000..83cfb6d
--- /dev/null
+++ twisted/python/topfiles/7628.feature
@@ -0,0 +1 @@
+twisted.python._release The NEWS builder gets a new change type 'port' for porting efforts
