Index: doc/core/howto/pb-usage.xhtml
===================================================================
--- doc/core/howto/pb-usage.xhtml	(revision 25153)
+++ doc/core/howto/pb-usage.xhtml	(working copy)
@@ -212,8 +212,7 @@
 exit.  If you start up a server, you can see:</p>
 
 <pre class="shell">
-% <em>mktap qotd --pb 8787</em>
-% <em>twistd -f qotd.tap</em>
+% <em>twistd qotd --pb 8787</em>
 % <em>python -c 'import TwistedQuotes.pbquoteclient'</em>
 An apple a day keeps the doctor away.
 </pre>
Index: doc/core/howto/telnet.xhtml
===================================================================
--- doc/core/howto/telnet.xhtml	(revision 25153)
+++ doc/core/howto/telnet.xhtml	(working copy)
@@ -10,13 +10,10 @@
 gives you remote access to a Python interpreter. We will use a telnet client
 to access this server.</p>
 
-<p>Run <code class="shell">mktap telnet -p 4040 -u admin -w admin</code> at
-your shell prompt. If you list the contents of your current directory,
-you'll notice a new file -- <code>telnet.tap</code>.  After you do this, run
-<code class="shell">twistd -f telnet.tap</code>. Since the Application has a
-telnet server that you specified to be on port 4040, it will start listening
-for connections on this port. Try connecting with your favorite telnet
-utility to 127.0.0.1 port 4040.</p>
+<p>Run <code class="shell">twistd telnet -p 4040 -u admin -w admin</code> at
+your shell prompt. The Application has a telnet server that you specified to 
+be on port 4040, and it will start listening for connections on this port. Try 
+connecting with your favorite telnet utility to 127.0.0.1 port 4040.</p>
 
 <pre class="shell">
 $ <em>telnet localhost 4040</em>
@@ -59,18 +56,17 @@
 <code class="API" base="twisted.manhole.telnet">ShellFactory</code>. 
 Its parent is a <code class="python">twisted.application.service.MultiService</code>,
 a collection of services. We can keep getting the parent attribute
-of services until we hit the root of all services in this tap.</p>
+of services until we hit the root of all services.</p>
 
 <p>As you can see, this is quite useful - we can introspect a
 running process, see the internal objects, and even change
-their attributes. We can add telnet support to existing
-tap like so: <code class="shell">mktap --append=foo.tap telnet -p 4040 -u user -w pass</code>. 
-The telnet server can of coursed be used from straight Python code as well. You can
-see how to do this by reading the code for <code class="API">twisted.tap.telnet</code>.</p>
+their attributes. The telnet server can of course be used from straight 
+Python code; you can see how to do this by reading the code for 
+<code class="API">twisted.tap.telnet</code>.</p>
 
 <p>A final note - if you want access to be more secure, you can even
 have the telnet server use SSL. Assuming you have the appropriate
-certificate and private key files, you can <code class="shell">mktap
+certificate and private key files, you can <code class="shell">twistd
 telnet -p ssl:443:privateKey=mykey.pem:certKey=cert.pem -u admin -w
 admin</code>.  See <code
 class="API">twisted.application.strports</code> for more examples of
Index: doc/core/man/tapconvert.1
===================================================================
--- doc/core/man/tapconvert.1	(revision 24967)
+++ doc/core/man/tapconvert.1	(working copy)
@@ -38,5 +38,3 @@
 .br
 This is free software; see the source for copying conditions.  There is NO
 warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
-.SH "SEE ALSO"
-mktap(1)
Index: doc/core/man/twistd.1
===================================================================
--- doc/core/man/twistd.1	(revision 25153)
+++ doc/core/man/twistd.1	(working copy)
@@ -120,5 +120,3 @@
 .br
 This is free software; see the source for copying conditions.  There is NO
 warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
-.SH "SEE ALSO"
-mktap(1)
Index: doc/web/howto/web-development.xhtml
===================================================================
--- doc/web/howto/web-development.xhtml	(revision 24991)
+++ doc/web/howto/web-development.xhtml	(working copy)
@@ -32,8 +32,7 @@
 <p>into a directory, and then running:</p>
 
 <pre class="shell">
-% mktap web --path=/directory
-% twistd -f web.tap
+% twistd web --path=/directory
 </pre>
 
 <p>You can also write a Python script like:</p>
Index: twisted/conch/tap.py
===================================================================
--- twisted/conch/tap.py	(revision 24967)
+++ twisted/conch/tap.py	(working copy)
@@ -8,7 +8,7 @@
 
 
 """
-I am a support module for making SSH servers with mktap.
+I am a support module for making SSH servers with twistd.
 """
 
 from twisted.conch import checkers, unix
@@ -19,8 +19,8 @@
 
 
 class Options(usage.Options):
-    synopsis = "Usage: mktap conch [-i <interface>] [-p <port>] [-d <dir>] "
-    longdesc = "Makes a Conch SSH server.."
+    synopsis = "[-i <interface>] [-p <port>] [-d <dir>] "
+    longdesc = "Makes a Conch SSH server."
     optParameters = [
          ["interface", "i", "", "local interface to which we listen"],
          ["port", "p", "22", "Port on which to listen"],
Index: twisted/mail/tap.py
===================================================================
--- twisted/mail/tap.py	(revision 24967)
+++ twisted/mail/tap.py	(working copy)
@@ -3,7 +3,7 @@
 # See LICENSE for details.
 
 
-"""I am the support module for creating mail servers with 'mktap'
+"""I am the support module for creating mail servers with twistd
 """
 
 import os
@@ -22,7 +22,7 @@
 
 
 class Options(usage.Options):
-    synopsis = "Usage: mktap mail [options]"
+    synopsis = "[options]"
 
     optParameters = [
         ["pop3", "p", 8110, "Port to start the POP3 server on (0 to disable).", usage.portCoerce],
Index: twisted/news/tap.py
===================================================================
--- twisted/news/tap.py	(revision 24967)
+++ twisted/news/tap.py	(working copy)
@@ -72,7 +72,7 @@
 
 
 class Options(usage.Options):
-    synopsis = "Usage: mktap news [options]"
+    synopsis = "[options]"
     
     groups = None
     servers = None
Index: twisted/tap/ftp.py
===================================================================
--- twisted/tap/ftp.py	(revision 24967)
+++ twisted/tap/ftp.py	(working copy)
@@ -4,7 +4,7 @@
 
 
 """
-I am the support module for making a ftp server with mktap.
+I am the support module for making a ftp server with twistd.
 """
 
 from twisted.protocols import ftp
@@ -16,7 +16,7 @@
 
 
 class Options(usage.Options):
-    synopsis = """Usage: mktap ftp [options].
+    synopsis = """[options].
     WARNING: This FTP server is probably INSECURE do not use it.
     """
     optParameters = [
Index: twisted/tap/manhole.py
===================================================================
--- twisted/tap/manhole.py	(revision 24967)
+++ twisted/tap/manhole.py	(working copy)
@@ -4,7 +4,7 @@
 
 
 """
-I am the support module for making a manhole server with mktap.
+I am the support module for making a manhole server with twistd.
 """
 
 from twisted.manhole import service
@@ -15,7 +15,7 @@
 import os, sys
 
 class Options(usage.Options):
-    synopsis = "mktap manhole [options]"
+    synopsis = "[options]"
     optParameters = [
            ["user", "u", "admin", "Name of user to allow to log in"],
            ["port", "p", str(pb.portno), "Port to listen on"],
Index: twisted/tap/portforward.py
===================================================================
--- twisted/tap/portforward.py	(revision 24967)
+++ twisted/tap/portforward.py	(working copy)
@@ -3,14 +3,14 @@
 # See LICENSE for details.
 
 """
-Support module for making a port forwarder with mktap.
+Support module for making a port forwarder with twistd.
 """
 from twisted.protocols import portforward
 from twisted.python import usage
 from twisted.application import strports
 
 class Options(usage.Options):
-    synopsis = "Usage: mktap portforward [options]"
+    synopsis = "[options]"
     longdesc = 'Port Forwarder.'
     optParameters = [
           ["port", "p", "6666","Set the port number."],
Index: twisted/tap/socks.py
===================================================================
--- twisted/tap/socks.py	(revision 24967)
+++ twisted/tap/socks.py	(working copy)
@@ -4,7 +4,7 @@
 
 
 """
-I am a support module for making SOCKSv4 servers with mktap.
+I am a support module for making SOCKSv4 servers with twistd.
 """
 
 from twisted.protocols import socks
@@ -13,7 +13,7 @@
 import sys
 
 class Options(usage.Options):
-    synopsis = "Usage: mktap socks [-i <interface>] [-p <port>] [-l <file>]"
+    synopsis = "[-i <interface>] [-p <port>] [-l <file>]"
     optParameters = [["interface", "i", "127.0.0.1", "local interface to which we listen"],
                   ["port", "p", 1080, "Port on which to listen"],
                   ["log", "l", None, "file to log connection data to"]]
Index: twisted/tap/telnet.py
===================================================================
--- twisted/tap/telnet.py	(revision 24967)
+++ twisted/tap/telnet.py	(working copy)
@@ -4,7 +4,7 @@
 
 
 """
-Support module for making a telnet server with mktap.
+Support module for making a telnet server with twistd.
 """
 
 from twisted.manhole import telnet
@@ -12,7 +12,7 @@
 from twisted.application import strports
 
 class Options(usage.Options):
-    synopsis = "Usage: mktap telnet [options]"
+    synopsis = "[options]"
     longdesc = "Makes a telnet server to a Python shell."
     optParameters = [
          ["username", "u", "admin","set the login username"],
Index: twisted/web/tap.py
===================================================================
--- twisted/web/tap.py	(revision 24967)
+++ twisted/web/tap.py	(working copy)
@@ -16,7 +16,7 @@
 
 
 class Options(usage.Options):
-    synopsis = "Usage: mktap web [options]"
+    synopsis = "[options]"
     optParameters = [["port", "p", "8080","Port to start the server on."],
                      ["logfile", "l", None, "Path to web CLF (Combined Log Format) log file."],
                      ["https", None, None, "Port to listen on for Secure HTTP."],
@@ -35,9 +35,8 @@
 
 
     longdesc = """\
-This creates a web.tap file that can be used by twistd.  If you specify
-no arguments, it will be a demo webserver that has the Test class from
-twisted.web.demo in it."""
+This starts a webserver.  If you specify no arguments, it will be a 
+demo webserver that has the Test class from twisted.web.demo in it."""
 
     def __init__(self):
         usage.Options.__init__(self)
Index: twisted/words/toctap.py
===================================================================
--- twisted/words/toctap.py	(revision 24967)
+++ twisted/words/toctap.py	(working copy)
@@ -4,7 +4,7 @@
 
 
 """
-Support module for making TOC servers with mktap.
+Support module for making TOC servers with twistd.
 """
 
 from twisted.words.protocols import toc
@@ -12,7 +12,7 @@
 from twisted.application import strports
 
 class Options(usage.Options):
-    synopsis = "Usage: mktap toc [-p <port>]"
+    synopsis = "[-p <port>]"
     optParameters = [["port", "p", "5190"]]
     longdesc = "Makes a TOC server."
 
