Index: doc/core/howto/design.xhtml
===================================================================
--- doc/core/howto/design.xhtml	(revision 24967)
+++ doc/core/howto/design.xhtml	(working copy)
@@ -94,8 +94,8 @@
 <p> The <code class="python">QOTDFactory</code>'s role is to specify to the
 Twisted framework how to create a <code class="python">Protocol</code> instance
 that will handle the connection.  Twisted will not instantiate a <code
-class="python">QOTDFactory</code>; you will do that yourself later, in the
-<code class="shell">mktap</code> plug-in below.
+class="python">QOTDFactory</code>; you will do that yourself later, in a  
+<code class="shell">twistd</code> plug-in.
 </p>
 
 <p>Note: you can read more specifics of <code class="python">Protocol</code> and
