Index: twisted/conch/test/test_filetransfer.py
===================================================================
--- twisted/conch/test/test_filetransfer.py	(revision 27295)
+++ twisted/conch/test/test_filetransfer.py	(working copy)
@@ -1,5 +1,5 @@
 # -*- test-case-name: twisted.conch.test.test_filetransfer -*-
-# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2009 Twisted Matrix Laboratories.
 # See LICENSE file for details.
 
 
@@ -21,6 +21,7 @@
         # Hooray.
         pass
 
+from twisted.conch import error
 from twisted.conch import avatar
 from twisted.conch.ssh import common, connection, filetransfer, session
 from twisted.internet import defer
@@ -572,7 +573,28 @@
         self.assertSFTPConnectionLost()
 
 
+    def test_SSHChannelOpenError(self):
+        """
+        When L{SSHConnection.ssh_CHANNEL_OPEN}'s
+        C{transport.avatar.lookupChannel} raises a L{ConchError}, the exception
+        is logged and the packet describing the failure is sent without causing
+        an exception.
+        """
+        conn = self.buildServerConnection()
+        packet = common.NS('session') + struct.pack('>L', 0) * 3
+        # Force the avatar's lookupChannel to raise a ConchError
+        conn.transport.avatar.channelLookup = {}
 
+        conn.ssh_CHANNEL_OPEN(packet)
+
+        # Confirm that the error is coming from the avatar's lookupChannel
+        errs = self.flushLoggedErrors()
+        self.assertEquals(len(errs), 1)
+        self.assertEquals(errs[0].type, error.ConchError)
+        self.assertEquals(errs[0].value[1], connection.OPEN_UNKNOWN_CHANNEL_TYPE)
+
+
+
 class TestConstants(unittest.TestCase):
     """
     Tests for the constants used by the SFTP protocol implementation.
