Index: twisted/internet/process.py
===================================================================
--- twisted/internet/process.py	(revision 12668)
+++ twisted/internet/process.py	(working copy)
@@ -229,6 +229,33 @@
             mapping that opens the usual stdin/stdout/stderr pipes.
         """
 
+        # These are all checks done by the C implementation of
+        # os.execvpe in the posix module. Duplicate them here, because
+        # the forking makes it difficult to get exceptions later on.
+        assert isinstance(args, (tuple, list)), "args must be a tuple or list"
+        assert args, "args must not be empty"
+        # it is likely this is not entirely correct. C uses
+        # PyArg_Parse with an "et" format here, which wants a "string,
+        # Unicode object or character buffer compatible object". But
+        # what is a "character buffer compatible object"?
+        for arg in args:
+            assert isinstance(arg, (str, unicode)), \
+                "args must contain only strings"
+        
+        if environment is not None:
+            # C does a PyMapping_Check here. I'm not sure how to do
+            # this from python. So let's just hope that something we
+            # can do items() on is sufficiently dict-like.
+            for key, val in environment.items():
+                assert isinstance(key, (str, unicode)), \
+                    "env contains a non-string key"
+                assert isinstance(val, (str, unicode)), \
+                    "env contains a non-string value"
+        # perhaps we should also check whether command is executable?
+        # that would require some PATH-walking though (see _execvpe in
+        # os.py) if we don't want to complain about things execvpe
+        # handles.
+        
         if not proto:
             assert 'r' not in childFDs.values()
             assert 'w' not in childFDs.values()
