Index: twisted/protocols/ftp.py
===================================================================
--- twisted/protocols/ftp.py	(revision 35769)
+++ twisted/protocols/ftp.py	(working copy)
@@ -1179,7 +1179,7 @@
 
 
     def ftp_SIZE(self, path):
-        '''File SIZE
+        """File SIZE
 
         The FTP command, SIZE OF FILE (SIZE), is used to obtain the transfer
         size of a file from the server-FTP process.  This is the exact number
@@ -1202,7 +1202,7 @@
         precise information is acceptable.
 
         http://tools.ietf.org/html/rfc3659
-        '''
+        """
         try:
             newsegs = toSegments(self.workingDirectory, path)
         except InvalidPath:
@@ -1215,7 +1215,7 @@
 
 
     def ftp_MDTM(self, path):
-        '''File Modification Time (MDTM)
+        """File Modification Time (MDTM)
 
         The FTP command, MODIFICATION TIME (MDTM), can be used to determine
         when a file in the server NVFS was last modified.  This command has
@@ -1225,7 +1225,7 @@
         the new MLST command is recommended as a superior alternative.
 
         http://tools.ietf.org/html/rfc3659
-        '''
+        """
         try:
             newsegs = toSegments(self.workingDirectory, path)
         except InvalidPath:
@@ -1238,7 +1238,7 @@
 
 
     def ftp_TYPE(self, type):
-        '''REPRESENTATION TYPE (TYPE)
+        """REPRESENTATION TYPE (TYPE)
 
         The argument specifies the representation type as described
         in the Section on Data Representation and Storage.  Several
@@ -1248,7 +1248,7 @@
         for local byte is a decimal integer to indicate Bytesize.
         The parameters are separated by a <SP> (Space, ASCII code
         32).
-        '''
+        """
         p = type.upper()
         if p:
             f = getattr(self, 'type_' + p[0], None)
@@ -1342,20 +1342,20 @@
 
 
     def ftp_FEAT(self):
-        '''Advertise the features supported by the server.
+        """Advertise the features supported by the server.
 
         http://tools.ietf.org/html/rfc2389
-        '''
+        """
         self.sendLine(RESPONSE[FEAT_OK][0])
         for feature in self.FEATURES:
             self.sendLine(' ' + feature)
         self.sendLine(RESPONSE[FEAT_OK][1])
 
     def ftp_OPTS(self, option):
-        '''Handle OPTS command.
+        """Handle OPTS command.
 
         http://tools.ietf.org/html/draft-ietf-ftpext-utf-8-option-00
-        '''
+        """
         return self.reply(OPTS_NOT_IMPLEMENTED, option)
 
     def ftp_QUIT(self):
