--- abstract.py.orig	2009-12-19 04:06:49.000000000 -0800
+++ abstract.py	2009-12-19 04:09:06.000000000 -0800
@@ -207,6 +207,9 @@     def writeSequence(self, iovec):
         streaming producer is registered, it will be paused until the buffered
         data is written to the underlying file descriptor.
         """
+        for i in iovec:
+            if isinstance(i, unicode): # no, really, I mean it
+                raise TypeError("Data must not be unicode")
         if not self.connected or not iovec or self._writeDisconnected:
             return
         self._tempDataBuffer.extend(iovec)
