=== modified file 'twisted/trial/unittest.py'
--- twisted/trial/unittest.py	2009-05-16 14:04:45 +0000
+++ twisted/trial/unittest.py	2009-05-20 14:44:18 +0000
@@ -761,7 +761,8 @@
         return d
 
     def shortDescription(self):
-        desc = super(TestCase, self).shortDescription()
+        # Python 2.7 now returns additional information, so split and ignore it.
+        desc = super(TestCase, self).shortDescription().split(' ')[0]
         if desc is None:
             return self._testMethodName
         return desc

