Index: suppress-warnings-3639-2/twisted/test/test_sip.py
===================================================================
--- suppress-warnings-3639-2/twisted/test/test_sip.py	(revision 27382)
+++ suppress-warnings-3639-2/twisted/test/test_sip.py	(working copy)
@@ -487,6 +487,9 @@
         self.proxy.registry = self.proxy.locator = self.registry
         self.sent = []
         self.proxy.sendMessage = lambda dest, msg: self.sent.append((dest, msg))
+    setUp = utils.suppressWarnings(setUp, 
+        util.suppress(category=DeprecationWarning,
+            message=r'twisted.protocols.sip.DigestAuthorizer was deprecated'))
 
     def tearDown(self):
         for d, uri in self.registry.users.values():
@@ -671,6 +674,9 @@
         self.clientPort = reactor.listenUDP(0, self.client, interface="127.0.0.1")
         self.serverAddress = (self.serverPort.getHost().host,
                               self.serverPort.getHost().port)
+    setUp = utils.suppressWarnings(setUp, 
+        util.suppress(category=DeprecationWarning,
+            message=r'twisted.protocols.sip.DigestAuthorizer was deprecated'))
 
     def tearDown(self):
         for d, uri in self.registry.users.values():
