Index: suppress-warnings-3639/twisted/test/test_sip.py
===================================================================
--- suppress-warnings-3639/twisted/test/test_sip.py	(revision 27378)
+++ suppress-warnings-3639/twisted/test/test_sip.py	(working copy)
@@ -5,9 +5,9 @@
 
 """Session Initialization Protocol tests."""
 
-from twisted.trial import unittest
+from twisted.trial import unittest, util
 from twisted.protocols import sip
-from twisted.internet import defer, reactor
+from twisted.internet import defer, reactor, utils
 from twisted.python.versions import Version
 
 from twisted.test import proto_helpers
@@ -814,15 +814,6 @@
         self.proxy = sip.RegisterProxy(host="intarweb.us")
         self.proxy.authorizers = self.proxy.authorizers.copy()
         self.proxy.authorizers['digest'] = FakeDigestAuthorizer()
-        warnings = self.flushWarnings(
-            offendingFunctions=[self.setUp])
-        self.assertEqual(len(warnings), 1)
-        self.assertEqual(
-            warnings[0]['message'],
-            'twisted.protocols.sip.DigestAuthorizer was deprecated in Twisted 9.0.0')
-        self.assertEqual(
-            warnings[0]['category'],
-            DeprecationWarning)
 
         self.registry = FakeRegistry("intarweb.us")
         self.proxy.registry = self.proxy.locator = self.registry
@@ -835,6 +826,9 @@
         c.addUser('exarkun@intarweb.us', 'password')
         p.registerChecker(c)
         self.proxy.portal = p
+    setUp = utils.suppressWarnings(setUp, 
+        util.suppress(category=DeprecationWarning,
+            message=r'twisted.protocols.sip.DigestAuthorizer was deprecated'))
 
     def tearDown(self):
         for d, uri in self.registry.users.values():
@@ -851,25 +845,16 @@
         self.transport.written = []
 
         self.proxy.datagramReceived(authRequest, ("127.0.0.1", 5632))
-        warnings = self.flushWarnings(
-            offendingFunctions=[sip.DigestAuthorizer.decode,
-                                sip.DigestedCredentials.checkPassword,
-                                sip.RegisterProxy.__init__])
 
-        deprecated_classes = ['DigestAuthorizer', 'DigestedCredentials', 'DigestCalcHA1', 'DigestCalcResponse']
-        self.assertEqual(len(warnings), 4)
-        for w in warnings:
-            self.assertEqual(
-                w['message'],
-               'twisted.protocols.sip.%s was deprecated in Twisted 9.0.0' % deprecated_classes.pop(0))
-            self.assertEqual(
-                w['category'],
-                DeprecationWarning)
-
         self.assertEquals(
             self.transport.written[-1],
             ((okResponse, ("127.0.0.1", 5632)))
         )
+    testChallenge.suppress = [
+        util.suppress(category=DeprecationWarning, message=r'twisted.protocols.sip.DigestAuthorizer was deprecated'), 
+        util.suppress(category=DeprecationWarning, message=r'twisted.protocols.sip.DigestedCredentials was deprecated'), 
+        util.suppress(category=DeprecationWarning, message=r'twisted.protocols.sip.DigestCalcHA1 was deprecated'), 
+        util.suppress(category=DeprecationWarning, message=r'twisted.protocols.sip.DigestCalcResponse was deprecated')]
 
 
 
