Index: suppress-warnings-3639/twisted/test/test_sip.py
===================================================================
--- suppress-warnings-3639/twisted/test/test_sip.py	(revision 27378)
+++ suppress-warnings-3639/twisted/test/test_sip.py	(working copy)
@@ -857,7 +857,6 @@
                                 sip.RegisterProxy.__init__])
 
         deprecated_classes = ['DigestAuthorizer', 'DigestedCredentials', 'DigestCalcHA1', 'DigestCalcResponse']
-        self.assertEqual(len(warnings), 4)
         for w in warnings:
             self.assertEqual(
                 w['message'],
@@ -865,6 +864,7 @@
             self.assertEqual(
                 w['category'],
                 DeprecationWarning)
+        self.assertEqual(len(warnings), 4)
 
         self.assertEquals(
             self.transport.written[-1],
