Index: suppress-warnings-3639/twisted/test/test_sip.py
===================================================================
--- suppress-warnings-3639/twisted/test/test_sip.py	(revision 27376)
+++ suppress-warnings-3639/twisted/test/test_sip.py	(working copy)
@@ -816,6 +816,13 @@
         self.proxy.authorizers['digest'] = FakeDigestAuthorizer()
         warnings = self.flushWarnings(
             offendingFunctions=[self.setUp])
+        self.assertEqual(len(warnings), 1)
+        self.assertEqual(
+            warnings[0]['message'],
+            'twisted.protocols.sip.DigestAuthorizer was deprecated in Twisted 9.0.0')
+        self.assertEqual(
+            warnings[0]['category'],
+            DeprecationWarning)
 
         self.registry = FakeRegistry("intarweb.us")
         self.proxy.registry = self.proxy.locator = self.registry
@@ -847,6 +854,14 @@
         warnings = self.flushWarnings(
             offendingFunctions=[sip.DigestAuthorizer.decode,
                                 sip.DigestedCredentials.checkPassword])
+        self.assertEqual(len(warnings), 3)
+        self.assertEqual(
+            warnings[0]['message'],
+            'twisted.protocols.sip.DigestedCredentials was deprecated in Twisted 9.0.0')
+        self.assertEqual(
+            warnings[0]['category'],
+            DeprecationWarning)
+
         self.assertEquals(
             self.transport.written[-1],
             ((okResponse, ("127.0.0.1", 5632)))
