Index: suppress-warnings-3639/twisted/protocols/sip.py
===================================================================
--- suppress-warnings-3639/twisted/protocols/sip.py	(revision 27377)
+++ suppress-warnings-3639/twisted/protocols/sip.py	(working copy)
@@ -1144,13 +1144,13 @@
 
     registry = None # should implement IRegistry
 
-    authorizers = {
-        'digest': DigestAuthorizer(),
-    }
+    authorizers = {}
 
     def __init__(self, *args, **kw):
         Proxy.__init__(self, *args, **kw)
         self.liveChallenges = {}
+        if "digest" not in self.authorizers:
+            self.authorizers["digest"] = DigestAuthorizer()
         
     def handle_ACK_request(self, message, (host, port)):
         # XXX
Index: suppress-warnings-3639/twisted/test/test_sip.py
===================================================================
--- suppress-warnings-3639/twisted/test/test_sip.py	(revision 27377)
+++ suppress-warnings-3639/twisted/test/test_sip.py	(working copy)
@@ -853,15 +853,19 @@
         self.proxy.datagramReceived(authRequest, ("127.0.0.1", 5632))
         warnings = self.flushWarnings(
             offendingFunctions=[sip.DigestAuthorizer.decode,
-                                sip.DigestedCredentials.checkPassword])
-        self.assertEqual(len(warnings), 3)
-        self.assertEqual(
-            warnings[0]['message'],
-            'twisted.protocols.sip.DigestedCredentials was deprecated in Twisted 9.0.0')
-        self.assertEqual(
-            warnings[0]['category'],
-            DeprecationWarning)
+                                sip.DigestedCredentials.checkPassword,
+                                sip.RegisterProxy.__init__])
 
+        deprecated_classes = ['DigestAuthorizer', 'DigestedCredentials', 'DigestCalcHA1', 'DigestCalcResponse']
+        self.assertEqual(len(warnings), 4)
+        for w in warnings:
+            self.assertEqual(
+                w['message'],
+               'twisted.protocols.sip.%s was deprecated in Twisted 9.0.0' % deprecated_classes.pop(0))
+            self.assertEqual(
+                w['category'],
+                DeprecationWarning)
+
         self.assertEquals(
             self.transport.written[-1],
             ((okResponse, ("127.0.0.1", 5632)))
