diff --git twisted/test/test_sslverify.py twisted/test/test_sslverify.py
index 0178735..47ad270 100644
--- twisted/test/test_sslverify.py
+++ twisted/test/test_sslverify.py
@@ -31,7 +31,7 @@ from twisted.test.test_twisted import SetAsideModule
 from twisted.test.iosim import connectedServerAndClient
 
 from twisted.internet.error import ConnectionClosed
-from twisted.python.compat import nativeString
+from twisted.python.compat import nativeString, _PY3
 from twisted.python.constants import NamedConstant, Names
 from twisted.python.filepath import FilePath
 
@@ -2410,12 +2410,25 @@ class SelectVerifyImplementationTests(unittest.SynchronousTestCase):
 
             sslverify._selectVerifyImplementation(_postTwelveOpenSSL)
 
-            [warning] = list(
-                warning
-                for warning
-                in self.flushWarnings()
-                if warning["category"] == UserWarning)
+        [warning] = list(
+            warning
+            for warning
+            in self.flushWarnings()
+            if warning["category"] == UserWarning)
 
+        if _PY3:
+            expectedMessage = (
+                "You do not have a working installation of the "
+                "service_identity module: "
+                "'import of 'service_identity' halted; None in sys.modules'.  "
+                "Please install it from "
+                "<https://pypi.python.org/pypi/service_identity> "
+                "and make sure all of its dependencies are satisfied.  "
+                "Without the service_identity module and a recent enough "
+                "pyOpenSSL to support it, Twisted can perform only "
+                "rudimentary TLS client hostname verification.  Many valid "
+                "certificate/hostname mappings may be rejected.")
+        else:
             expectedMessage = (
                 "You do not have a working installation of the "
                 "service_identity module: "
@@ -2428,8 +2441,8 @@ class SelectVerifyImplementationTests(unittest.SynchronousTestCase):
                 "rudimentary TLS client hostname verification.  Many valid "
                 "certificate/hostname mappings may be rejected.")
 
-            self.assertEqual(
-                (warning["message"], warning["filename"], warning["lineno"]),
-                # See the comment in test_pyOpenSSLTooOldWarning.
-                (expectedMessage, "", 0))
+        self.assertEqual(
+            (warning["message"], warning["filename"], warning["lineno"]),
+            # See the comment in test_pyOpenSSLTooOldWarning.
+            (expectedMessage, "", 0))
 
diff --git twisted/topfiles/7355.misc twisted/topfiles/7355.misc
new file mode 100644
index 0000000..e69de29
