Index: twisted/internet/defer.py
===================================================================
--- twisted/internet/defer.py	(revision 30920)
+++ twisted/internet/defer.py	(working copy)
@@ -17,6 +17,7 @@
 """
 
 import traceback
+import types
 import warnings
 from sys import exc_info
 
@@ -1138,7 +1139,10 @@
                 raise Exception('DESTROY ALL LIFE')
     """
     def unwindGenerator(*args, **kwargs):
-        return _inlineCallbacks(None, f(*args, **kwargs), Deferred())
+        gen = f(*args, **kwargs)
+        if not isinstance(gen, types.GeneratorType):
+            raise TypeError("inlineCallbacks requires a generator")
+        return _inlineCallbacks(None, gen, Deferred())
     return mergeFunctionMetadata(f, unwindGenerator)
 
 
