Index: twisted/internet/defer.py
===================================================================
--- twisted/internet/defer.py	(revision 24259)
+++ twisted/internet/defer.py	(working copy)
@@ -782,8 +782,9 @@
 
     WARNING: this function will not work in Python 2.4 and earlier!
 
-    inlineCallbacks helps you write Deferred-using code that looks like a
-    regular sequential function. This function uses features of Python 2.5
+    inlineCallbacks is a decorator for functions which return generators
+    which helps you write Deferred-using code that looks like a regular
+    sequential function.  This function uses features of Python 2.5
     generators.  If you need to be compatible with Python 2.4 or before, use
     the L{deferredGenerator} function instead, which accomplishes the same
     thing, but with somewhat more boilerplate.  For example::
