Index: twisted/test/test_defer.py
===================================================================
--- twisted/test/test_defer.py	(revision 22074)
+++ twisted/test/test_defer.py	(working copy)
@@ -567,6 +567,29 @@
 
 
 
+class CircularDebugInfoTestCase(unittest.TestCase):
+    def setUp(self):
+        self._deferredWasDebugging = defer.getDebugging()
+        defer.setDebugging(True)
+
+    def tearDown(self):
+        defer.setDebugging(self._deferredWasDebugging)
+
+    def testCircularFailureReference(self):
+        """
+        When debug is enabled, Deferred should not create gc.garbage,
+        even if circular references between the Deferred and its
+        contained failure is used.
+        """
+        d = defer.Deferred()
+        f = failure.Failure(Exception(d))
+        d.errback(f)
+        self.assertEqual(gc.garbage, [])
+        del d, f
+        gc.collect()
+        self.assertEqual(gc.garbage, [])
+
+
 class LogTestCase(unittest.TestCase):
     """
     Test logging of unhandled errors.
