Index: doc/core/examples/courier.py
===================================================================
--- doc/core/examples/courier.py	(revision 25782)
+++ doc/core/examples/courier.py	(working copy)
@@ -55,7 +55,7 @@
     delimiter = '\n'
     
     def connectionMade(self):
-    log.msg('Connection from %r' % self.transport)
+        log.msg('Connection from %r' % self.transport)
         self.state = 'connected'
         self.metaInfo = []
 
