Index: twisted/web/test/test_webclient.py
===================================================================
--- twisted/web/test/test_webclient.py	(revision 30711)
+++ twisted/web/test/test_webclient.py	(working copy)
@@ -612,6 +612,24 @@
         return d
 
 
+    def test_downloadAfterFoundGet(self):
+        """
+        Test to make sure C{client.downloadPage} works with
+        C{afterFoundGet} works"
+        """
+        url = self.getURL('extendedRedirect?code=302')
+
+        def gotPage(page):
+            self.assertEquals(
+                self.extendedRedirect.lastMethod,
+                "GET",
+                "With afterFoundGet, the HTTP method must change to GET")
+
+        d = client.downloadPage(url, "downloadTemp",
+            followRedirect=True, afterFoundGet=True, method="POST")
+        d.addCallback(gotPage)
+        return d
+
     def test_afterFoundGetMakesOneRequest(self):
         """
         When C{afterFoundGet} is C{True}, L{client.getPage} only issues one
Index: twisted/web/client.py
===================================================================
--- twisted/web/client.py	(revision 30711)
+++ twisted/web/client.py	(working copy)
@@ -361,7 +361,7 @@
                  method='GET', postdata=None, headers=None,
                  agent="Twisted client", supportPartial=0,
                  timeout=0, cookies=None, followRedirect=1,
-                 redirectLimit=20):
+                 redirectLimit=20, afterFoundGet=False):
         self.requestedPartial = 0
         if isinstance(fileOrName, types.StringTypes):
             self.fileName = fileOrName
@@ -378,7 +378,8 @@
         HTTPClientFactory.__init__(
             self, url, method=method, postdata=postdata, headers=headers,
             agent=agent, timeout=timeout, cookies=cookies,
-            followRedirect=followRedirect, redirectLimit=redirectLimit)
+            followRedirect=followRedirect, redirectLimit=redirectLimit,
+            afterFoundGet=afterFoundGet)
 
 
     def gotHeaders(self, headers):
