diff --git twisted/internet/interfaces.py twisted/internet/interfaces.py
index fad315b..74256f5 100644
--- twisted/internet/interfaces.py
+++ twisted/internet/interfaces.py
@@ -655,7 +655,7 @@ class IReactorTCP(Interface):
 
         @param host: A hostname or an IPv4 or IPv6 address literal.
 
-        @type host: L{bytes}
+        @type host: L{unicode}
 
         @param port: a port number
 
diff --git twisted/internet/topfiles/7956.doc twisted/internet/topfiles/7956.doc
new file mode 100644
index 0000000..6dfe87b
--- /dev/null
+++ twisted/internet/topfiles/7956.doc
@@ -0,0 +1,2 @@
+The host parameter of IReactorTCP.connectTCP in twisted.internet.interfaces is
+now documented as unicode instead of bytes.
