Index: twisted/trial/reporter.py
===================================================================
--- twisted/trial/reporter.py	(revision 27686)
+++ twisted/trial/reporter.py	(working copy)
@@ -926,31 +926,27 @@
 
 
     def startTest(self, test):
-        return self._subunit.startTest(test)
+        pass
 
 
     def stopTest(self, test):
-        return self._subunit.stopTest(test)
+        pass
 
 
     def addSuccess(self, test):
-        return self._subunit.addSuccess(test)
+        pass
 
 
     def addSkip(self, test, reason):
-        return self._subunit.addSkip(test, reason)
+        pass
 
 
     def addError(self, test, err):
         self._successful = False
-        return self._subunit.addError(
-            test, util.excInfoOrFailureToExcInfo(err))
 
 
     def addFailure(self, test, err):
-        self._successful = False
-        return self._subunit.addFailure(
-            test, util.excInfoOrFailureToExcInfo(err))
+        pass
 
 
     def addExpectedFailure(self, test, failure, todo):
@@ -960,14 +956,14 @@
         failure = util.excInfoOrFailureToExcInfo(failure)
         addExpectedFailure = getattr(self._subunit, 'addExpectedFailure', None)
         if addExpectedFailure is None:
-            self.addSuccess(test)
+            pass
         else:
             addExpectedFailure(test, failure)
 
 
     def addUnexpectedSuccess(self, test, todo):
         # Not represented in pyunit/subunit.
-        self.addSuccess(test)
+        pass
 
 
 
