=== modified file 'twisted/plugins/twisted_trial.py'
--- twisted/plugins/twisted_trial.py	2006-03-12 19:59:12 +0000
+++ twisted/plugins/twisted_trial.py	2009-09-06 09:26:30 +0000
@@ -51,3 +51,9 @@
                    shortOpt=None,
                    klass="TimingTextReporter")
 
+Subunit = _Reporter("Subunit Reporter",
+                    "twisted.trial.reporter",
+                    description="subunit output",
+                    longOpt="subunit",
+                    shortOpt=None,
+                    klass="SubunitReporter")

=== modified file 'twisted/trial/reporter.py'
--- twisted/trial/reporter.py	2008-11-02 13:59:21 +0000
+++ twisted/trial/reporter.py	2009-09-06 09:40:39 +0000
@@ -20,6 +20,10 @@
 from twisted.python.util import untilConcludes
 from twisted.trial import itrial, util
 
+try:
+    from subunit import TestProtocolClient
+except ImportError:
+    TestProtocolClient = None
 from zope.interface import implements
 
 pyunit = __import__('unittest')
@@ -885,6 +889,18 @@
         self.stream.write(text)
 
 
+if TestProtocolClient is not None:
+    class SubunitReporter(TestProtocolClient):
+        """Reports test output via Subunit."""
+        implements(itrial.IReporter)
+
+        def __init__(self, stream=sys.stdout, tbformat='default',
+            realtime=False, publisher=None):
+            super(SubunitReporter, self).__init__(stream)
+else:
+    def __inform_missing():raise Exception("Subunit not available.")
+    SubunitReporter = lambda:__inform_missing()
+
 
 class TreeReporter(Reporter):
     """

