Index: b/twisted/python/dist3.py
===================================================================
--- a/twisted/python/dist3.py
+++ b/twisted/python/dist3.py
@@ -49,6 +49,7 @@ modules = [
     "twisted.conch",
     "twisted.conch._version",
     "twisted.conch.ssh",
+    "twisted.conch.ssh._cryptography_backports",
     "twisted.conch.ssh.common",
     "twisted.conch.ssh.keys",
     "twisted.conch.ssh.sexpy",
