Index: twisted/vfs/adapters/sftp.py
===================================================================
--- twisted/vfs/adapters/sftp.py	(revision 20355)
+++ twisted/vfs/adapters/sftp.py	(revision 20363)
@@ -133,7 +133,7 @@
         """
         Rename C{oldpath} to C{newpath}.
 
-        See L{twisted.conch.interfaces.ISFTPServer.renameFile.
+        See L{twisted.conch.interfaces.ISFTPServer.renameFile}.
         """
         old = self.filesystem.fetch(oldpath)
         old.rename(newpath)
Index: twisted/vfs/backends/osfs.py
===================================================================
--- twisted/vfs/backends/osfs.py	(revision 20355)
+++ twisted/vfs/backends/osfs.py	(revision 20363)
@@ -71,7 +71,7 @@
         try:
             os.rename(self.realPath, newPath)
         except OSError, e:
-            if e.errno in (errno.EISDIR, errno.ENOTEMPTY):
+            if e.errno in (errno.EISDIR, errno.ENOTEMPTY, errno.EEXIST):
                 raise ivfs.AlreadyExistsError(
                     "Can't rename %s to %s: %s already exists"
                     % (self.realPath, newPath, newPath))
