Index: twisted/conch/test/test_userauth.py
===================================================================
--- twisted/conch/test/test_userauth.py	(revision 34751)
+++ twisted/conch/test/test_userauth.py	(working copy)
@@ -37,7 +37,7 @@
     class userauth:
         class SSHUserAuthClient:
             """
-            A stub class so that leter class definitions won't die.
+            A stub class so that later class definitions won't die.
             """
 else:
     from twisted.conch.ssh.common import NS
@@ -1064,6 +1064,8 @@
 
 
 class ModuleInitializationTestCase(unittest.TestCase):
+    if keys is None:
+        skip = "cannot run w/o PyCrypto or PyASN1"
 
 
     def test_messages(self):
