Index: internet/protocol.py
===================================================================
--- internet/protocol.py	(revision 24208)
+++ internet/protocol.py	(working copy)
@@ -215,7 +215,7 @@
     factor = 2.7182818284590451 # (math.e)
     # Phi = 1.6180339887498948 # (Phi is acceptable for use as a
     # factor if e is too large for your application.)
-    jitter = 0.11962656492 # molar Planck constant times c, Jule meter/mole
+    jitter = 0.11962656572 # molar Planck constant times c, joule meter/mole
 
     delay = initialDelay
     retries = 0
