Index: sux.py
===================================================================
RCS file: /cvs/Twisted/twisted/protocols/sux.py,v
retrieving revision 1.36
diff -u -3 -p -r1.36 sux.py
--- sux.py	31 Oct 2003 01:56:21 -0000	1.36
+++ sux.py	5 Mar 2004 05:25:15 -0000
@@ -254,7 +254,20 @@ class XMLParser(Protocol):
             else:
                 self._parseError("Invalid '[' in tag-name")
         else:
+	    if self.beExtremelyLenient:
+		self.bodydata = '<'
+		return 'unentity'
             self._parseError('Invalid tag character: %r'% byte)
+
+    def begin_unentity(self, byte):
+	self.bodydata += byte
+
+    def do_unentity(self, byte):
+	self.bodydata += byte
+	return 'bodydata'
+
+    def end_unentity(self):
+	self.gotText(self.bodydata)
 
     def begin_expectcdata(self, byte):
         self.cdatabuf = byte
