Index: test_xml.py
===================================================================
RCS file: /cvs/Twisted/twisted/test/test_xml.py,v
retrieving revision 1.32
diff -u -3 -p -r1.32 test_xml.py
--- test_xml.py	29 Oct 2003 06:56:59 -0000	1.32
+++ test_xml.py	5 Mar 2004 05:27:03 -0000
@@ -54,6 +54,16 @@ class SUXTest(TestCase):
 
 class MicroDOMTest(TestCase):
 
+    def testUnEntities(self):
+	s = """
+		<HTML>
+		    This HTML goes between Stupid <=CrAzY!=> Dumb. 
+		</HTML>
+	    """
+	d = microdom.parseString(s, beExtremelyLenient=1)
+	n = domhelpers.gatherTextNodes(d)
+	self.assertNotEquals(n.find('>'), -1)
+
     def testEmptyError(self):
         self.assertRaises(sux.ParseError, microdom.parseString, "")
 
