diff --git twisted/words/im/basesupport.py twisted/words/im/basesupport.py
index 5c8b424..43b4c1a 100644
--- twisted/words/im/basesupport.py
+++ twisted/words/im/basesupport.py
@@ -10,8 +10,7 @@ You will find these useful if you're adding a new protocol to IM.
 
 # Abstract representation of chat "model" classes
 
-from twisted.words.im.locals import ONLINE, OFFLINE, OfflineError
-from twisted.words.im import interfaces
+from twisted.words.im.locals import OFFLINE, OfflineError
 
 from twisted.internet.protocol import Protocol
 
diff --git twisted/words/im/interfaces.py twisted/words/im/interfaces.py
index 8f34fb1..b1d48cd 100644
--- twisted/words/im/interfaces.py
+++ twisted/words/im/interfaces.py
@@ -8,8 +8,6 @@ Pan-protocol chat client.
 
 from zope.interface import Interface, Attribute
 
-from twisted.words.im import locals
-
 # (Random musings, may not reflect on current state of code:)
 #
 # Accounts have Protocol components (clients)
diff --git twisted/words/iwords.py twisted/words/iwords.py
index c8ce09f..c4728d0 100644
--- twisted/words/iwords.py
+++ twisted/words/iwords.py
@@ -2,7 +2,8 @@
 # Copyright (c) Twisted Matrix Laboratories.
 # See LICENSE for details.
 
-from zope.interface import Interface, Attribute, implements
+from zope.interface import Interface, Attribute
+
 
 class IProtocolPlugin(Interface):
     """Interface for plugins providing an interface to a Words service
@@ -262,5 +263,5 @@ class IChatService(Interface):
         """
 
 __all__ = [
-    'IChatInterface', 'IGroup', 'IChatClient', 'IUser', 'IChatService',
+    'IGroup', 'IChatClient', 'IUser', 'IChatService',
     ]
diff --git twisted/words/protocols/irc.py twisted/words/protocols/irc.py
index 6a35ae6..8fbe46f 100644
--- twisted/words/protocols/irc.py
+++ twisted/words/protocols/irc.py
@@ -1760,6 +1760,7 @@ class IRCClient(basic.LineReceiver):
             "XXX!!! Help!  I need to bind a socket, have it listen, and tell me its address.  "
             "(and stop accepting once we've made a single connection.)")
 
+        my_address = None
         my_address = struct.pack("!I", my_address)
 
         args = ['SEND', name, my_address, str(port)]
diff --git twisted/words/protocols/jabber/client.py twisted/words/protocols/jabber/client.py
index 2a37bcb..93d0851 100644
--- twisted/words/protocols/jabber/client.py
+++ twisted/words/protocols/jabber/client.py
@@ -275,7 +275,7 @@ class SessionInitializer(xmlstream.BaseFeatureInitiatingInitializer):
 
     def start(self):
         iq = xmlstream.IQ(self.xmlstream, 'set')
-        session = iq.addElement((NS_XMPP_SESSION, 'session'))
+        iq.addElement((NS_XMPP_SESSION, 'session'))
         return iq.send()
 
 
diff --git twisted/words/protocols/oscar.py twisted/words/protocols/oscar.py
index 14908c6..f9ee9bc 100644
--- twisted/words/protocols/oscar.py
+++ twisted/words/protocols/oscar.py
@@ -1092,12 +1092,11 @@ class ChatService(OSCARService):
                 self.members.remove(u)
         self.bos.chatMemberLeft(self,user)
 
-    def oscar_0E_06(self,snac):
-        data = snac[3]
-        user,rest=self.bos.parseUser(snac[3][14:],1)
+    def oscar_0E_06(self, snac):
+        user, rest = self.bos.parseUser(snac[3][14:], 1)
         tlvs = readTLVs(rest[8:])
-        message=tlvs[1]
-        self.bos.chatReceiveMessage(self,user,message)
+        message = tlvs[1]
+        self.bos.chatReceiveMessage(self, user, message)
 
     def sendMessage(self,message):
         tlvs=TLV(0x02,"us-ascii")+TLV(0x03,"en")+TLV(0x01,message)
diff --git twisted/words/test/test_domish.py twisted/words/test/test_domish.py
index 275afb7..eb6633e 100644
--- twisted/words/test/test_domish.py
+++ twisted/words/test/test_domish.py
@@ -5,6 +5,7 @@
 Tests for L{twisted.words.xish.domish}, a DOM-like library for XMPP.
 """
 
+from twisted.python.reflect import requireModule
 from twisted.trial import unittest
 from twisted.words.xish import domish
 
@@ -256,10 +257,10 @@ class DomishExpatStreamTestCase(DomishStreamTestsMixin, unittest.TestCase):
     """
     streamClass = domish.ExpatElementStream
 
-    try:
-        import pyexpat
-    except ImportError:
+    if requireModule('pyexpat', default=None) is None:
         skip = "pyexpat is required for ExpatElementStream tests."
+    else:
+        skip = None
 
 
 
diff --git twisted/words/test/test_jabberxmlstream.py twisted/words/test/test_jabberxmlstream.py
index caa2ba6..93d4665 100644
--- twisted/words/test/test_jabberxmlstream.py
+++ twisted/words/test/test_jabberxmlstream.py
@@ -1141,7 +1141,6 @@ class StreamManagerTest(unittest.TestCase):
         sm = self.streamManager
         handler = FailureReasonXMPPHandler()
         handler.setHandlerParent(sm)
-        xs = xmlstream.XmlStream(xmlstream.Authenticator())
         sm._disconnected(failure.Failure(Exception("no reason")))
         self.assertEqual(True, handler.gotFailureReason)
 
diff --git twisted/words/test/test_msn.py twisted/words/test/test_msn.py
index 669c9aa..bc4ec66 100644
--- twisted/words/test/test_msn.py
+++ twisted/words/test/test_msn.py
@@ -5,18 +5,21 @@
 Test cases for L{twisted.words.protocols.msn}.
 """
 
-# System imports
 import StringIO
 from hashlib import md5
 
-# Twisted imports
+from twisted.internet.defer import Deferred
+from twisted.protocols import loopback
+from twisted.python.reflect import requireModule
+from twisted.test.proto_helpers import StringTransport, StringIOWithoutClosing
+from twisted.trial import unittest
 
 # t.w.p.msn requires an HTTP client
 try:
     # So try to get one - do it directly instead of catching an ImportError
     # from t.w.p.msn so that other problems which cause that module to fail
     # to import don't cause the tests to be skipped.
-    from twisted.web import client
+    requireModule('twisted.web.client')
 except ImportError:
     # If there isn't one, we're going to skip all the tests.
     msn = None
@@ -24,12 +27,6 @@ else:
     # Otherwise importing it should work, so do it.
     from twisted.words.protocols import msn
 
-
-from twisted.protocols import loopback
-from twisted.internet.defer import Deferred
-from twisted.trial import unittest
-from twisted.test.proto_helpers import StringTransport, StringIOWithoutClosing
-
 def printError(f):
     print f
 
diff --git twisted/words/test/test_xpath.py twisted/words/test/test_xpath.py
index 9dbda0f..ed37af8 100644
--- twisted/words/test/test_xpath.py
+++ twisted/words/test/test_xpath.py
@@ -3,7 +3,6 @@
 
 
 from twisted.trial import unittest
-import sys, os
 
 from twisted.words.xish.domish import Element
 from twisted.words.xish.xpath import XPathQuery
diff --git twisted/words/topfiles/6994.misc twisted/words/topfiles/6994.misc
new file mode 100644
index 0000000..e69de29
