Index: doc/web/howto/client.xhtml
===================================================================
--- doc/web/howto/client.xhtml	(revision 38232)
+++ doc/web/howto/client.xhtml	(working copy)
@@ -187,10 +187,10 @@
       The <code>Response</code> object has several attributes giving the
       response information: its code, version, phrase, and headers, as well as
       the length of the body to expect.  The <code>Response</code> object also
-      has a method which makes the response body available: <code class="API"
-      base="twisted.web.client.Response">deliverBody</code>.  Using the
-      attributes of the response object and this method, here's an example which
-      displays part of the response to a request:
+      has a method which makes the response body available: 
+      <a href="http://twistedmatrix.com/documents/current/api/twisted.web.client.Response.html#deliverBody">
+      deliverBody</a>. Using the attributes of the response object and this
+      method, here's an example which displays part of the response to a request:
     </p>
 
     <a href="listings/client/response.py" class="py-listing">
