From c294d9c94944fe409e8e6bc64d54dba269a226d3 Mon Sep 17 00:00:00 2001
From: Dave Peticolas <dave@krondo.com>
Date: Sat, 7 Jun 2008 17:49:25 -0700
Subject: [PATCH] Fix example.

---
 doc/core/examples/ampserver.py |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/doc/core/examples/ampserver.py b/doc/core/examples/ampserver.py
index 393cfec..7b5adf0 100644
--- a/doc/core/examples/ampserver.py
+++ b/doc/core/examples/ampserver.py
@@ -21,8 +21,8 @@ class Math(amp.AMP):
     Sum.responder(sum)
 
     def divide(self, numerator, denominator):
-        result = numerator / denominator
-        print 'Divided: %d / %d = %d' % (numerator, denominator, total)
+        result = float(numerator) / denominator
+        print 'Divided: %d / %d = %f' % (numerator, denominator, result)
         return {'result': result}
     Divide.responder(divide)
 
-- 
1.5.5.3

