Index: twisted/names/cache.py
===================================================================
--- twisted/names/cache.py	(revision 34432)
+++ twisted/names/cache.py	(working copy)
@@ -95,7 +95,10 @@
         if self.verbose > 1:
             log.msg('Adding %r to cache' % query)
 
-        self.cache[query] = (cacheTime or self._reactor.seconds(), payload)
+        if cacheTime is None:
+            self.cache[query] = (self._reactor.seconds(), payload)
+        else:
+            self.cache[query] = (cacheTime, payload)
 
         if self.cancel.has_key(query):
             self.cancel[query].cancel()
