=== modified file 'twisted/lore/test/test_lore.py'
--- twisted/lore/test/test_lore.py	2011-02-14 04:45:15 +0000
+++ twisted/lore/test/test_lore.py	2011-06-01 02:28:32 +0000
@@ -1197,32 +1197,3 @@
         """
         processor = lore.getProcessor("lore", "html", options)
         self.assertNotIdentical(processor, None)
-
-
-
-class DeprecationTests(unittest.TestCase):
-    """
-    Tests for deprecated APIs in L{twisted.lore.tree}.
-    """
-    def test_comparePosition(self):
-        """
-        L{tree.comparePosition} is deprecated.
-        """
-        from twisted.web.microdom import parseString
-        element = parseString('<foo/>').documentElement
-        self.assertEqual(
-            self.callDeprecated(
-                Version('Twisted', 9, 0, 0),
-                tree.comparePosition, element, element),
-            0)
-
-
-    def test_compareMarkPos(self):
-        """
-        L{tree.compareMarkPos} is deprecated.
-        """
-        self.assertEqual(
-            self.callDeprecated(
-                Version('Twisted', 9, 0, 0),
-                tree.compareMarkPos, [0, 1], [1, 2]),
-            -1)

=== modified file 'twisted/lore/tree.py'
--- twisted/lore/tree.py	2011-04-06 02:20:28 +0000
+++ twisted/lore/tree.py	2011-06-01 02:27:53 +0000
@@ -12,7 +12,6 @@
 
 from twisted.python import htmlizer, text
 from twisted.python.filepath import FilePath
-from twisted.python.deprecate import deprecated
 from twisted.python.versions import Version
 from twisted.web import domhelpers
 import process, latex, indexer, numberer, htmlbook
@@ -471,34 +470,6 @@
 
 
 
-def compareMarkPos(a, b):
-    """
-    Perform in every way identically to L{cmp} for valid inputs.
-    """
-    linecmp = cmp(a[0], b[0])
-    if linecmp:
-        return linecmp
-    return cmp(a[1], b[1])
-compareMarkPos = deprecated(Version('Twisted', 9, 0, 0))(compareMarkPos)
-
-
-
-def comparePosition(firstElement, secondElement):
-    """
-    Compare the two elements given by their position in the document or
-    documents they were parsed from.
-
-    @type firstElement: C{dom.Element}
-    @type secondElement: C{dom.Element}
-
-    @return: C{-1}, C{0}, or C{1}, with the same meanings as the return value
-    of L{cmp}.
-    """
-    return cmp(firstElement._markpos, secondElement._markpos)
-comparePosition = deprecated(Version('Twisted', 9, 0, 0))(comparePosition)
-
-
-
 def findNodeJustBefore(target, nodes):
     """
     Find the last Element which is a sibling of C{target} and is in C{nodes}.

