Index: twisted/runner/test/test_procmontap.py
===================================================================
--- twisted/runner/test/test_procmontap.py	(revision 43542)
+++ twisted/runner/test/test_procmontap.py	(working copy)
@@ -11,7 +11,7 @@
 from twisted.runner import procmontap as tap
 
 
-class ProcessMonitorTapTest(unittest.TestCase):
+class ProcessMonitorTapTests(unittest.TestCase):
     """
     Tests for L{twisted.runner.procmontap}'s option parsing and makeService
     method.
Index: twisted/runner/topfiles/7726.misc
===================================================================
