Index: interfaces.py
===================================================================
--- interfaces.py	(revision 25612)
+++ interfaces.py	(working copy)
@@ -944,7 +944,11 @@
                        failure may be of other classes as well.
         """
 
+
 class IReadDescriptor(IFileDescriptor):
+    """
+    A L{FileDescriptor<twisted.internet.abstract.FileDescriptor> that can read.
+    """
 
     def doRead():
         """
@@ -953,6 +957,10 @@
 
 
 class IWriteDescriptor(IFileDescriptor):
+    """
+    A L{FileDescriptor<twisted.internet.abstract.FileDescriptor> that can
+    write.
+    """
 
     def doWrite():
         """
@@ -962,8 +970,8 @@
 
 class IReadWriteDescriptor(IReadDescriptor, IWriteDescriptor):
     """
-    I am a L{FileDescriptor<twisted.internet.abstract.FileDescriptor>} that
-    can both read and write.
+    A L{FileDescriptor<twisted.internet.abstract.FileDescriptor>} that can both
+    read and write.
     """
 
 
