Index: doc/core/examples/pbgtk2.py
===================================================================
--- doc/core/examples/pbgtk2.py	(revision 27376)
+++ doc/core/examples/pbgtk2.py	(working copy)
@@ -104,7 +104,7 @@
         w.add(vb)
         map(vb.add, [b, self.entry, self.outry])
         b.connect('clicked', self.clicked)
-        w.connect('destroy', reactor.stop)
+        w.connect('destroy', self.stop)
         w.show_all()
 
     def clicked(self, b):
@@ -112,6 +112,9 @@
         self.entry.set_text("")
         self.echoer.callRemote('echo',txt).addCallback(self.outry.set_text)
 
+    def stop(self, b):
+        reactor.stop()
+
 d = defer.Deferred()
 LoginDialog(d)
 d.addCallbacks(EchoClient,
