Index: twisted/internet/_newtls.py
===================================================================
--- twisted/internet/_newtls.py (revision 40726)
+++ twisted/internet/_newtls.py (working copy)
@@ -142,6 +142,7 @@

     transport.getHandle = tlsProtocol.getHandle
     transport.getPeerCertificate = tlsProtocol.getPeerCertificate
+    transport.whenHandshakeDone = tlsProtocol.whenHandshakeDone

     # Mark the transport as secure.
     directlyProvides(transport, ISSLTransport)
