Index: test-standard.xhtml
===================================================================
--- test-standard.xhtml	(revision 28396)
+++ test-standard.xhtml	(working copy)
@@ -312,12 +312,12 @@
       Python</a>.</li>
 
       <li><a href=
-      "http://docs.python.org/lib/module-unittest.html"
+      "http://docs.python.org/library/unittest.html"        
       ><code>unittest</code></a> module documentation, in the <a href=
       "http://docs.python.org/library">Python Library
       Reference</a>.</li>
 
-      <li><a href="http://c2.com/cgi/wiki?UnitTests">UnitTests</a> on
+      <li><a href="http://c2.com/cgi/wiki?UnitTest">UnitTest</a> on
       the <a href="http://c2.com/cgi/wiki">PortlandPatternRepository
       Wiki</a>, where all the cool <a href=
       "http://c2.com/cgi/wiki?ExtremeProgramming"
@@ -340,9 +340,9 @@
 
       <li><a href="http://pyunit.sourceforge.net">PyUnit's homepage</a>.</li>
 
-      <li>The <a href=
-      "http://svn.twistedmatrix.com/cvs/trunk/twisted/?root=Twisted"
-      >twisted/test directory</a> in Subversion.</li>
+      <li>The top-level tests directory, <a href=
+      "http://twistedmatrix.com/trac/browser/trunk/twisted/test"
+      >twisted/test</a>, in Subversion.</li>
 
     </ul>
 

