Index: doc/core/howto/endpoints.xhtml
===================================================================
--- doc/core/howto/endpoints.xhtml	(revision 33285)
+++ doc/core/howto/endpoints.xhtml	(working copy)
@@ -128,7 +128,7 @@
 implemented like this:
 <pre class="python">
 attempt = myEndpoint.connect(myFactory)
-reactor.callback(30, attempt.cancel)
+reactor.callLater(30, attempt.cancel)
 </pre>
 </p>
 
