Index: twisted/web/tap.py
===================================================================
--- twisted/web/tap.py	(revision 20668)
+++ twisted/web/tap.py	(working copy)
@@ -172,7 +172,7 @@
                  = pwd.getpwuid(os.getuid())
         i = internet.UNIXServer(os.path.join(pw_dir,
                                    distrib.UserDirectory.userSocketName),
-                      pb.BrokerFactory(distrib.ResourcePublisher(site)))
+                      pb.PBServerFactory(distrib.ResourcePublisher(site)))
         i.setServiceParent(s)
     else:
         if config['https']:
