diff --git twisted/web/topfiles/5509.misc twisted/web/topfiles/5509.misc
new file mode 100644
index 0000000..e69de29
diff --git twisted/web/util.py twisted/web/util.py
index ce87f72..efc9269 100644
--- twisted/web/util.py
+++ twisted/web/util.py
@@ -22,7 +22,7 @@ from twisted.python.versions import Version
 
 from twisted.web import html, resource
 from twisted.web.template import (
-    TagLoader, XMLFile, Element, renderer, flattenString)
+    TagLoader, XMLString, Element, renderer, flattenString)
 
 
 def redirectTo(URL, request):
@@ -368,7 +368,7 @@ class FailureElement(Element):
 
     @since: 12.1
     """
-    loader = XMLFile(FilePath(__file__).sibling('failure.xhtml').open())
+    loader = XMLString(FilePath(__file__).sibling('failure.xhtml').getContent())
 
     def __init__(self, failure, loader=None):
         Element.__init__(self, loader)
