Index: doc/core/howto/tutorial/library.xhtml
===================================================================
--- doc/core/howto/tutorial/library.xhtml	(revision 27295)
+++ doc/core/howto/tutorial/library.xhtml	(working copy)
@@ -13,10 +13,12 @@
 <p> This is the tenth part of the Twisted tutorial <a
 href="index.xhtml">Twisted from Scratch, or The Evolution of Finger</a>.</p>
 
-<p>In this part, we separate the application code that launches a finger
-service from the library code which defines a finger service, placing the
-application in a Twisted Application Configuration (.tac) file. We also move
-configuration (such as HTML templates) into separate files.</p>
+<p>In this part, we separate the application code that launches a finger service
+from the library code which defines a finger service, placing the application in
+a Twisted Application Configuration (.tac) file. We also move configuration
+(such as HTML templates) into separate files. Configuration and deployment with
+.tac and twistd are introduced in <a href="../application.xhtml">Using
+Application</a>.</p>
 
 <h2>Organization</h2>
 
Index: doc/core/howto/tutorial/components.xhtml
===================================================================
--- doc/core/howto/tutorial/components.xhtml	(revision 27295)
+++ doc/core/howto/tutorial/components.xhtml	(working copy)
@@ -14,7 +14,9 @@
 href="index.xhtml">Twisted from Scratch, or The Evolution of Finger</a>.</p>
 
 <p>In this section of the tutorial, we'll move our code to a component
-architecture so that adding new features is trivial.</p>
+architecture so that adding new features is trivial.
+See <a href="../components.xhtml">Interfaces and Adapters</a> for a more complete
+discussion of components.</p>
 
 <h2>Write Maintainable Code</h2>
 
Index: doc/core/howto/tutorial/configuration.xhtml
===================================================================
--- doc/core/howto/tutorial/configuration.xhtml	(revision 27295)
+++ doc/core/howto/tutorial/configuration.xhtml	(working copy)
@@ -14,7 +14,9 @@
 href="index.xhtml">Twisted from Scratch, or The Evolution of Finger</a>.</p>
 
 <p>In this part, we make it easier for non-programmers to configure a finger
-server, and show how to package it in the .deb and RPM package formats.</p>
+server, and show how to package it in the .deb and RPM package formats. Plugins,
+tap, .tac, and packaging are discussed further throughout the High Level Twisted
+section of the <a href="../index.xhtml">core howto index</a>.</p>
 
 <h2>Plugins</h2>
 
Index: doc/core/howto/tutorial/factory.xhtml
===================================================================
--- doc/core/howto/tutorial/factory.xhtml	(revision 27295)
+++ doc/core/howto/tutorial/factory.xhtml	(working copy)
@@ -15,8 +15,9 @@
 <p> This is the eighth part of the Twisted tutorial <a
 href="index.xhtml">Twisted from Scratch, or The Evolution of Finger</a>.</p>
 
-<p>In this part, we add HTTPS support to our web frontend, showing how to have
-a single factory listen on multiple ports.</p>
+<p>In this part, we add HTTPS support to our web frontend, showing how to have a
+single factory listen on multiple ports. More information on using SSL in
+Twisted can be found in the <a href="../ssl.xhtml">SSL howto</a>.</p>
 
 <h2>Support HTTPS</h2>
 
Index: doc/core/howto/tutorial/index.xhtml
===================================================================
--- doc/core/howto/tutorial/index.xhtml	(revision 27295)
+++ doc/core/howto/tutorial/index.xhtml	(working copy)
@@ -28,6 +28,19 @@
 demonstrate some of Twisted's higher-level features.
 </p>
 
+<p>
+Each section of the tutorial dives straight into applications for various
+Twisted topics. These topics have their own introductory howtos listed in
+the <a href="../index.xhtml">core howto index</a> and in the documentation for
+other Twisted projects like Twisted Web and Twisted Words. There are at least
+three ways to use this tutorial: you may find it useful to read through the rest
+of the topics listed in the <a href="../index.xhtml">core howto index</a> before
+working through the finger tutorial, work through the finger tutorial and then
+go back and hit the introductory material that is relevant to the Twisted
+project you're working on, or read the introductory material one piece at a time
+as it comes up in the finger tutorial.
+</p>
+
 <h2>Contents</h2>
 
 <p>
Index: doc/core/howto/tutorial/intro.xhtml
===================================================================
--- doc/core/howto/tutorial/intro.xhtml	(revision 27295)
+++ doc/core/howto/tutorial/intro.xhtml	(working copy)
@@ -36,6 +36,13 @@
 organization. You can also define pseudo "users", which are essentially
 keywords.</p>
 
+<p>This portion of the tutorial makes use of factories and protocols as
+introduced in the <a href="../servers.xhtml">Writing a TCP Server howto</a> and
+deferreds as introduced in <a href="../defer.xhtml">Using Deferreds</a>
+and <a href="../gendefer.xhtml">Generating Deferreds</a>. Services and
+applications are discussed in <a href="../application.xhtml">Using
+Application</a>.</p>
+
 <p>By the end of this section of the tutorial, our finger server will answer
 TCP finger requests on port 1079, and will read data from the web.</p>
 
Index: doc/core/howto/tutorial/pb.xhtml
===================================================================
--- doc/core/howto/tutorial/pb.xhtml	(revision 27295)
+++ doc/core/howto/tutorial/pb.xhtml	(working copy)
@@ -14,7 +14,9 @@
 href="index.xhtml">Twisted from Scratch, or The Evolution of Finger</a>.</p>
 
 <p>In this part, we add a Perspective Broker service to the finger application
-so that Twisted clients can access the finger server.</p>
+so that Twisted clients can access the finger server. Perspective Broker is
+introduced in depth in its own section of the <a href="../index.xhtml">core
+howto index</a>.</p>
 
 <h2>Use Perspective Broker</h2>
 
Index: doc/core/howto/tutorial/protocol.xhtml
===================================================================
--- doc/core/howto/tutorial/protocol.xhtml	(revision 27295)
+++ doc/core/howto/tutorial/protocol.xhtml	(working copy)
@@ -15,7 +15,12 @@
 
 <p>In this section of the tutorial, our finger server will continue to sprout
 features: the ability for users to set finger announces, and using our finger
-service to send those announcements on the web, on IRC and over XML-RPC.</p>
+service to send those announcements on the web, on IRC and over XML-RPC.
+Resources and XML-RPC are introduced in the Web Applications portion of
+the <a href="../../../web/howto/index.xhtml">Twisted Web howto</a>. More examples
+using <code class="API">twisted.words.protocols.irc</code> can be found
+in <a href="../clients.xhtml">Writing a TCP Client</a> and
+the <a href="../../../words/examples/index.xhtml">Twisted Words examples</a>.</p>
 
 <h2>Setting Message By Local Users</h2>
 
