diff --git a/twisted/web/static.py b/twisted/web/static.py
index a8932e6..3ff2d7a 100644
--- a/twisted/web/static.py
+++ b/twisted/web/static.py
@@ -166,6 +166,7 @@ class File(resource.Resource, styles.Versioned, filepath.FilePath):
     return the contents of /tmp/foo/bar.html .
 
     @cvar childNotFound: L{Resource} used to render 404 Not Found error pages.
+    @cvar forbidden: L{Resource} used to render 403 Forbidden error pages.
     """
 
     contentTypes = loadMimeTypes()
@@ -267,6 +268,7 @@ class File(resource.Resource, styles.Versioned, filepath.FilePath):
         self.ignoredExts.append(ext)
 
     childNotFound = resource.NoResource("File not found.")
+    forbidden = resource.ForbiddenResource()
 
     def directoryListing(self):
         return DirectoryLister(self.path,
@@ -615,7 +617,7 @@ class File(resource.Resource, styles.Versioned, filepath.FilePath):
         except IOError, e:
             import errno
             if e[0] == errno.EACCES:
-                return resource.ForbiddenResource().render(request)
+                return self.forbidden.render(request)
             else:
                 raise
 
diff --git a/twisted/web/test/test_static.py b/twisted/web/test/test_static.py
index 07fb9af..df8296f 100644
--- a/twisted/web/test/test_static.py
+++ b/twisted/web/test/test_static.py
@@ -4,6 +4,7 @@
 """
 Tests for L{twisted.web.static}.
 """
+import errno
 import inspect
 import mimetypes
 import os
@@ -153,6 +154,34 @@ class StaticFileTests(TestCase):
         test_forbiddenResource.skip = "Cannot remove read permission on Windows"
 
 
+    def test_forbiddenResource_customize(self):
+        """
+        The resource rendered for forbidden requests is stored as an instance
+        member so that users can customize it.
+        """
+        base = FilePath(self.mktemp())
+        base.setContent('')
+        marker_resonse = 'custom-forbidden-response'
+
+        def failingOpenForReading():
+            raise IOError(errno.EACCES)
+
+        class CustomForbiddenResource(resource.Resource):
+            def render(self, request):
+                return marker_resonse
+
+        class CustomStaticFile(static.File):
+            forbidden = CustomForbiddenResource()
+
+        file_resource = CustomStaticFile(base.path)
+        file_resource.openForReading = failingOpenForReading
+        request = DummyRequest([''])
+
+        result = file_resource.render(request)
+
+        self.assertEqual(marker_resonse, result)
+
+
     def test_indexNames(self):
         """
         If a request is made which encounters a L{File} before a final empty
diff --git a/twisted/web/topfiles/6951.misc b/twisted/web/topfiles/6951.misc
new file mode 100644
index 0000000..e69de29
