Index: doc/core/howto/gendefer.xhtml
===================================================================
--- doc/core/howto/gendefer.xhtml	(revision 37993)
+++ doc/core/howto/gendefer.xhtml	(working copy)
@@ -169,7 +169,7 @@
 <p>Sometimes you might wish to return a Deferred from a synchronous function.
 There are several reasons why, the major two are maintaining API compatibility
 with another version of your function which returns a Deferred, or allowing
-for the possiblity that in the future your function might need to be
+for the possibility that in the future your function might need to be
 asynchronous.</p>
 
 <p>In the <a href="defer.xhtml">Using Deferreds</a> reference, we gave the
