=== modified file 'twisted/web/test/test_http.py'
--- twisted/web/test/test_http.py	2013-03-06 03:23:18 +0000
+++ twisted/web/test/test_http.py	2013-03-29 10:54:06 +0000
@@ -1482,6 +1482,15 @@
             req.received_cookies, {b"test": b'"lemur"', b"test2": b'"panda"'})
 
 
+    def test_parseCookiesNoCookie(self):
+        """
+        L{http.Request.parseCookies} returns None if the Cookies are not part 
+        of the header.
+        """
+        req = http.Request(DummyChannel(), None)
+        self.assertEqual(req.parseCookies(), None)
+
+
     def test_connectionLost(self):
         """
         L{http.Request.connectionLost} closes L{Request.content} and drops the

