Index: spread/util.py
===================================================================
--- spread/util.py	(revision 22462)
+++ spread/util.py	(working copy)
@@ -156,8 +156,9 @@
     implements(interfaces.IConsumer)
 
     def __init__(self, collector, fd, callback=None, *args, **kw):
+        self.fd = fd
+        self.startedProducing = False
         self.chunks = []
-        self.startProducing(fd)
         Pager.__init__(self, collector, callback, *args, **kw)
 
     def startProducing(self, fd):
@@ -179,6 +180,9 @@
         """
         Get the first chunk read and send it to collector.
         """
+        if not self.startedProducing:
+            self.startProducing(self.fd)
+            self.startedProducing = True
         if not self.chunks:
             return
         val = self.chunks.pop(0)
