Index: test/test_pb.py
===================================================================
--- test/test_pb.py	(revision 22474)
+++ test/test_pb.py	(working copy)
@@ -835,6 +835,29 @@
         self.assertEquals(''.join(l[0]), bigString,
                           "Pages received not equal to pages sent!")
 
+    def test_emptyFilePaging(self):
+        """
+        Test L{util.FilePager}, sending an empty file
+        """
+        filenameEmpty = self.mktemp()
+        fd = file(filenameEmpty, 'w')
+        fd.close()
+        c, s, pump = connectedServerAndClient()
+        pagerizer = FilePagerizer(filenameEmpty, None)
+        s.setNameForLocal("bar", pagerizer)
+        x = c.remoteForName("bar")
+        l = []
+        util.getAllPages(x, "getPages").addCallback(l.append)
+        ttl = 10
+        while not l and ttl > 0:
+            pump.pump()
+            ttl -= 1
+        print ttl
+        if not ttl:
+            self.fail('getAllPages timed out')
+        self.assertEquals(''.join(l[0]), '',
+                          "Pages received not equal to pages sent!")
+
     def test_filePagingWithCallback(self):
         """
         Test L{util.FilePager}, passing a callback to fire when all pages
