Index: twisted/cred/portal.py
===================================================================
--- twisted/cred/portal.py	(revision 19824)
+++ twisted/cred/portal.py	(working copy)
@@ -102,12 +102,11 @@
         this will not be in connectionLost (such as in a web-based session), it
         will always be at the end of a user's interactive session.
         """
-        ifac = providedBy(credentials)
-        for i in ifac:
-            c = self.checkers.get(i)
-            if c is not None:
-                return maybeDeferred(c.requestAvatarId, credentials
+        for i in self.checkers:
+            if i.providedBy(credentials):
+                return maybeDeferred(self.checkers[i].requestAvatarId, credentials
                     ).addCallback(self.realm.requestAvatar, mind, *interfaces
                     )
+        ifac = providedBy(credentials)
         return defer.fail(failure.Failure(error.UnhandledCredentials(
             "No checker for %s" % ', '.join(map(reflect.qual, ifac)))))
