Index: twisted/internet/pollreactor.py
===================================================================
--- twisted/internet/pollreactor.py	(revision 41970)
+++ twisted/internet/pollreactor.py	(working copy)
@@ -69,14 +69,16 @@
         self._writes = {}
         posixbase.PosixReactorBase.__init__(self)
 
-
-    def _updateRegistration(self, fd):
-        """Register/unregister an fd with the poller."""
+    def _unregister(self, fd): 
+        """Unregister an fd with the poller."""
         try:
             self._poller.unregister(fd)
         except KeyError:
             pass
+        self._updateRegistration(fd)
 
+    def _updateRegistration(self, fd):
+        """Register an fd with the poller."""
         mask = 0
         if fd in self._reads:
             mask = mask | POLLIN
@@ -107,7 +109,7 @@
                 return
         if fd in mdict:
             del mdict[fd]
-            self._updateRegistration(fd)
+            self._unregister(fd)
 
     def addReader(self, reader):
         """Add a FileDescriptor for notification of data available to read.
