diff --git a/twisted/internet/address.py b/twisted/internet/address.py
index 6b34b99..e80f460 100644
--- a/twisted/internet/address.py
+++ b/twisted/internet/address.py
@@ -50,6 +50,20 @@ class _IPAddress(FancyEqMixin, object):
         return hash((self.type, self.host, self.port))
 
 
+    def __getitem__(self, index):
+        if (index == 0):
+            return self.type
+        elif (index == 1):
+            return self.host
+        elif (index == 2):
+            return self.port
+        else:
+            if isinstance(index, int):
+                raise IndexError('Out of bounds key index = %s.' % index)
+            else:
+                raise TypeError('Invalid index type %s' % type(index))
+
+
 
 class IPv4Address(_IPAddress):
     """
