Index: twisted/test/test_log.py
===================================================================
--- twisted/test/test_log.py	(revision 32312)
+++ twisted/test/test_log.py	(working copy)
@@ -231,6 +231,36 @@
                                        line))
 
 
+    def test_removeObserver(self):
+        """
+        Removing an observer during C{log.msg} call has no ill effects.
+        """
+        observers = []
+        def cleanup():
+            for observer in observers:
+                log.removeObserver(observer)
+
+        self.addCleanup(cleanup)
+
+        def dummyObserver(eventDict):
+            pass
+
+        def evilObserver(eventdict):
+            for i in range(2):
+                log.removeObserver(dummyObserver)
+                observers.remove(dummyObserver)
+
+        for i in range(2):
+            log.addObserver(dummyObserver)
+            observers.append(dummyObserver)
+        log.addObserver(evilObserver)
+        observers.append(evilObserver)
+
+        log.msg("Hello!")
+        self.assertEqual(observers, [evilObserver])
+
+
+
 class FakeFile(list):
     def write(self, bytes):
         self.append(bytes)
