Index: stream.py
===================================================================
--- stream.py	(revision 15328)
+++ stream.py	(working copy)
@@ -989,7 +989,7 @@
         the entire stream into a string."""
         def gotdata():
             data = self.data
-            if size is not None and len(data) >= size:
+            if size is not None and size >= 0 and len(data) >= size:
                 pre,post = data[:size], data[size:]
                 self.data=post
                 return pre
