Index: twisted/python/test/test_versions.py
===================================================================
--- twisted/python/test/test_versions.py	(revision 25177)
+++ twisted/python/test/test_versions.py	(working copy)
@@ -40,7 +40,24 @@
 """
 
 
+VERSION_9_ENTRIES = """\
+9
 
+dir
+22715
+svn+ssh://svn.twistedmatrix.com/svn/Twisted/trunk
+"""
+
+
+VERSION_10_ENTRIES = """\
+10
+
+dir
+22715
+svn+ssh://svn.twistedmatrix.com/svn/Twisted/trunk
+"""
+
+
 class VersionsTest(unittest.TestCase):
 
     def test_versionComparison(self):
@@ -160,6 +177,24 @@
         version = Version("dummy", 1, 0, 0)
         self.assertEqual(
             version._parseSVNEntries_8(StringIO(VERSION_8_ENTRIES)), '22715')
+        
+        
+    def test_goodSVNEntries_9(self):
+        """
+        Version should be able to parse an SVN format 9 entries file.
+        """
+        version = Version("dummy", 1, 0, 0)
+        self.assertEqual(
+            version._parseSVNEntries_8(StringIO(VERSION_8_ENTRIES)), '22715')
+        
+        
+    def test_goodSVNEntries_10(self):
+        """
+        Version should be able to parse an SVN format 10 entries file.
+        """
+        version = Version("dummy", 1, 0, 0)
+        self.assertEqual(
+            version._parseSVNEntries_8(StringIO(VERSION_8_ENTRIES)), '22715')
 
 
     def test_getVersionString(self):
@@ -258,6 +293,22 @@
         parsed.
         """
         self.checkSVNFormat("8", VERSION_8_ENTRIES, '22715')
+        
+    
+    def test_detectVersion9(self):
+        """
+        Verify that version 9 format files will be properly detected and
+        parsed.
+        """
+        self.checkSVNFormat("9", VERSION_10_ENTRIES, '22715')
+        
+        
+    def test_detectVersion10(self):
+        """
+        Verify that version 10 format files will be properly detected and
+        parsed.
+        """
+        self.checkSVNFormat("10", VERSION_10_ENTRIES, '22715')
 
 
     def test_detectUnknownVersion(self):
Index: twisted/python/versions.py
===================================================================
--- twisted/python/versions.py	(revision 25177)
+++ twisted/python/versions.py	(working copy)
@@ -182,6 +182,11 @@
         entriesFile.readline()
         entriesFile.readline()
         return entriesFile.readline().strip()
+    
+    
+    #Add handlers for version 9 and 10 formats
+    _parseSVNEntries_9 = _parseSVNEntries_8
+    _parseSVNEntries_10 = _parseSVNEntries_8
 
 
     def _getSVNVersion(self):
