Index: distrib.py
===================================================================
RCS file: /cvs/Twisted/twisted/web/distrib.py,v
retrieving revision 1.52
diff -u -r1.52 distrib.py
--- distrib.py	16 Oct 2003 16:42:39 -0000	1.52
+++ distrib.py	15 Mar 2004 03:28:11 -0000
@@ -291,4 +291,7 @@
             self.putChild(name, rs)
             return rs
         else:
+            path = os.path.join(pw_dir, self.userDirName)
+            if not os.path.exists(path):
+                return error.NoResource()
             return static.File(os.path.join(pw_dir, self.userDirName))
