Index: twisted/internet/interfaces.py
===================================================================
--- twisted/internet/interfaces.py	(revision 26633)
+++ twisted/internet/interfaces.py	(working copy)
@@ -454,9 +454,15 @@
                      sequence of strings. The first string should be the
                      executable's name.
 
-        @param env: the environment variables to pass to the processs; a
-                    dictionary of strings. If 'None', use os.environ.
-
+        @param env: Environment variables to pass to the process;
+            A dictionary of strings. If 'None', os.environ will be used.
+            On Posix:
+                If not specified, an empty environment is passed in.
+            On Windows:
+                If not specified os.environ is used.
+                Otherwise, os.environ will be updated with the value and passed
+                in.
+            
         @param path: the path to run the subprocess in - defaults to the
                      current directory.
