diff --git twisted/python/reflect.py twisted/python/reflect.py
index f4f3432..f7db391 100644
--- twisted/python/reflect.py
+++ twisted/python/reflect.py
@@ -132,12 +132,12 @@ def isinst(inst,clazz):
 ## the following were factored out of usage
 
 if not _PY3:
-    # These functions are still imported by libraries used in turn by Twisted,
-    # like Nevow 0.10. Since they are deprecated, there's no need to port them
-    # to Python 3 (hence the condition above).
-    # https://code.launchpad.net/~multani/divmod.org/remove-deprecated-twisted.python.reflect
-    # removes the dependency in Nevow. Once this gets merged and released, these
-    # functions can be safely removed from Twisted.
+    # These functions are still imported by libraries used in turn by the
+    # Twisted unit tests, like Nevow 0.10. Since they are deprecated,
+    # there's no need to port them to Python 3 (hence the condition above).
+    # https://bazaar.launchpad.net/~divmod-dev/divmod.org/trunk/revision/2716
+    # removed the dependency in Nevow. Once this released, these functions can
+    # be safely removed from Twisted.
 
     @deprecated(Version("Twisted", 11, 0, 0), "inspect.getmro")
     def allYourBase(classObj, baseClass=None):
@@ -292,7 +292,7 @@ __all__ = [
 
     'funcinfo', 'fullFuncName', 'qual', 'getcurrent', 'getClass', 'isinst',
     'namedModule', 'namedObject', 'namedClass', 'namedAny',
-    'safe_repr', 'safe_str',
+    'safe_repr', 'safe_str', 'allYourBase', 'accumulateBases',
     'prefixedMethodNames', 'addMethodNamesToDict', 'prefixedMethods',
     'accumulateMethods',
     'accumulateClassDict', 'accumulateClassList', 'isSame', 'isLike',
diff --git twisted/test/test_reflect.py twisted/test/test_reflect.py
index 50229a2..a7945de 100644
--- twisted/test/test_reflect.py
+++ twisted/test/test_reflect.py
@@ -179,3 +179,27 @@ class GetClass(unittest.TestCase):
         new = NewClass()
         self.assertEqual(reflect.getClass(NewClass).__name__, 'type')
         self.assertEqual(reflect.getClass(new).__name__, 'NewClass')
+
+
+class DeprecationTestCase(unittest.TestCase):
+    """
+    Test deprecations in twisted.python.reflect
+    """
+
+    def test_allYourBase(self):
+        """
+        Test deprecation of L{reflect.allYourBase}. See #5481 for removal.
+        """
+        self.callDeprecated(
+            (Version("Twisted", 11, 0, 0), "inspect.getmro"),
+            reflect.allYourBase, DeprecationTestCase)
+
+
+    def test_accumulateBases(self):
+        """
+        Test deprecation of L{reflect.accumulateBases}. See #5481 for removal.
+        """
+        l = []
+        self.callDeprecated(
+            (Version("Twisted", 11, 0, 0), "inspect.getmro"),
+            reflect.accumulateBases, DeprecationTestCase, l, None)
