--- test_internet.py	Mon May 22 16:21:54 2006
+++ test_internet2.py	Mon May 22 16:25:21 2006
@@ -193,6 +193,33 @@
         self.failUnlessRaises(TypeError,
                               r.removeSystemEventTrigger, None)
 
+    def addSelfRemovingTrigger(self, phase, event, func):
+        class TriggerHolder: pass
+        th = TriggerHolder()
+        def remove(self=self, th=th, func=func):
+            try:
+                return func()
+            finally:
+                self.removeTrigger(th.trigger)
+        t = reactor.addSystemEventTrigger(phase, event, remove)
+        th.trigger = t
+        self.triggers.append(t)
+        return t
+
+    def testSelfRemovingTrigger(self):
+        l = []
+        def _appendToList(l=l):
+            l.append(1)
+        
+        r = reactor
+
+        self.addSelfRemovingTrigger("before", "test", _appendToList)
+        self.addSelfRemovingTrigger("before", "test", _appendToList)
+        self.assertEquals(len(l), 0, "Nothing happened yet.")
+        r.fireSystemEvent("test")
+        r.iterate()
+        self.assertEquals(len(l), 2, "Should have filled the list (got %s)." % len(l))
+
 
 class InterfaceTestCase(unittest.TestCase):
 
