Index: twisted/python/log.py
===================================================================
--- twisted/python/log.py	(revision 34526)
+++ twisted/python/log.py	(working copy)
@@ -531,6 +531,15 @@
             return
         self.logger.log(level, text)
 
+        extraDict = eventDict.copy()
+        for key in extraDict.keys():
+            if key in ['asctime', 'created', 'filename', 'funcName',
+                       'levelname', 'levelno', 'lineno', 'module', 'msecs',
+                       'message', 'name', 'pathname', 'process', 'processName',
+                       'relativeCreated', 'thread', 'threadName']:
+                del extraDict[key]
+        self.logger.log(level, text, extra=extraDict)
+
     def start(self):
         """
         Start observing log events.
