Index: twisted/test/test_twistd.py
===================================================================
--- twisted/test/test_twistd.py (revision 28978)
+++ twisted/test/test_twistd.py (working copy)
@@ -1,4 +1,4 @@
-# Copyright (c) 2007-2009 Twisted Matrix Laboratories.
+# Copyright (c) 2007-2010 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 """
@@ -911,20 +911,6 @@
             sys.modules.update(savedModules)
 
 
-    def test_nothotshotDeprecation(self):
-        """
-        Check that switching on the C{nothotshot} option produces a warning and
-        sets the profiler to B{profile}.
-        """
-        config = twistd.ServerOptions()
-        config['nothotshot'] = True
-        profiler = self.assertWarns(DeprecationWarning,
-            "The --nothotshot option is deprecated. Please specify the "
-            "profiler name using the --profiler option",
-            app.__file__, app.AppProfiler, config)
-        self.assertEquals(profiler.profiler, "profile")
-
-
     def test_hotshotPrintStatsError(self):
         """
         When an error happens while printing the stats, C{sys.stdout}
@@ -1042,51 +1028,7 @@
         self.assertEquals(profiler.profiler, "hotshot")
 
 
-    def test_oldRunWithProfiler(self):
-        """
-        L{app.runWithProfiler} should print a C{DeprecationWarning} pointing
-        at L{AppProfiler}.
-        """
-        class DummyProfiler(object):
-            called = False
-            def run(self, reactor):
-                self.called = True
-        profiler = DummyProfiler()
-        self.patch(app, "AppProfiler", lambda conf: profiler)
 
-        def runWithProfiler():
-            return app.runWithProfiler(DummyReactor(), {})
-
-        self.assertWarns(DeprecationWarning,
-                "runWithProfiler is deprecated since Twisted 8.0. "
-                "Use ProfileRunner instead.", __file__,
-                runWithProfiler)
-        self.assertTrue(profiler.called)
-
-
-    def test_oldRunWithHotshot(self):
-        """
-        L{app.runWithHotshot} should print a C{DeprecationWarning} pointing
-        at L{AppProfiler}.
-        """
-        class DummyProfiler(object):
-            called = False
-            def run(self, reactor):
-                self.called = True
-        profiler = DummyProfiler()
-        self.patch(app, "AppProfiler", lambda conf: profiler)
-
-        def runWithHotshot():
-            return app.runWithHotshot(DummyReactor(), {})
-
-        self.assertWarns(DeprecationWarning,
-                "runWithHotshot is deprecated since Twisted 8.0. "
-                "Use HotshotRunner instead.", __file__,
-                runWithHotshot)
-        self.assertTrue(profiler.called)
-
-
-
 def _patchFileLogObserver(patch):
     """
     Patch L{log.FileLogObserver} to record every call and keep a reference to
Index: twisted/application/app.py
===================================================================
--- twisted/application/app.py  (revision 28978)
+++ twisted/application/app.py  (working copy)
@@ -174,11 +174,6 @@
         saveStats = options.get("savestats", False)
         profileOutput = options.get("profile", None)
         self.profiler = options.get("profiler", "hotshot").lower()
-        if options.get("nothotshot", False):
-            warnings.warn("The --nothotshot option is deprecated. Please "
-                          "specify the profiler name using the --profiler "
-                          "option", category=DeprecationWarning)
-            self.profiler = "profile"
         if self.profiler in self.profilers:
             profiler = self.profilers[self.profiler](profileOutput, saveStats)
             self.run = profiler.run
@@ -187,32 +182,6 @@
 
 
 
-def runWithProfiler(reactor, config):
-    """
-    DEPRECATED in Twisted 8.0.
-
-    Run reactor under standard profiler.
-    """
-    warnings.warn("runWithProfiler is deprecated since Twisted 8.0. "
-                  "Use ProfileRunner instead.", DeprecationWarning, 2)
-    item = AppProfiler(config)
-    return item.run(reactor)
-
-
-
-def runWithHotshot(reactor, config):
-    """
-    DEPRECATED in Twisted 8.0.
-
-    Run reactor under hotshot profiler.
-    """
-    warnings.warn("runWithHotshot is deprecated since Twisted 8.0. "
-                  "Use HotshotRunner instead.", DeprecationWarning, 2)
-    item = AppProfiler(config)
-    return item.run(reactor)
-
-
-
 class AppLogger(object):
     """
     Class managing logging faciliy of the application.
@@ -330,12 +299,6 @@
         if config['profile']:
             if profiler is not None:
                 profiler.run(reactor)
-            else:
-                # Backward compatible code
-                if not config['nothotshot']:
-                    runWithHotshot(reactor, config)
-                else:
-                    runWithProfiler(reactor, config)
         elif config['debug']:
             sys.stdout = oldstdout
             sys.stderr = oldstderr
@@ -580,10 +543,7 @@
                  "the profiler."],
                 ['no_save','o',   "do not save state on shutdown"],
                 ['encrypted', 'e',
-                 "The specified tap/aos file is encrypted."],
-                ['nothotshot', None,
-                 "DEPRECATED. Don't use the hotshot profiler even if "
-                 "it's available."]]
+                 "The specified tap/aos file is encrypted."]]
 
     optParameters = [['logfile','l', None,
                       "log to a specified file, - for stdout"],
