--- /usr/lib/python2.3/site-packages/twisted/xish/domish.py	2004-10-06 09:06:29.000000000 +0200
+++ lib/python/twisted/xish/domish.py	2005-03-24 17:38:07.000000000 +0100
@@ -88,7 +88,7 @@
         for k,v in elem.attributes.items():
             # If the attribute name is a list, it's a qualified attribute
             if isinstance(k, types.TupleType):
-                write((" %s:%s='%s'" % (self.getPrefix[k[0]], k[1], escapeToXml(v, 1))).encode("utf-8"))
+                write((" %s:%s='%s'" % (self.getPrefix(k[0]), k[1], escapeToXml(v, 1))).encode("utf-8"))
             else:
                 write((" %s='%s'" % ( k, escapeToXml(v, 1))).encode("utf-8"))
 
@@ -165,7 +165,7 @@
         for k,v in elem.attributes.items():
             # If the attribute name is a list, it's a qualified attribute
             if isinstance(k, types.TupleType):
-                write(" %s:%s='%s'" % (self.getPrefix[k[0]], k[1], escapeToXml(v, 1)))
+                write(" %s:%s='%s'" % (self.getPrefix(k[0]), k[1], escapeToXml(v, 1)))
             else:
                 write((" %s='%s'" % ( k, escapeToXml(v, 1))))
 
