diff --git twisted/web/server.py twisted/web/server.py
index bc3f07c..408c585 100644
--- twisted/web/server.py
+++ twisted/web/server.py
@@ -216,8 +216,11 @@ class Request(Copyable, http.Request, components.Componentized):
         # though, so it's still okay.
         if not self._inFakeHead:
             if self._encoder:
+                print 'encode'
                 data = self._encoder.encode(data)
-            http.Request.write(self, data)
+            print data
+            if data is not None:
+                http.Request.write(self, data)
 
 
     def finish(self):
