Index: _glibbase.py
===================================================================
--- _glibbase.py	(revision 33585)
+++ _glibbase.py	(working copy)
@@ -12,6 +12,7 @@
 """
 
 import signal
+from functools import partial
 
 from twisted.internet import base, posixbase, selectreactor
 from twisted.internet.interfaces import IReactorFDSet
@@ -272,7 +273,7 @@
         self.wakeUp()
 
 
-    def run(self, installSignalHandlers=True):
+    def run(self, installSignalHandlers=False):
         """
         Run the reactor.
         """
@@ -314,7 +315,14 @@
         self._reschedule()
 
 
+    def registerGtkApplication(self, app):
+        """
+        Support GtkApplication.run main loops, instead of the default main loop.
+        """
+        self._run = partial(app.run, None)
 
+
+
 class PortableGlibReactorBase(GlibSignalMixin, selectreactor.SelectReactor):
     """
     Base class for GObject event loop reactors that works on Windows.
@@ -348,7 +356,7 @@
         self._crash()
 
 
-    def run(self, installSignalHandlers=True):
+    def run(self, installSignalHandlers=False):
         self.startRunning(installSignalHandlers=installSignalHandlers)
         self._timeout_add(0, self.simulate)
         if self._started:
@@ -366,3 +374,10 @@
         if timeout is None:
             timeout = 0.01
         self._simtag = self._timeout_add(int(timeout * 1000), self.simulate)
+
+
+    def registerGApplication(self, app):
+        """
+        Support GApplication.run main loops, instead of the default main loop.
+        """
+        self._run = partial(app.run, None)
