Index: twisted/enterprise/adbapi.py
===================================================================
--- twisted/enterprise/adbapi.py	(revision 22997)
+++ twisted/enterprise/adbapi.py	(working copy)
@@ -9,7 +9,6 @@
 
 from twisted.internet import defer, threads
 from twisted.python import reflect, log
-from twisted.enterprise.util import safe # backwards compat
 
 
 class ConnectionLost(Exception):
