diff --git a/twisted/test/test_log.py b/twisted/test/test_log.py
index d07ca6a..100161d 100644
--- a/twisted/test/test_log.py
+++ b/twisted/test/test_log.py
@@ -367,8 +367,12 @@ class FileObserverTestCase(LogPublisherTestCaseMixin, unittest.SynchronousTestCa
         by setting C{TZ} to various well-known values and asserting that the
         reported offset is correct.
         """
-        localDaylightTuple = (2006, 6, 30, 0, 0, 0, 4, 181, 1)
-        utcDaylightTimestamp = time.mktime(localDaylightTuple)
+        if time.daylight:
+            # only attempt/request DST adjustment when TZ has DST (not the case for UTC)
+            localDaylightTuple = (2006, 6, 30, 0, 0, 0, 4, 181, 1)
+            utcDaylightTimestamp = time.mktime(localDaylightTuple)
+        else:
+            utcDaylightTimestamp = None
         localStandardTuple = (2007, 1, 31, 0, 0, 0, 2, 31, 0)
         utcStandardTimestamp = time.mktime(localStandardTuple)
 
@@ -377,9 +381,10 @@ class FileObserverTestCase(LogPublisherTestCaseMixin, unittest.SynchronousTestCa
             # Test something west of UTC
             os.environ['TZ'] = 'America/New_York'
             time.tzset()
-            self.assertEqual(
-                self.flo.getTimezoneOffset(utcDaylightTimestamp),
-                14400)
+            if utcDaylightTimestamp is not None:
+                self.assertEqual(
+                    self.flo.getTimezoneOffset(utcDaylightTimestamp),
+                    14400)
             self.assertEqual(
                 self.flo.getTimezoneOffset(utcStandardTimestamp),
                 18000)
@@ -387,9 +392,10 @@ class FileObserverTestCase(LogPublisherTestCaseMixin, unittest.SynchronousTestCa
             # Test something east of UTC
             os.environ['TZ'] = 'Europe/Berlin'
             time.tzset()
-            self.assertEqual(
-                self.flo.getTimezoneOffset(utcDaylightTimestamp),
-                -7200)
+            if utcDaylightTimestamp is not None:
+                self.assertEqual(
+                    self.flo.getTimezoneOffset(utcDaylightTimestamp),
+                    -7200)
             self.assertEqual(
                 self.flo.getTimezoneOffset(utcStandardTimestamp),
                 -3600)
@@ -397,9 +403,10 @@ class FileObserverTestCase(LogPublisherTestCaseMixin, unittest.SynchronousTestCa
             # Test a timezone that doesn't have DST
             os.environ['TZ'] = 'Africa/Johannesburg'
             time.tzset()
-            self.assertEqual(
-                self.flo.getTimezoneOffset(utcDaylightTimestamp),
-                -7200)
+            if utcDaylightTimestamp is not None:
+                self.assertEqual(
+                    self.flo.getTimezoneOffset(utcDaylightTimestamp),
+                    -7200)
             self.assertEqual(
                 self.flo.getTimezoneOffset(utcStandardTimestamp),
                 -7200)
