diff --git a/twisted/test/test_log.py b/twisted/test/test_log.py
index d07ca6a..3ecf5ef 100644
--- a/twisted/test/test_log.py
+++ b/twisted/test/test_log.py
@@ -367,8 +367,10 @@ class FileObserverTestCase(LogPublisherTestCaseMixin, unittest.SynchronousTestCa
         by setting C{TZ} to various well-known values and asserting that the
         reported offset is correct.
         """
-        localDaylightTuple = (2006, 6, 30, 0, 0, 0, 4, 181, 1)
-        utcDaylightTimestamp = time.mktime(localDaylightTuple)
+        if time.daylight:
+            # only attempt/request DST adjustment when TZ has DST (not the case for UTC)
+            localDaylightTuple = (2006, 6, 30, 0, 0, 0, 4, 181, 1)
+            utcDaylightTimestamp = time.mktime(localDaylightTuple)
         localStandardTuple = (2007, 1, 31, 0, 0, 0, 2, 31, 0)
         utcStandardTimestamp = time.mktime(localStandardTuple)
 
