diff --git a/tox.ini b/tox.ini
new file mode 100644
index 0000000..c28e5d3
--- /dev/null
+++ b/tox.ini
@@ -0,0 +1,51 @@
+[tox]
+envlist=py{26,27,33},py{26,27}-{select_gc,without_modules},pyflakes,twistedchecker,api_docs,narrative_docs
+skipsdist=True
+
+[testenv]
+deps=
+    py{26,27},py{26,27}-{select_gc,without_modules}: zope.interface >= 3.6.1
+    py33: zope.interface >= 4.0.0
+commands =
+    py{26,27}: {toxinidir}/bin/trial {posargs:twisted}
+    py33: {toxinidir}/admin/run-python3-tests
+    py{26,27}-without_modules: {toxinidir}/bin/trial --without-module OpenSSL --without-module Crypto {posargs:twisted}
+    py{26,27}-select_gc: {toxinidir}/bin/trial --force-gc {posargs:twisted}
+
+[testenv:pyflakes]
+basepython=python2.7
+deps= pyflakes
+commands=pyflakes {toxinidir}/{posargs:twisted}
+
+[testenv:coverage]
+basepython=python2.7
+deps=
+    coverage
+    zope.interface >= 3.6.1
+commands=
+    coverage erase
+    coverage run --omit {toxinidir}/*_trial_temp/* --branch {toxinidir}/bin/trial --reporter=bwverbose {posargs:twisted}
+    coverage report
+
+[testenv:twistedchecker]
+basepython=python2.7
+deps=twistedchecker
+commands=twistedchecker {posargs:twisted}
+
+#####################
+# Build documentation
+#####################
+
+[testenv:api_docs]
+basepython=python2.7
+deps =
+    pydoctor
+    epydoc
+    nevow
+commands={toxinidir}/bin/admin/build-apidocs {toxinidir} apidocs
+
+[testenv:narrative_docs]
+basepython=python2.7
+deps=sphinx >= 1.2.1
+changedir=docs
+commands=sphinx-build -W -b html -d {toxinidir}/docs/_build . {toxinidir}/docs/_build/
diff --git a/twisted/topfiles/7228.feature b/twisted/topfiles/7228.feature
new file mode 100644
index 0000000..401f375
--- /dev/null
+++ b/twisted/topfiles/7228.feature
@@ -0,0 +1 @@
+Add a tox.ini file to simplify testing with virtualenvs
\ No newline at end of file
