diff --git tox.ini tox.ini
new file mode 100644
index 0000000..10c15f8
--- /dev/null
+++ tox.ini
@@ -0,0 +1,32 @@
+[tox]
+envlist = py26, py27, py33, lint, twistedchecker
+skipsdist = True
+
+[base]
+deps = zope.interface
+
+[testenv]
+deps = {[base]deps}
+commands =
+        {envpython} --version
+        {toxinidir}/bin/trial {posargs:twisted}
+
+[testenv:py33]
+deps = {[base]deps}
+commands =
+        {envpython} --version
+        {toxinidir}/admin/run-python3-tests
+
+[testenv:lint]
+basepython = python2.7
+deps = pyflakes
+commands =
+        {envpython} --version
+        pyflakes {toxinidir}/{posargs:twisted}
+
+[testenv:twistedchecker]
+basepython = python2.7
+deps = git+https://github.com/twisted/twistedchecker#egg=twistedchecker
+commands =
+         {envpython} --version
+         twistedchecker {posargs:twisted}
diff --git twisted/topfiles/7228.feature twisted/topfiles/7228.feature
new file mode 100644
index 0000000..401f375
--- /dev/null
+++ twisted/topfiles/7228.feature
@@ -0,0 +1 @@
+Add a tox.ini file to simplify testing with virtualenvs
\ No newline at end of file
