Index: twisted/trial/unittest.py
===================================================================
--- twisted/trial/unittest.py	(revision 34731)
+++ twisted/trial/unittest.py	(working copy)
@@ -537,19 +537,16 @@
     def _add(self):
         if self._added == 0:
             log.addObserver(self.gotEvent)
-            self._oldFE, log._flushErrors = (log._flushErrors, self.flushErrors)
-            self._oldIE, log._ignore = (log._ignore, self._ignoreErrors)
-            self._oldCI, log._clearIgnores = (log._clearIgnores,
-                                              self._clearIgnores)
+            log._flushErrors = self.flushErrors
+            log._ignore = self._ignoreErrors
+            log._clearIgnores = self._clearIgnores
         self._added += 1
 
+
     def _remove(self):
         self._added -= 1
         if self._added == 0:
             log.removeObserver(self.gotEvent)
-            log._flushErrors = self._oldFE
-            log._ignore = self._oldIE
-            log._clearIgnores = self._oldCI
 
 
     def _ignoreErrors(self, *errorTypes):
Index: twisted/python/log.py
===================================================================
--- twisted/python/log.py	(revision 34731)
+++ twisted/python/log.py	(working copy)
@@ -22,10 +22,10 @@
 
 class ILogContext:
     """
-    Actually, this interface is just a synoym for the dictionary interface,
+    Actually, this interface is just a synonym for the dictionary interface,
     but it serves as a key for the default information in a log.
 
-    I do not inherit from Interface because the world is a cruel place.
+    I do not inherit from C{Interface} because the world is a cruel place.
     """
 
 
@@ -93,85 +93,6 @@
 _keptErrors = []
 _ignoreErrors = []
 
-def startKeepingErrors():
-    """
-    DEPRECATED in Twisted 2.5.
-
-    Support function for testing frameworks.
-
-    Start keeping errors in a buffer which can be retrieved (and emptied) with
-    flushErrors.
-    """
-    warnings.warn("log.startKeepingErrors is deprecated since Twisted 2.5",
-                  category=DeprecationWarning, stacklevel=2)
-    global _keepErrors
-    _keepErrors = 1
-
-
-def flushErrors(*errorTypes):
-    """
-    DEPRECATED in Twisted 2.5.  See L{TestCase.flushLoggedErrors}.
-
-    Support function for testing frameworks.
-
-    Return a list of errors that occurred since the last call to flushErrors().
-    (This will return None unless startKeepingErrors has been called.)
-    """
-
-    warnings.warn("log.flushErrors is deprecated since Twisted 2.5. "
-                  "If you need to flush errors from within a unittest, "
-                  "use TestCase.flushLoggedErrors instead.",
-                  category=DeprecationWarning, stacklevel=2)
-    return _flushErrors(*errorTypes)
-
-
-def _flushErrors(*errorTypes):
-    """
-    PRIVATE. DEPRECATED. DON'T USE.
-    """
-    global _keptErrors
-    k = _keptErrors
-    _keptErrors = []
-    if errorTypes:
-        for erk in k:
-            shouldReLog = 1
-            for errT in errorTypes:
-                if erk.check(errT):
-                    shouldReLog = 0
-            if shouldReLog:
-                err(erk)
-    return k
-
-def ignoreErrors(*types):
-    """
-    DEPRECATED
-    """
-    warnings.warn("log.ignoreErrors is deprecated since Twisted 2.5",
-                  category=DeprecationWarning, stacklevel=2)
-    _ignore(*types)
-
-def _ignore(*types):
-    """
-    PRIVATE. DEPRECATED. DON'T USE.
-    """
-    for type in types:
-        _ignoreErrors.append(type)
-
-def clearIgnores():
-    """
-    DEPRECATED
-    """
-    warnings.warn("log.clearIgnores is deprecated since Twisted 2.5",
-                  category=DeprecationWarning, stacklevel=2)
-    _clearIgnores()
-
-def _clearIgnores():
-    """
-    PRIVATE. DEPRECATED. DON'T USE.
-    """
-    global _ignoreErrors
-    _ignoreErrors = []
-
 
 def err(_stuff=None, _why=None, **kw):
     """
Index: twisted/topfiles/5765.removal
===================================================================
--- twisted/topfiles/5765.removal	(revision 0)
+++ twisted/topfiles/5765.removal	(revision 0)
@@ -0,0 +1 @@
+startKeepingErrors, flushErrors, ignoreErrors, and clearIgnores in twisted.python.log (deprecated since Twisted 2.5) are removed now.
