diff -ruN twistedweb2-20070612-svn.clean/twisted/web2/channel/http.py twistedweb2-20070612-svn/twisted/web2/channel/http.py
--- twistedweb2-20070612-svn.clean/twisted/web2/channel/http.py	2007-06-15 13:33:31.000000000 -0700
+++ twistedweb2-20070612-svn/twisted/web2/channel/http.py	2007-06-15 13:39:24.000000000 -0700
@@ -512,8 +512,19 @@
                 self._cleanup()
 
     def getHostInfo(self):
-        t=self.channel.transport
-        secure = interfaces.ISSLTransport(t, None) is not None
+        t = self.channel.transport
+        # This change is required to get GNUTLS working with the correct
+        # https schema
+        try:
+            import gnutls.interfaces.twisted
+            has_gnutls = True
+        except ImportError:
+            has_gnutls = False
+        if has_gnutls:
+            if isinstance(t, gnutls.interfaces.twisted.TLSServer):
+                secure = True
+        else:
+            secure = interfaces.ISSLTransport(t, None) is not None
         host = t.getHost()
         host.host = _cachedGetHostByAddr(host.host)
         return host, secure
