Index: twisted/names/test/test_cache.py
===================================================================
--- twisted/names/test/test_cache.py	(revision 43547)
+++ twisted/names/test/test_cache.py	(working copy)
@@ -17,7 +17,7 @@
 from twisted.internet import task, interfaces
 
 
-class Caching(unittest.TestCase):
+class CachingTests(unittest.TestCase):
     """
     Tests for L{cache.CacheResolver}.
     """
Index: twisted/names/test/test_names.py
===================================================================
--- twisted/names/test/test_names.py	(revision 43547)
+++ twisted/names/test/test_names.py	(working copy)
@@ -141,7 +141,7 @@
     )
 
 
-class ServerDNSTestCase(unittest.TestCase):
+class ServerDNSTests(unittest.TestCase):
     """
     Test cases for DNS server and client.
     """
@@ -434,7 +434,7 @@
 
 
 
-class HelperTestCase(unittest.TestCase):
+class HelperTests(unittest.TestCase):
     def testSerialGenerator(self):
         f = self.mktemp()
         a = authority.getSerial(f)
@@ -444,7 +444,7 @@
             a = b
 
 
-class AXFRTest(unittest.TestCase):
+class AXFRTests(unittest.TestCase):
     def setUp(self):
         self.results = None
         self.d = defer.Deferred()
@@ -501,7 +501,7 @@
 
 
 
-class ResolvConfHandling(unittest.TestCase):
+class ResolvConfHandlingTests(unittest.TestCase):
     def testMissing(self):
         resolvConf = self.mktemp()
         r = client.Resolver(resolv=resolvConf)
@@ -825,7 +825,7 @@
 
 
 
-class NoInitialResponseTestCase(unittest.TestCase):
+class NoInitialResponseTests(unittest.TestCase):
 
     def test_no_answer(self):
         """
Index: twisted/names/test/test_dns.py
===================================================================
--- twisted/names/test/test_dns.py	(revision 43547)
+++ twisted/names/test/test_dns.py	(working copy)
@@ -293,7 +293,7 @@
 
 
 
-class RoundtripDNSTestCase(unittest.TestCase):
+class RoundtripDNSTests(unittest.TestCase):
     """
     Encoding and then decoding various objects.
     """
@@ -596,7 +596,7 @@
 
 
 
-class MessageTestCase(unittest.SynchronousTestCase):
+class MessageTests(unittest.SynchronousTestCase):
     """
     Tests for L{twisted.names.dns.Message}.
     """
@@ -1116,7 +1116,7 @@
 
 
 
-class DatagramProtocolTestCase(unittest.TestCase):
+class DatagramProtocolTests(unittest.TestCase):
     """
     Test various aspects of L{dns.DNSDatagramProtocol}.
     """
@@ -1239,7 +1239,7 @@
 
 
 
-class DNSProtocolTestCase(unittest.TestCase):
+class DNSProtocolTests(unittest.TestCase):
     """
     Test various aspects of L{dns.DNSProtocol}.
     """
@@ -3727,8 +3727,8 @@
 
 
 
-class EDNSMessageSpecificsTestCase(ConstructorTestsMixin,
-                                   unittest.SynchronousTestCase):
+class EDNSMessageSpecificsTests(ConstructorTestsMixin,
+                                unittest.SynchronousTestCase):
     """
     Tests for L{dns._EDNSMessage}.
 
Index: twisted/names/test/test_srvconnect.py
===================================================================
--- twisted/names/test/test_srvconnect.py	(revision 43547)
+++ twisted/names/test/test_srvconnect.py	(working copy)
@@ -54,7 +54,7 @@
 
 
 
-class SRVConnectorTest(unittest.TestCase):
+class SRVConnectorTests(unittest.TestCase):
     """
     Tests for L{srvconnect.SRVConnector}.
     """
Index: twisted/names/test/test_client.py
===================================================================
--- twisted/names/test/test_client.py	(revision 43547)
+++ twisted/names/test/test_client.py	(working copy)
@@ -796,7 +796,7 @@
 
 
 
-class ClientTestCase(unittest.TestCase):
+class ClientTests(unittest.TestCase):
 
     def setUp(self):
         """
@@ -1139,7 +1139,7 @@
 
 
 
-class RetryLogic(unittest.TestCase):
+class RetryLogicTests(unittest.TestCase):
     """
     Tests for query retrying implemented by L{client.Resolver}.
     """
Index: twisted/names/test/test_hosts.py
===================================================================
--- twisted/names/test/test_hosts.py	(revision 43547)
+++ twisted/names/test/test_hosts.py	(working copy)
@@ -121,7 +121,7 @@
 
 
 
-class HostsTestCase(TestCase, GoodTempPathMixin):
+class HostsTests(TestCase, GoodTempPathMixin):
     """
     Tests for the I{hosts(5)}-based L{twisted.names.hosts.Resolver}.
     """
Index: twisted/names/topfiles/7728.misc
===================================================================
