Index: doc/core/howto/constants.xhtml
===================================================================
--- doc/core/howto/constants.xhtml	(revision 33677)
+++ doc/core/howto/constants.xhtml	(working copy)
@@ -316,7 +316,7 @@
 >>>
   </pre>
 
-  <p>Constants can also be combined using the logical operators <code>&</code>
+  <p>Constants can also be combined using the logical operators <code>&amp;</code>
   (<em>and</em>), <code>|</code> (<em>or</em>), and <code>^</code>
   (<em>exclusive or</em>).
   </p>
@@ -324,7 +324,7 @@
   <pre class="shell">
 >>> Permission.USER_READ | Permission.USER_WRITE
 &lt;Permission={USER_READ,USER_WRITE}>
->>> (Permission.USER_READ | Permission.USER_WRITE) & Permission.USER_WRITE
+>>> (Permission.USER_READ | Permission.USER_WRITE) &amp; Permission.USER_WRITE
 &lt;Permission=USER_WRITE>
 >>> (Permission.USER_READ | Permission.USER_WRITE) ^ Permission.USER_WRITE
 &lt;Permission=USER_READ>
