diff --git a/twisted/test/test_failure.py b/twisted/test/test_failure.py
index b31150b..be51ff4 100644
--- a/twisted/test/test_failure.py
+++ b/twisted/test/test_failure.py
@@ -1,3 +1,4 @@
+# -*- coding: utf-8 -*-
 # Copyright (c) Twisted Matrix Laboratories.
 # See LICENSE for details.
 
@@ -410,6 +411,17 @@ class FailureTests(SynchronousTestCase):
         self.assertDetailedTraceback(captureVars=True, cleanFailure=True)
 
 
+    def test_printTracebackUnicode(self):
+        unicode_text =  u'Ӭ  ӭ  Ӯ  ӯ  Ӱ  ӱ  Ӳ  ӳ  Ӵ  ӵ  Ӷ'
+        try:
+            raise Exception(unicode_text)
+        except Exception:
+            f = failure.Failure()
+        buf = NativeStringIO()
+        f.printTraceback(buf)
+        self.assertIn(unicode_text, buf.getvalue())
+
+
     def test_invalidFormatFramesDetail(self):
         """
         L{failure.format_frames} raises a L{ValueError} if the supplied
