Index: twisted/application/app.py
===================================================================
--- twisted/application/app.py	(revision 22956)
+++ twisted/application/app.py	(working copy)
@@ -2,7 +2,7 @@
 # Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
-import sys, os, pdb, getpass, traceback, signal, warnings
+import sys, os, pdb, getpass, traceback, warnings
 
 from twisted.python import runtime, log, usage, failure, util, logfile
 from twisted.persisted import sob
@@ -261,6 +261,7 @@
             sys.stdout = oldstdout
             sys.stderr = oldstderr
             if runtime.platformType == 'posix':
+                import signal
                 signal.signal(signal.SIGUSR2, lambda *args: pdb.set_trace())
                 signal.signal(signal.SIGINT, lambda *args: pdb.set_trace())
             fixPdb()
