--- smtp.py	2011-06-30 21:34:01.821961994 +0200
+++ smtp.py.patch.py	2011-07-01 01:14:19.266016196 +0200
@@ -1736,6 +1736,11 @@
 from twisted.mail.imap4 import LOGINCredentials as lcredentials
 
 class LOGINCredentials(lcredentials):
+    """
+    LOGINCredentials inherits from imap4.LOGINCredentials and 
+    is used to set credentials to authenticate the sender
+    """
+
     def __init__(self):
         lcredentials.__init__(self)
         self.challenges = ['Password:','Username:']
