--- test_smtp.py	2011-06-29 01:14:35.481479052 +0200
+++ test_smtp.py.patch.py	2011-07-01 01:11:58.278528691 +0200
@@ -639,12 +639,7 @@
         p = cred.portal.Portal(realm)
         p.registerChecker(DummyChecker())
 
-        class WeirdLOGIN(imap4.LOGINCredentials):
-            def __init__(self):
-                imap4.LOGINCredentials.__init__(self)
-                self.challenges[1] = 'Username:'
-
-        server = DummyESMTP({'LOGIN': WeirdLOGIN})
+        server = DummyESMTP({'LOGIN': smtp.LOGINCredentials})
         server.portal = p
 
         client = MyESMTPClient('testpassword')
