--- smtp.py	2011-06-29 01:14:36.233491447 +0200
+++ smtp.py.patch.py	2011-06-30 19:13:40.349798492 +0200
@@ -1733,6 +1733,12 @@
 
 from twisted.mail.imap4 import IClientAuthentication
 from twisted.mail.imap4 import CramMD5ClientAuthenticator, LOGINAuthenticator
+from twisted.mail.imap4 import LOGINCredentials as lcredentials
+
+class LOGINCredentials(lcredentials):
+    def __init__(self):
+        lcredentials.__init__(self)
+        self.challenges = ['Password:','Username:']
 
 class PLAINAuthenticator:
     implements(IClientAuthentication)
