Index: doc/core/howto/async.xhtml
===================================================================
--- doc/core/howto/async.xhtml	(revision 26633)
+++ doc/core/howto/async.xhtml	(working copy)
@@ -211,8 +211,7 @@
     print "The Deferred has called printContents with the following contents:"
     print contents
 
-    # Stop the Twisted event handling system -- this is usually handled
-    # in higher level ways
+    # Stop the Twisted event handling system using the reactor API
     reactor.stop()
 
 # call getPage, which returns immediately with a Deferred, promising to
@@ -223,8 +222,7 @@
 # the page content has been downloaded
 deferred.addCallback(printContents)
 
-# Begin the Twisted event handling system to manage the process -- again this
-# isn't the usual way to do this
+# Begin the Twisted event handling system to manage the process using the reactor API
 reactor.run()
 </pre>
 
