Index: twisted/python/dist.py
===================================================================
--- twisted/python/dist.py	(revision 37872)
+++ twisted/python/dist.py	(working copy)
@@ -367,8 +367,13 @@
             self.define_macros = [("WIN32", 1)]
         else:
             self.define_macros = []
+
+        if sys.platform.startswith('sunos'):
+            self.define_macros.append(('_XPG4_2', 1))
+
         self.extensions = [x for x in self.conditionalExtensions
                            if x.condition(self)]
+
         for ext in self.extensions:
             ext.define_macros.extend(self.define_macros)
 
