Index: twisted/application/app.py
===================================================================
RCS file: /cvs/Twisted/twisted/application/app.py,v
retrieving revision 1.5
diff -u -r1.5 app.py
--- twisted/application/app.py	17 Oct 2003 13:57:28 -0000	1.5
+++ twisted/application/app.py	7 Nov 2003 02:27:19 -0000
@@ -90,6 +90,15 @@
         log.msg("Loaded.")
     except Exception, e:
         s = "Failed to load application: %s" % e
+        if isinstance(e, KeyError):
+            s += """
+Could not find 'application' in the file. To use 'twistd -y', your .tac
+file must create a suitable object (e.g., by calling service.Application())
+and store it in a variable named 'application'. twistd loads your .tac file
+and scans the global variables for one of this name.
+
+Please read the 'Using Application' HOWTO for details.
+"""
         traceback.print_exc(file=log.logfile)
         log.msg(s)
         log.deferr()
