Index: twisted/conch/scripts/cftp.py
===================================================================
--- twisted/conch/scripts/cftp.py	(revision 26656)
+++ twisted/conch/scripts/cftp.py	(working copy)
@@ -458,14 +458,14 @@
         chunks = []
         startTime = time.time()
         for i in range(numRequests):
-            d = self._cbPutWrite(None, rf, lf, chunks, startTime)
-            if d:
-                dList.append(d)
+            d = defer.Deferred()
+            self._cbPutWrite(None, rf, lf, chunks, startTime, d)
+            dList.append(d)
         dl = defer.DeferredList(dList, fireOnOneErrback=1)
         dl.addCallback(self._cbPutDone, rf, lf)
         return dl
 
-    def _cbPutWrite(self, ignored, rf, lf, chunks, startTime):
+    def _cbPutWrite(self, ignored, rf, lf, chunks, startTime, deferred):
         chunk = self._getNextChunk(chunks)
         start, size = chunk
         lf.seek(start)
@@ -475,10 +475,10 @@
             self._printProgessBar(lf, startTime)
         if data:
             d = rf.writeChunk(start, data)
-            d.addCallback(self._cbPutWrite, rf, lf, chunks, startTime)
-            return d
+            d.addCallback(self._cbPutWrite, rf, lf, chunks, startTime, deferred)
+            d.addErrback(deferred.errback)
         else:
-            return
+            deferred.callback(None)
 
     def _cbPutDone(self, ignored, rf, lf):
         lf.close()
