=== modified file 'twisted/web/test/test_http.py'
--- twisted/web/test/test_http.py	2013-04-06 15:07:00 +0000
+++ twisted/web/test/test_http.py	2013-07-17 09:30:12 +0000
@@ -1699,6 +1699,53 @@
             'clientproto=(no clientproto yet)>' % (id(request),))
 
 
+    def test_unregisterNonQueuedNonStreamingProducer(self):
+        """
+        L{Request.unregisterProducer} unregisters a non-queued non-streaming
+        producer. It also unregisters the request's transport producer.
+        """
+        req = http.Request(DummyChannel(), None)
+        req.transport = StringTransport()
+        req.registerProducer(DummyProducer(), False)
+        req.unregisterProducer()
+        self.assertEqual(req.transport.producer, None)
+        self.assertEqual(req.producer, None)
+
+
+    def test_unregisterNonQueuedStreamingProducer(self):
+        """
+        L{Request.unregisterProducer} unregisters a non-queued streaming
+        producer. Its also unregisters the request's transport producer.
+        """
+        req = http.Request(DummyChannel(), None)
+        req.transport = StringTransport()
+        req.registerProducer(DummyProducer(), True)
+        req.unregisterProducer()
+        self.assertEqual(req.transport.producer, None)
+        self.assertEqual(req.producer, None)
+
+
+    def test_unregisterQueuedNonStreamingProducer(self):
+        """
+        L{Request.unregisterProducer} unregisters a queued non-streaming
+        producer.
+        """
+        req = http.Request(DummyChannel(), True)
+        req.registerProducer(DummyProducer(), False)
+        req.unregisterProducer()
+        self.assertEqual(req.producer, None)
+
+
+    def test_unregisterQueuedStreamingProducer(self):
+        """
+        L{Request.unregisterProducer} unregisters a queued streaming producer.
+        """
+        req = http.Request(DummyChannel(), True)
+        req.registerProducer(DummyProducer(), True)
+        req.unregisterProducer()
+        self.assertEqual(req.producer, None)
+
+
 
 class MultilineHeadersTestCase(unittest.TestCase):
     """

=== added file 'twisted/web/topfiles/6121.misc'
