diff --git a/twisted/internet/address.py b/twisted/internet/address.py
index f017f55..755ffb4 100644
--- a/twisted/internet/address.py
+++ b/twisted/internet/address.py
@@ -59,14 +59,6 @@ class IPv4Address(_IPAddress):
     @type host: C{str}
     """
 
-    def __init__(self, type, host, port, _bwHack=None):
-        _IPAddress.__init__(self, type, host, port)
-        if _bwHack is not None:
-            warnings.warn("twisted.internet.address.IPv4Address._bwHack "
-                          "is deprecated since Twisted 11.0",
-                          DeprecationWarning, stacklevel=2)
-
-
 
 class IPv6Address(_IPAddress):
     """
@@ -126,11 +118,8 @@ class UNIXAddress(FancyEqMixin, object):
 
     compareAttributes = ('name', )
 
-    def __init__(self, name, _bwHack = None):
+    def __init__(self, name):
         self.name = name
-        if _bwHack is not None:
-            warnings.warn("twisted.internet.address.UNIXAddress._bwHack is deprecated since Twisted 11.0",
-                    DeprecationWarning, stacklevel=2)
 
 
     if getattr(os.path, 'samefile', None) is not None:
diff --git a/twisted/internet/test/test_address.py b/twisted/internet/test/test_address.py
index 9f212a7..fd805b2 100644
--- a/twisted/internet/test/test_address.py
+++ b/twisted/internet/test/test_address.py
@@ -159,20 +159,6 @@ class IPv4AddressTCPTestCase(unittest.SynchronousTestCase,
         return IPv4Address("TCP", "127.0.0.2", 0)
 
 
-    def test_bwHackDeprecation(self):
-        """
-        If a value is passed for the C{_bwHack} parameter to L{IPv4Address},
-        a deprecation warning is emitted.
-        """
-        # Construct this for warning side-effects, disregard the actual object.
-        IPv4Address("TCP", "127.0.0.3", 0, _bwHack="TCP")
-
-        message = (
-            "twisted.internet.address.IPv4Address._bwHack is deprecated "
-            "since Twisted 11.0")
-        return self.assertDeprecations(self.test_bwHackDeprecation, message)
-
-
 
 class IPv4AddressUDPTestCase(unittest.SynchronousTestCase,
                              IPv4AddressTestCaseMixin):
@@ -192,20 +178,6 @@ class IPv4AddressUDPTestCase(unittest.SynchronousTestCase,
         return IPv4Address("UDP", "127.0.0.2", 0)
 
 
-    def test_bwHackDeprecation(self):
-        """
-        If a value is passed for the C{_bwHack} parameter to L{IPv4Address},
-        a deprecation warning is emitted.
-        """
-        # Construct this for warning side-effects, disregard the actual object.
-        IPv4Address("UDP", "127.0.0.3", 0, _bwHack="UDP")
-
-        message = (
-            "twisted.internet.address.IPv4Address._bwHack is deprecated "
-            "since Twisted 11.0")
-        return self.assertDeprecations(self.test_bwHackDeprecation, message)
-
-
 
 class IPv6AddressTestCase(unittest.SynchronousTestCase, AddressTestCaseMixin):
     addressArgSpec = (("type", "%s"), ("host", "%r"), ("port", "%d"))
@@ -276,20 +248,6 @@ class UNIXAddressTestCase(unittest.SynchronousTestCase, AddressTestCaseMixin):
     test_hashOfLinkedFiles.skip = symlinkSkip
 
 
-    def test_bwHackDeprecation(self):
-        """
-        If a value is passed for the C{_bwHack} parameter to L{UNIXAddress},
-        a deprecation warning is emitted.
-        """
-        # Construct this for warning side-effects, disregard the actual object.
-        UNIXAddress(self.mktemp(), _bwHack='UNIX')
-
-        message = (
-            "twisted.internet.address.UNIXAddress._bwHack is deprecated "
-            "since Twisted 11.0")
-        return self.assertDeprecations(self.test_bwHackDeprecation, message)
-
-
 
 class EmptyUNIXAddressTestCase(unittest.SynchronousTestCase,
                                AddressTestCaseMixin):
diff --git a/twisted/topfiles/7050.misc b/twisted/topfiles/7050.misc
new file mode 100644
index 0000000..e69de29
