Index: doc/core/howto/components.xhtml
===================================================================
--- doc/core/howto/components.xhtml	(revision 31535)
+++ doc/core/howto/components.xhtml	(working copy)
@@ -327,7 +327,8 @@
 
 <p>As you can see, the <code>AmericanSocket</code> instance claims to
 implement <code>IAmericanSocket</code>, but the <code>UKSocket</code>
-does not. If we wanted to use the <code>HairDryer</code> with the <code>AmericanSocket</code>, we could know that it would be safe to do so by
+does not. If we wanted to use the <code>HairDryer</code> with the
+<code>AmericanSocket</code>, we could know that it would be safe to do so by
 checking whether it implements <code>IAmericanSocket</code>. However, if we
 decide we want to use <code>HairDryer</code> with a <code>UKSocket</code>
 instance, we must <em>adapt</em> it to <code>IAmericanSocket</code> before
@@ -342,7 +343,8 @@
 looks in the adapter registry for an adapter which implements the interface for
 the given instance's class. If it finds one, it constructs an instance of the
 Adapter class, passing the constructor the original instance, and returns it.
-Now the <code>HairDryer</code> can safely be used with the adapted  <code>UKSocket</code>. But what happens if we attempt to adapt an object
+Now the <code>HairDryer</code> can safely be used with the adapted 
+<code>UKSocket</code>. But what happens if we attempt to adapt an object
 which already implements <code>IAmericanSocket</code>? We simply get back the
 original instance:</p>
 
@@ -432,7 +434,8 @@
 True
 </pre>
 
-<p>If you want <code>MyThing</code> to inherit from <code>pb.Root</code> but <em>not</em> implement <code>IPBRoot</code> like <code>pb.Root</code> does,
+<p>If you want <code>MyThing</code> to inherit from <code>pb.Root</code>
+but <em>not</em> implement <code>IPBRoot</code> like <code>pb.Root</code> does,
 use <code>implementOnly</code>:</p>
 
 <pre class="python">
Index: doc/core/howto/defer.xhtml
===================================================================
--- doc/core/howto/defer.xhtml	(revision 31535)
+++ doc/core/howto/defer.xhtml	(working copy)
@@ -44,7 +44,8 @@
 available (this series of functions is known as a series of 
 <strong>callbacks</strong>, or a <strong>callback chain</strong>), together
 with a series of functions to be called if there is an error in the
-asychronous request (known as a series of <strong>errbacks</strong> or an  <strong>errback chain</strong>). The asychronous library code calls the first
+asychronous request (known as a series of <strong>errbacks</strong> or an
+<strong>errback chain</strong>). The asychronous library code calls the first
 callback when the result is available, or the first errback when an error
 occurs, and the <code>Deferred</code> object then hands the results of each
 callback or errback function to the next function in the chain.</p> 
@@ -390,7 +391,8 @@
     return d
 </pre>
 
-<p> Our original implementation of <code>authenticateUser</code> expected  <code>isValidUser</code> to be synchronous, but now we need to change it to handle both
+<p> Our original implementation of <code>authenticateUser</code> expected
+<code>isValidUser</code> to be synchronous, but now we need to change it to handle both
 synchronous and asynchronous implementations of <code>isValidUser</code>. For this, we
 use <code class="API" base="twisted.internet.defer">maybeDeferred</code> to
 call <code>isValidUser</code>, ensuring that the result of <code>isValidUser</code> is a Deferred,
@@ -434,7 +436,8 @@
 dl = defer.DeferredList([deferred1, deferred2, deferred3])
 </pre>
 
-<p>You can now treat the DeferredList like an ordinary Deferred; you can call  <code>addCallbacks</code> and so on.  The DeferredList will call its callback
+<p>You can now treat the DeferredList like an ordinary Deferred; you can call
+<code>addCallbacks</code> and so on.  The DeferredList will call its callback
 when all the deferreds have completed.  The callback will be called with a list
 of the results of the Deferreds it contains, like so:</p>
 
@@ -548,7 +551,8 @@
 class="footnote">Unless of course a later callback starts a fresh error &mdash;
 but as we've already noted, adding callbacks to a Deferred after its used in a
 DeferredList is confusing and usually avoided.</span>.  Passing a true value
-for the <code>consumeErrors</code> parameter will not change the behavior of <code>fireOnOneCallback</code> or <code>fireOnOneErrback</code>.</p>
+for the <code>consumeErrors</code> parameter will not change the behavior of
+<code>fireOnOneCallback</code> or <code>fireOnOneErrback</code>.</p>
 
 <a name="class"></a>
 
@@ -559,7 +563,8 @@
 substitute for the docstrings in the Deferred class, but can provide guidelines
 for its use.</p>
 
-<p>There is a parallel overview of functions used by the Deferred's  <em>creator</em> in <a href="gendefer.xhtml#class">Generating Deferreds</a>.</p>
+<p>There is a parallel overview of functions used by the Deferred's
+<em>creator</em> in <a href="gendefer.xhtml#class">Generating Deferreds</a>.</p>
 
 <h3>Basic Callback Functions</h3>
 
Index: doc/core/howto/pb-intro.xhtml
===================================================================
--- doc/core/howto/pb-intro.xhtml	(revision 31535)
+++ doc/core/howto/pb-intro.xhtml	(working copy)
@@ -1,7 +1,9 @@
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
+
 <html xmlns="http://www.w3.org/1999/xhtml">
 <head>
 <title>Introduction to Perspective Broker</title>
-<style src="stylesheet-unprocessed.css"></style>
+<link href="stylesheet-unprocessed.css" rel="stylesheet" type="text/css"/>
 </head>
 
 <body>
@@ -141,7 +143,7 @@
 flavors has a rule for how the <code class="python">callRemote</code>
 message is transformed into a local method call on the server.  In
 order to use one of these <q>flavors</q>, subclass them and name your
-published methods with the appropriate prefix.
+published methods with the appropriate prefix.</p>
 
 <ul>
   <li><code class="API">twisted.spread.pb.IPerspective</code> implementors
@@ -198,7 +200,6 @@
 
 </ul>
 
-</p>
 
 <h2>Things you can Copy Remotely</h2>
 
@@ -211,7 +212,6 @@
  <code class="python">__getstate__</code> or specialized method calls for that
 flavor.</p>
 
-<p>
 <ul>
   <li><code class="API">twisted.spread.pb.Copyable</code>
 
@@ -232,7 +232,7 @@
   actually be copied, you must declare a <code class="python">RemoteCopy</code>
   handler for
   that Copyable subclass.  The easiest way to do this is to declare both in the
-  same module, like so:
+  same module, like so:</p>
 
   <pre class="python">
 from twisted.spread import flavors
@@ -243,7 +243,7 @@
 flavors.setUnjellyableForClass(Foo, RemoteFoo)
   </pre>
 
-  In this case, each time a Foo is copied between peers, a RemoteFoo will be
+  <p>In this case, each time a Foo is copied between peers, a RemoteFoo will be
   instantiated and populated with the Foo's state.  If you do not do this, PB
   will complain that there have been security violations, and it may close the
   connection.
@@ -308,6 +308,5 @@
   Any further calls to this observer object will be invalid.</p>
   </li>
 </ul>
-</p>
 
 </body></html>
Index: doc/core/howto/deferredindepth.xhtml
===================================================================
--- doc/core/howto/deferredindepth.xhtml	(revision 31535)
+++ doc/core/howto/deferredindepth.xhtml	(working copy)
@@ -254,7 +254,8 @@
 </pre>
 
 <p>Two things to note here. First, &quot;- A -&quot; was skipped, like we wanted it to,
-and the second thing is that after &quot;- A -&quot;, noDecision is called, because  <strong>it is the next errback that exists in the chain</strong>. It returns a
+and the second thing is that after &quot;- A -&quot;, noDecision is called, because
+<strong>it is the next errback that exists in the chain</strong>. It returns a
 non-failure, so processing continues with the next callback at &quot;- B -&quot;, and
 the errback at the end of the chain is never called </p>
 
Index: doc/core/howto/pb-copyable.xhtml
===================================================================
--- doc/core/howto/pb-copyable.xhtml	(revision 31535)
+++ doc/core/howto/pb-copyable.xhtml	(working copy)
@@ -1,7 +1,8 @@
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
 <html xmlns="http://www.w3.org/1999/xhtml">
 <head>
 <title>PB Copyable: Passing Complex Types</title>
-<style src="stylesheet-unprocessed.css"></style>
+<link href="stylesheet-unprocessed.css" rel="stylesheet" type="text/css"/>
 </head>
 
 <body>
Index: doc/core/howto/pb.xhtml
===================================================================
--- doc/core/howto/pb.xhtml	(revision 31535)
+++ doc/core/howto/pb.xhtml	(working copy)
@@ -2,7 +2,7 @@
 <html xmlns="http://www.w3.org/1999/xhtml"> <head>
 <title>Overview of Twisted Spread</title>
 
-<style src="stylesheet-unprocessed.css"></style>
+<link href="stylesheet-unprocessed.css" rel="stylesheet" type="text/css"/>
 
 </head>
 
Index: doc/core/howto/application.xhtml
===================================================================
--- doc/core/howto/application.xhtml	(revision 31535)
+++ doc/core/howto/application.xhtml	(working copy)
@@ -79,12 +79,12 @@
 <p>See <a href="servers.xhtml">Writing Servers</a> for an explanation of
 EchoFactory.</p>
 
-<p>This example creates a simple hierarchy:
+<p>This example creates a simple hierarchy:</p>
 <pre>
    application
    |
    `- echoService
-</pre> More complicated hierarchies of services can be created using
+</pre> <p>More complicated hierarchies of services can be created using
 IServiceCollection.  You will most likely want to do this to manage Services
 which are dependent on other Services.  For example, a proxying Twisted
 application might want its server Service to only start up after the associated
Index: doc/core/howto/pb-cred.xhtml
===================================================================
--- doc/core/howto/pb-cred.xhtml	(revision 31535)
+++ doc/core/howto/pb-cred.xhtml	(working copy)
@@ -7,7 +7,7 @@
 <head>
 <title>Authentication with Perspective Broker</title>
 
-<style src="stylesheet-unprocessed.css"></style>
+<link href="stylesheet-unprocessed.css" rel="stylesheet" type="text/css"/>
 
 </head>
 
@@ -729,7 +729,8 @@
 class behaves a lot like <code>Referenceable</code>: it turns into a 
  <code>RemoteReference</code> when sent over the wire, and certain methods
 can be invoked by the holder of that reference. However, the methods that
-can be called have names that start with <code>view_</code> instead of  <code>remote_</code>, and those methods are always called with an extra 
+can be called have names that start with <code>view_</code> instead of
+<code>remote_</code>, and those methods are always called with an extra 
 <code>perspective</code> argument that points to the Avatar through which
 the reference was sent:</p>
 
Index: doc/core/howto/gendefer.xhtml
===================================================================
--- doc/core/howto/gendefer.xhtml	(revision 31535)
+++ doc/core/howto/gendefer.xhtml	(working copy)
@@ -285,12 +285,14 @@
 
 <h3>Firing Deferreds more than once is impossible</h3>
 
-<p>Deferreds are one-shot. You can only call <code>Deferred.callback</code> or <code>Deferred.errback</code> once. The processing chain continues each time
+<p>Deferreds are one-shot. You can only call <code>Deferred.callback</code> or
+<code>Deferred.errback</code> once. The processing chain continues each time
 you add new callbacks to an already-called-back-to Deferred.</p>
 
 <h3>Synchronous callback execution</h3>
 
-<p>If a Deferred already has a result available, <code>addCallback</code> <strong>may</strong> call the callback synchronously: that is, immediately
+<p>If a Deferred already has a result available, <code>addCallback</code>
+<strong>may</strong> call the callback synchronously: that is, immediately
 after it's been added.  In situations where callbacks modify state, it is
 might be desirable for the chain of processing to halt until all callbacks are
 added. For this, it is possible to <code>pause</code> and <code>unpause</code>
@@ -298,7 +300,8 @@
 
 <p>Be careful when you use these methods! If you <code>pause</code> a
 Deferred, it is <em>your</em> responsibility to make sure that you unpause it.
-The function adding the callbacks must unpause a paused Deferred, it should <em>never</em> be the responsibility of the code that actually fires the
+The function adding the callbacks must unpause a paused Deferred, it should
+<em>never</em> be the responsibility of the code that actually fires the
 callback chain by calling <code>callback</code> or <code>errback</code> as
 this would negate its usefulness!</p>
 
Index: doc/core/howto/logging.xhtml
===================================================================
--- doc/core/howto/logging.xhtml	(revision 31535)
+++ doc/core/howto/logging.xhtml	(working copy)
@@ -94,23 +94,23 @@
     should be useful to you.
     </p>
 
-    <p>You just start it like any other observer:
+    <p>You just start it like any other observer:</p>
         <pre class="python">
 observer = log.PythonLoggingObserver()
 observer.start()
         </pre>
 
-    Then <a href="http://docs.python.org/library/logging.html">configure the
+    <p>Then <a href="http://docs.python.org/library/logging.html">configure the
     standard library logging module</a> to behave as you want.
     </p>
 
     <p>This method allows you to customize the log level received by the
-    standard library logging module using the <code>logLevel</code> keyword:
+    standard library logging module using the <code>logLevel</code> keyword:</p>
         <pre class="python">
 log.msg("This is important!", logLevel=logging.CRITICAL)
 log.msg("Don't mind", logLevel=logging.DEBUG)
         </pre>
-    Unless <code>logLevel</code> is provided, logging.INFO is used for <code>log.msg</code>
+    <p>Unless <code>logLevel</code> is provided, logging.INFO is used for <code>log.msg</code>
     and <code>logging.ERROR</code> is used for <code>log.err</code>.
     </p>
 
Index: doc/core/howto/pb-clients.xhtml
===================================================================
--- doc/core/howto/pb-clients.xhtml	(revision 31535)
+++ doc/core/howto/pb-clients.xhtml	(working copy)
@@ -1,3 +1,4 @@
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
 <html xmlns="http://www.w3.org/1999/xhtml">
 <head>
 <title>Managing Clients of Perspectives</title>
Index: doc/core/howto/pb-usage.xhtml
===================================================================
--- doc/core/howto/pb-usage.xhtml	(revision 31535)
+++ doc/core/howto/pb-usage.xhtml	(working copy)
@@ -1,7 +1,8 @@
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
 <html xmlns="http://www.w3.org/1999/xhtml">
 <head>
 <title>Using Perspective Broker</title>
-<style src="stylesheet-unprocessed.css"></style>
+<link href="stylesheet-unprocessed.css" rel="stylesheet" type="text/css"/>
 </head>
 
 <body>
@@ -393,15 +394,15 @@
 taken off the wire as a stream of bytes and turned back into an object: a
 living, breathing instance of some class): one reason is that it does not
 know which local class ought to be used to create an instance that
-corresponds to the remote object<span class="footnote"><p>The naive approach
+corresponds to the remote object<span class="footnote">The naive approach
 of simply doing <code>import SomeClass</code> to match a remote caller who
 claims to have an object of type <q>SomeClass</q> could have nasty consequences
 for some modules that do significant operations in their <code>__init__</code>
 methods (think <code>telnetlib.Telnet(host='localhost', port='chargen')</code>,
  or even more powerful classes that you have available in your server program).
 Allowing a remote entity to create arbitrary classes in your namespace is
-nearly equivalent to allowing them to run arbitrary code.</p>
-<p>The <code class="API" base="twisted.spread.jelly">InsecureJelly</code>
+nearly equivalent to allowing them to run arbitrary code.<br/><br/>
+The <code class="API" base="twisted.spread.jelly">InsecureJelly</code>
 exception arises because the class being sent over the wire has not been
 registered with the serialization layer (known as <code class="API"
 base="twisted.spread">jelly</code>). The easiest way to make it possible to
@@ -409,7 +410,7 @@
 class="API" base="twisted.spread">pb.Copyable</code>, and then to use 
  <code>setUnjellyableForClass(remoteClass, localClass)</code> on the
 receiving side. See <a href="pb-copyable.xhtml">Passing Complex Types</a>
-for an example.</p></span>.</p>
+for an example.</span>.</p>
 
 <p>The receiving end of the connection gets to decide what to accept and what
 to reject. It indicates its disapproval by raising a <code class="API"
Index: doc/core/howto/endpoints.xhtml
===================================================================
--- doc/core/howto/endpoints.xhtml	(revision 31535)
+++ doc/core/howto/endpoints.xhtml	(working copy)
@@ -125,12 +125,11 @@
 interface does not guarantee one. If you don't have any way for the
 application to cancel a wayward connection attempt, the attempt may just
 keep waiting forever.  For example, a very simple 30-second timeout could be
-implemented like this:
+implemented like this:</p>
 <pre class="python">
 attempt = myEndpoint.connect(myFactory)
 reactor.callback(30, attempt.cancel)
 </pre>
-</p>
 
 <h2>Maximizing the Return on your Endpoint Investment</h2>
 
Index: doc/core/howto/row.xhtml
===================================================================
--- doc/core/howto/row.xhtml	(revision 31535)
+++ doc/core/howto/row.xhtml	(working copy)
@@ -187,7 +187,8 @@
 class="python">loadObjectsFrom</code> is called for a table, it will
 automatically load all the children rows for the rows from the specified
 table. The child rows will be put into a list member variable of the
-rowObject instance with the name <code>childRows</code> or if a <em>containerMethod</em> is specified for the foreign key relationship,
+rowObject instance with the name <code>childRows</code> or if a
+<em>containerMethod</em> is specified for the foreign key relationship,
 that method will be called on the parent row object for each row that is
 being added to it as a child.</p>
 
Index: doc/core/howto/trial.xhtml
===================================================================
--- doc/core/howto/trial.xhtml	(revision 31535)
+++ doc/core/howto/trial.xhtml	(working copy)
@@ -38,12 +38,12 @@
 already familiar with unit test in Python, jump straight to the section
 specific to <a href="#twisted">testing Twisted code</a>.</p>
 
-<p><div class="note">In what follows we will make a series of refinements
+<div class="note"><p>In what follows we will make a series of refinements
 to some simple classes. In order to keep the examples and source code links
 complete and to allow you to run Trial on the intermediate results at every
 stage, I add <code>_N</code> (where the <code>N</code> are successive
 integers) to file names to keep them separate. This is a minor visual
-distraction that should be ignored.</div></p>
+distraction that should be ignored.</p></div>
 
 <h2>Creating an API and writing tests</h2>
 
@@ -77,7 +77,7 @@
 
 <a href="listings/trial/calculus/test/test_base_1.py" class="py-listing">test_base_1.py</a>
 
-<p>You should now have the following 4 files:
+<p>You should now have the following 4 files:</p>
 
 <pre class="shell">
     calculus/__init__.py
@@ -85,7 +85,6 @@
     calculus/test/__init__.py
     calculus/test/test_base_1.py
 </pre>
-</p>
 
 <p>To run the tests, there are two things you must get set up. Make sure
 you get these both done - nothing below will work unless you do.</p>
@@ -195,7 +194,7 @@
 reporters.</p>
 
 <p>
-The tests can be run by <code class="shell">trial</code> in multiple ways:
+The tests can be run by <code class="shell">trial</code> in multiple ways:</p>
 <ul>
   <li><code class="shell">trial calculus</code>: run all the tests for the
   calculus package.</li>
@@ -219,7 +218,7 @@
   specific test file (not recommended).</li>
 </ul>
 
-The first 3 versions using full qualified names are strongly encouraged: they
+<p>The first 3 versions using full qualified names are strongly encouraged: they
 are much more reliable and they allow you to easily be more selective in your
 test runs.
 </p>
@@ -317,13 +316,13 @@
 
 <p>Here the <code class="python">_make_ints</code> helper function tries to
 convert a list into a list of equivalent integers, and raises a <code
-class="python">TypeError</code> in case the conversion goes wrong.
+class="python">TypeError</code> in case the conversion goes wrong.</p>
 
 <div class="note">The <code class="python">int</code> conversion can also
 raise a <code class="python">TypeError</code> if passed something of the
 wrong type, such as a list. We'll just let that exception go by as <code
 class="python">TypeError</code> is already what we want in case something
-goes wrong.</div></p>
+goes wrong.</div>
 
 
 <a name="twisted"></a>
@@ -607,7 +606,7 @@
 So the first step is adding a failing test.</p>
 
 <a href="listings/trial/calculus/test/test_client_3.py" class="py-listing">test_client_3.py</a>
-<p>What have we done here ?
+<p>What have we done here ?</p>
 <ul>
     <li>We switched to StringTransportWithDisconnection. This transport manages
     <code class="python">loseConnection</code> and forwards it to its protocol.</li>
@@ -615,7 +614,6 @@
    </code> attribute.</li>
     <li>We check that after a timeout our connection has closed.</li>
 </ul>
-</p>
 
 <p>For doing that, we then use the <code class="python">TimeoutMixin</code>
 class, that does almost everything we want. The great thing is that it almost
@@ -647,7 +645,7 @@
 
 <h2>Conclusion</h2>
 
-<p>So what did you learn in this document?
+<p>So what did you learn in this document?</p>
 <ul>
     <li>How to use the trial command-line tool to run your tests</li>
     <li>How to use string transports to test individual clients and servers
@@ -659,7 +657,6 @@
 If one of the topics still looks cloudy to you, please give us your feedback!
 You can file tickets to improve this document
 <a href="http://twistedmatrix.com/">on the Twisted web site</a>.
-</p>
 
 </body>
 </html>
Index: doc/core/howto/process.xhtml
===================================================================
--- doc/core/howto/process.xhtml	(revision 31535)
+++ doc/core/howto/process.xhtml	(working copy)
@@ -14,13 +14,12 @@
 
 <p>Along with connection to servers across the internet, Twisted also
 connects to local processes with much the same API. The API is described in
-more detail in the documentation of:
+more detail in the documentation of:</p>
 <ul>
 <li><code class="API">twisted.internet.interfaces.IReactorProcess</code></li>
 <li><code class="API">twisted.internet.interfaces.IProcessTransport</code></li>
 <li><code class="API">twisted.internet.interfaces.IProcessProtocol</code></li>
 </ul>
-</p>
 
     <h2>Running Another Process</h2>
 
Index: doc/core/howto/upgrading.xhtml
===================================================================
--- doc/core/howto/upgrading.xhtml	(revision 31535)
+++ doc/core/howto/upgrading.xhtml	(working copy)
@@ -299,7 +299,8 @@
 mixed with <code class="API"
 base="twisted.persisted.styles">Versioned</code>. <code>rebuild</code> does
 not run any of the classes' methods, whereas <code>Versioned</code> works by
-running <code>__setstate__</code> during the load process and  <code>doUpgrade</code> afterwards. This means <code>rebuild</code> can only
+running <code>__setstate__</code> during the load process and
+<code>doUpgrade</code> afterwards. This means <code>rebuild</code> can only
 be used to process upgrades that do not change the data attributes of any of
 the involved classes. Any time attributes are added or removed, the program
 must be shut down, persisted, and restarted, with <code>upgradeToVersionNN</code> methods
Index: doc/core/development/policy/test-standard.xhtml
===================================================================
--- doc/core/development/policy/test-standard.xhtml	(revision 31535)
+++ doc/core/development/policy/test-standard.xhtml	(working copy)
@@ -302,7 +302,7 @@
 well as <code>twisted-dev.el</code>'s F9 command) use this to automatically
 run the right tests.</p>
 
-<h2 id="links">Links</h2><a name="links"></a>
+<h2 id="links">Links</h2>
 
 <ul>
       <li>A chapter on <a href=
Index: doc/core/development/policy/coding-standard.xhtml
===================================================================
--- doc/core/development/policy/coding-standard.xhtml	(revision 31535)
+++ doc/core/development/policy/coding-standard.xhtml	(working copy)
@@ -156,11 +156,11 @@
     that module.</p>
 
     <p>Depending on the situation, it is acceptable to have imports that
-    look like this:
+    look like this:</p>
     <pre class="python">from twisted.internet.defer import Deferred</pre>
-    or like this:
+    <p>or like this:</p>
     <pre class="python">from twisted.internet import defer</pre>
-    That is, modules should import <em>modules</em> or <em>classes and
+    <p>That is, modules should import <em>modules</em> or <em>classes and
     functions</em>, but not <em>packages</em>.</p>
 
     <p>Wildcard import syntax may not be used by code in Twisted.  These
@@ -177,8 +177,8 @@
     task of refactoring more difficult.</p>
 
     <p>In case of local names conflicts due to import, use the <code>as</code>
-    syntax, for example:
-    <pre class="python">from twisted.trial import util as trial_util</pre></p>
+    syntax, for example:</p>
+    <pre class="python">from twisted.trial import util as trial_util</pre>
 
     <p>The encoding must always be ASCII, so no coding cookie is necessary.</p>
 
@@ -257,7 +257,7 @@
     are documenting.</p>
 
     <p>Docstrings should be triple-quoted. The opening and the closing of the
-    docstrings should be on a line by themselves. For example:
+    docstrings should be on a line by themselves. For example:</p>
 <pre class="python">
 class Ninja(object):
     """
@@ -268,7 +268,6 @@
         Attack C{someone} with this L{Ninja}'s shuriken.
         """
 </pre>
-    </p>
 
     <p>Docstrings should be written in epytext format; more
     documentation is available in the
@@ -347,7 +346,7 @@
       for the Conch sub-project is in <code>bin/conch/ckeygen</code>.
 <pre class="python">
 #!/usr/bin/env python
-</pre></li>
+</pre>
 
       <p>To make sure that the script is portable across different UNIX like
       operating systems we use the <code>/usr/bin/env</code> command. The env
@@ -357,7 +356,7 @@
       foolproof method. Always make sure that <code>/usr/bin/env</code> exists or
       use a softlink/symbolic link to point it to the correct path. Python's
       distutils will rewrite the shebang line upon installation so this policy
-      only covers the source files in version control.</p>
+      only covers the source files in version control.</p></li>
 
       <li>Add the Twisted running-from-SVN header:
 <pre class="python">
@@ -402,7 +401,7 @@
     a Python version, place the import statement inside a try/except block,
     and import the Python version if the import fails.  This allows code to
     work on platforms where the extension version is not available.  For
-    example:
+    example:</p>
 
 <pre class="python">
 try:
@@ -411,7 +410,7 @@
     import pickle
 </pre>
 
-    Use the &quot;as&quot; syntax of the import statement as well, to set
+    <p>Use the &quot;as&quot; syntax of the import statement as well, to set
     the name of the extension module to the name of the Python module.</p>
 
     <p>Some modules don't exist across all supported Python versions. For
