Index: plugin.py
===================================================================
--- plugin.py	(revision 15717)
+++ plugin.py	(working copy)
@@ -183,7 +183,12 @@
                 stage = 'closing'
                 f.close()
                 stage = 'renaming'
-                os.rename(tmpCacheFile, dropcache)
+                try:
+                    os.rename(tmpCacheFile, dropcache)
+                except OSError, e:
+                    # Windows won't allow an existing file to be renamed
+                    if e.errno == errno.EEXIST:
+                        os.remove(dropcache)
             except (OSError, IOError), e:
                 # A large number of errors can occur here.  There's nothing we
                 # can really do about any of them, but they are also non-fatal
