Index: twisted/plugin.py
===================================================================
--- twisted/plugin.py	(revision 15840)
+++ twisted/plugin.py	(working copy)
@@ -183,7 +183,17 @@
                 stage = 'closing'
                 f.close()
                 stage = 'renaming'
-                os.rename(tmpCacheFile, dropcache)
+                try:
+                    os.rename(tmpCacheFile, dropcache)
+                except OSError, e:
+                    # Windows won't allow an existing file to be renamed
+                    if e.errno == errno.EEXIST:
+                        stage = '(win32 special-case) removing'
+                        os.remove(dropcache)
+                        stage = '(win32 special-case) renaming'
+                        os.rename(tmpCacheFile, dropcache)
+                    else:
+                        raise
             except (OSError, IOError), e:
                 # A large number of errors can occur here.  There's nothing we
                 # can really do about any of them, but they are also non-fatal
@@ -192,6 +202,7 @@
                 # had not occurred.
                 log.msg("Error %s plugin cache file %r (%r): %r" % (
                     stage, tmpCacheFile, dropcache, os.strerror(e.errno)))
+                log.err()
 
     return topcache
 
