diff --git a/twisted/python/dist.py b/twisted/python/dist.py
index 9566039..3fbb8b0 100644
--- a/twisted/python/dist.py
+++ b/twisted/python/dist.py
@@ -62,7 +62,8 @@ class ConditionalExtension(Extension):
         things about the platform.
     """
     def __init__(self, *args, **kwargs):
-        self.condition = kwargs.pop("condition", lambda builder: True)
+        condition = kwargs.pop("condition", lambda builder: True)
+        self.condition = lambda builder: builder._check_header("Python.h") and condition(builder)
         Extension.__init__(self, *args, **kwargs)
 
 
