*** proxy.py.old	Thu Jul 24 14:56:06 2003
--- web/proxy.py	Thu Jul 24 14:56:31 2003
***************
*** 131,137 ****
  
          from twisted.protocols import http
          f = http.HTTPFactory()
!         f.HTTPChannel = Proxy
  
      Make the HTTPFactory a listener on a port as per usual, and you have
      a fully-functioning web proxy!
--- 131,137 ----
  
          from twisted.protocols import http
          f = http.HTTPFactory()
!         f.protocol = Proxy
  
      Make the HTTPFactory a listener on a port as per usual, and you have
      a fully-functioning web proxy!
