Index: twisted/web/__init__.py
===================================================================
--- twisted/web/__init__.py	(revision 24269)
+++ twisted/web/__init__.py	(working copy)
@@ -1,12 +1,10 @@
-
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
 """
-
-Twisted Web: a Twisted Web Server.
-
+Twisted Web: a Twisted Web Server that also provides
+basic HTTP L{client<twisted.web.client>} support.
 """
 
 from twisted.web._version import version
