Index: twisted/application/app.py
===================================================================
--- twisted/application/app.py	(revision 17783)
+++ twisted/application/app.py	(working copy)
@@ -21,6 +21,7 @@
     'win32': 'twisted.internet.win32eventreactor',
     'win': 'twisted.internet.win32eventreactor',
     'poll': 'twisted.internet.pollreactor',
+    'epoll': 'twisted.internet.epollreactor',
     'qt': 'twisted.internet.qtreactor',
     'cf' : 'twisted.internet.cfreactor',
     'kqueue': 'twisted.internet.kqreactor',
