Index: twisted/topfiles/setup.py
===================================================================
--- twisted/topfiles/setup.py	(revision 17783)
+++ twisted/topfiles/setup.py	(working copy)
@@ -58,6 +58,10 @@
                                 ["twisted/internet/iocpreactor/_iocp.c"],
                                 libraries=["ws2_32", "mswsock"],
                                 define_macros=builder.define_macros))
+    if sys.platform == 'linux2':
+        exts.append( Extension("twisted.python._epoll",
+                                ["twisted/python/_epoll.c"],
+                                define_macros=builder.define_macros))
 
     return exts
 
