Index: twisted/web/_newclient.py
===================================================================
--- twisted/web/_newclient.py	(revision 47025)
+++ twisted/web/_newclient.py	(working copy)
@@ -919,6 +919,14 @@
 
 
 
+# This proxy class is used only in the private constructor of the Response
+# class below, in order to prevent users relying on any property of the
+# concrete request object: they can only use what is provided by
+# IClientRequest.
+_ClientRequestProxy = proxyForInterface(IClientRequest)
+
+
+
 @implementer(IResponse)
 class Response:
     """
@@ -1012,7 +1020,7 @@
         @return: L{Response} instance.
         """
         response = Response(version, code, phrase, headers, _transport)
-        response.request = proxyForInterface(IClientRequest)(request)
+        response.request = _ClientRequestProxy(request)
         return response
 
 
