From 953b53234a481e210ed724ac4051181b31e3ba9a Mon Sep 17 00:00:00 2001
From: Torste Aikio <torste.aikio@f-secure.com>
Date: Mon, 18 Feb 2019 16:41:34 +0200
Subject: [PATCH] Use None as default value for DisconnectSelectable

---
 src/twisted/internet/posixbase.py | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/src/twisted/internet/posixbase.py b/src/twisted/internet/posixbase.py
index 1b0e50964..cd0d8e277 100644
--- a/src/twisted/internet/posixbase.py
+++ b/src/twisted/internet/posixbase.py
@@ -234,16 +234,18 @@ class _DisconnectSelectableMixin(object):
     Mixin providing the C{_disconnectSelectable} method.
     """
 
-    def _disconnectSelectable(self, selectable, why, isRead, faildict={
-        error.ConnectionDone: failure.Failure(error.ConnectionDone()),
-        error.ConnectionLost: failure.Failure(error.ConnectionLost())
-        }):
+    def _disconnectSelectable(self, selectable, why, isRead, faildict=None):
         """
         Utility function for disconnecting a selectable.
 
         Supports half-close notification, isRead should be boolean indicating
         whether error resulted from doRead().
         """
+        if faildict is None:
+            faildict = {
+                error.ConnectionDone: failure.Failure(error.ConnectionDone()),
+                error.ConnectionLost: failure.Failure(error.ConnectionLost())
+            }
         self.removeReader(selectable)
         f = faildict.get(why.__class__)
         if f:
-- 
2.11.0

