diff --git twisted/scripts/trial.py twisted/scripts/trial.py
index 7cdaa3c..38f8a76 100644
--- twisted/scripts/trial.py
+++ twisted/scripts/trial.py
@@ -3,8 +3,8 @@
 # Copyright (c) Twisted Matrix Laboratories.
 # See LICENSE for details.
 
+from __future__ import print_function,division,absolute_import
 
-from __future__ import print_function
 import gc
 import inspect
 import os
@@ -599,8 +599,8 @@ def run():
     config = Options()
     try:
         config.parseOptions()
-    except usage.error, ue:
-        raise SystemExit, "%s: %s" % (sys.argv[0], ue)
+    except usage.error as ue:
+        raise SystemExit("%s: %s" % (sys.argv[0], ue))
     _initialDebugSetup(config)
 
     try:
