diff --git a/endpoints.py b/cygdrive/c/Python27/lib/site-packages/twisted/internet/endpoints.py
old mode 100644
new mode 100755
index 7cb6723..ca2f770
--- a/endpoints.py
+++ b/cygdrive/c/Python27/lib/site-packages/twisted/internet/endpoints.py
@@ -141,7 +141,26 @@ class _WrappingFactory(ClientFactory):
         Errback the C{self._onConnection} L{Deferred} when the
         client connection fails.
         """
-        self._onConnection.errback(reason)
+        try:
+            self._wrappedFactory.clientConnectionFailed(connector, reason)
+            
+            if not self._onConnection.called:
+                self._onConnection.errback(reason)
+        except:
+            pass
+        
+        
+    def clientConnectionLost(self, connector, reason):
+        """Called when an established connection is lost.
+
+        It may be useful to call connector.connect() - this will reconnect.
+
+        @type reason: L{twisted.python.failure.Failure}
+        """
+        try:
+            self._wrappedFactory.clientConnectionLost(connector, reason)
+        except:
+            pass
 
 
 
