diff --git twisted/lore/indexer.py twisted/lore/indexer.py
index 528e7d6..60e9b7b 100644
--- twisted/lore/indexer.py
+++ twisted/lore/indexer.py
@@ -1,6 +1,9 @@
 # Copyright (c) Twisted Matrix Laboratories.
 # See LICENSE for details.
 
+indexFilename = None
+entries = None
+
 
 def setIndexFilename(filename='index.xhtml'):
     global indexFilename
diff --git twisted/lore/man2lore.py twisted/lore/man2lore.py
index fbcba1c..b9ecca3 100644
--- twisted/lore/man2lore.py
+++ twisted/lore/man2lore.py
@@ -197,7 +197,7 @@ class ManConverter(object):
             if arg.lower() == "ar":
                 text += " \\fU%s\\fR" % (args.pop(0),)
             elif arg.lower() == "op":
-                ign = args.pop(0)
+                args.pop(0)
                 text += " [\\fU%s\\fR]" % (args.pop(0),)
 
         self.text(text)
diff --git twisted/lore/numberer.py twisted/lore/numberer.py
index f91cc28..3ef0736 100644
--- twisted/lore/numberer.py
+++ twisted/lore/numberer.py
@@ -1,6 +1,9 @@
 # Copyright (c) Twisted Matrix Laboratories.
 # See LICENSE for details.
 
+numberSections = None
+filenum = None
+
 
 def reset():
     resetFilenum()
diff --git twisted/lore/test/test_lint.py twisted/lore/test/test_lint.py
index db257b8..f28b050 100644
--- twisted/lore/test/test_lint.py
+++ twisted/lore/test/test_lint.py
@@ -141,7 +141,7 @@ class DefaultTagCheckerTests(TestCase):
                 checker.check(document, filename)
             finally:
                 patch.restore()
-        except ProcessingFailure, e:
+        except ProcessingFailure:
             if expectSuccess:
                 raise
         else:
diff --git twisted/lore/texi.py twisted/lore/texi.py
index 8f8f3cb..991b3f3 100644
--- twisted/lore/texi.py
+++ twisted/lore/texi.py
@@ -21,7 +21,6 @@ class TexiSpitter(latex.BaseLatexSpitter):
     baseLevel = 1
 
     def writeNodeData(self, node):
-        buf = StringIO()
         latex.getLatexText(node, self.writer, texiEscape, entities)
 
     def visitNode_title(self, node):
diff --git twisted/lore/topfiles/setup.py twisted/lore/topfiles/setup.py
index a04f563..b9c29ca 100644
--- twisted/lore/topfiles/setup.py
+++ twisted/lore/topfiles/setup.py
@@ -1,8 +1,6 @@
 # Copyright (c) Twisted Matrix Laboratories.
 # See LICENSE for details.
 
-import sys
-
 try:
     from twisted.python import dist
 except ImportError:
