Index: /home/wolf/twisted-trunk/doc/core/howto/async.xhtml
===================================================================
--- /home/wolf/twisted-trunk/doc/core/howto/async.xhtml	(revision 23536)
+++ /home/wolf/twisted-trunk/doc/core/howto/async.xhtml	(working copy)
@@ -15,7 +15,7 @@
 
 <p>This document is for readers new to Twisted who are familiar with the
 Python programming language and, at least conceptually, with core networking
-conepts such as servers, clients and sockets. This document will give you a
+concepts such as servers, clients and sockets. This document will give you a
 high level overview of concurrent programming (interleaving several tasks) and
 of Twisted's concurrency model: <strong>non-blocking code</strong> or
 <strong>asynchronous code</strong>.</p>
