Index: doc/web/howto/web-in-60/wsgi.xhtml
===================================================================
--- doc/web/howto/web-in-60/wsgi.xhtml	(revision 33396)
+++ doc/web/howto/web-in-60/wsgi.xhtml	(working copy)
@@ -17,7 +17,7 @@
 serve <a href="http://www.python.org/dev/peps/pep-0333/">WSGI applications</a>
 in a Twisted Web server.</p>
 
-<p>Note thate <code>WSGIResource</code> is a multithreaded WSGI container. Like
+<p>Note that a <code>WSGIResource</code> is a multithreaded WSGI container. Like
 any other WSGI container, you can't do anything asynchronous in your WSGI
 applications, even though this is a Twisted WSGI container.</p>
 
