diff --git a/twisted/test/test_pb.py b/twisted/test/test_pb.py
index e17e2b5..15e2363 100644
--- a/twisted/test/test_pb.py
+++ b/twisted/test/test_pb.py
@@ -19,7 +19,7 @@ from zope.interface import implements, Interface
 from twisted.python.versions import Version
 from twisted.trial import unittest
 from twisted.spread import pb, util, publish, jelly
-from twisted.internet import protocol, main, reactor
+from twisted.internet import protocol, main, reactor, task
 from twisted.internet.error import ConnectionRefusedError
 from twisted.internet.defer import Deferred, gatherResults, succeed
 from twisted.protocols.policies import WrappingFactory
@@ -1336,10 +1336,16 @@ class NewCredTestCase(unittest.TestCase):
         d = factory.login(
             credentials.UsernamePassword('foo', 'bar'), "BRAINS!")
         def cbLoggedIn(avatar):
+            # Because frames above the current stack frame have a reference
+            # to avatar, we must escape the current stack.
+            return task.deferLater(reactor, 0, lambda: None)
+        d.addCallback(cbLoggedIn)
+        def garbageCollect(_):
+            gc.collect()
             # Just wait for the logout to happen, as it should since the
             # reference to the avatar will shortly no longer exists.
             return loggedOut
-        d.addCallback(cbLoggedIn)
+        d.addCallback(garbageCollect)
         def cbLoggedOut(ignored):
             # Verify that the server broker's _localCleanup dict isn't growing
             # without bound.
