Index: twisted/test/test_log.py
===================================================================
--- twisted/test/test_log.py	(revision 22180)
+++ twisted/test/test_log.py	(working copy)
@@ -168,7 +168,7 @@
         originalTimezone = os.environ.get('TZ', None)
         try:
             # Test something west of UTC
-            os.environ['TZ'] = 'US/Eastern'
+            os.environ['TZ'] = 'America/New_York'
             time.tzset()
             self.assertEqual(
                 self.flo.getTimezoneOffset(utcDaylightTimestamp),
