diff --git twisted/conch/test/test_agent.py twisted/conch/test/test_agent.py
index 6c768f1..772cf46 100644
--- twisted/conch/test/test_agent.py
+++ twisted/conch/test/test_agent.py
@@ -7,14 +7,8 @@ Tests for L{twisted.conch.ssh.agent}.
 
 import struct
 
-from twisted.internet import reactor
-from twisted.internet.interfaces import ITLSTransport
 from twisted.trial import unittest
-
-if not ITLSTransport.providedBy(reactor):
-    iosim = None
-else:
-    from twisted.test import iosim
+from twisted.test import iosim
 
 try:
     import Crypto.Cipher.DES3
diff --git twisted/conch/topfiles/7977.misc twisted/conch/topfiles/7977.misc
new file mode 100644
index 0000000..e69de29
