diff --git twisted/trial/_asynctest.py twisted/trial/_asynctest.py
index 9e7b0df..c793ed5 100644
--- twisted/trial/_asynctest.py
+++ twisted/trial/_asynctest.py
@@ -190,10 +190,10 @@ class TestCase(SynchronousTestCase):
 
 
     def _cbDeferRunCleanups(self, cleanupResults, result):
-        for flag, failure in cleanupResults:
+        for flag, testFailure in cleanupResults:
             if flag == defer.FAILURE:
-                result.addError(self, failure)
-                if failure.check(KeyboardInterrupt):
+                result.addError(self, testFailure)
+                if testFailure.check(KeyboardInterrupt):
                     result.stop()
                 self._passed = False
 
diff --git twisted/trial/runner.py twisted/trial/runner.py
index 4275c24..0838891 100644
--- twisted/trial/runner.py
+++ twisted/trial/runner.py
@@ -422,8 +422,8 @@ class TestLoader(object):
         if not hasattr(module, '__doctests__'):
             return suite
         docSuite = self.suiteFactory()
-        for doctest in module.__doctests__:
-            docSuite.addTest(self.loadDoctests(doctest))
+        for docTest in module.__doctests__:
+            docSuite.addTest(self.loadDoctests(docTest))
         return self.suiteFactory([suite, docSuite])
     loadTestsFromModule = loadModule
 
diff --git twisted/trial/test/suppression.py twisted/trial/test/suppression.py
index de9287c..e611ee0 100644
--- twisted/trial/test/suppression.py
+++ twisted/trial/test/suppression.py
@@ -14,11 +14,9 @@ from __future__ import division, absolute_import
 
 import warnings
 
-from twisted.python.compat import _PY3
 from twisted.trial import unittest, util
 
 
-
 METHOD_WARNING_MSG = "method warning message"
 CLASS_WARNING_MSG = "class warning message"
 MODULE_WARNING_MSG = "module warning message"
diff --git twisted/trial/test/test_output.py twisted/trial/test/test_output.py
index 5f8c7f9..2f52610 100644
--- twisted/trial/test/test_output.py
+++ twisted/trial/test/test_output.py
@@ -22,7 +22,7 @@ def runTrial(*args):
         stream=output,
         workingDirectory=config['temp-directory'])
     suite = trial._getSuite(config)
-    result = myRunner.run(suite)
+    myRunner.run(suite)
     return output.getvalue()
 
 
diff --git twisted/trial/test/test_runner.py twisted/trial/test/test_runner.py
index e30743c..84692d6 100644
--- twisted/trial/test/test_runner.py
+++ twisted/trial/test/test_runner.py
@@ -579,7 +579,7 @@ class TestTrialSuite(unittest.SynchronousTestCase):
     def test_imports(self):
         # FIXME, HTF do you test the reactor can be cleaned up ?!!!
         from twisted.trial.runner import TrialSuite
-
+        TrialSuite  # Silence the linter.
 
 
 
diff --git twisted/trial/test/test_suppression.py twisted/trial/test/test_suppression.py
index 8dbf046..e5086db 100644
--- twisted/trial/test/test_suppression.py
+++ twisted/trial/test/test_suppression.py
@@ -160,3 +160,8 @@ class SynchronousSuppressionTest(SuppressionMixin, unittest.SynchronousTestCase)
         SynchronousTestTearDownSuppression as TestTearDownSuppression,
         SynchronousTestSuppression as TestSuppression,
         SynchronousTestSuppression2 as TestSuppression2)
+    # Silence the linter.
+    TestSetUpSuppression
+    TestTearDownSuppression
+    TestSuppression
+    TestSuppression2
diff --git twisted/trial/test/test_tests.py twisted/trial/test/test_tests.py
index 1e013f2..b081e16 100644
--- twisted/trial/test/test_tests.py
+++ twisted/trial/test/test_tests.py
@@ -223,6 +223,10 @@ class SynchronousSkipMethodTests(SkipMethodsMixin, unittest.SynchronousTestCase)
         SynchronousSkipping as Skipping,
         SynchronousSkippingSetUp as SkippingSetUp,
         SynchronousDeprecatedReasonlessSkip as DeprecatedReasonlessSkip)
+    # Silence the linter.
+    Skipping
+    SkippingSetUp
+    DeprecatedReasonlessSkip
 
 
 class AsynchronousSkipMethodTests(SkipMethodsMixin, unittest.TestCase):
@@ -235,6 +239,10 @@ class AsynchronousSkipMethodTests(SkipMethodsMixin, unittest.TestCase):
         AsynchronousSkipping as Skipping,
         AsynchronousSkippingSetUp as SkippingSetUp,
         AsynchronousDeprecatedReasonlessSkip as DeprecatedReasonlessSkip)
+    # Silence the linter.
+    Skipping
+    SkippingSetUp
+    DeprecatedReasonlessSkip
 
 
 class SkipClassesMixin(ResultsTestMixin):
@@ -295,6 +303,7 @@ class SynchronousSkipClassTests(SkipClassesMixin, unittest.SynchronousTestCase):
     """
     from twisted.trial.test.skipping import (
         SynchronousSkippedClass as SkippedClass)
+    SkippedClass  # Silence the linter.
 
 
 class AsynchronousSkipClassTests(SkipClassesMixin, unittest.TestCase):
@@ -305,6 +314,7 @@ class AsynchronousSkipClassTests(SkipClassesMixin, unittest.TestCase):
     """
     from twisted.trial.test.skipping import (
         AsynchronousSkippedClass as SkippedClass)
+    SkippedClass  # Silence the linter.
 
 
 class TodoMixin(ResultsTestMixin):
@@ -406,6 +416,10 @@ class SynchronousTodoTests(TodoMixin, unittest.SynchronousTestCase):
         SynchronousTodo as Todo,
         SynchronousSetUpTodo as SetUpTodo,
         SynchronousTearDownTodo as TearDownTodo)
+    # Silence the linter.
+    Todo
+    SetUpTodo
+    TearDownTodo
 
 
 class AsynchronousTodoTests(TodoMixin, unittest.TestCase):
@@ -418,6 +432,10 @@ class AsynchronousTodoTests(TodoMixin, unittest.TestCase):
         AsynchronousTodo as Todo,
         AsynchronousSetUpTodo as SetUpTodo,
         AsynchronousTearDownTodo as TearDownTodo)
+    # Silence the linter.
+    Todo
+    SetUpTodo
+    TearDownTodo
 
 
 class ClassTodoMixin(ResultsTestMixin):
@@ -484,6 +502,7 @@ class SynchronousClassTodoTests(ClassTodoMixin, unittest.SynchronousTestCase):
     """
     from twisted.trial.test.skipping import (
         SynchronousTodoClass as TodoClass)
+    TodoClass  # Silence the linter.
 
 
 class AsynchronousClassTodoTests(ClassTodoMixin, unittest.TestCase):
@@ -494,6 +513,7 @@ class AsynchronousClassTodoTests(ClassTodoMixin, unittest.TestCase):
     """
     from twisted.trial.test.skipping import (
         AsynchronousTodoClass as TodoClass)
+    TodoClass  # Silence the linter.
 
 
 class StrictTodoMixin(ResultsTestMixin):
@@ -559,7 +579,7 @@ class SynchronousStrictTodoTests(StrictTodoMixin, unittest.SynchronousTestCase):
     """
     from twisted.trial.test.skipping import (
         SynchronousStrictTodo as StrictTodo)
-
+    StrictTodo  # Silence the linter.
 
 
 class AsynchronousStrictTodoTests(StrictTodoMixin, unittest.TestCase):
@@ -571,7 +591,7 @@ class AsynchronousStrictTodoTests(StrictTodoMixin, unittest.TestCase):
     """
     from twisted.trial.test.skipping import (
         AsynchronousStrictTodo as StrictTodo)
-
+    StrictTodo  # Silence the linter.
 
 
 class TestReactorCleanup(unittest.SynchronousTestCase):
@@ -661,6 +681,9 @@ class SynchronousFixtureTest(FixtureMixin, unittest.SynchronousTestCase):
     from twisted.trial.test.erroneous import (
         SynchronousTestFailureInSetUp as TestFailureInSetUp,
         SynchronousTestFailureInTearDown as TestFailureInTearDown)
+    # Silence the linter.
+    TestFailureInSetUp
+    TestFailureInTearDown
 
 
 class AsynchronousFixtureTest(FixtureMixin, unittest.TestCase):
@@ -672,6 +695,9 @@ class AsynchronousFixtureTest(FixtureMixin, unittest.TestCase):
     from twisted.trial.test.erroneous import (
         AsynchronousTestFailureInSetUp as TestFailureInSetUp,
         AsynchronousTestFailureInTearDown as TestFailureInTearDown)
+    # Silence the linter.
+    TestFailureInSetUp
+    TestFailureInTearDown
 
 
 class AsynchronousSuppressionTest(SuppressionMixin, unittest.TestCase):
@@ -686,6 +712,11 @@ class AsynchronousSuppressionTest(SuppressionMixin, unittest.TestCase):
         AsynchronousTestTearDownSuppression as TestTearDownSuppression,
         AsynchronousTestSuppression as TestSuppression,
         AsynchronousTestSuppression2 as TestSuppression2)
+    # Silence the linter.
+    TestSetUpSuppression
+    TestTearDownSuppression
+    TestSuppression
+    TestSuppression2
 
 
 class GCMixin:
@@ -928,6 +959,7 @@ class SynchronousAddCleanupTests(AddCleanupMixin, unittest.SynchronousTestCase):
     See: L{twisted.trial.test.test_tests.AddCleanupMixin}
     """
     from twisted.trial.test.skipping import SynchronousAddCleanup as AddCleanup
+    AddCleanup  # Silence the linter.
 
 
 class AsynchronousAddCleanupTests(AddCleanupMixin, unittest.TestCase):
@@ -937,6 +969,7 @@ class AsynchronousAddCleanupTests(AddCleanupMixin, unittest.TestCase):
     See: L{twisted.trial.test.test_tests.AddCleanupMixin}
     """
     from twisted.trial.test.skipping import AsynchronousAddCleanup as AddCleanup
+    AddCleanup  # Silence the linter.
 
     def test_addCleanupWaitsForDeferreds(self):
         """
@@ -1126,7 +1159,7 @@ class TestDecoratorMixin(object):
         test = self.TestCase()
         suite = unittest.TestSuite([test])
         count1 = getrefcount(test)
-        decoratedTest = unittest.decorate(suite, unittest.TestDecorator)
+        unittest.decorate(suite, unittest.TestDecorator)
         count2 = getrefcount(test)
         self.assertEqual(count1, count2)
 
diff --git twisted/trial/unittest.py twisted/trial/unittest.py
index 8f07556..a8889e4 100644
--- twisted/trial/unittest.py
+++ twisted/trial/unittest.py
@@ -12,6 +12,8 @@ from __future__ import division, absolute_import
 from twisted.trial._synctest import (
     FailTest, SkipTest, SynchronousTestCase, PyUnitResultAdapter, Todo,
     makeTodo)
+# Silence the linter.
+PyUnitResultAdapter
 from twisted.trial._asynctest import TestCase
 
 from twisted.python.compat import _PY3
@@ -21,6 +23,11 @@ if not _PY3:
         TestSuite, TestDecorator, decorate)
     from twisted.trial._asyncrunner import (
         _ForceGarbageCollectionDecorator, _iterateTests, _clearSuite)
+    # Silence the linter.
+    _clearSuite
+    _ForceGarbageCollectionDecorator
+    _iterateTests
+    TestDecorator
 
 # Further obscure the origins of these objects, to reduce surprise (and this is
 # what the values were before code got shuffled around between files, but was
@@ -33,6 +40,11 @@ FailTest.__module__ = SkipTest.__module__ = __name__
 # from (is this a good idea?)
 from twisted.trial._synctest import (
     _LogObserver, _logObserver, _collectWarnings, _setWarningRegistryToNone)
+# Silence the linter.
+_collectWarnings
+_logObserver
+_LogObserver
+_setWarningRegistryToNone
 
 
 __all__ = [
