diff --git twisted/trial/_asynctest.py twisted/trial/_asynctest.py
index 8e0927a..c7a7090 100644
--- twisted/trial/_asynctest.py
+++ twisted/trial/_asynctest.py
@@ -196,10 +196,10 @@ class TestCase(SynchronousTestCase):
 
 
     def _cbDeferRunCleanups(self, cleanupResults, result):
-        for flag, failure in cleanupResults:
+        for flag, testFailure in cleanupResults:
             if flag == defer.FAILURE:
-                result.addError(self, failure)
-                if failure.check(KeyboardInterrupt):
+                result.addError(self, testFailure)
+                if testFailure.check(KeyboardInterrupt):
                     result.stop()
                 self._passed = False
 
diff --git twisted/trial/runner.py twisted/trial/runner.py
index 4275c24..0838891 100644
--- twisted/trial/runner.py
+++ twisted/trial/runner.py
@@ -422,8 +422,8 @@ class TestLoader(object):
         if not hasattr(module, '__doctests__'):
             return suite
         docSuite = self.suiteFactory()
-        for doctest in module.__doctests__:
-            docSuite.addTest(self.loadDoctests(doctest))
+        for docTest in module.__doctests__:
+            docSuite.addTest(self.loadDoctests(docTest))
         return self.suiteFactory([suite, docSuite])
     loadTestsFromModule = loadModule
 
diff --git twisted/trial/test/suppression.py twisted/trial/test/suppression.py
index de9287c..3a405af 100644
--- twisted/trial/test/suppression.py
+++ twisted/trial/test/suppression.py
@@ -14,7 +14,6 @@ from __future__ import division, absolute_import
 
 import warnings
 
-from twisted.python.compat import _PY3
 from twisted.trial import unittest, util
 
 
diff --git twisted/trial/test/test_output.py twisted/trial/test/test_output.py
index 5f8c7f9..2f52610 100644
--- twisted/trial/test/test_output.py
+++ twisted/trial/test/test_output.py
@@ -22,7 +22,7 @@ def runTrial(*args):
         stream=output,
         workingDirectory=config['temp-directory'])
     suite = trial._getSuite(config)
-    result = myRunner.run(suite)
+    myRunner.run(suite)
     return output.getvalue()
 
 
diff --git twisted/trial/test/test_runner.py twisted/trial/test/test_runner.py
index e30743c..c5ecb16 100644
--- twisted/trial/test/test_runner.py
+++ twisted/trial/test/test_runner.py
@@ -13,6 +13,7 @@ from twisted.trial.itrial import IReporter, ITestCase
 from twisted.trial import unittest, runner, reporter, util
 from twisted.python import failure, log, reflect
 from twisted.python.filepath import FilePath
+from twisted.python.reflect import namedAny
 from twisted.scripts import trial
 from twisted.plugins import twisted_trial
 from twisted import plugin
@@ -578,8 +579,7 @@ class TestTrialSuite(unittest.SynchronousTestCase):
 
     def test_imports(self):
         # FIXME, HTF do you test the reactor can be cleaned up ?!!!
-        from twisted.trial.runner import TrialSuite
-
+        namedAny('twisted.trial.runner.TrialSuite')
 
 
 
diff --git twisted/trial/test/test_suppression.py twisted/trial/test/test_suppression.py
index 8dbf046..6999b64 100644
--- twisted/trial/test/test_suppression.py
+++ twisted/trial/test/test_suppression.py
@@ -9,6 +9,7 @@ from __future__ import division, absolute_import
 
 import unittest as pyunit
 
+from twisted.python.reflect import namedAny
 from twisted.trial import unittest
 from twisted.trial.test import suppression
 
@@ -155,8 +156,11 @@ class SynchronousSuppressionTest(SuppressionMixin, unittest.SynchronousTestCase)
     """
     @see: L{twisted.trial.test.test_tests}
     """
-    from twisted.trial.test.suppression import (
-        SynchronousTestSetUpSuppression as TestSetUpSuppression,
-        SynchronousTestTearDownSuppression as TestTearDownSuppression,
-        SynchronousTestSuppression as TestSuppression,
-        SynchronousTestSuppression2 as TestSuppression2)
+    TestSetUpSuppression = namedAny(
+        'twisted.trial.test.suppression.SynchronousTestSetUpSuppression')
+    TestTearDownSuppression = namedAny(
+        'twisted.trial.test.suppression.SynchronousTestTearDownSuppression')
+    TestSuppression = namedAny(
+        'twisted.trial.test.suppression.SynchronousTestSuppression')
+    TestSuppression2 = namedAny(
+        'twisted.trial.test.suppression.SynchronousTestSuppression2')
diff --git twisted/trial/test/test_tests.py twisted/trial/test/test_tests.py
index 609ead5..9e049bd 100644
--- twisted/trial/test/test_tests.py
+++ twisted/trial/test/test_tests.py
@@ -27,6 +27,7 @@ import gc, sys, weakref
 import unittest as pyunit
 
 from twisted.python.compat import _PY3, NativeStringIO
+from twisted.python.reflect import namedAny
 from twisted.internet import defer, reactor
 from twisted.trial import unittest, reporter, util
 if not _PY3:
@@ -219,10 +220,11 @@ class SynchronousSkipMethodTests(SkipMethodsMixin, unittest.SynchronousTestCase)
 
     See: L{twisted.trial.test.test_tests.SkipMethodsMixin}
     """
-    from twisted.trial.test.skipping import (
-        SynchronousSkipping as Skipping,
-        SynchronousSkippingSetUp as SkippingSetUp,
-        SynchronousDeprecatedReasonlessSkip as DeprecatedReasonlessSkip)
+    Skipping = namedAny('twisted.trial.test.skipping.SynchronousSkipping')
+    SkippingSetUp = namedAny(
+        'twisted.trial.test.skipping.SynchronousSkippingSetUp')
+    DeprecatedReasonlessSkip = namedAny(
+        'twisted.trial.test.skipping.SynchronousDeprecatedReasonlessSkip')
 
 
 class AsynchronousSkipMethodTests(SkipMethodsMixin, unittest.TestCase):
@@ -231,10 +233,11 @@ class AsynchronousSkipMethodTests(SkipMethodsMixin, unittest.TestCase):
 
     See: L{twisted.trial.test.test_tests.SkipMethodsMixin}
     """
-    from twisted.trial.test.skipping import (
-        AsynchronousSkipping as Skipping,
-        AsynchronousSkippingSetUp as SkippingSetUp,
-        AsynchronousDeprecatedReasonlessSkip as DeprecatedReasonlessSkip)
+    Skipping = namedAny('twisted.trial.test.skipping.AsynchronousSkipping')
+    SkippingSetUp = namedAny(
+        'twisted.trial.test.skipping.AsynchronousSkippingSetUp')
+    DeprecatedReasonlessSkip = namedAny(
+        'twisted.trial.test.skipping.AsynchronousDeprecatedReasonlessSkip')
 
 
 class SkipClassesMixin(ResultsTestMixin):
@@ -287,14 +290,16 @@ class SkipClassesMixin(ResultsTestMixin):
         self.assertEqual(expectedReasons, reasonsGiven)
 
 
+
 class SynchronousSkipClassTests(SkipClassesMixin, unittest.SynchronousTestCase):
     """
     Test the class skipping features in the synchronous case.
 
     See: L{twisted.trial.test.test_tests.SkipClassesMixin}
     """
-    from twisted.trial.test.skipping import (
-        SynchronousSkippedClass as SkippedClass)
+    SkippedClass = namedAny(
+        'twisted.trial.test.skipping.SynchronousSkippedClass')
+
 
 
 class AsynchronousSkipClassTests(SkipClassesMixin, unittest.TestCase):
@@ -303,8 +308,9 @@ class AsynchronousSkipClassTests(SkipClassesMixin, unittest.TestCase):
 
     See: L{twisted.trial.test.test_tests.SkipClassesMixin}
     """
-    from twisted.trial.test.skipping import (
-        AsynchronousSkippedClass as SkippedClass)
+    SkippedClass = namedAny(
+        'twisted.trial.test.skipping.AsynchronousSkippedClass')
+
 
 
 class TodoMixin(ResultsTestMixin):
@@ -396,16 +402,18 @@ class TodoMixin(ResultsTestMixin):
         self.assertEqual(self.reporter.successes, 0)
 
 
+
 class SynchronousTodoTests(TodoMixin, unittest.SynchronousTestCase):
     """
     Test the class skipping features in the synchronous case.
 
     See: L{twisted.trial.test.test_tests.TodoMixin}
     """
-    from twisted.trial.test.skipping import (
-        SynchronousTodo as Todo,
-        SynchronousSetUpTodo as SetUpTodo,
-        SynchronousTearDownTodo as TearDownTodo)
+    Todo = namedAny('twisted.trial.test.skipping.SynchronousTodo')
+    SetUpTodo = namedAny('twisted.trial.test.skipping.SynchronousSetUpTodo')
+    TearDownTodo = namedAny(
+        'twisted.trial.test.skipping.SynchronousTearDownTodo')
+
 
 
 class AsynchronousTodoTests(TodoMixin, unittest.TestCase):
@@ -414,10 +422,11 @@ class AsynchronousTodoTests(TodoMixin, unittest.TestCase):
 
     See: L{twisted.trial.test.test_tests.TodoMixin}
     """
-    from twisted.trial.test.skipping import (
-        AsynchronousTodo as Todo,
-        AsynchronousSetUpTodo as SetUpTodo,
-        AsynchronousTearDownTodo as TearDownTodo)
+    Todo = namedAny('twisted.trial.test.skipping.AsynchronousTodo')
+    SetUpTodo = namedAny('twisted.trial.test.skipping.AsynchronousSetUpTodo')
+    TearDownTodo = namedAny(
+        'twisted.trial.test.skipping.AsynchronousTearDownTodo')
+
 
 
 class ClassTodoMixin(ResultsTestMixin):
@@ -476,14 +485,15 @@ class ClassTodoMixin(ResultsTestMixin):
         self.assertEqual(expectedReasons, reasonsGiven)
 
 
+
 class SynchronousClassTodoTests(ClassTodoMixin, unittest.SynchronousTestCase):
     """
     Tests for the class-wide I{expected failure} features in the synchronous case.
 
     See: L{twisted.trial.test.test_tests.ClassTodoMixin}
     """
-    from twisted.trial.test.skipping import (
-        SynchronousTodoClass as TodoClass)
+    TodoClass = namedAny('twisted.trial.test.skipping.SynchronousTodoClass')
+
 
 
 class AsynchronousClassTodoTests(ClassTodoMixin, unittest.TestCase):
@@ -492,8 +502,8 @@ class AsynchronousClassTodoTests(ClassTodoMixin, unittest.TestCase):
 
     See: L{twisted.trial.test.test_tests.ClassTodoMixin}
     """
-    from twisted.trial.test.skipping import (
-        AsynchronousTodoClass as TodoClass)
+    TodoClass = namedAny('twisted.trial.test.skipping.AsynchronousTodoClass')
+
 
 
 class StrictTodoMixin(ResultsTestMixin):
@@ -550,6 +560,7 @@ class StrictTodoMixin(ResultsTestMixin):
         self.assertEqual(expectedReasons, reasonsGotten)
 
 
+
 class SynchronousStrictTodoTests(StrictTodoMixin, unittest.SynchronousTestCase):
     """
     Tests for the expected failure case when the exact failure that is expected
@@ -557,8 +568,7 @@ class SynchronousStrictTodoTests(StrictTodoMixin, unittest.SynchronousTestCase):
 
     See: L{twisted.trial.test.test_tests.StrictTodoMixin}
     """
-    from twisted.trial.test.skipping import (
-        SynchronousStrictTodo as StrictTodo)
+    StrictTodo = namedAny('twisted.trial.test.skipping.SynchronousStrictTodo')
 
 
 
@@ -569,8 +579,7 @@ class AsynchronousStrictTodoTests(StrictTodoMixin, unittest.TestCase):
 
     See: L{twisted.trial.test.test_tests.StrictTodoMixin}
     """
-    from twisted.trial.test.skipping import (
-        AsynchronousStrictTodo as StrictTodo)
+    StrictTodo = namedAny('twisted.trial.test.skipping.AsynchronousStrictTodo')
 
 
 
@@ -652,15 +661,18 @@ class FixtureMixin(object):
         self.assertEqual(0, self.reporter.successes)
 
 
+
 class SynchronousFixtureTest(FixtureMixin, unittest.SynchronousTestCase):
     """
     Tests for broken fixture helper methods in the synchronous case
 
     See: L{twisted.trial.test.test_tests.FixtureMixin}
     """
-    from twisted.trial.test.erroneous import (
-        SynchronousTestFailureInSetUp as TestFailureInSetUp,
-        SynchronousTestFailureInTearDown as TestFailureInTearDown)
+    TestFailureInSetUp = namedAny(
+        'twisted.trial.test.erroneous.SynchronousTestFailureInSetUp')
+    TestFailureInTearDown = namedAny(
+        'twisted.trial.test.erroneous.SynchronousTestFailureInTearDown')
+
 
 
 class AsynchronousFixtureTest(FixtureMixin, unittest.TestCase):
@@ -669,9 +681,11 @@ class AsynchronousFixtureTest(FixtureMixin, unittest.TestCase):
 
     See: L{twisted.trial.test.test_tests.FixtureMixin}
     """
-    from twisted.trial.test.erroneous import (
-        AsynchronousTestFailureInSetUp as TestFailureInSetUp,
-        AsynchronousTestFailureInTearDown as TestFailureInTearDown)
+    TestFailureInSetUp = namedAny(
+        'twisted.trial.test.erroneous.AsynchronousTestFailureInSetUp')
+    TestFailureInTearDown = namedAny(
+        'twisted.trial.test.erroneous.AsynchronousTestFailureInTearDown')
+
 
 
 class AsynchronousSuppressionTest(SuppressionMixin, unittest.TestCase):
@@ -681,11 +695,15 @@ class AsynchronousSuppressionTest(SuppressionMixin, unittest.TestCase):
 
     See L{twisted.trial.test.test_suppression.SuppressionMixin}
     """
-    from twisted.trial.test.suppression import (
-        AsynchronousTestSetUpSuppression as TestSetUpSuppression,
-        AsynchronousTestTearDownSuppression as TestTearDownSuppression,
-        AsynchronousTestSuppression as TestSuppression,
-        AsynchronousTestSuppression2 as TestSuppression2)
+    TestSetUpSuppression = namedAny(
+        'twisted.trial.test.suppression.AsynchronousTestSetUpSuppression')
+    TestTearDownSuppression = namedAny(
+        'twisted.trial.test.suppression.AsynchronousTestTearDownSuppression')
+    TestSuppression = namedAny(
+        'twisted.trial.test.suppression.AsynchronousTestSuppression')
+    TestSuppression2 = namedAny(
+        'twisted.trial.test.suppression.AsynchronousTestSuppression2')
+
 
 
 class GCMixin:
@@ -921,13 +939,15 @@ class AddCleanupMixin(object):
         self.assertEqual(error2.getErrorMessage(), 'foo')
 
 
+
 class SynchronousAddCleanupTests(AddCleanupMixin, unittest.SynchronousTestCase):
     """
     Test the addCleanup method of TestCase in the synchronous case
 
     See: L{twisted.trial.test.test_tests.AddCleanupMixin}
     """
-    from twisted.trial.test.skipping import SynchronousAddCleanup as AddCleanup
+    AddCleanup = namedAny('twisted.trial.test.skipping.SynchronousAddCleanup')
+
 
 
 class AsynchronousAddCleanupTests(AddCleanupMixin, unittest.TestCase):
@@ -936,7 +956,7 @@ class AsynchronousAddCleanupTests(AddCleanupMixin, unittest.TestCase):
 
     See: L{twisted.trial.test.test_tests.AddCleanupMixin}
     """
-    from twisted.trial.test.skipping import AsynchronousAddCleanup as AddCleanup
+    AddCleanup = namedAny('twisted.trial.test.skipping.AsynchronousAddCleanup')
 
     def test_addCleanupWaitsForDeferreds(self):
         """
@@ -1126,7 +1146,7 @@ class TestDecoratorMixin(object):
         test = self.TestCase()
         suite = unittest.TestSuite([test])
         count1 = getrefcount(test)
-        decoratedTest = unittest.decorate(suite, unittest.TestDecorator)
+        unittest.decorate(suite, unittest.TestDecorator)
         count2 = getrefcount(test)
         self.assertEqual(count1, count2)
 
diff --git twisted/trial/unittest.py twisted/trial/unittest.py
index 8f07556..61fc7c4 100644
--- twisted/trial/unittest.py
+++ twisted/trial/unittest.py
@@ -36,6 +36,21 @@ from twisted.trial._synctest import (
 
 
 __all__ = [
-    'FailTest', 'SkipTest', 'SynchronousTestCase', 'Todo', 'makeTodo',
-
-    'TestCase', 'TestSuite', 'decorate']
+    '_clearSuite',
+    '_ForceGarbageCollectionDecorator',
+    '_iterateTests',
+    '_collectWarnings',
+    '_logObserver',
+    '_LogObserver',
+    '_setWarningRegistryToNone',
+    'FailTest',
+    'SkipTest',
+    'SynchronousTestCase',
+    'Todo',
+    'makeTodo',
+    'PyUnitResultAdapter',
+    'TestCase',
+    'TestDecorator',
+    'TestSuite',
+    'decorate',
+    ]
