Index: pb.py
===================================================================
--- pb.py	(revision 17428)
+++ pb.py	(working copy)
@@ -1023,8 +1023,9 @@
     protocol = Broker
     unsafeTracebacks = 0
 
-    def __init__(self):
+    def __init__(self,security=globalSecurity):
         self._reset()
+        self.security=security
 
     def _reset(self):
         self.rootObjectRequests = [] # list of deferred
@@ -1052,6 +1053,7 @@
 
     def clientConnectionMade(self, broker):
         self._broker = broker
+        self._broker.security=self.security
         self._root = broker.remoteForName("root")
         ds = self.rootObjectRequests
         self.rootObjectRequests = []
@@ -1116,14 +1118,15 @@
     # object broker factory
     protocol = Broker
 
-    def __init__(self, root, unsafeTracebacks=False):
+    def __init__(self, root, unsafeTracebacks=False,security=globalSecurity):
         self.root = IPBRoot(root)
         self.unsafeTracebacks = unsafeTracebacks
+        self.security=security
 
     def buildProtocol(self, addr):
         """Return a Broker attached to me (as the service provider).
         """
-        proto = self.protocol(0)
+        proto = self.protocol(0, security=self.security)
         proto.factory = self
         proto.setNameForLocal("root", self.root.rootObject(proto))
         return proto
