Index: twisted/words/protocols/irc.py
===================================================================
--- twisted/words/protocols/irc.py	(revision 40276)
+++ twisted/words/protocols/irc.py	(working copy)
@@ -265,7 +265,7 @@
         First argument is the command, all subsequent arguments are parameters
         to that command.  If a prefix is desired, it may be specified with the
         keyword argument 'prefix'.
-        
+
         Message tags may be specified using the keyword argument 'tags', passing
         a list of 1- or 2-tuples.  One-tuples should be passed for tags whose
         presence has a meaning (i.e. those that do not show a value).
@@ -284,7 +284,7 @@
             line = ":%s %s" % (prefix['prefix'], line)
         if 'tags' in prefix:
             line = "@%s %s" % (";".join(
-               ["=".join(tag) for tag in prefix['tags']]), line)
+               ["=".join(tag[:2]) for tag in prefix['tags']]), line)
         self.sendLine(line)
 
         if len(parameter_list) > 15:
