Index: twisted/web/script.py
===================================================================
--- twisted/web/script.py	(revision 34886)
+++ twisted/web/script.py	(working copy)
@@ -71,7 +71,11 @@
     return rsrc
 
 def ResourceTemplate(path, registry):
-    from quixote import ptl_compile
+    # The newer version of quixote has ptl_compile in the ptl module.
+    try:
+        from quixote import ptl_compile
+    except ImportError:
+        from quixote.ptl import ptl_compile
 
     glob = {'__file__': path,
             'resource': resource.ErrorPage(500, "Whoops! Internal Error",
Index: doc/web/examples/simple.rtl
===================================================================
--- doc/web/examples/simple.rtl	(revision 34886)
+++ doc/web/examples/simple.rtl	(working copy)
@@ -2,13 +2,13 @@
 # See LICENSE for details.
 
 """
-This example demostrates how to render a page using a third-party template
-system.
+This example demostrates how to render a page using ResourceTemplate.
 
 Usage:
-    $ twistd -n web --process=.rtl=twisted.web.script.ResourceTemplate --path /path/to/examples/
+    $ twistd -n web --path=/path/to/examples/ --process=.rtl=twisted.web.script.ResourceTemplate
 
-And make sure Quixote is installed.
+And make sure Quixote is installed:
+    $ pip install Quixote
 """
 
 from twisted.web.resource import Resource
@@ -16,17 +16,24 @@
 
 class ExampleResource(Resource):
 
+    template = """
+<html>
+  <head><title>Welcome To Twisted Python</title></head>
+  <body>
+    <h1>%(brief)s</h1>
+    <ul>%(detail)s</ul>
+  </body>
+</html>
+"""
+
     def render_GET(self, request):
-        """\
-<HTML>
-    <HEAD><TITLE> Welcome To Twisted Python </title></head>
-
-    <BODY><ul>"""
+        n_list = ''
         for i in range(10):
-            '<LI>';i
-        """</ul></body>
-</html>"""
+            n_list += "<li>%s</li>" % i
 
+        return self.template % dict(
+            brief="Welcome!",
+            detail=n_list)
 
+
 resource = ExampleResource()
-
