diff --git a/twisted/trial/test/test_runner.py b/twisted/trial/test/test_runner.py
index 8dfc75f..1dc5fb7 100644
--- a/twisted/trial/test/test_runner.py
+++ b/twisted/trial/test/test_runner.py
@@ -533,7 +533,7 @@ class TestRunner(unittest.SynchronousTestCase):
         debugger = my_runner.debugger = CapturingDebugger()
         result = self.runSampleSuite(my_runner)
         self.assertEqual(self.standardReport, result._calls)
-        self.assertEqual(['runcall'], debugger._calls)
+        self.assertEqual(['runcall'], debugger.calls)
 
 
     def test_runnerDebuggerDefaultsToPdb(self):
@@ -589,7 +589,7 @@ class TestRunner(unittest.SynchronousTestCase):
         my_runner = self.getRunner()
         result = self.runSampleSuite(my_runner)
         self.assertEqual(self.standardReport, result._calls)
-        self.assertEqual(['runcall'], my_runner.debugger._calls)
+        self.assertEqual(['runcall'], my_runner.debugger.calls)
 
 
 
