Index: twisted/protocols/tls.py
===================================================================
--- twisted/protocols/tls.py	(revision 37814)
+++ twisted/protocols/tls.py	(working copy)
@@ -294,6 +294,11 @@
         """
         tlsContext = self.factory._contextFactory.getContext()
         self._tlsConnection = Connection(tlsContext, None)
+        tlsContext.set_cipher_list(':'.join([
+                    cipher
+                    for cipher
+                    in self._tlsConnection.get_cipher_list()
+                    if '3DES' not in cipher]))
         if self.factory._isClient:
             self._tlsConnection.set_connect_state()
         else:
