Index: twisted/test/test_twistd.py
===================================================================
--- twisted/test/test_twistd.py	(revision 29003)
+++ twisted/test/test_twistd.py	(working copy)
@@ -454,45 +454,7 @@
             reactor.called, "startReactor did not call reactor.run()")
 
 
-    def test_legacyApplicationRunnerGetLogObserver(self):
-        """
-        L{app.ApplicationRunner} subclasses can have a getLogObserver that used
-        to return a log observer. This test is there to ensure that it's
-        supported but it raises a warning when used.
-        """
-        observer = []
-        self.addCleanup(log.removeObserver, observer.append)
-        class GetLogObserverRunner(app.ApplicationRunner):
-            def getLogObserver(self):
-                return observer.append
 
-            def startLogging(self, observer):
-                """
-                Override C{startLogging} to call L{log.addObserver} instead of
-                L{log.startLoggingWithObserver}.
-                """
-                log.addObserver(observer)
-                self.logger._initialLog()
-
-            def preApplication(self):
-                pass
-
-            def postApplication(self):
-                pass
-
-            def createOrGetApplication(self):
-                pass
-
-        conf = twistd.ServerOptions()
-        runner = GetLogObserverRunner(conf)
-        self.assertWarns(DeprecationWarning,
-            "Specifying a log observer with getLogObserver is "
-            "deprecated. Please use a loggerFactory instead.",
-            app.__file__, runner.run)
-        self.assertEquals(len(observer), 3)
-
-
-
 class UnixApplicationRunnerSetupEnvironmentTests(unittest.TestCase):
     """
     Tests for L{UnixApplicationRunner.setupEnvironment}.
Index: twisted/topfiles/4092.removal
===================================================================
--- twisted/topfiles/4092.removal	(revision 0)
+++ twisted/topfiles/4092.removal	(revision 0)
@@ -0,0 +1,2 @@
+Removed twisted.application.app.ApplicationRunner.startLogging, which has been
+deprecated (doesn't say since when).
Index: twisted/application/app.py
===================================================================
--- twisted/application/app.py	(revision 29003)
+++ twisted/application/app.py	(working copy)
@@ -410,30 +410,12 @@
         self.preApplication()
         self.application = self.createOrGetApplication()
 
+        self.logger.start(self.application)
 
-        getLogObserverLegacy = getattr(self, 'getLogObserver', None)
-        if getLogObserverLegacy is not None:
-            warnings.warn("Specifying a log observer with getLogObserver is "
-                          "deprecated. Please use a loggerFactory instead.",
-                          category=DeprecationWarning)
-            self.startLogging(self.getLogObserver())
-        else:
-            self.logger.start(self.application)
-
         self.postApplication()
         self.logger.stop()
 
 
-    def startLogging(self, observer):
-        """
-        Initialize the logging system. DEPRECATED.
-
-        @param observer: The observer to add to the logging system.
-        """
-        log.startLoggingWithObserver(observer)
-        self.logger._initialLog()
-
-
     def startReactor(self, reactor, oldstdout, oldstderr):
         """
         Run the reactor with the given configuration.  Subclasses should
