--- test/test_process.py	(revision 26921)
+++ test/test_process.py	(working copy)
@@ -1222,16 +1222,16 @@
 
     def __init__(self):
         """
-        Initialiaze data structures.
+        Initialize data structures.
         """
         self.actions = []
         self.closed = []
         self.pipeCount = 0
-        self.O_RDWR = os.O_RDWR
-        self.O_NOCTTY = os.O_NOCTTY
-        self.WNOHANG = os.WNOHANG
-        self.WEXITSTATUS = os.WEXITSTATUS
-        self.WIFEXITED = os.WIFEXITED
+        self.O_RDWR = -1
+        self.O_NOCTTY = -2
+        self.WNOHANG = -4
+        self.WEXITSTATUS = lambda x: 0
+        self.WIFEXITED = lambda x: 1
         self.seteuidCalls = []
         self.setegidCalls = []