diff -urN Twisted-10.0.0.orig/twisted/test/test_jelly.py Twisted-10.0.0/twisted/test/test_jelly.py
--- Twisted-10.0.0.orig/twisted/test/test_jelly.py	2010-05-11 13:55:25.320728000 +0200
+++ Twisted-10.0.0/twisted/test/test_jelly.py	2010-05-11 15:38:33.998927000 +0200
@@ -84,7 +84,23 @@
     Dummy new-style class.
     """
 
+class E(object):
+    """
+    Dummy new-style class with slots.
+    """
+
+    __slots__ = ("x", "y")
+
+    def __init__(self, x=None, y=None):
+        self.x = x
+        self.y = y
+
+    def __getstate__(self):
+        return {"x" : self.x, "y" : self.y}
 
+    def __setstate__(self, state):
+        self.x = state["x"]
+        self.y = state["y"]
 
 class SimpleJellyTest:
     def __init__(self, x, y):
@@ -162,6 +178,13 @@
         self.assertIsInstance(m, D)
         self.assertIdentical(m.n2, m.n3)
 
+    def test_newStyleWithSlots(self):
+        n = E()
+        n.x = 1
+        c = jelly.jelly(n)
+        m = jelly.unjelly(c)
+        self.assertIsInstance(m, E)
+        self.assertEquals(n.x, 1)
 
     def test_typeOldStyle(self):
         """
