Index: twisted/manhole/test/test_explorer.py
===================================================================
--- twisted/manhole/test/test_explorer.py	(revision 43486)
+++ twisted/manhole/test/test_explorer.py	(working copy)
@@ -20,7 +20,7 @@
     """
 
 
-class PoolTestCase(unittest.TestCase):
+class PoolTests(unittest.TestCase):
     """
     Tests for the Pool class.
     """
@@ -38,7 +38,7 @@
 
 
 
-class CRUFTWatchyThingieTestCase(unittest.TestCase):
+class CRUFTWatchyThingieTests(unittest.TestCase):
     """
     Tests for the CRUFT_WatchyThingie class.
     """
@@ -71,7 +71,7 @@
 
 
 
-class WatchMonkeyTestCase(unittest.TestCase):
+class WatchMonkeyTests(unittest.TestCase):
     """
     Tests for the _WatchMonkey class.
     """
Index: twisted/topfiles/7719.misc
===================================================================
