Index: doc/lore/howto/extend-lore.xhtml
===================================================================
--- doc/lore/howto/extend-lore.xhtml	(revision 28086)
+++ doc/lore/howto/extend-lore.xhtml	(working copy)
@@ -74,7 +74,7 @@
 
   <p>Create this file in a <code class="shell">twisted/plugins/</code>
   directory (<em>not</em> a package) which is located in a directory in the
-  Python module search path.  See the <a href="../../core/plugin.xhtml">Twisted
+  Python module search path.  See the <a href="../../core/howto/plugin.xhtml">Twisted
   plugin howto</a> for more details on plugins.</p>
 
   <p>Users of your extension will pass the value of your plugin's <code
