Index: twisted/runner/topfiles/setup.py
===================================================================
--- twisted/runner/topfiles/setup.py	(revision 27523)
+++ twisted/runner/topfiles/setup.py	(working copy)
@@ -1,21 +1,16 @@
 # Copyright (c) 2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
+from distutils.core import Extension
 try:
-    from twisted.python.dist import setup, ConditionalExtension as Extension
+    from twisted.python import dist
 except ImportError:
     raise SystemExit("twisted.python.dist module not found.  Make sure you "
                      "have installed the Twisted core package before "
                      "attempting to install any other Twisted projects.")
 
-extensions = [
-    Extension("twisted.runner.portmap",
-              ["twisted/runner/portmap.c"],
-              condition=lambda builder: builder._check_header("rpc/rpc.h")),
-]
-
 if __name__ == '__main__':
-    setup(
+    dist.setup(
         twisted_subproject="runner",
         # metadata
         name="Twisted Runner",
@@ -31,5 +26,5 @@
 with Python and Twisted, and has an almost full replacement for inetd.
 """,
         # build stuff
-        conditionalExtensions=extensions,
+        detectExtensions=detectExtensions,
     )
