Index: twisted/python/filepath.py
===================================================================
--- twisted/python/filepath.py	(revision 36417)
+++ twisted/python/filepath.py	(working copy)
@@ -970,7 +970,7 @@
     def getPermissions(self):
         """
         Returns the permissions of the file.  Should also work on Windows,
-        however, those permissions may not what is expected in Windows.
+        however, those permissions may not be what is expected in Windows.
 
         @return: the permissions for the file
         @rtype: L{Permissions}
Index: doc/words/examples/msn_example.py
===================================================================
--- doc/words/examples/msn_example.py	(revision 36417)
+++ doc/words/examples/msn_example.py	(working copy)
@@ -13,7 +13,7 @@
 import sys, getpass
 
 """
-This example connects to the MSN chat service and
+This simple twisted example connects to the MSN chat service and
 prints out information about all the users on your
 contact list (both online and offline).
 
@@ -21,6 +21,9 @@
 the connection process.
 
 @author Samuel Jordan
+
+To run the script:
+$ python msn_example.py
 """
 
 
Index: doc/words/examples/oscardemo.py
===================================================================
--- doc/words/examples/oscardemo.py	(revision 36417)
+++ doc/words/examples/oscardemo.py	(working copy)
@@ -3,6 +3,13 @@
 # Copyright (c) Twisted Matrix Laboratories.
 # See LICENSE for details.
 
+"""
+A very simple examaple of twisted.oscar code
+
+To run the script:
+$ python oscardemo.py
+"""
+
 from twisted.words.protocols import oscar
 from twisted.internet import protocol, reactor
 import getpass
Index: doc/words/examples/pb_client.py
===================================================================
--- doc/words/examples/pb_client.py	(revision 36417)
+++ doc/words/examples/pb_client.py	(working copy)
@@ -10,6 +10,10 @@
 authenticates with a username and password (given by argv[3] and argv[4]),
 joins a group (argv[5]) sends a simple message, leaves the group, and quits
 the server.
+
+To run the script:
+$ python pb_client.py <host> <port> <username> <password> <group>
+
 """
 
 import sys
Index: doc/words/examples/xmpp_client.py
===================================================================
--- doc/words/examples/xmpp_client.py	(revision 36417)
+++ doc/words/examples/xmpp_client.py	(working copy)
@@ -2,6 +2,13 @@
 # Copyright (c) Twisted Matrix Laboratories.
 # See LICENSE for details.
 
+"""
+A very simple twisted xmpp-client (Jabber ID)
+
+To run the script:
+$ python xmpp_client.py <jid> <secret>
+"""
+
 import sys
 
 from twisted.internet.defer import Deferred
Index: doc/words/examples/cursesclient.py
===================================================================
--- doc/words/examples/cursesclient.py	(revision 36417)
+++ doc/words/examples/cursesclient.py	(working copy)
@@ -15,6 +15,10 @@
 block.
 
 Remember to call nodelay(1) in curses, to make getch() non-blocking.
+
+To run the script:
+$ python cursesclient.py
+
 """
 
 # System Imports
Index: doc/words/examples/jabber_client.py
===================================================================
--- doc/words/examples/jabber_client.py	(revision 36417)
+++ doc/words/examples/jabber_client.py	(working copy)
@@ -1,6 +1,13 @@
 # Copyright (c) Twisted Matrix Laboratories.
 # See LICENSE for details.
 
+"""
+A very simple twisted jabber client
+
+To run the script:
+$ python jabber_client.py
+"""
+
 # Originally written by Darryl Vandorp
 # http://randomthoughts.vandorp.ca/
 
Index: doc/words/examples/ircLogBot.py
===================================================================
--- doc/words/examples/ircLogBot.py	(revision 36417)
+++ doc/words/examples/ircLogBot.py	(working copy)
@@ -20,6 +20,9 @@
   $ python ircLogBot.py test test.log
 
 will log channel #test to the file 'test.log'.
+
+To run the script:
+$ python ircLogBot.py <channel> <file>
 """
 
 
