Index: twisted/python/failure.py
===================================================================
--- twisted/python/failure.py	(revision 11598)
+++ twisted/python/failure.py	(working copy)
@@ -226,7 +226,13 @@
         return error
 
     def check(self, *errorTypes):
-        for error in errorTypes:
+        """Check if this failure's type is in a predetermined list.
+
+        @type errorTypes: L{Exception}
+
+	@returns: the matching L{Exception} type, or None if no match.
+        """
+         for error in errorTypes:
             err = error
             if isinstance(error, types.ClassType) and issubclass(error, Exception):
                 err = reflect.qual(error)
