Index: twisted/application/service.py
===================================================================
--- twisted/application/service.py	(Revision 32181)
+++ twisted/application/service.py	(Arbeitskopie)
@@ -39,13 +39,13 @@
         "Twisted application plugin.")
 
     options = Attribute(
-        "A C{twisted.python.usage.Options} subclass defining the"
+        "A C{twisted.python.usage.Options} subclass defining the "
         "configuration options for this application.")
 
 
     def makeService(options):
         """
-        Create and return an object providing
+        Create and return an object providing  \
         L{twisted.application.service.IService}.
 
         @param options: A mapping (typically a C{dict} or
