Index: twisted/application/service.py
===================================================================
--- twisted/application/service.py	(Revision 32181)
+++ twisted/application/service.py	(Arbeitskopie)
@@ -39,7 +39,7 @@
         "Twisted application plugin.")
 
     options = Attribute(
-        "A C{twisted.python.usage.Options} subclass defining the"
+        "A C{twisted.python.usage.Options} subclass defining the "
         "configuration options for this application.")
 
 
