Index: twisted/conch/telnet.py
===================================================================
--- twisted/conch/telnet.py	(revision 31396)
+++ twisted/conch/telnet.py	(working copy)
@@ -620,7 +620,8 @@
         d = state.him.onResult
         state.him.onResult = None
         d.callback(True)
-        assert self.enableRemote(option), "enableRemote must return True in this context (for option %r)" % (option,)
+        if not self.enableRemote(option):
+            raise TelnetError("enableRemote(%r) failed" % (option,))
 
     def will_yes_false(self, state, option):
         # He is unilaterally offering to enable an already-enabled option.
