Index: interfaces.py
===================================================================
--- interfaces.py	(revision 22498)
+++ interfaces.py	(working copy)
@@ -1003,8 +1003,7 @@
         stops blocking and a socket has been received.  If you need to
         send any greeting or initial message, do it here.
         """
-
-
+    
 class IHalfCloseableProtocol(Interface):
     """Implemented to indicate they want notification of half-closes.
 
@@ -1097,6 +1096,29 @@
         """
 
     def getPeer():
+        """DEPRECATED - moved to ISymmetricTwoEndpointTransport
+        Get the remote address of this connection.
+
+        Treat this method with caution.  It is the unfortunate result of the
+        CGI and Jabber standards, but should not be considered reliable for
+        the usual host of reasons; port forwarding, proxying, firewalls, IP
+        masquerading, etc.
+
+        @return: An L{IAddress} provider.
+        """
+
+    def getHost():
+        """DEPRECATED. - moved to ISymmetricTwoEndpointTransport
+        Similar to getPeer, but returns an address describing this side of the
+        connection.
+
+        @return: An L{IAddress} provider.
+        """
+
+class ISymmetricTwoEndpointTransport(ITransport):
+    """ Allow access to host information at both end points of the transport"""
+    
+    def getPeer():
         """Get the remote address of this connection.
 
         Treat this method with caution.  It is the unfortunate result of the
@@ -1115,8 +1137,7 @@
         @return: An L{IAddress} provider.
         """
 
-
-class ITCPTransport(ITransport):
+class ITCPTransport(ISymmetricTwoEndpointTransport):
     """A TCP based transport."""
 
     def loseWriteConnection():
