--- twisted\internet\_dumbwin32proc.py~	2008-07-26 22:37:37.000000000 +0200
+++ twisted\internet\_dumbwin32proc.py	2009-06-29 13:55:55.879196500 +0200
@@ -205,8 +205,6 @@
                                     origcmd, sheb))
                         raise OSError(pwte2)
 
-        win32file.CloseHandle(self.hThread)
-
         # close handles which only the child will use
         win32file.CloseHandle(hStderrW)
         win32file.CloseHandle(hStdoutW)
@@ -310,6 +308,10 @@
 
     def maybeCallProcessEnded(self):
         if self.closedNotifies == 3 and self.lostProcess:
+            win32file.CloseHandle(self.hProcess)
+            win32file.CloseHandle(self.hThread)
+            self.hProcess = None
+            self.hThread = None
             BaseProcess.maybeCallProcessEnded(self)
 
 
