Index: twisted/topfiles/5590.doc
===================================================================
--- twisted/topfiles/5590.doc   (revision 0)
+++ twisted/topfiles/5590.doc   (revision 0)
@@ -0,0 +1,2 @@
+Added brief description for window.tac and telnet_echo.tac in
+/doc/conch/examples
Index: doc/conch/examples/telnet_echo.tac
===================================================================
--- doc/conch/examples/telnet_echo.tac  (revision 34048)
+++ doc/conch/examples/telnet_echo.tac  (working copy)
@@ -1,6 +1,15 @@
 # Copyright (c) Twisted Matrix Laboratories.
 # See LICENSE for details.
 
+# You can run this .tac file directly with:
+#    twistd -ny demo.tac
+
+"""Simple echo server echoes back client input
+
+This demo sets up a listening port on 6023 which accepts telnet 
+connections.  No login for the telnet server is required.
+"""
+
 from twisted.conch.telnet import TelnetTransport, TelnetProtocol
 from twisted.internet.protocol import ServerFactory
 from twisted.application.internet import TCPServer
Index: doc/conch/examples/window.tac
===================================================================
--- doc/conch/examples/window.tac   (revision 34048)
+++ doc/conch/examples/window.tac   (working copy)
@@ -4,6 +4,16 @@
 # You can run this .tac file directly with:
 #    twistd -ny window.tac
 
+"""Widgets demo
+Demonstrates various widgets or buttons, such as scrollable regions, 
+drawable canvas, etc.
+
+This demo sets up two listening ports: one on 6022 which accepts ssh
+connections; one on 6023 which accepts telnet connections.  No login
+for the telnet server is required; for the ssh server, \"username\" is
+the username and \"password\" is the password.
+"""
+
 from __future__ import division
 
 import string, random
