diff --git a/twisted/python/syslog.py b/twisted/python/syslog.py
index 4e0ce8a..f84297e 100644
--- a/twisted/python/syslog.py
+++ b/twisted/python/syslog.py
@@ -23,17 +23,10 @@ class SyslogObserver:
         syslog.openlog(prefix)
 
     def emit(self, eventDict):
-        edm = eventDict['message']
-        if not edm:
-            if eventDict['isError'] and eventDict.has_key('failure'):
-                text = eventDict['failure'].getTraceback()
-            elif eventDict.has_key('format'):
-                text = eventDict['format'] % eventDict
-            else:
-                # we don't know how to log this
-                return
-        else:
-            text = ' '.join(map(str, edm))
+        # Figure out what the message-text is.
+        text = log.textFromEventDict(eventDict)
+        if text is None:
+            return
 
         lines = text.split('\n')
         while lines[-1:] == ['']:
