Index: twisted/internet/_baseprocess.py
===================================================================
--- twisted/internet/_baseprocess.py	(revision 32118)
+++ twisted/internet/_baseprocess.py	(working copy)
@@ -34,7 +34,10 @@
                 _missingProcessExited % (qual(self.proto.__class__),),
                 DeprecationWarning, stacklevel=0)
         else:
-            processExited(Failure(reason))
+            try:
+                processExited(Failure(reason))
+            except:
+                err(None, "unexpected error in processExited")
 
 
     def processEnded(self, status):
Index: twisted/internet/test/test_process.py
===================================================================
--- twisted/internet/test/test_process.py	(revision 32118)
+++ twisted/internet/test/test_process.py	(working copy)
@@ -298,6 +298,37 @@
         self.runReactor(reactor)
 
 
+    def test_processExitedRaises(self):
+        """
+        Test that a protocol raising an exception inside processExited
+        doesn't cause the reactor to explode.
+        """
+        reactor = self.buildReactor()
+
+        class TestException(Exception):
+            pass
+
+        class Protocol(ProcessProtocol):
+            def processExited(self, reason):
+                reactor.callLater(0.01, lambda: reactor.stop())
+                raise TestException("processedExited raised")
+
+        protocol = Protocol()
+        transport = reactor.spawnProcess(
+               protocol, sys.executable, [sys.executable, "-c", ""],
+               usePTY=self.usePTY)
+        reactor.run()
+        self.flushLoggedErrors()
+
+        # Manually clean-up broken process handler
+        for pid, handler in process.reapProcessHandlers.items():
+            if handler is not transport:
+                continue
+            process.unregisterReapProcessHandler(pid, handler)
+            self.fail("After processExited raised, transport was left in"
+                      " reapProcessHandlers")
+
+
 class ProcessTestsBuilder(ProcessTestsBuilderBase):
     """
     Builder defining tests relating to L{IReactorProcess} for child processes
Index: NEWS
===================================================================
--- NEWS	(revision 32118)
+++ NEWS	(working copy)
@@ -69,6 +69,9 @@
  - twisted.words.services referenced nonexistent
    twisted.words.protocols.irc.IRC_NOSUCHCHANNEL. This has been fixed.
    Related code has also received test cases. (#4915)
+ - Broken process handlers will no longer be left in the global state
+   if a ProcessProtocol raises an exception from within processExited.
+   (#5151)
 
 Improved Documentation
 ----------------------
