Index: twisted/internet/posixbase.py
===================================================================
--- twisted/internet/posixbase.py	(revision 32087)
+++ twisted/internet/posixbase.py	(working copy)
@@ -226,7 +226,10 @@
         method.
         """
         _FDWaker.doRead(self)
-        process.reapAllProcesses()
+        try:
+            process.reapAllProcesses()
+        except:
+            log.err()
 
 
 
Index: twisted/internet/_baseprocess.py
===================================================================
--- twisted/internet/_baseprocess.py	(revision 32087)
+++ twisted/internet/_baseprocess.py	(working copy)
@@ -34,7 +34,10 @@
                 _missingProcessExited % (qual(self.proto.__class__),),
                 DeprecationWarning, stacklevel=0)
         else:
-            processExited(Failure(reason))
+            try:
+                processExited(Failure(reason))
+            except:
+                err(None, "unexpected error in processExited")
 
 
     def processEnded(self, status):
