Index: twisted/internet/defer.py
===================================================================
--- twisted/internet/defer.py	(revision 15595)
+++ twisted/internet/defer.py	(working copy)
@@ -355,11 +355,15 @@
         return self.timeoutCall
 
     def __str__(self):
-        cname = self.__class__.__name__
+        res = ["<"]
+        res.extend([self.__class__.__name__, " at ", str(hex(unsignedID(self))),])
+        if self.callbacks:
+            res.append(" ")
+            res.append(', '.join([cb.__name__ for (cb, bA, cbK), errstuff in self.callbacks]))
         if hasattr(self, 'result'):
-            return "<%s at %s  current result: %r>" % (cname, hex(unsignedID(self)),
-                                                       self.result)
-        return "<%s at %s>" % (cname, hex(unsignedID(self)))
+            res.extend(["  current result: ", str(self.result),])
+        res.append(">")
+        return ''.join(res)
     __repr__ = __str__
 
 class DebugInfo:
