Index: twisted/test/test_internet.py
===================================================================
--- twisted/test/test_internet.py	(revision 28948)
+++ twisted/test/test_internet.py	(working copy)
@@ -870,7 +870,146 @@
         dc.reset(3)
         self.assertEquals(dc.getTime(), 13)
 
+class DelayedCallRichComparison(unittest.TestCase):
+    """
+    Test  L{base.DelayedCall}'s "rich comparison" methods.
+    """
+    def _getDelayedCallAt(self, time):
+        """
+        Get L{base.DelayedCall} instance at a given C{time}.
+        
+        @param time: passed directly to L{base.DelayedCall}
+        """
+        return base.DelayedCall(time, lambda: None, (), {}, lambda dc: None,
+                                lambda dc: None, None)
 
+    def setUp(self):
+        """
+        Create L{base.DelayedCall} instances that fail
+        C{super(base.DelayedCall).__lt__}-based comparison.  Otherwise
+        the tests won't work.
+        """
+        self.zero, self.one = map(self._getDelayedCallAt, (0, 1))
+
+        def lt():
+            """
+            Emulate C{super(base.DelayedCall).__lt__}-based comparison.
+
+            Current implementation assumes L{id}-based comparison.
+            """
+            return id(self.zero) < id(self.one)
+
+        if lt():
+            # make sure that `super(base.DelayedCall).__lt__`-based
+            # comparison based on `id` fails otherwise tests won't
+            # catch erronious fallback. 
+            self.zero, self.one = self.one, self.zero
+            self.zero.time, self.one.time = 0, 1
+        assert not lt() # it fails half of the time without the above swap
+
+    def test_lt(self):
+        """
+        Zero should be less than one.
+        """
+        zero, one = self.zero, self.one
+        self.assertTrue(zero < one)
+        self.assertFalse(one < zero)
+        self.assertFalse(zero < zero)
+        self.assertFalse(one < one)
+        # document current behavior
+        self.assertRaises(AttributeError, lambda: zero < 1)
+        self.assertFalse(0 < one)
+
+    def test_le(self):
+        """
+        Zero should be less than or equal one.
+        """
+        zero, one = self.zero, self.one
+        self.assertTrue(zero <= one)
+        self.assertFalse(one <= zero)
+        self.assertTrue(zero <= zero)
+        self.assertTrue(one <= one)
+        # document current behavior
+        self.assertRaises(AttributeError, lambda: zero <= 1)
+        self.assertFalse(0 <= one)
+    
+    def test_gt(self):
+        """
+        One should be greater than zero.
+        """
+        zero, one = self.zero, self.one
+        self.assertTrue(one > zero)
+        self.assertFalse(zero > one)
+        self.assertFalse(zero > zero)
+        self.assertFalse(one > one)
+        # document current behavior
+        self.assertRaises(AttributeError, lambda: zero < 1)
+        self.assertFalse(0 < one)
+
+    def test_ge(self):
+        """
+        One should be greater than or equal zero.
+        """
+        zero, one = self.zero, self.one
+        self.assertTrue(one >= zero)
+        self.assertFalse(zero >= one)
+        self.assertTrue(zero >= zero)
+        self.assertTrue(one >= one)
+        # document current behavior
+        self.assertRaises(AttributeError, lambda: 1 >= zero)
+        self.assertFalse(one >= 0)
+
+    def test_eq(self):
+        """
+        Object should be equal to itself.
+        Different L{base.DelayedCall} objects are not equal.        
+        """
+        #NOTE: use explicit `==` here, don't use `assertEquals()`. We
+        #test here `==` method. There is no guarantie that current or
+        #future `assertEquals()` uses `==` consistently (it might use
+        #`!=`)
+        self.assertFalse(self.zero == self.one)
+        self.assertTrue(self.zero == self.zero)
+        self.assertTrue(self.one == self.one)
+        # document current behavior
+        self.assertFalse(self.zero == self._getDelayedCallAt(self.zero.time))
+        self.assertFalse(self.one == self._getDelayedCallAt(self.one.time))
+
+    def test_ne(self):
+        """
+        Object should not be unequal to itself.
+        Different L{base.DelayedCall} objects are unequal.        
+        """
+        #NOTE: use explicit `!=` here, don't use `assertNotEquals()`
+        #We test here `!=` method. There is no guarantie that current
+        #or future `assertNotEquals()` uses `!=` consistently (it
+        #might use `==`)
+        self.assertTrue(self.zero != self.one)
+        self.assertFalse(self.zero != self.zero)
+        self.assertFalse(self.one != self.one)
+        # document current behavior
+        self.assertTrue(self.zero != self._getDelayedCallAt(self.zero.time))
+        self.assertTrue(self.one != self._getDelayedCallAt(self.one.time))
+
+    def test_hash(self):
+        """
+        Equal objects should have equal hashes therefore if hashes are
+        not equal then the objects should not be equal.        
+        Hash should not change during object lifetime.
+        """
+        # hash(a) != hash(b) => a != b
+        if hash(self.zero) != hash(self.one):
+            self.assertNotEquals(self.zero, self.one)
+        else: # nothing  follows from hash(a) == hash(b)
+            self.assertTrue(True)
+
+        # hash must not change
+        hash_zero, hash_one = map(hash ,(self.zero, self.one))
+        self.zero.time += 1 
+        self.one.time *= 10
+        self.assertEquals(hash(self.zero), hash_zero)
+        self.assertEquals(hash(self.one), hash_one)
+
 class CallFromThreadTests(unittest.TestCase):
     def testWakeUp(self):
         # Make sure other threads can wake up the reactor
Index: twisted/topfiles/3359.bugfix
===================================================================
--- twisted/topfiles/3359.bugfix	(revision 0)
+++ twisted/topfiles/3359.bugfix	(revision 0)
@@ -0,0 +1 @@
+twisted.internet.base.DelayedCall now behaves consistently when used in rich comparison operations
\ No newline at end of file
Index: twisted/internet/base.py
===================================================================
--- twisted/internet/base.py	(revision 28948)
+++ twisted/internet/base.py	(working copy)
@@ -151,8 +151,18 @@
         return not (self.cancelled or self.called)
 
     def __le__(self, other):
+        """
+        Implement `<=` comparison
+        """
+        #NOTE: don't use `not (other < self)` implementation. It
+        #affects whether AttributeError is raised or not
         return self.time <= other.time
 
+    def __lt__(self, other):
+        """
+        Implement `<` comparison
+        """
+        return self.time < other.time
 
     def __str__(self):
         if self._str is not None:
