Index: twisted/internet/base.py
===================================================================
--- twisted/internet/base.py	(revision 28724)
+++ twisted/internet/base.py	(working copy)
@@ -153,6 +153,8 @@
     def __le__(self, other):
         return self.time <= other.time
 
+    def __lt__(self, other):
+        return self.time < other.time
 
     def __str__(self):
         if self._str is not None:
