Index: pb-copyable.xhtml
===================================================================
--- pb-copyable.xhtml	(revision 38186)
+++ pb-copyable.xhtml	(working copy)
@@ -173,7 +173,7 @@
 <a href="listings/pb/copy_receiver.tac" class="py-listing">copy_receiver.tac</a>
 
 <p>The sending side has a class called <code>LilyPond</code>. To make this
-eligble for transport through <code>callRemote</code> (either as an
+eligible for transport through <code>callRemote</code> (either as an
 argument, a return value, or something referenced by either of those [like a
 dictionary value]), it must inherit from one of the four <code class="API"
 base="twisted.spread.pb">Serializable</code> classes. In this section,
