Index: doc/core/howto/plugin.xhtml
===================================================================
--- doc/core/howto/plugin.xhtml	(revision 32735)
+++ doc/core/howto/plugin.xhtml	(working copy)
@@ -156,7 +156,10 @@
     <code class="shell">PYTHONPATH</code> environment variable) contains a
     <em>directory</em> named <code class="shell">twisted/plugins/</code>,
     each <code class="shell">.py</code> file in that directory will be loaded
-    as a source of plugins.  This directory <em>must not</em> be a Python
+    as a source of plugins. Note that the filename of this plugin should be
+    unique; if there are multiple plugins with the same filename (but located in
+    different source directories) then the remaining plugins will be ignored.
+    The plugins directory <em>must not</em> be a Python
     package: including <code class="shell">__init__.py</code> will cause the
     directory to be skipped and no plugins loaded from it.  Second, each
     module in the installed version of Twisted's <code class="shell">
