diff --git a/doc/web/howto/web-in-60/http-auth.xhtml b/doc/web/howto/web-in-60/http-auth.xhtml
index 36f3921..7d61541 100644
--- a/doc/web/howto/web-in-60/http-auth.xhtml
+++ b/doc/web/howto/web-in-60/http-auth.xhtml
@@ -74,10 +74,10 @@ this case, it will be a <code>File</code>. In general, with <code>Guard</code>,
 the avatar must be a resource of some sort.</p>
 
 <pre class="python">
-def requestAvatar(self, avatarId, mind, *interfaces):
-    if IResource in interfaces:
-        return (IResource, File("/home/%s/public_html" % (avatarId,)), lambda: None)
-    raise NotImplementedError()
+    def requestAvatar(self, avatarId, mind, *interfaces):
+        if IResource in interfaces:
+            return (IResource, File("/home/%s/public_html" % (avatarId,)), lambda: None)
+        raise NotImplementedError()
 </pre>
 
 <p>A few notes on this method:</p>
