--- window.tac.orig	2010-07-26 07:05:41.000000000 -0700
+++ window.tac	2010-07-26 07:03:18.000000000 -0700
@@ -156,7 +157,8 @@
 
 
 def makeService(args):
-    checker = checkers.InMemoryUsernamePasswordDatabaseDontUse(username="password")
+    checker = checkers.InMemoryUsernamePasswordDatabaseDontUse()
+    checker.addUser("username", "password")
 
     f = protocol.ServerFactory()
     f.protocol = lambda: TelnetTransport(TelnetBootstrapProtocol,
