--- /usr/lib/python2.6/dist-packages/twisted/conch/insults/window.py	2008-07-29 13:13:54.000000000 -0700
+++ window.py	2010-07-26 06:24:21.000000000 -0700
@@ -730,7 +730,7 @@
 
 class ScrolledArea(Widget):
     def __init__(self, containee):
-        Widget.__init__(self, containee)
+        Widget.__init__(self)
         self._viewport = Viewport(containee)
         self._horiz = HorizontalScrollbar(self._horizScroll)
         self._vert = VerticalScrollbar(self._vertScroll)
