diff --git twisted/python/_release.py twisted/python/_release.py
index 8d80d45..970dc9f 100644
--- twisted/python/_release.py
+++ twisted/python/_release.py
@@ -1122,7 +1122,7 @@ class BuildAPIDocsScript(object):
     A thing for building API documentation. See L{main}.
     """
 
-    def buildAPIDocs(self, projectRoot, output):
+    def buildAPIDocs(self, projectRoot, output, versionString):
         """
         Build the API documentation of Twisted, with our project policy.
 
@@ -1130,8 +1130,6 @@ class BuildAPIDocsScript(object):
             checkout.
         @param output: A L{FilePath} pointing to the desired output directory.
         """
-        version = Project(projectRoot.child("twisted")).getVersion()
-        versionString = version.base()
         sourceURL = ("http://twistedmatrix.com/trac/browser/tags/releases/"
                      "twisted-%s" % (versionString,))
         apiBuilder = APIBuilder()
@@ -1139,7 +1137,7 @@ class BuildAPIDocsScript(object):
             "Twisted",
             "http://twistedmatrix.com/",
             sourceURL,
-            projectRoot.child("twisted"),
+            projectRoot,
             output)
 
 
@@ -1155,4 +1153,11 @@ class BuildAPIDocsScript(object):
         if len(args) != 2:
             sys.exit("Must specify two arguments: "
                      "Twisted checkout and destination path")
-        self.buildAPIDocs(FilePath(args[0]), FilePath(args[1]))
+
+        if args[0] == '.':
+            version = Project(projectRoot.child("twisted")).getVersion()
+            versionString = 'version.base()'
+        else:
+            versionString = 'development'
+
+        self.buildAPIDocs(FilePath(args[0]), FilePath(args[1]), versionString)
