Index: twisted/test/test_banana.py
===================================================================
--- twisted/test/test_banana.py	(revision 27772)
+++ twisted/test/test_banana.py	(working copy)
@@ -63,6 +63,7 @@
                 self.enc.sendEncoded(n)
                 self.enc.dataReceived(self.io.getvalue())
                 self.assertEqual(self.result, n)
+                self.assertEqual(isinstance(self.result, long), n > sys.maxint)
 
 
     def _getSmallest(self):
Index: twisted/spread/banana.py
===================================================================
--- twisted/spread/banana.py	(revision 27772)
+++ twisted/spread/banana.py	(working copy)
@@ -41,11 +41,9 @@
         num = ord(char)
         i = i + (num * (oneHundredAndTwentyEight ** place))
         place = place + 1
-    if i <= 2147483647:
-        return int(i)
-    else:
-        return i
 
+    return int(i)
+
 # delimiter characters.
 LIST     = chr(0x80)
 INT      = chr(0x81)
@@ -188,15 +186,15 @@
             elif typebyte == INT:
                 buffer = rest
                 num = b1282int(num)
-                gotItem(int(num))
+                gotItem(num)
             elif typebyte == LONGINT:
                 buffer = rest
                 num = b1282int(num)
-                gotItem(long(num))
+                gotItem(num)
             elif typebyte == LONGNEG:
                 buffer = rest
                 num = b1282int(num)
-                gotItem(-long(num))
+                gotItem(-num)
             elif typebyte == NEG:
                 buffer = rest
                 num = -b1282int(num)
