Index: twisted/test/test_banana.py
===================================================================
--- twisted/test/test_banana.py	(revision 27895)
+++ twisted/test/test_banana.py	(working copy)
@@ -58,15 +58,16 @@
         """
         for exp in (32, 64, 128, 256):
             for add in (0, 1):
-                n = 2 ** exp + add
-                self.io.truncate(0)
-                self.enc.sendEncoded(n)
-                self.enc.dataReceived(self.io.getvalue())
-                self.assertEquals(self.result, n)
-                if n > sys.maxint:
-                    self.assertIsInstance(self.result, long)
-                else:
-                    self.assertIsInstance(self.result, int)
+                m = 2 ** exp + add
+                for n in (m, -m-1):
+                    self.io.truncate(0)
+                    self.enc.sendEncoded(n)
+                    self.enc.dataReceived(self.io.getvalue())
+                    self.assertEquals(self.result, n)
+                    if n > sys.maxint or n < -sys.maxint - 1:
+                        self.assertIsInstance(self.result, long)
+                    else:
+                        self.assertIsInstance(self.result, int)
 
 
     def _getSmallest(self):
Index: twisted/spread/banana.py
===================================================================
--- twisted/spread/banana.py	(revision 27895)
+++ twisted/spread/banana.py	(working copy)
@@ -185,7 +185,7 @@
             elif typebyte == INT:
                 buffer = rest
                 num = b1282int(num)
-                gotItem(int(num))
+                gotItem(num)
             elif typebyte == LONGINT:
                 buffer = rest
                 num = b1282int(num)
@@ -193,7 +193,7 @@
             elif typebyte == LONGNEG:
                 buffer = rest
                 num = b1282int(num)
-                gotItem(-long(num))
+                gotItem(-num)
             elif typebyte == NEG:
                 buffer = rest
                 num = -b1282int(num)
