--- ftp.py.orig	Thu Jul  6 18:00:31 2006
+++ ftp.py	Thu Jul  6 18:26:58 2006
@@ -38,6 +38,10 @@
 from twisted.cred import error as cred_error, portal, credentials, checkers
 
 # constants
+# PASV ports range
+PASV_MINPORT = 44000
+PASV_MAXPORT = 44050
+
 # response codes
 
 RESTART_MARKER_REPLY                    = "100"
@@ -576,6 +580,8 @@
     dtpInstance = None
     binary      = True     # binary transfers? False implies ASCII. defaults to True
 
+    pasvMinPort = PASV_MINPORT
+    pasvMaxPort = PASV_MAXPORT
 
     def reply(self, key, *args):
         msg = RESPONSE[key] % args
@@ -674,6 +680,15 @@
             else:
                 return BAD_CMD_SEQ, "RNTO required after RNFR"
 
+    def getDtpPort(self, factory):
+        for portn in range(self.pasvMinPort, self.pasvMaxPort):
+            try:
+                dtpPort = reactor.listenTCP(portn, factory)
+            except error.CannotListenError:
+                continue
+            else:
+                return dtpPort
+        return None
 
     def ftp_USER(self, username):
         """
@@ -744,7 +759,7 @@
             self.cleanupDTP()
         self.dtpFactory = DTPFactory(pi=self)
         self.dtpFactory.setTimeout(self.dtpTimeout)
-        self.dtpPort = reactor.listenTCP(0, self.dtpFactory)
+        self.dtpPort = self.getDtpPort(self.dtpFactory)
 
         host = self.transport.getHost().host
         port = self.dtpPort.getHost().port
@@ -1154,6 +1169,9 @@
     userAnonymous = 'anonymous'
     timeOut = 600
 
+    pasvMinPort = PASV_MINPORT
+    pasvMaxPort = PASV_MAXPORT
+
     welcomeMessage = "Twisted %s FTP Server" % (copyright.version,)
 
     def __init__(self, portal=None, userAnonymous='anonymous'):
@@ -1166,6 +1184,8 @@
         if p is not None:
             p.wrappedProtocol.portal = self.portal
             p.wrappedProtocol.timeOut = self.timeOut
+        p.pasvMinPort = self.pasvMinPort
+        p.pasvMaxPort = self.pasvMaxPort
         return p
 
     def stopFactory(self):
