diff --git a/twisted/web/http_headers.py b/twisted/web/http_headers.py
index f0790b9..f4f3242 100644
--- a/twisted/web/http_headers.py
+++ b/twisted/web/http_headers.py
@@ -10,7 +10,7 @@ from __future__ import division, absolute_import
 
 from collections import MutableMapping
 
-from twisted.python.compat import comparable, cmp
+from twisted.python.compat import comparable, cmp, StringType
 
 
 def _dashCapitalize(name):
@@ -54,6 +54,8 @@ class _DictHeaders(MutableMapping):
         """
         Set the given header.
         """
+        if isinstance(key, StringType):
+            key = key.encode('ascii')
         self._headers.setRawHeaders(key, [value])
 
 
diff --git a/twisted/web/test/test_http_headers.py b/twisted/web/test/test_http_headers.py
index 7a12a06..3aabcba 100644
--- a/twisted/web/test/test_http_headers.py
+++ b/twisted/web/test/test_http_headers.py
@@ -560,9 +560,6 @@ class HeaderDictTests(TestCase):
         self.assertEqual(headers.getRawHeaders(b"foo"), [b"panda"])
         self.assertEqual(headers.getRawHeaders(b"bar"), [b"marmot"])
 
-    if _PY3:
-        test_updateWithKeywords.skip = "Not yet supported on Python 3; see #6082."
-
 
     def test_setdefaultMissing(self):
         """
