Index: twisted/python/log.py
===================================================================
--- twisted/python/log.py	(revision 27376)
+++ twisted/python/log.py	(working copy)
@@ -18,8 +18,12 @@
 
 from twisted.python import util, context, reflect
 
+logSwitch = True
+"""
+This enables turning on or off the logging system,
+saving some cpu cycles and improving somewhat performances
+"""	
 
-
 class ILogContext:
     """
     Actually, this interface is just a synoym for the dictionary interface,
@@ -73,22 +77,24 @@
     Utility method which wraps a function in a try:/except:, logs a failure if
     one occurrs, and uses the system's logPrefix.
     """
-    try:
-        lp = logger.logPrefix()
-    except KeyboardInterrupt:
-        raise
-    except:
-        lp = '(buggy logPrefix method)'
-        err(system=lp)
-    try:
-        return callWithContext({"system": lp}, func, *args, **kw)
-    except KeyboardInterrupt:
-        raise
-    except:
-        err(system=lp)
+    if logSwitch:    
+		try:
+		    lp = logger.logPrefix()
+		except KeyboardInterrupt:
+		    raise
+		except:
+		    lp = '(buggy logPrefix method)'
+		    err(system=lp)
+		try:
+			return callWithContext({"system": lp}, func, *args, **kw)
+		except KeyboardInterrupt:
+		    raise
+		except:
+		    err(system=lp)
+    else:
+        func(*args,**kw)
 
 
-
 _keepErrors = 0
 _keptErrors = []
 _ignoreErrors = []
