Index: twisted/conch/client/knownhosts.py
===================================================================
--- twisted/conch/client/knownhosts.py	(revision 28294)
+++ twisted/conch/client/knownhosts.py	(working copy)
@@ -9,10 +9,10 @@
 """
 
 from binascii import Error as DecodeError, b2a_base64
+import hmac
 
 from zope.interface import implements
 
-from Crypto.Hash.HMAC import HMAC
 from Crypto.Hash import SHA
 
 from twisted.python.randbytes import secureRandom
@@ -202,7 +202,7 @@
     """
     Return the SHA-1 HMAC hash of the given key and string.
     """
-    hash = HMAC(key, digestmod=SHA)
+    hash = hmac.new(key, digestmod=SHA)
     hash.update(string)
     return hash.digest()
 
