Index: twisted/protocols/amp.py
===================================================================
--- twisted/protocols/amp.py	(revision 30779)
+++ twisted/protocols/amp.py	(working copy)
@@ -2405,10 +2405,7 @@
     @return: The converted dictionary mapping names to encoded argument
     strings (identical to C{strings}).
     """
-    myObjects = {}
-    for (k, v) in objects.items():
-        myObjects[k] = v
-
+    myObjects = objects.copy()
     for argname, argparser in arglist:
         argparser.toBox(argname, strings, myObjects, proto)
     return strings
