Index: doc/conch/examples/sshsimpleserver.py
===================================================================
--- doc/conch/examples/sshsimpleserver.py	(revision 23101)
+++ doc/conch/examples/sshsimpleserver.py	(working copy)
@@ -89,10 +89,10 @@
 
 class ExampleFactory(factory.SSHFactory):
     publicKeys = {
-        'ssh-rsa': keys.getPublicKeyString(data=publicKey)
+        'ssh-rsa': keys.Key.fromString(data=publicKey)
     }
     privateKeys = {
-        'ssh-rsa': keys.getPrivateKeyObject(data=privateKey)
+        'ssh-rsa': keys.Key.fromString(data=privateKey)
     }
     services = {
         'ssh-userauth': userauth.SSHUserAuthServer,
