Index: twisted/test/test_plugin.py
===================================================================
--- twisted/test/test_plugin.py	(revision 13207)
+++ twisted/test/test_plugin.py	(working copy)
@@ -70,7 +70,7 @@
         try:
             os.remove(sibpath(plugins.__file__, 'dropin.cache'))
         except OSError, ose:
-            if ose.errno == errno.EPERM:
+            if ose.errno in (errno.EACCES, errno.ENOENT):
                 print 'Testing in deployed mode.'
             else:
                 raise
@@ -124,9 +124,9 @@
         writeFileName = sibpath(plugins.__file__, 'pluginextra.py')
         try:
             wf = file(writeFileName, 'w')
-        except OSError, ose:
-            if ose.errno == errno.EPERM:
-                raise unittest.SkiptTest(
+        except IOError, ioe:
+            if ioe.errno == errno.EACCES:
+                raise unittest.SkipTest(
                     "No permission to add things to twisted.plugins")
             else:
                 raise
@@ -161,8 +161,15 @@
 
     def _testDetectFilesChanged(self):
         writeFileName = sibpath(plugins.__file__, 'pluginextra.py')
-        writeFile = file(writeFileName, 'w')
         try:
+            writeFile = file(writeFileName, 'w')
+        except IOError, ioe:
+            if ioe.errno == errno.EACCES:
+                raise unittest.SkipTest(
+                    "No permission to add things to twisted.plugins")
+            else:
+                raise
+        try:
             writeFile.write(begintest)
             writeFile.flush()
             plgs = list(plugin.getPlugIns(plugin.ITestPlugin))
