Index: twisted/web/client.py
===================================================================
--- twisted/web/client.py	(revision 22257)
+++ twisted/web/client.py	(working copy)
@@ -308,6 +308,7 @@
         self.waiting = 1
 
     def gotHeaders(self, headers):
+        HTTPClientFactory.gotHeaders(self, headers)
         if self.requestedPartial:
             contentRange = headers.get("content-range", None)
             if not contentRange:
