Index: twisted/conch/ssh/filetransfer.py
===================================================================
--- twisted/conch/ssh/filetransfer.py	(revision 37793)
+++ twisted/conch/ssh/filetransfer.py	(working copy)
@@ -472,14 +472,14 @@
         to be sent to the server.
         """
         FileTransferBase.__init__(self)
-        self.extData = {}
+        self.extData = extData
         self.counter = 0
         self.openRequests = {} # id -> Deferred
         self.wasAFile = {} # Deferred -> 1 TERRIBLE HACK
 
     def connectionMade(self):
         data = struct.pack('!L', max(self.versions))
-        for k,v in self.extData.itervalues():
+        for k,v in self.extData.iteritems():
             data += NS(k) + NS(v)
         self.sendPacket(FXP_INIT, data)
 
