Index: twisted/conch/test/test_filetransfer.py
===================================================================
--- twisted/conch/test/test_filetransfer.py	(revision 37793)
+++ twisted/conch/test/test_filetransfer.py	(working copy)
@@ -763,3 +763,30 @@
         """
         self.assertEqual(result[0], 'msg')
         self.assertEqual(result[1], '')
+
+
+class TestFileTransferClientMakeConnection(SFTPTestBase):
+    """
+    Test for L{filetransfer.FileTransferClient} makeConnection().
+    """
+    def setUp(self):
+        SFTPTestBase.setUp(self)
+        self.avatar = FileTransferTestAvatar(self.testDir)
+        self.server = filetransfer.FileTransferServer(avatar=self.avatar)
+        self.clientTransport = loopback.LoopbackRelay(self.server)
+        extData = {"test_key":"test_value"}
+        self.client = filetransfer.FileTransferClient(extData)
+        self.assertEqual(self.client.extData, extData)
+        self.serverTransport = loopback.LoopbackRelay(self.client)
+
+    def tearDown(self):
+        self.serverTransport.loseConnection()
+        self.clientTransport.loseConnection()
+        self.serverTransport.clearBuffer()
+        self.clientTransport.clearBuffer()
+
+    def test_makeConnection(self):
+        self.client.makeConnection(self.clientTransport)
+        self.server.makeConnection(self.serverTransport)
+
+
