*** conch/ssh/filetransfer.py	2011-09-09 13:48:29.000000000 -0400
--- conch/ssh/filetransfer-new.py	2013-03-22 05:03:55.400744923 -0400
***************
*** 472,485 ****
          to be sent to the server.
          """
          FileTransferBase.__init__(self)
!         self.extData = {}
          self.counter = 0
          self.openRequests = {} # id -> Deferred
          self.wasAFile = {} # Deferred -> 1 TERRIBLE HACK
  
      def connectionMade(self):
          data = struct.pack('!L', max(self.versions))
!         for k,v in self.extData.itervalues():
              data += NS(k) + NS(v)
          self.sendPacket(FXP_INIT, data)
  
--- 472,485 ----
          to be sent to the server.
          """
          FileTransferBase.__init__(self)
!         self.extData = extData
          self.counter = 0
          self.openRequests = {} # id -> Deferred
          self.wasAFile = {} # Deferred -> 1 TERRIBLE HACK
  
      def connectionMade(self):
          data = struct.pack('!L', max(self.versions))
!         for k,v in self.extData.iteritems():
              data += NS(k) + NS(v)
          self.sendPacket(FXP_INIT, data)
  
