Index: doc/web/howto/client.xhtml
===================================================================
--- doc/web/howto/client.xhtml	(revision 28074)
+++ doc/web/howto/client.xhtml	(working copy)
@@ -60,7 +60,7 @@
       responsible for resolving the hostname into an IP address and connecting
       to it on port 80.  It is also responsible for cleaning up the connection
       afterwards.  This code sends a request which includes one custom header,
-      I{User-Agent}.  The last argument passed to <code>Agent.request</code> is
+      <em>User-Agent</em>.  The last argument passed to <code>Agent.request</code> is
       <code>None</code>, though, so the request has no body.
     </p>
 
