diff --git a/docs/conch/examples/sshsimpleserver.py b/docs/conch/examples/sshsimpleserver.py
index 1212db1f8..1e9e757ad 100755
--- a/docs/conch/examples/sshsimpleserver.py
+++ b/docs/conch/examples/sshsimpleserver.py
@@ -7,8 +7,10 @@ from twisted.cred import portal
 from twisted.cred.checkers import InMemoryUsernamePasswordDatabaseDontUse
 from twisted.conch import avatar
 from twisted.conch.checkers import SSHPublicKeyChecker, InMemorySSHKeyDB
-from twisted.conch.ssh import factory, userauth, connection, keys, session
+from twisted.conch.ssh import factory, filetransfer, userauth, connection, keys, session
 from twisted.conch.ssh.transport import SSHServerTransport
+from twisted.conch.unix import SFTPServerForUnixConchUser
+
 from twisted.internet import reactor, protocol
 from twisted.python import log
 from twisted.python import components
@@ -100,8 +102,24 @@ class ExampleAvatar(avatar.ConchUser):
         avatar.ConchUser.__init__(self)
         self.username = username
         self.channelLookup.update({b'session':session.SSHSession})
+        self.subsystemLookup.update(
+                {b"sftp": filetransfer.FileTransferServer})
+
+    def getHomeDir(self):
+        return b'/tmp/test'
 
+    def _runAsUser(self, method, *args, **kwargs):
+        try:
+            calls = iter(method)
+        except TypeError:
+            calls = [(method, args, kwargs)]
 
+        for call in calls:
+            method = call[0]
+            args = len(call) > 1 and call[1] or ()
+            kwargs = len(call) > 2 and call[2] or {}
+            result = method(*args, **kwargs)
+        return result
 
 @implementer(portal.IRealm)
 class ExampleRealm(object):
@@ -184,7 +202,8 @@ class ExampleSession(object):
 
 
 components.registerAdapter(ExampleSession, ExampleAvatar, session.ISession)
-
+components.registerAdapter(
+    SFTPServerForUnixConchUser, ExampleAvatar, filetransfer.ISFTPServer)
 
 
 class ExampleFactory(factory.SSHFactory):
diff --git a/src/twisted/conch/unix.py b/src/twisted/conch/unix.py
index d9c3f05c8..6a49435c7 100644
--- a/src/twisted/conch/unix.py
+++ b/src/twisted/conch/unix.py
@@ -371,7 +371,7 @@ class SFTPServerForUnixConchUser:
 
     def _absPath(self, path):
         home = self.avatar.getHomeDir()
-        return os.path.join(nativeString(home.path), nativeString(path))
+        return os.path.join(nativeString(home), nativeString(path))
 
 
     def gotVersion(self, otherVersion, extData):
