Index: utils.py
===================================================================
RCS file: /cvs/Twisted/twisted/web/woven/utils.py,v
retrieving revision 1.15
diff -u -r1.15 utils.py
--- utils.py	23 Jul 2003 19:55:28 -0000	1.15
+++ utils.py	6 Aug 2003 23:19:22 -0000
@@ -10,7 +10,9 @@
     if not fail:
         fail = failure.Failure()
     log.err(fail)
-    request.write(webutil.formatFailure(fail))
+    
+    if request.site.displayTraceback:
+        request.write(webutil.formatFailure(fail))
     #request.finish()
 
 
