Index: server.py
===================================================================
RCS file: /cvs/Twisted/twisted/web/server.py,v
retrieving revision 1.101
diff -u -r1.101 server.py
--- server.py	21 Jul 2003 04:09:52 -0000	1.101
+++ server.py	7 Aug 2003 00:04:55 -0000
@@ -235,10 +235,15 @@
 
     def processingFailed(self, reason):
         log.err(reason)
-        body = ("<html><head><title>web.Server Traceback (most recent call last)</title></head>"
-                "<body><b>web.Server Traceback (most recent call last):</b>\n\n"
-                "%s\n\n</body></html>\n"
-                % webutil.formatFailure(reason))
+        if self.site.displayTraceback:
+            body = ("<html><head><title>web.Server Traceback (most recent call last)</title></head>"
+                    "<body><b>web.Server Traceback (most recent call last):</b>\n\n"
+                    "%s\n\n</body></html>\n"
+                    % webutil.formatFailure(reason))
+        else:
+            body = ("<html><head><title>Processing Failed</title></head><body>"
+                  "<b>Processing Failed</b></body></html>")
+
         self.setResponseCode(http.INTERNAL_SERVER_ERROR)
         self.setHeader('content-type',"text/html")
         self.setHeader('content-length', str(len(body)))
@@ -417,7 +422,8 @@
 
     counter = 0
     requestFactory = Request
-
+    displayTraceback = True
+    
     def __init__(self, resource, logPath=None, timeout=60*60*12):
         """Initialize.
         """
