Index: win32/build.aap
===================================================================
--- win32/build.aap	(révision 25387)
+++ win32/build.aap	(copie de travail)
@@ -1,5 +1,5 @@
 
-@supported_pythons = ('py2.5', 'py2.4')
+@supported_pythons = ('py2.6', 'py2.5', 'py2.4')
 
 all: ../main.aap
     @for p in _no.supported_pythons:
Index: win32/main.aap
===================================================================
--- win32/main.aap	(révision 25387)
+++ win32/main.aap	(copie de travail)
@@ -57,6 +57,10 @@
 # PyVersion is variant rather than a regular string variable so the script
 # will accept only one of these two values
 :variant PyVersion
+    py2.6
+        pyversion=2.6
+        pyversion10=26
+        pyhome=`_no.pathdb["python26"]`
     py2.5
         pyversion=2.5
         pyversion10=25
Index: win32/aapwin.py
===================================================================
--- win32/aapwin.py	(révision 25387)
+++ win32/aapwin.py	(copie de travail)
@@ -8,7 +8,7 @@
         return default
     if valuename is None:
         return key.value
-    try:    
+    try:
         return key.values[valuename].value
     except winreg.ValueNotFound:
         return default
@@ -49,7 +49,7 @@
         _innohome = getValueFromReg(r'%s\%s' % (regkey, _four),
                                     "Inno Setup: App Path",
                                     None)
-    if _innohome is None: 
+    if _innohome is None:
         return fallback
     else:
         return _innohome
@@ -58,6 +58,7 @@
 
 pathdb = dict(innohome=innohome,
               iscc=os.path.join(innohome, "ISCC.exe"),
+              python26=getPythonHomeForVersion('2.6'),
               python25=getPythonHomeForVersion('2.5'),
               python24=getPythonHomeForVersion('2.4'),
               )
