Index: twisted/mail/smtp.py
===================================================================
--- twisted/mail/smtp.py	（版本 27359）
+++ twisted/mail/smtp.py	（工作副本）
@@ -1416,7 +1416,7 @@
             self._okresponse = self.smtpState_from
             self.sendLine(encode_base64(resp, eol=""))
 
-        if auth.getName() == "LOGIN" and challenge == "Username:":
+        if auth.getName() == "LOGIN" and challenge.lower() == "username:":
             self._expected = [334]
             self._authinfo = auth
             self._okresponse = self.esmtpState_challenge
@@ -1762,9 +1762,9 @@
         return "LOGIN"
 
     def challengeResponse(self, secret, chal):
-        if chal== "Username:":
+        if chal.lower() == "username:":
             return self.user
-        elif chal == 'Password:':
+        elif chal.lower() == 'password:':
             return secret
 
 class PLAINAuthenticator:
