Index: twisted/web/http.py
===================================================================
--- twisted/web/http.py	(revision 27546)
+++ twisted/web/http.py	(working copy)
@@ -710,7 +710,22 @@
         # serialized and sent with the request so CGIs will work remotely
         self.client = self.channel.transport.getPeer()
         self.host = self.channel.transport.getHost()
+        
+        self.parseRequestBody()
+        
+        self.process()
 
+
+    def __repr__(self):
+        return '<%s %s %s>'% (self.method, self.uri, self.clientproto)
+
+    def parseRequestBody(self):
+        """
+        Read the request body and parse any available POST data.
+        
+        Can be overridden in subclasses.
+        This method is not intended for users.
+        """
         # Argument processing
         args = self.args
         ctype = self.requestHeaders.getRawHeaders('content-type')
@@ -738,12 +753,6 @@
                     raise
             self.content.seek(0, 0)
 
-        self.process()
-
-
-    def __repr__(self):
-        return '<%s %s %s>'% (self.method, self.uri, self.clientproto)
-
     def process(self):
         """Override in subclasses.
 
