=== modified file 'twisted/protocols/_c_urlarg.c'
--- twisted/protocols/_c_urlarg.c	2004-08-25 09:02:51 +0000
+++ twisted/protocols/_c_urlarg.c	2009-12-01 07:24:28 +0000
@@ -23,7 +23,7 @@
 
 static PyObject* UrlargError;
 
-#define OUTPUTCHAR(c,n) PycStringIO->cwrite(output, c, n)
+#define OUTPUTCHAR(c,n) PycStringIO->cwrite(output, (const char *)c, n)
 
 #define STATE_INITIAL 0
 #define STATE_PERCENT 1

