Index: interfaces.py
===================================================================
--- interfaces.py	(revision 18499)
+++ interfaces.py	(working copy)
@@ -566,12 +566,19 @@
         """Fire 'startup' System Events, move the reactor to the 'running'
         state, then run the main loop until it is stopped with stop() or
         crash().
+
+        @return: The process exit status. 
+                 This will be the value passed to L{stop}, unless a
+                 signal is catched, in which case 1 is returned.
         """
 
-    def stop():
+    def stop(exitStatus=0):
         """Fire 'shutdown' System Events, which will move the reactor to the
-        'stopped' state and cause reactor.run() to exit. """
+        'stopped' state and cause reactor.run() to exit. 
 
+        @param exitStatus: The exitStatus that should be returned by L{run}.
+        """
+
     def crash():
         """Stop the main loop *immediately*, without firing any system events.
 
