Index: twisted/internet/win32eventreactor.py
===================================================================
--- twisted/internet/win32eventreactor.py	(revision 14837)
+++ twisted/internet/win32eventreactor.py	(working copy)
@@ -211,8 +211,9 @@
     main.installReactor(r)
 
 _cmdLineQuoteRe = re.compile(r'(\\*)"')
-def _cmdLineQuote(s):
-    return '"' + _cmdLineQuoteRe.sub(r'\1\1\\"', s) + '"'
+def _cmdLineQuote(s):
+    quote = ((" " in s) or ("\t" in s)) and '"' or ''
+    return quote + _cmdLineQuoteRe.sub(r'\1\1\\"', s) + quote
 
 class Process(abstract.FileDescriptor):
     """A process that integrates with the Twisted event loop.
