Index: twisted/mail/mail.py
===================================================================
--- twisted/mail/mail.py	(revision 17515)
+++ twisted/mail/mail.py	(working copy)
@@ -62,16 +62,16 @@
         return len(self.domains)
 
     def __str__(self):
-        return '<DomainWithDefaultsDict %s>' % (self.domains,)
+        return '<DomainWithDefaultDict %s>' % (self.domains,)
 
     def __repr__(self):
-        return 'DomainWithDefaultsDict(%s)>' % (self.domains,)
+        return 'DomainWithDefaultDict(%s)>' % (self.domains,)
 
     def get(self, key, default=None):
         return self.domains.get(key, default)
 
     def copy(self):
-        return DomainWithDefaultsDict(self.domains.copy(), self.default)
+        return DomainWithDefaultDict(self.domains.copy(), self.default)
 
     def iteritems(self):
         return self.domains.iteritems()
