Index: twisted/internet/test/test_posixprocess.py
===================================================================
--- twisted/internet/test/test_posixprocess.py	(revision 29557)
+++ twisted/internet/test/test_posixprocess.py	(working copy)
@@ -43,10 +43,12 @@
         for f in sys.stdin, sys.stdout, sys.stderr:
             self.assertIn(f.fileno(), openfds)
 
+        # See http://twistedmatrix.com/trac/ticket/4522#comment:17
         f = file(os.devnull)
-        new_openfds = process._listOpenFDs()
-        self.assertTrue(len(new_openfds) > len(openfds))
-
-        f.close()
-        new_openfds = process._listOpenFDs()
-        self.assertEqual(len(new_openfds), len(openfds))
+        fd = os.dup(f.fileno())
+        try:
+            f.close()
+            self.assertIn(fd, process._listOpenFDs())
+        finally:
+            os.close(fd)
+        self.assertNotIn(fd, process._listOpenFDs())
