diff --git a/twisted/internet/process.py b/twisted/internet/process.py
index c5f1ac3..4a34ad0 100644
--- a/twisted/internet/process.py
+++ b/twisted/internet/process.py
@@ -463,6 +463,33 @@ class _BaseProcess(BaseProcess, object):
         return "<%s pid=%s status=%s>" % (self.__class__.__name__,
                                           self.pid, self.status)
 
+def listOpenFDs():
+    """
+    Return an iterable of potentially open file descriptors.
+
+    This function returns an iterable over the contents of /dev/fd or
+    /proc/<pid>/fd, if they're available on the platform. If they're not, the
+    returned value is the range [0, maxfds], where 'maxfds' is at least 256.
+    """
+    dname = "/dev/fd"
+    try:
+        return tuple(int(fd) for fd in os.listdir(dname))
+    except:
+        dname = "/proc/%d/fd" % os.getpid()
+        try:
+            return tuple(int(fd) for fd in os.listdir(dname))
+        except:
+            try:
+                import resource
+                maxfds = resource.getrlimit(resource.RLIMIT_NOFILE)[1] + 1
+                # OS-X reports 9223372036854775808. That's a lot of fds
+                # to close
+                if maxfds > 1024:
+                    maxfds = 1024
+            except:
+                maxfds = 256
+            return xrange(maxfds)
+
 class Process(_BaseProcess):
     """
     An operating-system Process.
@@ -601,7 +628,8 @@ class Process(_BaseProcess):
         This is accomplished in two steps::
 
             1. close all file descriptors that aren't values of fdmap.  This
-               means 0 .. maxfds.
+               means 0 .. maxfds (or just the open fds within that range, if
+               the platform supports '/proc/<pid>/fd').
 
             2. for each childFD::
 
@@ -625,16 +653,7 @@ class Process(_BaseProcess):
             errfd.write("starting _setupChild\n")
 
         destList = fdmap.values()
-        try:
-            import resource
-            maxfds = resource.getrlimit(resource.RLIMIT_NOFILE)[1] + 1
-            # OS-X reports 9223372036854775808. That's a lot of fds to close
-            if maxfds > 1024:
-                maxfds = 1024
-        except:
-            maxfds = 256
-
-        for fd in xrange(maxfds):
+        for fd in listOpenFDs():
             if fd in destList:
                 continue
             if debug and fd == errfd.fileno():
diff --git a/twisted/internet/test/test_process.py b/twisted/internet/test/test_process.py
index 06197c0..ac2fa67 100644
--- a/twisted/internet/test/test_process.py
+++ b/twisted/internet/test/test_process.py
@@ -476,6 +476,42 @@ class ProcessTestsBuilder(ProcessTestsBuilderBase):
 
         reactor.callWhenRunning(spawnChild)
         self.runReactor(reactor)
+
+
+    def test_listOpenFDs(self):
+        """
+        A spawned process has only stdin, stdout and stderr open
+        (file descriptor 3 is also reported as open, because of the call to
+        'os.listdir()').
+        """
+        here = os.path.dirname(__file__)
+        top = os.path.normpath(os.path.join(here, "..", "..", ".."))
+        source = (
+            "import sys",
+            "sys.path.insert(0, '%s')" % top,
+            "from twisted.internet import process",
+            "sys.stdout.write(str(process.listOpenFDs()))",
+            "sys.stdout.flush()")
+
+        reactor = self.buildReactor()
+
+        def processFinished(output):
+            self.assertEqual('(0, 1, 2, 3)', output)
+
+        def shutdown(result):
+            reactor.stop()
+            return result
+
+        def spawnChild():
+            print top
+            d = succeed(None)
+            d.addCallback(lambda dummy: utils.getProcessOutput(
+                sys.executable, ["-c", "\n".join(source)], reactor=reactor))
+            d.addCallback(processFinished)
+            d.addBoth(shutdown)
+
+        reactor.callWhenRunning(spawnChild)
+        self.runReactor(reactor)
 globals().update(ProcessTestsBuilder.makeTestCaseClasses())
 
 
diff --git a/twisted/test/test_process.py b/twisted/test/test_process.py
index 48918be..00cec9d 100644
--- a/twisted/test/test_process.py
+++ b/twisted/test/test_process.py
@@ -1533,6 +1533,12 @@ class MockOS(object):
         """
         return 0, 0, 1, 2
 
+    def listdir(self, path):
+        """
+        Override C{os.listdir}, returning fake contents of '/dev/fd'
+        """
+        return "-1", "-2"
+
 
 
 if process is not None:
