diff --git a/twisted/mail/bounce.py b/twisted/mail/bounce.py
index c0f55d5..5d5bde9 100644
--- a/twisted/mail/bounce.py
+++ b/twisted/mail/bounce.py
@@ -6,7 +6,6 @@
 
 import StringIO
 import rfc822
-import string
 import time
 import os
 
@@ -40,7 +39,7 @@ I've given up, and I will not retry the message again.
 
     boundary = "%s_%s_%s" % (time.time(), os.getpid(), 'XXXXX')
     failedAddress = rfc822.AddressList(failedTo)[0][1]
-    failedDomain = string.split(failedAddress, '@', 1)[1]
+    failedDomain = failedAddress.split('@', 1)[1]
     messageID = smtp.messageid(uniq='bounce')
     ctime = time.ctime(time.time())
 
diff --git a/twisted/mail/imap4.py b/twisted/mail/imap4.py
index d9fde29..6ca8384 100644
--- a/twisted/mail/imap4.py
+++ b/twisted/mail/imap4.py
@@ -1919,7 +1919,7 @@ class IMAP4Server(basic.LineReceiver, policies.TimeoutMixin):
                 sign = "+"
             else:
                 sign = "-"
-            odate = odate + sign + string.zfill(str(((abs(ttup[9]) / 3600) * 100 + (abs(ttup[9]) % 3600) / 60)), 4)
+            odate = odate + sign + str(((abs(ttup[9]) / 3600) * 100 + (abs(ttup[9]) % 3600) / 60)).zfill(4)
         _w('INTERNALDATE ' + _quote(odate))
 
     def spew_rfc822header(self, id, msg, _w=None, _f=None):
diff --git a/twisted/mail/pop3.py b/twisted/mail/pop3.py
index 48c91fa..3b65242 100644
--- a/twisted/mail/pop3.py
+++ b/twisted/mail/pop3.py
@@ -11,7 +11,6 @@ Post-office Protocol version 3
 @author: Jp Calderone
 """
 
-import string
 import base64
 import binascii
 import warnings
@@ -351,7 +350,7 @@ class POP3(basic.LineOnlyReceiver, policies.TimeoutMixin):
             self.blocked.append((command, args))
             return
 
-        command = string.upper(command)
+        command = command.upper()
         authCmd = command in self.AUTH_CMDS
         if not self.mbox and not authCmd:
             raise POP3Error("not authenticated yet: cannot do " + command)
