Index: doc/core/howto/plugin.xhtml
===================================================================
--- doc/core/howto/plugin.xhtml	(revision 30565)
+++ doc/core/howto/plugin.xhtml	(working copy)
@@ -33,16 +33,15 @@
 
     <h2>Installing a plugin</h2>
 
-    <p>Before using a Twisted plugin, make sure the following packages
-    are installed:</p>
-    <p><ul>
-    <li><a href='http://pypi.python.org/pypi/zope.interface'>zope.interface</a></li>
-    </ul></p>
+    <p>
+      Before using a Twisted plugin, make sure the <a
+      href='http://pypi.python.org/pypi/zope.interface'>zope.interface</a>
+      package is installed. You can install it using <a
+      href="http://pip.openplans.org/">pip</a>or, less ideally, using the
+      <code>easy_install</code> tool provided by <a
+      href='http://pypi.python.org/pypi/setuptools'>setuptools</a>.
+    </p>
 
-    <p>You can install these packages using the <code>easy_install</code>
-    tool provided by <a href='http://pypi.python.org/pypi/setuptools'>setuptools</a>
-    .</p>
-
     <h2>Writing Extensible Programs</h2>
 
     <p>Taking advantage of <code class="API">twisted.plugin</code> is
@@ -57,10 +56,10 @@
         </p>
 
         <p>
-        A convention for defining interfaces is do so in a file named like
+        A convention for defining interfaces is to do so in a file named like
         <em>ProjectName/projectname/iprojectname.py</em>.  The rest of this
         document will follow that convention: consider the following
-        interface definition be in <code>Matsim/matsim/imatsim.py</code>, an
+        interface definition to be in <code>Matsim/matsim/imatsim.py</code>, an
         interface definition module for a hypothetical material simulation
         package.
         </p>
