Index: config/application
===================================================================
RCS file: /cvs/TwistedWebDeb/config/application,v
retrieving revision 1.3
diff -u -u -r1.3 application
--- config/application	1 Oct 2003 09:13:53 -0000	1.3
+++ config/application	14 Oct 2003 13:22:05 -0000
@@ -1,6 +1,6 @@
 # You will not normally need to modify this file.
 import pwd, grp
-from twisted.application import service, strports, 
+from twisted.application import service, strports
 
 # Calculate user-id and group-id from names
 uid = pwd.getpwnam(username)[2]
Index: debian/rules
===================================================================
RCS file: /cvs/TwistedWebDeb/debian/rules,v
retrieving revision 1.16
diff -u -u -r1.16 rules
--- debian/rules	29 May 2003 16:32:54 -0000	1.16
+++ debian/rules	14 Oct 2003 13:22:05 -0000
@@ -26,7 +26,7 @@
 	dh_installdirs
 	dh_installdebconf
 	dh_installman man/rptwisted.8 man/twconfcheck.8
-	install -m 644 config/ports config/ssl-ports config/twisted-web \
+	install -m 644 config/ports config/user config/twisted-web \
 	               config/default config/root config/site \
 	               config/application config/directories \
 	               ${DIR}/etc/twisted-web/
