Index: log.py
===================================================================
--- log.py	(revision 13941)
+++ log.py	(working copy)
@@ -238,6 +238,7 @@
     removeObserver = theLogPublisher.removeObserver
     msg = theLogPublisher.msg
 
+import locale
 
 class FileLogObserver:
     """Log observer that writes to a file-like object.
@@ -245,10 +246,12 @@
     @ivar timeFormat: Format string passed to strftime()
     """
     timeFormat = "%Y/%m/%d %H:%M %Z"
+    localeEncoding = locale.getpreferredencoding()
 
     def __init__(self, f):
         self.write = f.write
         self.flush = f.flush
+        self.encoding = f.encoding
 
     def emit(self, eventDict):
         edm = eventDict['message']
@@ -273,6 +276,10 @@
             text = ' '.join(map(str, edm))
 
         timeStr = time.strftime(self.timeFormat, time.localtime(eventDict['time']))
+        if sys.platform == 'win32':
+            timeStr = unicode(timeStr, self.localeEncoding)
+            timeStr = timeStr.encode(self.encoding)
+
         fmtDict = {'system': eventDict['system'], 'text': text.replace("\n", "\n\t")}
         msgStr = " [%(system)s] %(text)s\n" % fmtDict
 
