diff --git a/twisted/web/client.py b/twisted/web/client.py
index d1b85e1..9b47a2d 100644
--- a/twisted/web/client.py
+++ b/twisted/web/client.py
@@ -750,7 +750,7 @@ if not _PY3:
 try:
     from OpenSSL import SSL
 
-    from twisted.internet.ssl import CertificateOptions
+    from twisted.internet.ssl import CertificateOptions, OpenSSLDefaultPaths
 except ImportError:
     class WebClientContextFactory(object):
         """
@@ -766,7 +766,10 @@ else:
         certificate verification.
         """
         def __init__(self):
-            self._contextFactory = CertificateOptions(method=SSL.SSLv23_METHOD)
+            self._contextFactory = CertificateOptions(
+                method=SSL.SSLv23_METHOD,
+                trustRoot=OpenSSLDefaultPaths(),
+            )
 
         def getContext(self, hostname, port):
             """
diff --git a/twisted/web/test/test_agent.py b/twisted/web/test/test_agent.py
index 0f880b8..3ab7519 100644
--- a/twisted/web/test/test_agent.py
+++ b/twisted/web/test/test_agent.py
@@ -26,6 +26,7 @@ from twisted.internet.error import ConnectionLost
 from twisted.internet.protocol import Protocol, Factory
 from twisted.internet.defer import Deferred, succeed, CancelledError
 from twisted.internet.endpoints import TCP4ClientEndpoint, SSL4ClientEndpoint
+from twisted.internet.ssl import OpenSSLDefaultPaths
 from twisted.web.client import FileBodyProducer, Request, HTTPConnectionPool
 from twisted.web.client import _WebToNormalContextFactory, ResponseDone
 from twisted.web.client import WebClientContextFactory, _HTTP11ClientFactory
@@ -1154,6 +1155,16 @@ class WebClientContextFactoryTests(TestCase):
         self.assertIsInstance(ctx, ssl.SSL.Context)
 
 
+    def test_setsTrustRootOnContextToDefaultTrustRoot(self):
+        """
+        The L{CertificateOptions} has C{trustRoot} set to the default trust
+        roots.
+        """
+        ctx = WebClientContextFactory()
+        self.assertIsInstance(
+            ctx._contextFactory.trustRoot, OpenSSLDefaultPaths)
+
+
     if ssl is None:
         test_returnsContext.skip = "SSL not present, cannot run SSL tests."
     else:
diff --git a/twisted/web/topfiles/7042.bugfix b/twisted/web/topfiles/7042.bugfix
new file mode 100644
index 0000000..1dcf343
--- /dev/null
+++ b/twisted/web/topfiles/7042.bugfix
@@ -0,0 +1 @@
+twisted.web.client.Agent now verifies that the provided server certificate in a TLS connection is trusted by the platform.
