Index: twisted/web/http.py
===================================================================
--- twisted/web/http.py	(revision 35841)
+++ twisted/web/http.py	(working copy)
@@ -1519,6 +1519,7 @@
         self.setTimeout(self.timeOut)
 
     def lineReceived(self, line):
+        print line
         self.resetTimeout()
 
         if self.__first_line:
Index: twisted/web/test/test_http.py
===================================================================
--- twisted/web/test/test_http.py	(revision 35841)
+++ twisted/web/test/test_http.py	(working copy)
@@ -692,7 +692,14 @@
             channel.transport.value(),
             "HTTP/1.1 400 Bad Request\r\n\r\n")
 
+    def test_invalidHeaders(self):
+        requestLines = ["GET / HTTP/1.0", "Content-Length: x"]
+        channel = self.runRequest("\n".join(requestLines), http.Request, 0)
+        self.assertEqual(
+            channel.transport.value(),
+            "HTTP/1.1 400 Bad Request\r\n\r\n")
 
+
     def test_headerLimitPerRequest(self):
         """
         L{HTTPChannel} enforces the limit of C{HTTPChannel.maxHeaders} per
@@ -864,7 +871,7 @@
                 # The tempfile API used to create content returns an
                 # instance of a different type depending on what platform
                 # we're running on.  The point here is to verify that the
-                # request body is in a file that's on the filesystem. 
+                # request body is in a file that's on the filesystem.
                 # Having a fileno method that returns an int is a somewhat
                 # close approximation of this. -exarkun
                 testcase.assertIsInstance(self.content.fileno(), int)
