--- main.py	Mon Nov 22 09:02:54 2004
+++ new_main.py	Wed Dec 14 20:43:50 2005
@@ -20,8 +20,12 @@
     # this stuff should be common to all reactors.
     import twisted.internet
     import sys
-    assert not sys.modules.has_key('twisted.internet.reactor'), \
-           "reactor already installed"
+    if sys.modules.has_key('twisted.internet.reactor'):
+        if twisted.internet.reactor != reactor:
+            raise RuntimeError, "reactor already installed"
+        else:
+            # the same reactor is already installed, so do nothing
+            return
     twisted.internet.reactor = reactor
     sys.modules['twisted.internet.reactor'] = reactor
 
