Index: test/test_http_headers.py
===================================================================
--- test/test_http_headers.py	(revision 19182)
+++ test/test_http_headers.py	(working copy)
@@ -494,6 +494,14 @@
     def testUserAgent(self):
         self.runRoundtripTest("User-Agent", (("CERN-LineMode/2.15 libwww/2.17b3", "CERN-LineMode/2.15 libwww/2.17b3"),))
 
+    def testContentType(self):
+        table = (
+            ('multipart/form-data; boundary="===============1318316892"',
+                http_headers.MimeType('multipart', 'form-data', {'boundary': '===============1318316892'}),
+                ['multipart/form-data;boundary="===============1318316892"']
+            ),
+        )
+        self.runRoundtripTest("Content-Type", table)
 
 class ResponseHeaderParsingTests(HeaderParsingTestBase):
     def testAcceptRanges(self):
