Index: http_headers.py
===================================================================
--- http_headers.py	(revision 19182)
+++ http_headers.py	(working copy)
@@ -402,7 +402,33 @@
         if v is None:
             l.append('%s' % k)
         else:
-            l.append('%s=%s' % (k,v))
+            # FROM http://tools.ietf.org/html/rfc2046#section-5.1.1
+            # 
+            # WARNING TO IMPLEMENTORS:  The grammar for parameters on the Content-
+            # type field is such that it is often necessary to enclose the boundary
+            # parameter values in quotes on the Content-type line.  This is not
+            # always necessary, but never hurts. Implementors should be sure to
+            # study the grammar carefully in order to avoid producing invalid
+            # Content-type fields.  Thus, a typical "multipart" Content-Type header
+            # field might look like this:
+            #
+            # Content-Type: multipart/mixed; boundary=gc0p4Jq0M2Yt08j34c0p
+            #
+            # But the following is not valid:
+            #
+            # Content-Type: multipart/mixed; boundary=gc0pJq0M:08jU534c0p
+            #
+            # (because of the colon) and must instead be represented as
+            #
+            # Content-Type: multipart/mixed; boundary="gc0pJq0M:08jU534c0p"
+            
+            # FROM http://tools.ietf.org/html/rfc822#section-3.3
+            #
+            # quoted-string = <"> *(qtext/quoted-pair) <">
+            # qtext =  <any CHAR excepting <">, "\" & CR, and including linear-white-space>
+            # quoted-pair =  "\" CHAR
+            v = v.replace('\\', r'\\').replace('"', r'\"')#.replace('\r', '\\\r')
+            l.append('%s="%s"' % (k,v))
     return ";".join(l)
 
 
