Index: twisted/test/test_compat.py
===================================================================
--- twisted/test/test_compat.py	(revision 38418)
+++ twisted/test/test_compat.py	(working copy)
@@ -638,7 +638,7 @@
         """
         self.callDeprecated(
             (Version("Twisted", 13, 1, 0), socket.inet_ntop),
-            compat.inet_ntop, socket.AF_INET, "\x01\x00\x01\x00")
+            compat.inet_ntop, socket.AF_INET, b'\x01\x00\x01\x00')
 
     def test_inet_pton(self):
         """
