Index: doc/core/howto/tutorial/listings/finger/finger14.tac
===================================================================
--- doc/core/howto/tutorial/listings/finger/finger14.tac	(revision 28884)
+++ doc/core/howto/tutorial/listings/finger/finger14.tac	(working copy)
@@ -20,7 +20,6 @@
     def __init__(self, filename):
         self.users = {}
         self.filename = filename
-        self._read()
 
     def _read(self):
         for line in file(self.filename):
