Index: setup.py
===================================================================
RCS file: /cvs/Twisted/setup.py,v
retrieving revision 1.149
diff -u -r1.149 setup.py
--- setup.py	17 Oct 2003 21:08:01 -0000	1.149
+++ setup.py	16 Nov 2003 16:39:14 -0000
@@ -85,7 +85,13 @@
         self._detect_modules()
         build_ext.build_extensions(self)
 
-
+    def _remove_conftest(self):
+        for filename in ("conftest.c", "conftest.o", "conftest.obj"):
+            try:
+                os.unlink(filename)
+            except:
+                pass
+                
     def _compile_helper(self, content):
         conftest = open("conftest.c", "w")
         try:
@@ -98,12 +104,8 @@
                 return False
             return True
         finally:
-            try:
-                os.unlink("conftest.c")
-                os.unlink("conftest.o")
-            except:
-                pass
-
+            self._remove_conftest()
+            
     def _check_define(self, include_files, define_name):
         """
         Check if the given name has been #define'd by trying to compile a
