=== modified file 'twisted/mail/mail.py'
--- twisted/mail/mail.py	2011-02-14 04:45:15 +0000
+++ twisted/mail/mail.py	2012-06-19 10:31:46 +0000
@@ -137,11 +137,6 @@
     def addUser(user, password):
         """Add a username/password to this domain."""
 
-    def startMessage(user):
-        """Create and return a new message to be delivered to the given user.
-
-        DEPRECATED.  Implement validateTo() correctly instead.
-        """
 
     def getCredentialsCheckers():
         """Return a list of ICredentialsChecker implementors for this domain.

=== added file 'twisted/mail/topfiles/4151.removal'
--- twisted/mail/topfiles/4151.removal	1970-01-01 00:00:00 +0000
+++ twisted/mail/topfiles/4151.removal	2012-06-19 10:38:42 +0000
@@ -0,0 +1,1 @@
+twisted.mail.mail.IDomain.startMessage has been removed.

