Index: twisted/runner/topfiles/setup.py
===================================================================
--- twisted/runner/topfiles/setup.py	(revision 40627)
+++ twisted/runner/topfiles/setup.py	(working copy)
@@ -11,7 +11,8 @@
 extensions = [
     Extension("twisted.runner.portmap",
               ["twisted/runner/portmap.c"],
-              condition=lambda builder: builder._check_header("rpc/rpc.h")),
+              condition=lambda builder:
+                  builder._check_header("rpc/rpc.h") and builder._check_header("Python.h")),
 ]
 
 if __name__ == '__main__':
