--- /home/rgl/twisted/INSTALL/lib/python2.5/site-packages/twisted/protocols/portforward.py	2005-01-29 01:48:43.000000000 +0000
+++ portforward.py	2008-07-12 13:15:41.000000000 +0000
@@ -31,6 +31,14 @@
 class ProxyClient(Proxy):
     def connectionMade(self):
         self.peer.setPeer(self)
+
+        # Wire this and the peer transport together to enable
+        # flow control (this stops connections from filling
+        # this proxy memory when one side produces data at a
+        # higher rate than the other can consume).
+        self.transport.registerProducer(self.peer.transport, True)
+        self.peer.transport.registerProducer(self.transport, True)
+
         # We're connected, everybody can read to their hearts content.
         self.peer.transport.resumeProducing()
 
