Index: twisted/python/log.py
===================================================================
--- twisted/python/log.py	(revision 39660)
+++ twisted/python/log.py	(working copy)
@@ -181,6 +181,10 @@
         >>> log.msg('Hello ', 'world.')
 
         This form only works (sometimes) by accident.
+
+        @param message: The message to log.
+        @param kw: Keyword arguments that will be converted to keys in the
+            event dict, which is in turn passed to L{ILogObserver}s.
         """
         actualEventDict = (context.get(ILogContext) or {}).copy()
         actualEventDict.update(kw)
