Index: twisted/conch/ssh/filetransfer.py
===================================================================
--- twisted/conch/ssh/filetransfer.py	(revision 25177)
+++ twisted/conch/ssh/filetransfer.py	(working copy)
@@ -868,7 +868,7 @@
 FILEXFER_ATTR_SIZE        = 0x00000001
 FILEXFER_ATTR_OWNERGROUP  = 0x00000002
 FILEXFER_ATTR_PERMISSIONS = 0x00000004
-FILEXFER_ATTR_ACMODTIME   = 0x00000009
+FILEXFER_ATTR_ACMODTIME   = 0x00000008
 FILEXFER_ATTR_EXTENDED    = 0x80000000L
 
 FILEXFER_TYPE_REGULAR        = 1
