Index: test_process.py
===================================================================
--- test_process.py	(revision 22460)
+++ test_process.py	(working copy)
@@ -73,6 +73,13 @@
         p = OutProtocol()
         p.childDataReceived(1, bytes)
         self.assertEqual(received, [bytes])
+        
+    def test_implmentsInterface(self):
+        """
+        Verifies ProcessProtocol implements IProcessProtocol
+        
+        """
+        self.assertTrue(interfaces.IProcessProtocol.implementedBy(StubProcessProtocol))
 
 
     def test_errReceived(self):
