diff --git a/twisted/conch/ssh/keys.py b/twisted/conch/ssh/keys.py
--- a/twisted/conch/ssh/keys.py
+++ b/twisted/conch/ssh/keys.py
@@ -402,7 +402,7 @@ class Key(object):
         lines = ['<%s %s (%s bits)' % (self.type(),
             self.isPublic() and 'Public Key' or 'Private Key',
             self.keyObject.size())]
-        for k, v in self.data().items():
+        for k, v in sorted(self.data().items()):
             lines.append('attr %s:' % k)
             by = common.MP(v)[4:]
             while by:
diff --git a/twisted/conch/test/test_keys.py b/twisted/conch/test/test_keys.py
--- a/twisted/conch/test/test_keys.py
+++ b/twisted/conch/test/test_keys.py
@@ -484,15 +484,15 @@ SUrCyZXsNh6VXwjs3gKQ
         """
         self.assertEqual(repr(keys.Key(self.rsaObj)),
 """<RSA Private Key (0 bits)
+attr d:
+\t03
 attr e:
 \t02
-attr d:
-\t03
 attr n:
 \t01
+attr p:
+\t04
 attr q:
 \t05
-attr p:
-\t04
 attr u:
 \t04>""")
