Index: doc/core/howto/testing.xhtml
===================================================================
--- doc/core/howto/testing.xhtml	(revision 37642)
+++ doc/core/howto/testing.xhtml	(working copy)
@@ -164,6 +164,25 @@
 scope in a test module or a module imported by a test module) then it 
  <em>will</em> raise an exception.</p>
 
+<h3>Parallel test</h3>
+
+<p>In many situations, your unit tests may run faster if they are allowed to
+run in parallel, such that blocking I/O calls allow other tests to continue.
+Trial, like unittest, supports the -j parameter.  Run <code>trial -j 3</code>
+to run 3 test runners at the same time.</p>
+
+<p>This requires care in your test creation.  Obviously, you need to ensure that
+your code is otherwise content to work in a parallel fashion while working within
+Twisted... and if you are using weird global variables in places, parallel tests
+might reveal this.</p>
+
+<p>However, if you have a test that fires up a schema on an external database
+in the <code>setUp</code> function, does some operations on it in the test, and
+then deletes that schema in the tearDown function, your tests will behave in an
+unpredictable fashion as they tromp upon each other if they have their own 
+schema.  And this won't actually indicate a real error in your code, merely a 
+testing-specific race-condition.</p>
+
   </body>
 </html>
 
