Index: twisted/internet/task.py
===================================================================
--- twisted/internet/task.py	(revision 33404)
+++ twisted/internet/task.py	(working copy)
@@ -36,6 +36,9 @@
         something else, but it probably ought to be set *before*
         calling L{start}.
 
+    @type running: C{boolean}
+    @ivar running: A flag indicating whether this instance is running.
+
     @type _expectNextCallAt: C{float}
     @ivar _expectNextCallAt: The time at which this instance most recently
         scheduled itself to run.
