Index: twisted/web/static.py
===================================================================
--- twisted/web/static.py	(revision 24969)
+++ twisted/web/static.py	(working copy)
@@ -249,7 +249,10 @@
             return self.childNotFound
 
         if path:
-            fpath = self.child(path)
+            try:            
+                fpath = self.child(path)
+            except filepath.InsecurePath:
+                return self.childNotFound
         else:
             fpath = self.childSearchPreauth(*self.indexNames)
             if fpath is None:
