Index: twisted/python/usage.py
===================================================================
--- twisted/python/usage.py	(revision 14251)
+++ twisted/python/usage.py	(working copy)
@@ -139,7 +139,8 @@
         print self.__str__()
         sys.exit(0)
 
-    def opt_version(self):
+    def opt_version(self):
+        """Display the version of Twisted."""
         from twisted import copyright
         print "Twisted version:", copyright.version
         sys.exit(0)
Index: twisted/scripts/trial.py
===================================================================
--- twisted/scripts/trial.py	(revision 14251)
+++ twisted/scripts/trial.py	(working copy)
@@ -95,8 +95,8 @@
                 ["until-failure", "u", "Repeat test until it fails"],
                 ["recurse", "R", "Search packages recursively"],
                 ['psyco', None, 'run tests with psyco.full() (EXPERIMENTAL)'],
-                ['verbose', 'v', 'verbose color output (default)'],
-                ['bwverbose', 'o', 'Colorless verbose output'],
+                ['verbose', 'v', 'verbose color output (default on non-windows)'],
+                ['bwverbose', 'o', 'Colorless verbose output (default on windows)'],
                 ['summary', 's', 'minimal summary output'],
                 ['text', 't', 'terse text output'],
                 ['timing', None, 'Timing output'],
@@ -128,7 +128,9 @@
         # by this point
         assert "twisted.internet.reactor" not in sys.modules
         usage.Options.__init__(self)
-        self._logObserver = None
+        self._logObserver = None
+        if sys.platform == "win32":
+            self["bwverbose"] = True
         self['modules'] = []
         self['packages'] = []
         self['testcases'] = []
