Index: twisted/mail/test/test_pop3.py
===================================================================
--- twisted/mail/test/test_pop3.py	(revision 18805)
+++ twisted/mail/test/test_pop3.py	(working copy)
@@ -1055,3 +1055,17 @@
             d, a = self.pop3Server.mbox.waiting.pop()
             d.callback(a)
         ValueErrorCommandTestCase._flush(self)
+
+class POP3MiscTestCase(unittest.TestCase):
+    """
+    Miscellaneous tests more to do with module/package structure than
+    anything to do with the Post Office Protocol.
+    """
+    def test_all(self):
+        """
+        This test checks that all names listed in
+        twisted.mail.pop3.__all__ are actually present in the module.
+        """
+        mod = twisted.mail.pop3
+        for attr in mod.__all__:
+            self.failUnless(hasattr(mod, attr))
Index: twisted/mail/pop3.py
===================================================================
--- twisted/mail/pop3.py	(revision 18805)
+++ twisted/mail/pop3.py	(working copy)
@@ -1068,7 +1068,7 @@
     'ServerErrorResponse', 'LineTooLong',
 
     # Protocol classes
-    'POP3', 'POP3Client', 'AdvancedPOP3Client'
+    'POP3', 'POP3Client', 'AdvancedPOP3Client',
 
     # Misc
     'APOPCredentials', 'Mailbox']
