Index: twisted/mail/pop3.py
===================================================================
--- twisted/mail/pop3.py	(revision 18713)
+++ twisted/mail/pop3.py	(working copy)
@@ -1063,7 +1063,7 @@
     'ServerErrorResponse', 'LineTooLong',
 
     # Protocol classes
-    'POP3', 'POP3Client', 'AdvancedPOP3Client'
+    'POP3', 'POP3Client', 'AdvancedPOP3Client',
 
     # Misc
     'APOPCredentials', 'Mailbox']
