Index: doc/core/howto/tutorial/listings/finger/finger20.tac
===================================================================
--- doc/core/howto/tutorial/listings/finger/finger20.tac	(revision 30465)
+++ doc/core/howto/tutorial/listings/finger/finger20.tac	(working copy)
@@ -237,7 +237,11 @@
     def getUsers(self):
         return defer.succeed(self.users.keys())
 
+    def stopService(self):
+        service.Service.stopService(self)
+        self.call.cancel()
 
+
 application = service.Application('finger', uid=1, gid=1)
 f = FingerService('/etc/users')
 serviceCollection = service.IServiceCollection(application)
Index: doc/core/howto/tutorial/listings/finger/finger21.tac
===================================================================
--- doc/core/howto/tutorial/listings/finger/finger21.tac	(revision 30465)
+++ doc/core/howto/tutorial/listings/finger/finger21.tac	(working copy)
@@ -264,7 +264,11 @@
     def getUsers(self):
         return defer.succeed(self.users.keys())
 
+    def stopService(self):
+        service.Service.stopService(self)
+        self.call.cancel()
 
+
 application = service.Application('finger', uid=1, gid=1)
 f = FingerService('/etc/users')
 serviceCollection = service.IServiceCollection(application)
Index: doc/core/howto/tutorial/listings/finger/finger14.tac
===================================================================
--- doc/core/howto/tutorial/listings/finger/finger14.tac	(revision 30465)
+++ doc/core/howto/tutorial/listings/finger/finger14.tac	(working copy)
@@ -46,6 +46,7 @@
         f.getUser = self.getUser
         return f
 
+
 application = service.Application('finger', uid=1, gid=1)
 f = FingerService('/etc/users')
 finger = internet.TCPServer(79, f.getFingerFactory())
Index: doc/core/howto/tutorial/listings/finger/finger15.tac
===================================================================
--- doc/core/howto/tutorial/listings/finger/finger15.tac	(revision 30465)
+++ doc/core/howto/tutorial/listings/finger/finger15.tac	(working copy)
@@ -67,6 +67,11 @@
         r = FingerResource(self.users)
         return r
 
+    def stopService(self):
+        service.Service.stopService(self)
+        self.call.cancel()
+
+
 application = service.Application('finger', uid=1, gid=1)
 f = FingerService('/etc/users')
 serviceCollection = service.IServiceCollection(application)
Index: doc/core/howto/tutorial/listings/finger/finger16.tac
===================================================================
--- doc/core/howto/tutorial/listings/finger/finger16.tac	(revision 30465)
+++ doc/core/howto/tutorial/listings/finger/finger16.tac	(working copy)
@@ -80,6 +80,11 @@
         f.getUser = self.getUser
         return f
 
+    def stopService(self):
+        service.Service.stopService(self)
+        self.call.cancel()
+
+
 application = service.Application('finger', uid=1, gid=1)
 f = FingerService('/etc/users')
 serviceCollection = service.IServiceCollection(application)
Index: doc/core/howto/tutorial/listings/finger/finger17.tac
===================================================================
--- doc/core/howto/tutorial/listings/finger/finger17.tac	(revision 30465)
+++ doc/core/howto/tutorial/listings/finger/finger17.tac	(working copy)
@@ -80,6 +80,11 @@
         f.getUser = self.getUser
         return f
 
+    def stopService(self):
+        service.Service.stopService(self)
+        self.call.cancel()
+
+
 application = service.Application('finger', uid=1, gid=1)
 f = FingerService('/etc/users')
 serviceCollection = service.IServiceCollection(application)
Index: doc/core/howto/tutorial/listings/finger/finger18.tac
===================================================================
--- doc/core/howto/tutorial/listings/finger/finger18.tac	(revision 30465)
+++ doc/core/howto/tutorial/listings/finger/finger18.tac	(working copy)
@@ -125,7 +125,11 @@
         f.getUser = self.getUser
         return f
 
+    def stopService(self):
+        service.Service.stopService(self)
+        self.call.cancel()
 
+
 application = service.Application('finger', uid=1, gid=1)
 f = FingerService('/etc/users')
 serviceCollection = service.IServiceCollection(application)
Index: doc/core/howto/tutorial/listings/finger/finger19b.tac
===================================================================
--- doc/core/howto/tutorial/listings/finger/finger19b.tac	(revision 30465)
+++ doc/core/howto/tutorial/listings/finger/finger19b.tac	(working copy)
@@ -230,6 +230,11 @@
     def getUsers(self):
         return defer.succeed(self.users.keys())
 
+    def stopService(self):
+        service.Service.stopService(self)
+        self.call.cancel()
+
+
 # Another back-end
 
 class LocalFingerService(service.Service):
Index: doc/core/howto/tutorial/listings/finger/finger19c.tac
===================================================================
--- doc/core/howto/tutorial/listings/finger/finger19c.tac	(revision 30465)
+++ doc/core/howto/tutorial/listings/finger/finger19c.tac	(working copy)
@@ -231,6 +231,11 @@
     def getUsers(self):
         return defer.succeed(self.users.keys())
 
+    def stopService(self):
+        service.Service.stopService(self)
+        self.call.cancel()
+
+
 # Yet another back-end
 
 class LocalFingerService(service.Service):
Index: doc/core/howto/tutorial/listings/finger/finger22.py
===================================================================
--- doc/core/howto/tutorial/listings/finger/finger22.py	(revision 30465)
+++ doc/core/howto/tutorial/listings/finger/finger22.py	(working copy)
@@ -265,7 +265,11 @@
     def getUsers(self):
         return defer.succeed(self.users.keys())
 
+    def stopService(self):
+        service.Service.stopService(self)
+        self.call.cancel()
 
+
 class ServerContextFactory:
 
     def getContext(self):
