Index: twisted/words/protocols/jabber/xmlstream.py
===================================================================
--- twisted/words/protocols/jabber/xmlstream.py	(revision 25272)
+++ twisted/words/protocols/jabber/xmlstream.py	(working copy)
@@ -391,6 +391,7 @@
 
         self.xmlstream.removeObserver('/failure', self.onFailure)
         ctx = ssl.CertificateOptions()
+        ctx.getContext().set_options(0x4000)
         self.xmlstream.transport.startTLS(ctx)
         self.xmlstream.reset()
         self.xmlstream.sendHeader()
