Index: twisted/words/protocols/jabber/xmlstream.py
===================================================================
--- twisted/words/protocols/jabber/xmlstream.py	(revision 24958)
+++ twisted/words/protocols/jabber/xmlstream.py	(working copy)
@@ -388,7 +388,10 @@
         """
 
         self.xmlstream.removeObserver('/failure', self.onFailure)
-        ctx = ssl.CertificateOptions()
+        # Use a SSLv2 handshake which indicates that SSLv3 and TLSv1 are also
+        # support. Doing the default TLSv1 handshake doesn't work with gtalk
+        # :(
+        ctx = ssl.CertificateOptions(method = ssl.SSL.SSLv23_METHOD)
         self.xmlstream.transport.startTLS(ctx)
         self.xmlstream.reset()
         self.xmlstream.sendHeader()
