Index: scripts/trial.py
===================================================================
--- scripts/trial.py    (revision 18550)
+++ scripts/trial.py    (working copy)
@@ -6,6 +6,7 @@
 
 import sys, os, random, gc, time, sets
 
+from twisted.internet import reactor
 from twisted.internet import defer
 from twisted.application import app
 from twisted.python import usage, reflect, failure
@@ -337,7 +338,7 @@
                               workingDirectory=config['temp-directory'])
 
 
-def run():
+def _run():
     if len(sys.argv) == 1:
         sys.argv.append("--help")
     config = Options()
@@ -359,3 +360,9 @@
                               coverdir=config.coverdir)
     sys.exit(not test_result.wasSuccessful())
 
+def run():
+    reactor.callLater( 0, _run )
+    reactor.run( )
