diff --git a/twisted/python/components.py b/twisted/python/components.py
index 2e715ba..001545e 100644
--- a/twisted/python/components.py
+++ b/twisted/python/components.py
@@ -370,7 +370,7 @@ class _ProxiedClassMethod(object):
     @type orginalAttribute: C{str}
     """
     def __init__(self, methodName, originalAttribute):
-        self.methodName = methodName
+        self.methodName = self.__name__ = methodName
         self.originalAttribute = originalAttribute
 
 
diff --git a/twisted/python/test/test_components.py b/twisted/python/test/test_components.py
index 1b09b6f..1519052 100644
--- a/twisted/python/test/test_components.py
+++ b/twisted/python/test/test_components.py
@@ -8,6 +8,8 @@ Test cases for Twisted component architecture.
 
 from __future__ import division, absolute_import
 
+from functools import wraps
+
 from zope.interface import Interface, implementer, Attribute
 from zope.interface.adapter import AdapterRegistry
 
@@ -655,6 +657,41 @@ class ProxyForInterfaceTests(unittest.SynchronousTestCase):
         self.assertEqual(yayable.yays, 2)
 
 
+    def test_decoratedProxyMethod(self):
+        """
+        Methods of the class created from L{proxyForInterface} can be used with
+        the decorator-helper L{functools.wraps}.
+        """
+        base = proxyForInterface(IProxiedInterface)
+        class klass(base):
+            @wraps(base.yay)
+            def yay(self):
+                self.original.yays += 1
+                return base.yay(self)
+
+        original = Yayable()
+        yayable = klass(original)
+        yayable.yay()
+        self.assertEqual(2, original.yays)
+
+
+    def test_decoratedProxyMethod_returns_wrappees_docstring(self):
+        """
+        Methods of the class created from L{proxyForInterface} used with
+        the decorator-helper L{functools.wraps} return the wrappee's docstring.
+        """
+        base = proxyForInterface(IProxiedInterface)
+        class klass(base):
+            @wraps(base.yay)
+            def yay(self):
+                """
+                I say yay when called.
+                """
+                pass
+
+        self.assertEqual(klass.yay.__doc__, u"I say yay when called")
+
+
     def test_proxyAttribute(self):
         """
         Proxy objects should proxy declared attributes, but not other
