Index: twisted/protocols/htb.py
===================================================================
--- twisted/protocols/htb.py	(revision 46300)
+++ twisted/protocols/htb.py	(working copy)
@@ -15,12 +15,13 @@
     HOWTO<http://lartc.org/howto/lartc.qdisc.classless.html#AEN682>}
 """
 
+from __future__ import division, print_function
 
 # TODO: Investigate whether we should be using os.times()[-1] instead of
 # time.time.  time.time, it has been pointed out, can go backwards.  Is
 # the same true of os.times?
 from time import time
-from zope.interface import implements, Interface
+from zope.interface import implementer, Interface
 
 from twisted.protocols import pcp
 
@@ -117,6 +118,8 @@
         @returntype: L{Bucket}
         """
 
+
+@implementer(IBucketFilter)
 class HierarchicalBucketFilter:
     """
     Filter things into buckets that can be nested.
@@ -127,8 +130,6 @@
     @type sweepInterval: C{int}
     """
 
-    implements(IBucketFilter)
-
     bucketFactory = Bucket
     sweepInterval = None
 
Index: twisted/protocols/pcp.py
===================================================================
--- twisted/protocols/pcp.py	(revision 46300)
+++ twisted/protocols/pcp.py	(working copy)
@@ -6,11 +6,14 @@
 Producer-Consumer Proxy.
 """
 
-from zope.interface import implements
+from __future__ import division, print_function
 
+from zope.interface import implementer
+
 from twisted.internet import interfaces
 
 
+@implementer(interfaces.IProducer, interfaces.IConsumer)
 class BasicProducerConsumerProxy:
     """
     I can act as a man in the middle between any Producer and Consumer.
@@ -22,7 +25,6 @@
     @ivar paused: As a Producer, am I paused?
     @type paused: bool
     """
-    implements(interfaces.IProducer, interfaces.IConsumer)
 
     consumer = None
     producer = None
Index: twisted/python/dist3.py
===================================================================
--- twisted/python/dist3.py	(revision 46300)
+++ twisted/python/dist3.py	(working copy)
@@ -139,6 +139,7 @@
     "twisted.positioning.nmea",
     "twisted.protocols",
     "twisted.protocols.basic",
+    "twisted.protocols.htb",
     "twisted.protocols.policies",
     "twisted.protocols.test",
     "twisted.protocols.tls",
@@ -321,6 +322,7 @@
     "twisted.test.testutils",
     "twisted.test.test_failure",
     "twisted.test.test_fdesc",
+    "twisted.test.test_htb",
     "twisted.test.test_internet",
     "twisted.test.test_iosim",
     "twisted.test.test_iutils",
Index: twisted/test/test_htb.py
===================================================================
--- twisted/test/test_htb.py	(revision 46300)
+++ twisted/test/test_htb.py	(working copy)
@@ -1,10 +1,16 @@
-# -*- Python -*-
+# Copyright (c) Twisted Matrix Laboratories.
+# See LICENSE for details.
 
-__version__ = '$Revision: 1.3 $'[11:-2]
+"""
+Tests for C{twisted.protocols.htb}.
+"""
 
+from __future__ import division, print_function
+
 from twisted.trial import unittest
 from twisted.protocols import htb
 
+
 class DummyClock:
     time = 0
     def set(self, when):
@@ -13,10 +19,12 @@
     def __call__(self):
         return self.time
 
+
 class SomeBucket(htb.Bucket):
     maxburst = 100
     rate = 2
 
+
 class TestBucketBase(unittest.TestCase):
     def setUp(self):
         self._realTimeFunc = htb.time
@@ -26,6 +34,7 @@
     def tearDown(self):
         htb.time = self._realTimeFunc
 
+
 class BucketTests(TestBucketBase):
     def testBucketSize(self):
         """Testing the size of the bucket."""
@@ -54,6 +63,7 @@
         empty = b.drip()
         self.assertTrue(empty)
 
+
 class BucketNestingTests(TestBucketBase):
     def setUp(self):
         TestBucketBase.setUp(self)
@@ -83,8 +93,9 @@
 
 # TODO: Test the Transport stuff?
 
-from test_pcp import DummyConsumer
+from twisted.test.test_pcp import DummyConsumer
 
+
 class ConsumerShaperTests(TestBucketBase):
     def setUp(self):
         TestBucketBase.setUp(self)
Index: twisted/topfiles/8121.bugfix
===================================================================
--- twisted/topfiles/8121.bugfix	(revision 0)
+++ twisted/topfiles/8121.bugfix	(working copy)
@@ -0,0 +1 @@
+twisted.protocols.htb and twisted.protocols.pcp are ported to Python 3.
