Index: twisted/python/dist3.py
===================================================================
--- twisted/python/dist3.py	(revision 46313)
+++ twisted/python/dist3.py	(working copy)
@@ -140,6 +140,7 @@
     "twisted.protocols",
     "twisted.protocols.basic",
     "twisted.protocols.htb",
+    "twisted.protocols.pcp",
     "twisted.protocols.policies",
     "twisted.protocols.test",
     "twisted.protocols.tls",
@@ -332,6 +333,7 @@
     "twisted.test.test_loopback",
     "twisted.test.test_modules",
     "twisted.test.test_monkey",
+    "twisted.test.test_pcp",
     "twisted.test.test_paths",
     "twisted.test.test_plugin",
     "twisted.test.test_persisted",
