Index: twisted/internet/error.py
===================================================================
--- twisted/internet/error.py	(revision 23937)
+++ twisted/internet/error.py	(working copy)
@@ -268,7 +268,10 @@
 
 
 class ProcessExitedAlready(Exception):
-    """The process has already excited, and the operation requested can no longer be performed."""
+    """
+    The process has already exited and the operation requested can no longer
+    be performed.
+    """
 
 
 class NotConnectingError(RuntimeError):
