Index: twisted/conch/test/test_insults.py
===================================================================
--- twisted/conch/test/test_insults.py	(revision 29097)
+++ twisted/conch/test/test_insults.py	(working copy)
@@ -1,5 +1,5 @@
 # -*- test-case-name: twisted.conch.test.test_insults -*-
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2010 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 from twisted.trial import unittest
@@ -458,3 +458,27 @@
         protocol.makeConnection(transport)
         protocol.nextLine()
         self.assertEqual(transport.value(), "\r\n")
+
+
+
+class Deprecations(unittest.TestCase):
+    """
+    Tests to ensure deprecation of L{colors} and {client}.
+    """
+
+    def test_colors(self):
+        """
+        Test deprecation of L{insults.colors}
+        """
+        from twisted.conch.insults import colors
+        warningsShown = self.flushWarnings()
+        # For some reason, something in import and/or deprecatedModuleAttribute
+        # seems to cause flushWarnings() to produce 2 of the exact same warning.
+        # When flushWarnings() is not used, only one warning is displayed.
+        self.assertTrue(len(warningsShown) >= 1)
+        for warning in warningsShown:
+            self.assertIdentical(warning['category'], DeprecationWarning)
+            self.assertEquals(
+                warning['message'],
+                "twisted.conch.insults.colors was deprecated in Twisted "
+                "10.1.0: Please use twisted.conch.insults.helper instead.")
Index: twisted/conch/topfiles/4096.removal
===================================================================
--- twisted/conch/topfiles/4096.removal	(revision 0)
+++ twisted/conch/topfiles/4096.removal	(revision 0)
@@ -0,0 +1,2 @@
+Deprecated twisted.conch.insults.colors.  twisted.conch.insults.helper should
+be used instead.
Index: twisted/conch/insults/__init__.py
===================================================================
--- twisted/conch/insults/__init__.py	(revision 29097)
+++ twisted/conch/insults/__init__.py	(working copy)
@@ -2,3 +2,11 @@
 Insults: a replacement for Curses/S-Lang.
 
 Very basic at the moment."""
+
+from twisted.python import deprecate, versions
+
+deprecate.deprecatedModuleAttribute(
+    versions.Version("Twisted", 10, 1, 0),
+    "Please use twisted.conch.insults.helper instead.",
+    __name__, "colors")
+
