diff --git twisted/python/logfile.py twisted/python/logfile.py
index f652271..0f50b17 100644
--- twisted/python/logfile.py
+++ twisted/python/logfile.py
@@ -7,6 +7,7 @@
 A rotating, browsable log file.
 """
 
+from __future__ import print_function,division,absolute_import
 # System Imports
 import os, glob, time, stat
 
@@ -66,7 +67,7 @@ class BaseLogFile:
         else:
             if self.defaultMode is not None:
                 # Set the lowest permissions
-                oldUmask = os.umask(0777)
+                oldUmask = os.umask(int("777",8))
                 try:
                     self._file = file(self.path, "w+", 1)
                 finally:
@@ -178,7 +179,7 @@ class LogFile(BaseLogFile):
         """
         filename = "%s.%d" % (self.path, identifier)
         if not os.path.exists(filename):
-            raise ValueError, "no such logfile exists"
+            raise ValueError("no such logfile exists")
         return LogReader(filename)
 
     def write(self, data):
@@ -266,7 +267,7 @@ class DailyLogFile(BaseLogFile):
             return self.getCurrentLog()
         filename = "%s.%s" % (self.path, self.suffix(identifier))
         if not os.path.exists(filename):
-            raise ValueError, "no such logfile exists"
+            raise ValueError("no such logfile exists")
         return LogReader(filename)
 
     def write(self, data):
