commit 9f82e2691111f9eac89266ebb3c2a0fc59690df8
Author: Wolfgang Rohdewald <wolfgang@rohdewald.de>
Date:   Fri Oct 3 23:28:27 2014 +0200

    spread.banana: selecting an unsupported dialect now raises BananaError

diff --git twisted/spread/banana.py twisted/spread/banana.py
index 82b7256..0b215eb 100644
--- twisted/spread/banana.py
+++ twisted/spread/banana.py
@@ -110,6 +110,8 @@ class Banana(protocol.Protocol, styles.Ephemeral):
         """
 
     def _selectDialect(self, dialect):
+        if dialect not in Banana.knownDialects:
+            raise BananaError("%r is an unknown dialect" % (dialect,))
         self.currentDialect = dialect
         self.connectionReady()
 
diff --git twisted/spread/topfiles/7661.bugfix twisted/spread/topfiles/7661.bugfix
new file mode 100644
index 0000000..0958f3a
--- /dev/null
+++ twisted/spread/topfiles/7661.bugfix
@@ -0,0 +1 @@
+twisted.spread.banana now raises BananaError when selecting an unsupported dialect
diff --git twisted/test/test_banana.py twisted/test/test_banana.py
index c4b69de..d914f47 100644
--- twisted/test/test_banana.py
+++ twisted/test/test_banana.py
@@ -45,6 +45,16 @@ class BananaTestCase(unittest.TestCase):
         self.enc.dataReceived(self.io.getvalue())
         assert self.result == 'hello'
 
+
+    def test_validDialect(self):
+        """
+        banana only knows dialects none and pb
+        """
+        self.enc._selectDialect(b'none')
+        self.enc._selectDialect(b'pb')
+        self.assertRaises(banana.BananaError, self.enc._selectDialect, b'abc')
+
+
     def test_int(self):
         """
         A positive integer less than 2 ** 32 should round-trip through
