Index: xish/utility.py
===================================================================
--- xish/utility.py	(revision 16633)
+++ xish/utility.py	(working copy)
@@ -232,6 +232,8 @@
                 if query.matches(object):
                     callbacklist.callback(object)
                     foundTarget = True
+                if not len(callbacklist.callbacks):
+                    del self._xpathObservers[(priority, query)]
 
         self._dispatchDepth = self._dispatchDepth -1
 
