--- /Users/jack/Sources/white.deployment/sources/twisted-2.5.0/TwistedWords-0.5.0/twisted/words/xish/utility.py	2007-01-19 23:10:52.000000000 -0700
+++ utility.py	2007-01-29 14:12:33.000000000 -0700
@@ -37,6 +37,9 @@
             methodwrapper(*args, **kwargs)
             if onetime:
                 del self.callbacks[key]
+    
+    def isEmpty(self):
+        return len(self.callbacks.keys()) == 0
 
 from twisted.words.xish import xpath, domish
 
@@ -232,6 +235,8 @@
                 if query.matches(object):
                     callbacklist.callback(object)
                     foundTarget = True
+                    if callbacklist.isEmpty():
+                        del self._xpathObservers[(priority, query)]
 
         self._dispatchDepth = self._dispatchDepth -1
 
