--- /Users/jack/Sources/white.deployment/sources/twisted-2.5.0/TwistedWords-0.5.0/twisted/words/protocols/jabber/client.py	2007-01-19 23:10:52.000000000 -0700
+++ protocols/jabber/client.py	2007-01-29 14:49:53.000000000 -0700
@@ -66,8 +66,10 @@
         """
         if to != None:
             self["to"] = to
-        self._xmlstream.addOnetimeObserver("/iq[@id='%s']" % self["id"], \
-                                                             self._resultEvent)
+            
+        if self.hasAttribute('type') and self['type'] in ('get', 'set'):
+            self._xmlstream.addOnetimeObserver("/iq[@id='%s']" % self["id"], \
+                                                                self._resultEvent)
         self._xmlstream.send(self)
 
     def _resultEvent(self, iq):
