--- utility.py  2007-01-29 16:03:14.000000000 -0600
+++ utility.py  2007-01-29 17:38:39.000000000 -0600
@@ -231,6 +231,8 @@
                     foundTarget = True
         else:
             for priority, query in self._orderedXpathObserverKeys:
+                if not self._xpathObservers.has_key((priority, query)):
+                    continue
                 callbacklist = self._xpathObservers[(priority, query)]
                 if query.matches(object):
                     callbacklist.callback(object)
