Index: TwistedWords-0.5.0/twisted/words/protocols/jabber/client.py
===================================================================
--- TwistedWords-0.5.0/twisted/words/protocols/jabber/client.py	(revision 4433)
+++ TwistedWords-0.5.0/twisted/words/protocols/jabber/client.py	(revision 4434)
@@ -71,7 +71,8 @@
         self._xmlstream.send(self)
 
     def _resultEvent(self, iq):
-        self.callbacks.callback(iq)
+        if self.callbacks != None:
+            self.callbacks.callback(iq)
         self.callbacks = None
 
 
