Index: ftp.py
===================================================================
--- ftp.py	(revision 16335)
+++ ftp.py	(working copy)
@@ -1450,7 +1450,7 @@
         try:
             p.listdir()
         except (IOError, OSError), e:
-            return errnoToFailure(e.errno, path)
+            raise PermissionDeniedError("Cannot find path. ")
         except:
             return defer.fail()
         else:
