Index: twisted/protocols/htb.py
===================================================================
--- twisted/protocols/htb.py	(revision 42640)
+++ twisted/protocols/htb.py	(working copy)
@@ -192,7 +192,7 @@
     sweepInterval = 60 * 20
 
     def getBucketKey(self, transport):
-        return transport.getPeer()[1]
+        return transport.getPeer()
 
 
 class FilterByServer(HierarchicalBucketFilter):
@@ -202,7 +202,7 @@
     sweepInterval = None
 
     def getBucketKey(self, transport):
-        return transport.getHost()[2]
+        return transport.getHost()
 
 
 class ShapedConsumer(pcp.ProducerConsumerProxy):
