Index: twisted/conch/topfiles/README
===================================================================
--- twisted/conch/topfiles/README	(revision 27607)
+++ twisted/conch/topfiles/README	(working copy)
@@ -1,4 +1,4 @@
 Twisted Conch 9.0.0
 
-Conch depends on Python Crypto (<http://www.amk.ca/python/code/crypto>)
+Conch depends on Python Crypto (<http://www.pycrypto.org>)
 extensions and the pyasn1 module (<http://pyasn1.sourceforge.net/>).
