Index: doc/core/examples/index.xhtml
===================================================================
--- doc/core/examples/index.xhtml	(revision 34294)
+++ doc/core/examples/index.xhtml	(working copy)
@@ -87,7 +87,7 @@
 
     <h2>POSIX Specific Tricks</h2>
     <ul>
-      <li><a href="">sendfd.py</a>, <a href="">recvfd.py</a> - send and receive
+      <li><a href="sendfd.py">sendfd.py</a>, <a href="recvfd.py">recvfd.py</a> - send and receive
       file descriptors over UNIX domain sockets
       </li>
     </ul>
