Index: twisted/protocols/basic.py
===================================================================
--- twisted/protocols/basic.py	(revision 22391)
+++ twisted/protocols/basic.py	(working copy)
@@ -394,7 +394,11 @@
 
     This is a mixin for string protocols (Int32StringReceiver,
     NetstringReceiver) which translates stringReceived into a callback
-    (prefixed with 'proto_') depending on state."""
+    (prefixed with 'proto_') depending on state.
+    
+    The state 'done' is magic; if a proto_* method returns it, the connection
+    will be closed immediately.
+    """
 
     state = 'init'
 
