Index: twisted/protocols/test/test_basic.py
===================================================================
--- twisted/protocols/test/test_basic.py	(revision 38689)
+++ twisted/protocols/test/test_basic.py	(working copy)
@@ -326,6 +326,23 @@
         self.assertTrue(transport.disconnecting)
 
 
+    def test_notQuiteMaximumLineLengthUnfinished(self):
+        """
+        C{LineReceiver} doesn't disconnect the transport it if
+        receives a non-finished line whose length, counting the
+        delimiter, is longer than its C{MAX_LENGTH} but shorter than
+        its C{MAX_LENGTH} + len(delimiter). (When the first part that
+        exceeds the max is the beginning of the delimiter.)
+        """
+        proto = basic.LineReceiver()
+        # '\r\n' is the default, but we set it just to be explicit in this test.
+        proto.delimiter = '\r\n'
+        transport = proto_helpers.StringTransport()
+        proto.makeConnection(transport)
+        proto.dataReceived((b'x' * proto.MAX_LENGTH) + proto.delimiter[:len(proto.delimiter)-1])
+        self.assertFalse(transport.disconnecting)
+
+
     def test_rawDataError(self):
         """
         C{LineReceiver.dataReceived} forwards errors returned by
