Index: Twisted/twisted/cred/portal.py
--- Twisted~issue740/twisted/cred/portal.py	2004-10-06 12:55:50.201742800 +0200
+++ Twisted/twisted/cred/portal.py	2004-10-06 12:59:12.188036216 +0200
@@ -22,6 +22,13 @@
         argument.  The 'avatarAspect' will implement that interface.
         The 'logout' object is a callable which will detach the mind
         from the avatar.
+
+	@type avatarId: string
+	@type mind: L{twisted.spread.pb.RemoteReference}
+	@param mind: remote reference on which the PB server can call remotely
+	@type interfaces: tuple of L{twisted.python.components.Interface}
+
+	@rtype: L{twisted.spread.pb.Avatar}
         """
 
 
