Index: twisted/vfs/test/test_vfs.py
===================================================================
--- twisted/vfs/test/test_vfs.py	(revision 24139)
+++ twisted/vfs/test/test_vfs.py	(working copy)
@@ -1,12 +1,9 @@
 
 import os
-import os.path
-import shutil
 
 from twisted.trial import unittest
 
 from twisted.vfs.backends import osfs, inmem
-from twisted.vfs.ivfs import IFileSystemContainer, IFileSystemLeaf
 
 
 class OSVFSTest(unittest.TestCase):
@@ -18,9 +15,6 @@
         open(os.path.join(self.tmpdir, 'file.txt'), 'w').write('wobble\n')
         self.root = osfs.OSDirectory(self.tmpdir)
 
-    def tearDown(self):
-        shutil.rmtree(self.tmpdir)
-
     def test_listdir(self):
         nodes = self.root.children()
         paths = [path for (path, node) in nodes]
