--- twisted/web/test/test_http.py	Mon Dec 27 03:14:08 2004
+++ twisted/web/test/test_http.py	Mon Mar  6 11:02:44 2006
@@ -341,7 +341,7 @@
 
     def testGET(self):
         httpRequest = '''\
-GET /?key=value&multiple=two+words&multiple=more%20words&empty= HTTP/1.0
+GET /?key=value&multiple=two+words&multiple=more%20words&?=?&empty= HTTP/1.0
 
 '''
         testcase = self
@@ -350,7 +350,8 @@
                 testcase.assertEquals(self.method, "GET")
                 testcase.assertEquals(self.args["key"], ["value"])
                 testcase.assertEquals(self.args["empty"], [""])
+                testcase.assertEquals(self.args["?"], ["?"])
                 testcase.assertEquals(self.args["multiple"], ["two words", "more words"])
                 testcase.didRequest = 1
                 self.finish()
