Index: doc/core/howto/glossary.xhtml
===================================================================
--- doc/core/howto/glossary.xhtml	(revision 26477)
+++ doc/core/howto/glossary.xhtml	(working copy)
@@ -199,7 +199,7 @@
 
 <dt><a name="Lore">Lore</a></dt>
 
-<dd><a href="http://twistedmatrix.com/projects/lore/">Lore</a> is
+<dd><a href="http://twistedmatrix.com/trac/wiki/TwistedLore/">Lore</a> is
 Twisted's documentation system.  The source format is a subset of
 XHTML, and output formats include HTML and LaTeX.</dd>
 
@@ -332,12 +332,11 @@
 
 <dt><a name="TwistedReality">Twisted Reality</a></dt>
 <dd>
-  In days of old, the <a
-  href="http://twistedmatrix.com/products/reality">Twisted Reality</a>
-  multiplayer text-based interactive-fiction system was the main focus of Twisted
-  Matrix Labs; Twisted, the general networking framework, grew out of Reality's
-  need for better network functionality. Twisted Reality has since been broken
-  off into a separate project.
+In days of old, the Twisted Reality multiplayer text-based interactive-fiction
+system was the main focus of Twisted Matrix Labs; Twisted, the general networking
+framework, grew out of Reality's need for better network functionality. Twisted
+Reality has been superseded by the
+<a href="http://www.divmod.org/trac/wiki/DivmodImaginary">Imaginary</a> project.
 </dd>
 
 <dt><a name="usage"><code noexpand="1" class="API" base="twisted.python">usage</code></a></dt>
