Index: twisted/internet/gtk2reactor.py
===================================================================
--- twisted/internet/gtk2reactor.py	(revisión: 28756)
+++ twisted/internet/gtk2reactor.py	(copia de trabajo)
@@ -306,6 +306,13 @@
         self.simulate() # fire Twisted timers
         return 1 # 1=don't auto-remove the source
 
+    def callLater (self, _seconds, _f, *args, **kw):
+        ret = posixbase.PosixReactorBase.callLater(self, _seconds, _f, *args, **kw)
+        if self._simtag is None:
+            timeout = min(self.timeout(), 0.1)
+            self._simtag = gobject.timeout_add(int(timeout * 1010), self.simulate)
+        return ret
+        
 
     def simulate(self):
         """
@@ -313,10 +320,11 @@
         """
         if self._simtag is not None:
             gobject.source_remove(self._simtag)
+            self._simtag = None
         self.runUntilCurrent()
         timeout = min(self.timeout(), 0.1)
         if timeout is None:
-            timeout = 0.1
+            return
         # grumble
         self._simtag = gobject.timeout_add(int(timeout * 1010), self.simulate)
 
