Index: twisted/conch/ssh/common.py
===================================================================
--- twisted/conch/ssh/common.py	(revision 34751)
+++ twisted/conch/ssh/common.py	(working copy)
@@ -9,7 +9,7 @@
 Maintainer: Paul Swartz
 """
 
-import struct, warnings
+import struct, warnings, __builtin__
 
 try:
     from Crypto import Util
@@ -107,7 +107,7 @@
         if type(x) in (long, int):
             x = mpz(x)
         return pyPow(x, y, z)
-    __builtins__['pow'] = _fastpow # evil evil
+    __builtin__.pow = _fastpow # evil evil
 
 try:
     import gmpy
