=== modified file 'twisted/internet/base.py'
--- twisted/internet/base.py	2009-06-01 18:07:14 +0000
+++ twisted/internet/base.py	2009-11-17 18:44:20 +0000
@@ -16,7 +16,7 @@
 
 import traceback
 
-from twisted.python.compat import set
+from twisted.python.compat import callable, set
 from twisted.python.util import unsignedID
 from twisted.internet.interfaces import IReactorCore, IReactorTime, IReactorThreads
 from twisted.internet.interfaces import IResolverSimple, IReactorPluggableResolver

=== modified file 'twisted/internet/defer.py'
--- twisted/internet/defer.py	2009-06-29 21:12:48 +0000
+++ twisted/internet/defer.py	2009-11-17 18:44:43 +0000
@@ -13,6 +13,7 @@
 
 # Twisted imports
 from twisted.python import log, failure, lockfile
+from twisted.python.compat import callable
 from twisted.python.util import unsignedID, mergeFunctionMetadata
 
 class AlreadyCalledError(Exception):

=== modified file 'twisted/internet/task.py'
--- twisted/internet/task.py	2009-11-08 15:54:28 +0000
+++ twisted/internet/task.py	2009-11-17 18:45:06 +0000
@@ -15,6 +15,7 @@
 from zope.interface import implements
 
 from twisted.python import reflect
+from twisted.python.compat import callable
 from twisted.python.failure import Failure
 
 from twisted.internet import base, defer

=== modified file 'twisted/mail/imap4.py'
--- twisted/mail/imap4.py	2009-11-15 14:19:58 +0000
+++ twisted/mail/imap4.py	2009-11-17 18:45:56 +0000
@@ -43,6 +43,7 @@
 from twisted.internet import error
 from twisted.internet.defer import maybeDeferred
 from twisted.python import log, text
+from twisted.python.compat import callable
 from twisted.internet import interfaces
 
 from twisted import cred

=== modified file 'twisted/mail/test/test_mail.py'
--- twisted/mail/test/test_mail.py	2009-05-03 22:17:54 +0000
+++ twisted/mail/test/test_mail.py	2009-11-17 18:45:32 +0000
@@ -30,6 +30,7 @@
 from twisted.internet.error import ProcessDone, ProcessTerminated
 from twisted.internet import address
 from twisted.python import failure
+from twisted.python.compat import callable
 from twisted.python.filepath import FilePath
 from twisted.python.hashlib import md5
 

=== modified file 'twisted/manhole/service.py'
--- twisted/manhole/service.py	2006-07-01 16:08:17 +0000
+++ twisted/manhole/service.py	2009-11-17 18:46:20 +0000
@@ -10,6 +10,7 @@
 from twisted import copyright
 from twisted.spread import pb
 from twisted.python import log, failure
+from twisted.python.compat import callable
 from twisted.cred import portal
 from twisted.application import service
 from zope.interface import implements, Interface

=== modified file 'twisted/python/compat.py'
--- twisted/python/compat.py	2008-01-06 08:30:50 +0000
+++ twisted/python/compat.py	2009-11-17 18:54:12 +0000
@@ -166,3 +166,10 @@
 except NameError:
     from sets import ImmutableSet as frozenset
 
+
+try:
+    callable = callable
+except NameError:
+    import collections
+    def callable(obj):
+        return isinstance(obj, collections.Callable)

=== modified file 'twisted/python/log.py'
--- twisted/python/log.py	2009-08-10 12:44:10 +0000
+++ twisted/python/log.py	2009-11-17 18:46:52 +0000
@@ -17,6 +17,7 @@
 from zope.interface import Interface
 
 from twisted.python import util, context, reflect
+from twisted.python.compat import callable
 
 
 

=== modified file 'twisted/python/zshcomp.py'
--- twisted/python/zshcomp.py	2008-03-04 20:36:22 +0000
+++ twisted/python/zshcomp.py	2009-11-17 18:47:29 +0000
@@ -137,6 +137,7 @@
 import itertools, sys, commands, os.path
 
 from twisted.python import reflect, util, usage
+from twisted.python.compat import callable
 from twisted.scripts.mktap import IServiceMaker
 
 class MyOptions(usage.Options):

=== modified file 'twisted/spread/pb.py'
--- twisted/spread/pb.py	2008-11-19 18:56:50 +0000
+++ twisted/spread/pb.py	2009-11-17 18:47:55 +0000
@@ -36,6 +36,7 @@
 # Twisted Imports
 from twisted.python import log, failure, reflect
 from twisted.python.versions import Version
+from twisted.python.compat import callable
 from twisted.python.deprecate import deprecated
 from twisted.python.hashlib import md5
 from twisted.internet import defer, protocol

=== modified file 'twisted/spread/util.py'
--- twisted/spread/util.py	2008-02-16 08:00:12 +0000
+++ twisted/spread/util.py	2009-11-17 18:48:21 +0000
@@ -10,6 +10,7 @@
 
 from twisted.internet import defer
 from twisted.python.failure import Failure
+from twisted.python.compat import callable
 from twisted.spread import pb
 from twisted.protocols import basic
 from twisted.internet import interfaces

=== modified file 'twisted/trial/test/test_assertions.py'
--- twisted/trial/test/test_assertions.py	2008-11-02 13:59:21 +0000
+++ twisted/trial/test/test_assertions.py	2009-11-17 18:48:55 +0000
@@ -9,6 +9,7 @@
 from pprint import pformat
 
 from twisted.python import reflect, failure
+from twisted.python.compat import callable
 from twisted.python.deprecate import deprecated, getVersionString
 from twisted.python.versions import Version
 from twisted.python.util import dsu

=== modified file 'twisted/trial/test/test_pyunitcompat.py'
--- twisted/trial/test/test_pyunitcompat.py	2008-07-29 20:13:54 +0000
+++ twisted/trial/test/test_pyunitcompat.py	2009-11-17 18:49:20 +0000
@@ -10,6 +10,7 @@
 from zope.interface import implements
 
 from twisted.python import reflect
+from twisted.python.compat import callable
 from twisted.python.failure import Failure
 from twisted.trial import util
 from twisted.trial.unittest import TestCase, PyUnitResultAdapter

=== modified file 'twisted/vfs/backends/osfs.py'
--- twisted/vfs/backends/osfs.py	2007-05-22 09:24:14 +0000
+++ twisted/vfs/backends/osfs.py	2009-11-17 18:49:41 +0000
@@ -7,6 +7,7 @@
 import os.path
 import errno
 
+from twisted.python.compat import callable
 from twisted.vfs import ivfs
 
 from zope.interface import implements

=== modified file 'twisted/web/soap.py'
--- twisted/web/soap.py	2008-07-29 20:13:54 +0000
+++ twisted/web/soap.py	2009-11-17 18:50:04 +0000
@@ -19,6 +19,7 @@
 import SOAPpy
 
 # twisted imports
+from twisted.python.compat import callable
 from twisted.web import server, resource, client
 from twisted.internet import defer
 

=== modified file 'twisted/web/xmlrpc.py'
--- twisted/web/xmlrpc.py	2009-01-21 17:10:00 +0000
+++ twisted/web/xmlrpc.py	2009-11-17 18:50:25 +0000
@@ -15,6 +15,7 @@
 from twisted.web import resource, server, http
 from twisted.internet import defer, protocol, reactor
 from twisted.python import log, reflect, failure
+from twisted.python.compat import callable
 
 # These are deprecated, use the class level definitions
 NOT_FOUND = 8001

=== modified file 'twisted/web2/xmlrpc.py'
--- twisted/web2/xmlrpc.py	2008-07-29 20:13:54 +0000
+++ twisted/web2/xmlrpc.py	2009-11-17 18:50:41 +0000
@@ -18,6 +18,7 @@
 from twisted.web2 import responsecode, http, http_headers
 from twisted.internet import defer
 from twisted.python import log, reflect
+from twisted.python.compat import callable
 
 # Useful so people don't need to import xmlrpclib directly
 Fault = xmlrpclib.Fault

=== modified file 'twisted/words/protocols/irc.py'
--- twisted/words/protocols/irc.py	2009-11-14 00:02:57 +0000
+++ twisted/words/protocols/irc.py	2009-11-17 18:51:00 +0000
@@ -37,6 +37,7 @@
 from twisted.persisted import styles
 from twisted.protocols import basic
 from twisted.python import log, reflect, text
+from twisted.python.compat import callable
 
 NUL = chr(0)
 CR = chr(015)

