diff --git twisted/mail/test/test_options.py twisted/mail/test/test_options.py
index d32e107..66cb694 100644
--- twisted/mail/test/test_options.py
+++ twisted/mail/test/test_options.py
@@ -11,13 +11,11 @@ from twisted.python.usage import UsageError
 from twisted.mail import protocols
 from twisted.mail.tap import Options, makeService
 from twisted.python.filepath import FilePath
-from twisted.python.reflect import namedAny
+from twisted.python.reflect import requireModule
 from twisted.internet import endpoints, defer
 
-try:
-    namedAny('OpenSSL')
-except ImportError, e:
-    sslSkip = str(e)
+if requireModule('OpenSSL', default=None) is None:
+    sslSkip = 'Missing OpenSSL package.'
 else:
     sslSkip = None
 
