Index: TwistedCore-2.4.0/twisted/test/stdio_test_halfclose.py
===================================================================
RCS file: /cvs/netsec-dev/twisted/TwistedCore-2.4.0/twisted/test/stdio_test_halfclose.py,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 stdio_test_halfclose.py
--- TwistedCore-2.4.0/twisted/test/stdio_test_halfclose.py      25 Sep 2006 21:56:17 -0000      1.1.1.1
+++ TwistedCore-2.4.0/twisted/test/stdio_test_halfclose.py      25 Sep 2006 22:14:57 -0000
@@ -17,7 +17,8 @@
 class ProducerChild(protocol.Protocol):
     _paused = False

-    ???
+    # ???
+    raise NotImplementedError()

     def connectionLost(self, reason):
         reactor.stop()
