Index: twisted/python/runtime.py
===================================================================
--- twisted/python/runtime.py	(revision 23532)
+++ twisted/python/runtime.py	(working copy)
@@ -69,6 +69,22 @@
     def isWindows(self):
         return self.getType() == 'win32'
 
+
+    def isVista(self):
+        """
+        Check if current platform is Windows Vista or Windows Server 2008.
+
+        @return: C{True} if the current platform has been detected as Vista
+        @rtype: C{bool}
+        """
+        try:
+            import win32api
+        except ImportError:
+             return False
+        else:
+            return win32api.GetVersionEx()[0] == 6
+
+
     def supportsThreads(self):
         """Can threads be created?
         """
Index: twisted/test/test_udp.py
===================================================================
--- twisted/test/test_udp.py	(revision 23532)
+++ twisted/test/test_udp.py	(working copy)
@@ -705,7 +705,7 @@
         return self.assertFailure(
             self.client.transport.joinGroup("127.0.0.1"),
             error.MulticastJoinError)
-    if runtime.platform.isWindows():
+    if runtime.platform.isWindows() and not runtime.platform.isVista():
         test_joinFailure.todo = "Windows' multicast is wonky"
 
 
