
diff -r c8fa1e90dc51 foolscap/test/test_interfaces.py
--- a/foolscap/test/test_interfaces.py	Mon Jul 02 02:19:58 2007 -0700
+++ b/foolscap/test/test_interfaces.py	Tue Jul 10 17:38:13 2007 +0200
@@ -121,7 +121,7 @@ class TestInterface(TargetMixin, unittes
             for i in range(len(s)):
                 line = s[i]
                 #print line
-                if ("test/test_interfaces.py" in line
+                if ("test_interfaces.py" in line
                     and i+1 < len(s)
                     and "rr.callRemote" in s[i+1]):
                     return # all good
