Index: twisted/web/test/test_http.py
===================================================================
--- twisted/web/test/test_http.py	(revision 36864)
+++ twisted/web/test/test_http.py	(working copy)
@@ -1653,6 +1653,15 @@
         self.assertRaises(RuntimeError, req.finish)
 
 
+    def test_noLongerQueuedRaisesError(self):
+        """
+        Calling L{Request.noLongerQueued} for an object that is not queued
+        results in a L{RuntimeError} being raised.
+        """
+        request = http.Request(DummyChannel(), False)
+        self.assertRaises(RuntimeError, request.noLongerQueued)
+
+
 
 class MultilineHeadersTestCase(unittest.TestCase):
     """
