commit 7cb29ae2206af4d41c7e854388d45ff83b16eacd
Author: Anthony Martinez <anthony@flowroute.com>
Date:   Wed Sep 23 10:53:22 2015 -0700

    Correct documentation for DeferredSemaphore

diff --git a/twisted/internet/defer.py b/twisted/internet/defer.py
index 1623e16..fe9f0a0 100644
--- a/twisted/internet/defer.py
+++ b/twisted/internet/defer.py
@@ -1390,13 +1390,13 @@ class DeferredSemaphore(_ConcurrencyPrimitive):
     If you are looking into this as a means of limiting parallelism, you might
     find L{twisted.internet.task.Cooperator} more useful.
 
-    @ivar tokens: At most this many users may acquire this semaphore at
+    @ivar limit: At most this many users may acquire this semaphore at
         once.
-    @type tokens: C{int}
+    @type limit: C{int}
 
-    @ivar limit: The difference between C{tokens} and the number of users
+    @ivar tokens: The difference between C{limit} and the number of users
         which have currently acquired this semaphore.
-    @type limit: C{int}
+    @type tokens: C{int}
     """
 
     def __init__(self, tokens):
diff --git a/twisted/topfiles/8024.doc b/twisted/topfiles/8024.doc
new file mode 100644
index 0000000..2d372fd
--- /dev/null
+++ b/twisted/topfiles/8024.doc
@@ -0,0 +1 @@
+Now the documentation for `twisted.defer.DeferredSemaphore` refers to the correct instance attributes.
