Index: twisted/words/protocols/irc.py
===================================================================
--- twisted/words/protocols/irc.py	(revision 29927)
+++ twisted/words/protocols/irc.py	(working copy)
@@ -2903,6 +2903,7 @@
 RPL_LISTEND = '323'
 RPL_UNIQOPIS = '325'
 RPL_CHANNELMODEIS = '324'
+RPL_WHOISACCOUNT = '330'
 RPL_NOTOPIC = '331'
 RPL_TOPIC = '332'
 RPL_INVITING = '341'
@@ -3045,6 +3046,7 @@
     "RPL_LISTEND": '323',
     "RPL_UNIQOPIS": '325',
     "RPL_CHANNELMODEIS": '324',
+    "RPL_WHOISACCOUNT": '330',
     "RPL_NOTOPIC": '331',
     "RPL_TOPIC": '332',
     "RPL_INVITING": '341',
