Index: python/filepath.py
===================================================================
--- python/filepath.py	(revision 22975)
+++ python/filepath.py	(working copy)
@@ -499,7 +499,7 @@
         utime(self.path, None)
 
     def remove(self):
-        if self.isdir():
+        if self.isdir() and not self.islink():
             for child in self.children():
                 child.remove()
             os.rmdir(self.path)
Index: test/test_paths.py
===================================================================
--- test/test_paths.py	(revision 22975)
+++ test/test_paths.py	(working copy)
@@ -383,6 +383,13 @@
         self.assertIn(ts, self.path.parent().children())
 
     def testRemove(self):
+        if hasattr(os, "symlink"): 
+            #if symlinks not supported, don't test this part
+            link = self.path.child("sub1.link")
+            os.symlink(self.path.child("sub1").path, link.path)
+            link.remove()
+            self.failIf(link.exists())
+            self.failUnless(self.path.child("sub1").exists())
         self.path.remove()
         self.failIf(self.path.exists())
 
