=== modified file 'twisted/conch/test/test_ssh.py'
--- twisted/conch/test/test_ssh.py	2009-10-09 10:24:59 +0000
+++ twisted/conch/test/test_ssh.py	2009-10-09 10:51:34 +0000
@@ -35,11 +35,13 @@
     """
     implements(portal.IRealm)
 
-    def __init__(self, avatar):
+    def __init__(self, avatar=None):
         """
         Initialize class with a avatar.
         @param avatar: an instance of C{avatar.ConchUser}.
         """
+        if avatar is None:
+            avatar = ConchTestAvatar()
         self.avatar = avatar
 
 
@@ -1008,8 +1010,6 @@
 
         @paran: avatar: an instance of C{avatar.ConchUser}.
         """
-        if avatar is None:
-            avatar = ConchTestAvatar()
         if clientConnection is None:
             clientConnection = ConchTestClientConnection()
         realm = ConchTestRealm(avatar)

