Index: twisted/conch/test/test_filetransfer.py
===================================================================
--- twisted/conch/test/test_filetransfer.py	(revision 22951)
+++ twisted/conch/test/test_filetransfer.py	(working copy)
@@ -408,8 +408,11 @@
 
 
 class FakeConn:
+    def logPrefix(self):
+        return 'Fake Connection'
+    
     def sendClose(self, channel):
-        pass
+        channel.closed()
 
 
 class TestFileTransferClose(unittest.TestCase):
