Index: doc/conch/examples/sshsimpleserver.py
===================================================================
--- doc/conch/examples/sshsimpleserver.py	(revision 22390)
+++ doc/conch/examples/sshsimpleserver.py	(working copy)
@@ -13,13 +13,20 @@
 log in with username "user" and password "password".
 """
 
+class Subsys(protocol.Protocol):
+    def __init__(self, *args, **kw):
+        self.avatar = kw['avatar']
+
+    def connectionLost(self, reason):
+        log.msg('Subsys.connectionLost')
+
 class ExampleAvatar(avatar.ConchUser):
 
     def __init__(self, username):
         avatar.ConchUser.__init__(self)
         self.username = username
         self.channelLookup.update({'session':session.SSHSession})
-
+        self.subsystemLookup.update({'subsys':Subsys})
 class ExampleRealm:
     implements(portal.IRealm)
 
