Index: twisted/conch/ssh/session.py
===================================================================
--- twisted/conch/ssh/session.py	(revision 22794)
+++ twisted/conch/ssh/session.py	(working copy)
@@ -127,8 +127,6 @@
     #    self.loseConnection() # don't know what to do with this
 
     def loseConnection(self):
-        if self.client:
-            self.client.transport.loseConnection()
         channel.SSHChannel.loseConnection(self)
 
 class _ProtocolWrapper(protocol.ProcessProtocol):
