Index: twisted/protocols/basic.py
===================================================================
--- twisted/protocols/basic.py	(Revision 29821)
+++ twisted/protocols/basic.py	(Arbeitskopie)
@@ -20,14 +20,30 @@
 
 # Twisted imports
 from twisted.internet import protocol, defer, interfaces, error
-from twisted.python import log
+from twisted.python import log, deprecate, versions
 
 
 LENGTH, DATA, COMMA = range(3)
 NUMBER = re.compile('(\d*)(:?)')
+
+_deprecatedSince = versions.Version("Twisted", 10, 1, 0)
+_useInstead = "Use the NetstringReceiver attribute %s instead."
+_attributesToDeprecate = [
+    ("LENGTH", _useInstead % ("_PARSING_LENGTH",)),
+    ("DATA",
+     (_useInstead % ("_PARSING_PAYLOAD",) + (" Beware that the parser "
+      "remains in this state until the trailing comma was processed."))),
+    ("COMMA",
+     "This module attribute is not used anymore: The parser now has only two "
+     "states: _PARSING_LENGTH and _PARSING_PAYLOAD."),
+    ("NUMBER", _useInstead % ("_LENGTH",))
+    ]
+for attrib, reason in _attributesToDeprecate:
+    deprecate.deprecatedModuleAttribute(
+        _deprecatedSince, reason, __name__, attrib)
+
 DEBUG = 0
 
-
 class NetstringParseError(ValueError):
     """
     The incoming data is not in valid Netstring format.
