Index: twisted/test/test_protocols.py
===================================================================
--- twisted/test/test_protocols.py	(Revision 29882)
+++ twisted/test/test_protocols.py	(Arbeitskopie)
@@ -650,6 +650,40 @@
                           self.netstringReceiver._consumeLength)
 
 
+    def test_deprecatedModuleAttributes(self):
+        """
+        Accessing a deprecated module attribute raises the appropriate warning.
+        """
+        length, data, comma, number = (basic.LENGTH, basic.DATA, basic.COMMA,
+                                       basic.NUMBER)
+        warnings = self.flushWarnings(
+            offendingFunctions = [self.test_deprecatedModuleAttributes])
+
+        self.assertEquals(len(warnings), 4)
+        for warning in warnings:
+            self.assertEquals(warning['category'], DeprecationWarning)
+        self.assertEquals(
+            warnings[0]['message'],
+            ("twisted.protocols.basic.LENGTH was deprecated in Twisted 10.2.0: "
+             "Use the NetstringReceiver attribute _PARSING_LENGTH instead."))
+        self.assertEquals(
+            warnings[1]['message'],
+            ("twisted.protocols.basic.DATA was deprecated in Twisted 10.2.0: "
+             "Use the NetstringReceiver attribute _PARSING_PAYLOAD instead.\n"
+             "Beware that the parser remains in this state until the trailing "
+             "comma was processed."))
+        self.assertEquals(
+            warnings[2]['message'],
+            ("twisted.protocols.basic.COMMA was deprecated in Twisted 10.2.0: "
+             "This module attribute is not used anymore: The parser now has "
+             "only two states: _PARSING_LENGTH and _PARSING_PAYLOAD."))
+        self.assertEquals(
+            warnings[3]['message'],
+            ("twisted.protocols.basic.NUMBER was deprecated in Twisted 10.2.0: "
+             "Use the NetstringReceiver attribute _LENGTH instead."))
+
+
+
 class IntNTestCaseMixin(LPTestCaseMixin):
     """
     TestCase mixin for int-prefixed protocols.
Index: twisted/protocols/basic.py
===================================================================
--- twisted/protocols/basic.py	(Revision 29882)
+++ twisted/protocols/basic.py	(Arbeitskopie)
@@ -20,14 +20,34 @@
 
 # Twisted imports
 from twisted.internet import protocol, defer, interfaces, error
-from twisted.python import log
+from twisted.python import log, deprecate, versions
 
 
 LENGTH, DATA, COMMA = range(3)
 NUMBER = re.compile('(\d*)(:?)')
+
+_deprecatedSince = versions.Version("Twisted", 10, 2, 0)
+deprecate.deprecatedModuleAttribute(
+    _deprecatedSince,
+    "Use the NetstringReceiver attribute _PARSING_LENGTH instead.", __name__,
+    "LENGTH")
+deprecate.deprecatedModuleAttribute(
+    _deprecatedSince,
+    ("Use the NetstringReceiver attribute _PARSING_PAYLOAD instead.\n"
+     "Beware that the parser remains in this state until the trailing comma "
+     "was processed."), __name__, "DATA")
+deprecate.deprecatedModuleAttribute(
+    _deprecatedSince,
+    ("This module attribute is not used anymore: The parser now has only two "
+     "states: _PARSING_LENGTH and _PARSING_PAYLOAD."),
+    __name__, "COMMA")
+deprecate.deprecatedModuleAttribute(
+    _deprecatedSince, "Use the NetstringReceiver attribute _LENGTH instead.",
+    __name__, "NUMBER")
+del _deprecatedSince
+
 DEBUG = 0
 
-
 class NetstringParseError(ValueError):
     """
     The incoming data is not in valid Netstring format.
Index: twisted/topfiles/4541.bugfix
===================================================================
--- twisted/topfiles/4541.bugfix	(Revision 0)
+++ twisted/topfiles/4541.bugfix	(Revision 0)
@@ -0,0 +1,3 @@
+The module attributes `LENGTH`, `DATA`, `COMMA`, and `NUMBER` of
+twisted.protocols.basic (previously used by `NetstringReceiver`)
+are now deprecated.
