Index: twisted/test/test_defer.py
===================================================================
--- twisted/test/test_defer.py  (revision 24017)
+++ twisted/test/test_defer.py  (working copy)
@@ -625,6 +625,7 @@
         that exception will be logged immediately.
         """
         defer.Deferred().addCallback(lambda x: 1/0).callback(1)
+        gc.collect()
         self._check()
 
     def test_errorLogWithInnerFrameRef(self):
