Index: twisted/conch/topfiles/setup.py
===================================================================
--- twisted/conch/topfiles/setup.py	(revision 13339)
+++ twisted/conch/topfiles/setup.py	(working copy)
@@ -1,6 +1,11 @@
 import sys
 
-from twisted.python import dist
+try:
+    from twisted.python import dist
+except ImportError:
+    raise SystemExit("twisted.python.dist module not found.  Make sure you "
+                     "have installed the Twisted core package before "
+                     "attempting to install any other Twisted projects.")
 
 if __name__ == '__main__':
     dist.setup(
Index: twisted/flow/topfiles/setup.py
===================================================================
--- twisted/flow/topfiles/setup.py	(revision 13339)
+++ twisted/flow/topfiles/setup.py	(working copy)
@@ -1,6 +1,11 @@
 import sys
 
-from twisted.python import dist
+try:
+    from twisted.python import dist
+except ImportError:
+    raise SystemExit("twisted.python.dist module not found.  Make sure you "
+                     "have installed the Twisted core package before "
+                     "attempting to install any other Twisted projects.")
 
 if __name__ == '__main__':
     dist.setup(
Index: twisted/lore/topfiles/setup.py
===================================================================
--- twisted/lore/topfiles/setup.py	(revision 13339)
+++ twisted/lore/topfiles/setup.py	(working copy)
@@ -1,6 +1,11 @@
 import sys
 
-from twisted.python import dist
+try:
+    from twisted.python import dist
+except ImportError:
+    raise SystemExit("twisted.python.dist module not found.  Make sure you "
+                     "have installed the Twisted core package before "
+                     "attempting to install any other Twisted projects.")
 
 if __name__ == '__main__':
     dist.setup(
Index: twisted/mail/topfiles/setup.py
===================================================================
--- twisted/mail/topfiles/setup.py	(revision 13339)
+++ twisted/mail/topfiles/setup.py	(working copy)
@@ -1,6 +1,11 @@
 import sys
 
-from twisted.python import dist
+try:
+    from twisted.python import dist
+except ImportError:
+    raise SystemExit("twisted.python.dist module not found.  Make sure you "
+                     "have installed the Twisted core package before "
+                     "attempting to install any other Twisted projects.")
 
 if __name__ == '__main__':
     dist.setup(
Index: twisted/names/topfiles/setup.py
===================================================================
--- twisted/names/topfiles/setup.py	(revision 13339)
+++ twisted/names/topfiles/setup.py	(working copy)
@@ -1,6 +1,11 @@
 import sys
 
-from twisted.python import dist
+try:
+    from twisted.python import dist
+except ImportError:
+    raise SystemExit("twisted.python.dist module not found.  Make sure you "
+                     "have installed the Twisted core package before "
+                     "attempting to install any other Twisted projects.")
 
 if __name__ == '__main__':
     dist.setup(
Index: twisted/news/topfiles/setup.py
===================================================================
--- twisted/news/topfiles/setup.py	(revision 13339)
+++ twisted/news/topfiles/setup.py	(working copy)
@@ -1,6 +1,11 @@
 import sys
 
-from twisted.python import dist
+try:
+    from twisted.python import dist
+except ImportError:
+    raise SystemExit("twisted.python.dist module not found.  Make sure you "
+                     "have installed the Twisted core package before "
+                     "attempting to install any other Twisted projects.")
 
 if __name__ == '__main__':
     dist.setup(
Index: twisted/pair/topfiles/setup.py
===================================================================
--- twisted/pair/topfiles/setup.py	(revision 13339)
+++ twisted/pair/topfiles/setup.py	(working copy)
@@ -1,6 +1,11 @@
 import sys
 
-from twisted.python import dist
+try:
+    from twisted.python import dist
+except ImportError:
+    raise SystemExit("twisted.python.dist module not found.  Make sure you "
+                     "have installed the Twisted core package before "
+                     "attempting to install any other Twisted projects.")
 
 if __name__ == '__main__':
     dist.setup(
Index: twisted/runner/topfiles/setup.py
===================================================================
--- twisted/runner/topfiles/setup.py	(revision 13339)
+++ twisted/runner/topfiles/setup.py	(working copy)
@@ -2,7 +2,12 @@
 
 from distutils.core import Extension
 
-from twisted.python import dist
+try:
+    from twisted.python import dist
+except ImportError:
+    raise SystemExit("twisted.python.dist module not found.  Make sure you "
+                     "have installed the Twisted core package before "
+                     "attempting to install any other Twisted projects.")
 
 def detectExtensions(builder):
     if builder._check_header("rpc/rpc.h"):
Index: twisted/web2/topfiles/setup.py
===================================================================
--- twisted/web2/topfiles/setup.py	(revision 13339)
+++ twisted/web2/topfiles/setup.py	(working copy)
@@ -1,6 +1,11 @@
 import sys
 
-from twisted.python import dist
+try:
+    from twisted.python import dist
+except ImportError:
+    raise SystemExit("twisted.python.dist module not found.  Make sure you "
+                     "have installed the Twisted core package before "
+                     "attempting to install any other Twisted projects.")
 
 if __name__ == '__main__':
     dist.setup(
Index: twisted/web/topfiles/setup.py
===================================================================
--- twisted/web/topfiles/setup.py	(revision 13339)
+++ twisted/web/topfiles/setup.py	(working copy)
@@ -1,6 +1,11 @@
 import sys
 
-from twisted.python import dist
+try:
+    from twisted.python import dist
+except ImportError:
+    raise SystemExit("twisted.python.dist module not found.  Make sure you "
+                     "have installed the Twisted core package before "
+                     "attempting to install any other Twisted projects.")
 
 if __name__ == '__main__':
     dist.setup(
Index: twisted/words/topfiles/setup.py
===================================================================
--- twisted/words/topfiles/setup.py	(revision 13339)
+++ twisted/words/topfiles/setup.py	(working copy)
@@ -1,6 +1,11 @@
 import sys
 
-from twisted.python import dist
+try:
+    from twisted.python import dist
+except ImportError:
+    raise SystemExit("twisted.python.dist module not found.  Make sure you "
+                     "have installed the Twisted core package before "
+                     "attempting to install any other Twisted projects.")
 
 if __name__ == '__main__':
     dist.setup(
Index: twisted/xish/topfiles/setup.py
===================================================================
--- twisted/xish/topfiles/setup.py	(revision 13339)
+++ twisted/xish/topfiles/setup.py	(working copy)
@@ -1,6 +1,11 @@
 import sys
 
-from twisted.python import dist
+try:
+    from twisted.python import dist
+except ImportError:
+    raise SystemExit("twisted.python.dist module not found.  Make sure you "
+                     "have installed the Twisted core package before "
+                     "attempting to install any other Twisted projects.")
 
 if __name__ == '__main__':
     dist.setup(
