Index: enterprise/row.py
===================================================================
--- enterprise/row.py	(revision 24235)
+++ enterprise/row.py	(working copy)
@@ -10,7 +10,7 @@
 
 This is an extremely thin wrapper.
 
-Maintainer: U{Dave Peticolas<mailto:davep@twistedmatrix.com>}
+Maintainer: Dave Peticolas
 """
 
 import warnings
Index: python/dist.py
===================================================================
--- python/dist.py	(revision 24235)
+++ python/dist.py	(working copy)
@@ -3,7 +3,7 @@
 
 Don't use this outside of Twisted.
 
-Maintainer: U{Christopher Armstrong<mailto:radix@twistedmatrix.com>}
+Maintainer: Christopher Armstrong
 """
 
 import sys, os
Index: python/release.py
===================================================================
--- python/release.py	(revision 24235)
+++ python/release.py	(working copy)
@@ -3,7 +3,7 @@
 
 Don't use this outside of Twisted.
 
-Maintainer: U{Christopher Armstrong<mailto:radix@twistedmatrix.com>}
+Maintainer: Christopher Armstrong
 """
 
 import os
Index: protocols/gps/nmea.py
===================================================================
--- protocols/gps/nmea.py	(revision 24235)
+++ protocols/gps/nmea.py	(working copy)
@@ -4,7 +4,7 @@
 
 """NMEA 0183 implementation
 
-Maintainer: U{Bob Ippolito<mailto:bob@redivi.com>}
+Maintainer: Bob Ippolito
 
 The following NMEA 0183 sentences are currently understood::
     GPGGA (fix)
Index: protocols/gps/rockwell.py
===================================================================
--- protocols/gps/rockwell.py	(revision 24235)
+++ protocols/gps/rockwell.py	(working copy)
@@ -5,7 +5,7 @@
 """Rockwell Semiconductor Zodiac Serial Protocol
 Coded from official protocol specs (Order No. GPS-25, 09/24/1996, Revision 11)
 
-Maintainer: U{Bob Ippolito<mailto:bob@redivi.com>}
+Maintainer: Bob Ippolito
 
 The following Rockwell Zodiac messages are currently understood::
     EARTHA\\r\\n (a hack to "turn on" a DeLorme Earthmate)
Index: protocols/htb.py
===================================================================
--- protocols/htb.py	(revision 24235)
+++ protocols/htb.py	(working copy)
@@ -13,7 +13,7 @@
   <http://luxik.cdi.cz/~devik/qos/htb/manual/userg.htm>}
 @seealso: U{Token Bucket Filter in Linux Advanced Routing & Traffic Control
     HOWTO<http://lartc.org/howto/lartc.qdisc.classless.html#AEN682>}
-@author: U{Kevin Turner<mailto:acapnotic@twistedmatrix.com>}
+@author: Kevin Turner
 """
 
 from __future__ import nested_scopes
Index: protocols/basic.py
===================================================================
--- protocols/basic.py	(revision 24235)
+++ protocols/basic.py	(working copy)
@@ -6,7 +6,7 @@
 """
 Basic protocols, such as line-oriented, netstring, and int prefixed strings.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 # System imports
Index: protocols/dict.py
===================================================================
--- protocols/dict.py	(revision 24235)
+++ protocols/dict.py	(working copy)
@@ -5,7 +5,7 @@
 """
 Dict client protocol implementation.
 
-@author: U{Pavel Pergamenshchik<mailto:pp64@cornell.edu>}
+@author: Pavel Pergamenshchik
 """
 
 from twisted.protocols import basic
Index: protocols/ftp.py
===================================================================
--- protocols/ftp.py	(revision 24235)
+++ protocols/ftp.py	(working copy)
@@ -5,9 +5,9 @@
 """
 An FTP protocol implementation
 
-@author: U{Itamar Shtull-Trauring<mailto:itamarst@twistedmatrix.com>}
-@author: U{Jp Calderone<mailto:exarkun@divmod.com>}
-@author: U{Andrew Bennetts<mailto:spiv@twistedmatrix.com>}
+@author: Itamar Shtull-Trauring
+@author: Jp Calderone
+@author: Andrew Bennetts
 """
 
 # System Imports
Index: protocols/ident.py
===================================================================
--- protocols/ident.py	(revision 24235)
+++ protocols/ident.py	(working copy)
@@ -6,7 +6,7 @@
 """
 Ident protocol implementation.
 
-@author: U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
+@author: Jp Calderone
 """
 
 from __future__ import generators
Index: plugin.py
===================================================================
--- plugin.py	(revision 24235)
+++ plugin.py	(working copy)
@@ -6,8 +6,8 @@
 """
 Plugin system for Twisted.
 
-@author: U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
-@author: U{Glyph Lefkowitz<mailto:glyph@twistedmatrix.com>}
+@author: Jp Calderone
+@author: Glyph Lefkowitz
 """
 
 import os
Index: scripts/twistd.py
===================================================================
--- scripts/twistd.py	(revision 24235)
+++ scripts/twistd.py	(working copy)
@@ -5,7 +5,7 @@
 """
 The Twisted Daemon: platform-independent interface.
 
-@author: U{Christopher Armstrong<mailto:radix@twistedmatrix.com>}
+@author: Christopher Armstrong
 """
 
 from twisted.application import app
Index: conch/error.py
===================================================================
--- conch/error.py	(revision 24235)
+++ conch/error.py	(working copy)
@@ -7,7 +7,7 @@
 """
 An error to represent bad things happening in Conch.
 
-Maintainer: U{Paul Swartz<mailto:z3p@twistedmatrix.com>}
+Maintainer: Paul Swartz
 """
 
 class ConchError(Exception):
Index: conch/telnet.py
===================================================================
--- conch/telnet.py	(revision 24235)
+++ conch/telnet.py	(working copy)
@@ -5,7 +5,7 @@
 """
 Telnet protocol implementation.
 
-@author: U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
+@author: Jp Calderone
 """
 
 import struct
Index: conch/openssh_compat/__init__.py
===================================================================
--- conch/openssh_compat/__init__.py	(revision 24235)
+++ conch/openssh_compat/__init__.py	(working copy)
@@ -6,6 +6,6 @@
 """
 Support for OpenSSH configuration files.
 
-Maintainer: U{Paul Swartz<mailto:z3p@twistedmatrix.com>}
+Maintainer: Paul Swartz
 """
 
Index: conch/openssh_compat/primes.py
===================================================================
--- conch/openssh_compat/primes.py	(revision 24235)
+++ conch/openssh_compat/primes.py	(working copy)
@@ -6,7 +6,7 @@
 """
 Parsing for the moduli file, which contains Diffie-Hellman prime groups.
 
-Maintainer: U{Paul Swartz<mailto:z3p@twistedmatrix.com>}
+Maintainer: Paul Swartz
 """
 
 def parseModuliFile(filename):
Index: conch/__init__.py
===================================================================
--- conch/__init__.py	(revision 24235)
+++ conch/__init__.py	(working copy)
@@ -11,7 +11,7 @@
 Currently this contains the SSHv2 implementation, but it may work over other
 protocols in the future. (i.e. Telnet)
 
-Maintainer: U{Paul Swartz<mailto:z3p@twistedmatrix.com>}
+Maintainer: Paul Swartz
 """
 
 from twisted.conch._version import version
Index: conch/manhole_ssh.py
===================================================================
--- conch/manhole_ssh.py	(revision 24235)
+++ conch/manhole_ssh.py	(working copy)
@@ -5,7 +5,7 @@
 """
 insults/SSH integration support.
 
-@author: U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
+@author: Jp Calderone
 """
 
 from zope.interface import implements
Index: conch/ssh/common.py
===================================================================
--- conch/ssh/common.py	(revision 24235)
+++ conch/ssh/common.py	(working copy)
@@ -6,7 +6,7 @@
 """
 Common functions for the SSH classes.
 
-Maintainer: U{Paul Swartz<mailto:z3p@twistedmatrix.com>}
+Maintainer: Paul Swartz
 """
 
 import struct, warnings
Index: conch/ssh/transport.py
===================================================================
--- conch/ssh/transport.py	(revision 24235)
+++ conch/ssh/transport.py	(working copy)
@@ -8,7 +8,7 @@
 encryption and the compression.  The transport layer is described in
 RFC 4253.
 
-Maintainer: U{Paul Swartz<mailto:z3p@twistedmatrix.com>}
+Maintainer: Paul Swartz
 """
 
 # base library imports
Index: conch/ssh/agent.py
===================================================================
--- conch/ssh/agent.py	(revision 24235)
+++ conch/ssh/agent.py	(working copy)
@@ -5,7 +5,7 @@
 """
 Implements the old SSHv1 key agent protocol.
 
-Maintainer: U{Paul Swartz<mailto:z3p@twistedmatrix.com>}
+Maintainer: Paul Swartz
 """
 
 import struct
Index: conch/ssh/connection.py
===================================================================
--- conch/ssh/connection.py	(revision 24235)
+++ conch/ssh/connection.py	(working copy)
@@ -8,7 +8,7 @@
 This module contains the implementation of the ssh-connection service, which
 allows access to the shell and port-forwarding.
 
-Maintainer: U{Paul Swartz<mailto:z3p@twistedmatrix.com>}
+Maintainer: Paul Swartz
 """
 
 import struct
Index: conch/ssh/__init__.py
===================================================================
--- conch/ssh/__init__.py	(revision 24235)
+++ conch/ssh/__init__.py	(working copy)
@@ -6,5 +6,5 @@
 """
 An SSHv2 implementation for Twisted.  Part of the Twisted.Conch package.
 
-Maintainer: U{Paul Swartz<mailto:z3p@twistedmatrix.com>}
+Maintainer: Paul Swartz
 """
Index: conch/ssh/userauth.py
===================================================================
--- conch/ssh/userauth.py	(revision 24235)
+++ conch/ssh/userauth.py	(working copy)
@@ -7,7 +7,7 @@
 Implementation of the ssh-userauth service.
 Currently implemented authentication types are public-key and password.
 
-Maintainer: U{Paul Swartz<mailto:z3p@twistedmatrix.com>}
+Maintainer: Paul Swartz
 """
 
 import struct
Index: conch/ssh/service.py
===================================================================
--- conch/ssh/service.py	(revision 24235)
+++ conch/ssh/service.py	(working copy)
@@ -5,7 +5,7 @@
 The parent class for all the SSH services.  Currently implemented services
 are ssh-userauth and ssh-connection.
 
-Maintainer: U{Paul Swartz<mailto:z3p@twistedmatrix.com>}
+Maintainer: Paul Swartz
 """
 
 
Index: conch/ssh/forwarding.py
===================================================================
--- conch/ssh/forwarding.py	(revision 24235)
+++ conch/ssh/forwarding.py	(working copy)
@@ -7,7 +7,7 @@
 This module contains the implementation of the TCP forwarding, which allows
 clients and servers to forward arbitrary TCP data across the connection.
 
-Maintainer: U{Paul Swartz<mailto:z3p@twistedmatrix.com>}
+Maintainer: Paul Swartz
 """
 
 import struct
Index: conch/ssh/asn1.py
===================================================================
--- conch/ssh/asn1.py	(revision 24235)
+++ conch/ssh/asn1.py	(working copy)
@@ -6,7 +6,7 @@
 """
 A basic ASN.1 parser to parse private SSH keys.
 
-Maintainer: U{Paul Swartz<mailto:z3p@twistedmatrix.com>}
+Maintainer: Paul Swartz
 """
 
 from Crypto.Util import number
Index: conch/ssh/factory.py
===================================================================
--- conch/ssh/factory.py	(revision 24235)
+++ conch/ssh/factory.py	(working copy)
@@ -5,7 +5,7 @@
 A Factory for SSH servers, along with an OpenSSHFactory to use the same
 data sources as OpenSSH.
 
-Maintainer: U{Paul Swartz<mailto:z3p@twistedmatrix.com>}
+Maintainer: Paul Swartz
 """
 
 try:
Index: conch/ssh/channel.py
===================================================================
--- conch/ssh/channel.py	(revision 24235)
+++ conch/ssh/channel.py	(working copy)
@@ -7,7 +7,7 @@
 The parent class for all the SSH Channels.  Currently implemented channels
 are session. direct-tcp, and forwarded-tcp.
 
-Maintainer: U{Paul Swartz<mailto:z3p@twistedmatrix.com>}
+Maintainer: Paul Swartz
 """
 
 from twisted.python import log
Index: conch/ssh/session.py
===================================================================
--- conch/ssh/session.py	(revision 24235)
+++ conch/ssh/session.py	(working copy)
@@ -8,7 +8,7 @@
 This module contains the implementation of SSHSession, which (by default)
 allows access to a shell and a python interpreter over SSH.
 
-Maintainer: U{Paul Swartz<mailto:z3p@twistedmatrix.com>}
+Maintainer: Paul Swartz
 """
 
 import struct
Index: conch/recvline.py
===================================================================
--- conch/recvline.py	(revision 24235)
+++ conch/recvline.py	(working copy)
@@ -5,7 +5,7 @@
 """
 Basic line editing support.
 
-@author: U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
+@author: Jp Calderone
 """
 
 import string
Index: conch/insults/helper.py
===================================================================
--- conch/insults/helper.py	(revision 24235)
+++ conch/insults/helper.py	(working copy)
@@ -5,7 +5,7 @@
 """
 Partial in-memory terminal emulator
 
-@author: U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
+@author: Jp Calderone
 """
 
 import re, string
Index: conch/insults/insults.py
===================================================================
--- conch/insults/insults.py	(revision 24235)
+++ conch/insults/insults.py	(working copy)
@@ -5,7 +5,7 @@
 """
 VT102 and VT220 terminal manipulation.
 
-@author: U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
+@author: Jp Calderone
 """
 
 from zope.interface import implements, Interface
Index: conch/insults/text.py
===================================================================
--- conch/insults/text.py	(revision 24235)
+++ conch/insults/text.py	(working copy)
@@ -49,7 +49,7 @@
 
  | A.underline[A.fg.red[\"Hello\", -A.underline[\" world\"]]]
 
-@author: U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
+@author: Jp Calderone
 """
 
 from twisted.conch.insults import helper, insults
Index: conch/insults/window.py
===================================================================
--- conch/insults/window.py	(revision 24235)
+++ conch/insults/window.py	(working copy)
@@ -3,7 +3,7 @@
 """
 Simple insults-based widget library
 
-@author: U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
+@author: Jp Calderone
 """
 
 import array
Index: conch/stdio.py
===================================================================
--- conch/stdio.py	(revision 24235)
+++ conch/stdio.py	(working copy)
@@ -5,7 +5,7 @@
 """
 Asynchronous local terminal input handling
 
-@author: U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
+@author: Jp Calderone
 """
 
 import os, tty, sys, termios
Index: conch/manhole.py
===================================================================
--- conch/manhole.py	(revision 24235)
+++ conch/manhole.py	(working copy)
@@ -10,7 +10,7 @@
 code with support for VT102 terminals, control code handling (^C, ^D, ^Q),
 and reasonable handling of Deferreds.
 
-@author: U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
+@author: Jp Calderone
 """
 
 import code, sys, StringIO, tokenize
Index: conch/mixin.py
===================================================================
--- conch/mixin.py	(revision 24235)
+++ conch/mixin.py	(working copy)
@@ -8,7 +8,7 @@
 This module provides a single mixin class which allows protocols to
 collapse numerous small writes into a single larger one.
 
-@author: U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
+@author: Jp Calderone
 """
 
 from twisted.internet import reactor
Index: conch/ui/tkvt100.py
===================================================================
--- conch/ui/tkvt100.py	(revision 24235)
+++ conch/ui/tkvt100.py	(working copy)
@@ -5,7 +5,7 @@
 
 """Module to emulate a VT100 terminal in Tkinter.
 
-Maintainer: U{Paul Swartz <mailto:z3p@twistedmatrix.com>}
+Maintainer: Paul Swartz
 """
 
 import Tkinter, tkFont
Index: conch/ui/ansi.py
===================================================================
--- conch/ui/ansi.py	(revision 24235)
+++ conch/ui/ansi.py	(working copy)
@@ -4,7 +4,7 @@
 #
 """Module to parse ANSI escape sequences
 
-Maintainer: U{Jean-Paul Calderone <exarkun@twistedmatrix.com>}
+Maintainer: Jean-Paul Calderone
 """
 
 import string
Index: conch/ui/__init__.py
===================================================================
--- conch/ui/__init__.py	(revision 24235)
+++ conch/ui/__init__.py	(working copy)
@@ -7,5 +7,5 @@
 """
 twisted.conch.ui is home to the UI elements for tkconch.
 
-Maintainer: U{Paul Swartz<mailto:z3p@twistedmatrix.com>}
+Maintainer: Paul Swartz
 """
Index: conch/client/agent.py
===================================================================
--- conch/client/agent.py	(revision 24235)
+++ conch/client/agent.py	(working copy)
@@ -5,7 +5,7 @@
 """
 Accesses the key agent for user authentication.
 
-Maintainer: U{Paul Swartz<mailto:z3p@twistedmatrix.com>}
+Maintainer: Paul Swartz
 """
 
 from twisted.conch.ssh import agent, channel
Index: conch/client/__init__.py
===================================================================
--- conch/client/__init__.py	(revision 24235)
+++ conch/client/__init__.py	(working copy)
@@ -5,5 +5,5 @@
 """
 Client support code for Conch.
 
-Maintainer: U{Paul Swartz<mailto:z3p@twistedmatrix.com>}
+Maintainer: Paul Swartz
 """
Index: conch/manhole_tap.py
===================================================================
--- conch/manhole_tap.py	(revision 24235)
+++ conch/manhole_tap.py	(working copy)
@@ -4,7 +4,7 @@
 """
 TAP plugin for creating telnet- and ssh-accessible manhole servers.
 
-@author: U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
+@author: Jp Calderone
 """
 
 from zope.interface import implements
Index: pair/__init__.py
===================================================================
--- pair/__init__.py	(revision 24235)
+++ pair/__init__.py	(working copy)
@@ -12,7 +12,7 @@
 See also twisted.protocols.ethernet, twisted.protocols.ip,
 twisted.protocols.raw and twisted.protocols.rawudp.
 
-Maintainer: U{Tommi Virtanen <mailto:tv@twistedmatrix.com>}
+Maintainer: Tommi Virtanen
 
 """
 
Index: words/protocols/irc.py
===================================================================
--- words/protocols/irc.py	(revision 24235)
+++ words/protocols/irc.py	(working copy)
@@ -20,7 +20,7 @@
 
 Test coverage needs to be better.
 
-@author: U{Kevin Turner<mailto:acapnotic@twistedmatrix.com>}
+@author: Kevin Turner
 
 @see: RFC 1459: Internet Relay Chat Protocol
 @see: RFC 2812: Internet Relay Chat: Client Protocol
Index: words/protocols/msn.py
===================================================================
--- words/protocols/msn.py	(revision 24235)
+++ words/protocols/msn.py	(working copy)
@@ -70,7 +70,7 @@
 - font handling
 - switchboard factory
 
-@author: U{Sam Jordan<mailto:sam@twistedmatrix.com>}
+@author: Sam Jordan
 """
 
 from __future__ import nested_scopes
Index: words/protocols/toc.py
===================================================================
--- words/protocols/toc.py	(revision 24235)
+++ words/protocols/toc.py	(working copy)
@@ -13,7 +13,7 @@
 
 This module is deprecated.
 
-Maintainer: U{Paul Swartz<mailto:z3p@twistedmatrix.com>}
+Maintainer: Paul Swartz
 """
 
 # twisted imports
Index: words/protocols/oscar.py
===================================================================
--- words/protocols/oscar.py	(revision 24235)
+++ words/protocols/oscar.py	(working copy)
@@ -6,7 +6,7 @@
 """
 An implementation of the OSCAR protocol, which AIM and ICQ use to communcate.
 
-Maintainer: U{Paul Swartz<mailto:z3p@twistedmatrix.com>}
+Maintainer: Paul Swartz
 """
 
 from __future__ import nested_scopes
Index: words/xish/xmlstream.py
===================================================================
--- words/xish/xmlstream.py	(revision 24235)
+++ words/xish/xmlstream.py	(working copy)
@@ -13,7 +13,7 @@
 The most prominent use of XML Streams is Jabber, but this module is generically
 usable. See Twisted Words for Jabber specific protocol support.
 
-Maintainer: U{Ralph Meijer<mailto:twisted@ralphm.ik.nu>}
+Maintainer: Ralph Meijer
 """
 
 from twisted.internet import protocol
Index: vfs/__init__.py
===================================================================
--- vfs/__init__.py	(revision 24235)
+++ vfs/__init__.py	(working copy)
@@ -2,5 +2,5 @@
 
 This package is a work in progress and may change incompatibly at any time.
 
-Maintainer: U{Andy Gayton<mailto:andy@thecablelounge.com>}
+Maintainer: Andy Gayton
 """
Index: web2/xmlrpc.py
===================================================================
--- web2/xmlrpc.py	(revision 24235)
+++ web2/xmlrpc.py	(working copy)
@@ -7,7 +7,7 @@
 """
 A generic resource for publishing objects via XML-RPC.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 # System Imports
Index: web2/http.py
===================================================================
--- web2/http.py	(revision 24235)
+++ web2/http.py	(working copy)
@@ -6,7 +6,7 @@
 
 The second coming.
 
-Maintainer: U{James Y Knight <mailto:foom@fuhm.net>}
+Maintainer: James Y Knight
 
 """
 #        import traceback; log.msg(''.join(traceback.format_stack()))
Index: persisted/dirdbm.py
===================================================================
--- persisted/dirdbm.py	(revision 24235)
+++ persisted/dirdbm.py	(working copy)
@@ -17,7 +17,7 @@
 No files should be placed in the working directory of a DirDBM save those
 created by the DirDBM itself!
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 
Index: persisted/journal/__init__.py
===================================================================
--- persisted/journal/__init__.py	(revision 24235)
+++ persisted/journal/__init__.py	(working copy)
@@ -6,5 +6,5 @@
 """
 Command-journalling persistence framework inspired by Prevayler.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
Index: persisted/sob.py
===================================================================
--- persisted/sob.py	(revision 24235)
+++ persisted/sob.py	(working copy)
@@ -6,7 +6,7 @@
 """
 Save and load Small OBjects to and from files, using various formats.
 
-Maintainer: U{Moshe Zadka<mailto:moshez@twistedmatrix.com>}
+Maintainer: Moshe Zadka
 """
 
 import os, md5, sys
Index: names/server.py
===================================================================
--- names/server.py	(revision 24235)
+++ names/server.py	(working copy)
@@ -15,7 +15,7 @@
 This violates the most basic RFC and is just plain annoying
 for resolvers to deal with.  Fix it.
 
-@author: U{Jp Calderone <mailto:exarkun@twistedmatrix.com>}
+@author: Jp Calderone
 """
 
 from __future__ import nested_scopes
Index: names/client.py
===================================================================
--- names/client.py	(revision 24235)
+++ names/client.py	(working copy)
@@ -16,7 +16,7 @@
 Future plans: Proper nameserver acquisition on Windows/MacOS,
 better caching, respect timeouts
 
-@author: U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
+@author: Jp Calderone
 """
 
 from __future__ import nested_scopes
Index: names/resolve.py
===================================================================
--- names/resolve.py	(revision 24235)
+++ names/resolve.py	(working copy)
@@ -8,7 +8,7 @@
 Future Plans: This needs someway to specify which resolver answered
 the query, or someway to specify (authority|ttl|cache behavior|more?)
 
-@author: U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
+@author: Jp Calderone
 """
 
 from twisted.internet import defer, interfaces
Index: names/root.py
===================================================================
--- names/root.py	(revision 24235)
+++ names/root.py	(working copy)
@@ -7,7 +7,7 @@
 Resolver implementation for querying successive authoritative servers to
 lookup a record, starting from the root nameservers.
 
-@author: U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
+@author: Jp Calderone
 
 todo::
     robustify it
Index: names/dns.py
===================================================================
--- names/dns.py	(revision 24235)
+++ names/dns.py	(working copy)
@@ -10,8 +10,8 @@
     - Get rid of some toplevels, maybe.
     - Put in a better lookupRecordType implementation.
 
-@author: U{Moshe Zadka<mailto:moshez@twistedmatrix.com>},
-         U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
+@author: Moshe Zadka
+@author: Jp Calderone
 """
 
 # System imports
Index: application/internet.py
===================================================================
--- application/internet.py	(revision 24235)
+++ application/internet.py	(working copy)
@@ -29,7 +29,7 @@
 C{TCPServer(8080, server.Site(r))}.  See the documentation for the
 reactor.listen/connect* methods for more information.
 
-Maintainer: U{Moshe Zadka<mailto:moshez@twistedmatrix.com>}
+Maintainer: Moshe Zadka
 """
 
 from twisted.python import log
Index: application/service.py
===================================================================
--- application/service.py	(revision 24235)
+++ application/service.py	(working copy)
@@ -11,7 +11,7 @@
 there is need to access them through the hierarchy (from a parent or
 a sibling).
 
-Maintainer: U{Moshe Zadka<mailto:moshez@twistedmatrix.com>}
+Maintainer: Moshe Zadka
 """
 
 from zope.interface import implements, Interface, Attribute
Index: application/strports.py
===================================================================
--- application/strports.py	(revision 24235)
+++ application/strports.py	(working copy)
@@ -31,7 +31,7 @@
 
 See specific function documentation for more information.
 
-Maintainer: U{Moshe Zadka<mailto:moshez@twistedmatrix.com>}
+Maintainer: Moshe Zadka
 """
 from __future__ import generators
 
Index: flow/wrap.py
===================================================================
--- flow/wrap.py	(revision 24235)
+++ flow/wrap.py	(working copy)
@@ -2,7 +2,7 @@
 # See LICENSE for details.
 
 #
-# Author: Clark Evans  (cce@clarkevans.com)
+# Author: Clark Evans
 # 
  
 """ flow.wrap 
Index: flow/base.py
===================================================================
--- flow/base.py	(revision 24235)
+++ flow/base.py	(working copy)
@@ -2,7 +2,7 @@
 # See LICENSE for details.
 
 #
-# Author: Clark Evans  (cce@clarkevans.com)
+# Author: Clark Evans
 #
 """
 flow.base
Index: flow/threads.py
===================================================================
--- flow/threads.py	(revision 24235)
+++ flow/threads.py	(working copy)
@@ -2,7 +2,7 @@
 # See LICENSE for details.
 
 #
-# Author: Clark Evans  (cce@clarkevans.com)
+# Author: Clark Evans
 # Stability: The API is stable, but the implementation may still
 #            have one or more bugs; threads are tough.
 #
Index: flow/pipe.py
===================================================================
--- flow/pipe.py	(revision 24235)
+++ flow/pipe.py	(working copy)
@@ -2,7 +2,7 @@
 # See LICENSE for details.
 
 #
-# Author: Clark Evans  (cce@clarkevans.com)
+# Author: Clark Evans
 
 """
 flow.pipe
Index: flow/web.py
===================================================================
--- flow/web.py	(revision 24235)
+++ flow/web.py	(working copy)
@@ -2,7 +2,7 @@
 # See LICENSE for details.
 
 #
-# Author: Clark Evans  (cce@clarkevans.com)
+# Author: Clark Evans
 
 """ flow.web
 
Index: flow/protocol.py
===================================================================
--- flow/protocol.py	(revision 24235)
+++ flow/protocol.py	(working copy)
@@ -2,7 +2,7 @@
 # See LICENSE for details.
 
 #
-# Author: Clark Evans  (cce@clarkevans.com)
+# Author: Clark Evans
 #
 
 """
Index: flow/controller.py
===================================================================
--- flow/controller.py	(revision 24235)
+++ flow/controller.py	(working copy)
@@ -2,7 +2,7 @@
 # See LICENSE for details.
 
 #
-# Author: Clark Evans  (cce@clarkevans.com)
+# Author: Clark Evans
 # 
 
 """
Index: flow/stage.py
===================================================================
--- flow/stage.py	(revision 24235)
+++ flow/stage.py	(working copy)
@@ -2,7 +2,7 @@
 # See LICENSE for details.
 
 #
-# Author: Clark Evans  (cce@clarkevans.com)
+# Author: Clark Evans
 #
 
 """
Index: flow/flow.py
===================================================================
--- flow/flow.py	(revision 24235)
+++ flow/flow.py	(working copy)
@@ -2,7 +2,7 @@
 # See LICENSE for details.
 
 #
-# Author: Clark Evans  (cce@clarkevans.com)
+# Author: Clark Evans
 #
 
 """
Index: news/news.py
===================================================================
--- news/news.py	(revision 24235)
+++ news/news.py	(working copy)
@@ -3,7 +3,7 @@
 
 
 """
-Maintainer: U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
+Maintainer: Jp Calderone
 """
 
 from twisted.news import nntp
Index: news/nntp.py
===================================================================
--- news/nntp.py	(revision 24235)
+++ news/nntp.py	(working copy)
@@ -6,7 +6,7 @@
 """
 NNTP protocol support.
 
-Maintainer: U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
+Maintainer: Jp Calderone
 
 The following protocol commands are currently understood::
 
Index: news/database.py
===================================================================
--- news/database.py	(revision 24235)
+++ news/database.py	(working copy)
@@ -6,7 +6,7 @@
 """
 News server backend implementations
 
-Maintainer: U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
+Maintainer: Jp Calderone
 
 Future Plans: A PyFramer-based backend and a new backend interface that is
 less NNTP specific
Index: web/http.py
===================================================================
--- web/http.py	(revision 24235)
+++ web/http.py	(working copy)
@@ -12,7 +12,7 @@
  - Accept chunked data from clients in server.
  - Other missing HTTP features from the RFC.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 # system imports
Index: web/soap.py
===================================================================
--- web/soap.py	(revision 24235)
+++ web/soap.py	(working copy)
@@ -8,7 +8,7 @@
 
 Requires SOAPpy 0.10.1 or later.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 
 Future plans:
 SOAPContext support of some kind.
Index: web/xmlrpc.py
===================================================================
--- web/xmlrpc.py	(revision 24235)
+++ web/xmlrpc.py	(working copy)
@@ -5,7 +5,7 @@
 """
 A generic resource for publishing objects via XML-RPC.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 # System Imports
Index: cred/util.py
===================================================================
--- cred/util.py	(revision 24235)
+++ cred/util.py	(working copy)
@@ -8,7 +8,7 @@
 
 These are currently for challenge-response shared secret authentication.
 
-Maintainer: U{Glyph Lefkowitz<mailto:glyph@twistedmatrix.com>}
+Maintainer: Glyph Lefkowitz
 """
 
 # System Imports
Index: test/test_ftp.py
===================================================================
--- test/test_ftp.py	(revision 24235)
+++ test/test_ftp.py	(working copy)
@@ -4,7 +4,7 @@
 """
 FTP tests.
 
-Maintainer: U{Andrew Bennetts<mailto:spiv@twistedmatrix.com>}
+Maintainer: Andrew Bennetts
 """
 
 import os.path
Index: plugins/__init__.py
===================================================================
--- plugins/__init__.py	(revision 24235)
+++ plugins/__init__.py	(working copy)
@@ -8,8 +8,8 @@
 this is the only place where an __init__.py is necessary, thanks to
 the __path__ variable.
 
-@author: U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
-@author: U{Glyph Lefkowitz<mailto:glyph@twistedmatrix.com>}
+@author: Jp Calderone
+@author: Glyph Lefkowitz
 """
 
 from twisted.plugin import pluginPackagePaths
Index: mail/alias.py
===================================================================
--- mail/alias.py	(revision 24235)
+++ mail/alias.py	(working copy)
@@ -7,7 +7,7 @@
 """
 Support for aliases(5) configuration files
 
-@author: U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
+@author: Jp Calderone
 
 TODO::
     Monitor files for reparsing
Index: mail/__init__.py
===================================================================
--- mail/__init__.py	(revision 24235)
+++ mail/__init__.py	(working copy)
@@ -7,7 +7,7 @@
 
 Twisted Mail: a Twisted E-Mail Server.
 
-Maintainer: U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
+Maintainer: Jp Calderone
 
 """
 
Index: mail/imap4.py
===================================================================
--- mail/imap4.py	(revision 24235)
+++ mail/imap4.py	(working copy)
@@ -6,7 +6,7 @@
 """
 An IMAP4 protocol implementation
 
-@author: U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
+@author: Jp Calderone
 
 To do::
   Suspend idle timeout while server is processing
Index: mail/pop3client.py
===================================================================
--- mail/pop3client.py	(revision 24235)
+++ mail/pop3client.py	(working copy)
@@ -7,7 +7,7 @@
 
 Don't use this module directly.  Use twisted.mail.pop3 instead.
 
-@author: U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
+@author: Jp Calderone
 """
 
 import re, md5
Index: mail/pop3.py
===================================================================
--- mail/pop3.py	(revision 24235)
+++ mail/pop3.py	(working copy)
@@ -7,8 +7,8 @@
 """
 Post-office Protocol version 3
 
-@author: U{Glyph Lefkowitz<mailto:glyph@twistedmatrix.com>}
-@author: U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
+@author: Glyph Lefkowitz
+@author: Jp Calderone
 """
 
 import string
Index: internet/udp.py
===================================================================
--- internet/udp.py	(revision 24235)
+++ internet/udp.py	(working copy)
@@ -9,7 +9,7 @@
 
 Please do not use this module directly.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 # System Imports
Index: internet/threads.py
===================================================================
--- internet/threads.py	(revision 24235)
+++ internet/threads.py	(working copy)
@@ -6,7 +6,7 @@
 
 For basic support see reactor threading API docs.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 import Queue
Index: internet/epollreactor.py
===================================================================
--- internet/epollreactor.py	(revision 24235)
+++ internet/epollreactor.py	(working copy)
@@ -10,7 +10,7 @@
     from twisted.internet import epollreactor
     epollreactor.install()
 
-Maintainer: U{Jp Calderone <mailto:exarkun@twistedmatrix.com>}
+Maintainer: Jp Calderone
 """
 
 import sys, errno
Index: internet/wxsupport.py
===================================================================
--- internet/wxsupport.py	(revision 24235)
+++ internet/wxsupport.py	(working copy)
@@ -18,7 +18,7 @@
 On Windows the Twisted event loop might block when dialogs are open
 or menus are selected.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 import warnings
Index: internet/posixbase.py
===================================================================
--- internet/posixbase.py	(revision 24235)
+++ internet/posixbase.py	(working copy)
@@ -7,7 +7,7 @@
 """
 Posix reactor base class
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 import warnings
Index: internet/ssl.py
===================================================================
--- internet/ssl.py	(revision 24235)
+++ internet/ssl.py	(working copy)
@@ -20,7 +20,7 @@
     - support for switching TCP into SSL
     - more options
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 # If something goes wrong, most notably an OpenSSL import failure,
Index: internet/task.py
===================================================================
--- internet/task.py	(revision 24235)
+++ internet/task.py	(working copy)
@@ -5,7 +5,7 @@
 """
 Scheduling utility methods and classes.
 
-@author: U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
+@author: Jp Calderone
 """
 
 __metaclass__ = type
Index: internet/abstract.py
===================================================================
--- internet/abstract.py	(revision 24235)
+++ internet/abstract.py	(working copy)
@@ -6,7 +6,7 @@
 """
 Support for generic select()able objects.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 from zope.interface import implements
Index: internet/cfreactor.py
===================================================================
--- internet/cfreactor.py	(revision 24235)
+++ internet/cfreactor.py	(working copy)
@@ -19,7 +19,7 @@
 otherwise the current CFRunLoop.  Either way, if one doesn't exist, it will
 be created).
 
-Maintainer: U{Bob Ippolito<mailto:bob@redivi.com>}
+Maintainer: Bob Ippolito
 """
 
 __all__ = ['install']
Index: internet/tksupport.py
===================================================================
--- internet/tksupport.py	(revision 24235)
+++ internet/tksupport.py	(working copy)
@@ -5,7 +5,7 @@
 """
 This module integrates Tkinter with twisted.internet's mainloop.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 
 To use, do::
 
Index: internet/interfaces.py
===================================================================
--- internet/interfaces.py	(revision 24235)
+++ internet/interfaces.py	(working copy)
@@ -5,7 +5,7 @@
 """
 Interface documentation.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 from zope.interface import Interface, Attribute
Index: internet/unix.py
===================================================================
--- internet/unix.py	(revision 24235)
+++ internet/unix.py	(working copy)
@@ -8,7 +8,7 @@
 
 End users shouldn't use this module directly - use the reactor APIs instead.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 # System imports
Index: internet/pyuisupport.py
===================================================================
--- internet/pyuisupport.py	(revision 24235)
+++ internet/pyuisupport.py	(working copy)
@@ -5,7 +5,7 @@
 """
 This module integrates PyUI with twisted.internet's mainloop.
 
-Maintainer: U{Jp Calderone<mailto:exarkun@twistedmatrix.com>}
+Maintainer: Jp Calderone
 
 See doc/examples/pyuidemo.py for example usage.
 """
Index: internet/main.py
===================================================================
--- internet/main.py	(revision 24235)
+++ internet/main.py	(working copy)
@@ -8,7 +8,7 @@
 In general, this module should not be used, other than by reactor authors
 who need to use the 'installReactor' method.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 import error
Index: internet/default.py
===================================================================
--- internet/default.py	(revision 24235)
+++ internet/default.py	(working copy)
@@ -8,7 +8,7 @@
 """
 Deprecated module that used to contain SelectReactor and PosixReactorBase
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 import warnings
Index: internet/win32eventreactor.py
===================================================================
--- internet/win32eventreactor.py	(revision 24235)
+++ internet/win32eventreactor.py	(working copy)
@@ -7,7 +7,7 @@
 
 This requires win32all or ActivePython to be installed.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 
 
 LIMITATIONS:
Index: internet/defer.py
===================================================================
--- internet/defer.py	(revision 24235)
+++ internet/defer.py	(working copy)
@@ -6,7 +6,7 @@
 """
 Support for results that aren't immediately available.
 
-Maintainer: U{Glyph Lefkowitz<mailto:glyph@twistedmatrix.com>}
+Maintainer: Glyph Lefkowitz
 """
 
 from __future__ import nested_scopes, generators
@@ -628,8 +628,6 @@
 
 def deferredGenerator(f):
     """
-    Maintainer: U{Christopher Armstrong<mailto:radix@twistedmatrix.com>}
-
     deferredGenerator and waitForDeferred help you write Deferred-using code
     that looks like a regular sequential function. If your code has a minimum
     requirement of Python 2.5, consider the use of L{inlineCallbacks} instead,
@@ -778,8 +776,6 @@
 
 def inlineCallbacks(f):
     """
-    Maintainer: U{Christopher Armstrong<mailto:radix@twistedmatrix.com>}
-
     WARNING: this function will not work in Python 2.4 and earlier!
 
     inlineCallbacks helps you write Deferred-using code that looks like a
Index: internet/pollreactor.py
===================================================================
--- internet/pollreactor.py	(revision 24235)
+++ internet/pollreactor.py	(working copy)
@@ -11,7 +11,7 @@
     from twisted.internet import pollreactor
     pollreactor.install()
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 # System imports
Index: internet/tcp.py
===================================================================
--- internet/tcp.py	(revision 24235)
+++ internet/tcp.py	(working copy)
@@ -7,7 +7,7 @@
 
 End users shouldn't use this module directly - use the reactor APIs instead.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 
Index: internet/gtk2reactor.py
===================================================================
--- internet/gtk2reactor.py	(revision 24235)
+++ internet/gtk2reactor.py	(working copy)
@@ -19,7 +19,7 @@
 event loop or the GTK+ event loop which is based on it but adds GUI
 integration.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 # System Imports
Index: internet/error.py
===================================================================
--- internet/error.py	(revision 24235)
+++ internet/error.py	(working copy)
@@ -5,7 +5,7 @@
 """
 Exceptions and errors for use in twisted.internet modules.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 import socket
Index: internet/kqreactor.py
===================================================================
--- internet/kqreactor.py	(revision 24235)
+++ internet/kqreactor.py	(working copy)
@@ -14,7 +14,7 @@
 This reactor only works on FreeBSD and requires PyKQueue 1.3, which is
 available at:  U{http://people.freebsd.org/~dwhite/PyKQueue/}
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 
 
 
Index: internet/selectreactor.py
===================================================================
--- internet/selectreactor.py	(revision 24235)
+++ internet/selectreactor.py	(working copy)
@@ -6,7 +6,7 @@
 """
 Select reactor
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 from time import sleep
Index: internet/protocol.py
===================================================================
--- internet/protocol.py	(revision 24235)
+++ internet/protocol.py	(working copy)
@@ -10,7 +10,7 @@
 Start here if you are looking to write a new protocol implementation for
 Twisted.  The Protocol class contains some introductory material.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 import random
Index: internet/_posixstdio.py
===================================================================
--- internet/_posixstdio.py	(revision 24235)
+++ internet/_posixstdio.py	(working copy)
@@ -8,7 +8,7 @@
     Rewrite to use the reactor instead of an ad-hoc mechanism for connecting
         protocols to transport.
 
-Maintainer: U{James Y Knight <mailto:foom@fuhm.net>}
+Maintainer: James Y Knight
 """
 
 import warnings
Index: internet/gtkreactor.py
===================================================================
--- internet/gtkreactor.py	(revision 24235)
+++ internet/gtkreactor.py	(working copy)
@@ -12,7 +12,7 @@
 Then use twisted.internet APIs as usual.  The other methods here are not
 intended to be called directly.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 import sys
Index: internet/base.py
===================================================================
--- internet/base.py	(revision 24235)
+++ internet/base.py	(working copy)
@@ -6,7 +6,7 @@
 """
 Very basic functionality for a Reactor implementation.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 import socket # needed only for sync-dns
Index: internet/process.py
===================================================================
--- internet/process.py	(revision 24235)
+++ internet/process.py	(working copy)
@@ -7,7 +7,7 @@
 
 Do NOT use this module directly - use reactor.spawnProcess() instead.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 # System Imports
Index: internet/wxreactor.py
===================================================================
--- internet/wxreactor.py	(revision 24235)
+++ internet/wxreactor.py	(working copy)
@@ -22,7 +22,7 @@
 expected and probably does not reflect on the reactor's ability to run
 real applications.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 import Queue
Index: internet/glib2reactor.py
===================================================================
--- internet/glib2reactor.py	(revision 24235)
+++ internet/glib2reactor.py	(working copy)
@@ -17,7 +17,7 @@
 event loop or the GTK+ event loop which is based on it but adds GUI
 integration.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 from twisted.internet import gtk2reactor
Index: internet/_threadedselect.py
===================================================================
--- internet/_threadedselect.py	(revision 24235)
+++ internet/_threadedselect.py	(working copy)
@@ -9,7 +9,7 @@
 """
 Threaded select reactor
 
-Maintainer: U{Bob Ippolito<mailto:bob@redivi.com>}
+Maintainer: Bob Ippolito
 
 
 The threadedselectreactor is a specialized reactor for integrating with
Index: internet/stdio.py
===================================================================
--- internet/stdio.py	(revision 24235)
+++ internet/stdio.py	(working copy)
@@ -20,7 +20,7 @@
         protocols to transport.
 
 
-Maintainer: U{James Y Knight <mailto:foom@fuhm.net>}
+Maintainer: James Y Knight
 """
 
 from twisted.python.runtime import platform
Index: lore/__init__.py
===================================================================
--- lore/__init__.py	(revision 24235)
+++ lore/__init__.py	(working copy)
@@ -5,7 +5,7 @@
 '''
 The Twisted Documentation Generation System
 
-Maintainer: U{Andrew Bennetts<mailto:spiv@twistedmatrix.com>}
+Maintainer: Andrew Bennetts
 '''
 
 # TODO
Index: lore/nevowlore.py
===================================================================
--- lore/nevowlore.py	(revision 24235)
+++ lore/nevowlore.py	(working copy)
@@ -14,7 +14,7 @@
 
   lore -inevow --config pageclass=some.module.SomePageSubclass [other-opts]
 
-Maintainer: U{Christopher Armstrong<mailto:radix@twistedmatrix.com>}
+Maintainer: Christopher Armstrong
 
 """
 
Index: runner/inetdtap.py
===================================================================
--- runner/inetdtap.py	(revision 24235)
+++ runner/inetdtap.py	(working copy)
@@ -6,7 +6,7 @@
 """
 Twisted inetd TAP support
 
-Maintainer: U{Andrew Bennetts<mailto:spiv@twistedmatrix.com>}
+Maintainer: Andrew Bennetts
 
 Future Plans: more configurability.
 """
Index: runner/inetdconf.py
===================================================================
--- runner/inetdconf.py	(revision 24235)
+++ runner/inetdconf.py	(working copy)
@@ -5,7 +5,7 @@
 """
 Parser for inetd.conf files
 
-Maintainer: U{Andrew Bennetts<mailto:spiv@twistedmatrix.com>}
+Maintainer: Andrew Bennetts
 
 Future Plans: xinetd configuration file support?
 """
Index: runner/__init__.py
===================================================================
--- runner/__init__.py	(revision 24235)
+++ runner/__init__.py	(working copy)
@@ -1,7 +1,7 @@
 """
 Twisted runer: run and monitor processes
 
-Maintainer: U{Andrew Bennetts<mailto:spiv@twistedmatrix.com>}
+Maintainer: Andrew Bennetts
 
 classic inetd(8) support:
 Future Plans: The basic design should be final.  There are some bugs that need
Index: runner/inetd.py
===================================================================
--- runner/inetd.py	(revision 24235)
+++ runner/inetd.py	(working copy)
@@ -6,7 +6,7 @@
 """
 Twisted inetd.
 
-Maintainer: U{Andrew Bennetts<mailto:spiv@twistedmatrix.com>}
+Maintainer: Andrew Bennetts
 
 Future Plans: Bugfixes.  Specifically for UDP and Sun-RPC, which don't work
 correctly yet.
Index: trial/test/test_pyunitcompat.py
===================================================================
--- trial/test/test_pyunitcompat.py	(revision 24235)
+++ trial/test/test_pyunitcompat.py	(working copy)
@@ -1,7 +1,7 @@
 # Copyright (c) 2001-2007 Twisted Matrix Laboratories.
 # See LICENSE for details.
 #
-# Maintainer: Jonathan Lange <jml@twistedmatrix.com>
+# Maintainer: Jonathan Lange
 
 
 import sys
Index: trial/test/test_reporter.py
===================================================================
--- trial/test/test_reporter.py	(revision 24235)
+++ trial/test/test_reporter.py	(working copy)
@@ -1,7 +1,7 @@
 # Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 #
-# Maintainer: Jonathan Lange <jml@twistedmatrix.com>
+# Maintainer: Jonathan Lange
 
 """
 Tests for L{twisted.trial.reporter}.
Index: trial/test/test_runner.py
===================================================================
--- trial/test/test_runner.py	(revision 24235)
+++ trial/test/test_runner.py	(working copy)
@@ -1,8 +1,8 @@
 # Copyright (c) 2005-2007 Twisted Matrix Laboratories.
 # See LICENSE for details.
 #
-# Maintainer: Jonathan Lange <jml@twistedmatrix.com>
-# Author: Robert Collins <robertc@robertcollins.net>
+# Maintainer: Jonathan Lange
+# Author: Robert Collins
 
 
 import StringIO
Index: trial/runner.py
===================================================================
--- trial/runner.py	(revision 24235)
+++ trial/runner.py	(working copy)
@@ -7,7 +7,7 @@
 """
 A miscellany of code used to run Trial tests.
 
-Maintainer: Jonathan Lange <jml@twistedmatrix.com>
+Maintainer: Jonathan Lange
 """
 
 
Index: trial/util.py
===================================================================
--- trial/util.py	(revision 24235)
+++ trial/util.py	(working copy)
@@ -12,7 +12,7 @@
 
 Any non-Trial Twisted code that uses this module will be shot.
 
-Maintainer: U{Jonathan Lange<mailto:jml@twistedmatrix.com>}
+Maintainer: Jonathan Lange
 """
 
 import traceback, sys
Index: trial/__init__.py
===================================================================
--- trial/__init__.py	(revision 24235)
+++ trial/__init__.py	(working copy)
@@ -1,7 +1,7 @@
 # Copyright (c) 2001-2004 Twisted Matrix Laboratories.
 # See LICENSE for details.
 #
-# Maintainer: Jonathan Lange <jml@twistedmatrix.com>
+# Maintainer: Jonathan Lange
 
 """
 Asynchronous unit testing framework.
@@ -9,7 +9,7 @@
 Trial extends Python's builtin C{unittest} to provide support for asynchronous
 tests.
 
-Maintainer: Jonathan Lange <jml@twistedmatrix.com>
+Maintainer: Jonathan Lange
 
 Trial strives to be compatible with other Python xUnit testing frameworks.
 "Compatibility" is a difficult things to define. In practice, it means that:
Index: trial/unittest.py
===================================================================
--- trial/unittest.py	(revision 24235)
+++ trial/unittest.py	(working copy)
@@ -6,7 +6,7 @@
 """
 Things likely to be used by writers of unit tests.
 
-Maintainer: Jonathan Lange <jml@twistedmatrix.com>
+Maintainer: Jonathan Lange
 """
 
 
Index: trial/reporter.py
===================================================================
--- trial/reporter.py	(revision 24235)
+++ trial/reporter.py	(working copy)
@@ -3,7 +3,7 @@
 # Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 #
-# Maintainer: Jonathan Lange <jml@twistedmatrix.com>
+# Maintainer: Jonathan Lange
 
 """
 Defines classes that handle the results of tests.
Index: trial/itrial.py
===================================================================
--- trial/itrial.py	(revision 24235)
+++ trial/itrial.py	(working copy)
@@ -4,7 +4,7 @@
 
 # Copyright (c) 2001-2007 Twisted Matrix Laboratories.
 # See LICENSE for details.
-# Maintainer: Jonathan Lange <jml@twistedmatrix.com>
+# Maintainer: Jonathan Lange
 
 import zope.interface as zi
 from zope.interface import Attribute
Index: spread/refpath.py
===================================================================
--- spread/refpath.py	(revision 24235)
+++ spread/refpath.py	(working copy)
@@ -7,7 +7,7 @@
 """
 Path-based references for PB, and other reference-based protocols.
 
-Maintainer: U{Glyph Lefkowitz<mailto:glyph@twistedmatrix.com>}
+Maintainer: Glyph Lefkowitz
 
 Future Plans: None at this point besides a final overview and finalization
 pass.
Index: spread/flavors.py
===================================================================
--- spread/flavors.py	(revision 24235)
+++ spread/flavors.py	(working copy)
@@ -13,7 +13,7 @@
 Optimizations to this module should not affect external-use semantics at all,
 but may have a small impact on users who subclass and override methods.
 
-@author: U{Glyph Lefkowitz<mailto:glyph@twistedmatrix.com>}
+@author: Glyph Lefkowitz
 """
 
 # NOTE: this module should NOT import pb; it is supposed to be a module which
Index: spread/__init__.py
===================================================================
--- spread/__init__.py	(revision 24235)
+++ spread/__init__.py	(working copy)
@@ -8,5 +8,5 @@
 
 Future Plans: PB, Jelly and Banana need to be optimized.
 
-@author: U{Glyph Lefkowitz<mailto:glyph@twistedmatrix.com>}
+@author: Glyph Lefkowitz
 """
Index: spread/jelly.py
===================================================================
--- spread/jelly.py	(revision 24235)
+++ spread/jelly.py	(working copy)
@@ -59,7 +59,7 @@
 
 The same rule applies for C{frozenset} and C{sets.ImmutableSet}.
 
-@author: U{Glyph Lefkowitz<mailto:glyph@twistedmatrix.com>}
+@author: Glyph Lefkowitz
 """
 
 # System Imports
Index: spread/banana.py
===================================================================
--- spread/banana.py	(revision 24235)
+++ spread/banana.py	(working copy)
@@ -10,7 +10,7 @@
 to it as applies to L{twisted.spread.jelly}, however.  Read its future plans
 for more details.
 
-@author: U{Glyph Lefkowitz<mailto:glyph@twistedmatrix.com>}
+@author: Glyph Lefkowitz
 """
 
 __version__ = "$Revision: 1.37 $"[11:-2]
Index: spread/pb.py
===================================================================
--- spread/pb.py	(revision 24235)
+++ spread/pb.py	(working copy)
@@ -24,7 +24,7 @@
 
 To get started, begin with L{PBClientFactory} and L{PBServerFactory}.
 
-@author: U{Glyph Lefkowitz<mailto:glyph@twistedmatrix.com>}
+@author: Glyph Lefkowitz
 """
 
 import md5
Index: spread/publish.py
===================================================================
--- spread/publish.py	(revision 24235)
+++ spread/publish.py	(working copy)
@@ -5,7 +5,7 @@
 """
 Persistently cached objects for PB.
 
-Maintainer: U{Glyph Lefkowitz<mailto:glyph@twistedmatrix.com>}
+Maintainer: Glyph Lefkowitz
 
 Future Plans: None known.
 """
