Index: protocols/basic.py
===================================================================
--- protocols/basic.py	(revision 24203)
+++ protocols/basic.py	(working copy)
@@ -6,7 +6,7 @@
 """
 Basic protocols, such as line-oriented, netstring, and int prefixed strings.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 # System imports
Index: web2/xmlrpc.py
===================================================================
--- web2/xmlrpc.py	(revision 24203)
+++ web2/xmlrpc.py	(working copy)
@@ -7,7 +7,7 @@
 """
 A generic resource for publishing objects via XML-RPC.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 # System Imports
Index: persisted/dirdbm.py
===================================================================
--- persisted/dirdbm.py	(revision 24203)
+++ persisted/dirdbm.py	(working copy)
@@ -17,7 +17,7 @@
 No files should be placed in the working directory of a DirDBM save those
 created by the DirDBM itself!
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 
Index: persisted/journal/__init__.py
===================================================================
--- persisted/journal/__init__.py	(revision 24203)
+++ persisted/journal/__init__.py	(working copy)
@@ -6,5 +6,5 @@
 """
 Command-journalling persistence framework inspired by Prevayler.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
Index: web/http.py
===================================================================
--- web/http.py	(revision 24203)
+++ web/http.py	(working copy)
@@ -12,7 +12,7 @@
  - Accept chunked data from clients in server.
  - Other missing HTTP features from the RFC.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 # system imports
Index: web/soap.py
===================================================================
--- web/soap.py	(revision 24203)
+++ web/soap.py	(working copy)
@@ -8,7 +8,7 @@
 
 Requires SOAPpy 0.10.1 or later.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 
 Future plans:
 SOAPContext support of some kind.
Index: web/xmlrpc.py
===================================================================
--- web/xmlrpc.py	(revision 24203)
+++ web/xmlrpc.py	(working copy)
@@ -5,7 +5,7 @@
 """
 A generic resource for publishing objects via XML-RPC.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 # System Imports
Index: internet/udp.py
===================================================================
--- internet/udp.py	(revision 24203)
+++ internet/udp.py	(working copy)
@@ -9,7 +9,7 @@
 
 Please do not use this module directly.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 # System Imports
Index: internet/threads.py
===================================================================
--- internet/threads.py	(revision 24203)
+++ internet/threads.py	(working copy)
@@ -6,7 +6,7 @@
 
 For basic support see reactor threading API docs.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 import Queue
Index: internet/wxsupport.py
===================================================================
--- internet/wxsupport.py	(revision 24203)
+++ internet/wxsupport.py	(working copy)
@@ -18,7 +18,7 @@
 On Windows the Twisted event loop might block when dialogs are open
 or menus are selected.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 import warnings
Index: internet/posixbase.py
===================================================================
--- internet/posixbase.py	(revision 24203)
+++ internet/posixbase.py	(working copy)
@@ -7,7 +7,7 @@
 """
 Posix reactor base class
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 import warnings
Index: internet/ssl.py
===================================================================
--- internet/ssl.py	(revision 24203)
+++ internet/ssl.py	(working copy)
@@ -20,7 +20,7 @@
     - support for switching TCP into SSL
     - more options
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 # If something goes wrong, most notably an OpenSSL import failure,
Index: internet/abstract.py
===================================================================
--- internet/abstract.py	(revision 24203)
+++ internet/abstract.py	(working copy)
@@ -6,7 +6,7 @@
 """
 Support for generic select()able objects.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 from zope.interface import implements
Index: internet/tksupport.py
===================================================================
--- internet/tksupport.py	(revision 24203)
+++ internet/tksupport.py	(working copy)
@@ -5,7 +5,7 @@
 """
 This module integrates Tkinter with twisted.internet's mainloop.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 
 To use, do::
 
Index: internet/interfaces.py
===================================================================
--- internet/interfaces.py	(revision 24203)
+++ internet/interfaces.py	(working copy)
@@ -5,7 +5,7 @@
 """
 Interface documentation.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 from zope.interface import Interface, Attribute
Index: internet/unix.py
===================================================================
--- internet/unix.py	(revision 24203)
+++ internet/unix.py	(working copy)
@@ -8,7 +8,7 @@
 
 End users shouldn't use this module directly - use the reactor APIs instead.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 # System imports
Index: internet/main.py
===================================================================
--- internet/main.py	(revision 24203)
+++ internet/main.py	(working copy)
@@ -8,7 +8,7 @@
 In general, this module should not be used, other than by reactor authors
 who need to use the 'installReactor' method.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 import error
Index: internet/default.py
===================================================================
--- internet/default.py	(revision 24203)
+++ internet/default.py	(working copy)
@@ -8,7 +8,7 @@
 """
 Deprecated module that used to contain SelectReactor and PosixReactorBase
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 import warnings
Index: internet/win32eventreactor.py
===================================================================
--- internet/win32eventreactor.py	(revision 24203)
+++ internet/win32eventreactor.py	(working copy)
@@ -7,7 +7,7 @@
 
 This requires win32all or ActivePython to be installed.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 
 
 LIMITATIONS:
Index: internet/pollreactor.py
===================================================================
--- internet/pollreactor.py	(revision 24203)
+++ internet/pollreactor.py	(working copy)
@@ -11,7 +11,7 @@
     from twisted.internet import pollreactor
     pollreactor.install()
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 # System imports
Index: internet/tcp.py
===================================================================
--- internet/tcp.py	(revision 24203)
+++ internet/tcp.py	(working copy)
@@ -7,7 +7,7 @@
 
 End users shouldn't use this module directly - use the reactor APIs instead.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 
Index: internet/gtk2reactor.py
===================================================================
--- internet/gtk2reactor.py	(revision 24203)
+++ internet/gtk2reactor.py	(working copy)
@@ -19,7 +19,7 @@
 event loop or the GTK+ event loop which is based on it but adds GUI
 integration.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 # System Imports
Index: internet/error.py
===================================================================
--- internet/error.py	(revision 24203)
+++ internet/error.py	(working copy)
@@ -5,7 +5,7 @@
 """
 Exceptions and errors for use in twisted.internet modules.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 import socket
Index: internet/kqreactor.py
===================================================================
--- internet/kqreactor.py	(revision 24203)
+++ internet/kqreactor.py	(working copy)
@@ -14,7 +14,7 @@
 This reactor only works on FreeBSD and requires PyKQueue 1.3, which is
 available at:  U{http://people.freebsd.org/~dwhite/PyKQueue/}
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 
 
 
Index: internet/selectreactor.py
===================================================================
--- internet/selectreactor.py	(revision 24203)
+++ internet/selectreactor.py	(working copy)
@@ -6,7 +6,7 @@
 """
 Select reactor
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 from time import sleep
Index: internet/protocol.py
===================================================================
--- internet/protocol.py	(revision 24203)
+++ internet/protocol.py	(working copy)
@@ -10,7 +10,7 @@
 Start here if you are looking to write a new protocol implementation for
 Twisted.  The Protocol class contains some introductory material.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 import random
Index: internet/gtkreactor.py
===================================================================
--- internet/gtkreactor.py	(revision 24203)
+++ internet/gtkreactor.py	(working copy)
@@ -12,7 +12,7 @@
 Then use twisted.internet APIs as usual.  The other methods here are not
 intended to be called directly.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 import sys
Index: internet/base.py
===================================================================
--- internet/base.py	(revision 24203)
+++ internet/base.py	(working copy)
@@ -6,7 +6,7 @@
 """
 Very basic functionality for a Reactor implementation.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 import socket # needed only for sync-dns
Index: internet/process.py
===================================================================
--- internet/process.py	(revision 24203)
+++ internet/process.py	(working copy)
@@ -7,7 +7,7 @@
 
 Do NOT use this module directly - use reactor.spawnProcess() instead.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 # System Imports
Index: internet/wxreactor.py
===================================================================
--- internet/wxreactor.py	(revision 24203)
+++ internet/wxreactor.py	(working copy)
@@ -22,7 +22,7 @@
 expected and probably does not reflect on the reactor's ability to run
 real applications.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 import Queue
Index: internet/glib2reactor.py
===================================================================
--- internet/glib2reactor.py	(revision 24203)
+++ internet/glib2reactor.py	(working copy)
@@ -17,7 +17,7 @@
 event loop or the GTK+ event loop which is based on it but adds GUI
 integration.
 
-Maintainer: U{Itamar Shtull-Trauring<mailto:twisted@itamarst.org>}
+Maintainer: Itamar Shtull-Trauring
 """
 
 from twisted.internet import gtk2reactor
