Index: using-twistedweb.xhtml
===================================================================
--- using-twistedweb.xhtml	(revision 29003)
+++ using-twistedweb.xhtml	(working copy)
@@ -745,8 +745,8 @@
 <p>Sometimes, you want to be able to send headers and status directly. While
 you can do this with a
 <code base="twisted.web.script" class="API">ResourceScript</code>, an easier
-way is to use <code base="twisted.web.static" class="API">AsIsProcessor</code>.
-Use it by, for example, addding it as a processor for the <code>.asis</code>
+way is to use <code base="twisted.web.static" class="API">ASISProcessor</code>.
+Use it by, for example, adding it as a processor for the <code>.asis</code>
 extension. Here is a sample file:</p>
 
 <pre>
