diff --git twisted/names/root.py twisted/names/root.py
index 61aab5b..86c136c 100644
--- twisted/names/root.py
+++ twisted/names/root.py
@@ -13,8 +13,6 @@ todo::
     documentation
 """
 
-import warnings
-
 from twisted.python.failure import Failure
 from twisted.internet import defer
 from twisted.names import dns, common, error
diff --git twisted/names/test/test_rootresolve.py twisted/names/test/test_rootresolve.py
index bb8959a..cffb46c 100644
--- twisted/names/test/test_rootresolve.py
+++ twisted/names/test/test_rootresolve.py
@@ -187,24 +187,26 @@ class RootResolverTests(TestCase):
         # And a DNS packet sent.
         [(packet, address)] = transport._sentPackets
 
-        msg = Message()
-        msg.fromStr(packet)
+        message = Message()
+        message.fromStr(packet)
 
         # It should be a query with the parameters used above.
-        self.assertEqual(msg.queries, [Query(b'foo.example.com', A, IN)])
-        self.assertEqual(msg.answers, [])
-        self.assertEqual(msg.authority, [])
-        self.assertEqual(msg.additional, [])
+        self.assertEqual(message.queries, [Query(b'foo.example.com', A, IN)])
+        self.assertEqual(message.answers, [])
+        self.assertEqual(message.authority, [])
+        self.assertEqual(message.additional, [])
 
         response = []
         d.addCallback(response.append)
         self.assertEqual(response, [])
 
         # Once a reply is received, the Deferred should fire.
-        del msg.queries[:]
-        msg.answer = 1
-        msg.answers.append(RRHeader(b'foo.example.com', payload=Record_A('5.8.13.21')))
-        transport._protocol.datagramReceived(msg.toStr(), ('1.1.2.3', 1053))
+        del message.queries[:]
+        message.answer = 1
+        message.answers.append(RRHeader(
+            b'foo.example.com', payload=Record_A('5.8.13.21')))
+        transport._protocol.datagramReceived(
+            message.toStr(), ('1.1.2.3', 1053))
         return response[0]
 
 
