diff --git a/twisted/manhole/explorer.py b/twisted/manhole/explorer.py
index b52693c..23bfd68 100644
--- a/twisted/manhole/explorer.py
+++ b/twisted/manhole/explorer.py
@@ -16,27 +16,23 @@ L{inspect} module available in Python 2.1.
 
 # System Imports
 import inspect, string, sys, types
-import UserDict
 
 # Twisted Imports
 from twisted.spread import pb
 from twisted.python import reflect
 
 
-True=(1==1)
-False=not True
-
-class Pool(UserDict.UserDict):
+class Pool(dict):
     def getExplorer(self, object, identifier):
         oid = id(object)
-        if oid in self.data:
+        if oid in self:
             # XXX: This potentially returns something with
             # 'identifier' set to a different value.
-            return self.data[oid]
+            return self[oid]
         else:
             klass = typeTable.get(type(object), ExplorerGeneric)
             e = types.InstanceType(klass, {})
-            self.data[oid] = e
+            self[oid] = e
             klass.__init__(e, object, identifier)
             return e
 
diff --git a/twisted/protocols/postfix.py b/twisted/protocols/postfix.py
index 7a2079d..43797bf 100644
--- a/twisted/protocols/postfix.py
+++ b/twisted/protocols/postfix.py
@@ -7,7 +7,6 @@ Postfix mail transport agent related protocols.
 """
 
 import sys
-import UserDict
 import urllib
 
 from twisted.protocols import basic
@@ -92,8 +91,7 @@ class PostfixTCPMapServer(basic.LineReceiver, policies.TimeoutMixin):
                 self.sendCode(500, 'put is not implemented yet.')
 
 
-class PostfixTCPMapDictServerFactory(protocol.ServerFactory,
-                                     UserDict.UserDict):
+class PostfixTCPMapDictServerFactory(protocol.ServerFactory, dict):
     """An in-memory dictionary factory for PostfixTCPMapServer."""
 
     protocol = PostfixTCPMapServer
diff --git a/twisted/topfiles/5789.misc b/twisted/topfiles/5789.misc
new file mode 100644
index 0000000..e69de29
