Index: twisted/topfiles/6485.doc
===================================================================
--- twisted/topfiles/6485.doc	(revision 0)
+++ twisted/topfiles/6485.doc	(revision 0)
@@ -0,0 +1,5 @@
+The following example files now use "y in x" sytanx instead of the deprecated "x.has_key(y)" syntax:
+./doc/core/howto/tutorial/listings/finger/finger_config.py
+./doc/core/howto/tutorial/listings/finger/finger/finger.py
+./doc/core/howto/listings/servers/chat.py
+./doc/core/howto/listings/pb/chatserver.py
Index: doc/core/howto/listings/pb/chatserver.py
===================================================================
--- doc/core/howto/listings/pb/chatserver.py	(revision 38313)
+++ doc/core/howto/listings/pb/chatserver.py	(working copy)
@@ -14,7 +14,7 @@
         self.groups = {} # indexed by name
 
     def joinGroup(self, groupname, user, allowMattress):
-        if not self.groups.has_key(groupname):
+        if groupname not in self.groups:
             self.groups[groupname] = Group(groupname, allowMattress)
         self.groups[groupname].addUser(user)
         return self.groups[groupname]
Index: doc/core/howto/listings/servers/chat.py
===================================================================
--- doc/core/howto/listings/servers/chat.py	(revision 38313)
+++ doc/core/howto/listings/servers/chat.py	(working copy)
@@ -13,7 +13,7 @@
         self.sendLine("What's your name?")
 
     def connectionLost(self, reason):
-        if self.users.has_key(self.name):
+        if self.name in self.users:
             del self.users[self.name]
 
     def lineReceived(self, line):
@@ -23,7 +23,7 @@
             self.handle_CHAT(line)
 
     def handle_GETNAME(self, name):
-        if self.users.has_key(name):
+        if name in self.users:
             self.sendLine("Name taken, please choose another.")
             return
         self.sendLine("Welcome, %s!" % (name,))
Index: doc/core/howto/tutorial/listings/finger/finger/finger.py
===================================================================
--- doc/core/howto/tutorial/listings/finger/finger/finger.py	(revision 38313)
+++ doc/core/howto/tutorial/listings/finger/finger/finger.py	(working copy)
@@ -351,7 +351,7 @@
 #        k.setServiceParent(s)
 
     # irc fingerbot
-    if config.has_key('ircnick'):
+    if 'ircnick' in config:
         i = IIRCClientFactory(f)
         i.nickname = config['ircnick']
         ircserver = config['ircserver']
@@ -359,7 +359,7 @@
         b.setServiceParent(s)
 
     # Pespective Broker on port 8889
-    if config.has_key('pbport'):
+    if 'pbport' in config:
         m = internet.TCPServer(
             int(config['pbport']),
             pb.PBServerFactory(IPerspectiveFinger(f)))
Index: doc/core/howto/tutorial/listings/finger/finger_config.py
===================================================================
--- doc/core/howto/tutorial/listings/finger/finger_config.py	(revision 38313)
+++ doc/core/howto/tutorial/listings/finger/finger_config.py	(working copy)
@@ -21,7 +21,7 @@
         k.setServiceParent(s)
 
     # irc fingerbot
-    if config.has_key('ircnick'):
+    if 'ircnick' in config:
         i = IIRCClientFactory(f)
         i.nickname = config['ircnick']
         ircserver = config['ircserver']
@@ -29,7 +29,7 @@
         b.setServiceParent(s)
 
     # Pespective Broker on port 8889
-    if config.has_key('pbport'):
+    if 'pbport' in config:
         m = internet.TCPServer(
             int(config['pbport']),
             pb.PBServerFactory(IPerspectiveFinger(f)))
