Index: manhole/ui/gtkmanhole.py
===================================================================
--- manhole/ui/gtkmanhole.py	(revision 24203)
+++ manhole/ui/gtkmanhole.py	(working copy)
@@ -70,8 +70,6 @@
     def remote_receiveExplorer(self, xplorer):
         if _GNOME_POWER:
             self.display.receiveExplorer(xplorer)
-        else:
-            XXX # text display?
 
     def remote_listCapabilities(self):
         return self.capabilities
