diff --git twisted/manhole/_inspectro.py twisted/manhole/_inspectro.py
index e397393..430ae7b 100644
--- twisted/manhole/_inspectro.py
+++ twisted/manhole/_inspectro.py
@@ -16,7 +16,7 @@ from twisted.manhole.ui import gtk2manhole
 from twisted.python.components import Adapter, registerAdapter
 from twisted.python import log
 from twisted.protocols import policies
-from zope.interface import implementer, Interface
+from zope.interface import implements, Interface
 
 # the glade file uses stock icons, which requires gnome to be installed
 import gnome
@@ -51,8 +51,8 @@ class INode(Interface):
             return InstanceNode(adaptable)
         return AttributesNode(adaptable)
 
-@implementer(INode)
 class InspectorNode(Adapter):
+    implements(INode)
 
     def postInit(self, offset, parent, slot):
         self.offset = offset
diff --git twisted/manhole/service.py twisted/manhole/service.py
index 74ef7cc..c9d4679 100644
--- twisted/manhole/service.py
+++ twisted/manhole/service.py
@@ -12,7 +12,7 @@ from twisted.spread import pb
 from twisted.python import log, failure
 from twisted.cred import portal
 from twisted.application import service
-from zope.interface import implementer, Interface
+from zope.interface import implements, Interface
 
 # sibling imports
 import explorer
@@ -349,9 +349,10 @@ class Perspective(pb.Avatar):
                               self.receiveExplorer)
 
 
-@implementer(portal.IRealm)
 class Realm:
 
+    implements(portal.IRealm)
+
     def __init__(self, service):
         self.service = service
         self._cache = {}
diff --git twisted/manhole/ui/gtk2manhole.py twisted/manhole/ui/gtk2manhole.py
index fcff7c1..2c6a532 100644
--- twisted/manhole/ui/gtk2manhole.py
+++ twisted/manhole/ui/gtk2manhole.py
@@ -16,7 +16,7 @@ from twisted.spread import pb
 from twisted.spread.ui import gtk2util
 
 from twisted.manhole.service import IManholeClient
-from zope.interface import implementer
+from zope.interface import implements
 
 # The pygtk.require for version 2.0 has already been done by the reactor.
 import gtk
@@ -335,8 +335,8 @@ class _Notafile:
     def flush(self):
         pass
 
-@implementer(IManholeClient)
 class ManholeClient(components.Adapter, pb.Referenceable):
+    implements(IManholeClient)
 
     capabilities = {
 #        "Explorer": 'Set',
