--- client.py.orig	2010-03-25 09:49:19.000000000 +0100
+++ client.py.new	2010-03-25 10:05:20.000000000 +0100
@@ -108,9 +108,6 @@
             return
         url = l[0]
         if self.followRedirect:
-            scheme, host, port, path = \
-                _parse(url, defaultPort=self.transport.getPeer().port)
-
             self.factory._redirectCount += 1
             if self.factory._redirectCount >= self.factory.redirectLimit:
                 err = error.InfiniteRedirection(
@@ -122,6 +119,10 @@
                 self.transport.loseConnection()
                 return
 
+            # If the given url is relative (no scheme),
+            # we add the current path at the beginning.
+            if _parse(url)[0] == "":
+            	url = self.factory.path.rsplit("/", 1)[0] + "/" + url
             self.factory.setURL(url)
 
             if self.factory.scheme == 'https':
