--- imap4.py	Mon Nov  8 14:38:51 2004
+++ imap.twisted	Mon Nov  8 13:13:28 2004
@@ -1934,7 +1934,7 @@
 class IllegalServerResponse(IMAP4Exception): pass
 
 
-class IMAP4Client(basic.LineReceiver):
+class IMAP4Client(basic.LineReceiver, policies.TimeoutMixin):
     """IMAP4 client protocol implementation
 
     @ivar state: A string representing the state the connection is currently
@@ -1950,6 +1950,10 @@
 
     startedTLS = False
 
+    #Number of seconds to wait before timing out a connection.
+    #If the number is <= 0 no timeout checking will be performed.
+    timeout = 0
+
     # Capabilities are not allowed to change during the session
     # So cache the first response and use that for all later
     # lookups
@@ -1992,6 +1996,9 @@
         self.authenticators[auth.getName().upper()] = auth
 
     def rawDataReceived(self, data):
+        if self.timeout > 0:
+            self.setTimeout(self.timeout)
+
         self._pendingSize -= len(data)
         if self._pendingSize > 0:
             self._pendingBuffer.write(data)
@@ -2017,8 +2024,20 @@
         self._parts = [rest, '\r\n']
         self.setRawMode()
 
+    def connectionMade(self):
+        if self.timeout > 0:
+            self.setTimeout(self.timeout)
+
+    def connectionLost(self, reason):
+        """We are no longer connected"""
+        if self.timeout > 0:
+            self.setTimeout(None)
+
     def lineReceived(self, line):
 #        print 'C: ' + repr(line)
+        if self.timeout > 0:
+            self.resetTimeout()
+
         if self._parts is None:
             lastPart = line.rfind(' ')
             if lastPart != -1:
@@ -2045,6 +2064,15 @@
             tag, rest = self._tag, ''.join(self._parts)
             self._tag = self._parts = None
             self.dispatchCommand(tag, rest)
+
+    def timeoutConnection(self):
+        if self.queued:
+            e = error.TimeoutError("Communication with IMAP Server timed out. Please try again later.")
+
+            for cmd in self.queued:
+                cmd.defer.errback(e)
+
+        self.transport.loseConnection()
 
     def _regularDispatch(self, line):
         parts = line.split(None, 1)
