diff --git twisted/mail/alias.py twisted/mail/alias.py
index 1741b1f..8bd1f0c 100644
--- twisted/mail/alias.py
+++ twisted/mail/alias.py
@@ -327,7 +327,7 @@ class ProcessAlias(AliasBase):
 
     def __init__(self, path, *args):
         AliasBase.__init__(self, *args)
-        self.path = path.split()
+        self.path = [path]
         self.program = self.path[0]
 
 
diff --git twisted/mail/topfiles/6426.bugfix twisted/mail/topfiles/6426.bugfix
new file mode 100644
index 0000000..987fe10
--- /dev/null
+++ twisted/mail/topfiles/6426.bugfix
@@ -0,0 +1 @@
+twisted.mail.alias.ProcessAlias now correctly handles paths
diff --git twisted/mail/topfiles/6426.misc twisted/mail/topfiles/6426.misc
new file mode 100644
index 0000000..e69de29
