--- ftp.py.orig	2010-02-04 21:21:53.000000000 +0100
+++ ftp.py	2010-02-04 21:22:33.000000000 +0100
@@ -868,7 +868,7 @@
 
         self.dtpFactory = DTPFactory(pi=self, peerHost=self.transport.getPeer().host)
         self.dtpFactory.setTimeout(self.dtpTimeout)
-        self.dtpPort = reactor.connectTCP(ip, port, self.dtpFactory)
+        self.dtpPort = reactor.connectTCP(ip, port, self.dtpFactory,bindAddress=(self.transport.getHost().host,0))
 
         def connected(ignored):
             return ENTERING_PORT_MODE
