Index: static.py
===================================================================
RCS file: /cvs/Twisted/twisted/web/static.py,v
retrieving revision 1.106
diff -u -r1.106 static.py
--- static.py	28 Jul 2003 00:01:13 -0000	1.106
+++ static.py	28 Jul 2003 06:26:32 -0000
@@ -450,6 +450,8 @@
         # dirs allows usage of the File to specify what gets listed
         self.dirs = dirs
         self.path = pathname
+        self.directoryButton = "<b>D</b> "
+        self.fileButton = "<b>F</B> "
 
     def getTitle(self, request):
         return "Directory Listing For %s" % request.path
@@ -465,7 +467,13 @@
             url = urllib.quote(path, "/:")
             if os.path.isdir(os.path.join(self.path, path)):
                 url = url + '/'
-            write('<li><a href="%s">%s</a></li>' % (url, path))
+                write('<li>%s<a href="%s">%s</a></li>'
+                      % (self.directoryButton, url, path))
+        for path in directory:
+            url = urllib.quote(path, "/:")
+            if not os.path.isdir(os.path.join(self.path, path)):
+                write('<li>%s<a href="%s">%s</a></li>'
+                      % (self.fileButton, url, path))
         write("</ul>\n")
 
     def __repr__(self):
