Index: twisted/words/protocols/oscar.py
===================================================================
--- twisted/words/protocols/oscar.py	(revision 39080)
+++ twisted/words/protocols/oscar.py	(working copy)
@@ -15,6 +15,8 @@
 import random
 import types
 import re
+import functools
+
 from hashlib import md5
 
 from twisted.internet import reactor, defer, protocol
@@ -199,7 +201,7 @@
         user.group = self
 
     def oscarRep(self, groupID, buddyID):
-        tlvData = TLV(0xc8, reduce(lambda x,y:x+y, [struct.pack('!H',self.usersToID[x]) for x in self.users]))
+        tlvData = TLV(0xc8, functools.reduce(lambda x,y:x+y, [struct.pack('!H',self.usersToID[x]) for x in self.users]))
         return struct.pack('!H', len(self.name)) + self.name + \
                struct.pack('!HH', groupID, buddyID) + '\000\001' + tlvData
 
@@ -230,7 +232,7 @@
                 self.alertSound = v
  
     def oscarRep(self, groupID, buddyID):
-        tlvData = reduce(lambda x,y: x+y, map(lambda (k,v):TLV(k,v), self.tlvs.items()), '\000\000')
+        tlvData = functools.reduce(lambda x,y: x+y, map(lambda (k,v):TLV(k,v), self.tlvs.items()), '\000\000')
         return struct.pack('!H', len(self.name)) + self.name + \
                struct.pack('!HH', groupID, buddyID) + '\000\000' + tlvData
 
Index: twisted/words/protocols/irc.py
===================================================================
--- twisted/words/protocols/irc.py	(revision 39080)
+++ twisted/words/protocols/irc.py	(working copy)
@@ -39,6 +39,7 @@
 import string, socket
 import textwrap
 import shlex
+import functools
 from os import path
 
 from twisted.internet import reactor, protocol, task
@@ -3042,11 +3043,11 @@
 
     @param z: Initial value.
 
-    @param xs: Sequence to reduce.
+    @param xs: Sequence to functools.reduce.
 
     @return: Single value resulting from reducing C{xs}.
     """
-    return reduce(lambda x, y: f(y, x), reversed(xs), z)
+    return functools.reduce(lambda x, y: f(y, x), reversed(xs), z)
 
 
 
Index: twisted/words/protocols/msn.py
===================================================================
--- twisted/words/protocols/msn.py	(revision 39080)
+++ twisted/words/protocols/msn.py	(working copy)
@@ -71,7 +71,7 @@
 @author: Sam Jordan
 """
 
-import types, operator, os
+import types, operator, os, functools
 from random import randint
 from urllib import quote, unquote
 from hashlib import md5
@@ -371,7 +371,7 @@
         used to calculte the number to send
         as the message length when sending a message.
         """
-        return reduce(operator.add, [len(x[0]) + len(x[1]) + 4  for x in self.headers.items()]) + len(self.message) + 2
+        return functools.reduce(operator.add, [len(x[0]) + len(x[1]) + 4  for x in self.headers.items()]) + len(self.message) + 2
 
     def setHeader(self, header, value):
         """ set the desired header """
