diff --git a/twisted/lore/latex.py b/twisted/lore/latex.py
index 61189e1..ed843ed 100644
--- a/twisted/lore/latex.py
+++ b/twisted/lore/latex.py
@@ -6,7 +6,7 @@ LaTeX output support for Lore.
 """
 
 from xml.dom import minidom as dom
-import os.path, re, string
+import os.path, re
 from cStringIO import StringIO
 import urlparse
 
@@ -207,7 +207,7 @@ class LatexSpitter(BaseLatexSpitter):
     def visitNode_a_listing(self, node):
         fileName = os.path.join(self.currDir, node.getAttribute('href'))
         self.writer('\\begin{verbatim}\n')
-        lines = map(string.rstrip, open(fileName).readlines())
+        lines = map(str.rstrip, open(fileName).readlines())
         skipLines = int(node.getAttribute('skipLines') or 0)
         lines = lines[skipLines:]
         self.writer(text.removeLeadingTrailingBlanks('\n'.join(lines)))
diff --git a/twisted/lore/tree.py b/twisted/lore/tree.py
index 1b6dbe1..37ffa3f 100755
--- a/twisted/lore/tree.py
+++ b/twisted/lore/tree.py
@@ -3,7 +3,7 @@
 
 
 from itertools import count
-import re, os, cStringIO, time, cgi, string, urlparse
+import re, os, cStringIO, time, cgi, urlparse
 from xml.dom import minidom as dom
 from xml.sax.handler import ErrorHandler, feature_validation
 from xml.dom.pulldom import SAX2DOM
@@ -199,7 +199,7 @@ def addPyListings(document, dir):
                                                      "py-listing"):
         filename = node.getAttribute("href")
         outfile = cStringIO.StringIO()
-        lines = map(string.rstrip, open(os.path.join(dir, filename)).readlines())
+        lines = map(str.rstrip, open(os.path.join(dir, filename)).readlines())
 
         skip = node.getAttribute('skipLines') or 0
         lines = lines[int(skip):]
