Index: doc/core/howto/index.xhtml
===================================================================
--- doc/core/howto/index.xhtml	(revision 30464)
+++ doc/core/howto/index.xhtml	(working copy)
@@ -32,6 +32,8 @@
 
         <li><a href="clients.xhtml">Writing a TCP client</a></li>
 
+        <li><a href="endpoints.xhtml">Getting Connected with Endpoints</a></li>
+
         <li><a href="quotes.xhtml">Setting up the TwistedQuotes
         application</a></li>
 
