Index: twisted/conch/test/test_ssh.py
===================================================================
--- twisted/conch/test/test_ssh.py	(revision 23031)
+++ twisted/conch/test/test_ssh.py	(working copy)
@@ -690,9 +690,11 @@
     if not Crypto:
         skip = "can't run w/o PyCrypto"
 
-    def testOurServerOurClient(self):
-        """test the Conch server against the Conch client
+    def setUp(self):
         """
+        Create Conch client and server protocols and loopback transports for
+        them.
+        """
         realm = ConchTestRealm()
         p = portal.Portal(realm)
         sshpc = ConchTestSSHChecker()
@@ -707,6 +709,11 @@
         self.client = ConchTestClient()
         self.serverTransport = LoopbackRelay(self.client)
 
+
+    def _runServerClient(self):
+        """
+        Connect and run server and client.
+        """
         self.server.makeConnection(self.serverTransport)
         self.client.makeConnection(self.clientTransport)
 
@@ -718,6 +725,14 @@
         self.failIf(self.server.done and self.client.done)
 
 
+    def test_ourServerOurClient(self):
+        """
+        Test the Conch server against the Conch client.
+        """
+        self._runServerClient()
+
+
+
 class TestSSHFactory(unittest.TestCase):
 
     if not Crypto:
