Index: twisted/conch/test/test_ssh.py
===================================================================
--- twisted/conch/test/test_ssh.py	(revision 22951)
+++ twisted/conch/test/test_ssh.py	(working copy)
@@ -690,9 +690,7 @@
     if not Crypto:
         skip = "can't run w/o PyCrypto"
 
-    def testOurServerOurClient(self):
-        """test the Conch server against the Conch client
-        """
+    def setUp(self):
         realm = ConchTestRealm()
         p = portal.Portal(realm)
         sshpc = ConchTestSSHChecker()
@@ -707,6 +705,10 @@
         self.client = ConchTestClient()
         self.serverTransport = LoopbackRelay(self.client)
 
+    def _runServerClient(self):
+        """
+        Connect and run server and client.
+        """
         self.server.makeConnection(self.serverTransport)
         self.client.makeConnection(self.clientTransport)
 
@@ -717,7 +719,14 @@
                                 self.clientTransport.clearBuffer)
         self.failIf(self.server.done and self.client.done)
 
+    def testOurServerOurClient(self):
+        """
+        Test the Conch server against the Conch client.
+        """
+        self._runServerClient()
+    
 
+
 class TestSSHFactory(unittest.TestCase):
 
     if not Crypto:
