diff --git a/twisted/internet/error.py b/twisted/internet/error.py
--- a/twisted/internet/error.py
+++ b/twisted/internet/error.py
@@ -97,8 +97,8 @@ class ConnectError(Exception):
         s = self.__doc__ or self.__class__.__name__
         if self.osError:
             s = '%s: %s' % (s, self.osError)
-        if self[0]:
-            s = '%s: %s' % (s, self[0])
+        if self.args[0]:
+            s = '%s: %s' % (s, self.args[0])
         s = '%s.' % s
         return s
 
diff --git a/twisted/internet/pollreactor.py b/twisted/internet/pollreactor.py
--- a/twisted/internet/pollreactor.py
+++ b/twisted/internet/pollreactor.py
@@ -152,7 +152,7 @@ class PollReactor(posixbase.PosixReactor
         try:
             l = self._poller.poll(timeout)
         except SelectError, e:
-            if e[0] == errno.EINTR:
+            if e.args[0] == errno.EINTR:
                 return
             else:
                 raise
diff --git a/twisted/internet/tcp.py b/twisted/internet/tcp.py
--- a/twisted/internet/tcp.py
+++ b/twisted/internet/tcp.py
@@ -403,13 +403,13 @@ class Client(BaseClient):
         try:
             skt = self.createInternetSocket()
         except socket.error, se:
-            err = error.ConnectBindError(se[0], se[1])
+            err = error.ConnectBindError(se.args[0], se.args[1])
             whenDone = None
         if whenDone and bindAddress is not None:
             try:
                 skt.bind(bindAddress)
             except socket.error, se:
-                err = error.ConnectBindError(se[0], se[1])
+                err = error.ConnectBindError(se.args[0], se.args[1])
                 whenDone = None
         self._finishInit(whenDone, skt, err, reactor)
 
