Index: twisted/web2/test/test_log.py
===================================================================
--- twisted/web2/test/test_log.py	(revision 16041)
+++ twisted/web2/test/test_log.py	(working copy)
@@ -107,6 +107,12 @@
 
         return d
 
+    def testLogResponse(self):
+        class Response: 
+            stream = None
+        response = Response()
+        log.logFilter(None, response)
+        assert response.stream == None, "Should not wrap a None response.stream"
         
 
         
Index: twisted/web2/log.py
===================================================================
--- twisted/web2/log.py	(revision 16041)
+++ twisted/web2/log.py	(working copy)
@@ -58,6 +58,9 @@
     startTime=None
     
 def logFilter(request, response, startTime=None):
+    if not response.stream:
+        return response
+
     if startTime is None:
         startTime = time.time()
         
