Index: twisted/mail/topfiles/README
===================================================================
--- twisted/mail/topfiles/README	(revision 27776)
+++ twisted/mail/topfiles/README	(working copy)
@@ -1,5 +1,3 @@
 Twisted Mail 9.0.0
 
-Mail was recently split out of Twisted.
-
-Twisted Mail depends on Twisted and (sometimes) Twisted Names.
+Twisted Mail depends on Twisted and (sometimes) Twisted Names.
\ No newline at end of file
Index: twisted/pair/topfiles/README
===================================================================
--- twisted/pair/topfiles/README	(revision 27776)
+++ twisted/pair/topfiles/README	(working copy)
@@ -1,4 +1 @@
-Twisted Pair 9.0.0
-
-Twisted Pair was recently split out of Twisted.
-
+Twisted Pair 9.0.0
\ No newline at end of file
