--- ftp.py	2005-08-04 15:55:49.067884532 +0200
+++ ftp_ok.py	2005-08-04 10:57:53.000000000 +0200
@@ -1802,10 +1802,9 @@
             return None
 
     def queueLogin(self, username, password):
-        """Login: send the username, send the password. If the password in None, don't send it"""
+        """Login: send the username, send the password."""
         for command in ('USER ' + username,
                                       'PASS ' + password):
-            if password == None and command.startswith('PASS'): continue
             d = self.queueStringCommand(command, public=0)
             # If something goes wrong, call fail
             d.addErrback(self.fail)
