--- a/twisted/trial/runner.py
+++ b/twisted/trial/runner.py
@@ -579,10 +579,10 @@ class TestLoader(object):
             discovery = pkgobj.walkModules()
         else:
             discovery = pkgobj.iterModules()
-        discovered = []
+        discovered = set()
         for disco in discovery:
             if disco.name.split(".")[-1].startswith(self.modulePrefix):
-                discovered.append(disco)
+                discovered.add(disco)
         suite = self.suiteFactory()
         for modinfo in self.sort(discovered):
             try:
--- a/twisted/python/modules.py
+++ b/twisted/python/modules.py
@@ -402,6 +402,12 @@ class PythonModule(_ModuleIteratorHelper):
             return True
         return other.name != self.name

+    def __hash__(self):
+        """
+        PythonModules are hashed by their name.
+        """
+        return hash(self.name)
+
     def walkModules(self, importPackages=False):
         if importPackages and self.isPackage():
             self.load()
