Index: twisted/web/resource.py
===================================================================
--- twisted/web/resource.py	(revision 43942)
+++ twisted/web/resource.py	(working copy)
@@ -293,12 +293,14 @@
             L{ErrorPage.__init__}.
 
     @ivar code: An integer status code which will be used for the response.
+    @type code: C{int}
 
-    @ivar brief: A short string which will be included in the response body.
+    @ivar brief: A short string which will be included in the response body as
+        the page title.
     @type brief: C{str}
 
     @ivar detail: A longer string which will be included in the response body.
-    @ivar detail: C{str}
+    @type detail: C{str}
     """
 
     template = """
