Index: twisted/conch/ssh/agent.py
===================================================================
--- twisted/conch/ssh/agent.py	(revision 46578)
+++ twisted/conch/ssh/agent.py	(working copy)
@@ -157,10 +157,10 @@
             packet, self.buf = self.buf[4:4 + packLen], self.buf[4 + packLen:]
             reqType = ord(packet[0])
             reqName = messages.get(reqType, None)
-            if not reqName:
+            f = getattr(self, 'agentc_%s' % reqName, None)
+            if reqName is None or f is None:
                 self.sendResponse(AGENT_FAILURE, '')
             else:
-                f = getattr(self, 'agentc_%s' % reqName)
                 if getattr(self.factory, 'keys', None) is None:
                     self.sendResponse(AGENT_FAILURE, '')
                     raise MissingKeyStoreError()
@@ -273,8 +273,11 @@
 
 AGENTC_REQUEST_RSA_IDENTITIES   = 1
 AGENT_RSA_IDENTITIES_ANSWER     = 2
+AGENTC_RSA_CHALLENGE            = 3
+AGENT_RSA_RESPONSE              = 4
 AGENT_FAILURE                   = 5
 AGENT_SUCCESS                   = 6
+AGENTC_ADD_RSA_IDENTITY         = 7
 
 AGENTC_REMOVE_RSA_IDENTITY         = 8
 AGENTC_REMOVE_ALL_RSA_IDENTITIES   = 9
@@ -287,6 +290,14 @@
 AGENTC_REMOVE_IDENTITY          = 18
 AGENTC_REMOVE_ALL_IDENTITIES    = 19
 
+AGENTC_ADD_SMARTCARD_KEY             = 20
+AGENTC_REMOVE_SMARTCARD_KEY          = 21
+AGENTC_LOCK                          = 22
+AGENTC_UNLOCK                        = 23
+AGENTC_ADD_RSA_ID_CONSTRAINED        = 24
+AGENTC_ADD_ID_CONSTRAINED            = 25
+AGENTC_ADD_SMARTCARD_KEY_CONSTRAINED = 26
+
 messages = {}
 for name, value in locals().copy().items():
     if name[:7] == 'AGENTC_':
