Index: doc/core/development/policy/test-standard.xhtml
===================================================================
--- doc/core/development/policy/test-standard.xhtml	(revision 27607)
+++ doc/core/development/policy/test-standard.xhtml	(working copy)
@@ -346,7 +346,7 @@
 
     </ul>
 
-  <p>See also <a href="../testing.xhtml">Tips for writing tests for Twisted
+  <p>See also <a href="../../howto/testing.xhtml">Tips for writing tests for Twisted
   code</a>.</p>
 
   </body>
