Index: tap.py
===================================================================
--- tap.py	(revision 33677)
+++ tap.py	(working copy)
@@ -118,10 +118,10 @@
     ca, cl = [], []
     if config['cache']:
         ca.append(cache.CacheResolver(verbose=config['verbose']))
+    if config['hosts-file']:
+        cl.append(hosts.Resolver(file=config['hosts-file']))
     if config['recursive']:
         cl.append(client.createResolver(resolvconf=config['resolv-conf']))
-    if config['hosts-file']:
-        cl.append(hosts.Resolver(file=config['hosts-file']))
 
     f = server.DNSServerFactory(config.zones, ca, cl, config['verbose'])
     p = dns.DNSDatagramProtocol(f)
