Index: twisted/web/client.py
===================================================================
--- twisted/web/client.py	(revision 27896)
+++ twisted/web/client.py	(working copy)
@@ -1,5 +1,5 @@
 # -*- test-case-name: twisted.web.test.test_webclient -*-
-# Copyright (c) 2001-2009 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2010 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 """
@@ -63,7 +63,7 @@
 
     def handleHeader(self, key, value):
         key = key.lower()
-        l = self.headers[key] = self.headers.get(key, [])
+        l = self.headers.setdefault(key, [])
         l.append(value)
 
     def handleStatus(self, version, status, message):
