Index: twisted/test/test_text.py
===================================================================
--- twisted/test/test_text.py	(revision 32062)
+++ twisted/test/test_text.py	(working copy)
@@ -15,6 +15,7 @@
 prominent place in chemistry - an aberration which is happily almost
 impossible - it would occasion a rapid and widespread degeneration of that
 science.
+
            --  Auguste Comte, Philosophie Positive, Paris, 1838
 """
 
@@ -148,7 +149,7 @@
 
     def test_self(self):
         self.assertEquals(True, text.strFile("this is a test string", self.io))
-    
+
     def test_insensitive(self):
         self.assertEquals(True, text.strFile("ThIs is A test STRING", self.io, False))
 
Index: twisted/python/text.py
===================================================================
--- twisted/python/text.py	(revision 32062)
+++ twisted/python/text.py	(working copy)
@@ -134,7 +134,7 @@
     if inString.find('\n\n') >= 0:
         paragraphs = string.split(inString, '\n\n')
         for para in paragraphs:
-            outLines.extend(greedyWrap(para) + [''])
+            outLines.extend(greedyWrap(para, width) + [''])
         return outLines
     inWords = string.split(inString)
 
