--- twisted/conch/test/test_conch.py
+++ twisted/conch/test/test_conch.py
@@ -402,7 +402,7 @@ class OpenSSHClientTestCase(ForwardingTestBase, unittest.TestCase):
                    ' 127.0.0.1 ' + remoteCommand
         port = self.conchServer.getHost().port
         cmds = (cmdline % port).split()
-        reactor.spawnProcess(process, "ssh", cmds)
+        reactor.spawnProcess(process, which("ssh")[0], cmds)
         return process.deferred
