*** /c/python/2.2/lib/site-packages/twisted/protocols/http.py	Fri Oct 17 20:44:19 2003
--- http.py	Fri Aug 13 12:22:57 2004
***************
*** 261,266 ****
--- 261,270 ----
  def stringToDatetime(dateString):
      """Convert an HTTP date string to seconds since epoch."""
      parts = dateString.split(' ')
+     if '-' in parts[1]:
+         # RFC850 format, further split date field.  Assume year 2000+
+         parts[1:2] = parts[1].split('-')
+         parts[3] = '20' + parts[3]
      day = int(parts[1])
      month = int(monthname.index(parts[2]))
      year = int(parts[3])
