--- F:\src\Web2\channel\scgi.py	Sun Jun 04 12:38:27 2006
+++ F:\python24\Lib\site-packages\twisted\web2\channel\scgi.py	Wed Oct 11 20:55:52 2006
@@ -15,7 +15,7 @@
         l = []
         code_message = responsecode.RESPONSES.get(code, "Unknown Status")
         
-        l.append("Status: %s %s\n" % (code, code_message))
+        l.append("Status: %s %s\r\n" % (code, code_message))
         if headers is not None:
             for name, valuelist in headers.getAllRawHeaders():
                 for value in valuelist:
