diff --git twisted/pair/ethernet.py twisted/pair/ethernet.py
index 2cefbc6..b432c6f 100644
--- twisted/pair/ethernet.py
+++ twisted/pair/ethernet.py
@@ -12,7 +12,7 @@ import struct
 
 from twisted.internet import protocol
 from twisted.pair import raw
-from zope.interface import implementer, Interface
+from zope.interface import implements, Interface
 
 
 class IEthernetProtocol(Interface):
@@ -29,8 +29,9 @@ class EthernetHeader:
         (self.dest, self.source, self.proto) \
                     = struct.unpack("!6s6sH", data[:6+6+2])
 
-@implementer(IEthernetProtocol)
 class EthernetProtocol(protocol.AbstractDatagramProtocol):
+
+    implements(IEthernetProtocol)
     
     def __init__(self):
         self.etherProtos = {}
diff --git twisted/pair/ip.py twisted/pair/ip.py
index e0f2549..de03bd4 100644
--- twisted/pair/ip.py
+++ twisted/pair/ip.py
@@ -12,7 +12,7 @@ import socket
 
 from twisted.internet import protocol
 from twisted.pair import raw
-from zope.interface import implementer
+from zope.interface import implements
 
 
 class IPHeader:
@@ -31,8 +31,8 @@ class IPHeader:
 
 MAX_SIZE = 2L**32
 
-@implementer(raw.IRawPacketProtocol)
 class IPProtocol(protocol.AbstractDatagramProtocol):
+    implements(raw.IRawPacketProtocol)
 
     def __init__(self):
         self.ipProtos = {}
diff --git twisted/pair/rawudp.py twisted/pair/rawudp.py
index 099b4ed..1425e6b 100644
--- twisted/pair/rawudp.py
+++ twisted/pair/rawudp.py
@@ -10,7 +10,7 @@ import struct
 
 from twisted.internet import protocol
 from twisted.pair import raw
-from zope.interface import implementer
+from zope.interface import implements
 
 class UDPHeader:
     def __init__(self, data):
@@ -18,8 +18,8 @@ class UDPHeader:
         (self.source, self.dest, self.len, self.check) \
                  = struct.unpack("!HHHH", data[:8])
 
-@implementer(raw.IRawDatagramProtocol)
 class RawUDPProtocol(protocol.AbstractDatagramProtocol):
+    implements(raw.IRawDatagramProtocol)
     def __init__(self):
         self.udpProtos = {}
 
diff --git twisted/pair/test/test_ethernet.py twisted/pair/test/test_ethernet.py
index 363477b..2b675fe 100644
--- twisted/pair/test/test_ethernet.py
+++ twisted/pair/test/test_ethernet.py
@@ -7,10 +7,10 @@ from twisted.trial import unittest
 from twisted.internet import protocol, reactor, error
 from twisted.python import failure, components
 from twisted.pair import ethernet, raw
-from zope.interface import implementer
+from zope.interface import implements
 
-@implementer(raw.IRawPacketProtocol)
 class MyProtocol:
+    implements(raw.IRawPacketProtocol)
     
     def __init__(self, expecting):
         self.expecting = list(expecting)
diff --git twisted/pair/test/test_ip.py twisted/pair/test/test_ip.py
index a4a1ed9..ed1623b 100644
--- twisted/pair/test/test_ip.py
+++ twisted/pair/test/test_ip.py
@@ -7,10 +7,10 @@ from twisted.trial import unittest
 from twisted.internet import protocol, reactor, error
 from twisted.python import failure, components
 from twisted.pair import ip, raw
-from zope.interface import implementer
+from zope import interface
 
-@implementer(raw.IRawDatagramProtocol)
 class MyProtocol:
+    interface.implements(raw.IRawDatagramProtocol)
     
     def __init__(self, expecting):
         self.expecting = list(expecting)
