Index: twisted/names/dns.py
===================================================================
--- twisted/names/dns.py	(revision 21254)
+++ twisted/names/dns.py	(working copy)
@@ -1217,7 +1217,10 @@
         """
         if not self.transport:
             # XXX transport might not get created automatically, use callLater?
-            self.startListening()
+            try:
+                self.startListening()
+            except:
+                return defer.fail()
 
         if id is None:
             id = self.pickID()
