Index: sasl_mechanisms.py
===================================================================
--- sasl_mechanisms.py	(revision 18175)
+++ sasl_mechanisms.py	(working copy)
@@ -71,7 +71,7 @@
     def __init__(self, serv_type, host, serv_name, username, password):
         self.username = username
         self.password = password
-
+        self.host = host
         self.digest_uri = '%s/%s' % (serv_type, host)
         if serv_name is not None:
             self.digest_uri += '/%s' % serv_name
@@ -90,7 +90,7 @@
             return ''
 
         return self._gen_response(directives['charset'],
-                                  directives['realm'],
+                                  directives.get('realm', self.host),
                                   directives['nonce'])
 
 
