Index: twisted/internet/win32eventreactor.py
===================================================================
--- twisted/internet/win32eventreactor.py	(revision 33398)
+++ twisted/internet/win32eventreactor.py	(working copy)
@@ -248,7 +248,7 @@
 
         handles = self._events.keys() or [self.dummyEvent]
         timeout = int(timeout * 1000)
-        val = MsgWaitForMultipleObjects(handles, 0, timeout, QS_ALLINPUT | QS_ALLEVENTS)
+        val = MsgWaitForMultipleObjects(handles, 0, timeout, QS_ALLINPUT)
         if val == WAIT_TIMEOUT:
             return
         elif val == WAIT_OBJECT_0 + len(handles):
