Index: win32/twisted_postinstall.py
===================================================================
--- win32/twisted_postinstall.py	(revision 23166)
+++ win32/twisted_postinstall.py	(working copy)
@@ -30,7 +30,7 @@
     scripts = join(prefix, 'scripts')
     pyexe = join(prefix, 'python.exe')
     for bat in """twistd.bat mktap.bat lore.bat 
-               manhole.bat tapconvert.bat trial.bat coil.bat""".split():
+               manhole.bat tapconvert.bat trial.bat""".split():
         f = join(scripts, bat)
         scriptpy = f.replace('.bat', '.py')
         file(f, 'w').write("@%s %s %%*" % (pyexe, scriptpy))
