Index: twisted/web2/dav/element/base.py
===================================================================
--- twisted/web2/dav/element/base.py	(revision 35)
+++ twisted/web2/dav/element/base.py	(working copy)
@@ -196,7 +196,12 @@
         return output.getvalue()
 
     def element(self, document):
-        element = document.createElementNS(self.namespace, self.name)
+        ename = self.name
+        # If element is part of DAV namespace and does not already have
+        # a prefix, add the "DAV:" prefix.
+        if ((self.namespace == dav_namespace) and (ename.find(':') == -1)):
+            ename = "DAV:" + ename
+        element = document.createElementNS(self.namespace, ename)
         if hasattr(self, "attributes"):
             for name, value in self.attributes.items():
                 namespace, name = decodeXMLName(name)
