Index: twisted/words/protocols/jabber/xmlstream.py
===================================================================
--- twisted/words/protocols/jabber/xmlstream.py	(révision 31801)
+++ twisted/words/protocols/jabber/xmlstream.py	(copie de travail)
@@ -457,7 +457,7 @@
     Wraps a transport with zlib compression, to implement XEP-0138 (stream
     compression). Used by L{CompressInitiatingInitializer}.
 
-    @since: 11.0
+    @since: 11.1
     """
 
     def __init__(self, wrappedProtocol):
@@ -515,7 +515,7 @@
     """
     Exception concerning stream compression (XEP-0138).
 
-    @since: 11.0
+    @since: 11.1
     """
 
 
@@ -536,7 +536,7 @@
         is already used.
     @type withTLS: C{bool}
 
-    @since: 11.0
+    @since: 11.1
     """
 
     feature = (NS_XMPP_FEATURE_COMPRESS, 'compression')
