=== modified file 'twisted/web/http.py'
--- twisted/web/http.py	2010-01-18 00:16:57 +0000
+++ twisted/web/http.py	2010-02-23 21:38:26 +0000
@@ -1734,9 +1734,9 @@
 
     logPath = None
 
-    timeOut = 60 * 60 * 12
+    timeOut = 15
 
-    def __init__(self, logPath=None, timeout=60*60*12):
+    def __init__(self, logPath=None, timeout=15):
         if logPath is not None:
             logPath = os.path.abspath(logPath)
         self.logPath = logPath

=== modified file 'twisted/web/server.py'
--- twisted/web/server.py	2009-09-14 17:25:32 +0000
+++ twisted/web/server.py	2010-02-23 21:39:12 +0000
@@ -441,7 +441,7 @@
     sessionFactory = Session
     sessionCheckTime = 1800
 
-    def __init__(self, resource, logPath=None, timeout=60*60*12):
+    def __init__(self, resource, logPath=None, timeout=15):
         """
         Initialize.
         """

