Index: core/examples/pairudp.py
===================================================================
--- core/examples/pairudp.py	(revision 24208)
+++ core/examples/pairudp.py	(working copy)
@@ -1,18 +0,0 @@
-from twisted.internet import reactor, protocol
-from twisted.pair import ethernet, rawudp, ip
-from twisted.pair import tuntap
-
-class MyProto(protocol.DatagramProtocol):
-    def datagramReceived(self, *a, **kw):
-        print a, kw
-
-p_udp = rawudp.RawUDPProtocol()
-p_udp.addProto(42, MyProto())
-p_ip = ip.IPProtocol()
-p_ip.addProto(17, p_udp)
-p_eth = ethernet.EthernetProtocol()
-p_eth.addProto(0x800, p_ip)
-
-reactor.listenWith(tuntap.TuntapPort,
-                   interface='tap0', proto=p_eth, reactor=reactor)
-reactor.run()
Index: core/examples/index.xhtml
===================================================================
--- core/examples/index.xhtml	(revision 24208)
+++ core/examples/index.xhtml	(working copy)
@@ -83,7 +83,6 @@
         <li><a href="example.html">example.html</a></li>
         <li><a href="longex.py">longex.py</a></li>
         <li><a href="longex2.py">longex2.py</a></li>
-        <li><a href="pairudp.py">pairudp.py</a></li>
         <li><a href="stdin.py">stdin.py</a></li>
         <li><a href="filewatch.py">filewatch.py</a></li>
         <li><a href="gpsfix.py">gpsfix.py</a></li>
Index: pair/examples/index.xhtml
===================================================================
--- pair/examples/index.xhtml	(revision 0)
+++ pair/examples/index.xhtml	(revision 0)
@@ -0,0 +1,20 @@
+<?xml version="1.0"?>
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
+    "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
+
+<html xmlns="http://www.w3.org/1999/xhtml">
+<head>
+<title>Twisted code examples</title>
+</head>
+
+<body>
+    <h1>Twisted code examples</h1>
+
+    <h2>Miscellaenous</h2>
+    <ul>
+        <li><a href="pairudp.py">pairudp.py</a></li>
+    </ul>
+    
+</body>
+</html>
+
