--- twisted/logger/_flatten.py.orig	2015-08-20 09:14:37.237133165 -0400
+++ twisted/logger/_flatten.py	2015-08-20 09:15:09.877132201 -0400
@@ -170,6 +170,9 @@
     keyFlattener = KeyFlattener()
     formatFields = aFormatter.parse(event["log_format"])
     for literalText, fieldName, formatSpec, conversion in formatFields:
-        key = keyFlattener.flatKey(fieldName, formatSpec, conversion or "s")
-        s.extend([literalText, unicode(fieldValues[key])])
+        s.append(literalText)
+        if fieldName is not None:
+            key = keyFlattener.flatKey(fieldName, formatSpec, conversion or "s")
+            s.append(unicode(fieldValues[key]))
+
     return u"".join(s)
