Index: twisted/words/test/test_irc.py
===================================================================
--- twisted/words/test/test_irc.py	(revision 32957)
+++ twisted/words/test/test_irc.py	(working copy)
@@ -1877,4 +1877,5 @@
         def privmsg(user, channel, message):
             self.fail("privmsg() should not have been called")
         self.protocol.privmsg = privmsg
-        self.protocol.irc_NOTICE('spam', "I don't want any spam!")
+        self.protocol.irc_NOTICE('spam',
+            ['#greasyspooncafe', "I don't want any spam!"])
