Index: twisted/test/test_amp.py
===================================================================
--- twisted/test/test_amp.py	(revision 20367)
+++ twisted/test/test_amp.py	(working copy)
@@ -1232,7 +1232,7 @@
             L.append(defer.maybeDeferred(self.serverPort.stopListening))
         if self.clientConn is not None:
             self.clientConn.disconnect()
-        return defer.DeferredList(L)
+        return defer.gatherResults(L)
 
 def show(x):
     import sys
@@ -1288,9 +1288,11 @@
                                    tls_verifyAuthorities=[cert]).addCallback(secured)
 
 class SlightlySmartTLS(SimpleSymmetricCommandProtocol):
-    def tlisfy(self):
+    def getTLSVars(self):
         return dict(tls_localCertificate=tempcert)
+    amp.StartTLS.responder(getTLSVars)
 
+
 class PlainVanillaLiveFire(LiveFireBase, unittest.TestCase):
 
     clientProto = SimpleSymmetricCommandProtocol
@@ -1315,4 +1317,6 @@
         """
         def secured(result):
             return self.cli.callRemote(SecuredPing)
-        return self.cli.callRemote(amp.StartTLS, tls_verifyAuthorities=[tempcert])
+        return self.cli.callRemote(amp.StartTLS, tls_verifyAuthorities=[tempcert]
+            ).addCallback(secured)
+
