Index: twisted/names/dns.py
===================================================================
--- twisted/names/dns.py	(revision 34050)
+++ twisted/names/dns.py	(working copy)
@@ -1410,12 +1410,6 @@
         self.name = Name()
         self.name.decode(strio)
 
-    def exchange(self):
-        warnings.warn("use Record_MX.name instead", DeprecationWarning, stacklevel=2)
-        return self.name
-
-    exchange = property(exchange)
-
     def __hash__(self):
         return hash((self.preference, self.name))
 
