Index: microdom.py
===================================================================
--- microdom.py	(revision 10433)
+++ microdom.py	(working copy)
@@ -47,6 +47,7 @@
 from types import ListType as NodeList
 from types import StringTypes, UnicodeType
 import sys
+
 dictsAreNotSequences=0
 try:
     'x' in {}
@@ -595,7 +596,7 @@
                     }
 
 
-    def __init__(self, beExtremelyLenient=0, caseInsensitive=1, preserveCase=0, soonClosers=soonClosers, laterClosers=laterClosers):
+    def __init__(self, beExtremelyLenient=0, caseInsensitive=1, preserveCase=0, preserveSpace=0, soonClosers=soonClosers, laterClosers=laterClosers):
         self.elementstack = []
         d = {'xmlns': 'xmlns', '':''}
         dr = _reverseDict(d)
@@ -605,11 +606,14 @@
         self.beExtremelyLenient = beExtremelyLenient
         self.caseInsensitive = caseInsensitive
         self.preserveCase = preserveCase or not caseInsensitive
+        self.preserveSpace = preserveSpace
         self.soonClosers = soonClosers
         self.laterClosers = laterClosers
         # self.indentlevel = 0
 
     def shouldPreserveSpace(self):
+        if self.preserveSpace: 
+            return 1
         for edx in xrange(len(self.elementstack)):
             el = self.elementstack[-edx]
             if el.tagName == 'pre' or el.getAttribute("xml:space", '') == 'preserve':
