Index: twisted/internet/_posixserialport.py
===================================================================
--- twisted/internet/_posixserialport.py	(revision 46111)
+++ twisted/internet/_posixserialport.py	(working copy)
@@ -6,13 +6,15 @@
 Serial Port Protocol
 """
 
+from __future__ import division, absolute_import
+
 # dependent on pyserial ( http://pyserial.sf.net/ )
 # only tested w/ 1.18 (5 Dec 2002)
 from serial import PARITY_NONE
 from serial import STOPBITS_ONE
 from serial import EIGHTBITS
 
-from serialport import BaseSerialPort
+from .serialport import BaseSerialPort
 
 from twisted.internet import abstract, fdesc
 
Index: twisted/internet/_win32serialport.py
===================================================================
--- twisted/internet/_win32serialport.py	(revision 46111)
+++ twisted/internet/_win32serialport.py	(working copy)
@@ -7,6 +7,8 @@
 Requires PySerial and pywin32.
 """
 
+from __future__ import division, absolute_import
+
 # system imports
 from serial import PARITY_NONE
 from serial import STOPBITS_ONE
@@ -17,7 +19,7 @@
 from twisted.internet import abstract
 
 # sibling imports
-from serialport import BaseSerialPort
+from .serialport import BaseSerialPort
 
 
 class SerialPort(BaseSerialPort, abstract.FileDescriptor):
Index: twisted/internet/serialport.py
===================================================================
--- twisted/internet/serialport.py	(revision 46111)
+++ twisted/internet/serialport.py	(working copy)
@@ -6,6 +6,8 @@
 Serial Port Protocol
 """
 
+from __future__ import division, absolute_import
+
 # http://twistedmatrix.com/trac/ticket/3725#comment:24
 # Apparently applications use these names even though they should
 # be imported from pyserial
Index: twisted/python/dist3.py
===================================================================
--- twisted/python/dist3.py	(revision 46111)
+++ twisted/python/dist3.py	(working copy)
@@ -59,6 +59,7 @@
     "twisted.internet._glibbase",
     "twisted.internet._newtls",
     "twisted.internet._posixstdio",
+    "twisted.internet._posixserialport",
     "twisted.internet._signals",
     "twisted.internet.abstract",
     "twisted.internet.address",
@@ -80,6 +81,7 @@
     "twisted.internet.protocol",
     "twisted.internet.reactor",
     "twisted.internet.selectreactor",
+    "twisted.internet.serialport",
     "twisted.internet.ssl",
     "twisted.internet.stdio",
     "twisted.internet.task",
@@ -257,6 +259,7 @@
     "twisted.internet.test.test_posixprocess",
     "twisted.internet.test.test_process",
     "twisted.internet.test.test_protocol",
+    "twisted.internet.test.test_serialport",
     "twisted.internet.test.test_sigchld",
     "twisted.internet.test.test_stdio",
     "twisted.internet.test.test_tcp",
