diff --git twisted/test/test_monkey.py twisted/test/test_monkey.py
index 95c0454..e31f020 100644
--- twisted/test/test_monkey.py
+++ twisted/test/test_monkey.py
@@ -151,7 +151,7 @@ class MonkeyPatcherTest(unittest.TestCase):
         def _():
             self.assertEqual(self.testObject.foo, 'haha')
             self.assertEqual(self.testObject.bar, 'blahblah')
-            raise RuntimeError, "Something went wrong!"
+            raise RuntimeError("Something went wrong!")
 
         self.monkeyPatcher.addPatch(self.testObject, 'foo', 'haha')
         self.monkeyPatcher.addPatch(self.testObject, 'bar', 'blahblah')
