diff --git twisted/plugins/twisted_trial.py twisted/plugins/twisted_trial.py
index debc8af..74fab78 100644
--- twisted/plugins/twisted_trial.py
+++ twisted/plugins/twisted_trial.py
@@ -1,11 +1,11 @@
 
-from zope.interface import implements
+from zope.interface import implementer
 
 from twisted.trial.itrial import IReporter
 from twisted.plugin import IPlugin
 
+@implementer(IPlugin, IReporter)
 class _Reporter(object):
-    implements(IPlugin, IReporter)
 
     def __init__(self, name, module, description, longOpt, shortOpt, klass):
         self.name = name
