Index: doc/core/howto/tutorial/intro.xhtml
===================================================================
--- doc/core/howto/tutorial/intro.xhtml	(revision 27896)
+++ doc/core/howto/tutorial/intro.xhtml	(working copy)
@@ -28,7 +28,7 @@
 </pre>
 
 <p>If the target computer does not have the <code>fingerd</code> <a
-href="glossary.xhtml#Daemon">daemon</a> running you'll get a "Connection
+href="../glossary.xhtml#Daemon">daemon</a> running you'll get a "Connection
 Refused" error. Paranoid sysadmins keep <code>fingerd</code> off or limit the
 output to hinder crackers and harassers. The above format is the standard
 <code>fingerd</code> default, but an alternate implementation can output
