--- imap4.py	2006-08-08 17:00:12.000000000 -0700
+++ imap4.py.new	2006-08-08 16:59:53.000000000 -0700
@@ -1635,8 +1635,11 @@
         except StopIteration:
             self.sendPositiveResponse(tag, 'FETCH completed')
             self._unblock()
-            self.setTimeout(self._oldTimeout)
-            del self._oldTimeout
+            self.setTimeout(getattr(self,'_oldTimeout',None))
+            try:
+                del self._oldTimeout
+            except AttributeError:
+                pass
         else:
             self.spewMessage(id, msg, query, uid
                 ).addCallback(lambda _: self.__cbFetch(results, tag, query, uid)
