Index: tcp.py
===================================================================
--- tcp.py	(revision 20276)
+++ tcp.py	(working copy)
@@ -524,6 +524,8 @@
 
         err = self.socket.getsockopt(socket.SOL_SOCKET, socket.SO_ERROR)
         if err:
+            del self.doWrite
+            del self.doRead
             self.failIfNotConnected(error.getConnectError((err, os.strerror(err))))
             return
 
@@ -550,6 +552,8 @@
                 self.startWriting()
                 return
             else:
+                del self.doWrite
+                del self.doRead
                 self.failIfNotConnected(error.getConnectError((connectResult, os.strerror(connectResult))))
                 return
 
