diff --git a/docs/Makefile b/docs/Makefile
index 8857977..0667116 100644
--- a/docs/Makefile
+++ b/docs/Makefile
@@ -16,8 +16,11 @@ endif
 PAPEROPT_a4     = -D latex_paper_size=a4
 PAPEROPT_letter = -D latex_paper_size=letter
 ALLSPHINXOPTS   = -d $(BUILDDIR)/doctrees $(PAPEROPT_$(PAPER)) $(SPHINXOPTS) .
+# Turn all warning into errors, complain about broken references and rebuild
+# all files. It runs in quite mode to reduce stored logs on buildbot.
+TESTSPHINXOPTS  = -W -n -a -E -q
 # the i18n builder cannot share the environment and doctrees with the others
-I18NSPHINXOPTS  = $(PAPEROPT_$(PAPER)) $(SPHINXOPTS) .
+I18NSPHINXOPTS  = $(PAPEROPT_$(PAPER)) $(SPHINXOPTS)
 
 .PHONY: help clean html dirhtml singlehtml pickle json htmlhelp qthelp devhelp epub latex latexpdf text man changes linkcheck doctest gettext
 
@@ -54,6 +57,11 @@ html:
 	@echo
 	@echo "Build finished. The HTML pages are in $(BUILDDIR)/html."
 
+test-html:
+	$(SPHINXBUILD) $(TESTSPHINXOPTS) -b html $(ALLSPHINXOPTS) $(BUILDDIR)/html
+	@echo
+	@echo "Test finished."
+
 dirhtml:
 	$(SPHINXBUILD) -b dirhtml $(ALLSPHINXOPTS) $(BUILDDIR)/dirhtml
 	@echo
diff --git a/docs/projects/web/howto/web-in-60/access-logging.rst b/docs/projects/web/howto/web-in-60/access-logging.rst
index 9dd3bdb..45f578c 100644
--- a/docs/projects/web/howto/web-in-60/access-logging.rst
+++ b/docs/projects/web/howto/web-in-60/access-logging.rst
@@ -27,7 +27,7 @@ Do this by passing ``logPath`` to the initializer:
 
 Or if you want to change the logging behavior of a server you're launching with ``twistd web`` then just pass the ``--logfile`` option:
 
-.. code-block:: shell
+.. code-block:: sh
 
     $ twistd -n web --logfile /tmp/access-logging-demo.log
 
