Index: twisted/internet/endpoints.py
===================================================================
--- twisted/internet/endpoints.py	(revision 39080)
+++ twisted/internet/endpoints.py	(working copy)
@@ -1599,6 +1599,8 @@
     @return: The result of calling C{connect} on the endpoint, i.e. a
     L{Deferred} that will fire with the protocol when connected, or an
     appropriate error.
+
+    @since: 13.1
     """
     class OneShotFactory(Factory):
         def buildProtocol(self, addr):
