diff --git a/twisted/web/http.py b/twisted/web/http.py
index 4024b0f..fe29fe7 100644
--- a/twisted/web/http.py
+++ b/twisted/web/http.py
@@ -781,7 +781,6 @@ class Request:
         """
         self.content.seek(0,0)
         self.args = {}
-        self.stack = []
 
         self.method, self.uri = command, path
         self.clientproto = version
diff --git a/twisted/web/topfiles/6959.misc b/twisted/web/topfiles/6959.misc
new file mode 100644
index 0000000..e69de29
