Index: twisted/mail/test/test_imap.py
===================================================================
--- twisted/mail/test/test_imap.py	(revision 29988)
+++ twisted/mail/test/test_imap.py	(working copy)
@@ -3652,19 +3652,23 @@
         self.messages = '21'
         self.msgObjs = [
             FakeyMessage(
-                {'content-type': 'multipart/alternative'}, (), '',
-                'Irrelevant', 12345, [
+                {'content-type': 'multipart/mixed'}, (), '',
+                'RootOf', 98765, [
                     FakeyMessage(
-                        {'content-type': 'text/plain'}, (), 'date',
-                        'Stuff', 54321,  None),
-                    FakeyMessage(
-                        {'content-type': 'text/html'}, (), 'date',
-                        'Things', 32415, None)])]
+                        {'content-type': 'multipart/alternative'}, (), '',
+                        'Irrelevant', 12345, [
+                            FakeyMessage(
+                                {'content-type': 'text/plain'}, (), 'date',
+                                'Stuff', 54321,  None),
+                            FakeyMessage(
+                                {'content-type': 'text/html'}, (), 'date',
+                                'Things', 32415, None)])])]
         self.expected = {
             0: {'BODY': [
-                    ['text', 'plain', [], None, None, None, '5', '1'],
-                    ['text', 'html', [], None, None, None, '6', '1'],
-                    'alternative']}}
+                    [['text', 'plain', [], None, None, None, '5', '1'],
+                     ['text', 'html', [], None, None, None, '6', '1'],
+                    'alternative'],
+                    'mixed']}}
 
         return self._fetchWork(False)
 
