From 8d94adef5b48ced5c59018933d4c614e07efdbfb Mon Sep 17 00:00:00 2001
From: Carlos Valiente <carlos.valiente@ecmwf.int>
Date: Tue, 29 Jun 2010 15:40:30 +0100
Subject: [PATCH] Git: Ignore generated files.

---
 .gitignore |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)
 create mode 100644 .gitignore

diff --git a/.gitignore b/.gitignore
new file mode 100644
index 0000000..b7d57fd
--- /dev/null
+++ b/.gitignore
@@ -0,0 +1,6 @@
+*.pyc
+/MANIFEST
+/_trial_temp/
+/build/
+/dist/
+/twisted/plugins/dropin.cache
-- 
1.7.1

