Index: twisted/web/__init__.py
===================================================================
--- twisted/web/__init__.py	(révision 32445)
+++ twisted/web/__init__.py	(copie de travail)
@@ -10,4 +10,11 @@
 """
 
 from twisted.web._version import version
+from twisted.python.versions import Version
+from twisted.python.deprecate import deprecatedModuleAttribute 
 __version__ = version.short()
+
+deprecatedModuleAttribute(
+    Version('Twisted', 11, 1, 0),
+    "Google module is deprecated. Use Google's API instead",
+    __name__, "google")
Index: twisted/web/test/test_web.py
===================================================================
--- twisted/web/test/test_web.py	(révision 32445)
+++ twisted/web/test/test_web.py	(copie de travail)
@@ -597,6 +597,14 @@
         return d
 
 
+    def test_deprecated(self):
+        """ 
+        Google module is deprecated since Twisted 11.1.0
+        """
+        from twisted.web import google
+        warnings = self.flushWarnings(offendingFunctions=[self.test_deprecated])
+        self.assertEqual(len(warnings),1)
+        self.assertEqual(warnings[0]['category'], DeprecationWarning)
Index: twisted/web/google.py
===================================================================
--- twisted/web/google.py	(révision 32445)
+++ twisted/web/google.py	(copie de travail)
@@ -7,6 +7,9 @@
 import urllib
 from twisted.internet import protocol, reactor, defer
 from twisted.web import http
+from twisted.python.deprecate import deprecated
+from twisted.python.versions import Version, getVersionString
+import warnings
 
 class GoogleChecker(http.HTTPClient):
