diff --git INSTALL INSTALL
index 6c08d02..cb23d1f 100644
--- INSTALL
+++ INSTALL
@@ -2,7 +2,7 @@ Requirements
 
   Python 2.6 or 2.7.
 
-  Zope Interfaces 3.3.0 or better (http://pypi.python.org/pypi/zope.interface)
+  Zope Interface 3.6.0 or better (http://pypi.python.org/pypi/zope.interface)
 
   pyOpenSSL (<http://launchpad.net/pyopenssl>) is required for any SSL APIs.  On
   Windows, version 0.10 or newer is required.  pyOpenSSL 0.10 or newer is also
diff --git twisted/topfiles/5683.feature twisted/topfiles/5683.feature
new file mode 100644
index 0000000..db28298
--- /dev/null
+++ twisted/topfiles/5683.feature
@@ -0,0 +1 @@
+The minimum required version of zope.interface is now 3.6.0.
\ No newline at end of file
