Index: twisted/web/test/test_newclient.py
===================================================================
--- twisted/web/test/test_newclient.py	(revision 30077)
+++ twisted/web/test/test_newclient.py	(working copy)
@@ -1795,6 +1795,14 @@
         d.addCallback(cbFinished)
         return d
 
+    def test_proxiesConsumerMethods(self):
+        """
+        The proxy should proxy methods on the wrapped L{IConsumer}.
+        """
+        self.assertEquals(self.enforcer.registerProducer,
+                          self.enforcer._consumer.registerProducer)
+        self.assertEquals(self.enforcer.unregisterProducer,
+                          self.enforcer._consumer.unregisterProducer)
 
 
 class RequestBodyConsumerTests(TestCase):
Index: twisted/web/_newclient.py
===================================================================
--- twisted/web/_newclient.py	(revision 30077)
+++ twisted/web/_newclient.py	(working copy)
@@ -790,6 +790,8 @@
             if self._length:
                 raise WrongBodyLength("too few bytes written")
 
+    def __getattr__(self, attr):
+        return getattr(self._consumer, attr)
 
 
 def makeStatefulDispatcher(name, template):
