Index: widgets.py
===================================================================
RCS file: /cvs/Twisted/twisted/web/woven/widgets.py,v
retrieving revision 1.104
diff -u -r1.104 widgets.py
--- widgets.py	28 Aug 2003 11:17:45 -0000	1.104
+++ widgets.py	16 Sep 2003 09:16:24 -0000
@@ -661,7 +661,7 @@
     tagName = 'select'
 
 
-class Option(Input):
+class Option(Widget):
     tagName = 'option'
     def initialize(self):
         self.text = ''
@@ -677,8 +677,10 @@
 
     def setUp(self, request, node, data):
         self.add(Text(self.text or data))
-        Input.setUp(self, request, node, data)
-
+        if data is None:
+            data = ''
+        if not self.attributes.has_key('value'):
+            self['value'] = str(data)
 
 class Anchor(Widget):
     tagName = 'a'
