Index: twisted/topfiles/6430.misc
===================================================================
Index: twisted/trial/test/test_loader.py
===================================================================
--- twisted/trial/test/test_loader.py	(revision 38110)
+++ twisted/trial/test/test_loader.py	(working copy)
@@ -179,8 +179,10 @@
         L{runner.filenameToModule} raises a C{SyntaxError} when a non-Python
         file is passed.
         """
-        self.failUnlessRaises(SyntaxError, runner.filenameToModule,
-                              util.sibpath(__file__, 'notpython'))
+        filename = filepath.FilePath(self.parent).child('notpython') 
+        filename.setContent("This isn't python") 
+        self.failUnlessRaises(
+            SyntaxError, runner.filenameToModule, filename.path) 
 
 
     def test_filenameMatchesPackage(self):
Index: twisted/trial/test/notpython
===================================================================
--- twisted/trial/test/notpython	(revision 38110)
+++ twisted/trial/test/notpython	(working copy)
@@ -1,2 +0,0 @@
-
-this isn't python
