Index: twisted/trial/test/test_loader.py
===================================================================
--- twisted/trial/test/test_loader.py	(revision 38098)
+++ twisted/trial/test/test_loader.py	(working copy)
@@ -181,6 +181,11 @@
         """
         self.failUnlessRaises(SyntaxError, runner.filenameToModule,
                               util.sibpath(__file__, 'notpython'))
+        filename = filepath.FilePath(self.parent).child('notpython2')
+        filename.setContent("This isn't python, either")
+        self.failUnlessRaises(SyntaxError, runner.filenameToModule,
+                              filename.path)
+        filename.remove()
 
 
     def test_filenameMatchesPackage(self):
