Index: interfaces.py
===================================================================
--- interfaces.py	(revision 16494)
+++ interfaces.py	(working copy)
@@ -149,6 +149,8 @@
         @type portType: type which implements L{IListeningPort}
         @param portType: The object given by C{portType(*args, **kw)}
         will be started listening.
+
+        @returns: an object that satisfies the L{IListeningPort} interface
         """
 
     def connectWith(self, connectorType, *args, **kw):
@@ -158,6 +160,8 @@
         @type connectorType: type which implements L{IConnector}
         @param connectorType: The object given by C{connectorType(*args, **kw)}
         will be started connecting.
+
+        @returns:  An object implementing the L{IConnector} interface
         """
 
 class IReactorTCP(Interface):
@@ -272,6 +276,8 @@
         @param mode: mode to set on the unix socket.
 
         @param wantPID: if True, create a pidfile for the socket.
+
+        @returns:  An object implementing the L{IListeningPort} interface
         """
 
 
@@ -290,6 +296,8 @@
         @param mode: mode to set on the unix socket.
 
         @param bindAddress: address to bind to
+
+        @returns: an L{IConnector}.
         """
 
     def listenUNIXDatagram(self, address, protocol, maxPacketSize=8192, mode=0666):
