Index: twisted/web/server.py
===================================================================
--- twisted/web/server.py	(revision 33693)
+++ twisted/web/server.py	(working copy)
@@ -34,7 +34,20 @@
 from twisted.web.error import UnsupportedMethod
 from twisted.web.microdom import escape
 
+from twisted.python.versions import Version
+from twisted.python.deprecate import deprecatedModuleAttribute
+
 # backwards compatability
+deprecatedModuleAttribute(
+    Version("Twisted", 12, 2, 0),
+    "No longer used",
+    "twisted.web.server",
+    "date_time_string")
+deprecatedModuleAttribute(
+    Version("Twisted", 12, 2, 0),
+    "No longer used",
+    "twisted.web.server",
+    "string_date_time")
 date_time_string = http.datetimeToString
 string_date_time = http.stringToDatetime
 
@@ -563,4 +576,3 @@
 
 
 import html
-
