Index: twisted/web/server.py
===================================================================
--- twisted/web/server.py	(revision 33732)
+++ twisted/web/server.py	(working copy)
@@ -34,7 +34,20 @@
 from twisted.web.error import UnsupportedMethod
 from twisted.web.microdom import escape
 
+from twisted.python.versions import Version
+from twisted.python.deprecate import deprecatedModuleAttribute
+
 # backwards compatability
+deprecatedModuleAttribute(
+    Version("Twisted", 12, 1, 0),
+    "Please use twisted.web.http.datetimeToString instead",
+    "twisted.web.server",
+    "date_time_string")
+deprecatedModuleAttribute(
+    Version("Twisted", 12, 1, 0),
+    "Please use twisted.web.http.stringToDatetime instead",
+    "twisted.web.server",
+    "string_date_time")
 date_time_string = http.datetimeToString
 string_date_time = http.stringToDatetime
 
@@ -563,4 +576,3 @@
 
 
 import html
-
Index: twisted/web/test/test_web.py
===================================================================
--- twisted/web/test/test_web.py	(revision 33732)
+++ twisted/web/test/test_web.py	(working copy)
@@ -1057,3 +1057,32 @@
         self.assertEqual(
             self.site.logFile.read(),
             '1.2.3.4 - - [25/Oct/2004:12:31:59 +0000] "GET /dummy HTTP/1.0" 123 - "-" "Malicious Web\\" Evil"\n')
+
+
+
+class ServerAttributesTest(unittest.TestCase):
+
+    def test_deprecatedAttributeDateTimeString(self):
+        deprecated_func = server.date_time_string
+        warnings = self.flushWarnings(
+            offendingFunctions=[self.test_deprecatedAttributeDateTimeString])
+
+        self.assertEqual(len(warnings), 1)
+        self.assertEqual(warnings[0]['category'], DeprecationWarning)
+        self.assertEqual(
+            warnings[0]['message'],
+            ("twisted.web.server.date_time_string was deprecated in Twisted "
+             "12.1.0: Please use twisted.web.http.datetimeToString instead"))
+
+
+    def test_deprecatedAttributeStringDateTime(self):
+        deprecated_func = server.string_date_time
+        warnings = self.flushWarnings(
+            offendingFunctions=[self.test_deprecatedAttributeStringDateTime])
+
+        self.assertEqual(len(warnings), 1)
+        self.assertEqual(warnings[0]['category'], DeprecationWarning)
+        self.assertEqual(
+            warnings[0]['message'],
+            ("twisted.web.server.string_date_time was deprecated in Twisted "
+             "12.1.0: Please use twisted.web.http.stringToDatetime instead"))
