--- failure_old.py	2008-03-27 18:08:36.000000000 -0400
+++ failure_new.py	2008-03-27 18:09:28.000000000 -0400
@@ -363,7 +363,7 @@
         # it is only really originating from
         # throwExceptionIntoGenerator if the bottom of the traceback
         # is a yield.
-        if lastFrame.f_code.co_code[lastTb.tb_lasti] != cls._yieldOpcode:
+        if (not lastFrame.f_code.co_code) or lastFrame.f_code.co_code[lastTb.tb_lasti] != cls._yieldOpcode:
             return
 
         # if the exception was caught above the generator.throw
