Index: twisted/web/test/test_webclient.py
===================================================================
--- twisted/web/test/test_webclient.py	(revision 30367)
+++ twisted/web/test/test_webclient.py	(working copy)
@@ -164,7 +164,19 @@
         return util.Redirect.render(self, request)
 
 
+class CountingResource(resource.Resource):
+    """
+    A resource that keeps track of the number of times it has been accessed.
+    """
+    def __init__(self):
+        resource.Resource.__init__(self)
+        self.count = 0
 
+    def render(self, request):
+        self.count += 1
+        return "Success"
+
+
 class ParseUrlTestCase(unittest.TestCase):
     """
     Test URL parsing facility and defaults values.
@@ -283,6 +295,10 @@
         r.putChild("broken", BrokenDownloadResource())
         r.putChild("cookiemirror", CookieMirrorResource())
 
+        self.afterFoundGetCounter = CountingResource()
+        r.putChild("afterFoundGetCounter", self.afterFoundGetCounter)
+        r.putChild("afterFoundGetRedirect", util.Redirect("/afterFoundGetCounter"))
+
         miscasedHead = static.Data("miscased-head GET response content", "major/minor")
         miscasedHead.render_Head = lambda request: "miscased-head content"
         r.putChild("miscased-head", miscasedHead)
@@ -596,6 +612,21 @@
         return d
 
 
+    def test_afterFoundGetMakesOneRequest(self):
+        """
+        Test for regression on bug 4760: when using afterFoundGet, make sure
+        only one request gets sent to the server.
+        """
+        def checkRedirectCount(*a):
+            self.assertEquals(self.afterFoundGetCounter.count, 1)
+
+        url = self.getURL('afterFoundGetRedirect')
+        d = client.getPage(
+            url, followRedirect=True, afterFoundGet=True, method="POST")
+        d.addCallback(checkRedirectCount)
+        return d
+
+
     def testPartial(self):
         name = self.mktemp()
         f = open(name, "wb")
Index: twisted/web/topfiles/4760.bugfix
===================================================================
--- twisted/web/topfiles/4760.bugfix	(revision 0)
+++ twisted/web/topfiles/4760.bugfix	(revision 0)
@@ -0,0 +1,2 @@
+twisted.web.client.HTTPPageGetter and http.web.client.getPage now no longer makes two requests when using afterFoundGet
+twisted.web.client.HTTPDownloader and http.web.client.downloadPage now accept afterFoundGet
Index: twisted/web/client.py
===================================================================
--- twisted/web/client.py	(revision 30367)
+++ twisted/web/client.py	(working copy)
@@ -144,7 +144,8 @@
     def handleStatus_302(self):
         if self.afterFoundGet:
             self.handleStatus_303()
-        self.handleStatus_301()
+        else:
+            self.handleStatus_301()
 
 
     def handleStatus_303(self):
@@ -360,7 +361,7 @@
                  method='GET', postdata=None, headers=None,
                  agent="Twisted client", supportPartial=0,
                  timeout=0, cookies=None, followRedirect=1,
-                 redirectLimit=20):
+                 redirectLimit=20, afterFoundGet=False):
         self.requestedPartial = 0
         if isinstance(fileOrName, types.StringTypes):
             self.fileName = fileOrName
@@ -377,7 +378,8 @@
         HTTPClientFactory.__init__(
             self, url, method=method, postdata=postdata, headers=headers,
             agent=agent, timeout=timeout, cookies=cookies,
-            followRedirect=followRedirect, redirectLimit=redirectLimit)
+            followRedirect=followRedirect, redirectLimit=redirectLimit,
+            afterFoundGet=afterFoundGet)
 
 
     def gotHeaders(self, headers):
