Index: twisted/internet/reactor.py
===================================================================
--- twisted/internet/reactor.py (revision 34322)
+++ twisted/internet/reactor.py (working copy)
@@ -6,10 +6,10 @@
 applications using Twisted. The reactor provides APIs for networking,
 threading, dispatching events, and more.

-The default reactor is based on C{select(2)} and will be installed if this
-module is imported without another reactor being explicitly installed.
-Regardless of which reactor is installed, importing this module is the correct
-way to get a reference to it.
+The default reactor depends on the platform (poll on posix, select on other
+platforms) and will be installed if this module is imported without another
+reactor being explicitly installed. Regardless of which reactor is installed,
+importing this module is the correct way to get a reference to it.

 New application code should prefer to pass and accept the reactor as a
 parameter where it is needed, rather than relying on being able to import this
