Index: twisted/internet/reactor.py
===================================================================
--- twisted/internet/reactor.py (revision 34339)
+++ twisted/internet/reactor.py (working copy)
@@ -6,10 +6,10 @@
 applications using Twisted. The reactor provides APIs for networking,
 threading, dispatching events, and more.

-The default reactor is based on C{select(2)} and will be installed if this
-module is imported without another reactor being explicitly installed.
-Regardless of which reactor is installed, importing this module is the correct
-way to get a reference to it.
+The default reactor depends on the platform (poll on posix, select on other
+platforms) and will be installed if this module is imported without another
+reactor being explicitly installed. Regardless of which reactor is installed,
+importing this module is the correct way to get a reference to it.

 New application code should prefer to pass and accept the reactor as a
 parameter where it is needed, rather than relying on being able to import this
Index: doc/core/howto/choosing-reactor.xhtml
===================================================================
--- doc/core/howto/choosing-reactor.xhtml   (revision 34339)
+++ doc/core/howto/choosing-reactor.xhtml   (working copy)
@@ -80,16 +80,15 @@

     <h3>Select()-based Reactor</h3><a name="select" />

-    <p>The <code>select</code> reactor is currently the default reactor on all
-    platforms. The following code will install it, if no other reactor has
+    <p>The <code>poll</code> reactor is currently the default reactor on posix and
+    <code>select</code> on all other platforms. The following code will install it, if no other reactor has
     been installed:</p>

 <pre class="python">
 from twisted.internet import reactor
 </pre>

-    <p>In the future, if another reactor becomes the default, but the
-    <code>select</code> reactor is desired, it may be installed via:</p>
+    <p>If the <code>select</code> reactor is desired on posix, it may be installed via:</p>

 <pre class="python">
 from twisted.internet import selectreactor
