Index: doc/mail/tutorial/smtpclient/smtpclient-11.tac
===================================================================
--- doc/mail/tutorial/smtpclient/smtpclient-11.tac	(revision 34432)
+++ doc/mail/tutorial/smtpclient/smtpclient-11.tac	(working copy)
@@ -46,7 +46,7 @@
 
 def getMailExchange(host):
     def cbMX(mxRecord):
-        return str(mxRecord.exchange)
+        return str(mxRecord.name)
     return relaymanager.MXCalculator().getMX(host).addCallback(cbMX)
 
 def cbMailExchange(exchange):
Index: doc/mail/tutorial/smtpclient/smtpclient.xhtml
===================================================================
--- doc/mail/tutorial/smtpclient/smtpclient.xhtml	(revision 34432)
+++ doc/mail/tutorial/smtpclient/smtpclient.xhtml	(working copy)
@@ -703,7 +703,7 @@
 <pre class="python">
 def getMailExchange(host):
     def cbMX(mxRecord):
-        return str(mxRecord.exchange)
+        return str(mxRecord.name)
     return relaymanager.MXCalculator().getMX(host).addCallback(cbMX)
 </pre>
 
