--- __init__.py	2013-03-26 00:01:40.470601547 +0530
+++ /home/dhanush/Desktop/__init__.py	2013-04-02 00:17:29.512953285 +0530
@@ -1 +1,8 @@
 "Chat protocols"
+
+#Deprecated twisted.words.protocols.msn
+twisted.python.deprecate.deprecatedModuleAttribute(
+        Version("Twisted", 12, 3, 0),
+        "MSN is being shut down. Client code has no usefulness on its own.",
+        "twisted.words.protocols.msn.py",
+        "msn")
