Index: twisted/words/protocols/__init__.py
===================================================================
--- twisted/words/protocols/__init__.py	(revision 42524)
+++ twisted/words/protocols/__init__.py	(working copy)
@@ -5,3 +5,9 @@
 """
 Chat protocols.
 """
+
+from twisted.python import deprecate, versions
+
+deprecate.deprecatedModuleAttribute(
+    versions.Version("Twisted", 13, 3, 0), "MSN has shutdown.", __name__, 
+    "msn")
\ No newline at end of file
Index: twisted/words/test/test_msn.py
===================================================================
--- twisted/words/test/test_msn.py	(revision 42524)
+++ twisted/words/test/test_msn.py	(working copy)
@@ -513,9 +513,30 @@
         d.addCallback(check)
         return d
 
+
+
+class DeprecationTests(unittest.TestCase):
+    """
+    Test deprecation of L{twisted.words.protocols.msn}
+    """
+    
+    def test_deprecation(self):
+        """
+        Accessing L{twisted.words.protocols.msn} emits a deprecation warning
+        """
+        from twisted.words.protocols import msn
+        warningsShown = self.flushWarnings([self.test_deprecation])
+        self.assertEqual(len(warningsShown), 1)
+        self.assertIdentical(warningsShown[0]['category'], DeprecationWarning)
+        self.assertEqual(
+            warningsShown[0]['message'],
+            'twisted.words.protocols.msn was deprecated in Twisted 13.3.0: ' +
+            'MSN has shutdown.')
+
 if msn is None:
     for testClass in [DispatchTests, PassportTests, NotificationTests,
-                      MessageHandlingTests, FileTransferTestCase]:
+                      MessageHandlingTests, FileTransferTestCase,
+                      DeprecationTests]:
         testClass.skip = (
             "MSN requires an HTTP client but none is available, "
             "skipping tests.")
Index: twisted/words/topfiles/6395.misc
===================================================================
--- twisted/words/topfiles/6395.misc	(revision 0)
+++ twisted/words/topfiles/6395.misc	(working copy)
@@ -0,0 +1 @@
+twisted.words.protocols.msn is now deprecated
\ No newline at end of file
