Index: twisted/words/protocols/msn.py
===================================================================
--- twisted/words/protocols/msn.py	(revision 37801)
+++ twisted/words/protocols/msn.py	(working copy)
@@ -71,6 +71,14 @@
 @author: Sam Jordan
 """
 
+import warnings
+warnings.warn(
+    "As of Twisted 13.0.0, twisted.words.protocols.msn is deprecated.  "
+    "Reason : MSN is being shut down",
+    DeprecationWarning,
+    stacklevel=2)
+
+
 import types, operator, os
 from random import randint
 from urllib import quote, unquote
Index: twisted/words/protocols/__init__.py
===================================================================
--- twisted/words/protocols/__init__.py	(revision 37801)
+++ twisted/words/protocols/__init__.py	(working copy)
@@ -1 +1,6 @@
+# Copyright (c) Twisted Matrix Laboratories.
+# See LICENSE for details.
+
 "Chat protocols"
+
+
Index: twisted/words/topfiles/6395.removal
===================================================================
--- twisted/words/topfiles/6395.removal	(revision 0)
+++ twisted/words/topfiles/6395.removal	(revision 0)
@@ -0,0 +1 @@
+twisted.words.protocols.msn is now deprecated.
Index: twisted/words/test/test_msn.py
===================================================================
--- twisted/words/test/test_msn.py	(revision 37801)
+++ twisted/words/test/test_msn.py	(working copy)
@@ -23,7 +23,7 @@
     # Otherwise importing it should work, so do it.
     from twisted.words.protocols import msn
 
-
+from twisted.trial import unittest
 from twisted.python.hashlib import md5
 from twisted.protocols import loopback
 from twisted.internet.defer import Deferred
@@ -520,3 +520,27 @@
         testClass.skip = (
             "MSN requires an HTTP client but none is available, "
             "skipping tests.")
+
+
+class Deprecations(unittest.TestCase):
+    """
+    Tests to ensure deprecation of L{twisted.words.protocols.msn} 
+    """
+
+    def ensureDeprecated(self, message):
+        """
+        Ensures that the correct deprecation warning was issued.
+        """
+        warnings = self.flushWarnings(self.test_msnDeprecated)
+        self.assertIdentical(warnings[0]['category'], DeprecationWarning)
+        self.assertEqual(warnings[0]['message'], message)
+        self.assertEqual(len(warnings), 1)
+
+
+    def test_msnDeprecated(self):
+        """
+        When importing twisted.words.protocols.msn L{DeprecationWarning} is reported. 
+        """
+        from twisted.words.protocols import msn
+        self.ensureDeprecated("twisted.words.protocols.msn is deprecated "
+                              "since Twisted 13.0.0")
