Index: docs/words/examples/msn_example.py
===================================================================
--- docs/words/examples/msn_example.py	(revision 42524)
+++ docs/words/examples/msn_example.py	(working copy)
@@ -1,70 +0,0 @@
-#!/usr/bin/env python
-
-# Copyright (c) Twisted Matrix Laboratories.
-# See LICENSE for details.
-
-# Twisted Imports
-from twisted.internet import reactor
-from twisted.internet.protocol import ClientFactory
-from twisted.words.protocols import msn
-from twisted.python import log
-
-# System Imports
-import sys, getpass
-
-"""
-This simple example connects to the MSN chat service and
-prints out information about all the users on your
-contact list (both online and offline).
-
-The main aim of this example is to demonstrate
-the connection process.
-
-@author Samuel Jordan
-
-To run the script::
-
-    $ python msn_example.py
-"""
-
-
-def _createNotificationFac():
-    fac = msn.NotificationFactory.forProtocol(Notification)
-    fac.userHandle = USER_HANDLE
-    fac.password = PASSWORD
-    return fac
-
-class Dispatch(msn.DispatchClient):
-
-    def __init__(self):
-        msn.DispatchClient.__init__(self)
-        self.userHandle = USER_HANDLE
-
-    def gotNotificationReferral(self, host, port):
-        self.transport.loseConnection()
-        reactor.connectTCP(host, port, _createNotificationFac())
-
-class Notification(msn.NotificationClient):
-
-    def loginFailure(self, message):
-        print 'Login failure:', message
-
-    def listSynchronized(self, *args):
-        contactList = self.factory.contacts
-        print 'Contact list has been synchronized, number of contacts = %s' % len(contactList.getContacts())
-        for contact in contactList.getContacts().values():
-            print 'Contact: %s' % (contact.screenName,)
-            print '    email: %s' % (contact.userHandle,)
-            print '   groups:'
-            for group in contact.groups:
-                print '      - %s' % contactList.groups[group]
-            print
-
-if __name__ == '__main__':
-    USER_HANDLE = raw_input("Email (passport): ")
-    PASSWORD = getpass.getpass()
-    log.startLogging(sys.stdout)
-    _dummy_fac = ClientFactory()
-    _dummy_fac.protocol = Dispatch
-    reactor.connectTCP('messenger.hotmail.com', 1863, _dummy_fac)
-    reactor.run()
Index: twisted/words/protocols/__init__.py
===================================================================
--- twisted/words/protocols/__init__.py	(revision 42524)
+++ twisted/words/protocols/__init__.py	(working copy)
@@ -5,3 +5,9 @@
 """
 Chat protocols.
 """
+
+from twisted.python import deprecate, versions
+
+deprecate.deprecatedModuleAttribute(
+    versions.Version("Twisted", 14, 1, 0), "MSN has shutdown.", __name__, 
+    "msn")
\ No newline at end of file
Index: twisted/words/test/test_msn.py
===================================================================
--- twisted/words/test/test_msn.py	(revision 42524)
+++ twisted/words/test/test_msn.py	(working copy)
@@ -513,9 +513,30 @@
         d.addCallback(check)
         return d
 
+
+
+class DeprecationTests(unittest.TestCase):
+    """
+    Test deprecation of L{twisted.words.protocols.msn}
+    """
+    
+    def test_deprecation(self):
+        """
+        Accessing L{twisted.words.protocols.msn} emits a deprecation warning
+        """
+        from twisted.words.protocols import msn
+        warningsShown = self.flushWarnings([self.test_deprecation])
+        self.assertEqual(len(warningsShown), 1)
+        self.assertIdentical(warningsShown[0]['category'], DeprecationWarning)
+        self.assertEqual(
+            warningsShown[0]['message'],
+            'twisted.words.protocols.msn was deprecated in Twisted 14.1.0: ' +
+            'MSN has shutdown.')
+
 if msn is None:
     for testClass in [DispatchTests, PassportTests, NotificationTests,
-                      MessageHandlingTests, FileTransferTestCase]:
+                      MessageHandlingTests, FileTransferTestCase,
+                      DeprecationTests]:
         testClass.skip = (
             "MSN requires an HTTP client but none is available, "
             "skipping tests.")
Index: twisted/words/topfiles/6395.misc
===================================================================
--- twisted/words/topfiles/6395.misc	(revision 0)
+++ twisted/words/topfiles/6395.misc	(working copy)
@@ -0,0 +1 @@
+twisted.words.protocols.msn is now deprecated
\ No newline at end of file
