Index: twisted/internet/udp.py
===================================================================
--- twisted/internet/udp.py	(revision 21254)
+++ twisted/internet/udp.py	(working copy)
@@ -129,10 +129,12 @@
                     log.err()
 
 
-    def write(self, datagram, addr=None):
+    def write(self, datagram, addr=None, retries=5):
         """Write a datagram.
 
         @param addr: should be a tuple (ip, port), can be None in connected mode.
+
+        @param retries: number of times to retry an interrupted write
         """
         if self._connectedAddr:
             assert addr in (None, self._connectedAddr)
@@ -141,7 +143,10 @@
             except socket.error, se:
                 no = se.args[0]
                 if no == EINTR:
-                    return self.write(datagram)
+                    # In rare cases, Windows will return EINTR forever.
+                    if retries <= 0:
+                        raise
+                    return self.write(datagram, retries-1)
                 elif no == EMSGSIZE:
                     raise error.MessageLengthError, "message too long"
                 elif no == ECONNREFUSED:
@@ -157,7 +162,9 @@
             except socket.error, se:
                 no = se.args[0]
                 if no == EINTR:
-                    return self.write(datagram, addr)
+                    if retries <= 0:
+                        raise
+                    return self.write(datagram, addr, retries-1)
                 elif no == EMSGSIZE:
                     raise error.MessageLengthError, "message too long"
                 elif no == ECONNREFUSED:
