Index: twisted/topfiles/5784.misc
===================================================================
Index: twisted/scripts/trial.py
===================================================================
--- twisted/scripts/trial.py	(revision 34835)
+++ twisted/scripts/trial.py	(working copy)
@@ -162,8 +162,8 @@
 
     def opt_coverage(self):
         """
-        Generate coverage information in the I{coverage} file in the
-        directory specified by the I{trial-temp} option.
+        Generate coverage information in the coverage file in the
+        directory specified by the temp-directory option.
         """
         import trace
         self.tracer = trace.Trace(count=1, trace=0)
