diff --git a/twisted/internet/test/test_unix.py b/twisted/internet/test/test_unix.py
--- a/twisted/internet/test/test_unix.py
+++ b/twisted/internet/test/test_unix.py
@@ -107,7 +107,7 @@ class UNIXTestsBuilder(UNIXFamilyMixin, 
         reactor = self.buildReactor()
         port = reactor.listenUNIX('\0' + path, ServerFactory())
         self.assertEqual(port.getHost(), UNIXAddress('\0' + path))
-    if platform != 'linux2':
+    if not platform.startswith('linux'):
         test_listenOnLinuxAbstractNamespace.skip = (
             'Abstract namespace UNIX sockets only supported on Linux.')
 
@@ -122,7 +122,7 @@ class UNIXTestsBuilder(UNIXFamilyMixin, 
         connector = reactor.connectUNIX('\0' + path, ClientFactory())
         self.assertEqual(
             connector.getDestination(), UNIXAddress('\0' + path))
-    if platform != 'linux2':
+    if not platform.startswith('linux'):
         test_connectToLinuxAbstractNamespace.skip = (
             'Abstract namespace UNIX sockets only supported on Linux.')
 
@@ -153,7 +153,7 @@ class UNIXDatagramTestsBuilder(UNIXFamil
         reactor = self.buildReactor()
         port = reactor.listenUNIXDatagram('\0' + path, DatagramProtocol())
         self.assertEqual(port.getHost(), UNIXAddress('\0' + path))
-    if platform != 'linux2':
+    if not platform.startswith('linux'):
         test_listenOnLinuxAbstractNamespace.skip = (
             'Abstract namespace UNIX sockets only supported on Linux.')
 
