diff --git a/twisted/internet/base.py b/twisted/internet/base.py
index 772810b..ae6a243 100644
--- a/twisted/internet/base.py
+++ b/twisted/internet/base.py
@@ -1112,6 +1112,7 @@ class _SignalReactorMixin:
 
         if platformType == 'posix':
             signal.signal(signal.SIGCHLD, self._handleSigchld)
+            self._handleSigchld(signal.SIGCHLD, None)
 
 
     def _handleSigchld(self, signum, frame, _threadSupport=platform.supportsThreads()):
