diff --git a/twisted/internet/endpoints.py b/twisted/internet/endpoints.py
index 8aedc05..5799ea0 100644
--- a/twisted/internet/endpoints.py
+++ b/twisted/internet/endpoints.py
@@ -1114,8 +1114,6 @@ def _parseSSL(factory, port, privateKey="server.pem", certKey=None,
     kw = {}
     if sslmethod is not None:
         kw['method'] = getattr(ssl.SSL, sslmethod)
-    else:
-        kw['method'] = ssl.SSL.SSLv23_METHOD
     certPEM = FilePath(certKey).getContent()
     keyPEM = FilePath(privateKey).getContent()
     privateCertificate = ssl.PrivateCertificate.loadPEM(certPEM + keyPEM)
diff --git a/twisted/internet/test/test_endpoints.py b/twisted/internet/test/test_endpoints.py
index b2ce8cc..04d32a2 100644
--- a/twisted/internet/test/test_endpoints.py
+++ b/twisted/internet/test/test_endpoints.py
@@ -58,7 +58,9 @@ try:
     from twisted.internet.ssl import PrivateCertificate, Certificate
     from twisted.internet.ssl import CertificateOptions, KeyPair
     from twisted.internet.ssl import DiffieHellmanParameters
-    from OpenSSL.SSL import ContextType, SSLv23_METHOD, TLSv1_METHOD
+    from OpenSSL.SSL import (
+        ContextType, SSLv23_METHOD, TLSv1_METHOD, OP_NO_SSLv3
+    )
     testCertificate = Certificate.loadPEM(pemPath.getContent())
     testPrivateCertificate = PrivateCertificate.loadPEM(pemPath.getContent())
 
@@ -2505,6 +2507,9 @@ class ServerStringTests(unittest.TestCase):
         self.assertEqual(server._backlog, 50)
         self.assertEqual(server._interface, "")
         self.assertEqual(server._sslContextFactory.method, SSLv23_METHOD)
+        self.assertTrue(
+            server._sslContextFactory._options & OP_NO_SSLv3,
+        )
         ctx = server._sslContextFactory.getContext()
         self.assertIsInstance(ctx, ContextType)
 
diff --git a/twisted/topfiles/7684.bugfix b/twisted/topfiles/7684.bugfix
new file mode 100644
index 0000000..e5dc6a3
--- /dev/null
+++ b/twisted/topfiles/7684.bugfix
@@ -0,0 +1 @@
+The SSL endpoint string parser now results in disallowing SSLv3, which has serious security issues.
