Index: twisted/internet/posixbase.py
===================================================================
--- twisted/internet/posixbase.py	(revision 15539)
+++ twisted/internet/posixbase.py	(working copy)
@@ -14,6 +14,7 @@
 
 import warnings
 import socket
+import errno
 import os
 from zope.interface import implements, classImplements
 
@@ -82,7 +83,11 @@
     def wakeUp(self):
         """Send a byte to my connection.
         """
-        self.w.send('x')
+        try:
+            util.untilConcludes(self.w.send, 'x')
+        except OSError, e:
+            if e.errno != errno.ENOBUFS:
+                raise
 
     def doRead(self):
         """Read some data from my connection.
@@ -125,7 +130,11 @@
         """Write one byte to the pipe, and flush it.
         """
         if self.o is not None:
-            util.untilConcludes(os.write, self.o, 'x')
+            try:
+                util.untilConcludes(os.write, self.o, 'x')
+            except OSError, e:
+                if e.errno != errno.EAGAIN:
+                    raise
 
     def connectionLost(self, reason):
         """Close both ends of my pipe.
