Index: twisted/protocols/ftp.py
===================================================================
--- twisted/protocols/ftp.py	(revision 15239)
+++ twisted/protocols/ftp.py	(working copy)
@@ -836,6 +836,8 @@
         def cbList(results):
             self.reply(DATA_CNX_ALREADY_OPEN_START_XFR)
             for (name, ignored) in results:
+                if name is None:
+                    continue
                 self.dtpInstance.sendLine(name)
             self.dtpInstance.transport.loseConnection()
             return (TXFR_COMPLETE_OK,)
