diff -ur Twisted/twisted/python/test/test_utilpy3.py Twisted-Test/twisted/python/test/test_utilpy3.py
--- Twisted/twisted/python/test/test_utilpy3.py	Sat Oct 27 04:03:05 2012
+++ Twisted-Test/twisted/python/test/test_utilpy3.py	Sat Oct 27 19:49:01 2012
@@ -396,8 +396,16 @@
             ('f', 'F'),
             ('fo', 'Fo'),
             ('foo', 'Foo'),
+            ('F', 'F'),
+            ('FO', 'FO'),
+            ('FOO', 'FOO'),
             ('fooBar', 'Foo Bar'),
             ('fooBarBaz', 'Foo Bar Baz'),
+            ('FooBar', 'Foo Bar'),
+            ('FooBarBaz', 'Foo Bar Baz'),
+            ('FOObar', 'FOO Bar'),
+            ('fooBAR', 'Foo BAR'),
+            ('rmswaveRMSWAVErmswaveRMSWAVE', 'Rmswave RMSWAVE Rmswave RMSWAVE'),
             ]
         for inp, out in nameData:
             got = util.nameToLabel(inp)
