Index: twisted/internet/base.py
===================================================================
--- twisted/internet/base.py	(revision 28467)
+++ twisted/internet/base.py	(working copy)
@@ -614,7 +614,7 @@
     def addSystemEventTrigger(self, _phase, _eventType, _f, *args, **kw):
         """See twisted.internet.interfaces.IReactorCore.addSystemEventTrigger.
         """
-        assert callable(_f), "%s is not callable" % _f
+        assert hasattr(_f,'__call__'), "%s is not callable" % _f
         if _eventType not in self._eventTriggers:
             self._eventTriggers[_eventType] = _ThreePhaseEvent()
         return (_eventType, self._eventTriggers[_eventType].addTrigger(
@@ -671,7 +671,7 @@
     def callLater(self, _seconds, _f, *args, **kw):
         """See twisted.internet.interfaces.IReactorTime.callLater.
         """
-        assert callable(_f), "%s is not callable" % _f
+        assert hasattr(_f,'__call__'), "%s is not callable" % _f
         assert sys.maxint >= _seconds >= 0, \
                "%s is not greater than or equal to 0 seconds" % (_seconds,)
         tple = DelayedCall(self.seconds() + _seconds, _f, args, kw,
@@ -895,7 +895,7 @@
             """
             See L{twisted.internet.interfaces.IReactorThreads.callFromThread}.
             """
-            assert callable(f), "%s is not callable" % (f,)
+            assert hasattr(f,'__call__'), "%s is not callable" % (f,)
             # lists are thread-safe in CPython, but not in Jython
             # this is probably a bug in Jython, but until fixed this code
             # won't work in Jython.
