Index: pb.py
===================================================================
RCS file: /cvs/Twisted/twisted/spread/pb.py,v
retrieving revision 1.150
diff -u -r1.150 pb.py
--- pb.py	12 Oct 2003 16:19:50 -0000	1.150
+++ pb.py	17 Oct 2003 06:17:37 -0000
@@ -370,7 +370,7 @@
         self.__init__(unjellier.invoker.unserializingPerspective, unjellier.invoker, unjellyList[1], 1)
         return self
 
-    def callRemote(self, name, *args, **kw):
+    def callRemote(self, _name, *args, **kw):
         """Asynchronously invoke a remote method.
 
         @type name:   C{string}
@@ -381,7 +381,7 @@
         @returns: a Deferred which will be fired when the result of
                   this remote call is received.
         """
-        return self.broker._sendMessage('',self.perspective, self.luid, name, args, kw)
+        return self.broker._sendMessage('',self.perspective, self.luid, _name, args, kw)
 
     def remoteMethod(self, key):
         """Get a L{RemoteMethod} for this key.
