--- client.py_original	2009-12-03 15:01:03.000000000 +0100
+++ client.py	2009-12-03 15:01:27.000000000 +0100
@@ -243,6 +243,8 @@
             self.queryTCP(q, t).chainDeferred(d)
         del self.pending[:]
 
+    def connectionLost(self, protocol):
+        pass
 
     def messageReceived(self, message, protocol, address = None):
         log.msg("Unexpected message (%d) received from %r" % (message.id, address))
