=== modified file 'twisted/internet/_glibbase.py'
--- old/twisted/internet/_glibbase.py	2013-01-16 17:44:26 +0000
+++ new/twisted/internet/_glibbase.py	2013-03-14 21:14:30 +0000
@@ -158,9 +158,17 @@ class GlibReactorBase(posixbase.PosixRea
         else:
             fileno = source
             wrapper = callback
-        return self._glib.io_add_watch(
-            fileno, condition, wrapper,
-            priority=self._glib.PRIORITY_DEFAULT_IDLE)
+        try:
+            from gi import version_info as gi_version
+        except ImportError:
+            gi_version = (2, 32, 0)
+        if gi_version < (3, 7, 3):
+            return self._glib.io_add_watch(
+                fileno, condition, wrapper,
+                priority=self._glib.PRIORITY_DEFAULT_IDLE)
+        else:
+            return self._glib.io_add_watch(
+                fileno, self._glib.PRIORITY_DEFAULT, condition, wrapper)
 
 
     def _ioEventCallback(self, source, condition):

