Index: twisted/internet/epollreactor.py
===================================================================
--- twisted/internet/epollreactor.py	(revision 33404)
+++ twisted/internet/epollreactor.py	(working copy)
@@ -58,8 +58,9 @@
         Initialize epoll object, file descriptor tracking dictionaries, and the
         base class.
         """
-        # Create the poller we're going to use.  The 1024 here is just a hint
-        # to the kernel, it is not a hard maximum.
+        # Create the poller we're going to use.  The 1024 here is just a hint to
+        # the kernel, it is not a hard maximum.  After Linux 2.6.8, the size
+        # argument is completely ignored.
         self._poller = _epoll.epoll(1024)
         self._reads = {}
         self._writes = {}
