Index: trial-test-discovery/twisted/python/reflect.py
===================================================================
--- trial-test-discovery/twisted/python/reflect.py   (revision 27382)
+++ trial-test-discovery/twisted/python/reflect.py   (working copy)
@@ -619,7 +619,7 @@
         for name in classObj.__dict__:
             method = getattr(classObj, name)
             optName = name[len(prefix):]
-            if (hasattr(method, "__call__") and hasattr(method, "im_class")
+            if (hasattr(method, "__call__")
                 and (name[:len(prefix)] == prefix)
                 and (len(optName))):
                 dict[optName] = 1
Index: trial-test-discovery/twisted/test/test_reflect.py
===================================================================
--- trial-test-discovery/twisted/test/test_reflect.py   (revision 27382)
+++ trial-test-discovery/twisted/test/test_reflect.py   (working copy)
@@ -824,8 +824,8 @@
     Test for L{reflect.addMethodNamesToDict}.
     """
     non_base = object()
-    groups = ['def', 'fdecor', 'lambda', 'class1', 'class2', 'class3', 'methdesc']
-    fail_groups = ['decor', 'desc', 'builtin']
+    groups = ['def', 'fdecor', 'lambda', 'class1', 'class2', 'class3', 'methdesc', 'decor', 'desc', 'builtin']
+    fail_groups = []
     prefix = "prefix_"
     postfix = "_method"
 
Index: trial-test-discovery/twisted/trial/unittest.py
===================================================================
--- trial-test-discovery/twisted/trial/unittest.py   (revision 27382)
+++ trial-test-discovery/twisted/trial/unittest.py   (working copy)
@@ -695,7 +695,7 @@
         self._testMethodName = methodName
         testMethod = getattr(self, methodName)
         self._parents = [testMethod, self]
-        self._parents.extend(util.getPythonContainers(testMethod))
+        self._parents.extend(util.getPythonContainers(testMethod, type(self)))
         self._shared = (hasattr(self, 'setUpClass') or
                         hasattr(self, 'tearDownClass'))
         if self._shared:
Index: trial-test-discovery/twisted/trial/util.py
===================================================================
--- trial-test-discovery/twisted/trial/util.py   (revision 27382)
+++ trial-test-discovery/twisted/trial/util.py   (working copy)
@@ -20,7 +20,7 @@
 from twisted.internet import defer, utils, interfaces
 from twisted.python.failure import Failure
 
-
+DUMMY = object()
 DEFAULT_TIMEOUT = object()
 DEFAULT_TIMEOUT_DURATION = 120.0
 
@@ -260,12 +260,14 @@
     return _
 
 
-def getPythonContainers(meth):
+def getPythonContainers(meth, cls=DUMMY):
     """Walk up the Python tree from method 'meth', finding its class, its module
     and all containing packages."""
+    if cls == DUMMY:
+        cls = meth.im_class
     containers = []
-    containers.append(meth.im_class)
-    moduleName = meth.im_class.__module__
+    containers.append(cls)
+    moduleName = cls.__module__
     while moduleName is not None:
         module = sys.modules.get(moduleName, None)
         if module is None:
