--- http.py	2010-08-04 12:22:44.980139175 +0200
+++ http_patch.py	2010-08-04 12:23:58.073465954 +0200
@@ -1564,7 +1564,7 @@
             self._command = command
             self._path = request
             self._version = version
-        elif line == '':
+        elif line == '' or '\n' in line:
             if self.__header:
                 self.headerReceived(self.__header)
             self.__header = ''
