diff --git a/twisted/scripts/tap2rpm.py b/twisted/scripts/tap2rpm.py
index 1132fe9..f3d2e0d 100755
--- a/twisted/scripts/tap2rpm.py
+++ b/twisted/scripts/tap2rpm.py
@@ -266,6 +266,8 @@ def run(options=None):
 
     #  build rpm
     job = subprocess.Popen([
+            "strace",
+            "-s", "1000",
             "rpmbuild",
             "-vv",
             "--define", "_topdir %s" % (tmp_dir,),
@@ -274,8 +276,9 @@ def run(options=None):
     stdout, _ = job.communicate()
 
     # If there was a problem, show people what it was.
-    if job.returncode != 0:
-        print stdout
+    #if job.returncode != 0:
+    #    print stdout
+    print stdout
     
     #  copy the RPMs to the local directory
     log.msg(os.listdir(tmp_dir))
