diff --git a/twisted/scripts/tap2rpm.py b/twisted/scripts/tap2rpm.py
index 96f6720..68e9cea 100755
--- a/twisted/scripts/tap2rpm.py
+++ b/twisted/scripts/tap2rpm.py
@@ -237,9 +237,11 @@ def run(options=None):
     #  build rpm
     job = subprocess.Popen([
             "rpmbuild",
+            "-vv",
             "--define", "_topdir %s" % (tmp_dir,),
             "-ta", tarfile_name,
-        ], stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+        ])
+    #    ], stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
     stdout, _ = job.communicate()
 
     # If there was a problem, show people what it was.
