Index: twisted/topfiles/4076.removal
===================================================================
--- twisted/topfiles/4076.removal	(revision 0)
+++ twisted/topfiles/4076.removal	(revision 0)
@@ -0,0 +1,3 @@
+twisted.internet.interfaces.IReactorTime.cancelCallLater (deprecated since 
+2007) and twisted.internet.interfaces.base.ReactorBase.cancelCallLater 
+(deprecated since 2002) have been removed.
\ No newline at end of file
Index: twisted/internet/interfaces.py
===================================================================
--- twisted/internet/interfaces.py	(revision 29213)
+++ twisted/internet/interfaces.py	(working copy)
@@ -565,18 +565,7 @@
                  C{reset()} methods.
         """
 
-    def cancelCallLater(callID):
-        """
-        This method is deprecated.
 
-        Cancel a call that would happen later.
-
-        @param callID: this is an opaque identifier returned from C{callLater}
-                       that will be used to cancel a specific call.
-
-        @raise ValueError: if the callID is not recognized.
-        """
-
     def getDelayedCalls():
         """
         Retrieve all currently scheduled delayed calls.
Index: twisted/internet/base.py
===================================================================
--- twisted/internet/base.py	(revision 29213)
+++ twisted/internet/base.py	(working copy)
@@ -721,13 +721,6 @@
     def _cancelCallLater(self, tple):
         self._cancellations+=1
 
-    def cancelCallLater(self, callID):
-        """See twisted.internet.interfaces.IReactorTime.cancelCallLater.
-        """
-        # DO NOT DELETE THIS - this is documented in Python in a Nutshell, so we
-        # we can't get rid of it for a long time.
-        warnings.warn("reactor.cancelCallLater(callID) is deprecated - use callID.cancel() instead")
-        callID.cancel()
 
     def getDelayedCalls(self):
         """Return all the outstanding delayed calls in the system.
