diff --git a/src/twisted/internet/iocpreactor/abstract.py b/src/twisted/internet/iocpreactor/abstract.py
index 1cae9f7..8442795 100644
--- a/src/twisted/internet/iocpreactor/abstract.py
+++ b/src/twisted/internet/iocpreactor/abstract.py
@@ -188,7 +188,7 @@ class FileHandle(_ConsumerMixin, _LogOwner):
 
 
     def _cbWrite(self, rc, numBytesWritten, evt):
-        if self._handleWrite(rc, numBytesWritten, evt):
+        if self._handleWrite(rc, numBytesWritten, evt) and self._writeScheduled is None:
             self.doWrite()
 
 
