diff --git twisted/spread/banana.py twisted/spread/banana.py
index edae9c6..82b7256 100644
--- twisted/spread/banana.py
+++ twisted/spread/banana.py
@@ -79,6 +79,8 @@ def setPrefixLimit(limit):
     """
     global _PREFIX_LIMIT
     _PREFIX_LIMIT = limit
+
+_PREFIX_LIMIT = None
 setPrefixLimit(64)
 
 SIZE_LIMIT = 640 * 1024   # 640k is all you'll ever need :-)
diff --git twisted/spread/ui/tktree.py twisted/spread/ui/tktree.py
index 8fbe462..15d1a91 100644
--- twisted/spread/ui/tktree.py
+++ twisted/spread/ui/tktree.py
@@ -16,7 +16,17 @@ What I want it to look like:
 """
 
 import os
-from Tkinter import *
+from Tkinter import (
+    END,
+    Listbox,
+    Tk,
+    Scrollbar,
+    LEFT,
+    BOTH,
+    RIGHT,
+    Y,
+    )
+
 
 class Node:
     def __init__(self):
diff --git twisted/spread/ui/tkutil.py twisted/spread/ui/tkutil.py
index 2aee67d..ac306ec 100644
--- twisted/spread/ui/tkutil.py
+++ twisted/spread/ui/tkutil.py
@@ -4,11 +4,27 @@
 
 """Utilities for building L{PB<twisted.spread.pb>} clients with L{Tkinter}.
 """
-from Tkinter import *
+from Tkinter import (
+    ACTIVE,
+    Button,
+    Canvas,
+    E,
+    END,
+    Entry,
+    Frame,
+    Label,
+    LEFT,
+    Listbox,
+    mainloop,
+    N,
+    S,
+    StringVar,
+    Toplevel,
+    Tk,
+    W,)
 from tkSimpleDialog import _QueryString
 from tkFileDialog import _Dialog
 from twisted.spread import pb
-from twisted.internet import reactor
 from twisted import copyright
 
 import string
