Index: twisted/internet/process.py
===================================================================
--- twisted/internet/process.py	(revision 38615)
+++ twisted/internet/process.py	(working copy)
@@ -676,7 +676,8 @@
 
             self._fork(path, uid, gid, executable, args, environment, fdmap=fdmap)
         except:
-            map(os.close, _openedPipes)
+            for p in _openedPipes:
+                os.close(p)
             raise
 
         # we are the parent process:
