--- twisted/internet/tcp.py	2008-12-10 15:34:01.000000000 +0000
+++ twisted/internet/tcp.py	2008-12-10 15:36:51.000000000 +0000
@@ -524,6 +524,8 @@
 
         err = self.socket.getsockopt(socket.SOL_SOCKET, socket.SO_ERROR)
         if err:
+            del self.doWrite
+            del self.doRead
             self.failIfNotConnected(error.getConnectError((err, strerror(err))))
             return
 
@@ -550,6 +552,8 @@
                 self.startWriting()
                 return
             else:
+                del self.doWrite
+                del self.doRead
                 self.failIfNotConnected(error.getConnectError((connectResult, strerror(connectResult))))
                 return
 
