Index: twisted/internet/_pollingfile.py
===================================================================
--- twisted/internet/_pollingfile.py	(wersja 27397)
+++ twisted/internet/_pollingfile.py	(kopia robocza)
@@ -249,6 +249,8 @@
         while self.outQueue:
             data = self.outQueue.pop(0)
             errCode = 0
+            if type(data) is unicode:
+                raise TypeError, 'unicode not allowed'
             try:
                 errCode, nBytesWritten = win32file.WriteFile(self.writePipe,
                                                              data, None)
Index: twisted/internet/test/test__pollingfile.py
===================================================================
--- twisted/internet/test/test__pollingfile.py	(wersja 0)
+++ twisted/internet/test/test__pollingfile.py	(wersja 0)
@@ -0,0 +1,22 @@
+# Copyright (c) 2009 Twisted Matrix Laboratories.
+# See LICENSE for details.
+
+"""
+Tests for L{twisted.internet._pollingfile}.
+"""
+
+from twisted.internet import _pollingfile
+from twisted.trial.unittest import TestCase
+
+class TestPollableWritePipe(TestCase):
+    def test_checkWork(self):
+
+        def _():
+            pass
+
+        p = _pollingfile._PollableWritePipe(1, _)
+        p.write('test')
+        p.checkWork()
+
+        p.write(u'test')
+        self.assertRaises(TypeError, p.checkWork)
