Index: _pollingfile.py
===================================================================
--- _pollingfile.py	(wersja 27183)
+++ _pollingfile.py	(kopia robocza)
@@ -248,7 +248,8 @@
                 return numBytesWritten
         while self.outQueue:
             data = self.outQueue.pop(0)
-            errCode = 0
+            errCode = 0
+            assert(type(data) != unicode), 'unicode not allowed'            
             try:
                 errCode, nBytesWritten = win32file.WriteFile(self.writePipe,
                                                              data, None)
