Index: twisted/protocols/ident.py
===================================================================
--- twisted/protocols/ident.py	(revision 28819)
+++ twisted/protocols/ident.py	(working copy)
@@ -89,8 +89,8 @@
         self.transport.loseConnection()
 
     def validQuery(self, portOnServer, portOnClient):
-        serverAddr = self.transport.getHost()[1], portOnServer
-        clientAddr = self.transport.getPeer()[1], portOnClient
+        serverAddr = self.transport.getHost().host, portOnServer
+        clientAddr = self.transport.getPeer().host, portOnClient
         defer.maybeDeferred(self.lookup, serverAddr, clientAddr
             ).addCallback(self._cbLookup, portOnServer, portOnClient
             ).addErrback(self._ebLookup, portOnServer, portOnClient
