Index: twisted/application/app.py
===================================================================
--- twisted/application/app.py	(revision 28956)
+++ twisted/application/app.py	(working copy)
@@ -523,15 +523,6 @@
 
 
 
-def reportProfile(report_profile, name):
-    """
-    DEPRECATED since Twisted 8.0. This does nothing.
-    """
-    warnings.warn("reportProfile is deprecated and a no-op since Twisted 8.0.",
-                  category=DeprecationWarning)
-
-
-
 def _reactorZshAction():
     return "(%s)" % " ".join([r.shortName for r in reactors.getReactorTypes()])
 
