Index: twisted/topfiles/5679.bugfix
===================================================================
--- twisted/topfiles/5679.bugfix	(revision 0)
+++ twisted/topfiles/5679.bugfix	(revision 0)
@@ -0,0 +1 @@
+twisted.internet.test.test_process.test_timeleyProcessExited is now renamed test_timelyProcessExited.
Index: twisted/internet/test/test_process.py
===================================================================
--- twisted/internet/test/test_process.py	(revision 34534)
+++ twisted/internet/test/test_process.py	(working copy)
@@ -384,7 +384,7 @@
         self.runReactor(reactor)
 
 
-    def test_timeleyProcessExited(self):
+    def test_timelyProcessExited(self):
         """
         If a spawned process exits, C{processExited} will be called in a
         timely manner.
