--- Twisted-1.0.7/MANIFEST.in.orig	2003-09-16 21:16:00.000000000 -0600
+++ Twisted-1.0.7/MANIFEST.in	2003-09-16 21:30:58.000000000 -0600
@@ -27,8 +27,11 @@
 recursive-include bin *
 recursive-include doc *.html
 recursive-include doc/img *
-recursive-include pyunit *.py
 recursive-include static *
 recursive-include twisted *.h
+recursive-include twisted *.c
+recursive-include twisted/web/woven *
+recursive-include twisted/lore *
+recursive-include twisted/test *
 
 global-exclude */CVS/*
