Index: twisted/application/app.py
===================================================================
--- twisted/application/app.py	(revision 28956)
+++ twisted/application/app.py	(working copy)
@@ -187,19 +187,6 @@
 
 
 
-def runWithProfiler(reactor, config):
-    """
-    DEPRECATED in Twisted 8.0.
-
-    Run reactor under standard profiler.
-    """
-    warnings.warn("runWithProfiler is deprecated since Twisted 8.0. "
-                  "Use ProfileRunner instead.", DeprecationWarning, 2)
-    item = AppProfiler(config)
-    return item.run(reactor)
-
-
-
 def runWithHotshot(reactor, config):
     """
     DEPRECATED in Twisted 8.0.
@@ -334,8 +321,6 @@
                 # Backward compatible code
                 if not config['nothotshot']:
                     runWithHotshot(reactor, config)
-                else:
-                    runWithProfiler(reactor, config)
         elif config['debug']:
             sys.stdout = oldstdout
             sys.stderr = oldstderr
