diff --git a/twisted/conch/ui/tkvt100.py b/twisted/conch/ui/tkvt100.py
index cd7581d..44b78ea 100644
--- a/twisted/conch/ui/tkvt100.py
+++ b/twisted/conch/ui/tkvt100.py
@@ -13,7 +13,7 @@ import ansi
 import string
 
 ttyFont = None#tkFont.Font(family = 'Courier', size = 10)
-fontWidth, fontHeight = None,None#max(map(ttyFont.measure, string.letters+string.digits)), int(ttyFont.metrics()['linespace'])
+fontWidth, fontHeight = None,None#max(map(ttyFont.measure, string.ascii_letters+string.digits)), int(ttyFont.metrics()['linespace'])
 
 colorKeys = (
     'b', 'r', 'g', 'y', 'l', 'm', 'c', 'w',
@@ -31,7 +31,7 @@ class VT100Frame(Tkinter.Frame):
     def __init__(self, *args, **kw):
         global ttyFont, fontHeight, fontWidth
         ttyFont = tkFont.Font(family = 'Courier', size = 10)
-        fontWidth, fontHeight = max(map(ttyFont.measure, string.letters+string.digits)), int(ttyFont.metrics()['linespace'])
+        fontWidth, fontHeight = max(map(ttyFont.measure, string.ascii_letters+string.digits)), int(ttyFont.metrics()['linespace'])
         self.width = kw.get('width', 80)
         self.height = kw.get('height', 25)
         self.callback = kw['callback']
