--- Twisted-2.4.0.orig/protocols/ftp.py	2006-04-10 05:21:33.000000000 +0200
+++ Twisted-2.4.0.patched/protocols/ftp.py	2006-06-05 20:04:05.000000000 +0200
@@ -1719,11 +1719,11 @@
     def write(self, data):
         self.transport.write(data)
 
-    def registerProducer(self):
-        pass
+    def registerProducer(self, producer, streaming):
+        self.transport.registerProducer(producer, streaming)
 
     def unregisterProducer(self):
-        pass
+        self.transport.unregisterProducer()
 
     def finish(self):
         self.transport.loseConnection()
