Index: twisted/python/_initgroups.c
===================================================================
--- twisted/python/_initgroups.c	(Revision 28089)
+++ twisted/python/_initgroups.c	(Arbeitskopie)
@@ -58,9 +58,28 @@
 
 #endif /* defined(__unix__) || defined(unix) */
 
+#if PY_MAJOR_VERSION < 3
 void
 init_initgroups(void)
 {
 	Py_InitModule("_initgroups", InitgroupsMethods);
 }
+#else
+static struct PyModuleDef _module = {
+	PyModuleDef_HEAD_INIT,
+	"_initgroups",
+	NULL,
+	-1,
+	InitgroupsMethods,
+	NULL,
+	NULL,
+	NULL,
+	NULL
+};
 
+PyMODINIT_FUNC
+PyInit__initgroups(void)
+{
+    return PyModule_Create(&_module);
+}
+#endif
