Index: _win32stdio.py
===================================================================
--- _win32stdio.py	(revision 27397)
+++ _win32stdio.py	(working copy)
@@ -10,6 +10,8 @@
 
 from twisted.internet import _pollingfile, main
 
+from twisted.python.failure import Failure
+
 class Win32PipeAddress(object):
     implements(IAddress)
 
@@ -70,7 +72,7 @@
         if self.connsLost >= 2:
             self.disconnecting = True
             self.disconnected = True
-            self.proto.connectionLost(main.CONNECTION_DONE)
+            self.proto.connectionLost(Failure(main.CONNECTION_DONE))
 
     # ITransport
 
