Index: twisted/test/test_amp.py
===================================================================
--- twisted/test/test_amp.py	(wersja 27170)
+++ twisted/test/test_amp.py	(kopia robocza)
@@ -861,6 +861,14 @@
         self.assertIdentical(self.stopReason.value.keyName, None)
 
 
+    def test_passException(self):
+        """
+        Test if L{amp.BinaryBoxProtocol}.connectionLost accepts Exceptions
+        """
+        protocol = amp.BinaryBoxProtocol(self)
+        protocol.connectionLost(error.ConnectionDone("simulated connection done"))
+
+
     def test_receiveBoxData(self):
         """
         When a binary box protocol receives the serialized form of an AMP box,
Index: twisted/protocols/amp.py
===================================================================
--- twisted/protocols/amp.py	(wersja 27170)
+++ twisted/protocols/amp.py	(kopia robocza)
@@ -1989,7 +1989,7 @@
                 self.innerProtocolClientFactory.clientConnectionLost(None, reason)
         if self._keyLengthLimitExceeded:
             failReason = Failure(TooLong(True, False, None, None))
-        elif reason.check(ConnectionClosed) and self._justStartedTLS:
+        elif hasattr(reason, 'check') and reason.check(ConnectionClosed) and self._justStartedTLS:
             # We just started TLS and haven't received any data.  This means
             # the other connection didn't like our cert (although they may not
             # have told us why - later Twisted should make 'reason' into a TLS
