=== modified file 'twisted/web/xmlrpc.py'
--- twisted/web/xmlrpc.py	2008-04-02 02:10:22 +0000
+++ twisted/web/xmlrpc.py	2008-04-03 05:28:38 +0000
@@ -314,16 +314,16 @@
         try:
             response = xmlrpclib.loads(contents)
         except:
-            self.deferred.errback(failure.Failure())
-            self.deferred = None
+            deferred, self.deferred = self.deferred, None
+            deferred.errback(failure.Failure())
         else:
-            self.deferred.callback(response[0][0])
-            self.deferred = None
+            deferred, self.deferred = self.deferred, None
+            deferred.callback(response[0][0])
 
     def clientConnectionLost(self, _, reason):
         if self.deferred is not None:
-            self.deferred.errback(reason)
-            self.deferred = None
+            deferred, self.deferred = self.deferred, None
+            deferred.errback(reason)
 
     clientConnectionFailed = clientConnectionLost
 

