--- domish.py	2004-03-03 15:58:16.000000000 -0800
+++ /cygdrive/c/Python22/Lib/site-packages/twisted/xish/domish.py	2004-03-03 16:05:47.107579200 -0800
@@ -87,7 +87,7 @@
         for k,v in elem.attributes.items():
             # If the attribute name is a list, it's a qualified attribute
             if isinstance(k, types.TupleType):
-                write(" %s:%s='%s'" % (self.getPrefix[k[0]], k[1], escapeToXml(v, 1)).encode("utf-8"))
+                write((" %s:%s='%s'" % (self.getPrefix(k[0]), k[1], escapeToXml(v, 1))).encode("utf-8"))
             else:
                 write((" %s='%s'" % ( k, escapeToXml(v, 1))).encode("utf-8"))
 
@@ -599,7 +599,7 @@
         # Process attributes
         for k, v in attrs.items():
             if k.find(" ") != -1:
-                attrs[k.split(" ")] = v
+                attrs[tuple(k.split(" "))] = v
                 del attrs[k]
 
         # Construct the new element
