diff --git a/twisted/python/deprecate.py b/twisted/python/deprecate.py
index f4b31b4..9136fe5 100644
--- a/twisted/python/deprecate.py
+++ b/twisted/python/deprecate.py
@@ -528,7 +528,4 @@ def warnAboutFunction(offender, warningString):
         registry=globals.setdefault("__warningregistry__", {}),
         module_globals=None)
 
-    if sys.version_info[:2] < (2, 5):
-        kwargs.pop('module_globals')
-
     warn_explicit(warningString, **kwargs)
