diff --git a/doc/core/howto/row.xhtml b/doc/core/howto/row.xhtml
index 46cc7f0..4fe1506 100644
--- a/doc/core/howto/row.xhtml
+++ b/doc/core/howto/row.xhtml
@@ -41,8 +41,9 @@ database, and slightly less suited to added database persistance
 to an existing python application.</p>
 
 <p><em>If row does not fit your model, you will be best off using
-the <a href="rdbms.xhtml">low-level database API</a> directly, 
-or writing your own object/relational layer on top of it.</em></p>
+the </em> <a href="rdbms.xhtml">low-level database API</a> <em>
+directly, or writing your own object/relational layer on top of
+it.</em></p>
 
 <h2>Class Definitions</h2>
 
@@ -181,14 +182,15 @@ from being changed by mistake.</p>
 <h2>Relationships Between Tables</h2>
 
 <p>Specifying a foreign key for a RowClass creates a relationship
-between database tables. When <code
-class="python">loadObjectsFrom</code> is called for a table, it will
-automatically load all the children rows for the rows from the specified
-table. The child rows will be put into a list member variable of the
-rowObject instance with the name <code>childRows</code> or if a
-<em>containerMethod</em> is specified for the foreign key relationship,
-that method will be called on the parent row object for each row that is
-being added to it as a child.</p>
+between database
+tables. When <code class="python">loadObjectsFrom</code> is called for
+a table, it will automatically load all the children rows for the rows
+from the specified table. The child rows will be put into a list
+member variable of the rowObject instance with the
+name <code>childRows</code> or if a <em>containerMethod</em> is
+specified for the foreign key relationship, that method will be called
+on the parent row object for each row that is being added to it as a
+child.</p>
 
 <p>The <em>autoLoad</em> member of the foreign key definition is a flag
 that specifies whether child rows should be auto-loaded for that
