diff --git a/doc/core/howto/design.xhtml b/doc/core/howto/design.xhtml
index 8d7d144..73e1505 100644
--- a/doc/core/howto/design.xhtml
+++ b/doc/core/howto/design.xhtml
@@ -89,23 +89,26 @@ client connects to this server, a <code class="python">QOTD</code> instance is
 created, and its <code class="python">connectionMade</code> method is called.
 </p>
 
-<p> The <code class="python">QOTDFactory</code>'s role is to specify to the
-Twisted framework how to create a <code class="python">Protocol</code> instance
-that will handle the connection.  Twisted will not instantiate a <code
-class="python">QOTDFactory</code>; you will do that yourself later, in a
-<code class="shell">twistd</code> plug-in.
+<p> The <code class="python">QOTDFactory</code>'s role is to specify
+to the Twisted framework how to create
+a <code class="python">Protocol</code> instance that will handle the
+connection.  Twisted will not instantiate
+a <code class="python">QOTDFactory</code>; you will do that yourself
+later, in a <code class="shell">twistd</code> plug-in.
 </p>
 
-<p>Note: you can read more specifics of <code class="python">Protocol</code> and
-<code class="python">Factory</code> in the <a href="servers.xhtml">Writing
-Servers</a> HOWTO.</p>
-
-<p>Once we have an abstraction -- a <code>Quoter</code> -- and we have a
-mechanism to connect it to the network -- the <code>QOTD</code> protocol -- the
-next thing to do is to put the last link in the chain of functionality between
-abstraction and user.  This last link will allow a user to choose a
-<code>Quoter</code> and configure the protocol. Writing this configuration is
-covered in the <a href="application.xhtml">Application HOWTO</a>.</p>
+<p>Note: you can read more specifics
+of <code class="python">Protocol</code>
+and <code class="python">Factory</code> in
+the <a href="servers.xhtml">Writing Servers</a> HOWTO.</p>
+
+<p>Once we have an abstraction -- a <code>Quoter</code> -- and we have
+a mechanism to connect it to the network -- the <code>QOTD</code>
+protocol -- the next thing to do is to put the last link in the chain
+of functionality between abstraction and user.  This last link will
+allow a user to choose a <code>Quoter</code> and configure the
+protocol. Writing this configuration is covered in
+the <a href="application.xhtml">Application HOWTO</a>.</p>
 
 </body>
 
diff --git a/doc/core/howto/internet-overview.xhtml b/doc/core/howto/internet-overview.xhtml
index 05a8766..46725e1 100644
--- a/doc/core/howto/internet-overview.xhtml
+++ b/doc/core/howto/internet-overview.xhtml
@@ -19,13 +19,13 @@ toolkits like GTK+ or Tk.</p>
 <p>Twisted Internet contains the various interfaces to the reactor
 API, whose usage is documented in the low-level chapter. Those APIs
 are <code class="API" base="twisted.internet.interfaces">IReactorCore</code>,
-<code class="API" base="twisted.internet.interfaces">IReactorTCP</code>,
-<code class="API" base="twisted.internet.interfaces">IReactorSSL</code>,
-<code class="API" base="twisted.internet.interfaces">IReactorUNIX</code>,
-<code class="API" base="twisted.internet.interfaces">IReactorUDP</code>,
-<code class="API" base="twisted.internet.interfaces">IReactorTime</code>,
-<code class="API" base="twisted.internet.interfaces">IReactorProcess</code>,
-<code class="API" base="twisted.internet.interfaces">IReactorMulticast</code>
+ <code class="API" base="twisted.internet.interfaces">IReactorTCP</code>,
+ <code class="API" base="twisted.internet.interfaces">IReactorSSL</code>,
+ <code class="API" base="twisted.internet.interfaces">IReactorUNIX</code>,
+ <code class="API" base="twisted.internet.interfaces">IReactorUDP</code>,
+ <code class="API" base="twisted.internet.interfaces">IReactorTime</code>,
+ <code class="API" base="twisted.internet.interfaces">IReactorProcess</code>,
+ <code class="API" base="twisted.internet.interfaces">IReactorMulticast</code>
 and <code class="API" base="twisted.internet.interfaces">IReactorThreads</code>.
 The reactor APIs allow non-persistent calls to be made.</p>
 
diff --git a/doc/core/howto/quotes.xhtml b/doc/core/howto/quotes.xhtml
index 4c7d1fd..735e88d 100644
--- a/doc/core/howto/quotes.xhtml
+++ b/doc/core/howto/quotes.xhtml
@@ -36,13 +36,13 @@ following:</p>
     <li><a href="listings/TwistedQuotes/quoteproto.py" class="py-listing">quoteproto.py</a>;</li>
     </ul>
 </li>
-<li>Add the <code>TwistedQuotes</code> directory's <em>parent</em> to your Python
-path. For example, if the TwistedQuotes directory's path is
-<code>/tmp/TwistedQuotes</code>
-add <code>/tmp</code> to your Python path. On UNIX this would be <code
-class="shell">export PYTHONPATH=/my/stuff:$PYTHONPATH</code>, on Microsoft
-Windows change the <code class="shell">PYTHONPATH</code> variable through the
-Systems Properites dialog to add <code class="shell">/my/stuff;</code> at the
+<li>Add the <code>TwistedQuotes</code> directory's <em>parent</em> to
+your Python path. For example, if the TwistedQuotes directory's path
+is <code>/tmp/TwistedQuotes</code> add <code>/tmp</code> to your
+Python path. On UNIX this would be <code class="shell">export
+PYTHONPATH=/my/stuff:$PYTHONPATH</code>, on Microsoft Windows change
+the <code class="shell">PYTHONPATH</code> variable through the Systems
+Properites dialog to add <code class="shell">/my/stuff;</code> at the
 beginning.</li>
 <li>
 Test your package by trying to import it in the Python interpreter:
