diff --git twisted/internet/test/test_endpoints.py twisted/internet/test/test_endpoints.py
index 3ae009b..55f80c7 100644
--- twisted/internet/test/test_endpoints.py
+++ twisted/internet/test/test_endpoints.py
@@ -1882,7 +1882,10 @@ class ServerStringTests(unittest.TestCase):
             ValueError,
             endpoints.serverFromString,
             object(),
-            self.SSL_CHAIN_TEMPLATE % (escapedPEMPathName, fp.path,)
+            self.SSL_CHAIN_TEMPLATE % (
+                escapedPEMPathName,
+                endpoints.quoteStringArgument(fp.path),
+            )
         )
 
 
