diff --git twisted/internet/endpoints.py twisted/internet/endpoints.py
index e94a839..eb37950 100644
--- twisted/internet/endpoints.py
+++ twisted/internet/endpoints.py
@@ -935,7 +935,7 @@ def _parseSSL(factory, port, privateKey="server.pem", certKey=None,
 
     @param extraCertChain: The path of a file containing one or more
         certificates in PEM format that establish the chain from a root CA to
-        the CA that signed your L{certKey}.
+        the CA that signed your C{certKey}.
     @type extraCertChain: L{bytes}
 
     @return: a 2-tuple of (args, kwargs), describing  the parameters to
diff --git twisted/internet/test/test_endpoints.py twisted/internet/test/test_endpoints.py
index 1504421..a8e83a2 100644
--- twisted/internet/test/test_endpoints.py
+++ twisted/internet/test/test_endpoints.py
@@ -1838,10 +1838,11 @@ class ServerStringTests(unittest.TestCase):
         self.assertIsInstance(ctx, ContextType)
 
 
-    # Use a class variable to ensure we use the exactly same endpoint string
-    # except for the chain file itself.
-    SSL_CHAIN_TEMPLATE = ("ssl:1234:privateKey=%s:extraCertChain=%%s"
-                          % (escapedPEMPathName,))
+    if not skipSSL:
+        # Use a class variable to ensure we use the exactly same endpoint
+        # string except for the chain file itself.
+        SSL_CHAIN_TEMPLATE = ("ssl:1234:privateKey=%s:extraCertChain=%%s"
+                              % (escapedPEMPathName,))
 
 
     def test_sslChainLoads(self):
