diff --git twisted/web/test/test_http.py twisted/web/test/test_http.py
index 959acfa..403ec65 100644
--- twisted/web/test/test_http.py
+++ twisted/web/test/test_http.py
@@ -878,6 +878,40 @@ Hello,
 
         self.runRequest(httpRequest, MyRequest)
 
+    def test_malformed_chunkedEncoding(self):
+        """
+        If a request uses the I{chunked} transfer encoding, the request body is
+        decoded accordingly before it is made available on the request. If the
+        request is malformed, we should get a 400 error.
+        """
+        # See test_chunkedEncoding for the correct form of this request.
+        httpRequest = '''\
+GET / HTTP/1.0
+Content-Type: text/plain
+Transfer-Encoding: chunked
+
+MALFORMED_LINE_THIS_SHOULD_BE_'6'
+Hello,
+14
+ spam,eggs spam spam
+0
+
+'''
+        testcase = self
+        didRequest = False
+        class MyRequest(http.Request):
+            def process(self):
+                # This request should fail, so this should never be called.
+                testcase.didRequest = True
+
+        try:
+            channel = self.runRequest(httpRequest, MyRequest, success=False)
+        except ValueError:
+            self.fail()
+        self.assertFalse(didRequest)
+        self.assertEqual(
+            channel.transport.value(),
+            "HTTP/1.1 400 Bad Request\r\n\r\n")
 
 
 class QueryArgumentsTestCase(unittest.TestCase):
