Index: doc/web/examples/httpclient.py
===================================================================
--- doc/web/examples/httpclient.py	(revision 34055)
+++ doc/web/examples/httpclient.py	(working copy)
@@ -2,6 +2,16 @@
 # Copyright (c) Twisted Matrix Laboratories.
 # See LICENSE for details.
 
+"""
+Example of retrieving the resource at a provided url.
+
+This script prints out the resource as well as session headers 
+and other metadata from the http response object.
+
+To demo this script, run this file:
+	python httpclient.py <url>
+"""
+
 import sys
 from pprint import pprint
 
Index: doc/web/examples/webguard.py
===================================================================
--- doc/web/examples/webguard.py	(revision 34055)
+++ doc/web/examples/webguard.py	(working copy)
@@ -1,6 +1,16 @@
 # Copyright (c) Twisted Matrix Laboratories.
 # See LICENSE for details.
 
+"""
+Example of HTTP authentication with username and password using portals and realms.
+
+To demo this script, run this file:
+    python webguard.py
+
+Visit the url http://localhost:8899 , at login prompt enter the username 'joe' and password
+'blow' to authenticate the site.
+"""
+
 import sys
 
 from zope.interface import implements
Index: twisted/web/topfiles/5592.misc
===================================================================
