Index: gtk2reactor.py
===================================================================
--- gtk2reactor.py	(revision 14195)
+++ gtk2reactor.py	(working copy)
@@ -231,13 +231,13 @@
 
     def crash(self):
         import gtk
-        gtk.mainquit()
+        gtk.main_quit()
 
     def run(self, installSignalHandlers=1):
         import gtk
         self.startRunning(installSignalHandlers=installSignalHandlers)
         self.simulate()
-        gtk.mainloop()
+        gtk.main()
 
     def simulate(self):
         """Run simulation loops and reschedule callbacks.
