Index: twisted/web/http.py
===================================================================
--- twisted/web/http.py	(revision 25717)
+++ twisted/web/http.py	(working copy)
@@ -666,7 +666,7 @@
         @type version: C{str}
         @param version: The HTTP version of this request.
         """
-        self.content.seek(0,0)
+        self.content.seek(0, 0)
         self.args = {}
         self.stack = []
 
@@ -711,6 +711,7 @@
                         return
                     raise
 
+        self.content.seek(0, 0)
         self.process()
 
 
