=== modified file 'twisted/mail/tap.py'
--- twisted/mail/tap.py	2012-03-13 10:09:50 +0000
+++ twisted/mail/tap.py	2012-06-13 07:15:44 +0000
@@ -112,19 +112,6 @@
     opt_s = opt_smtp
 
 
-    def opt_passwordfile(self, filename):
-        """
-        Specify a file containing username:password login info for authenticated
-        ESMTP connections. (DEPRECATED; see --help-auth instead)
-        """
-        ch = checkers.OnDiskUsernamePasswordDatabase(filename)
-        self.service.smtpPortal.registerChecker(ch)
-        msg = deprecate.getDeprecationWarningString(
-            self.opt_passwordfile, versions.Version('twisted.mail', 11, 0, 0))
-        warnings.warn(msg, category=DeprecationWarning, stacklevel=2)
-    opt_P = opt_passwordfile
-
-
     def opt_default(self):
         """Make the most recently specified domain the default domain."""
         if self.last_domain:

=== modified file 'twisted/mail/test/test_options.py'
--- twisted/mail/test/test_options.py	2012-03-13 10:09:50 +0000
+++ twisted/mail/test/test_options.py	2012-06-13 07:16:32 +0000
@@ -47,22 +47,6 @@
             '--aliases', self.aliasFilename])
 
 
-    def testPasswordfileDeprecation(self):
-        """
-        Test that the --passwordfile option will emit a correct warning.
-        """
-        passwd = FilePath(self.mktemp())
-        passwd.setContent("")
-        options = Options()
-        options.opt_passwordfile(passwd.path)
-        warnings = self.flushWarnings([self.testPasswordfileDeprecation])
-        self.assertEqual(warnings[0]['category'], DeprecationWarning)
-        self.assertEqual(len(warnings), 1)
-        msg = deprecate.getDeprecationWarningString(options.opt_passwordfile,
-                             versions.Version('twisted.mail', 11, 0, 0))
-        self.assertEqual(warnings[0]['message'], msg)
-
-
     def test_barePort(self):
         """
         A bare port passed to I{--pop3} results in deprecation warning in

=== added file 'twisted/mail/topfiles/5541.removal'
--- twisted/mail/topfiles/5541.removal	1970-01-01 00:00:00 +0000
+++ twisted/mail/topfiles/5541.removal	2012-06-13 07:19:08 +0000
@@ -0,0 +1,1 @@
+twiste.mail.tap.Options.opt_passwordfile is removed.

