Index: twisted/internet/cdefer/cdefer.c
===================================================================
--- twisted/internet/cdefer/cdefer.c	(revision 20676)
+++ twisted/internet/cdefer/cdefer.c	(working copy)
@@ -50,7 +50,7 @@
      * popping every item. It has to be per-deferred, because
      * runCallbacks can be called reentrantly, and must not repeat the
      * previously-called callbacks. */
-    Py_ssize_t callback_index;
+    int callback_index;
 } cdefer_Deferred;
 
 /* Prototypes */
