Index: twisted/internet/_dumbwin32proc.py
===================================================================
--- twisted/internet/_dumbwin32proc.py	(revision 44923)
+++ twisted/internet/_dumbwin32proc.py	(working copy)
@@ -16,6 +16,7 @@
 import win32pipe
 import win32process
 import win32security
+import winerror
 
 import pywintypes
 
@@ -172,8 +173,14 @@
         cmdline = quoteArguments(args)
         # TODO: error detection here.  See #2787 and #4184.
         def doCreate():
-            self.hProcess, self.hThread, self.pid, dwTid = win32process.CreateProcess(
-                command, cmdline, None, None, 1, 0, env, path, StartupInfo)
+            try:
+                self.hProcess, self.hThread, self.pid, dwTid = win32process.CreateProcess(
+                    command, cmdline, None, None, 1, 0, env, path, StartupInfo)
+            except pywintypes.error, pwte:
+                if pwte.args[0] != winerror.ERROR_FILE_NOT_FOUND:
+                    raise
+                self.hProcess, self.hThread, self.pid, dwTid = win32process.CreateProcess(
+                    None, cmdline, None, None, 1, 0, env, path, StartupInfo)
         try:
             try:
                 doCreate()
