Index: twisted/python/_release.py
===================================================================
--- twisted/python/_release.py	(revision 28364)
+++ twisted/python/_release.py	(working copy)
@@ -748,6 +748,7 @@
             newNews,
             self._headings.get(self._MISC),
             self._findChanges(path, self._MISC))
+        newNews.write('\n')
         newNews.write(oldNews)
         newNews.close()
         output.sibling('NEWS.new').moveTo(output)
Index: twisted/python/test/test_release.py
===================================================================
--- twisted/python/test/test_release.py	(revision 28364)
+++ twisted/python/test/test_release.py	(working copy)
@@ -1650,7 +1650,7 @@
             'Other\n'
             '-----\n'
             ' - #30, #35\n'
-            '\n' + self.existingText)
+            '\n\n' + self.existingText)
 
 
     def test_preserveTicketHint(self):
@@ -1695,7 +1695,7 @@
             'Other\n'
             '-----\n'
             ' - #30, #35\n'
-            '\n'
+            '\n\n'
             'Blah blah other stuff.\n')
 
 
@@ -1725,7 +1725,7 @@
             'Deprecations and Removals\n'
             '-------------------------\n'
             ' - Stupid stuff was deprecated. (#25)\n'
-            '\n'
+            '\n\n'
             'Here is stuff which was present previously.\n')
 
 
@@ -1764,7 +1764,7 @@
             'Other\n'
             '-----\n'
             ' - #30, #35\n'
-            '\n'
+            '\n\n'
             'Here is stuff which was present previously.\n')
 
 
