Index: twisted/runner/test/test_procmon.py
===================================================================
--- twisted/runner/test/test_procmon.py	(revision 33677)
+++ twisted/runner/test/test_procmon.py	(working copy)
@@ -475,54 +475,3 @@
         # all pending process restarts.
         self.assertEqual(self.pm.protocols, {})
 
-
-    def test_activeAttributeEqualsRunning(self):
-        """
-        L{ProcessMonitor.active} unneccessarily duplicates the standard
-        L{IService.running} flag.
-        """
-        self.assertEqual(self.pm.active, self.pm.running)
-        self.pm.startService()
-        self.assertEqual(self.pm.active, self.pm.running)
-
-
-    def test_activeAttributeDeprecation(self):
-        """
-        L{ProcessMonitor.active} unneccessarily duplicates the standard
-        L{IService.running} flag and is therefore deprecated.
-        """
-        def getActive():
-            return self.pm.active
-        expectedMessage = "active is deprecated since Twisted 10.1.0.  Use " \
-                          "running instead."
-
-        self.assertWarns(DeprecationWarning,
-                         expectedMessage, __file__, getActive)
-
-
-    def test_consistencyAttributeDeprecation(self):
-        """
-        L{ProcessMonitor.consistency} is no longer needed since the removal of
-        the ProcessMonitor._checkConsistency function and is therefore
-        deprecated.
-        """
-        def getConsistency():
-            return self.pm.consistency
-        expectedMessage = "consistency is deprecated since Twisted 10.1.0."
-
-        self.assertWarns(DeprecationWarning,
-                         expectedMessage, __file__, getConsistency)
-
-
-    def test_consistencyDelayAttributeDeprecation(self):
-        """
-        L{ProcessMonitor.consistencyDelay} is no longer needed since the
-        removal of the ProcessMonitor._checkConsistency function and is
-        therefore deprecated.
-        """
-        def getConsistencyDelay():
-            return self.pm.consistencyDelay
-        expectedMessage = "consistencyDelay is deprecated since Twisted 10.1.0."
-
-        self.assertWarns(DeprecationWarning,
-                         expectedMessage, __file__, getConsistencyDelay)
Index: twisted/runner/procmon.py
===================================================================
--- twisted/runner/procmon.py	(revision 33677)
+++ twisted/runner/procmon.py	(working copy)
@@ -108,31 +108,7 @@
         self.murder = {}
         self.restart = {}
 
-    def _getActive(self):
-        warnings.warn("active is deprecated since Twisted 10.1.0.  "
-                      "Use running instead.", category=DeprecationWarning,
-                                              stacklevel=2)
-        return self.running
 
-    active = property(_getActive, None)
-
-
-    def _getConsistency(self):
-        warnings.warn("consistency is deprecated since Twisted 10.1.0.",
-                      category=DeprecationWarning, stacklevel=2)
-        return None
-
-    consistency = property(_getConsistency, None)
-
-
-    def _getConsistencyDelay(self):
-        warnings.warn("consistencyDelay is deprecated since Twisted 10.1.0.",
-                      category=DeprecationWarning, stacklevel=2)
-        return 60
-
-    consistencyDelay = property(_getConsistencyDelay, None)
-
-
     def __getstate__(self):
         dct = service.Service.__getstate__(self)
         del dct['_reactor']
