Index: twisted/mail/test/test_smtp.py
===================================================================
--- twisted/mail/test/test_smtp.py	(revision 38792)
+++ twisted/mail/test/test_smtp.py	(working copy)
@@ -13,7 +13,7 @@
 from twisted.mail import smtp
 from twisted.internet import defer, protocol, reactor, interfaces
 from twisted.internet import address, error, task
-from twisted.test.proto_helpers import StringTransport
+from twisted.test.proto_helpers import MemoryReactor, StringTransport
 
 from twisted import cred
 import twisted.cred.error
@@ -1649,3 +1649,26 @@
             warningsShown[0]['message'],
             "tlsMode attribute of twisted.mail.smtp.ESMTPClient "
             "is deprecated since Twisted 13.0")
+
+
+
+class SendmailTestCase(unittest.TestCase):
+    """
+    Tests for L{twisted.mail.smtp.sendmail}.
+    """
+    def test_cancel(self):
+        """
+        A deferred returned by L{twisted.mail.smtp.sendmail} can be cancelled,
+        telling the L{twisted.mail.smtp.SMTPSenderFactory} not to 
+        retry and disconnecting the L{twisted.internet.interfaces.IConnector}. 
+        """
+        reactor = MemoryReactor()
+        d = smtp.sendmail("localhost", "source@address", "recipient@address",
+                          "message", _reactor=reactor)
+        factory = reactor.tcpClients[0][2]
+        connector = reactor.connectors[0]
+        d.cancel()
+        self.assertEqual(factory.sendFinished, 1)
+        self.assertEqual(connector._disconnected, True)
+        self.assertFailure(d, defer.CancelledError)
+        return d
Index: twisted/mail/smtp.py
===================================================================
--- twisted/mail/smtp.py	(revision 38792)
+++ twisted/mail/smtp.py	(working copy)
@@ -1894,7 +1894,8 @@
         p.timeout = self.timeout
         return p
 
-def sendmail(smtphost, from_addr, to_addrs, msg, senderDomainName=None, port=25):
+def sendmail(smtphost, from_addr, to_addrs, msg,
+             senderDomainName=None, port=25, _reactor=reactor):
     """Send an email
 
     This interface is intended to be a direct replacement for
@@ -1921,6 +1922,8 @@
 
     @param port: Remote port to which to connect.
 
+    @param _reactor: The reactor used to connect to host.
+
     @rtype: L{Deferred}
     @returns: A L{Deferred}, its callback will be called if a message is sent
         to ANY address, the errback if no message is sent.
@@ -1934,13 +1937,19 @@
         # It's not a file
         msg = StringIO(str(msg))
 
-    d = defer.Deferred()
+    def cancel(d):
+        """
+        Cancel the L{twisted.mail.smtp.sendmail} call.
+        """
+        factory.sendFinished = 1
+        connector.disconnect()
+    d = defer.Deferred(cancel)
     factory = SMTPSenderFactory(from_addr, to_addrs, msg, d)
 
     if senderDomainName is not None:
         factory.domain = senderDomainName
 
-    reactor.connectTCP(smtphost, port, factory)
+    connector = _reactor.connectTCP(smtphost, port, factory)
 
     return d
 
