--- test_wsgi_new.py	2007-04-30 15:51:36.000000000 +0200
+++ test_wsgi_new2.py	2007-04-30 15:34:13.000000000 +0200
@@ -278,6 +278,10 @@
             self.assertEnv('/foo', {'SERVER_PORT': '80'}, version=(1,0)),
         ], fireOnOneErrback=True)
 
+    def test_SCRIPT_NAME(self):
+        """SCRIPT_NAME"""
+        return self.assertEnv('http://host/', {'SCRIPT_NAME': ''})
+
 if WSGI is None:
     for cls in (TestContainer, TestWSGIEnvironment):
         setattr(cls, 'skip', 'Required thread support is missing, skipping')
