Index: doc/core/howto/pb-copyable.xhtml
===================================================================
--- doc/core/howto/pb-copyable.xhtml	(revision 32564)
+++ doc/core/howto/pb-copyable.xhtml	(working copy)
@@ -95,7 +95,7 @@
 events is the receipt of an object through a PB remote procedure call, which
 is a request to create an object in your <q>inside</q> namespace. The
 question is what to do in response to it. For this reason, you must
-explicitly specific what remote classes will be accepted, and how their
+explicitly specify what remote classes will be accepted, and how their
 local representatives are to be created.</p>
 
 <h3>What class to use?</h3>
