diff --git twisted/logger/test/test_json.py twisted/logger/test/test_json.py
index 738ed83..40fd4b0 100644
--- twisted/logger/test/test_json.py
+++ twisted/logger/test/test_json.py
@@ -280,6 +280,27 @@ class FileLogObserverTests(TestCase):
         self.assertObserverWritesJSON(recordSeparator=u"")
 
 
+    # TODO: Better name, document it, etc
+    def test_observerDoesntFailWithWhatever(self):
+        try:
+            1 / 0
+        except ZeroDivisionError:
+            f = Failure()
+        event = {
+            'log_namespace': 'twisted.internet.defer',
+            'log_level': LogLevel.critical,
+            'debugInfo': '',
+            'log_source': None,
+            'log_logger': None,  # do i need a logger instance?
+            'log_time': 1446731370.38742,
+            'log_format': None,
+            'log_failure': f,
+        }
+        with StringIO() as fileHandle:
+            observer = jsonFileLogObserver(fileHandle)
+            observer(event)
+
+
 
 class LogFileReaderTests(TestCase):
     """
