diff --git twisted/application/reactors.py twisted/application/reactors.py
index cccddd1..b4efd7e 100644
--- twisted/application/reactors.py
+++ twisted/application/reactors.py
@@ -7,7 +7,9 @@ Plugin-based system for enumerating available reactors and installing one of
 them.
 """
 
-from zope.interface import Interface, Attribute, implements
+from __future__ import print_function,division,absolute_import
+
+from zope.interface import Interface, Attribute, implementer
 
 from twisted.plugin import IPlugin, getPlugins
 from twisted.python.reflect import namedAny
@@ -41,12 +43,12 @@ class NoSuchReactor(KeyError):
     """
 
 
+@implementer(IPlugin, IReactorInstaller)
 class Reactor(object):
     """
     @ivar moduleName: The fully-qualified Python name of the module of which
     the install callable is an attribute.
     """
-    implements(IPlugin, IReactorInstaller)
 
 
     def __init__(self, shortName, moduleName, description):
