Index: doc/core/howto/defer.xhtml
===================================================================
--- doc/core/howto/defer.xhtml	(revision 18770)
+++ doc/core/howto/defer.xhtml	(working copy)
@@ -548,7 +548,7 @@
 <ul>
   <li>
   <code class="py-prototype">addCallbacks(self, callback[, errback, callbackArgs,
-  errbackArgs, errbackKeywords, asDefaults])</code> 
+  callbackKeywords, errbackArgs, errbackKeywords])</code> 
   
   <p>This is the method you will use to interact
   with Deferred. It adds a pair of callbacks <q>parallel</q> to
