Index: twisted/spread/jelly.py
===================================================================
--- twisted/spread/jelly.py	(revision 24265)
+++ twisted/spread/jelly.py	(working copy)
@@ -65,6 +65,7 @@
 # System Imports
 import pickle
 import types
+import inspect
 import warnings
 from types import StringType
 from types import UnicodeType
@@ -263,7 +264,7 @@
 
     for i in dir(module):
         i_ = getattr(module, i)
-        if type(i_) == types.ClassType:
+        if inspect.isclass(i_):
             if issubclass(i_, baseClass):
                 setUnjellyableForClass('%s%s' % (prefix, i), i_)
 
