Index: twisted/names/test/test_cache.py
===================================================================
--- twisted/names/test/test_cache.py	(revision 34157)
+++ twisted/names/test/test_cache.py	(working copy)
@@ -73,8 +73,8 @@
 
         def cbLookup(result):
             self.assertEquals(result[0][0].ttl, 0)
-            self.assertEquals(result[0][0].ttl, 0)
-            self.assertEquals(result[0][0].ttl, 0)
+            self.assertEquals(result[1][0].ttl, 0)
+            self.assertEquals(result[2][0].ttl, 0)
             self.assertEquals(result[0][0].name.name, "example.com")
 
         return c.lookupAddress("example.com").addCallback(cbLookup)
