diff --git a/twisted/internet/endpoints.py b/twisted/internet/endpoints.py
index 5799ea0..3c7c2e6 100644
--- a/twisted/internet/endpoints.py
+++ b/twisted/internet/endpoints.py
@@ -1631,7 +1631,6 @@ def _parseClientSSL(*args, **kwargs):
         verify = False
         caCerts = None
     kwargs['sslContextFactory'] = ssl.CertificateOptions(
-        method=ssl.SSL.SSLv23_METHOD,
         certificate=certx509,
         privateKey=privateKey,
         verify=verify,
diff --git a/twisted/internet/test/test_endpoints.py b/twisted/internet/test/test_endpoints.py
index 04d32a2..6ce068c 100644
--- a/twisted/internet/test/test_endpoints.py
+++ b/twisted/internet/test/test_endpoints.py
@@ -2908,6 +2908,8 @@ class SSLClientStringTests(unittest.TestCase):
         self.assertEqual(client._bindAddress, (b"10.0.0.3", 0))
         certOptions = client._sslContextFactory
         self.assertIsInstance(certOptions, CertificateOptions)
+        self.assertEqual(certOptions.method, SSLv23_METHOD)
+        self.assertTrue(certOptions._options & OP_NO_SSLv3)
         ctx = certOptions.getContext()
         self.assertIsInstance(ctx, ContextType)
         self.assertEqual(Certificate(certOptions.certificate), testCertificate)
diff --git a/twisted/topfiles/7686.bugfix b/twisted/topfiles/7686.bugfix
new file mode 100644
index 0000000..556a5f7
--- /dev/null
+++ b/twisted/topfiles/7686.bugfix
@@ -0,0 +1 @@
+The SSL client string endpoint parser (twisted.internet.endpoints.clientFromString) now constructs endpoints which, by default, disable the insecure SSLv3 protocol.
