Index: twisted/conch/test/test_ckeygen.py
===================================================================
--- twisted/conch/test/test_ckeygen.py	(revision 35300)
+++ twisted/conch/test/test_ckeygen.py	(working copy)
@@ -15,7 +15,9 @@
     skip = "PyCrypto and pyasn1 required for twisted.conch.scripts.ckeygen."
 else:
     from twisted.conch.ssh.keys import Key
-    from twisted.conch.scripts.ckeygen import printFingerprint, _saveKey
+    from twisted.conch.scripts.ckeygen import (
+        displayPublicKey, printFingerprint, _saveKey
+    )
 
 from twisted.python.filepath import FilePath
 from twisted.trial.unittest import TestCase
@@ -78,3 +80,17 @@
             Key.fromString(base.child('id_rsa.pub').getContent()),
             key.public())
 
+
+    def test_displayPublicKey(self):
+        """
+        L{displayPublicKey} prints out the public key associated with a given
+        private key
+        """
+        filename = self.mktemp()
+        FilePath(filename).setContent(privateRSA_openssh)
+        displayPublicKey({'filename': filename})
+        self.assertEqual(
+            self.stdout.getvalue().strip('\n'),
+            publicRSA_openssh.strip(' comment')
+        )
+
Index: twisted/conch/scripts/ckeygen.py
===================================================================
--- twisted/conch/scripts/ckeygen.py	(revision 35300)
+++ twisted/conch/scripts/ckeygen.py	(working copy)
@@ -147,7 +147,7 @@
                 options['pass'] = getpass.getpass('Enter passphrase: ')
             key = keys.Key.fromFile(
                 options['filename'], passphrase = options['pass']).keyObject
-    print keys.Key(key).public().toString()
+    print keys.Key(key).public().toString('openssh')
 
 
 def _saveKey(key, options):
