diff --git twisted/python/win32.py twisted/python/win32.py
index ca04fc0..cd318ba 100644
--- twisted/python/win32.py
+++ twisted/python/win32.py
@@ -12,7 +12,6 @@ See also twisted.python.shortcut.
 """
 
 import re
-import exceptions
 import os
 
 try:
@@ -31,19 +30,16 @@ ERROR_DIRECTORY = 267
 
 O_BINARY = getattr(os, "O_BINARY", 0)
 
-def _determineWindowsError():
-    """
-    Determine which WindowsError name to export.
-    """
-    return getattr(exceptions, 'WindowsError', FakeWindowsError)
-
 class FakeWindowsError(OSError):
     """
     Stand-in for sometimes-builtin exception on platforms for which it
     is missing.
     """
 
-WindowsError = _determineWindowsError()
+try:
+    WindowsError = WindowsError
+except NameError:
+    WindowsError = FakeWindowsError
 
 # XXX fix this to use python's builtin _winreg?
 
