--- defer.xhtml	Mon Jan 19 05:48:34 2004
+++ defer.xhtml_	Mon Jan 19 05:47:40 2004
@@ -144,13 +144,13 @@
     import sys
     sys.stderr.write(str(failure))
 
-# this will print an error message
+# this will print "<p>Result: 9</p>"
 g = Getter()
 d = g.getData(3)
 d.addCallback(printData)
 d.addErrback(printError)
 
-# this will print "<p>Result: 12</p>"
+# this will print an error message
 g = Getter()
 d = g.getData(4)
 d.addCallback(printData)
