Index: doc/core/howto/quotes.xhtml
===================================================================
--- doc/core/howto/quotes.xhtml	(revision 27914)
+++ doc/core/howto/quotes.xhtml	(working copy)
@@ -30,7 +30,7 @@
     <ul>
     <li><a href="listings/TwistedQuotes/__init__.py" class="py-listing">__init__.py</a> (this
         file marks it as a package, see <a
-        href="http://www.python.org/doc/current/tut/node8.html#SECTION008400000000000000000"
+        href="http://docs.python.org/tutorial/modules.html#packages"
         >this section</a> of the Python tutorial for more on packages);</li>
     <li><a href="listings/TwistedQuotes/quoters.py" class="py-listing">quoters.py</a>;</li>
     <li><a href="listings/TwistedQuotes/quoteproto.py" class="py-listing">quoteproto.py</a>;</li>
