Index: test/test_doctest.py
===================================================================
--- test/test_doctest.py	(revision 23315)
+++ test/test_doctest.py	(working copy)
@@ -53,14 +53,29 @@
         self.assertEqual(7, suite.countTestCases())
 
 
-    def test_expectedResults(self):
+    def _testRun(self, suite):
         """
-        Trial can correctly run doctests with its xUnit test APIs.
+        Run C{suite} and check the result.
         """
-        suite = self.makeDocSuite(mockdoctest)
         result = reporter.TestResult()
         suite.run(result)
         self.assertEqual(5, result.successes)
         # doctest reports failures as errors in 2.3
         self.assertEqual(2, len(result.errors) + len(result.failures))
 
+
+    def test_expectedResults(self, count=1):
+        """
+        Trial can correctly run doctests with its xUnit test APIs.
+        """
+        suite = runner.TestLoader().loadDoctests(mockdoctest)
+        self._testRun(suite)
+
+
+    def test_repeatable(self):
+        """
+        Doctests should be runnable repeatably.
+        """
+        suite = runner.TestLoader().loadDoctests(mockdoctest)
+        self._testRun(suite)
+        self._testRun(suite)
Index: runner.py
===================================================================
--- runner.py	(revision 23315)
+++ runner.py	(working copy)
@@ -565,7 +565,14 @@
         if not inspect.ismodule(module):
             warnings.warn("trial only supports doctesting modules")
             return
-        return doctest.DocTestSuite(module)
+        if (2, 4) <= sys.version_info:
+            # Work around Python issue2604: DocTestCase.tearDown clobbers globs
+            def saveGlobs(test):
+                test._saved_globs = getattr(test, '_saved_globs', test.globs)
+                test.globs = test._saved_globs.copy()
+            return doctest.DocTestSuite(module, setUp=saveGlobs)
+        else:
+            return doctest.DocTestSuite(module)
 
     def loadAnything(self, thing, recurse=False):
         """
