Index: twisted/mail/test/test_mail.py
===================================================================
--- twisted/mail/test/test_mail.py	(revision 39143)
+++ twisted/mail/test/test_mail.py	(working copy)
@@ -33,6 +33,7 @@
 from twisted.internet import address
 from twisted.python import failure
 from twisted.python.filepath import FilePath
+from twisted.python.deprecate import deprecatedModuleAttribute
 
 from twisted import mail
 import twisted.mail.mail
@@ -81,7 +82,6 @@
             self.assertEqual(d[x], x + 10)
             self.assertEqual(d.get(x), x + 10)
             self.failUnless(x in d)
-            self.failUnless(d.has_key(x))
 
         del d[2], d[4], d[6]
 
@@ -141,8 +141,24 @@
         """
         self._stringificationTest(repr)
 
+    def test_has_key(self):
+        """
+        L{DomainWithDefaultDict.has_key} is deprecated. has_key always
+        returns 1.
+        """
+        domain = mail.mail.DomainWithDefaultDict({}, 'Default')
+        self.failUnless(domain.has_key('key'))
+        warningsShown = self.flushWarnings()
+        self.assertEqual(len(warningsShown), 1)
+        self.assertIdentical(warningsShown[0]['category'], DeprecationWarning)
+        self.assertEqual(
+            warningsShown[0]['message'],
+            "twisted.mail.mail.DomainWithDefaultDict.has_key is "
+            "deprecated since Twisted 13.1.0. Please use the in keyword "
+            "instead.")
 
 
+
 class BounceTestCase(unittest.TestCase):
     def setUp(self):
         self.domain = mail.mail.BounceDomain()
Index: twisted/mail/mail.py
===================================================================
--- twisted/mail/mail.py	(revision 39143)
+++ twisted/mail/mail.py	(working copy)
@@ -19,6 +19,7 @@
 
 # System imports
 import os
+import warnings
 from zope.interface import implements, Interface
 
 
@@ -34,6 +35,12 @@
         self.default = domain
 
     def has_key(self, name):
+        warnings.warn(
+            "twisted.mail.mail.DomainWithDefaultDict.has_key is "
+            "deprecated since Twisted 13.1.0. Please use the in keyword "
+            "instead.",
+            DeprecationWarning,
+            stacklevel=2)
         return 1
 
     def fromkeys(klass, keys, value=None):
