Index: twisted/conch/error.py
===================================================================
--- twisted/conch/error.py	(revision 27280)
+++ twisted/conch/error.py	(working copy)
@@ -12,10 +12,19 @@
 
 
 class ConchError(Exception):
-    def __init__(self, value, data = None):
-        Exception.__init__(self, value, data)
+    """
+    Raised as a general conch exception if a more specific exception doesn't
+    exist.
+
+    @ivar data: textual information about the error.
+    @type data: C{str}
+    @ivar value: the error code.
+    @type value: C{int}
+    """
+    def __init__(self, data, value = None):
+        Exception.__init__(self, data, value)
+        self.data = data
         self.value = value
-        self.data = data
 
 
 
