diff --git twisted/conch/scripts/ckeygen.py twisted/conch/scripts/ckeygen.py
index 5e7184a..113f362 100644
--- twisted/conch/scripts/ckeygen.py
+++ twisted/conch/scripts/ckeygen.py
@@ -31,11 +31,12 @@ class GeneralOptions(usage.Options):
                      ['type', 't', None, 'Specify type of key to create.'],
                      ['comment', 'C', None, 'Provide new comment.'],
                      ['newpass', 'N', None, 'Provide new passphrase.'],
-                     ['pass', 'P', None, 'Provide old passphrase']]
+                     ['pass', 'P', None, 'Provide old passphrase.']]
 
     optFlags = [['fingerprint', 'l', 'Show fingerprint of key file.'],
                 ['changepass', 'p', 'Change passphrase of private key file.'],
                 ['quiet', 'q', 'Quiet.'],
+                ['no-passphrase', None, "Create the key with no passphrase."],
                 ['showpub', 'y', 'Read private key file and print public key.']]
 
     compData = usage.Completions(
@@ -193,7 +194,9 @@ def _saveKey(key, options):
         yn = raw_input('Overwrite (y/n)? ')
         if yn[0].lower() != 'y':
             sys.exit()
-    if not options['pass']:
+    if options.get('no-passphrase'):
+        options['pass'] = b''
+    elif not options['pass']:
         while 1:
             p1 = getpass.getpass('Enter passphrase (empty for no passphrase): ')
             p2 = getpass.getpass('Enter same passphrase again: ')
diff --git twisted/conch/test/test_ckeygen.py twisted/conch/test/test_ckeygen.py
index b4df2d7..272f947 100644
--- twisted/conch/test/test_ckeygen.py
+++ twisted/conch/test/test_ckeygen.py
@@ -99,6 +99,24 @@ class KeyGenTests(TestCase):
             key.public())
 
 
+    def test_saveKeyEmptyPassphrase(self):
+        """
+        L{_saveKey} will choose an empty string for the passphrase if
+        no-passphrase is C{True}.
+        """
+        base = FilePath(self.mktemp())
+        base.makedirs()
+        filename = base.child('id_rsa').path
+        key = Key.fromString(privateRSA_openssh)
+        _saveKey(
+            key.keyObject,
+            {'filename': filename, 'no-passphrase': True})
+        self.assertEqual(
+            key.fromString(
+                base.child('id_rsa').getContent(), None, b''),
+            key)
+
+
     def test_displayPublicKey(self):
         """
         L{displayPublicKey} prints out the public key associated with a given
diff --git twisted/conch/topfiles/5998.feature twisted/conch/topfiles/5998.feature
new file mode 100644
index 0000000..1bcc8af
--- /dev/null
+++ twisted/conch/topfiles/5998.feature
@@ -0,0 +1 @@
+ckeygen now accepts --no-passphrase as a means of setting the passphrase to an empty string.
\ No newline at end of file
