Index: twisted/test/test_dirdbm.py
===================================================================
--- twisted/test/test_dirdbm.py	(revision 24123)
+++ twisted/test/test_dirdbm.py	(working copy)
@@ -18,9 +18,6 @@
         self.dbm = dirdbm.open(self.path)
         self.items = (('abc', 'foo'), ('/lalal', '\000\001'), ('\000\012', 'baz'))
 
-    def tearDown(self):
-        shutil.rmtree(self.path)
-
     def testAll(self):
         k = "//==".decode("base64")
         self.dbm[k] = "a"
