Index: /home/mark/workspace/Twisted/web/http.py
===================================================================
--- /home/mark/workspace/Twisted/web/http.py	(revision 19722)
+++ /home/mark/workspace/Twisted/web/http.py	(working copy)
@@ -357,6 +357,7 @@
     based virtual hosting.
     """
     length = None
+    ignoreLength = False
     firstLine = 1
     __buffer = None
 
@@ -361,6 +362,9 @@
     __buffer = None
 
     def sendCommand(self, command, path):
+        # If HTTP method is HEAD, there might be a Content-length but no body
+        self.ignoreLength = (command == 'HEAD')
+        
         self.transport.write('%s %s HTTP/1.0\r\n' % (command, path))
 
     def sendHeader(self, name, value):
@@ -386,7 +390,7 @@
             key, val = line.split(':', 1)
             val = val.lstrip()
             self.handleHeader(key, val)
-            if key.lower() == 'content-length':
+            if not self.ignoreLength and key.lower() == 'content-length':
                 self.length = int(val)
         else:
             self.__buffer = StringIO()
