Index: /Software/fsw/Fetch/trunk/ThirdParty/twisted/twisted/plugin.py
===================================================================
--- /Software/fsw/Fetch/trunk/ThirdParty/twisted/twisted/plugin.py	(revision 24620)
+++ /Software/fsw/Fetch/trunk/ThirdParty/twisted/twisted/plugin.py	(working copy)
@@ -110,10 +110,15 @@
 
 _exts = fromkeys(['.py', '.so', '.pyd', '.dll'])
 
+def locateDropinCache(path):
+	return os.path.join(path, "dropin.cache")
+
+dropinCacheLocator = locateDropinCache
+
 def getCache(module):
     topcache = {}
     for p in module.__path__:
-        dropcache = os.path.join(p, "dropin.cache")
+        dropcache = dropinCacheLocator(p)
         try:
             cache = pickle.load(file(dropcache))
             lastCached = getmtime(dropcache)
