Index: web2/examples/demo.py
===================================================================
--- web2/examples/demo.py	(revision 24190)
+++ web2/examples/demo.py	(working copy)
@@ -1,5 +1,5 @@
 #!/usr/bin/env python
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 """I am a simple test resource.
Index: lore/man/lore.1
===================================================================
--- lore/man/lore.1	(revision 24190)
+++ lore/man/lore.1	(working copy)
@@ -51,7 +51,7 @@
 .SH "REPORTING BUGS"
 To report a bug, visit \fIhttp://twistedmatrix.com/bugs/\fR
 .SH COPYRIGHT
-Copyright \(co 2003 Matthew W. Lefkowitz
+Copyright \(co 2003-2008 Twisted Matrix Laboratories.
 .br
 This is free software; see the source for copying conditions.  There is NO
 warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Index: mail/man/mailmail.1
===================================================================
--- mail/man/mailmail.1	(revision 24190)
+++ mail/man/mailmail.1	(working copy)
@@ -15,7 +15,7 @@
 .SH "REPORTING BUGS"
 To report a bug, visit \fIhttp://twistedmatrix.com/bugs/\fR
 .SH COPYRIGHT
-Copyright \(co 2003 Matthew W. Lefkowitz
+Copyright \(co 2003-2008 Twisted Matrix Laboratories.
 .br
 This is free software; see the source for copying conditions.  There is NO
 warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Index: mail/examples/imap4client.py
===================================================================
--- mail/examples/imap4client.py	(revision 24190)
+++ mail/examples/imap4client.py	(working copy)
@@ -1,6 +1,6 @@
 #!/usr/bin/python
 
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
Index: mail/examples/emailserver.tac
===================================================================
--- mail/examples/emailserver.tac	(revision 24190)
+++ mail/examples/emailserver.tac	(working copy)
@@ -1,5 +1,5 @@
 
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 # You can run this module directly with:
Index: core/howto/listings/pb/pbAnonServer.py
===================================================================
--- core/howto/listings/pb/pbAnonServer.py	(revision 24190)
+++ core/howto/listings/pb/pbAnonServer.py	(working copy)
@@ -1,5 +1,5 @@
 #!/usr/bin/python
-# Copyright (c) 2007 Twisted Matrix Laboratories.
+# Copyright (c) 2007-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 """
Index: core/howto/listings/pb/pbAnonClient.py
===================================================================
--- core/howto/listings/pb/pbAnonClient.py	(revision 24190)
+++ core/howto/listings/pb/pbAnonClient.py	(working copy)
@@ -1,5 +1,5 @@
 #!/usr/bin/python
-# Copyright (c) 2007 Twisted Matrix Laboratories.
+# Copyright (c) 2007-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 """
Index: core/man/tap2deb.1
===================================================================
--- core/man/tap2deb.1	(revision 24190)
+++ core/man/tap2deb.1	(working copy)
@@ -49,7 +49,7 @@
 .SH "REPORTING BUGS"
 To report a bug, visit \fIhttp://twistedmatrix.com/bugs/\fR
 .SH COPYRIGHT
-Copyright \(co 2000 Matthew W. Lefkowitz
+Copyright \(co 2000-2008 Twisted Matrix Laboratories.
 .br
 This is free software; see the source for copying conditions.  There is NO
 warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Index: core/man/manhole.1
===================================================================
--- core/man/manhole.1	(revision 24190)
+++ core/man/manhole.1	(working copy)
@@ -10,7 +10,7 @@
 .SH "REPORTING BUGS"
 To report a bug, visit \fIhttp://twistedmatrix.com/bugs/\fR
 .SH COPYRIGHT
-Copyright \(co 2000 Matthew W. Lefkowitz
+Copyright \(co 2000-2008 Twisted Matrix Laboratories.
 .br
 This is free software; see the source for copying conditions.  There is NO
 warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Index: core/man/tap2rpm.1
===================================================================
--- core/man/tap2rpm.1	(revision 24190)
+++ core/man/tap2rpm.1	(working copy)
@@ -50,7 +50,7 @@
 .SH "REPORTING BUGS"
 To report a bug, visit \fIhttp://twistedmatrix.com/bugs/\fR
 .SH COPYRIGHT
-Copyright \(co 2000 Matthew W. Lefkowitz
+Copyright \(co 2000-2008 Twisted Matrix Laboratories.
 .br
 This is free software; see the source for copying conditions.  There is NO
 warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Index: core/man/tapconvert.1
===================================================================
--- core/man/tapconvert.1	(revision 24190)
+++ core/man/tapconvert.1	(working copy)
@@ -31,7 +31,7 @@
 .SH "REPORTING BUGS"
 To report a bug, visit \fIhttp://twistedmatrix.com/bugs/\fR
 .SH COPYRIGHT
-Copyright \(co 2000 Matthew W. Lefkowitz
+Copyright \(co 2000-2008 Twisted Matrix Laboratories.
 .br
 This is free software; see the source for copying conditions.  There is NO
 warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Index: core/man/trial.1
===================================================================
--- core/man/trial.1	(revision 24190)
+++ core/man/trial.1	(working copy)
@@ -118,7 +118,7 @@
 .SH "REPORTING BUGS"
 To report a bug, visit \fIhttp://twistedmatrix.com/bugs/\fR
 .SH COPYRIGHT
-Copyright \(co 2003-2007 Twisted Matrix Laboratories
+Copyright \(co 2003-2008 Twisted Matrix Laboratories
 .br
 This is free software; see the source for copying conditions.  There is NO
 warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Index: core/man/mktap.1
===================================================================
--- core/man/mktap.1	(revision 24190)
+++ core/man/mktap.1	(working copy)
@@ -213,7 +213,7 @@
 .SH "REPORTING BUGS"
 To report a bug, visit \fIhttp://twistedmatrix.com/bugs/\fR
 .SH COPYRIGHT
-Copyright \(co 2000 Matthew W. Lefkowitz
+Copyright \(co 2000-2008 Twisted Matrix Laboratories.
 .br
 This is free software; see the source for copying conditions.  There is NO
 warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Index: core/examples/ftpclient.py
===================================================================
--- core/examples/ftpclient.py	(revision 24190)
+++ core/examples/ftpclient.py	(working copy)
@@ -1,5 +1,5 @@
 
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
Index: core/examples/simpleclient.py
===================================================================
--- core/examples/simpleclient.py	(revision 24190)
+++ core/examples/simpleclient.py	(working copy)
@@ -1,5 +1,5 @@
 
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
Index: core/examples/threadedselect/Cocoa/SimpleWebClient/Twistzilla.py
===================================================================
--- core/examples/threadedselect/Cocoa/SimpleWebClient/Twistzilla.py	(revision 24190)
+++ core/examples/threadedselect/Cocoa/SimpleWebClient/Twistzilla.py	(working copy)
@@ -1,4 +1,4 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
Index: core/examples/threadedselect/blockingdemo.py
===================================================================
--- core/examples/threadedselect/blockingdemo.py	(revision 24190)
+++ core/examples/threadedselect/blockingdemo.py	(working copy)
@@ -1,4 +1,4 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
Index: core/examples/simpleserv.py
===================================================================
--- core/examples/simpleserv.py	(revision 24190)
+++ core/examples/simpleserv.py	(working copy)
@@ -1,5 +1,5 @@
 
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
Index: core/examples/pyuidemo.py
===================================================================
--- core/examples/pyuidemo.py	(revision 24190)
+++ core/examples/pyuidemo.py	(working copy)
@@ -1,5 +1,5 @@
 #!/usr/bin/python
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
Index: core/examples/pbecho.py
===================================================================
--- core/examples/pbecho.py	(revision 24190)
+++ core/examples/pbecho.py	(working copy)
@@ -1,5 +1,5 @@
 
-# Copyright (c) 2001-2006 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 from zope.interface import implements
Index: core/examples/pbinterop.py
===================================================================
--- core/examples/pbinterop.py	(revision 24190)
+++ core/examples/pbinterop.py	(working copy)
@@ -1,4 +1,4 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
Index: core/examples/pbsimpleclient.py
===================================================================
--- core/examples/pbsimpleclient.py	(revision 24190)
+++ core/examples/pbsimpleclient.py	(working copy)
@@ -1,5 +1,5 @@
 
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
Index: core/examples/echoclient.py
===================================================================
--- core/examples/echoclient.py	(revision 24190)
+++ core/examples/echoclient.py	(working copy)
@@ -1,5 +1,5 @@
 #!/usr/bin/python
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
Index: core/examples/echoserv.py
===================================================================
--- core/examples/echoserv.py	(revision 24190)
+++ core/examples/echoserv.py	(working copy)
@@ -1,5 +1,5 @@
 #!/usr/bin/python
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 #
Index: core/examples/stdin.py
===================================================================
--- core/examples/stdin.py	(revision 24190)
+++ core/examples/stdin.py	(working copy)
@@ -1,5 +1,5 @@
 
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
Index: core/examples/echoclient_udp.py
===================================================================
--- core/examples/echoclient_udp.py	(revision 24190)
+++ core/examples/echoclient_udp.py	(working copy)
@@ -1,5 +1,5 @@
 #!/usr/bin/python
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
Index: core/examples/echoserv_udp.py
===================================================================
--- core/examples/echoserv_udp.py	(revision 24190)
+++ core/examples/echoserv_udp.py	(working copy)
@@ -1,5 +1,5 @@
 #!/usr/bin/python
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 #
Index: core/examples/pbbenchserver.py
===================================================================
--- core/examples/pbbenchserver.py	(revision 24190)
+++ core/examples/pbbenchserver.py	(working copy)
@@ -1,4 +1,4 @@
-# Copyright (c) 2001-2006 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
Index: core/examples/pbgtk2.py
===================================================================
--- core/examples/pbgtk2.py	(revision 24190)
+++ core/examples/pbgtk2.py	(working copy)
@@ -1,5 +1,5 @@
 
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
Index: core/examples/rotatinglog.py
===================================================================
--- core/examples/rotatinglog.py	(revision 24190)
+++ core/examples/rotatinglog.py	(working copy)
@@ -1,5 +1,5 @@
 
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
Index: core/examples/echoclient_ssl.py
===================================================================
--- core/examples/echoclient_ssl.py	(revision 24190)
+++ core/examples/echoclient_ssl.py	(working copy)
@@ -1,5 +1,5 @@
 #!/usr/bin/python
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
Index: core/examples/pbechoclient.py
===================================================================
--- core/examples/pbechoclient.py	(revision 24190)
+++ core/examples/pbechoclient.py	(working copy)
@@ -1,5 +1,5 @@
 
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
Index: core/examples/echoserv_ssl.py
===================================================================
--- core/examples/echoserv_ssl.py	(revision 24190)
+++ core/examples/echoserv_ssl.py	(working copy)
@@ -1,5 +1,5 @@
 
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
Index: core/examples/wxdemo.py
===================================================================
--- core/examples/wxdemo.py	(revision 24190)
+++ core/examples/wxdemo.py	(working copy)
@@ -1,4 +1,4 @@
-# Copyright (c) 2001-2006 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 """Demo of wxPython integration with Twisted."""
Index: core/examples/wxacceptance.py
===================================================================
--- core/examples/wxacceptance.py	(revision 24190)
+++ core/examples/wxacceptance.py	(working copy)
@@ -1,4 +1,4 @@
-# Copyright (c) 2001-2006 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 """
Index: core/examples/filewatch.py
===================================================================
--- core/examples/filewatch.py	(revision 24190)
+++ core/examples/filewatch.py	(working copy)
@@ -1,4 +1,4 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 #
Index: core/examples/pbsimple.py
===================================================================
--- core/examples/pbsimple.py	(revision 24190)
+++ core/examples/pbsimple.py	(working copy)
@@ -1,5 +1,5 @@
 
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
Index: core/examples/cred.py
===================================================================
--- core/examples/cred.py	(revision 24190)
+++ core/examples/cred.py	(working copy)
@@ -1,5 +1,5 @@
 
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
Index: core/examples/bananabench.py
===================================================================
--- core/examples/bananabench.py	(revision 24190)
+++ core/examples/bananabench.py	(working copy)
@@ -1,4 +1,4 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
Index: core/examples/dbcred.py
===================================================================
--- core/examples/dbcred.py	(revision 24190)
+++ core/examples/dbcred.py	(working copy)
@@ -1,5 +1,5 @@
 #!/usr/bin/python
-# Copyright (c) 2001-2007 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 """
Index: core/examples/testlogging.py
===================================================================
--- core/examples/testlogging.py	(revision 24190)
+++ core/examples/testlogging.py	(working copy)
@@ -1,4 +1,4 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
Index: conch/man/ckeygen.1
===================================================================
--- conch/man/ckeygen.1	(revision 24190)
+++ conch/man/ckeygen.1	(working copy)
@@ -47,7 +47,7 @@
 .SH "REPORTING BUGS"
 To report a bug, visit \fIhttp://twistedmatrix.com/bugs/\fR
 .SH COPYRIGHT
-Copyright \(co 2002 Matthew W. Lefkowitz
+Copyright \(co 2002-2008 Twisted Matrix Laboratories.
 .br
 This is free software; see the source for copying conditions.  There is NO
 warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Index: conch/man/tkconch.1
===================================================================
--- conch/man/tkconch.1	(revision 24190)
+++ conch/man/tkconch.1	(working copy)
@@ -55,7 +55,7 @@
 .SH "REPORTING BUGS"
 To report a bug, visit \fIhttp://twistedmatrix.com/bugs/\fR
 .SH COPYRIGHT
-Copyright \(co 2002 Matthew W. Lefkowitz
+Copyright \(co 2002-2008 Twisted Matrix Laboratories.
 .br
 This is free software; see the source for copying conditions.  There is NO
 warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Index: conch/man/conch.1
===================================================================
--- conch/man/conch.1	(revision 24190)
+++ conch/man/conch.1	(working copy)
@@ -198,7 +198,7 @@
 .Sh "REPORTING BUGS"
 To report a bug, visit \fIhttp://twistedmatrix.com/bugs/\fR
 .Sh COPYRIGHT
-Copyright \(co 2002 Matthew W. Lefkowitz
+Copyright \(co 2002-2008 Twisted Matrix Laboratories.
 .br 
 This is free software; see the source for copying conditions.  There is NO
 warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Index: conch/man/cftp.1
===================================================================
--- conch/man/cftp.1	(revision 24190)
+++ conch/man/cftp.1	(working copy)
@@ -83,7 +83,7 @@
 .Sh "REPORTING BUGS"
 Report bugs to \fIhttp://twistedmatrix.com/bugs/\fR
 .Sh COPYRIGHT
-Copyright \(co 2005 Twisted Matrix Laboratories
+Copyright \(co 2005-2008 Twisted Matrix Laboratories
 .br 
 This is free software; see the source for copying conditions.  There is NO
 warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Index: conch/benchmarks/buffering_mixin.py
===================================================================
--- conch/benchmarks/buffering_mixin.py	(revision 24190)
+++ conch/benchmarks/buffering_mixin.py	(working copy)
@@ -1,4 +1,4 @@
-# Copyright (c) 2006 Twisted Matrix Laboratories.
+# Copyright (c) 2006-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 """
Index: conch/examples/demo_recvline.tac
===================================================================
--- conch/examples/demo_recvline.tac	(revision 24190)
+++ conch/examples/demo_recvline.tac	(working copy)
@@ -1,4 +1,4 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 """Demonstrates line-at-a-time handling with basic line-editing support.
Index: conch/examples/demo_draw.tac
===================================================================
--- conch/examples/demo_draw.tac	(revision 24190)
+++ conch/examples/demo_draw.tac	(working copy)
@@ -1,4 +1,4 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 """A trivial drawing application.
Index: conch/examples/demo_scroll.tac
===================================================================
--- conch/examples/demo_scroll.tac	(revision 24190)
+++ conch/examples/demo_scroll.tac	(working copy)
@@ -1,4 +1,4 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 """Simple echo-ish server that uses the scroll-region.
Index: conch/examples/demo.tac
===================================================================
--- conch/examples/demo.tac	(revision 24190)
+++ conch/examples/demo.tac	(working copy)
@@ -1,4 +1,4 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 """Nearly pointless demonstration of the manhole interactive interpreter.
Index: conch/examples/demo_insults.tac
===================================================================
--- conch/examples/demo_insults.tac	(revision 24190)
+++ conch/examples/demo_insults.tac	(working copy)
@@ -1,4 +1,4 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 """Various simple terminal manipulations using the insults module.
Index: conch/examples/demo_manhole.tac
===================================================================
--- conch/examples/demo_manhole.tac	(revision 24190)
+++ conch/examples/demo_manhole.tac	(working copy)
@@ -1,4 +1,4 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 """An interactive Python interpreter with syntax coloring.
Index: web/examples/soap.py
===================================================================
--- web/examples/soap.py	(revision 24190)
+++ web/examples/soap.py	(working copy)
@@ -1,4 +1,4 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 # 
Index: web/examples/advogato.py
===================================================================
--- web/examples/advogato.py	(revision 24190)
+++ web/examples/advogato.py	(working copy)
@@ -1,4 +1,4 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 #
Index: web/examples/web.py
===================================================================
--- web/examples/web.py	(revision 24190)
+++ web/examples/web.py	(working copy)
@@ -1,5 +1,5 @@
 
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
Index: web/examples/lj.rpy.py
===================================================================
--- web/examples/lj.rpy.py	(revision 24190)
+++ web/examples/lj.rpy.py	(working copy)
@@ -1,4 +1,4 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 #
Index: web/examples/proxy.py
===================================================================
--- web/examples/proxy.py	(revision 24190)
+++ web/examples/proxy.py	(working copy)
@@ -1,5 +1,5 @@
 
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
Index: web/examples/tkwebmon.py
===================================================================
--- web/examples/tkwebmon.py	(revision 24190)
+++ web/examples/tkwebmon.py	(working copy)
@@ -1,4 +1,4 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 #
Index: words/man/im.1
===================================================================
--- words/man/im.1	(revision 24190)
+++ words/man/im.1	(working copy)
@@ -10,7 +10,7 @@
 .SH "REPORTING BUGS"
 To report a bug, visit \fIhttp://twistedmatrix.com/bugs/\fR
 .SH COPYRIGHT
-Copyright \(co 2000 Matthew W. Lefkowitz
+Copyright \(co 2000-2008 Twisted Matrix Laboratories.
 .br
 This is free software; see the source for copying conditions.  There is NO
 warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Index: words/man/t-im.1
===================================================================
--- words/man/t-im.1	(revision 24190)
+++ words/man/t-im.1	(working copy)
@@ -10,7 +10,7 @@
 .SH "REPORTING BUGS"
 To report a bug, visit \fIhttp://twistedmatrix.com/bugs/\fR
 .SH COPYRIGHT
-Copyright \(co 2000 Matthew W. Lefkowitz
+Copyright \(co 2000-2008 Twisted Matrix Laboratories
 .br
 This is free software; see the source for copying conditions.  There is NO
 warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Index: words/examples/msn_example.py
===================================================================
--- words/examples/msn_example.py	(revision 24190)
+++ words/examples/msn_example.py	(working copy)
@@ -1,6 +1,6 @@
 #!/usr/bin/python
 
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 #
Index: words/examples/aimbot.py
===================================================================
--- words/examples/aimbot.py	(revision 24190)
+++ words/examples/aimbot.py	(working copy)
@@ -1,5 +1,5 @@
 #!/usr/bin/python
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 #
Index: words/examples/xmpp_client.py
===================================================================
--- words/examples/xmpp_client.py	(revision 24190)
+++ words/examples/xmpp_client.py	(working copy)
@@ -1,4 +1,4 @@
-# Copyright (c) 2001-2006 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 import sys
Index: words/examples/minchat.py
===================================================================
--- words/examples/minchat.py	(revision 24190)
+++ words/examples/minchat.py	(working copy)
@@ -1,4 +1,4 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
Index: words/examples/ircLogBot.py
===================================================================
--- words/examples/ircLogBot.py	(revision 24190)
+++ words/examples/ircLogBot.py	(working copy)
@@ -1,4 +1,4 @@
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
