Index: twisted/web/test/test_newclient.py
===================================================================
--- twisted/web/test/test_newclient.py	(revision 38215)
+++ twisted/web/test/test_newclient.py	(working copy)
@@ -410,22 +410,37 @@
         """
         header = {}
         finished = []
-        bodyDataFinished = []
+        reason = [] 
         protocol = HTTPClientParser(request, finished.append)
         protocol.headerReceived = header.__setitem__
         body = []
         transport = StringTransport()
         protocol.makeConnection(transport)
         protocol.dataReceived(status)
+
+        class StubConsumer(Protocol):
+            def dataReceived(self, data):
+                pass
+
+            def connectionLost(self, reason=None):
+                pass
+
+        consumer = StubConsumer()
+        protocol.response.deliverBody(consumer)
         protocol.response._bodyDataReceived = body.append
-        protocol.response._bodyDataFinished = (
-            lambda: bodyDataFinished.append(True))
+
+        def connectionLost(failure=None):
+            if failure:
+                reason.append(failure.value.message)
+
+        consumer.connectionLost = connectionLost  
+
         protocol.dataReceived(response)
         self.assertEqual(transport.producerState, 'producing')
         self.assertEqual(protocol.state, DONE)
         self.assertEqual(body, [])
         self.assertEqual(finished, [''])
-        self.assertEquals(bodyDataFinished, [True])
+        self.assertEqual(reason, ["Response with no body fully received"])
         self.assertEqual(protocol.response.length, 0)
         return header
 
Index: twisted/web/_newclient.py
===================================================================
--- twisted/web/_newclient.py	(revision 38215)
+++ twisted/web/_newclient.py	(working copy)
@@ -453,7 +453,8 @@
             or self.request.method == 'HEAD'):
             self.response.length = 0
             self._finished(self.clearLineBuffer())
-            self.response._bodyDataFinished()
+            self.response._bodyDataFinished(
+                Failure(ResponseDone("Response with no body fully received"))) 
         else:
             transferEncodingHeaders = self.connHeaders.getRawHeaders(
                 'transfer-encoding')
