Index: twisted/trial/test/test_assertions.py
===================================================================
--- twisted/trial/test/test_assertions.py	(révision 24660)
+++ twisted/trial/test/test_assertions.py	(copie de travail)
@@ -565,7 +565,9 @@
         """
         A = type('A', (object,), {})
         a = A()
-        self.assertRaises(self.failureException, self.assertNotIsInstance, a, A)
+        error = self.assertRaises(self.failureException,
+                                  self.assertNotIsInstance, a, A)
+        self.assertEquals(str(error), "%r is an instance of %s" % (a, A))
 
     def test_assertNotIsInstanceErrorMultipleClasses(self):
         """
Index: twisted/trial/unittest.py
===================================================================
--- twisted/trial/unittest.py	(révision 24660)
+++ twisted/trial/unittest.py	(copie de travail)
@@ -427,7 +427,7 @@
         @type classOrTuple: class, type, or tuple.
         """
         if isinstance(instance, classOrTuple):
-            self.fail("%r is not an instance of %s" % (instance, classOrTuple))
+            self.fail("%r is an instance of %s" % (instance, classOrTuple))
     assertNotIsInstance = failIfIsInstance
 
 
