--- twisted/conch/test/test_conch.py	2012-01-15 02:23:55.613313858 +0000
+++ twisted/conch/test/test_conch.py	2015-01-18 15:07:14.776638856 +0100
@@ -311,7 +311,8 @@
         self.conchFactory.expectedLoseConnection = 1
         self.conchServer = reactor.listenTCP(0, self.conchFactory,
                                              interface="127.0.0.1")
-        self.echoServer = reactor.listenTCP(0, EchoFactory())
+        self.echoServer = reactor.listenTCP(0, EchoFactory(),
+                                             interface="::")
         self.echoPort = self.echoServer.getHost().port
 
 
@@ -364,6 +365,22 @@
                          % (localPort, self.echoPort))
         d.addCallback(self.assertEqual, 'test\n')
         return d
+
+
+    def test_localToRemoteForwardingV6(self):
+        """
+        Test that we can use whatever client to forward a local port to a
+        specified port on the IPv6 server.
+        """
+        if not OpenSSHClientMixin in self.__class__.__bases__:
+                return True # TODO: should skip this test instead
+        localPort = self._getFreePort()
+        process = ConchTestForwardingProcess(localPort, 'test\n')
+        d = self.execute('', process,
+                         sshArgs='-N -L%i:[::1]:%i'
+                         % (localPort, self.echoPort))
+        d.addCallback(self.assertEqual, 'test\n')
+        return d
 
 
     def test_remoteToLocalForwarding(self):