diff --git twisted/conch/ssh/forwarding.py twisted/conch/ssh/forwarding.py
index 753f994..03b2444 100755
--- twisted/conch/ssh/forwarding.py
+++ twisted/conch/ssh/forwarding.py
@@ -13,6 +13,7 @@ Maintainer: Paul Swartz
 import struct
 
 from twisted.internet import protocol, reactor
+from twisted.internet.endpoints import HostnameEndpoint
 from twisted.python import log
 
 import common, channel
@@ -73,9 +74,10 @@ class SSHConnectForwardingChannel(channel.SSHChannel):
         self.clientBuf = ''
 
     def channelOpen(self, specificData):
-        cc = protocol.ClientCreator(reactor, SSHForwardingClient, self)
         log.msg("connecting to %s:%i" % self.hostport)
-        cc.connectTCP(*self.hostport).addCallbacks(self._setClient, self._close)
+        ep = HostnameEndpoint(reactor, self.hostport[0], self.hostport[1])
+        d = ep.connect(protocol.Factory.forProtocol(lambda: SSHForwardingClient(self)))
+        d.addCallbacks(self._setClient, self._close)
 
     def _setClient(self, client):
         self.client = client
diff --git twisted/conch/test/test_conch.py twisted/conch/test/test_conch.py
index 6e850a2..6a11cdf 100644
--- twisted/conch/test/test_conch.py
+++ twisted/conch/test/test_conch.py
@@ -323,7 +323,7 @@ class ConchServerSetupMixin:
         self.conchFactory.expectedLoseConnection = 1
         self.conchServer = reactor.listenTCP(0, self.conchFactory,
                                              interface="127.0.0.1")
-        self.echoServer = reactor.listenTCP(0, EchoFactory())
+        self.echoServer = reactor.listenTCP(0, EchoFactory(), interface="::")
         self.echoPort = self.echoServer.getHost().port
 
 
@@ -522,6 +522,17 @@ class OpenSSHClientForwardingTests(ForwardingMixin, OpenSSHClientMixin,
     """
     Connection forwarding tests run against the OpenSSL command line client.
     """
+    def test_localToRemoteForwardingV6(self):
+        """
+        Forwarding of arbitrary IPv6 TCP connections via SSH.
+        """
+        localPort = self._getFreePort()
+        process = ConchTestForwardingProcess(localPort, 'test\n')
+        d = self.execute('', process,
+                         sshArgs='-N -L%i:ip6-localhost.us.to:%i'
+                         % (localPort, self.echoPort))
+        d.addCallback(self.assertEqual, 'test\n')
+        return d
 
 
 
diff --git twisted/conch/topfiles/7751.feature twisted/conch/topfiles/7751.feature
new file mode 100644
index 0000000..006249a
--- /dev/null
+++ twisted/conch/topfiles/7751.feature
@@ -0,0 +1 @@
+twisted.conch.ssh.forwarding now supports local->remote forwarding of IPv6
