--- twisted/conch/ssh/forwarding.py	2011-02-14 04:45:15.000000000 +0000
+++ twisted/conch/ssh/forwarding.py	2015-01-18 13:44:16.000000000 +0100
@@ -13,6 +13,7 @@
 import struct
 
 from twisted.internet import protocol, reactor
+from twisted.internet.endpoints import HostnameEndpoint
 from twisted.python import log
 
 import common, channel
@@ -73,9 +74,10 @@
         self.clientBuf = ''
 
     def channelOpen(self, specificData):
-        cc = protocol.ClientCreator(reactor, SSHForwardingClient, self)
         log.msg("connecting to %s:%i" % self.hostport)
-        cc.connectTCP(*self.hostport).addCallbacks(self._setClient, self._close)
+        ep = HostnameEndpoint(reactor, self.hostport[0], self.hostport[1])
+        d = ep.connect(SSHForwardingClientFactory(self))
+        d.addCallbacks(self._setClient, self._close)
 
     def _setClient(self, client):
         self.client = client
@@ -130,6 +132,15 @@
             self.channel = None
 
 
+class SSHForwardingClientFactory(protocol.Factory):
+
+    channel = None
+    def __init__(self, channel):
+        self.channel = channel
+    def buildProtocol(self, addr):
+        return SSHForwardingClient(self.channel)
+
+
 def packOpen_direct_tcpip((connHost, connPort), (origHost, origPort)):
     """Pack the data suitable for sending in a CHANNEL_OPEN packet.
     """
