From f7b145a69f0bb25602a14775c01480fb819f753b Mon Sep 17 00:00:00 2001
From: Phillip Knauss <knauss@.telsasoft.com>
Date: Mon, 21 Jul 2014 15:12:08 -0500
Subject: [PATCH] Specify that default CHALENGE_LIFETIME_SECS is 15 minutes

---
 twisted/cred/credentials.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/twisted/cred/credentials.py b/twisted/cred/credentials.py
index 05ac2a4..3b92c12 100644
--- a/twisted/cred/credentials.py
+++ b/twisted/cred/credentials.py
@@ -179,7 +179,7 @@ class DigestCredentialFactory(object):
     Support for RFC2617 HTTP Digest Authentication
 
     @cvar CHALLENGE_LIFETIME_SECS: The number of seconds for which an
-        opaque should be valid.
+        opaque should be valid. Default is 15 minutes.
 
     @type privateKey: C{str}
     @ivar privateKey: A random string used for generating the secure opaque.
-- 
2.0.0

