Index: twisted/test/test_process.py
===================================================================
--- twisted/test/test_process.py	(revision 23394)
+++ twisted/test/test_process.py	(working copy)
@@ -15,6 +15,7 @@
 import gc
 import warnings
 import socket
+import shutil
 try:
     import fcntl
 except ImportError:
@@ -1819,7 +1820,7 @@
         """
         j = os.path.join
 
-        base = self.mktemp()
+        self.base = base = self.mktemp()
 
         self.foo = j(base, "foo")
         self.baz = j(base, "baz")
@@ -1856,6 +1857,7 @@
                 pass
         else:
             os.environ['PATH'] = self.oldPath
+        shutil.rmtree(self.base)
 
 
     def test_whichWithoutPATH(self):
