Index: twisted/names/authority.py
===================================================================
--- twisted/names/authority.py	(Revision 29347)
+++ twisted/names/authority.py	(Arbeitskopie)
@@ -94,9 +94,11 @@
                     results.append(
                         dns.RRHeader(name, record.TYPE, dns.IN, ttl, record, auth=True)
                     )
-                elif record.TYPE == dns.NS and type != dns.ALL_RECORDS:
+                elif record.TYPE == dns.NS and name.lower() != self.soa[0].lower():
+                    # NS record belong to a child zone: this is a referral.
+                    # As NS records are authoritative in the child zone, ours here are not
                     authority.append(
-                        dns.RRHeader(name, record.TYPE, dns.IN, ttl, record, auth=True)
+                        dns.RRHeader(name, record.TYPE, dns.IN, ttl, record, auth=False)
                     )
                 if record.TYPE == dns.CNAME:
                     cnames.append(
@@ -105,6 +107,13 @@
             if not results:
                 results = cnames
 
+            if not results:
+                # Empty response. Include SOA record to allow clients to cache this
+                # response.
+                authority.append(
+                    dns.RRHeader(self.soa[0], dns.SOA, dns.IN, ttl, self.soa[1], auth=True)
+                    )
+
             for record in results + authority:
                 section = {dns.NS: additional, dns.CNAME: results, dns.MX: additional}.get(record.type)
                 if section is not None:
