Index: ../Twisted/trunk/doc/core/development/policy/svn-dev.xhtml
===================================================================
--- ../Twisted/trunk/doc/core/development/policy/svn-dev.xhtml	(revision 32533)
+++ ../Twisted/trunk/doc/core/development/policy/svn-dev.xhtml	(working copy)
@@ -212,7 +212,7 @@
 <h2>Emacs</h2>
 
 <p>A minor mode for development with Twisted using Emacs is available.  See 
-<code>twisted-dev.el</code>, provided by <a href="http://launchpad.net/twisted-emacs">,
+<code>twisted-dev.el</code>, provided by <a href="http://launchpad.net/twisted-emacs">twisted-emacs</a>,
 for several utility functions which make it easier to grep for methods, run test cases, etc.</p>
 
 <h2>Building Debian packages</h2>
