=== modified file 'INSTALL'
--- INSTALL	2009-12-25 00:22:15 +0000
+++ INSTALL	2010-02-23 17:32:20 +0000
@@ -1,6 +1,6 @@
 Requirements
 
-  Python 2.3 or later.
+  Python 2.4 or later.
 
   Zope Interfaces 3.0.1 (http://zope.org/Products/ZopeInterface) - if
   you have ZopeX3 (at least version 3.0.0c1) installed that should
@@ -33,3 +33,38 @@
     is:
 
       python setup.py install
+
+Supported Platforms
+    
+    Twisted takes the word "support" very seriously.  We only consider a 
+    platform supported if we have a buildbot for it.  Go to
+    http://buildbot.twistedmatrix.com/boxes-supported to view which platforms
+    and versions of Python currently have buildbots.  
+
+    The following platforms were considered supported at the time of release:
+
+    * Apple OS X 10.6
+    * Debian GNU/Linux (ia64 bit)
+    * Fedora (ia32 and ia64 bit)
+    * Microsoft Windows XP (32 bit) - requires SP2 or higher
+    * Ubuntu Linux (ia32 and ia64 bit)
+
+
+    The following platforms are known to have active users in the Twisted 
+    community and are therefore expected to work but did not have active 
+    buildbots available at the time of release:
+    
+    * Apple OS X (versions other than 10.6)
+    * FreeBSD
+    * Microsoft Windows XP (64 bit)
+    * Microsoft Windows Vista
+    * Microsoft Windows 7
+    * Novell SUSE Linux Enterprise Server
+    * OpenSUSE
+    * Red Hat Enterprise Linux
+
+    Bug reports are accepted for all platforms, but the test cases may be
+    hard to reproduce without an proper environment to test them in.  The
+    Twisted team would appreciate additional buildbot client slaves for
+    other platforms or hardware donations.  More info available at
+    http://twistedmatrix.com/trac/wiki/ContributingToTwistedLabs

