diff --git a/twisted/conch/ssh/transport.py b/twisted/conch/ssh/transport.py
--- a/twisted/conch/ssh/transport.py
+++ b/twisted/conch/ssh/transport.py
@@ -1582,5 +1582,6 @@ DISCONNECT_ILLEGAL_USER_NAME = 15
 
 messages = {}
 for name, value in globals().items():
-    if name.startswith('MSG_'):
+    # Avoid legacy messages which overlap with never ones
+    if name.startswith('MSG_') and not name.startswith('MSG_KEXDH_'):
         messages[value] = name
