Index: flavors.py
===================================================================
--- flavors.py	(revision 28294)
+++ flavors.py	(working copy)
@@ -133,8 +133,6 @@
     When a L{pb.BrokerFactory} produces a L{pb.Broker}, it supplies that
     L{pb.Broker} with an object named \"root\".  That object is obtained
     by calling my rootObject method.
-
-    See also: L{pb.getObjectAt}
     """
 
     implements(IPBRoot)
