--- smtp.py.orig	2009-02-01 01:53:00.890625000 +0900
+++ smtp.py	2009-02-01 02:27:37.281250000 +0900
@@ -1653,7 +1653,7 @@
             toEmail = [toEmail]
         self.fromEmail = Address(fromEmail)
         self.nEmails = len(toEmail)
-        self.toEmail = iter(toEmail)
+        self.toEmail = toEmail
         self.file = file
         self.result = deferred
         self.result.addBoth(self._removeDeferred)
@@ -1676,6 +1676,7 @@
         if self.retries < self.sendFinished <= 0:
             log.msg("SMTP Client retrying server. Retry: %s" % -self.retries)
 
+            self.file.seek(0,0)
             connector.connect()
             self.retries += 1
         elif self.sendFinished <= 0:
