diff --git twisted/news/database.py twisted/news/database.py
index b5a1f22..137736a 100644
--- twisted/news/database.py
+++ twisted/news/database.py
@@ -11,7 +11,7 @@ import os
 import StringIO
 from email.Message import Message
 from email.Generator import Generator
-from zope.interface import implementer, Interface
+from zope.interface import implements, Interface
 
 from twisted.news.nntp import NNTPError
 from twisted.mail import smtp
@@ -279,7 +279,6 @@ class _ModerationMixin:
 
 
 
-@implementer(INewsStorage)
 class PickleStorage(_ModerationMixin):
     """
     A trivial NewsStorage implementation using pickles
@@ -288,6 +287,8 @@ class PickleStorage(_ModerationMixin):
     real applications.  Consider yourself warned!
     """
 
+    implements(INewsStorage)
+
     sharedDBs = {}
 
     def __init__(self, filename, groups=None, moderators=(),
@@ -497,12 +498,13 @@ class Group:
         self.articles = {}
 
 
-@implementer(INewsStorage)
 class NewsShelf(_ModerationMixin):
     """
     A NewStorage implementation using Twisted's dirdbm persistence module.
     """
 
+    implements(INewsStorage)
+
     def __init__(self, mailhost, path, sender=None):
         """
         @param mailhost: A C{str} giving the mail exchange host which will
@@ -730,12 +732,13 @@ class NewsShelf(_ModerationMixin):
             return defer.succeed((index, a.getHeader('Message-ID'), StringIO.StringIO(a.body)))
 
 
-@implementer(INewsStorage)
 class NewsStorageAugmentation:
     """
     A NewsStorage implementation using Twisted's asynchronous DB-API
     """
 
+    implements(INewsStorage)
+
     schema = """
 
     CREATE TABLE groups (
