diff --git twisted/python/runtime.py twisted/python/runtime.py
index 159d5dc..6f50e2e 100644
--- twisted/python/runtime.py
+++ twisted/python/runtime.py
@@ -11,11 +11,10 @@ import imp
 
 
 def shortPythonVersion():
-    hv = sys.hexversion
-    major = (hv & 0xff000000L) >> 24
-    minor = (hv & 0x00ff0000L) >> 16
-    teeny = (hv & 0x0000ff00L) >> 8
-    return "%s.%s.%s" % (major,minor,teeny)
+    """
+    Returns the Python version as a dot-separated string.
+    """
+    return "%s.%s.%s" % sys.version_info[:3]
 
 knownPlatforms = {
     'nt': 'win32',
diff --git twisted/python/test/test_runtime.py twisted/python/test/test_runtime.py
index 6d0d7cf..15c18ae 100644
--- twisted/python/test/test_runtime.py
+++ twisted/python/test/test_runtime.py
@@ -7,9 +7,21 @@ Tests for runtime checks.
 
 import sys
 
-from twisted.python.runtime import Platform
+from twisted.python.runtime import Platform, shortPythonVersion
 from twisted.trial.unittest import TestCase
 
+class PythonVersionTests(TestCase):
+    """
+    Tests the shortPythonVersion method.
+    """
+
+    def test_shortPythonVersion(self):
+        """
+        Verify if the Python version is returned correctly.
+        """
+        ver = shortPythonVersion().split('.')
+        for i in range(3):
+            assertEqual(int(ver[i]), sys.version_info[i])
 
 
 class PlatformTests(TestCase):
