diff -r cb14745ec3f5 -r de94b78f190a pytools/webserver/improved_proxy.py
--- a/twisted/web/proxy.py	Thu Oct 03 10:50:20 2013 -0700
+++ b/twisted/web/proxy.py	Thu Oct 03 11:52:11 2013 -0700
@@ -49,6 +49,7 @@
         headers.pop('keep-alive', None)
         self.headers = headers
         self.data = data
+        father.notifyFinish().addErrback(self.fatherFailed)
 
 
     def connectionMade(self):
@@ -89,6 +90,12 @@
             self.transport.loseConnection()
 
 
+    def fatherFailed(self, failure):
+        if not self._finished:
+            self.transport.loseConnection()
+            self._finished = True
+
+
 
 class ProxyClientFactory(ClientFactory):
     """
