Index: doc/core/howto/endpoints.xhtml
===================================================================
--- doc/core/howto/endpoints.xhtml	(revision 40050)
+++ doc/core/howto/endpoints.xhtml	(working copy)
@@ -267,5 +267,19 @@
   also <a href="systemd.xhtml">Deploying Twisted with systemd</a>.
   </li>
 </ul>
+
+<h2>Allowing Quoting in Endpoint Descriptions</h2>
+
+<p>Since Twisted 13.2 <code class="API" base="twisted.internet.endpoints">clientFromString</code>, <code class="API" base="twisted.internet.endpoints">serverFromString</code>, and <code class="API" base="twisted.application">strports.service</code> support the keyword parameter <code>quoting=True</code> to enable an expanded syntax for endpoint descriptions. When enabled, users can quote literals with braces. This is intended to make entering arguments with multiple special characters, such as IPv6 addresses, easier for end users. However, in this syntax it is impossible to escape braces and all opened braces must be closed.</p>
+
+<h3>Normal Syntax</h3>
+<code>tcp6:port=80:interface=2001\:0DB8\:f00e\:eb00\:\:1</code>
+
+<h3>Expanded Syntax</h3>
+<code>tcp6:port=80:interface={2001:0DB8:f00e:eb00::1}</code>
+
+<h3>Invalid Expanded Syntax</h3>
+<code>tcp6:port=80:interface={2001:0DB8:f00e:eb00::1</code>
+
 </body>
 </html>
Index: twisted/application/strports.py
===================================================================
--- twisted/application/strports.py	(revision 40050)
+++ twisted/application/strports.py	(working copy)
@@ -24,7 +24,7 @@
 
     @param description: The description of the listening port, in the syntax
         described by L{twisted.internet.endpoints.serverFromString}.
-    @type description: C{str}
+    @type description: L{str}
 
     @param factory: The protocol factory which will build protocols for
         connections to this service.
@@ -32,14 +32,16 @@
 
     @param default: Do not use this parameter. It has been deprecated since
         Twisted 10.2.0.
-    @type default: C{str} or C{None}
+    @type default: L{str} or L{None}
 
     @param quoting: Whether to allow quoting in the description string or not.
-    @type quoting: C{bool}
+    @type quoting: L{bool}
 
     @return: a 3-tuple of (plugin or name, arguments, keyword arguments)
 
     @see: L{twisted.internet.endpoints.serverFromString}
+
+    @since: 13.2
     """
     return endpoints._parseServer(description, factory, default, quoting)
 
@@ -59,7 +61,7 @@
 
     @param description: The description of the listening port, in the syntax
         described by L{twisted.internet.endpoints.serverFromString}.
-    @type description: C{str}
+    @type description: L{str}
 
     @param factory: The protocol factory which will build protocols for
         connections to this service.
@@ -67,19 +69,21 @@
 
     @param default: Do not use this parameter. It has been deprecated since
         Twisted 10.2.0.
-    @type default: C{str} or C{None}
+    @type default: L{str} or L{None}
 
     @param quoting: Whether to allow quoting in the description string or not.
-    @type quoting: C{bool}
+    @type quoting: L{bool}
 
     @param reactor: The server endpoint will be constructed with this reactor.
-    @type reactor: L{twisted.internet.interfaces.IReactorCore} or C{None}
+    @type reactor: L{twisted.internet.interfaces.IReactorCore} or L{None}
 
     @return: the service corresponding to a description of a reliable
         stream server.
-    @rtype: C{twisted.application.service.IService}
+    @rtype: L{twisted.application.service.IService}
 
     @see: L{twisted.internet.endpoints.serverFromString}
+
+    @since: 13.2
     """
     if reactor is None:
         from twisted.internet import reactor
@@ -106,9 +110,12 @@
     """
     Listen on a port corresponding to a description
 
+    See the documentation of the L{parse} function for description
+    of the semantics of the arguments.
+
     @param description: The description of the listening port, in the syntax
         described by L{twisted.internet.endpoints.serverFromString}.
-    @type description: C{str}
+    @type description: L{str}
 
     @param factory: The protocol factory which will build protocols for
         connections to this service.
@@ -116,17 +123,16 @@
 
     @param default: Do not use this parameter. It has been deprecated since
         Twisted 10.2.0.
-    @type default: C{str} or C{None}
+    @type default: L{str} or L{None}
 
     @param quoting: Whether to allow quoting in the description string or not.
-    @type quoting: C{bool}
+    @type quoting: L{bool}
 
     @return: the port corresponding to a description of a reliable
     virtual circuit server.
-    @rtype: C{twisted.internet.interfaces.IListeningPort}
+    @rtype: L{twisted.internet.interfaces.IListeningPort}
 
-    See the documentation of the C{parse} function for description
-    of the semantics of the arguments.
+    @since: 13.2
     """
     from twisted.internet import reactor
     name, args, kw = parse(description, factory, default, quoting)
Index: twisted/internet/endpoints.py
===================================================================
--- twisted/internet/endpoints.py	(revision 40050)
+++ twisted/internet/endpoints.py	(working copy)
@@ -1260,9 +1260,10 @@
 
     @param description: a string as described by L{serverFromString} or
         L{clientFromString} with quoting disabled.
+    @type description: L{str}
 
-    @return: an iterable of 2-tuples of (L{_OP} or L{_STRING}, string).  Tuples
-        starting with L{_OP} will contain a second element of either ':' (i.e.
+    @return: an iterable of 2-tuples of (C{_OP} or C{_STRING}, string).  Tuples
+        starting with C{_OP} will contain a second element of either ':' (i.e.
         'next parameter') or '=' (i.e. 'assign parameter value').  For example,
         the string 'hello:greet\=ing=world' would result in a generator
         yielding these values::
@@ -1297,9 +1298,10 @@
 
     @param description: a string as described by L{serverFromString} or
         L{clientFromString} with quoting enabled.
+    @type description: L{str}
 
-    @return: an iterable of 2-tuples of (L{_OP} or L{_STRING}, string).  Tuples
-        starting with L{_OP} will contain a second element of either ':' (i.e.
+    @return: an iterable of 2-tuples of (C{_OP} or C{_STRING}, string).  Tuples
+        starting with C{_OP} will contain a second element of either ':' (i.e.
         'next parameter') or '=' (i.e. 'assign parameter value').  For example,
         the string 'hello:{greet=ing}=world' would result in a generator
         yielding these values::
@@ -1335,6 +1337,8 @@
             current += description.next()
         else:
             current += n
+    if depth != 0:
+        raise ValueError("Malformed endpoint description: Unmatched braces.")
     yield _STRING, current
 
 
@@ -1346,13 +1350,15 @@
 
     @param description: a string as described by L{serverFromString} or
         L{clientFromString}.
+    @type description: L{str}
 
     @param quoting: a boolean that determines which tokenization function
         to use.
+    @type quoting: L{bool}
 
     @return: a 2-tuple of C{(args, kwargs)}, where 'args' is a list of all
-        ':'-separated C{str}s not containing an '=' and 'kwargs' is a map of
-        all C{str}s which do contain an '='.  For example, the result of
+        ':'-separated L{str}s not containing an '=' and 'kwargs' is a map of
+        all L{str}s which do contain an '='.  For example, the result of
         C{_parse('a:b:d=1:c')} would be C{(['a', 'b', 'c'], {'d': '1'})}.
     """
     tokenFunction = _tokenizeWithQuoting if quoting else _tokenize
@@ -1396,7 +1402,7 @@
 
     @param description: A description in the format explained by
         L{serverFromString}.
-    @type description: C{str}
+    @type description: L{str}
 
     @param factory: A 'factory' argument; this is left-over from
         twisted.application.strports, it's not really used.
@@ -1405,10 +1411,10 @@
     @param default: Deprecated argument, specifying the default parser mode to
         use for unqualified description strings (those which do not have a ':'
         and prefix).
-    @type default: C{str} or C{NoneType}
+    @type default: L{str} or L{None}
 
     @param quoting: Whether to allow quoting in the description string or not.
-    @type quoting: C{bool}
+    @type quoting: L{bool}
 
     @return: a 3-tuple of (plugin or name, arguments, keyword arguments)
     """
@@ -1449,15 +1455,15 @@
     @type reactor: L{twisted.internet.interfaces.IReactorCore}
 
     @param description: The strports description to parse.
-    @type description: C{str}
+    @type description: L{str}
 
     @param default: Deprecated argument, specifying the default parser mode to
         use for unqualified description strings (those which do not have a ':'
         and prefix).
-    @type default: C{str} or C{NoneType}
+    @type default: L{str} or L{None}
 
     @param quoting: Whether to allow quoting in the description string or not.
-    @type quoting: C{bool}
+    @type quoting: L{bool}
 
     @return: A new endpoint which can be used to listen with the parameters
         given by by C{description}.
@@ -1536,10 +1542,10 @@
     @type reactor: L{twisted.internet.interfaces.IReactorCore}
 
     @param description: The strports description to parse.
-    @type description: C{str}
+    @type description: L{str}
 
     @param quoting: Whether to allow quoting in the description string or not.
-    @type quoting: C{bool}
+    @type quoting: L{bool}
 
     @return: A new endpoint which can be used to listen with the parameters
         given by by C{description}.
@@ -1547,7 +1553,7 @@
 
     @raise ValueError: when the 'description' string cannot be parsed.
 
-    @since: 10.2
+    @since: 13.2
     """
     return _serverFromStringLegacy(reactor, description, _NO_DEFAULT, quoting)
 
@@ -1800,21 +1806,33 @@
         clientFromString(reactor, "unix:/var/foo/bar")
         clientFromString(reactor, "unix:/var/foo/bar:lockfile=1:timeout=9")
 
+    If quoting is set to True, you can declare literals by surrounding them in
+    braces. For example, you can specify the bindAddress for a TCP6 client
+    endpoint like this::
+
+        clientFromString(reactor, "tcp6:www.example.com:80:"
+                                  "bindAddress={::1}", quoting=True)
+
     This function is also extensible; new endpoint types may be registered as
     L{IStreamClientEndpointStringParser} plugins.  See that interface for more
     information.
 
     @param reactor: The client endpoint will be constructed with this reactor.
+    @type reactor: L{twisted.internet.interfaces.IReactorCore}
 
     @param description: The strports description to parse.
+    @type description: L{str}
 
     @param quoting: Whether to allow quoting in the description string or not.
+    @type quoting: L{bool}
 
     @return: A new endpoint which can be used to connect with the parameters
         given by by C{description}.
     @rtype: L{IStreamClientEndpoint<twisted.internet.interfaces.IStreamClientEndpoint>}
 
-    @since: 10.2
+    @raise ValueError: when the 'description' string cannot be parsed.
+
+    @since: 13.2
     """
     args, kwargs = _parse(description, quoting)
     aname = args.pop(0)
Index: twisted/internet/test/test_endpoints.py
===================================================================
--- twisted/internet/test/test_endpoints.py	(revision 40050)
+++ twisted/internet/test/test_endpoints.py	(working copy)
@@ -2317,7 +2317,15 @@
              {'mode': 0o666, 'backlog': 50, 'wantPID': True}))
 
 
+    def test_malformedQuoting(self):
+        """
+        When quoting=True, braces are unescapable and must be properly paired.
+        """
+        self.assertRaises(ValueError, self.parse, "unix:{/var/run/finger\\{}",
+            self.f, quoting=True)
 
+
+
 class ServerStringTests(unittest.TestCase):
     """
     Tests for L{twisted.internet.endpoints.serverFromString}.
Index: twisted/topfiles/6758.feature
===================================================================
--- twisted/topfiles/6758.feature	(revision 0)
+++ twisted/topfiles/6758.feature	(working copy)
@@ -0,0 +1 @@
+twisted.internet.endpoints.serverFromString, twisted.internet.endpoints.clientFromString, and twisted.application.strports.service now have a quoting argument for an enhanced endpoint description syntax.
\ No newline at end of file
Index: twisted/words/topfiles/6758.feature
===================================================================
--- twisted/words/topfiles/6758.feature	(revision 0)
+++ twisted/words/topfiles/6758.feature	(working copy)
@@ -0,0 +1 @@
+twisted.words.protocols.jabber.jstrports.client now has a quoting argument for an enhanced endpoint description syntax.
\ No newline at end of file
