Index: twisted/python/zsh/_mktap
===================================================================
--- twisted/python/zsh/_mktap	(revision 27834)
+++ twisted/python/zsh/_mktap	(working copy)
@@ -154,8 +154,8 @@
 "(-l)--logfile=[Path to web CLF (Combined Log Format) log file.]:logfile:_files -g '*.log'" \
 '(--mime-type)-m[Specify the default mime-type for static files.]:mime-type:_files' \
 '(-m)--mime-type=[Specify the default mime-type for static files.]:mime-type:_files' \
-'(--notracebacks)-n[Display tracebacks in broken web pages. Displaying tracebacks to users may be security risk!]' \
-'(-n)--notracebacks[Display tracebacks in broken web pages. Displaying tracebacks to users may be security risk!]' \
+'(--notracebacks)-n[Do not display tracebacks in broken web pages. Displaying tracebacks to users may be security risk!]' \
+'(-n)--notracebacks[Do not display tracebacks in broken web pages. Displaying tracebacks to users may be security risk!]' \
 '--path=[<path> is either a specific file or a directory to]:path:_files' \
 '--personal[Instead of generating a webserver, generate a ResourcePublisher which listens on ~/.twistd-web-pb]' \
 '(--port)-p[Port to start the server on.]:port:_files' \
Index: twisted/python/zsh/_twistd
===================================================================
--- twisted/python/zsh/_twistd	(revision 27834)
+++ twisted/python/zsh/_twistd	(working copy)
@@ -178,8 +178,8 @@
 "(-l)--logfile=[Path to web CLF (Combined Log Format) log file.]:logfile:_files -g '*.log'" \
 '(--mime-type)-m[Specify the default mime-type for static files.]:mime-type:_files' \
 '(-m)--mime-type=[Specify the default mime-type for static files.]:mime-type:_files' \
-'(--notracebacks)-n[Display tracebacks in broken web pages. Displaying tracebacks to users may be security risk!]' \
-'(-n)--notracebacks[Display tracebacks in broken web pages. Displaying tracebacks to users may be security risk!]' \
+'(--notracebacks)-n[Do not display tracebacks in broken web pages. Displaying tracebacks to users may be security risk!]' \
+'(-n)--notracebacks[Do not display tracebacks in broken web pages. Displaying tracebacks to users may be security risk!]' \
 '--path=[<path> is either a specific file or a directory to]:path:_files' \
 '--personal[Instead of generating a webserver, generate a ResourcePublisher which listens on ~/.twistd-web-pb]' \
 '(--port)-p[Port to start the server on.]:port:_files' \
Index: twisted/web/tap.py
===================================================================
--- twisted/web/tap.py	(revision 27834)
+++ twisted/web/tap.py	(working copy)
@@ -34,7 +34,7 @@
                  "ResourcePublisher which listens on  the port given by "
                  "--port, or ~/%s " % (distrib.UserDirectory.userSocketName,) +
                  "if --port is not specified."],
-                ["notracebacks", "n", "Display tracebacks in broken web pages. " +
+                ["notracebacks", "n", "Do not display tracebacks in broken web pages. " +
                  "Displaying tracebacks to users may be security risk!"],
                 ]
 
