Index: defer-intro.xhtml
===================================================================
--- defer-intro.xhtml	(revision 37049)
+++ defer-intro.xhtml	(working copy)
@@ -549,31 +549,31 @@
 <h2>Conclusion</h2>
 
 <p>
-  You have been introduced to asynchronous code and have seen how to
-  use <code base="twisted.internet.defer" class="API">Deferred</code>s to:
+  Having been introduced to asynchronous code using
+  <code base="twisted.internet.defer" class="API">Deferred</code> objects,
+  you have seen how to use them for...
 
   <ul>
-    <li>Do something after an asynchronous operation completes
+    <li>Doing something after an asynchronous operation completes
     successfully</li>
-    <li>Use the result of a successful asynchronous operation</li>
-    <li>Catch errors in asynchronous operations</li>
-    <li>Do one thing if an operation succeeds, and a different thing if it
+    <li>Using the result of a successful asynchronous operation</li>
+    <li>Catching errors in asynchronous operations</li>
+    <li>Doing one thing if an operation succeeds, and a different thing if it
       fails</li>
-    <li>Do something after an error has been handled successfully</li>
-    <li>Wrap multiple asynchronous operations with one error handler</li>
-    <li>Do something after an asynchronous operation, regardless of whether it
+    <li>Doing something after an error has been handled successfully</li>
+    <li>Wraping multiple asynchronous operations with one error handler</li>
+    <li>Doing something after an asynchronous operation, regardless of whether it
       succeeded or failed</li>
   </ul>
 </p>
 
 <p>
-  This covers the very basic use of
+  These are the very basic uses of
   <code base="twisted.internet.defer" class="API">Deferred</code>.  For
   detailed information about how they work, how to combine multiple Deferreds,
   and how to write code that mixes synchronous and asynchronous APIs, see
-  the <a href="defer.xhtml">reference documentation</a>.  Alternatively, read
-  about how to write functions that <a href="gendefer.xhtml">generate
-  Deferreds</a>.
+  the <a href="defer.xhtml">Deferred reference</a>.  Alternatively, read about
+  how to write functions that <a href="gendefer.xhtml">generate Deferreds</a>.
 </p>
 
 </body>
