diff --git .travis.yml .travis.yml
new file mode 100644
index 0000000..6099175
--- /dev/null
+++ .travis.yml
@@ -0,0 +1,65 @@
+#
+# Travis CI configuration file for Twisted tests.
+#
+language: python
+
+# Supported Python versions.
+python:
+  - "2.6"
+  - "2.7"
+
+# Run default tests on all supported Python versions..
+env:
+    - TEST=py-default
+
+# Start X server for GTK reactor.
+before_install:
+  - "export DISPLAY=:99.0"
+  - "sh -e /etc/init.d/xvfb start"
+
+# Dump trial log on failure.
+after_failure: "cat _trial_temp/test.log"
+
+# In addition, some tests are executed on a single Python version.
+matrix:
+  include:
+    # Python 3.3 tests are only executed once while Twisted is in migration
+    # process.
+    - python: "3.3"
+      env: TEST=py-3.3
+
+    - python: "2.7"
+      env: TEST=py-without-modules
+
+    # Run test with various reactors using a single Python version.
+    - python: "2.7"
+      env: TEST=py-select
+    - python: "2.7"
+      env: TEST=py-poll
+    - python: "2.7"
+      env: TEST=py-epoll
+    - python: "2.7"
+      env: TEST=py-glib2
+    - python: "2.7"
+      env: TEST=py-gi
+    - python: "2.7"
+      env: TEST=py-gtk2
+    - python: "2.7"
+      env: TEST=py-gtk3
+
+    - python: "2.7"
+      env: TEST=pyflakes
+
+    - python: "2.7"
+      env: TEST=api-reference
+    - python: "2.7"
+      env: TEST=documentation
+
+# Each builder has its own install requirements in order
+# to avoid installing unused packages.
+install:
+  - echo "Custom install steps for each builder."
+
+# Command used for executing the tests, where $TEST is extracted from `env`.
+script:
+  - ./bin/admin/run-test-suite $TEST
\ No newline at end of file
diff --git bin/admin/run-test-suite bin/admin/run-test-suite
new file mode 100755
index 0000000..2b3ecb0
--- /dev/null
+++ bin/admin/run-test-suite
@@ -0,0 +1,133 @@
+#!/bin/bash
+#
+# Helper to run predefined test suites.
+#
+
+COMMAND=$1
+
+# Run minimal report to reduce log size.
+TRIAL_COMMAND="./bin/trial --reporter=text --unclean-warnings"
+PYTHON_3_TEST="./admin/run-python3-tests"
+
+REQUIREMENTS="\
+    pyOpenSSL==0.14 \
+    service-identity==0.2 \
+    pycrypto==2.6.1 \
+    pyasn1==0.1.7 \
+    "
+
+#
+# Install project and dependencies with pip.
+#
+pip_install_dependencies() {
+    pip install .
+    pip install $REQUIREMENTS
+    # Removed compiled python files.
+    find . -name "*.pyc" -or -name "*$py.class" | xargs rm -f
+}
+
+
+#
+# Run trial inside the virtual environment.
+#
+run_venv_trial() {
+    pip_install_dependencies
+    echo "Running tests $@..."
+    python $TRIAL_COMMAND $@
+    exit_code=$?
+    return $exit_code
+}
+
+
+#
+# Run trial on using system Python.
+#
+run_system_trial() {
+    echo "Updating packages..."
+    sudo apt-get update -qq
+
+    echo "Installing dependencies..."
+    sudo apt-get install -qq python-gtk2 python-gobject
+
+    echo "Running tests $@..."
+    /usr/bin/python $TRIAL_COMMAND $@
+    exit_code=$?
+
+    return $exit_code
+}
+
+
+case "$COMMAND" in
+    "py-default")
+        run_venv_trial twisted
+        ;;
+
+    "py-without-modules")
+        run_venv_trial --without-module OpenSSL --without-module Crypto twisted
+        ;;
+
+    "py-select-gc")
+        run_venv_trial --force-gc --reactor=select twisted.internet
+        ;;
+
+    "py-select")
+        run_venv_trial --reactor=select twisted.internet
+        ;;
+
+    "py-poll")
+        run_venv_trial --reactor=poll twisted.internet
+        ;;
+
+    "py-epoll")
+        run_venv_trial --reactor=epoll twisted.internet
+        ;;
+
+    "py-glib2")
+        run_system_trial --reactor=glib2 twisted.internet
+        ;;
+
+    "py-gi")
+        run_system_trial --reactor=gi twisted.internet
+        ;;
+
+    "py-gtk2")
+        run_system_trial --reactor=gtk2 twisted.internet
+        ;;
+
+    "py-gtk3")
+        run_system_trial --reactor=gtk3 twisted.internet
+        ;;
+
+    "py-3.3")
+        pip_install_dependencies
+        python3.3 $PYTHON_3_TEST
+        exit_code=$?
+        exit $exit_code
+        ;;
+
+    "pyflakes")
+        pip install pyflakes==0.8.1
+        pyflakes twisted
+        ;;
+
+    "twistedchecker")
+        pip install pep8==1.5.1 twistedchecker==0.2.0
+        twistedchecker twisted
+        ;;
+
+    "documentation")
+        pip install . sphinx==1.2.1
+        python ./bin/admin/build-docs . doc/core/howto/template.
+        ;;
+
+    "api-reference")
+        pip install pydoctor==0.5
+        python ./bin/admin/build-apidocs . apidocs
+        ;;
+
+    *)
+        echo "Unknown command $COMMAND"
+        exit 1
+esac
+
+exit $?
diff --git twisted/internet/test/test_inotify.py twisted/internet/test/test_inotify.py
index a003562..24ea373 100644
--- twisted/internet/test/test_inotify.py
+++ twisted/internet/test/test_inotify.py
@@ -4,6 +4,7 @@
 """
 Tests for the inotify wrapper in L{twisted.internet.inotify}.
 """
+import os
 
 from twisted.internet import defer, reactor
 from twisted.python import filepath, runtime
@@ -16,6 +17,10 @@ except ImportError:
 else:
     from twisted.internet import inotify
 
+skipTravisCI = None
+if os.environ.get('TRAVIS') == 'true':
+    skipTravisCI = 'IN_DELETE_SELF is not working on Travis CI workers.'
+
 
 
 class TestINotify(unittest.TestCase):
@@ -219,6 +224,7 @@ class TestINotify(unittest.TestCase):
 
         return self._notificationTest(
             inotify.IN_DELETE_SELF, operation, expectedPath=self.dirname)
+    test_deleteSelf.skip = skipTravisCI
 
 
     def test_moveSelf(self):
@@ -298,6 +304,8 @@ class TestINotify(unittest.TestCase):
         d = defer.Deferred()
         subdir.createDirectory()
         return d
+    # autoAdd depends IN_DELETE_SELF which is not working on Travis.
+    test_simpleDeleteDirectory.skip = skipTravisCI
 
 
     def test_ignoreDirectory(self):
@@ -411,6 +419,7 @@ class TestINotify(unittest.TestCase):
         expectedPath.remove()
 
         return notified
+    test_seriesOfWatchAndIgnore.skip = skipTravisCI
 
 
     def test_ignoreFilePath(self):
@@ -446,6 +455,7 @@ class TestINotify(unittest.TestCase):
         expectedPath2.remove()
 
         return notified
+    test_ignoreFilePath.skip = skipTravisCI
 
 
     def test_ignoreNonWatchedFile(self):
