Index: twisted/web/server.py
===================================================================
--- twisted/web/server.py	(revision 33693)
+++ twisted/web/server.py	(working copy)
@@ -34,6 +34,15 @@
 from twisted.web.error import UnsupportedMethod
 from twisted.web.microdom import escape
 
+__all__ = [
+    'supportedMethods',
+    'Request',
+    'Session',
+    'Site',
+    'version',
+    'NOT_DONE_YET'
+]
+
 # backwards compatability
 date_time_string = http.datetimeToString
 string_date_time = http.stringToDatetime
