diff --git twisted/mail/tap.py twisted/mail/tap.py
index fe525f3..cd3f6b5 100644
--- twisted/mail/tap.py
+++ twisted/mail/tap.py
@@ -66,7 +66,11 @@ class Options(usage.Options, strcred.AuthOptionMixin):
                                "certificate" : usage.CompleteFiles("*.pem")}
                    )
 
-    longdesc = "This creates a mail.tap file that can be used by twistd."
+    longdesc = ("This creates a mail.tap file that can be used by twistd."
+                "\n\nFor example, you can start a server that accepts emails"
+                " to joe@foo.com and saves them to /tmp/foo.com with:"
+                "\n\n"
+                "  twistd mail -d 'foo.com=/tmp/foo.com' -u joe=password")
 
     def __init__(self):
         usage.Options.__init__(self)
