Index: twisted/web/client.py
===================================================================
--- twisted/web/client.py	(revisione 20822)
+++ twisted/web/client.py	(copia locale)
@@ -127,6 +127,11 @@
                 failure.Failure(
                     error.Error(
                         self.status, self.message, response)))
+        elif self.factory.method == 'HEAD':
+            # server might be broken and return a message-body in the
+            # response
+            #assert not response
+            self.factory.page('')
         elif self.length != None and self.length != 0:
             self.factory.noPage(failure.Failure(
                 PartialDownloadError(self.status, self.message, response)))
Index: twisted/web/test/test_webclient.py
===================================================================
--- twisted/web/test/test_webclient.py	(revisione 20822)
+++ twisted/web/test/test_webclient.py	(copia locale)
@@ -172,6 +172,9 @@
     def testGetPage(self):
         return client.getPage(self.getURL("file")).addCallback(self.assertEquals, "0123456789")
 
+    def testGetPageHead(self):
+        return client.getPage(self.getURL("file"), method='HEAD').addCallback(self.assertEquals, '')
+        
     def testTimeoutNotTriggering(self):
         # Test that when the timeout doesn't trigger, things work as expected.
         d = client.getPage(self.getURL("wait"), timeout=100)
