Index: TwistedWords-0.5.0/twisted/words/xish/utility.py
===================================================================
--- TwistedWords-0.5.0/twisted/words/xish/utility.py	(revision 5110)
+++ TwistedWords-0.5.0/twisted/words/xish/utility.py	(revision 5111)
@@ -209,7 +209,7 @@
             return
 
         event, observers = self._getEventAndObservers(event)
-
+        cbl = None
         if priority not in observers:
             cbl = CallbackList()
             observers[priority] = {event: cbl}
@@ -219,7 +219,8 @@
                 cbl = CallbackList()
                 observers[priority][event] = cbl
 
-        cbl.addCallback(onetime, observerfn, *args, **kwargs)
+        if cbl:
+            cbl.addCallback(onetime, observerfn, *args, **kwargs)
 
     def removeObserver(self, event, observerfn):
         """
