diff --git twisted/test/test_twisted.py twisted/test/test_twisted.py
index 9613114..ebf310d 100644
--- twisted/test/test_twisted.py
+++ twisted/test/test_twisted.py
@@ -630,8 +630,9 @@ class RealZopeInterfaceTests(TestCase, ZopeInterfaceTestsMixin):
                     "Cannot determine system version of zope.interface: %s" % (
                         e,))
             installed = pkg.version
-            versionTuple = tuple(int(part) for part in installed.split('.'))
-            if versionTuple[:len(version)] == version:
+            versionTuple = tuple(
+                int(part) for part in installed.split('.')[:len(version)])
+            if versionTuple == version:
                 pass
             else:
                 raise SkipTest("Mismatched system version of zope.interface")
diff --git twisted/topfiles/6007.misc twisted/topfiles/6007.misc
new file mode 100644
index 0000000..e69de29
