Index: doc/core/howto/trial.xhtml
===================================================================
--- doc/core/howto/trial.xhtml	(revision 36391)
+++ doc/core/howto/trial.xhtml	(working copy)
@@ -114,7 +114,8 @@
 
 <p>With those (one-time) preliminary steps out of the way, let's perform
 the tests. Run <code class="shell">trial calculus.test.test_base_1</code> from the
-command line from the <code class="shell">calculus</code> directory.
+command line when you are in the directory above the 
+<code class="shell">calculus</code> directory.
 
 You should see the following output (though your files are probably not in
  <code class="shell">/tmp</code>:</p>
