Index: twisted/test/test_usage.py
===================================================================
--- twisted/test/test_usage.py	(revision 38152)
+++ twisted/test/test_usage.py	(working copy)
@@ -1,4 +1,3 @@
-
 # Copyright (c) Twisted Matrix Laboratories.
 # See LICENSE for details.
 
@@ -348,6 +347,12 @@
         self.failUnlessIdentical(oBar.subOptions.parent, oBar)
 
 
+class GetUsageOptions(usage.Options):
+    longdesc = ("\nA test documentation string.\n"
+                "This line has more than 80 characters-"
+                "PADDINGXXPADDINGXXPADDINGXXPADDINGXXPADDINGXXPADDINGXX\n")
+
+
 class HelpStringTest(unittest.TestCase):
     def setUp(self):
         """
@@ -379,7 +384,15 @@
         self.failUnless(len(lines) > 0)
         self.failUnless(lines[0].find("flagallicious") >= 0)
 
+    def test_longdescNotWrapped(self):
+        """
+        getUsage does not wrap lines in longdesc.
+        """
+        opt = GetUsageOptions()
+        opt.getUsage(width=80)
+        self.assertTrue(len(opt.longdesc.splitlines()[2]) > 80)
 
+
 class PortCoerceTestCase(unittest.TestCase):
     """
     Test the behavior of L{usage.portCoerce}.
@@ -583,21 +596,3 @@
                 action = cls(None)
             self.assertRaises(NotImplementedError, action._shellCode,
                               None, "bad_shell_type")
-
-class GetUsageOptions(usage.Options):
-    longdesc = """
-A test documentation string. 
-This line has more than 80 characters-PADDINGXXPADDINGXXPADDINGXXPADDINGXXPADDINGXXPADDINGXX
-""" 
-
-class GetUsageTestCase(unittest.TestCase):
-    """
-    Test Options.getUsage for proper output.
-    """
-    def test_longdescNotWrapped(self):
-        """
-        Check that lines in longdesc are not wrapped.
-        """
-        opt = GetUsageOptions()
-        opt.getUsage(width=80)
-        self.assertGreater(len(opt.longdesc.splitlines()[2]), 80)
Index: twisted/python/usage.py
===================================================================
--- twisted/python/usage.py	(revision 38152)
+++ twisted/python/usage.py	(working copy)
@@ -534,7 +534,7 @@
                 longdesc = ''
 
         if longdesc:
-            longdesc = ('\n' + longdesc.strip() + '\n')
+            longdesc = '\n' + longdesc.strip() + '\n'
 
         if optDicts:
             chunks = docMakeChunks(optDicts, width)
