Index: twisted/python/usage.py
===================================================================
--- twisted/python/usage.py	(revision 30564)
+++ twisted/python/usage.py	(working copy)
@@ -497,9 +497,7 @@
                 longdesc = ''
 
         if longdesc:
-            longdesc = ('\n' +
-                        '\n'.join(text.wordWrap(longdesc, width)).strip()
-                        + '\n')
+            longdesc = ('\n' + longdesc.strip() + '\n')
 
         if optDicts:
             chunks = docMakeChunks(optDicts, width)
