Index: twisted/conch/ssh/keys.py
===================================================================
--- twisted/conch/ssh/keys.py	(revision 28138)
+++ twisted/conch/ssh/keys.py	(working copy)
@@ -1,5 +1,5 @@
 # -*- test-case-name: twisted.conch.test.test_keys -*-
-# Copyright (c) 2001-2009 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2010 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 """
@@ -768,7 +768,7 @@
     @rtype:         C{str}
     """
     warnings.warn("makePublicKeyString is deprecated since Twisted Conch 0.9."
-            "  Use Key(obj).toString().",
+            "  Use Key(obj).public().toString().",
             DeprecationWarning, stacklevel=2)
     return Key(obj).public().toString(kind, comment)
 
