Index: options.xhtml
===================================================================
--- options.xhtml	(revision 38186)
+++ options.xhtml	(working copy)
@@ -227,7 +227,7 @@
     <p>Sometimes, just setting an attribute on the basis of the
     options is not flexible enough. In those cases, Twisted does
     not even attempt to provide abstractions such as <q>counts</q> or
-    <q>lists</q>, but rathers lets you call your own method, which will
+    <q>lists</q>, but rather lets you call your own method, which will
     be called whenever the option is encountered.</p>
 
     <p>Here is an example of counting verbosity</p>
