--- Twisted-1.3.0/twisted/python/dispatch.py	2003-11-29 00:30:05.000000000 -0500
+++ Twisted-Local/twisted/python/dispatch.py	2004-12-20 16:59:30.000000000 -0500
@@ -31,11 +31,12 @@
         self.prefix = prefix
         self.callbacks = {}
 
-
     def registerHandler(self, name, meth):
         self.callbacks.setdefault(name, []).append(meth)
 
-
+    def unregisterHandler(self, name, meth):
+        self.callbacks[name].remove(meth)
+        
     def autoRegister(self, obj):
         from twisted.python import reflect
         d = {}
@@ -43,7 +44,13 @@
         for k,v in d.items():
             self.registerHandler(k, v)
 
-
+    def autoUnregister(self, obj):
+        from twisted.python import reflect
+        d = {}
+        reflect.accumulateMethods(obj, d, self.prefix)
+        for k,v in d.items():
+            self.unregisterHandler(k, v)
+            
     def publishEvent(self, name, *args, **kwargs):
         for cb in self.callbacks[name]:
             cb(*args, **kwargs)
