Index: twisted/test/test_process.py
===================================================================
--- twisted/test/test_process.py	(revision 28409)
+++ twisted/test/test_process.py	(working copy)
@@ -7,7 +7,7 @@
 
 import gzip
 import os
-import popen2
+from subprocess import Popen, PIPE
 import sys
 import signal
 import StringIO
@@ -2397,7 +2397,9 @@
     FDTest.skip = skipMessage
 else:
     # do this before running the tests: it uses SIGCHLD and stuff internally
-    lsOut = popen2.popen3("/bin/ls ZZXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX")[2].read()
+    p = Popen("/bin/ls ZZXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX", shell=True,
+          stdin=PIPE, stdout=PIPE, stderr=PIPE, close_fds=True)
+    lsOut = p.stderr.read()
 
 if (runtime.platform.getType() != 'win32') or (not interfaces.IReactorProcess(reactor, None)):
     Win32ProcessTestCase.skip = skipMessage
