diff -ru Twisted-13.1.0-orig/twisted/internet/iocpreactor/abstract.py Twisted-13.1.0/twisted/internet/iocpreactor/abstract.py
--- Twisted-13.1.0-orig/twisted/internet/iocpreactor/abstract.py	2014-05-26 11:22:31.044639688 +1000
+++ Twisted-13.1.0/twisted/internet/iocpreactor/abstract.py	2014-05-26 11:24:35.076634472 +1000
@@ -152,6 +152,7 @@
     _writeDisconnecting = False
     _writeDisconnected = False
     writeBufferSize = 2**2**2**2
+    waitingForCallback = False
 
 
     def loseWriteConnection(self):
@@ -190,6 +191,7 @@
 
 
     def _cbWrite(self, rc, bytes, evt):
+        self.waitingForCallback = False
         if self._handleWrite(rc, bytes, evt):
             self.doWrite()
 
@@ -235,6 +237,9 @@
 
 
     def doWrite(self):
+        if self.waitingForCallback:
+            return
+
         if len(self.dataBuffer) - self.offset < self.SEND_LIMIT:
             # If there is currently less than SEND_LIMIT bytes left to send
             # in the string, extend it with the array data.
@@ -254,6 +259,8 @@
         rc, bytes = self.writeToHandle(buff, evt)
         if rc and rc != ERROR_IO_PENDING:
             self._handleWrite(rc, bytes, evt)
+        else:
+            self.waitingForCallback = True
 
 
     def writeToHandle(self, buff, evt):
Only in Twisted-13.1.0/twisted/internet/iocpreactor: .abstract.py.swp
Only in Twisted-13.1.0/twisted/internet: _ssl.pyc
