Index: twisted/python/util.py
===================================================================
--- twisted/python/util.py	(revision 38002)
+++ twisted/python/util.py	(working copy)
@@ -346,6 +346,18 @@
 def println(*a):
     sys.stdout.write(' '.join(map(str, a))+'\n')
 
+def pretty_format(a,s):
+	""" a is a list of objects
+	    s is the or/and string"""
+        c=""
+        for i in a[:-1]:
+                c+=repr(i)
+                c+=", "
+        c+=s
+        c+=' '
+        c+=repr(a[-1])
+        return c
+
 # XXX
 # This does not belong here
 # But where does it belong?
