Index: twisted/test/test_twistd.py
===================================================================
--- twisted/test/test_twistd.py	(revision 33609)
+++ twisted/test/test_twistd.py	(working copy)
@@ -1547,20 +1547,3 @@
 
 
 
-
-class DeprecationTests(unittest.TestCase):
-    """
-    Tests for deprecated features.
-    """
-
-    def test_initialLog(self):
-        """
-        L{app.initialLog} is deprecated.
-        """
-        logs = []
-        log.addObserver(logs.append)
-        self.addCleanup(log.removeObserver, logs.append)
-        self.callDeprecated(Version("Twisted", 8, 2, 0), app.initialLog)
-        self.assertEqual(len(logs), 2)
-        self.assertIn("starting up", logs[0]["message"][0])
-        self.assertIn("reactor class", logs[1]["message"][0])
Index: twisted/application/app.py
===================================================================
--- twisted/application/app.py	(revision 33609)
+++ twisted/application/app.py	(working copy)
@@ -653,12 +653,6 @@
 
 
 
-def initialLog():
-    AppLogger({})._initialLog()
-initialLog = deprecated(Version("Twisted", 8, 2, 0))(initialLog)
-
-
-
 def convertStyle(filein, typein, passphrase, fileout, typeout, encrypt):
     application = service.loadApplication(filein, typein, passphrase)
     sob.IPersistable(application).setStyle(typeout)
