Index: TwistedCore-2.5.0/twisted/application/app.py
===================================================================
--- TwistedCore-2.5.0/twisted/application/app.py	(revision 5192)
+++ TwistedCore-2.5.0/twisted/application/app.py	(revision 5201)
@@ -18,7 +18,10 @@
 def runWithProfiler(reactor, config):
     """Run reactor under standard profiler."""
     try:
-        import profile
+	try:
+            import cProfile as profile
+        except:
+            import profile
     except ImportError, e:
         s = "Failed to import module profile: %s" % e
         s += """
