diff --git a/twisted/names/srvconnect.py b/twisted/names/srvconnect.py
--- a/twisted/names/srvconnect.py
+++ b/twisted/names/srvconnect.py
@@ -102,11 +102,9 @@ class SRVConnector:
             self.orderedServers.append((a.payload.priority, a.payload.weight,
                                         str(a.payload.target), a.payload.port))
 
-    def _serverCmp(self, a, b):
-        if a[0]!=b[0]:
-            return cmp(a[0], b[0])
-        else:
-            return cmp(a[1], b[1])
+    def _serverKey(self, serv):
+        # Priority, weight
+        return serv[:1]
 
     def pickServer(self):
         assert self.servers is not None
@@ -123,7 +121,7 @@ class SRVConnector:
 
         assert self.servers
 
-        self.servers.sort(self._serverCmp)
+        self.servers.sort(key=self._serverKey)
         minPriority=self.servers[0][0]
 
         weightIndex = zip(xrange(len(self.servers)), [x[1] for x in self.servers
