--- ../Twisted/twisted/mail/imap4.py	2006-07-31 17:00:51.000000000 -0700
+++ ./imap4.py	2006-07-31 17:32:04.000000000 -0700
@@ -1727,7 +1727,12 @@
         if _w is None:
             _w = self.transport.write
         for p in part.part:
-            msg = msg.getSubPart(p)
+            if msg.isMultipart():
+                msg = msg.getSubPart(p)
+            else:
+                if p > 0:
+                    raise TypeError, "Requested subpart of non-multipart message"
+
         if part.header:
             hdrs = msg.getHeaders(part.header.negate, *part.header.fields)
             hdrs = _formatHeaders(hdrs)
