Index: twisted/internet/main.py
===================================================================
--- twisted/internet/main.py	(revision 30564)
+++ twisted/internet/main.py	(working copy)
@@ -32,5 +32,4 @@
     sys.modules['twisted.internet.reactor'] = reactor
 
 
-__all__ = ["CONNECTION_LOST", "CONNECTION_DONE",
-           "ReactorAlreadyInstalledError", "installReactor"]
+__all__ = ["CONNECTION_LOST", "CONNECTION_DONE", "installReactor"]
