diff --git a/doc/core/development/policy/coding-standard.xhtml b/doc/core/development/policy/coding-standard.xhtml
index 1dd4af4..1723970 100644
--- a/doc/core/development/policy/coding-standard.xhtml
+++ b/doc/core/development/policy/coding-standard.xhtml
@@ -36,7 +36,7 @@
 
     <p>Twisted development should always be
      <a href="http://en.wikipedia.org/wiki/Test-driven_development">
-     test-driven</a>.  The complete test suite in trunk@HEAD is required to
+     test-driven</a>.  The complete test suite in the head of the SVN trunk is required to
      be passing on <a href="http://buildbot.twistedmatrix.com/supported">
      supported platforms</a> at all times.  Regressions in the test suite
      are addressed by reverting whatever revisions introduced them.  For
@@ -55,7 +55,7 @@
      the test suite are generally better examples than older parts - check
      when the code you are looking at was written before you use it as an
      example of what you should write).  The names of test modules should
-     begin with <q>test_</q> so that they are automatically discoverable by
+     begin with <code>test_</code> so that they are automatically discoverable by
      test runners such as Trial.  Twisted's unit tests are written using
      <code class="API">twisted.trial</code>, an xUnit library which has been
      extensively customized for use in testing Twisted and Twisted-based
@@ -527,10 +527,10 @@ myref.callRemote("addUser", name="bob", phone="555-1212")
 </pre>
 
     <p>In this case, <code>callRemote</code> (and any code that uses the
-    **kwargs syntax) must be careful to not use <q>name</q>, <q>phone</q>, or
-    any other name that might overlap with a user-provided named parameter.
-    Therefore, <code>callRemote</code> is implemented with the following
-    signature:</p>
+    <code class="python">**kwargs</code> syntax) must be careful to not use
+    <q>name</q>, <q>phone</q>, or any other name that might overlap with
+    a user-provided named parameter.  Therefore, <code>callRemote</code> is
+    implemented with the following signature:</p>
 
 <pre class="python">
 def callRemote(self, _name, *args, **kw):
@@ -545,7 +545,8 @@ def callRemote(self, _name, *args, **kw):
 
     <h2>Special Methods</h2>
 
-    <p>The augmented assignment protocol, defined by __iadd__ and other
+    <p>The augmented assignment protocol, defined by <code
+    class="python">__iadd__</code> and other
     similarly named methods, can be used to allow objects to be modified in
     place or to rebind names if an object is immutable -- both through use
     of the same operator.  This can lead to confusing code, which in turn
diff --git a/doc/core/development/policy/doc-standard.xhtml b/doc/core/development/policy/doc-standard.xhtml
index 68ef7c4..ff0d4ce 100644
--- a/doc/core/development/policy/doc-standard.xhtml
+++ b/doc/core/development/policy/doc-standard.xhtml
@@ -12,7 +12,7 @@
 
     <h2>Allowable Tags</h2>
 
-    <p>Please try to restrict your HTML usage to the following tags (all only for the original logical purpose, and not whatever visual effect you see): <code>&lt;html&gt;</code>, <code>&lt;title&gt;</code>, <code>&lt;head&gt;</code>, <code>&lt;body&gt;</code>, <code>&lt;h1&gt;</code>, <code>&lt;h2</code>, <code>&lt;h3&gt;</code>, <code>&lt;ol&gt;</code>, <code>&lt;ul&gt;</code>, <code>&lt;dl&gt;</code>, <code>&lt;li&gt;</code>,   <code>&lt;dt&gt;</code>, <code>&lt;dd&gt;</code>, <code>&lt;p&gt;</code>, <code>&lt;code&gt;</code>,  <code>&lt;img&gt;</code>,  <code>&lt;blockquote&gt;</code>,  <code>&lt;a&gt;</code>,  <code>&lt;cite&gt;</code>, <code>&lt;div&gt;</code>, <code>&lt;span&gt;</code>, <code>&lt;strong&gt;</code>, <code>&lt;em&gt;</code>, <code>&lt;pre&gt;</code>, <code>&lt;q&gt;</code>, <code>&lt;table&gt;</code>,<code>&lt;tr&gt;</code>, <code>&lt;td&gt;</code> and <code>&lt;th&gt;</code>.</p>
+    <p>Please try to restrict your HTML usage to the following tags (all only for the original logical purpose, and not whatever visual effect you see): <code>&lt;html&gt;</code>, <code>&lt;title&gt;</code>, <code>&lt;head&gt;</code>, <code>&lt;body&gt;</code>, <code>&lt;h1&gt;</code>, <code>&lt;h2</code>, <code>&lt;h3&gt;</code>, <code>&lt;ol&gt;</code>, <code>&lt;ul&gt;</code>, <code>&lt;dl&gt;</code>, <code>&lt;li&gt;</code>,   <code>&lt;dt&gt;</code>, <code>&lt;dd&gt;</code>, <code>&lt;p&gt;</code>, <code>&lt;code&gt;</code>,  <code>&lt;img&gt;</code>,  <code>&lt;blockquote&gt;</code>,  <code>&lt;a&gt;</code>,  <code>&lt;cite&gt;</code>, <code>&lt;div&gt;</code>, <code>&lt;span&gt;</code>, <code>&lt;strong&gt;</code>, <code>&lt;em&gt;</code>, <code>&lt;pre&gt;</code>, <code>&lt;q&gt;</code>, <code>&lt;table&gt;</code>, <code>&lt;tr&gt;</code>, <code>&lt;td&gt;</code> and <code>&lt;th&gt;</code>.</p>
 
     <p>Please avoid using the quote sign (<code>"</code>) for quoting, and use the relevant html tags (<code>&lt;q&gt;&lt;/q&gt;</code>) -- it is impossible to distinguish right and left quotes with the quote sign, and some more sophisticated output methods work better with that distinction.</p>
 
@@ -24,20 +24,25 @@
     and <q>shell</q>. For example:</p>
 
     <h3><q>python</q></h3>
+    <p>Original markup:</p>
+    <blockquote>
 <pre>
-    &lt;p&gt;
-    For example, this is how one defines a Resource:
-    &lt;/p&gt;
+&lt;p&gt;
+For example, this is how one defines a Resource:
+&lt;/p&gt;
 
-    &lt;pre class="python"&gt;
+&lt;pre class="python"&gt;
 from twisted.web import resource
 
 class MyResource(resource.Resource):
     def render_GET(self, request):
         return "Hello, world!"
-    &lt;/pre&gt;
+&lt;/pre&gt;
 </pre>
+    </blockquote>
 
+    <p>Rendered result:</p>
+    <blockquote>
     <p>For example, this is how one defines a Resource:</p>
 <pre class="python">
 from twisted.web import resource
@@ -47,24 +52,30 @@ class MyResource(resource.Resource):
         return "Hello, world!"
    
 </pre>
+    </blockquote>
 
     <p>Note that you should never have leading indentation inside a
     &lt;pre&gt; block -- this makes it hard for readers to
     copy/paste the code.</p>
 
     <h3><q>python-interpreter</q></h3>
+    <p>Original markup:</p>
+    <blockquote>
 <pre>
-    &lt;pre class="python-interpreter"&gt;
-    &amp;gt;&amp;gt;&amp;gt; from twisted.web import resource
-    &amp;gt;&amp;gt;&amp;gt; class MyResource(resource.Resource):
-    ...     def render_GET(self, request):
-    ...         return "Hello, world!"
-    ...
-    &amp;gt;&amp;gt;&amp;gt; MyResource().render_GET(None)
-    "Hello, world!"
-    &lt;/pre&gt;
+&lt;pre class="python-interpreter"&gt;
+&amp;gt;&amp;gt;&amp;gt; from twisted.web import resource
+&amp;gt;&amp;gt;&amp;gt; class MyResource(resource.Resource):
+...     def render_GET(self, request):
+...         return "Hello, world!"
+...
+&amp;gt;&amp;gt;&amp;gt; MyResource().render_GET(None)
+"Hello, world!"
+&lt;/pre&gt;
 </pre>
+    </blockquote>
 
+    <p>Rendered result:</p>
+    <blockquote>
 <pre class="python-interpreter">
 &gt;&gt;&gt; from twisted.web import resource
 &gt;&gt;&gt; class MyResource(resource.Resource):
@@ -74,17 +85,24 @@ class MyResource(resource.Resource):
 &gt;&gt;&gt; MyResource().render_GET(None)
 "Hello, world!"
 </pre>
+    </blockquote>
 
     <h3><q>shell</q></h3>
+    <p>Original markup:</p>
+    <blockquote>
 <pre>
     &lt;pre class="shell"&gt;
     $ twistd web --path /var/www
     &lt;/pre&gt;
 </pre>
+    </blockquote>
 
+    <p>Rendered result:</p>
+    <blockquote>
 <pre class="shell">
 $ twistd web --path /var/www
 </pre>
+    </blockquote>
 
     <h2>Code inside paragraph text</h2>
 
@@ -103,6 +121,8 @@ $ twistd web --path /var/www
     to the module or classname.  This is to help keep the documentation
     clearer and less cluttered by allowing links to API docs that don't
     need the module name.</p>
+    <p>Original markup:</p>
+    <blockquote>
 <pre>
         &lt;p&gt;
     To add a &lt;code class="API"&gt;twisted.web.widgets.Widget&lt;/code&gt;
@@ -120,8 +140,10 @@ $ twistd web --path /var/www
         &lt;/p&gt;
     
 </pre>
+    </blockquote>
 
-<div class="boxed">
+    <p>Rendered result:</p>
+    <blockquote>
         <p>
     To add a <code class="API">twisted.web.widgets.Widget</code>
     instance to a <code class="API" base="twisted.web.widgets">Gadget</code>
@@ -135,8 +157,7 @@ $ twistd web --path /var/www
     which is a
     list.)
         </p>
-
-</div>
+    </blockquote>
 
     <h2>Headers</h2>
 
diff --git a/doc/core/development/policy/svn-dev.xhtml b/doc/core/development/policy/svn-dev.xhtml
index ae4ea97..0ee68f9 100644
--- a/doc/core/development/policy/svn-dev.xhtml
+++ b/doc/core/development/policy/svn-dev.xhtml
@@ -22,10 +22,10 @@ edge.</p>
 
 <h2>Checkout</h2>
 
-<p>Subversion tutorials can be found elsewhere, see in particular
+<p>Subversion tutorials can be found elsewhere, see in particular 
 <a href="http://subversion.apache.org/">the Subversion homepage</a>. The
 relevant data you need to check out a copy of the Twisted tree is available on
-the <a href="http://twistedmatrix.com/trac/wiki/TwistedDevelopment">development
+the <a href="http://twistedmatrix.com/trac/wiki/TwistedDevelopment">development 
 page</a>, and is as follows:</p>
 
 <pre class="shell">
@@ -35,15 +35,15 @@ $ svn co svn://svn.twistedmatrix.com/svn/Twisted/trunk Twisted
 <h2>Alternate tree names</h2>
 
 <p>By using <code>svn co svn://svn.twistedmatrix.com/svn/Twisted/trunk
-otherdir</code>, you can put the workspace tree in a directory other than
+otherdir</code>, you can put the workspace tree in a directory other than 
 <q>Twisted</q>. I do this (with a name like <q>Twisted-Subversion</q>) to
 remind myself that this tree comes from Subversion and not from a released
 version (like <q>Twisted-1.0.5</q>). This practice can cause a few problems,
 because there are a few places in the Twisted tree that need to know where
 the tree starts, so they can add it to <code>sys.path</code> without
 requiring the user manually set their PYTHONPATH. These functions walk the
-current directory up to the root, looking for a directory named
-<q>Twisted</q> (sometimes exactly that, sometimes with a
+current directory up to the root, looking for a directory named 
+<q>Twisted</q> (sometimes exactly that, sometimes with a 
 <code>.startswith</code> test). Generally these are test scripts or other
 administrative tools which expect to be launched from somewhere inside the
 tree (but not necessarily from the top).</p>
@@ -58,31 +58,33 @@ problems.</p>
 
 <h2>Combinator</h2>
 
-<p>In order to simplify the use of Subversion, we typically use
+<p>In order to simplify the use of Subversion, we typically use 
 <a href="http://divmod.org/trac/wiki/DivmodCombinator">Divmod Combinator</a>.
 You may find it to be useful, too.  In particular, because Twisted uses
 branches for almost all feature development, if you plan to contribute to
 Twisted you will probably find Combinator very useful.  For more details,
-see the Combinator website, as well as the
+see the Combinator website, as well as the 
 <a href="http://divmod.org/trac/wiki/UltimateQualityDevelopmentSystem">
 UQDS</a> page.</p>
 
 <h2>Compiling C extensions</h2>
 
 <p>
-There are currently several C extension modules in Twisted:
-twisted.protocols._c_urlarg, twisted.internet.cfsupport,
-twisted.internet.iocpreactor._iocp, and twisted.python._epoll.  These modules
+There are currently several C extension modules in Twisted: 
+<code class="python">twisted.protocols._c_urlarg</code>, 
+<code class="python">twisted.internet.cfsupport</code>, 
+<code class="python">twisted.internet.iocpreactor._iocp</code>, 
+and <code class="python">twisted.python._epoll</code>.  These modules
 are optional, but you'll have to compile them if you want to experience their
 features, performance improvements, or bugs. There are two approaches.
 </p>
 
 <p>The first is to do a regular distutils <code>./setup.py build</code>, which
-will create a directory under <code>build/</code> to hold both the generated
+will create a directory under <code>build/</code> to hold both the generated 
 <code>.so</code> files as well as a copy of the 600-odd <code>.py</code> files
 that make up Twisted. If you do this, you will need to set your PYTHONPATH to
 something like <code>MyDir/Twisted/build/lib.linux-i686-2.5</code> in order to
-run code against the Subversion twisted (as opposed to whatever's installed in
+run code against the Subversion twisted (as opposed to whatever's installed in 
 <code>/usr/lib/python2.5</code> or wherever python usually looks). In
 addition, you will need to re-run the <code>build</code> command <em>every
 time</em> you change a <code>.py</code> file. The <code>build/lib.foo</code>
@@ -95,7 +97,7 @@ PYTHONPATH at the top of the tree, like <code>MyDir/Twisted</code>. This way
 you're using the .py files in place too, removing the confusion a forgotten
 rebuild could cause with the separate build/ directory above. To build the C
 modules in place, do <code>./setup.py build_ext -i</code>. You only need to
-re-run this command when you change the C files. Note that
+re-run this command when you change the C files. Note that 
 <code>setup.py</code> is not Make, it does not always get the dependencies
 right (<code>.h</code> files in particular), so if you are hacking on the
 cReactor you may need to manually delete the <code>.o</code> files before
@@ -127,13 +129,13 @@ do one of:</p>
 <p>This depends upon the <code>.py</code> file having an appropriate
 <q>test-case-name</q> tag that indicates which test cases provide coverage.
 See the <a href="test-standard.xhtml">Test Standards</a> document for
-details about using <q>test-case-name</q>. In this example, the
+details about using <q>test-case-name</q>. In this example, the 
 <code>twisted.mail.test.test_imap</code> test will be run.</p>
 
 <p>Many tests create temporary files in /tmp or ./_trial_temp, but
 everything in /tmp should be deleted when the test finishes. Sometimes these
-cleanup calls are commented out by mistake, so if you see a stray
-/tmp/@12345.1 directory, it is probably from test_dirdbm or test_popsicle.
+cleanup calls are commented out by mistake, so if you see a stray 
+<code>/tmp/@12345.1</code> directory, it is probably from <code>test_dirdbm</code> or <code>test_popsicle</code>.
 Look for an <code>rmtree</code> that has been commented out and complain to
 the last developer who touched that file.</p>
 
@@ -154,7 +156,7 @@ the index file will be placed in <code>doc/howto/index.html</code>.</p>
 ./bin/lore/lore -p --config template=doc/howto/template.tpl doc/howto/*.xhtml
 </pre>
 
-<p>To run hlint over a single Lore document, such as
+<p>To run hlint over a single Lore document, such as 
 <code>doc/development/policy/svn-dev.xhtml</code>, do the following. This is
 useful because the HTML conversion may bail without a useful explanation if
 it sees mismatched tags.</p>
@@ -164,7 +166,7 @@ it sees mismatched tags.</p>
 </pre>
 
 <p>To convert it to HTML (including markup, interpolation of examples,
-footnote processing, etc), do the following. The results will be placed in
+footnote processing, etc), do the following. The results will be placed in 
 <code>doc/development/policy/svn-dev.html</code>:</p>
 
 <pre class="shell">
@@ -176,13 +178,13 @@ footnote processing, etc), do the following. The results will be placed in
 include a <q>-l</q> argument to <code>bin/lore/lore</code>. Links in the
 .xhtml file are to .xhtml targets: when the .xhtml is turned into .html, the
 link targets are supposed to be turned into .html also. In addition to this,
-Lore markup of the form &lt;code class=&quot;API&quot;&gt; is supposed to
+Lore markup of the form <code>&lt;code class=&quot;API&quot;&gt;</code> is supposed to
 turn into a link to the corresponding API reference page. These links will
 probably be wrong unless the correct base URL is provided to Lore.</p>
 
 <h2>Committing and Post-commit Hooks</h2>
 
-<p>Twisted uses a customized
+<p>Twisted uses a customized 
 <a href="http://bazaar.launchpad.net/~exarkun/twisted-trac-integration/trunk/annotate/head%3A/trac-hooks/trac-post-commit-hook">
 trac-post-commit-hook</a> to enable ticket updates based on svn commit
 logs. When making a branch for a ticket, the branch name should end
@@ -205,12 +207,12 @@ Fixes: #9999
 My longer description of the changes made.
 </pre>
 
-<p>The <a href="coding-standard.xhtml#commits">Twisted Coding Standard</a>
+<p>The <a href="coding-standard.xhtml">Twisted Coding Standard</a>
 elaborates on commit messages and source control.</p>
 
 <h2>Emacs</h2>
 
-<p>A minor mode for development with Twisted using Emacs is available.  See
+<p>A minor mode for development with Twisted using Emacs is available.  See 
 <code>emacs/twisted-dev.el</code> for several utility functions which make
 it easier to grep for methods, run test cases, etc.</p>
 
@@ -218,7 +220,7 @@ it easier to grep for methods, run test cases, etc.</p>
 
 <p>Our support for building Debian packages has fallen into disrepair.  We
 would very much like to restore this functionality, but until we do so, if
-you are interested in this, you are on your own.  See
+you are interested in this, you are on your own.  See 
 <a href="http://github.com/astraw/stdeb">stdeb</a> for one possible approach to
 this.</p>
 
diff --git a/doc/core/development/policy/test-standard.xhtml b/doc/core/development/policy/test-standard.xhtml
index 8df03f6..e1efdfa 100644
--- a/doc/core/development/policy/test-standard.xhtml
+++ b/doc/core/development/policy/test-standard.xhtml
@@ -129,7 +129,7 @@ operating systems. The important common factor is that nobody considers
 these limitations to be a bug.</p>
 
 <p>To make it easy to test as much as possible, some tests may be skipped in
-certain situations. Individual test cases can raise the
+certain situations. Individual test cases can raise the 
 <code>SkipTest</code> exception to indicate that they should be skipped, and
 the remainder of the test is not run. In the summary (the very last thing
 printed, at the bottom of the test output) the test is counted as a
@@ -253,7 +253,7 @@ in the issue tracker would be more useful.</p>
 
 <p>Trial provides line coverage information, which is very useful to ensure
 old code has decent coverage. Passing the <code>--coverage</code> option to
-to Trial will generate the coverage information in a file called
+to Trial will generate the coverage information in a file called 
 <code>coverage</code> which can be found in the <code>_trial_temp</code>
 folder. This option requires Python 2.3.3 or newer.</p>
 
@@ -310,7 +310,7 @@ run the right tests.</p>
 
       <li><a href=
       "http://docs.python.org/library/unittest.html"
-      ><code>unittest</code></a> module documentation, in the <a href=
+      >unittest</a> module documentation, in the <a href=
       "http://docs.python.org/library">Python Library
       Reference</a>.</li>
 
