Index: basic.py
===================================================================
RCS file: /cvs/Twisted/twisted/protocols/basic.py,v
retrieving revision 1.42
diff -U3 -r1.42 basic.py
--- basic.py	24 Jul 2003 02:37:03 -0000	1.42
+++ basic.py	23 Aug 2003 10:31:15 -0000
@@ -143,8 +143,11 @@
         """
         
         lines  = (self._buffer+data).split(self.delimiter)
-        self._buffer = lines[-1]
-        for line in lines[:-1]:
+        self._buffer = lines.pop()
+        if len(self._buffer) > self.MAX_LENGTH:
+            line, self._buffer = self._buffer, ''
+            self.lineLengthExceeded(line)
+        for line in lines:
             if self.transport.disconnecting:
                 # this is necessary because the transport may be told to lose
                 # the connection by a line within a larger packet, and it is
