diff --git twisted/conch/endpoints.py twisted/conch/endpoints.py
index 691b9e9..353ef7f 100644
--- twisted/conch/endpoints.py
+++ twisted/conch/endpoints.py
@@ -700,6 +700,7 @@ class _NewConnectionHelper(object):
     establishing a brand new SSH connection, securing it, and authenticating.
     """
     _KNOWN_HOSTS = _KNOWN_HOSTS
+    port = 22
 
     def __init__(self, reactor, hostname, port, command, username, keys,
                  password, agentEndpoint, knownHosts, ui,
@@ -712,7 +713,8 @@ class _NewConnectionHelper(object):
         """
         self.reactor = reactor
         self.hostname = hostname
-        self.port = port
+        if port is not None:
+            self.port = port
         self.command = command
         self.username = username
         self.keys = keys
diff --git twisted/conch/test/test_endpoints.py twisted/conch/test/test_endpoints.py
index 6a9dd0f..2442429 100644
--- twisted/conch/test/test_endpoints.py
+++ twisted/conch/test/test_endpoints.py
@@ -384,6 +384,27 @@ class SSHCommandClientEndpointTestsMixin(object):
         self.assertTrue(verifyObject(IStreamClientEndpoint, endpoint))
 
 
+    def test_defaultPort(self):
+        """
+        Passing C{None} for C{port} makes the endpoint use the default port
+        number.
+        """
+        endpoint = SSHCommandClientEndpoint.newConnection(
+            self.reactor, b"dummy command", b"dummy user",
+            self.hostname, port=None)
+        self.assertEqual(22, endpoint._creator.port)
+
+
+    def test_specifiedPort(self):
+        """
+        Passing a C{port} makes the endpoint use it.
+        """
+        endpoint = SSHCommandClientEndpoint.newConnection(
+            self.reactor, b"dummy command", b"dummy user",
+            self.hostname, port=2222)
+        self.assertEqual(2222, endpoint._creator.port)
+
+
     def test_channelOpenFailure(self):
         """
         If a channel cannot be opened on the authenticated SSH connection, the
diff --git twisted/conch/topfiles/6631.bugfix twisted/conch/topfiles/6631.bugfix
new file mode 100644
index 0000000..6090234
--- /dev/null
+++ twisted/conch/topfiles/6631.bugfix
@@ -0,0 +1 @@
+twisted.conch.endpoints.SSHCommandClientEndpoint now uses the standard SSH port if no port is specified.
