Index: twisted/test/test_internet.py
===================================================================
--- twisted/test/test_internet.py	(revision 13746)
+++ twisted/test/test_internet.py	(working copy)
@@ -4,7 +4,10 @@
 
 from twisted.trial import unittest
 from twisted.trial.util import spinUntil
-from twisted.internet import reactor, protocol, error, app, abstract, interfaces
+from twisted.internet import reactor, protocol, error, app, abstract
+from twisted.internet import interfaces
+from twisted.internet.utils import getProcessOutput
+
 try:
     from twisted.internet import ssl
 except ImportError:
@@ -16,6 +19,7 @@
 from twisted.python import util, threadable, log
 threadable.init(1)
 
+import os
 import sys
 import time
 import threading
@@ -487,6 +491,53 @@
         self.assertEquals(dcall.active(), 0)
 
 
+resolve_helper = """
+import %(reactor)s
+%(reactor)s.install()
+from twisted.internet import reactor
+
+class Foo:
+    def __init__(self):
+        reactor.callLater(1, self.start)
+        self.timer = reactor.callLater(3, self.failed)
+    def start(self):
+        reactor.resolve('localhost').addBoth(self.done)
+    def done(self, res):
+        print 'done', res
+        reactor.stop()
+    def failed(self):
+        print 'failed'
+        self.timer = None
+        reactor.stop()
+f = Foo()
+reactor.run()
+"""
+
+class Resolve(unittest.TestCase):
+    # this uses t.i.util.getProcessOutput, and maybe it wants to live in
+    # test_iutils.py instead of here
+    def testChildResolve(self):
+        # I've seen problems with reactor.run under gtk2reactor. Spawn a
+        # child which just does reactor.resolve after a second of delay, and
+        # fail if it does not complete in a timely fashion.
+        helper = os.path.abspath(self.mktemp())
+        helperf = open(helper, 'wt')
+        reactorname = reactor.__module__
+        helperf.write(resolve_helper % {'reactor': reactorname})
+        helperf.close()
+        env = os.environ.copy()
+        env['PYTHONPATH'] = os.pathsep.join(sys.path)
+        d = getProcessOutput(sys.executable, (helper,), env, errortoo=1)
+        d.addCallback(self._testChildResolve_1)
+        return d
+    testChildResolve.timeout = 10
+    def _testChildResolve_1(self, res):
+        if res.startswith("failed\n"):
+            print "The child process timed out. Output is:"
+            print res
+            self.fail("The child process timed out.")
+        self.failUnlessEqual(res, "done 127.0.0.1\n")
+
 class Counter:
     index = 0
 
Index: twisted/internet/gtk2reactor.py
===================================================================
--- twisted/internet/gtk2reactor.py	(revision 13746)
+++ twisted/internet/gtk2reactor.py	(working copy)
@@ -40,6 +40,16 @@
 from twisted.python import log, threadable, runtime, failure, components
 from twisted.internet.interfaces import IReactorFDSet
 
+if hasattr(gobject, "threads_init"):
+    # initialize threads early, otherwise timers added before the reactor
+    # starts cause all sorts of problems
+    
+    # recent versions of python-gtk expose this. python-gtk=2.4.1
+    # (wrapping glib-2.4.7) does. python-gtk=2.0.0 (wrapping
+    # glib-2.2.3) does not.
+    gobject.threads_init()
+    threadable.init(1)
+
 # Sibling Imports
 from twisted.internet import main, posixbase, error, selectreactor
 
@@ -92,14 +102,6 @@
             self.__iteration = gtk.main_iteration
             self.__crash = _our_mainquit
             self.__run = gtk.main
-
-    def _initThreads(self):
-        if hasattr(gobject, "threads_init"):
-            # recent versions of python-gtk expose this. python-gtk=2.4.1
-            # (wrapping glib-2.4.7) does. python-gtk=2.0.0 (wrapping
-            # glib-2.2.3) does not.
-            gobject.threads_init()
-        posixbase.PosixReactorBase._initThreads(self)
     
     # The input_add function in pygtk1 checks for objects with a
     # 'fileno' method and, if present, uses the result of that method
