Index: twisted/internet/base.py
===================================================================
--- twisted/internet/base.py	(revision 13679)
+++ twisted/internet/base.py	(working copy)
@@ -256,7 +256,8 @@
 
         self.usingThreads = 0
         if platform.supportsThreads():
-            threadable.whenThreaded(self._initThreads)
+            self._initThreads()
+            # threadable.whenThreaded(self._initThreads)
 
     # override in subclasses
 
