=== added file 'twisted/topfiles/5554.removal'
--- twisted/topfiles/5554.removal	1970-01-01 00:00:00 +0000
+++ twisted/topfiles/5554.removal	2012-06-14 05:07:08 +0000
@@ -0,0 +1,1 @@
+twisted.trial.runner.DocTestCase is removed.

=== modified file 'twisted/trial/runner.py'
--- twisted/trial/runner.py	2012-03-14 22:44:18 +0000
+++ twisted/trial/runner.py	2012-06-14 05:01:43 +0000
@@ -11,7 +11,7 @@
 __all__ = [
     'suiteVisit', 'TestSuite',
 
-    'DestructiveTestSuite', 'DocTestCase', 'DryRunVisitor',
+    'DestructiveTestSuite', 'DryRunVisitor',
     'ErrorHolder', 'LoggedSuite', 'PyUnitTestCase',
     'TestHolder', 'TestLoader', 'TrialRunner', 'TrialSuite',
 
@@ -239,22 +239,6 @@
 
 
 
-class DocTestCase(PyUnitTestCase):
-    """
-    DEPRECATED in Twisted 8.0.
-    """
-
-    def id(self):
-        """
-        In Python 2.4, doctests have correct id() behaviour. In Python 2.3,
-        id() returns 'runit'.
-
-        Here we override id() so that at least it will always contain the
-        fully qualified Python name of the doctest.
-        """
-        return self._test.shortDescription()
-
-
 class TrialSuite(TestSuite):
     """
     Suite to wrap around every single test in a C{trial} run. Used internally

