Index: test_tls.py
===================================================================
--- test_tls.py	(revision 26940)
+++ test_tls.py	(working copy)
@@ -7,6 +7,8 @@
 
 try:
     from twisted.protocols.tls import TLSMemoryBIOProtocol, TLSMemoryBIOFactory
+    from twisted.internet.ssl import ClientContextFactory, PrivateCertificate
+    from twisted.internet.ssl import DefaultOpenSSLContextFactory
 except ImportError:
     # Skip the whole test module if it can't be imported.
     skip = "pyOpenSSL 0.10 or newer required for twisted.protocol.tls"
@@ -21,8 +23,6 @@
 from twisted.internet.error import ConnectionDone
 from twisted.internet.defer import Deferred, gatherResults
 from twisted.internet.protocol import Protocol, ClientFactory, ServerFactory
-from twisted.internet.ssl import ClientContextFactory, PrivateCertificate
-from twisted.internet.ssl import DefaultOpenSSLContextFactory
 from twisted.protocols.loopback import loopbackAsync, collapsingPumpPolicy
 from twisted.trial.unittest import TestCase
 from twisted.test.test_tcp import ConnectionLostNotifyingProtocol
@@ -563,3 +563,4 @@
             self.assertTrue(clientProtocol.transport.q.disconnect)
         handshakeDeferred.addCallback(cbConnectionDone)
         return handshakeDeferred
+
