--- guard.py		2003-09-01 09:31:41.000000000 +1200
+++ guard-new.py	2003-09-01 09:47:00.000000000 +1200
@@ -162,7 +162,16 @@
             cookieKey = "woven_session_" + _sessionCookie()
         self.cookieKey = cookieKey
         self.sessions = {}
+        self.lifetime = 1800
 
+    def setLifetime(self, lifetime):
+        """Set the approximate lifetime for wrapped sessions.
+
+        See  the L{GuardSession.setLifetime<twisted.web.woven.guard.GuardSession>}
+        documentation for discussion of "approximate".
+        """
+        self.lifetime = lifetime
+        
     def render(self, request):
         return redirectTo(addSlash(request), request)
 
@@ -186,7 +195,7 @@
         setupURL = request.setupSessionURL = urlToChild(INIT_SESSION, *([path]+request.postpath))
         request.setupSession = lambda: Redirect(setupURL)
         if self.sessions.has_key(path):
-            self.sessions[path].setLifetime(1800)
+            self.sessions[path].setLifetime(self.lifetime)
             if cookie == path:
                 # /sessionized-url/aef9c34aecc3d9148/foo
                 #                  ^
