From 56e23a9524fa26f520fdf44dceb98ce036dfe7a6 Mon Sep 17 00:00:00 2001
From: doadin <tbkizle@gmail.com>
Date: Tue, 8 Sep 2015 03:28:52 -0400
Subject: [PATCH] Update _pollingfile.py for zope.interface python3
 compatibility

---
 twisted/internet/_pollingfile.py | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/twisted/internet/_pollingfile.py b/twisted/internet/_pollingfile.py
index ff6481a..e2e064d 100644
--- a/twisted/internet/_pollingfile.py
+++ b/twisted/internet/_pollingfile.py
@@ -7,7 +7,7 @@
 select() - this is pretty much only useful on Windows.
 """
 
-from zope.interface import implements
+from zope.interface import implementer
 
 from twisted.internet.interfaces import IConsumer, IPushProducer
 
@@ -105,7 +105,7 @@ def _pollEvent(self):
 
 class _PollableReadPipe(_PollableResource):
 
-    implements(IPushProducer)
+    implementer(IPushProducer)
 
     def __init__(self, pipe, receivedCallback, lostCallback):
         # security attributes for pipes
@@ -161,7 +161,7 @@ def resumeProducing(self):
 
 class _PollableWritePipe(_PollableResource):
 
-    implements(IConsumer)
+    implementer(IConsumer)
 
     def __init__(self, writePipe, lostCallback):
         self.disconnecting = False
