Index: twisted/conch/test/test_window.py
===================================================================
--- twisted/conch/test/test_window.py	(revision 27909)
+++ twisted/conch/test/test_window.py	(working copy)
@@ -5,7 +5,7 @@
 
 from twisted.trial.unittest import TestCase
 
-from twisted.conch.insults.window import TopWindow
+from twisted.conch.insults.window import TopWindow, ScrolledArea, TextOutput
 
 
 class TopWindowTests(TestCase):
@@ -47,3 +47,9 @@
         root.repaint()
         self.assertEqual(len(paints), 1)
         self.assertEqual(len(scheduled), 1)
+
+
+class ScrolledAreaTests(TestCase):
+    def test_CanInstantiate(self):
+        ScrolledArea(TextOutput())
+
Index: twisted/conch/insults/window.py
===================================================================
--- twisted/conch/insults/window.py	(revision 27909)
+++ twisted/conch/insults/window.py	(working copy)
@@ -730,7 +730,7 @@
 
 class ScrolledArea(Widget):
     def __init__(self, containee):
-        Widget.__init__(self, containee)
+        Widget.__init__(self)
         self._viewport = Viewport(containee)
         self._horiz = HorizontalScrollbar(self._horizScroll)
         self._vert = VerticalScrollbar(self._vertScroll)
