Index: twisted/protocols/basic.py
===================================================================
--- twisted/protocols/basic.py	(revision 38110)
+++ twisted/protocols/basic.py	(working copy)
@@ -39,16 +39,6 @@
 
 
 
-LENGTH, DATA, COMMA = range(3)
-NUMBER = re.compile(b'(\d*)(:?)')
-
-deprecatedSince = versions.Version("Twisted", 10, 2, 0)
-message = "NetstringReceiver parser state is private."
-for attr in ["LENGTH", "DATA", "COMMA", "NUMBER"]:
-    deprecate.deprecatedModuleAttribute(
-        deprecatedSince, message, __name__, attr)
-del deprecatedSince, message, attr
-
 DEBUG = 0
 
 class NetstringParseError(ValueError):
Index: twisted/protocols/test/test_basic.py
===================================================================
--- twisted/protocols/test/test_basic.py	(revision 38110)
+++ twisted/protocols/test/test_basic.py	(working copy)
@@ -688,37 +688,7 @@
         self.assertRaises(NotImplementedError, proto.stringReceived, 'foo')
 
 
-    def test_deprecatedModuleAttributes(self):
-        """
-        Accessing one of the old module attributes used by the
-        NetstringReceiver parser emits a deprecation warning.
-        """
-        basic.LENGTH, basic.DATA, basic.COMMA, basic.NUMBER
-        warnings = self.flushWarnings(
-            offendingFunctions=[self.test_deprecatedModuleAttributes])
 
-        self.assertEqual(len(warnings), 4)
-        for warning in warnings:
-            self.assertEqual(warning['category'], DeprecationWarning)
-        self.assertEqual(
-            warnings[0]['message'],
-            ("twisted.protocols.basic.LENGTH was deprecated in Twisted 10.2.0: "
-             "NetstringReceiver parser state is private."))
-        self.assertEqual(
-            warnings[1]['message'],
-            ("twisted.protocols.basic.DATA was deprecated in Twisted 10.2.0: "
-             "NetstringReceiver parser state is private."))
-        self.assertEqual(
-            warnings[2]['message'],
-            ("twisted.protocols.basic.COMMA was deprecated in Twisted 10.2.0: "
-             "NetstringReceiver parser state is private."))
-        self.assertEqual(
-            warnings[3]['message'],
-            ("twisted.protocols.basic.NUMBER was deprecated in Twisted 10.2.0: "
-             "NetstringReceiver parser state is private."))
-
-
-
 class IntNTestCaseMixin(LPTestCaseMixin):
     """
     TestCase mixin for int-prefixed protocols.
