--- ftp.py.old	2011-02-23 16:48:39.000000000 +0000
+++ ftp.py	2011-02-22 13:34:26.000000000 +0000
@@ -1115,6 +1115,8 @@
         def ebSent(err):
             log.msg("Unexpected error receiving file from client:")
             log.err(err)
+            if isinstance(err.value, FTPCmdError):
+                return (err.value.errorCode, err.value.errorMessage)
             return (CNX_CLOSED_TXFR_ABORTED,)
 
         def cbConsumer(cons):
