Index: doc/core/howto/tutorial/intro.xhtml
===================================================================
--- doc/core/howto/tutorial/intro.xhtml	(revision 27020)
+++ doc/core/howto/tutorial/intro.xhtml	(working copy)
@@ -111,7 +111,17 @@
 However, usually people actually expect some data about users to be
 transmitted.</p>
 
+<p>In order to test the finger server, which you just wrote, you may want to
+telnet to the server and input an username. Something like this:</p>
 
+<pre class="shell">
+% telnet localhost 1079
+Connected to localhost.
+glyph
+Connection closed by foreign host.
+</pre>
+
+
 <h2>Read Username, Output Error, Drop Connections</h2>
 
 <a href="listings/finger/finger05.py" class="py-listing">finger05.py</a>
@@ -274,4 +284,4 @@
 the standard finger command: <code>finger moshez</code>.</p>
 
 </body>
-</html>
\ No newline at end of file
+</html>
