--- /home/users/pradu/formmethod.py	2003-09-29 11:58:16.000000000 +0200
+++ formmethod.py	2003-09-29 11:50:32.000000000 +0200
@@ -46,6 +46,7 @@
 
     def __init__(self, name, default=None, shortDesc=None,
                  longDesc=None, hints=None, allowNone=1):
+	print self.__class__.__name__, default
         self.name = name
         self.allowNone = allowNone
         if default is None:
@@ -83,8 +84,8 @@
     
     def __init__(self, name, default=None, shortDesc=None,
                  longDesc=None, hints=None, allowNone=1, min=0, max=None):
-        Argument.__init__(self, name, default=None, shortDesc=None,
-                          longDesc=None, hints=None, allowNone=1)
+        Argument.__init__(self, name, default, shortDesc,
+                          longDesc, hints, allowNone)
         self.min = min
         self.max = max
     
