Index: twisted/names/common.py
===================================================================
--- twisted/names/common.py	(revision 27345)
+++ twisted/names/common.py	(working copy)
@@ -201,9 +201,9 @@
         if r.type == dns.NS:
             from twisted.names import client
             r = client.Resolver(servers=[(str(r.payload.name), dns.PORT)])
-            return r.lookupAddress(str(name)
-                ).addCallback(lambda (ans, auth, add): extractRecord(r, name, ans + auth + add, level - 1)
-                ).addBoth(lambda passthrough: (r.protocol.transport.stopListening(), passthrough)[1])
+            return r.lookupAddress(str(name)).addCallback(
+                lambda (ans, auth, add): extractRecord(r, name, ans + auth + add, level - 1)
+                )
 
 typeToMethod = {
     dns.A:     'lookupAddress',
