Index: twisted/protocols/ftp.py
===================================================================
--- twisted/protocols/ftp.py	(revision 24576)
+++ twisted/protocols/ftp.py	(working copy)
@@ -1246,7 +1246,7 @@
 
     def __init__(self, portal=None, userAnonymous='anonymous'):
         self.portal = portal
-        self.userAnonymous = 'anonymous'
+        self.userAnonymous = userAnonymous
         self.instances = []
 
     def buildProtocol(self, addr):
