Index: twisted/test/test_ftp.py
===================================================================
--- twisted/test/test_ftp.py	(revision 24576)
+++ twisted/test/test_ftp.py	(working copy)
@@ -63,6 +63,7 @@
     @ivar clientFactory: class used as ftp client.
     """
     clientFactory = ftp.FTPClientBasic
+    userAnonymous = "anonymous"
 
     def setUp(self):
         # Create a directory
@@ -73,7 +74,8 @@
         p = portal.Portal(ftp.FTPRealm(self.directory))
         p.registerChecker(checkers.AllowAnonymousAccess(),
                           credentials.IAnonymous)
-        self.factory = ftp.FTPFactory(portal=p)
+        self.factory = ftp.FTPFactory(portal=p,
+                                      userAnonymous=self.userAnonymous)
         port = reactor.listenTCP(0, self.factory, interface="127.0.0.1")
         self.addCleanup(port.stopListening)
 
@@ -140,7 +142,30 @@
             ['230 Anonymous login ok, access restrictions apply.'],
             chainDeferred=d)
 
+class FTPAnonymousTestCase(FTPServerTestCase):
+    """Simple tests for an FTP server with different anonymous username.
+    The new anonymous username used in this test case is "guest"
+    """
+    userAnonymous = "guest"
 
+    def setUp(self):
+        # Running the FTPServerTestCase setUp, with different anonymous username
+        return FTPServerTestCase.setUp(self)
+
+    def test_anonymousLogin(self):
+        """Tests whether the changing of the anonymous username is working or
+        not. The FTP server should not comply about the need of password for
+        the username 'guest', letting it login as anonymous asking just an
+        email address as password.
+        """
+        d = self.assertCommandResponse(
+            'USER guest',
+            ['331 Guest login ok, type your email address as password.'])
+        return self.assertCommandResponse(
+            'PASS test@twistedmatrix.com',
+            ['230 Anonymous login ok, access restrictions apply.'],
+            chainDeferred=d)
+
 class BasicFTPServerTestCase(FTPServerTestCase):
     def testNotLoggedInReply(self):
         """When not logged in, all commands other than USER and PASS should
