diff --git a/twisted/internet/_sslverify.py b/twisted/internet/_sslverify.py
--- a/twisted/internet/_sslverify.py
+++ b/twisted/internet/_sslverify.py
@@ -103,7 +103,7 @@ class DistinguishedName(dict):
         lablen = 0
         def uniqueValues(mapping):
             return dict.fromkeys(mapping.itervalues()).keys()
-        for k in uniqueValues(_x509names):
+        for k in sorted(uniqueValues(_x509names)):
             label = util.nameToLabel(k)
             lablen = max(len(label), lablen)
             v = getattr(self, k, None)
diff --git a/twisted/test/test_sslverify.py b/twisted/test/test_sslverify.py
--- a/twisted/test/test_sslverify.py
+++ b/twisted/test/test_sslverify.py
@@ -232,22 +232,22 @@ class OpenSSLOptions(unittest.TestCase):
         self.assertEqual(
             c.inspect().split('\n'),
             ["Certificate For Subject:",
-             "  Organizational Unit Name: Security",
-             "         Organization Name: Twisted Matrix Labs",
              "               Common Name: example.twistedmatrix.com",
-             "    State Or Province Name: Massachusetts",
              "              Country Name: US",
              "             Email Address: nobody@twistedmatrix.com",
              "             Locality Name: Boston",
+             "         Organization Name: Twisted Matrix Labs",
+             "  Organizational Unit Name: Security",
+             "    State Or Province Name: Massachusetts",
              "",
              "Issuer:",
-             "  Organizational Unit Name: Security",
-             "         Organization Name: Twisted Matrix Labs",
              "               Common Name: example.twistedmatrix.com",
-             "    State Or Province Name: Massachusetts",
              "              Country Name: US",
              "             Email Address: nobody@twistedmatrix.com",
              "             Locality Name: Boston",
+             "         Organization Name: Twisted Matrix Labs",
+             "  Organizational Unit Name: Security",
+             "    State Or Province Name: Massachusetts",
              "",
              "Serial Number: 12345",
              "Digest: C4:96:11:00:30:C3:EC:EE:A3:55:AA:ED:8C:84:85:18",
