Index: doc/web/howto/client.xhtml
===================================================================
--- doc/web/howto/client.xhtml	(revision 28088)
+++ doc/web/howto/client.xhtml	(working copy)
@@ -177,7 +177,7 @@
       class="API">twisted.web.http.PotentialDataLoss</code> exception: this
       indicates that the server framed the response such that there is no way
       to know when the entire response body has been received.  Only
-      HTTP/1.0} servers should behave this way.  Finally, it is possible for
+      HTTP/1.0 servers should behave this way.  Finally, it is possible for
       the exception to be of another type, indicating guaranteed data loss for
       some reason (a lost connection, a memory error, etc).
     </p>
