--- trunk\client.py	2005-12-27 16:49:42.781250000 +0100
+++ patched\client.py	2005-12-27 17:07:27.609375000 +0100
@@ -28,6 +28,14 @@
 
 class HTTPPageGetter(http.HTTPClient):
 
+    # First class headers are sent before others to comply with broken proxies 
+    # and/or ease request dispatching.
+    _firstClassHeaders = {
+        'host': None, 
+        'user-agent': None, 
+        'cookie': None, 
+        'content-length': None
+    }
     quietLoss = 0
     followRedirect = 1 
     failed = 0
@@ -46,7 +54,7 @@
         if data is not None:
             self.sendHeader("Content-Length", str(len(data)))
         for (key, value) in self.factory.headers.items():
-            if key.lower() != "content-length":
+            if key.lower() not in self._firstClassHeaders:
                 # we calculated it on our own
                 self.sendHeader(key, value)
         self.endHeaders()
