diff --git a/twisted/python/deprecate.py b/twisted/python/deprecate.py
index f4b31b4..93b4655 100644
--- a/twisted/python/deprecate.py
+++ b/twisted/python/deprecate.py
@@ -512,9 +512,7 @@ def warnAboutFunction(offender, warningString):
 
     @since: 11.0
     """
-    # inspect.getmodule() is attractive, but somewhat
-    # broken in Python < 2.6.  See Python bug 4845.
-    offenderModule = sys.modules[offender.__module__]
+    offenderModule = inspect.getmodule(offender)
     filename = inspect.getabsfile(offenderModule)
     lineStarts = list(findlinestarts(offender.func_code))
     lastLineNo = lineStarts[-1][1]
