Index: twisted/python/test/test_dist.py
===================================================================
--- twisted/python/test/test_dist.py	(revision 34466)
+++ twisted/python/test/test_dist.py	(working copy)
@@ -8,6 +8,7 @@
 
 import os
 import shutil
+import sys
 
 from distutils.core import Distribution
 
@@ -88,6 +89,16 @@
         setup.setContent(self.setupTemplate % (name,))
 
 
+    def writeEmptySetup(self, *path):
+        """
+        Write out an empty C{setup.py} file to a location determined by
+        L{self.basedir} and L{path}.
+        """
+        outdir = self.basedir.descendant(path)
+        outdir.makedirs()
+        outdir.child("setup.py").setContent("")
+
+
     def assertExtensions(self, expected):
         """
         Assert that the given names match the (sorted) names of discovered
@@ -127,7 +138,26 @@
         self.assertExtensions([])
 
 
+    def test_getExtensionsNotSupportedOnJava(self):
+        """
+        Extensions are not supported on Java-based platforms.
+        """
+        self.addCleanup(setattr, sys, "platform", sys.platform)
+        sys.platform = "java"
+        self.writeSetup("twisted.sorcery", "topfiles")
+        self.assertExtensions([])
 
+
+    def test_getExtensionsExtensionsLocalIsOptional(self):
+        """
+        It is acceptable for extensions to not define the C{extensions} local
+        variable.
+        """
+        self.writeEmptySetup("twisted.necromancy", "topfiles")
+        self.assertExtensions([])
+
+
+
 class GetVersionTest(TestCase):
     """
     Tests for L{dist.getVersion}.
