Index: doc/core/howto/pb-intro.xhtml
===================================================================
--- doc/core/howto/pb-intro.xhtml	(revision 37979)
+++ doc/core/howto/pb-intro.xhtml	(working copy)
@@ -159,7 +159,7 @@
   perspective directly is being called on behalf of the user who is
   represented by that perspective.  (Services with unusual
   requirements for <q>on behalf of</q>, such as simulations with the
-  ability to posess another player's avatar, are accomplished by
+  ability to posses another player's avatar, are accomplished by
   providing indirected access to another user's perspective.)
 
   </p>
