Index: twisted/news/topfiles/7703.misc
===================================================================
Index: twisted/news/test/test_news.py
===================================================================
--- twisted/news/test/test_news.py	(revision 43485)
+++ twisted/news/test/test_news.py	(working copy)
@@ -29,7 +29,7 @@
  10:56pm up 4 days, 4:42, 1 user, load average: 0.08, 0.08, 0.12
 """ % (MESSAGE_ID)
 
-class NewsTestCase(unittest.TestCase):
+class NewsTests(unittest.TestCase):
     def setUp(self):
         self.backend = database.NewsShelf(None, 'news2.db')
         self.backend.addGroup('alt.test.nntp', 'y')
Index: twisted/news/test/test_nntp.py
===================================================================
--- twisted/news/test/test_nntp.py	(revision 43485)
+++ twisted/news/test/test_nntp.py	(working copy)
@@ -100,7 +100,7 @@
         raise AssertionError("fetchArticle() failed: %s" % (error,))
 
 
-class NNTPTestCase(unittest.TestCase):
+class NNTPTests(unittest.TestCase):
     def setUp(self):
         self.server = nntp.NNTPServer()
         self.server.factory = self
