--- abstract.py.old	Fri Sep 30 00:51:09 2005
+++ abstract.py	Fri Sep 30 00:51:39 2005
@@ -164,7 +164,7 @@
         ready for writing.
         """
         if isinstance(data, unicode): # no, really, I mean it
-            raise TypeError("Data must be not be unicode")
+            raise TypeError("Data must not be unicode")
         if not self.connected or self._writeDisconnected:
             return
         if data:
