Index: twisted/mail/tap.py
===================================================================
--- twisted/mail/tap.py	(revision 32422)
+++ twisted/mail/tap.py	(working copy)
@@ -234,6 +234,9 @@
                 self.addEndpoint(
                     'pop3', self['pop3s'], certificate=self['certificate'])
 
+        for ch in self['credCheckers']:
+            self.service.smtpPortal.registerChecker(ch)
+
         if not self['disable-anonymous']:
             self.service.smtpPortal.registerChecker(checkers.AllowAnonymousAccess())
 
