Index: twisted/mail/test/test_options.py
===================================================================
--- twisted/mail/test/test_options.py	(revision 33706)
+++ twisted/mail/test/test_options.py	(working copy)
@@ -187,7 +187,21 @@
         self.assertEqual("--esmtp requires --hostname", str(exc))
 
 
+    def test_auth(self):
+        """
+        Tests that the --auth option registers a checker.
+        """
+        options = Options()
+        options.parseOptions(['--auth', 'memory:admin:admin:bob:password'])
+        self.assertEqual(len(options['credCheckers']), 1)
+        checker = options['credCheckers'][0]
+        interfaces = checker.credentialInterfaces
+        registered_checkers = options.service.smtpPortal.checkers
+        for iface in interfaces:
+            self.assertEqual(checker, registered_checkers[iface])
 
+
+
 class SpyEndpoint(object):
     """
     SpyEndpoint remembers what factory it is told to listen with.
Index: twisted/mail/tap.py
===================================================================
--- twisted/mail/tap.py	(revision 33706)
+++ twisted/mail/tap.py	(working copy)
@@ -240,6 +240,12 @@
         if self['esmtp'] and self['hostname'] is None:
             raise usage.UsageError("--esmtp requires --hostname")
 
+        # If the --auth option was passed, this will be present -- otherwise,
+        # it won't be, which is also a perfectly valid state.
+        if 'credCheckers' in self:
+            for ch in self['credCheckers']:
+                self.service.smtpPortal.registerChecker(ch)
+
         if not self['disable-anonymous']:
             self.service.smtpPortal.registerChecker(checkers.AllowAnonymousAccess())
 
