From 0efb44c69231ff10fd7794163579905c2148151e Mon Sep 17 00:00:00 2001
From: Vadim Markovtsev <v.markovtsev@samsung.com>
Date: Fri, 23 May 2014 10:01:07 +0400
Subject: [PATCH 4/5] Add Python 3 specific checks to isIPAddress and
 isIPv6Address

---
 twisted/internet/abstract.py | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/twisted/internet/abstract.py b/twisted/internet/abstract.py
index 1e49053..9758fe0 100644
--- a/twisted/internet/abstract.py
+++ b/twisted/internet/abstract.py
@@ -493,6 +493,8 @@ def isIPAddress(addr):
     @return: C{True} if C{addr} represents an IPv4 address, C{False}
     otherwise.
     """
+    if _PY3 and isinstance(addr, bytes):
+        raise TypeError("addr must be a string, not bytes")
     dottedParts = addr.split('.')
     if len(dottedParts) == 4:
         for octet in dottedParts:
@@ -519,6 +521,8 @@ def isIPv6Address(addr):
         otherwise.
     @rtype: C{bool}
     """
+    if _PY3 and isinstance(addr, bytes):
+        raise TypeError("addr must be a string, not bytes")
     if '%' in addr:
         addr = addr.split('%', 1)[0]
     if not addr:
-- 
1.9.3
