Index: twisted/protocols/amp.py
===================================================================
--- twisted/protocols/amp.py	(revision 27345)
+++ twisted/protocols/amp.py	(working copy)
@@ -755,7 +755,7 @@
         this command, or None, as specified by requiresAnswer.
         """
         box = Box(kw)
-        return self._sendBoxCommand(command, box)
+        return self._sendBoxCommand(command, box, requiresAnswer)
 
 
     def callRemote(self, commandType, *a, **kw):
Index: twisted/test/test_amp.py
===================================================================
--- twisted/test/test_amp.py	(revision 27345)
+++ twisted/test/test_amp.py	(working copy)
@@ -1074,6 +1074,16 @@
         self.assertEquals(L[0]['Print'], HELLO_UNICODE)
 
 
+    def test_callRemoteStringRequiresAnswerFalse(self):
+        """
+        L{BoxDispatcher.callRemoteString} returns C{None} if C{requiresAnswer}
+        is C{False}.
+        """
+        c, s, p = connectedServerAndClient()
+        ret = c.callRemoteString("WTF", requiresAnswer=False)
+        self.assertIdentical(ret, None)
+
+
     def test_unknownCommandLow(self):
         """
         Verify that unknown commands using low-level APIs will be rejected with an
