commit 1a25be1bff6654c1de0bc434e9472f3d0c90a32b
Author: Wolfgang Rohdewald <wolfgang@rohdewald.de>
Date:   Sat Sep 27 14:47:46 2014 +0200

    spread.jelly: deprecate _unjelly_instance

diff --git twisted/spread/jelly.py twisted/spread/jelly.py
index 4173d28..de5eb80 100644
--- twisted/spread/jelly.py
+++ twisted/spread/jelly.py
@@ -867,21 +867,32 @@ class _Unjellier:
         else:
             return Unpersistable("Persistent callback not found")
 
-
-    def _unjelly_instance(self, rest):
-        clz = self.unjelly(rest[0])
-        if type(clz) is not types.ClassType:
-            raise InsecureJelly("Instance found with non-class class.")
-        if hasattr(clz, "__setstate__"):
-            inst = _newInstance(clz, {})
-            state = self.unjelly(rest[1])
-            inst.__setstate__(state)
-        else:
-            state = self.unjelly(rest[1])
-            inst = _newInstance(clz, state)
-        if hasattr(clz, 'postUnjelly'):
-            self.postCallbacks.append(inst.postUnjelly)
-        return inst
+    if not _PY3:
+        def _unjelly_instance(self, rest):
+            """
+            (internal) Unjelly an instance.
+            I am unused. No test ever comes here.
+            I am never called directly, only by the unjellier.
+            But the jellier never jellies an instance, so the jellier
+            cannot call me.
+            When removing me, also remove instance_atom
+            """
+            warnings.warn("_unjelly_instance() is deprecated since Twisted 14.1.0 and unsupported for Python 3"
+                DeprecationWarning, stacklevel=2)
+
+            clz = self.unjelly(rest[0])
+            if type(clz) is not types.ClassType:
+                raise InsecureJelly("Instance found with non-class class.")
+            if hasattr(clz, "__setstate__"):
+                inst = _newInstance(clz, {})
+                state = self.unjelly(rest[1])
+                inst.__setstate__(state)
+            else:
+                state = self.unjelly(rest[1])
+                inst = _newInstance(clz, state)
+            if hasattr(clz, 'postUnjelly'):
+                self.postCallbacks.append(inst.postUnjelly)
+            return inst
 
 
     def _unjelly_unpersistable(self, rest):
diff --git twisted/spread/topfiles/7653.removal twisted/spread/topfiles/7653.removal
new file mode 100644
index 0000000..368c45b
--- /dev/null
+++ twisted/spread/topfiles/7653.removal
@@ -0,0 +1 @@
+twisted.spread.jelly._unjelly_instance is now deprecated. It was unused and will not be ported to Python 3
