commit 1552fd235f4f3ca4566390290f5cd9fee0c03df9
Author: Wolfgang Rohdewald <wolfgang@rohdewald.de>
Date:   Mon Sep 29 11:37:24 2014 +0200

    Ticket 7653. spread.jelly: deprecate unjellying instances

diff --git twisted/spread/jelly.py twisted/spread/jelly.py
index 4173d28..dfff192 100644
--- twisted/spread/jelly.py
+++ twisted/spread/jelly.py
@@ -109,6 +109,9 @@ from twisted.persisted.crefutil import _Container
 
 from twisted.spread.interfaces import IJellyable, IUnjellyable
 
+from twisted.python.deprecate import deprecatedModuleAttribute
+from twisted.python.versions import Version
+
 DictTypes = (DictionaryType,)
 
 None_atom = "None"                  # N
@@ -134,6 +137,11 @@ instance_atom = 'instance'          # i
 frozenset_atom = 'frozenset'
 
 
+deprecatedModuleAttribute(
+    Version("Twisted", 14, 1, 0),
+    "jelly.instance_atom is unused within Twisted.",
+    "twisted.spread.jelly", "instance_atom")
+
 # errors
 unpersistable_atom = "unpersistable"# u
 unjellyableRegistry = {}
@@ -869,6 +877,18 @@ class _Unjellier:
 
 
     def _unjelly_instance(self, rest):
+        """
+        (internal) Unjelly an instance.
+        Unused.
+        Never called directly, only by the unjellier.
+        But the jellier never jellies an instance,
+        so the unjellier cannot call this.
+        When removing this, also remove instance_atom
+        """
+        warnings.warn("Unjellying an instance is deprecated since Twisted 14.1.0"
+            " and unsupported for Python 3",
+            DeprecationWarning)
+
         clz = self.unjelly(rest[0])
         if type(clz) is not types.ClassType:
             raise InsecureJelly("Instance found with non-class class.")
diff --git twisted/spread/topfiles/7653.removal twisted/spread/topfiles/7653.removal
new file mode 100644
index 0000000..d17f012
--- /dev/null
+++ twisted/spread/topfiles/7653.removal
@@ -0,0 +1 @@
+twisted.spread.jelly's support for unjellying "instance" atoms is now deprecated.
diff --git twisted/test/test_jelly.py twisted/test/test_jelly.py
index 184dc35..a8b0978 100644
--- twisted/test/test_jelly.py
+++ twisted/test/test_jelly.py
@@ -620,6 +620,48 @@ class JellyTestCase(unittest.TestCase):
 
 
 
+class JellyDeprecationTests(unittest.TestCase):
+    """
+    Tests for deprecated Jelly things
+    """
+
+    def test_deprecatedInstanceAtom(self):
+        """
+        Unjellying the instance atom is deprecated with 14.1.0.
+        Twisted itself does not jelly them anymore since at
+        least 2003.
+        """
+        jelly.instance_atom
+        warnings = self.flushWarnings()
+        self.assertEqual(len(warnings), 1)
+        self.assertEqual(
+            warnings[0]['message'],
+            'twisted.spread.jelly.instance_atom was deprecated in Twisted 14.1.0:'
+            ' jelly.instance_atom is unused within Twisted.')
+        self.assertEqual(
+            warnings[0]['category'],
+            DeprecationWarning)
+
+
+    def test_deprecatedUnjellyingInstanceAtom(self):
+        """
+        Unjellying the instance atom is deprecated with 14.1.0.
+        Twisted itself does not jelly them anymore since at
+        least 2003.
+        """
+        jelly.unjelly(["instance", ["class", "twisted.test.test_jelly.A"], ["dictionary"]])
+        warnings = self.flushWarnings()
+        self.assertEqual(len(warnings), 1)
+        self.assertEqual(
+            warnings[0]['message'],
+            'Unjellying an instance is deprecated since Twisted 14.1.0'
+            ' and unsupported for Python 3')
+        self.assertEqual(
+            warnings[0]['category'],
+            DeprecationWarning)
+
+
+
 class ClassA(pb.Copyable, pb.RemoteCopy):
     def __init__(self):
         self.ref = ClassB(self)
