diff --git twisted/conch/ssh/keys.py twisted/conch/ssh/keys.py
index bbd5f96..12836c9 100644
--- twisted/conch/ssh/keys.py
+++ twisted/conch/ssh/keys.py
@@ -391,13 +391,13 @@ class Key(object):
         methods.
         """
         if data.startswith('ssh-'):
-            return 'public_openssh'
+            return 'PUBLIC_OPENSSH'
         elif data.startswith('-----BEGIN'):
-            return 'private_openssh'
+            return 'PRIVATE_OPENSSH'
         elif data.startswith('{'):
-            return 'public_lsh'
+            return 'PUBLIC_LSH'
         elif data.startswith('('):
-            return 'private_lsh'
+            return 'PRIVATE_LSH'
         elif data.startswith('\x00\x00\x00\x07ssh-'):
             ignored, rest = common.getNS(data)
             count = 0
@@ -405,9 +405,9 @@ class Key(object):
                 count += 1
                 ignored, rest = common.getMP(rest)
             if count > 4:
-                return 'agentv3'
+                return 'AGENTV3'
             else:
-                return 'blob'
+                return 'BLOB'
     _guessStringType = classmethod(_guessStringType)
 
 
diff --git twisted/conch/topfiles/6806.misc twisted/conch/topfiles/6806.misc
new file mode 100644
index 0000000..069cb6c
--- /dev/null
+++ twisted/conch/topfiles/6806.misc
@@ -0,0 +1 @@
+t.conch.ssh.keys._guessStringType now returns uppercase values.
