diff --git twisted/conch/ssh/keys.py twisted/conch/ssh/keys.py
index bbd5f96..4c74f16 100644
--- twisted/conch/ssh/keys.py
+++ twisted/conch/ssh/keys.py
@@ -390,14 +390,15 @@ class Key(object):
         Guess the type of key in data.  The types map to _fromString_*
         methods.
         """
+        retval = None
         if data.startswith('ssh-'):
-            return 'public_openssh'
+            retval = 'public_openssh'
         elif data.startswith('-----BEGIN'):
-            return 'private_openssh'
+            retval = 'private_openssh'
         elif data.startswith('{'):
-            return 'public_lsh'
+            retval = 'public_lsh'
         elif data.startswith('('):
-            return 'private_lsh'
+            retval = 'private_lsh'
         elif data.startswith('\x00\x00\x00\x07ssh-'):
             ignored, rest = common.getNS(data)
             count = 0
@@ -405,9 +406,11 @@ class Key(object):
                 count += 1
                 ignored, rest = common.getMP(rest)
             if count > 4:
-                return 'agentv3'
+                retval = 'agentv3'
             else:
-                return 'blob'
+                retval = 'blob'
+
+        return retval.upper() if retval else retval
     _guessStringType = classmethod(_guessStringType)
 
 
diff --git twisted/conch/test/test_keys.py twisted/conch/test/test_keys.py
index 0923974..fe9174e 100644
--- twisted/conch/test/test_keys.py
+++ twisted/conch/test/test_keys.py
@@ -137,31 +137,31 @@ class KeyTestCase(unittest.TestCase):
         correctly.
         """
         self.assertEqual(keys.Key._guessStringType(keydata.publicRSA_openssh),
-                'public_openssh')
+                'public_openssh'.upper())
         self.assertEqual(keys.Key._guessStringType(keydata.publicDSA_openssh),
-                'public_openssh')
+                'public_openssh'.upper())
         self.assertEqual(keys.Key._guessStringType(
-            keydata.privateRSA_openssh), 'private_openssh')
+            keydata.privateRSA_openssh), 'private_openssh'.upper())
         self.assertEqual(keys.Key._guessStringType(
-            keydata.privateDSA_openssh), 'private_openssh')
+            keydata.privateDSA_openssh), 'private_openssh'.upper())
         self.assertEqual(keys.Key._guessStringType(keydata.publicRSA_lsh),
-                'public_lsh')
+                'public_lsh'.upper())
         self.assertEqual(keys.Key._guessStringType(keydata.publicDSA_lsh),
-                'public_lsh')
+                'public_lsh'.upper())
         self.assertEqual(keys.Key._guessStringType(keydata.privateRSA_lsh),
-                'private_lsh')
+                'private_lsh'.upper())
         self.assertEqual(keys.Key._guessStringType(keydata.privateDSA_lsh),
-                'private_lsh')
+                'private_lsh'.upper())
         self.assertEqual(keys.Key._guessStringType(
-            keydata.privateRSA_agentv3), 'agentv3')
+            keydata.privateRSA_agentv3), 'agentv3'.upper())
         self.assertEqual(keys.Key._guessStringType(
-            keydata.privateDSA_agentv3), 'agentv3')
+            keydata.privateDSA_agentv3), 'agentv3'.upper())
         self.assertEqual(keys.Key._guessStringType(
             '\x00\x00\x00\x07ssh-rsa\x00\x00\x00\x01\x01'),
-            'blob')
+            'blob'.upper())
         self.assertEqual(keys.Key._guessStringType(
             '\x00\x00\x00\x07ssh-dss\x00\x00\x00\x01\x01'),
-            'blob')
+            'blob'.upper())
         self.assertEqual(keys.Key._guessStringType('not a key'),
                 None)
 
diff --git twisted/conch/topfiles/6806.misc twisted/conch/topfiles/6806.misc
new file mode 100644
index 0000000..5ac299c
--- /dev/null
+++ twisted/conch/topfiles/6806.misc
@@ -0,0 +1 @@
+twisted.conch.ssh.keys.Key._guessStringtype now returns an upper case string or None.
