Index: twisted/mail/pop3.py
===================================================================
--- twisted/mail/pop3.py	(revision 30174)
+++ twisted/mail/pop3.py	(working copy)
@@ -506,7 +506,7 @@
         self._userIs = user
         self.successResponse('USER accepted, send PASS')
 
-    def do_PASS(self, password):
+    def do_PASS(self, password=''):
         if self._userIs is None:
             self.failResponse("USER required before PASS")
             return
