Index: twisted/mail/test/test_pop3.py
===================================================================
--- twisted/mail/test/test_pop3.py	(revision 34673)
+++ twisted/mail/test/test_pop3.py	(working copy)
@@ -409,7 +409,7 @@
         return d.addCallback(self._cbTestEmptyPASS, client, dummy)
 
     def _cbTestEmptyPASS(self, ignored, client, dummy):
-        expected_output = '+OK <moshez>\r\n-ERR USER required before PASS\r\n+OK \r\n'
+        expected_output = '+OK <moshez>\r\n-ERR Invalid password\r\n+OK \r\n'
         self.assertEqual(expected_output, '\r\n'.join(client.response) + '\r\n')
         dummy.connectionLost(failure.Failure(Exception("Test harness disconnect")))
 
Index: twisted/mail/pop3.py
===================================================================
--- twisted/mail/pop3.py	(revision 34673)
+++ twisted/mail/pop3.py	(working copy)
@@ -506,6 +506,9 @@
         self.successResponse('USER accepted, send PASS')
 
     def do_PASS(self, password):
+        if not password:
+            self.failResponse("Invalid password")
+            return
         if self._userIs is None:
             self.failResponse("USER required before PASS")
             return
