Index: twisted/names/client.py
===================================================================
--- twisted/names/client.py	(revision 26633)
+++ twisted/names/client.py	(working copy)
@@ -20,6 +20,7 @@
 
 import os
 import errno
+import warnings
 
 from zope.interface import implements
 
@@ -494,7 +495,9 @@
         if reactor is None:
             from twisted.internet import reactor
         _ThreadedResolverImpl.__init__(self, reactor)
-        # warnings.warn("twisted.names.client.ThreadedResolver is deprecated, use XXX instead.")
+        warnings.warn(
+            "twisted.names.client.ThreadedResolver is deprecated, "
+            "use twisted.internet.base.ThreadedResolver instead.")
 
 class DNSClientFactory(protocol.ClientFactory):
     def __init__(self, controller, timeout = 10):
