Index: twisted/spread/util.py
===================================================================
--- twisted/spread/util.py	(revision 17340)
+++ twisted/spread/util.py	(working copy)
@@ -143,7 +143,6 @@
     
     def __init__(self, collector, fd, callback=None, *args, **kw):
         self.chunks = []
-        self.pointer = 0
         self.startProducing(fd)
         Pager.__init__(self, collector, callback, *args, **kw)
 
@@ -163,10 +162,9 @@
         self.chunks.append(chunk)
 
     def sendNextPage(self):
-        if self.pointer >= len(self.chunks):
+        if len(self.chunks) == 0:
             return
-        val = self.chunks[self.pointer]
-        self.pointer += 1
+        val = self.chunks.pop(0)
         self.producer.resumeProducing()
         self.collector.callRemote("gotPage", val)
 
