diff --git a/twisted/web/error.py b/twisted/web/error.py
index 2d01da2..8a18947 100644
--- a/twisted/web/error.py
+++ b/twisted/web/error.py
@@ -24,6 +24,9 @@ class Error(Exception):
     @type response: C{str}
     @ivar response: A complete HTML document for an error page.
     """
+    
+    message = ""
+    
     def __init__(self, code, message=None, response=None):
         """
         Initializes a basic exception.
