Index: twisted/topfiles/5055.misc
===================================================================
Index: twisted/application/strports.py
===================================================================
--- twisted/application/strports.py	(revision 34048)
+++ twisted/application/strports.py	(working copy)
@@ -51,8 +51,8 @@
 
     @type default: C{str} or C{None}
 
-    @param default: Do not use this parameter.  It is deprecated since Twisted
-        10.2.0.
+    @param default: Do not use this parameter. It has been deprecated since
+        Twisted 10.2.0.
 
     @rtype: C{twisted.application.service.IService}
 
