diff --git a/twisted/internet/main.py b/twisted/internet/main.py
index 76c3024..6b21389 100644
--- a/twisted/internet/main.py
+++ b/twisted/internet/main.py
@@ -1,9 +1,10 @@
-# -*- test-case-name: twisted.test.test_app -*-
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# -*- test-case-name: twisted.internet.test.test_main -*-
+# Copyright (c) 2001-2010 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
-"""Backwards compatability, and utility functions.
+"""
+Backwards compatibility, and utility functions.
 
 In general, this module should not be used, other than by reactor authors
 who need to use the 'installReactor' method.
@@ -16,13 +17,27 @@ import error
 CONNECTION_DONE = error.ConnectionDone('Connection done')
 CONNECTION_LOST = error.ConnectionLost('Connection lost')
 
+class ReactorAlreadyInstalledError(AssertionError):
+    """
+    Could not install reactor because one is already installed.
+    """
+
+
+
 def installReactor(reactor):
+    """
+    Install reactor C{reactor}.
+
+    @param reactor: An object that provides one or more IReactor* interfaces.
+    """
     # this stuff should be common to all reactors.
     import twisted.internet
     import sys
-    assert not sys.modules.has_key('twisted.internet.reactor'), \
-           "reactor already installed"
+    if sys.modules.has_key('twisted.internet.reactor'):
+        raise ReactorAlreadyInstalledError("reactor already installed")
     twisted.internet.reactor = reactor
     sys.modules['twisted.internet.reactor'] = reactor
 
-__all__ = ["CONNECTION_LOST", "CONNECTION_DONE", "installReactor"]
+
+__all__ = ["CONNECTION_LOST", "CONNECTION_DONE",
+           "ReactorAlreadyInstalledError", "installReactor"]
diff --git a/twisted/internet/test/test_main.py b/twisted/internet/test/test_main.py
new file mode 100644
index 0000000..b235c34
--- /dev/null
+++ b/twisted/internet/test/test_main.py
@@ -0,0 +1,38 @@
+# Copyright (c) 2010 Twisted Matrix Laboratories.
+# See LICENSE for details.
+
+"""
+Tests for L{twisted.internet.main}.
+"""
+
+from twisted.trial import unittest
+from twisted.internet.main import installReactor, ReactorAlreadyInstalledError
+
+
+class InstallReactorTests(unittest.TestCase):
+    """
+    Tests for L{installReactor}
+    """
+
+    def test_alreadyInstalled(self):
+        """
+        If a reactor is already installed, L{installReactor} raises
+        L{ReactorAlreadyInstalledError}.
+        """
+        # Because this test runs in trial, assume a reactor is already
+        # installed.
+        self.assertRaises(ReactorAlreadyInstalledError, installReactor,
+                          object())
+
+
+
+    def test_errorIsAnAssertionError(self):
+        """
+        For backwards compatibility, L{ReactorAlreadyInstalledError} is an
+        L{AssertionError}.
+        """
+        self.assertTrue(issubclass(ReactorAlreadyInstalledError,
+                        AssertionError))
+
+
+
diff --git a/twisted/topfiles/4476.bugfix b/twisted/topfiles/4476.bugfix
new file mode 100644
index 0000000..b69c7c3
--- /dev/null
+++ b/twisted/topfiles/4476.bugfix
@@ -0,0 +1 @@
+twisted.internet.main.installReactor now blocks installation of another reactor when using python -O
