diff --git a/twisted/internet/main.py b/twisted/internet/main.py
index 76c3024..19fa87a 100644
--- a/twisted/internet/main.py
+++ b/twisted/internet/main.py
@@ -20,8 +20,8 @@ def installReactor(reactor):
     # this stuff should be common to all reactors.
     import twisted.internet
     import sys
-    assert not sys.modules.has_key('twisted.internet.reactor'), \
-           "reactor already installed"
+    if sys.modules.has_key('twisted.internet.reactor'):
+        raise AssertionError("reactor already installed")
     twisted.internet.reactor = reactor
     sys.modules['twisted.internet.reactor'] = reactor
 
