Index: twisted/words/test/test_jabberjstrports.py
===================================================================
--- twisted/words/test/test_jabberjstrports.py	(revision 0)
+++ twisted/words/test/test_jabberjstrports.py	(revision 0)
@@ -0,0 +1,24 @@
+# Copyright (c) 2001-2011 Twisted Matrix Laboratories.
+# See LICENSE for details.
+
+from twisted.trial import unittest
+
+from twisted.words.protocols.jabber import jstrports
+
+
+class JabberStrPortsPlaceHolderTest(unittest.TestCase):
+    """
+    Tests for L{jstrports}
+    """
+
+    def test_parse(self):
+        expected = ('TCP', ('DOMAIN', 65535, 'Factory'), {})
+        got = jstrports.parse("tcp:DOMAIN:65535", "Factory")
+        self.assertEqual(expected, got)
+
+    def test_client(self):
+        from twisted.application.internet import TCPClient
+        got = jstrports.client("tcp:DOMAIN:65535", "Factory")
+        self.assertTrue(isinstance(got, TCPClient))
+
+
Index: twisted/words/protocols/jabber/jstrports.py
===================================================================
--- twisted/words/protocols/jabber/jstrports.py	(revision 30414)
+++ twisted/words/protocols/jabber/jstrports.py	(working copy)
@@ -6,7 +6,7 @@
 """ A temporary placeholder for client-capable strports, until we
 sufficient use cases get identified """
 
-from twisted.application import strports
+from twisted.internet.endpoints import _parse
 
 def _parseTCPSSL(factory, domain, port):
     """ For the moment, parse TCP or SSL connections the same """
@@ -22,7 +22,7 @@
 
 
 def parse(description, factory):
-    args, kw = strports._parse(description)
+    args, kw = _parse(description)
     return (args[0].upper(),) + _funcs[args[0]](factory, *args[1:], **kw)
 
 def client(description, factory):
