diff -ur Twisted-12.2.0.orig//twisted/internet/tcp.py Twisted-12.2.0/twisted/internet/tcp.py
--- Twisted-12.2.0.orig//twisted/internet/tcp.py	2012-07-28 16:29:16.000000000 +0200
+++ Twisted-12.2.0/twisted/internet/tcp.py	2012-09-17 17:17:17.000000000 +0200
@@ -946,6 +946,8 @@
     def createInternetSocket(self):
         s = base.BasePort.createInternetSocket(self)
         if platformType == "posix" and sys.platform != "cygwin":
+            if self.addressFamily == socket.AF_INET6:
+                s.setsockopt(socket.IPPROTO_IPV6, socket.IPV6_V6ONLY, 1)
             s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
         return s
 
