Index: twisted/spread/pb.py
===================================================================
--- twisted/spread/pb.py	(revision 31650)
+++ twisted/spread/pb.py	(working copy)
@@ -408,10 +408,7 @@
             state['type'] = self.type
         else:
             state['type'] = reflect.qual(self.type) # Exception class
-        if self.unsafeTracebacks:
-            state['traceback'] = self.getTraceback()
-        else:
-            state['traceback'] = 'Traceback unavailable\n'
+        state['traceback'] = self.getTraceback()
         return state
 
 
