Index: suppress-test_test_visitor-warnings/twisted/trial/test/test_test_visitor.py
===================================================================
--- suppress-test_test_visitor-warnings/twisted/trial/test/test_test_visitor.py	(revision 27376)
+++ suppress-test_test_visitor-warnings/twisted/trial/test/test_test_visitor.py	(working copy)
@@ -1,4 +1,4 @@
-from twisted.trial import unittest
+from twisted.trial import unittest, util
 from twisted.trial.runner import TestSuite, suiteVisit
 
 pyunit = __import__('unittest')
@@ -16,6 +16,10 @@
 
 
 class TestTestVisitor(unittest.TestCase):
+
+    suppress = [util.suppress(action='ignore', category=DeprecationWarning,
+        message=r'Test visitors deprecated')]
+
     def setUp(self):
         self.visitor = MockVisitor()
 
