Index: imap4.py
===================================================================
--- imap4.py	(révision 15658)
+++ imap4.py	(copie de travail)
@@ -2437,7 +2437,7 @@
         and whose errback is invoked otherwise.
         """
         d = maybeDeferred(self.getCapabilities)
-        d.addCallback(self.__cbLoginCaps, username, password)
+        d.addCallback(self.__cbLoginCaps, username, _quote(password))
         return d
 
     def serverGreeting(self, caps):
