Index: test_imap.py
===================================================================
--- test_imap.py	(revision 15658)
+++ test_imap.py	(copie de travail)
@@ -793,7 +793,7 @@
         imap4.IMAP4Server.lineReceived(self, line)
 
     def authenticateLogin(self, username, password):
-        if username == 'testuser' and password == 'password-test':
+        if username == 'testuser' and (password == 'password-test' or password == 'pass{word test'):
             return imap4.IAccount, self.theAccount, lambda: None
         raise cred.error.UnauthorizedLogin()
 
@@ -912,6 +912,16 @@
         self.assertEquals(self.server.account, SimpleServer.theAccount)
         self.assertEquals(self.server.state, 'auth')
 
+    def testLoginWithQuotePassword(self):
+        def login():
+            d = self.client.login('testuser', 'pass{word test')
+            d.addBoth(self._cbStopClient)
+        self.connected.addCallback(strip(login)).addErrback(self._ebGeneral)
+        self.loopback()
+
+        self.assertEquals(self.server.account, SimpleServer.theAccount)
+        self.assertEquals(self.server.state, 'auth')
+
     def testFailedLogin(self):
         def login():
             d = self.client.login('testuser', 'wrong-password')
