Index: twisted/words/test/test_irc.py
===================================================================
--- twisted/words/test/test_irc.py	(revision 30895)
+++ twisted/words/test/test_irc.py	(working copy)
@@ -1273,7 +1273,22 @@
         self.lines.append(m)
 
 
+class ClientListTests(unittest.TestCase):
+    def setUp(self):
+        """
+        Create a L{DummyClient} to call C{list} on.
+        """
+        self.client = DummyClient()
 
+
+    def test_list(self):
+        """
+        Test if the LIST function actually through 
+        """
+        self.client.channel_list()
+        self.assertEquals(self.client.lines, ['LIST'])
+
+
 class ClientInviteTests(unittest.TestCase):
     """
     Tests for L{IRCClient.invite}.
@@ -1303,7 +1318,6 @@
         self.assertEquals(self.client.lines, ['INVITE foo #bar'])
 
 
-
 class ClientMsgTests(unittest.TestCase):
     def setUp(self):
         self.client = DummyClient()
Index: twisted/words/protocols/irc.py
===================================================================
--- twisted/words/protocols/irc.py	(revision 30895)
+++ twisted/words/protocols/irc.py	(working copy)
@@ -1386,6 +1386,14 @@
         self.sendLine("INVITE %s %s" % (user, channel))
 
 
+    def channel_list(self):
+        """
+        Attempts to ask for the LIST which is every channel on the server.
+
+        @since: 11.0
+        """
+        self.sendLine("LIST")
+
     def topic(self, channel, topic=None):
         """
         Attempt to set the topic of the given channel, or ask what it is.
