Index: doc/words/examples/index.xhtml
===================================================================
--- doc/words/examples/index.xhtml	(revision 20006)
+++ doc/words/examples/index.xhtml	(working copy)
@@ -16,6 +16,9 @@
       <li><a href="aimbot.py">aimbot.py</a> - (AOL Instant Messaging)</li>
       <li><a href="msn_example.py">msn_example.py</a></li>
       <li><a href="oscardemo.py">oscardemo.py</a></li>
+      <li><a href="jabber_client.py">jabber_client.py</a></li>
+      <li><a href="pb_client.py">pb_client.py</a></li>
+      <li><a href="xmpp_client.py">xmpp_client.py</a></li>
     </ul>
 
 </body>
