Index: twisted/words/protocols/irc.py
===================================================================
--- twisted/words/protocols/irc.py	(revision 23862)
+++ twisted/words/protocols/irc.py	(working copy)
@@ -1016,6 +1030,13 @@
     def irc_RPL_WELCOME(self, prefix, params):
         self.signedOn()

+    def irc_RPL_NAMREPLY(self, prefix, params):
+        chantype, channel, users = params[1:]
+        pusers = []
+        for user in users.split():
+            if user[0] in '@%+':
+                pusers.append((user[0], parse_user(user[1:])))
+            else:
+                pusers.append((None, parse_user(user)))
+        self.namreply(prefix, chantype, channel, pusers, users)
+
     def irc_JOIN(self, prefix, params):
         nick = string.split(prefix,'!')[0]
         channel = params[-1]
