Index: imap4.py
===================================================================
--- imap4.py	(revision 20503)
+++ imap4.py	(working copy)
@@ -2084,10 +2084,11 @@
         lastPart = line.rfind(' ')
         if lastPart != -1:
             lastPart = line[lastPart + 1:]
-            if lastPart.startswith('{') and lastPart.endswith('}'):
+            leftParenPos = lastPart.rfind('{')
+            if lastPart.endswith('}') and leftParenPos != -1:
                 # It's a literal a-comin' in
                 try:
-                    octets = int(lastPart[1:-1])
+                    octets = int(lastPart[leftParenPos+1:-1])
                 except ValueError:
                     raise IllegalServerResponse(line)
                 if self._parts is None:
