Index: twisted/internet/abstract.py
===================================================================
--- twisted/internet/abstract.py	(revision 30077)
+++ twisted/internet/abstract.py	(working copy)
@@ -149,8 +149,8 @@
                 return self._postLoseConnection()
             elif self._writeDisconnecting:
                 # I was previously asked to to half-close the connection.
+                self._writeDisconnected = True
                 result = self._closeWriteConnection()
-                self._writeDisconnected = True
                 return result
         return result
 
Index: twisted/internet/test/test_abstract.py
===================================================================
--- twisted/internet/test/test_abstract.py	(revision 0)
+++ twisted/internet/test/test_abstract.py	(revision 0)
@@ -0,0 +1,35 @@
+from twisted.internet import protocol, interfaces
+from zope.interface import implements
+
+class ListenerProtocol1(protocol.Protocol):
+    """ Protocol that knows how to deal with connections to our listener """
+
+    implements(interfaces.IHalfCloseableProtocol)
+
+    def __init__(self):
+        self.data = []
+
+    def dataReceived(self, data):
+        print "data Received %r" % data
+
+    def readConnectionLost(self):
+        print "readConnectionLost"
+        self.transport.loseWriteConnection()
+
+    def writeConnectionLost(self):
+        print "writeConnectionLost"
+        self.transport.loseConnection()
+
+    def connectionLost(self, reason):
+        print "connectionLost"
+
+class ListenerFactory(protocol.Factory):
+    protocol = ListenerProtocol1
+
+#from twisted.internet import epollreactor
+#epollreactor.install()
+
+from twisted.internet import reactor
+port = reactor.listenTCP(5555, ListenerFactory())
+reactor.run()
+
