Index: pollreactor.py
===================================================================
--- pollreactor.py	(revision 16175)
+++ pollreactor.py	(working copy)
@@ -18,7 +18,10 @@
 # System imports
 import select, errno, sys
 
+from zope.interface import implements
+
 # Twisted imports
+from twisted.internet.interfaces import IReactorFDSet
 from twisted.python import log, threadable, failure
 from twisted.internet import main, posixbase, error
 
@@ -33,6 +36,7 @@
 
 class PollReactor(posixbase.PosixReactorBase):
     """A reactor that uses poll(2)."""
+    implements(IReactorFDSet)
 
     def _updateRegistration(self, fd):
         """Register/unregister an fd with the poller."""
