Index: twisted/topfiles/4077.removal
===================================================================
--- twisted/topfiles/4077.removal	(revision 0)
+++ twisted/topfiles/4077.removal	(revision 0)
@@ -0,0 +1,2 @@
+twisted.internet.interfaces.IUDPConnectedTransport has been removed
+(deprecated since 2008).
Index: twisted/internet/interfaces.py
===================================================================
--- twisted/internet/interfaces.py	(revision 29213)
+++ twisted/internet/interfaces.py	(working copy)
@@ -1612,22 +1612,7 @@
         """
 
 
-class IUDPConnectedTransport(Interface):
-    """
-    DEPRECATED. Transport for UDP ConnectedPacketProtocols.
-    """
 
-    def write(packet):
-        """
-        Write packet to address we are connected to.
-        """
-
-    def getHost():
-        """
-        Returns L{UNIXAddress}.
-        """
-
-
 class IUNIXDatagramTransport(Interface):
     """
     Transport for UDP PacketProtocols.
