Index: twisted/topfiles/7727.misc
===================================================================
Index: twisted/scripts/test/test_tap2deb.py
===================================================================
--- twisted/scripts/test/test_tap2deb.py	(revision 43542)
+++ twisted/scripts/test/test_tap2deb.py	(working copy)
@@ -13,7 +13,7 @@
 
 
 
-class TestTap2DEB(TestCase):
+class Tap2DEBTests(TestCase):
     """
     Tests for the L{tap2deb} script.
     """
Index: twisted/scripts/test/test_tap2rpm.py
===================================================================
--- twisted/scripts/test/test_tap2rpm.py	(revision 43542)
+++ twisted/scripts/test/test_tap2rpm.py	(working copy)
@@ -121,7 +121,7 @@
 
 
 
-class TestTap2RPM(TestCase):
+class Tap2RPMTests(TestCase):
 
 
     def setUp(self):
Index: twisted/scripts/test/test_scripts.py
===================================================================
--- twisted/scripts/test/test_scripts.py	(revision 43542)
+++ twisted/scripts/test/test_scripts.py	(working copy)
@@ -186,7 +186,7 @@
 
 
 
-class ZshIntegrationTestCase(TestCase, ZshScriptTestMixin):
+class ZshIntegrationTests(TestCase, ZshScriptTestMixin):
     """
     Test that zsh completion functions are generated without error
     """
