diff --git twisted/scripts/test/test_scripts.py twisted/scripts/test/test_scripts.py
index 8c0dc59..a18de73 100644
--- twisted/scripts/test/test_scripts.py
+++ twisted/scripts/test/test_scripts.py
@@ -187,7 +187,7 @@ class ScriptTests(TestCase, ScriptTestsMixin):
 
 
 
-class ZshIntegrationTestCase(TestCase, ZshScriptTestMixin):
+class ZshIntegrationTests(TestCase, ZshScriptTestMixin):
     """
     Test that zsh completion functions are generated without error
     """
diff --git twisted/topfiles/7727.misc twisted/topfiles/7727.misc
new file mode 100644
index 0000000..e69de29
