Index: doc/core/howto/listings/udp/MulticastServer.py
===================================================================
--- doc/core/howto/listings/udp/MulticastServer.py	(revision 32036)
+++ doc/core/howto/listings/udp/MulticastServer.py	(working copy)
@@ -1,6 +1,5 @@
 from twisted.internet.protocol import DatagramProtocol
 from twisted.internet import reactor
-from twisted.application.internet import MulticastServer
 
 class MulticastServerUDP(DatagramProtocol):
     def startProtocol(self):
Index: doc/core/howto/listings/udp/MulticastClient.py
===================================================================
--- doc/core/howto/listings/udp/MulticastClient.py	(revision 32036)
+++ doc/core/howto/listings/udp/MulticastClient.py	(working copy)
@@ -1,6 +1,5 @@
 from twisted.internet.protocol import DatagramProtocol
 from twisted.internet import reactor
-from twisted.application.internet import MulticastServer
 
 class MulticastClientUDP(DatagramProtocol):
 
@@ -8,6 +7,6 @@
             print "Received:" + repr(datagram)
 
 # Send multicast on 224.0.0.1:8005, on our dynamically allocated port
-reactor.listenUDP(0, MulticastClientUDP()).write('UniqueID', 
-                                                 ('224.0.0.1', 8005))
+port = reactor.listenUDP(0, MulticastClientUDP())
+port.write('UniqueID', ('224.0.0.1', 8005))
 reactor.run()
