=== modified file 'twisted/conch/telnet.py'
--- twisted/conch/telnet.py	2011-02-14 04:45:15 +0000
+++ twisted/conch/telnet.py	2011-09-30 18:55:47 +0000
@@ -157,7 +157,7 @@
         """A command was received but not understood.
         """
 
-    def unhandledSubnegotiation(bytes):
+    def unhandledSubnegotiation(command, bytes):
         """A subnegotiation command was received but not understood.
         """
 

=== modified file 'twisted/conch/test/test_telnet.py'
--- twisted/conch/test/test_telnet.py	2011-07-14 18:05:14 +0000
+++ twisted/conch/test/test_telnet.py	2011-09-30 18:55:20 +0000
@@ -7,6 +7,7 @@
 """
 
 from zope.interface import implements
+from zope.interface.verify import verifyObject
 
 from twisted.internet import defer
 
@@ -69,6 +70,13 @@
 
 
 
+class TestInterfaces(unittest.TestCase):
+    def test_ITelnetProtocol(self):
+        p = telnet.TelnetProtocol()
+        verifyObject(telnet.ITelnetProtocol, p)
+
+
+
 class TelnetTransportTestCase(unittest.TestCase):
     """
     Tests for L{telnet.TelnetTransport}.

