diff -ruaN Twisted-1.3.0.orig/twisted/protocols/http.py Twisted-1.3.0/twisted/protocols/http.py
--- Twisted-1.3.0.orig/twisted/protocols/http.py	2004-04-24 04:43:51.000000000 +0100
+++ Twisted-1.3.0/twisted/protocols/http.py	2005-01-12 02:52:19.470051064 +0000
@@ -386,9 +386,6 @@
             self.setLineMode(rest)
 
 
-# response codes that must have empty bodies
-NO_BODY_CODES = (204, 304)
-
 class Request:
     """A HTTP request.
 
@@ -618,7 +615,7 @@
             self.write('')
 
         # debug code just in case I made a mistake
-        if self.chunked and self.code in NO_BODY_CODES:
+        if self.chunked and not self.bodyAllowed():
             raise RuntimeError, "we screwed up"
 
         if self.chunked:
@@ -633,6 +630,17 @@
         if not self.queued:
             self._cleanup()
 
+    def bodyAllowed(self):
+        """
+        Does the response the this request permit a message body?
+
+        RFC2616 specifies that 1xx, 204, 304 responses and all
+        responses to HEAD requests must not include a message body.
+        """
+        code = self.code
+        return (code not in (204, 304) and (code < 100 or code >= 200) and
+                self.method != "HEAD")
+
     def write(self, data):
         """
         Write some data as a result of an HTTP request.  The first
@@ -649,7 +657,7 @@
             # persistent connections.
             if ((version == "HTTP/1.1") and
                 (self.headers.get('content-length', None) is None) and
-                (self.code not in NO_BODY_CODES)):
+                self.bodyAllowed()):
                 l.append("%s: %s\r\n" % ('Transfer-encoding', 'chunked'))
                 self.chunked = 1
             if self.lastModified is not None:
@@ -669,13 +677,8 @@
 
             self.transport.writeSequence(l)
 
-            # if this is a "HEAD" request, we shouldn't return any data
-            if self.method == "HEAD":
-                self.write = lambda data: None
-                return
-
-            # for certain result codes, we should never return any data
-            if self.code in NO_BODY_CODES:
+            # in certain cases, we should never return any data
+            if not self.bodyAllowed():
                 self.write = lambda data: None
                 return
 
