Index: twisted/web/test/test_webclient.py
===================================================================
--- twisted/web/test/test_webclient.py	(revision 17764)
+++ twisted/web/test/test_webclient.py	(working copy)
@@ -72,6 +72,25 @@
         request.write('abc')
         return ''
 
+class ParseUrlTestCase(unittest.TestCase):
+    """Test URL parsing facility and defaults values."""
+    def testParse(self):
+        scheme, host, port, path = client._parse("http://127.0.0.1/")
+        self.assertEquals(path, "/")
+        self.assertEquals(port, 80)
+        scheme, host, port, path = client._parse("https://127.0.0.1/")
+        self.assertEquals(path, "/")
+        self.assertEquals(port, 443)
+        scheme, host, port, path = client._parse("http://spam:12345/")
+        self.assertEquals(port, 12345)
+        scheme, host, port, path = client._parse("http://foo ")
+        self.assertEquals(host, "foo")
+        self.assertEquals(path, "/")
+        scheme, host, port, path = client._parse("http://egg:7890")
+        self.assertEquals(port, 7890)
+        self.assertEquals(host, "egg")
+        self.assertEquals(path, "/")
+
 class WebClientTestCase(unittest.TestCase):
     def _listen(self, site):
         return reactor.listenTCP(0, site, interface="127.0.0.1")
