Index: test/test_webclient.py
===================================================================
--- test/test_webclient.py	(revision 17493)
+++ test/test_webclient.py	(working copy)
@@ -126,6 +126,15 @@
     def testGetPage(self):
         return client.getPage(self.getURL("file")).addCallback(self.assertEquals, "0123456789")
 
+    def testGetPageEmptyPath(self):
+        def cb(results):
+            self.assertEquals(results[0], results[2])
+            self.assertEquals(results[1], results[2])
+        return defer.gatherResults([
+            client.getPage("http://127.0.0.1:%d " % self.portno),
+            client.getPage("http://127.0.0.1:%d" % self.portno),
+            client.getPage("http://127.0.0.1:%d/" % self.portno)]).addCallback(cb)
+
     def testTimeoutNotTriggering(self):
         # Test that when the timeout doesn't trigger, things work as expected.
         d = client.getPage(self.getURL("wait"), timeout=100)
