Index: twisted/web/client.py
===================================================================
--- twisted/web/client.py	(revision 13972)
+++ twisted/web/client.py	(working copy)
@@ -375,6 +375,10 @@
     if ':' in host:
         host, port = host.split(':')
         port = int(port)
+    if path is None:
+        path = '/'
+    elif path.strip() is '':
+        path = '/'
     return scheme, host, port, path
 
 

