Index: twisted/web/proxy.py
===================================================================
--- twisted/web/proxy.py	(revision 21718)
+++ twisted/web/proxy.py	(working copy)
@@ -169,7 +169,7 @@
         self.path = path
 
     def getChild(self, path, request):
-        return ReverseProxyResource(self.host, self.port, self.path+'/'+path)
+        return ReverseProxyResource(self.host, self.port, self.path+path)
 
     def render(self, request):
         request.received_headers['host'] = self.host
