Index: twisted/web/woven/widgets.py
===================================================================
RCS file: /cvs/Twisted/twisted/web/woven/widgets.py,v
retrieving revision 1.92
diff -u -r1.92 widgets.py
--- twisted/web/woven/widgets.py	30 Jun 2003 02:47:34 -0000	1.92
+++ twisted/web/woven/widgets.py	15 Jul 2003 18:54:50 -0000
@@ -30,6 +30,7 @@
 import template
 import view
 import utils
+import interfaces
 
 from twisted.python import components, failure
 from twisted.python import log
@@ -975,8 +976,28 @@
 
         return macro
 
+class DeferredWidget(Widget):
+    def setDataCallback(self, result, request, node):
+        model = result
+        view = None
+        if isinstance(model, components.Componentized):
+            view = model.getAdapter(interfaces.IView)
+        if not view and hasattr(model, '__class__'):
+            view = components.getAdapter(model,
+                    interfaces.IView,
+                    None,
+                    components.getAdapterClassWithInheritance)
+        
+        if view:
+            view["id"] = self.attributes.get('id', '')
+            view.templateNode = node
+            view.controller = self.controller
+            return view.setDataCallback(result, request, node)
+        else:
+            return Widget.setDataCallback(self, result, request, node)
 
 view.registerViewForModel(Text, model.StringModel)
 view.registerViewForModel(List, model.ListModel)
 view.registerViewForModel(KeyedList, model.DictionaryModel)
 view.registerViewForModel(Link, model.Link)
+view.registerViewForModel(DeferredWidget, model.DeferredWrapper)
