diff --git a/twisted/web/util.py b/twisted/web/util.py
index 6fec248..b7b727c 100644
--- a/twisted/web/util.py
+++ b/twisted/web/util.py
@@ -336,7 +336,7 @@ def formatFailure(myFailure):
             first = 0
         else:
             w('<div class="frame">')
-        w(frameHTML % (filename, lineno, method))
+        w(frameHTML % (html.escape(filename), html.escape(str(lineno)), html.escape(method)))
 
         w('<div class="snippet">')
         textSnippet = ''
@@ -360,7 +360,7 @@ def formatFailure(myFailure):
                     w('<div class="variables"><b>Self</b>')
                     w('<table class="variables">')
                     for key, value in usedVars:
-                        w(variableHTML % (key, htmlrepr(value)))
+                        w(variableHTML % (html.escape(key), htmlrepr(value)))
                     w('</table></div>')
                 break
 
@@ -372,7 +372,7 @@ def formatFailure(myFailure):
             if usedVars:
                 w('<div class="variables"><b>%s</b><table class="variables">' % nm)
                 for name, var in usedVars:
-                    w(variableHTML % (name, htmlrepr(var)))
+                    w(variableHTML % (html.escape(name), htmlrepr(var)))
                 w('</table></div>')
 
         w('</div>') # frame
