Index: twisted/web/microdom.py
===================================================================
--- twisted/web/microdom.py	(revision 11259)
+++ twisted/web/microdom.py	(working copy)
@@ -715,6 +715,12 @@
     def gotText(self, data):
         if data.strip() or self.shouldPreserveSpace():
             self._gotStandalone(Text, data)
+        elif data and not data.strip():
+            parent = self._getparent()
+            if (parent
+                and parent.hasChildNodes()
+                and isinstance(parent.lastChild(), EntityReference)):
+                self._gotStandalone(Text, ' ')
 
     def gotComment(self, data):
         self._gotStandalone(Comment, data)
