Index: twisted/web/microdom.py
===================================================================
--- twisted/web/microdom.py	(revision 11259)
+++ twisted/web/microdom.py	(working copy)
@@ -715,6 +715,8 @@
     def gotText(self, data):
         if data.strip() or self.shouldPreserveSpace():
             self._gotStandalone(Text, data)
+        elif data and not data.strip():
+            self._gotStandalone(Text, ' ')
 
     def gotComment(self, data):
         self._gotStandalone(Comment, data)
