Index: irc.py
===================================================================
--- irc.py	(revision 30410)
+++ irc.py	(working copy)
@@ -1426,7 +1426,7 @@
         self.sendLine(line)
 
 
-    def say(self, channel, message, length = None):
+    def say(self, channel, message, MAX_COMMAND_LENGTH)
         """
         Send a message to a channel
 
@@ -1436,12 +1436,11 @@
         @type message: C{str}
         @param message: The message to say.
         @type length: C{int}
-        @param length: The maximum number of octets to send at a time.  This has
-            the effect of turning a single call to C{msg()} into multiple
-            commands to the server.  This is useful when long messages may be
-            sent that would otherwise cause the server to kick us off or
-            silently truncate the text we are sending.  If None is passed, the
-            entire message is always send in one command.
+
+        @param length: The maximum number of octets to send in a single
+            command, including the IRC protocol framing. If not supplied,
+            defaults to L{MAX_COMMAND_LENGTH}.
+        @type length: C{int}
         """
         if channel[0] not in CHANNEL_PREFIXES:
             channel = '#' + channel
