Index: twisted/python/usage.py
===================================================================
--- twisted/python/usage.py	(revision 27024)
+++ twisted/python/usage.py	(working copy)
@@ -136,7 +136,7 @@
     def __init__(self):
         super(Options, self).__init__()
 
-        self.opts = self
+        self.opts = {}
         self.defaults = {}
 
         # These are strings/lists we will pass to getopt
@@ -155,16 +154,18 @@
 
         for c in collectors:
             (longOpt, shortOpt, docs, settings, synonyms, dispatch) = c()
             self.longOpt.extend(longOpt)
             self.shortOpt = self.shortOpt + shortOpt
             self.docs.update(docs)
 
-            self.opts.update(settings)
             self.defaults.update(settings)
 
             self.synonyms.update(synonyms)
             self._dispatch.update(dispatch)
 
+        self.update(self.defaults)
+
     def __hash__(self):
         """
         Define a custom hash function so that Options instances can be used
@@ -238,6 +239,7 @@
             except TypeError:
                 raise UsageError("Wrong number of arguments.")
 
+        self.update(self.opts)
         self.postOptions()
 
     def postOptions(self):
