Index: twisted/test/test_usage.py
===================================================================
--- twisted/test/test_usage.py	(revision 27024)
+++ twisted/test/test_usage.py	(working copy)
@@ -70,7 +70,76 @@
         self.failUnlessEqual(self.nice.opts['myflag'], "PONY!")
         self.failUnlessEqual(self.nice.opts['myparam'], "Tofu WITH A PONY!")
 
+class ParsedOptionsTest(unittest.TestCase):
+    """
+    Tests Option's parsedOptions attribute.
 
+    This attribute should only contain options that were actually parsed, as
+    opposed to self.opts (or self), which contains default values for missing
+    options.
+    """
+    def setUp(self):
+        """
+        Instantiate and parseOptions a well-behaved Options class.
+        """
+
+        self.niceArgV = ("--long Alpha -n Beta "
+                         "--shortless Gamma -f --myflag "
+                         "--myparam Tofu").split()
+
+        self.nice = WellBehaved()
+
+        self.nice.parseOptions(self.niceArgV)
+
+    def test_parsedOptions(self):
+        """
+        Checking that parsedOptions contains the correct values.
+        """
+        self.failUnlessEqual(self.nice.parsedOptions['long'], "Alpha")
+        self.failUnlessEqual(self.nice.parsedOptions['another'], "Beta")
+        self.failUnlessEqual(self.nice.parsedOptions['shortless'], "Gamma")
+
+    def test_unparsedOptions(self):
+        """
+        Checking that parsedOptions doesn't contain unparsed options.
+        """
+        self.assertNotIn('longonly', self.nice.parsedOptions)
+        self.assertNotIn('flout', self.nice.parsedOptions)
+
+    def test_typicalUseCase(self):
+        """
+        Demonstrates a typical use case for the new behavior.
+
+        An app can have command line options, config file values and defaults,
+        which have different precedences: command line arguments are more
+        important than config file values, which in turn are more important
+        than the defaults.
+        """
+        fakeConfigFile = {
+            'long': "This will be overridden due to a command line argument.",
+            'longonly': "This, however, will not.",
+        }
+
+        finalOpts = {}
+        finalOpts.update(self.nice.defaults)
+        finalOpts.update(fakeConfigFile)
+        finalOpts.update(self.nice.parsedOptions)
+
+        self.failUnlessEqual(finalOpts['long'], "Alpha",
+            "parsed option set correctly, even if set in config file")
+        self.failUnlessEqual(finalOpts['another'], "Beta",
+            "parsed option set correctly")
+        self.failUnlessEqual(finalOpts['shortless'], "Gamma",
+            "parsed option set correctly")
+
+        self.failUnlessEqual(finalOpts['longonly'],
+            "This, however, will not.",
+            "unparsed option set correctly grabbed from config")
+
+        self.failUnlessEqual(finalOpts['flout'], 0,
+            "unparsed option not in config correctly grabbed from defaults")
+
+
 class TypedOptions(usage.Options):
     optParameters = [
         ['fooint', None, 392, 'Foo int', int],
Index: twisted/python/usage.py
===================================================================
--- twisted/python/usage.py	(revision 27024)
+++ twisted/python/usage.py	(working copy)
@@ -55,7 +55,7 @@
         except ValueError, e:
             raise UsageError("Parameter type enforcement failed: %s" % (e,))
 
-        self.options.opts[parameterName] = value
+        self.options.parsedOptions[parameterName] = value
 
 
 class Options(dict):
@@ -137,6 +137,7 @@
         super(Options, self).__init__()
 
         self.opts = self
+        self.parsedOptions = {}
         self.defaults = {}
 
         # These are strings/lists we will pass to getopt
@@ -146,7 +147,6 @@
         self.synonyms = {}
         self._dispatch = {}
 
-
         collectors = [
             self._gather_flags,
             self._gather_parameters,
@@ -155,16 +155,18 @@
 
         for c in collectors:
             (longOpt, shortOpt, docs, settings, synonyms, dispatch) = c()
+
             self.longOpt.extend(longOpt)
             self.shortOpt = self.shortOpt + shortOpt
             self.docs.update(docs)
 
-            self.opts.update(settings)
             self.defaults.update(settings)
 
             self.synonyms.update(synonyms)
             self._dispatch.update(dispatch)
 
+        self.update(self.defaults)
+
     def __hash__(self):
         """
         Define a custom hash function so that Options instances can be used
@@ -238,6 +240,7 @@
             except TypeError:
                 raise UsageError("Wrong number of arguments.")
 
+        self.update(self.parsedOptions)
         self.postOptions()
 
     def postOptions(self):
@@ -269,7 +272,7 @@
             raise UsageError("Flag '%s' takes no argument."
                              " Not even \"%s\"." % (flagName, value))
 
-        self.opts[flagName] = 1
+        self.parsedOptions[flagName] = 1
 
     def _gather_flags(self):
         """
@@ -408,7 +411,6 @@
 
         return longOpt, shortOpt, docs, settings, synonyms, dispatch
 
-
     def __str__(self):
         return self.getSynopsis() + '\n' + self.getUsage(width=None)
 
