Index: doc/core/howto/tutorial/listings/finger/finger12.tac
===================================================================
--- doc/core/howto/tutorial/listings/finger/finger12.tac	(revision 25518)
+++ doc/core/howto/tutorial/listings/finger/finger12.tac	(working copy)
@@ -20,14 +20,14 @@
     protocol = FingerProtocol
 
     def __init__(self, **kwargs):
-	self.users = kwargs
+        self.users = kwargs
 
     def getUser(self, user):
         return defer.succeed(self.users.get(user, "No such user"))
     
 class FingerSetterProtocol(basic.LineReceiver):
     def connectionMade(self):
-	self.lines = []
+        self.lines = []
 
     def lineReceived(self, line):
         self.lines.append(line)
@@ -40,9 +40,12 @@
 class FingerSetterFactory(protocol.ServerFactory):
     protocol = FingerSetterProtocol
 
-    def __init__(self, ff):
-        self.setUser = ff.users.__setitem__
+    def __init__(self, fingerFactory):
+        self.fingerFactory = fingerFactory
 
+    def setUser(self, user, status):
+        self.fingerFactory.users[user] = status
+
 ff = FingerFactory(moshez='Happy and well')
 fsf = FingerSetterFactory(ff)
 
Index: doc/core/howto/tutorial/listings/finger/finger13.tac
===================================================================
--- doc/core/howto/tutorial/listings/finger/finger13.tac	(revision 25518)
+++ doc/core/howto/tutorial/listings/finger/finger13.tac	(working copy)
@@ -35,6 +35,9 @@
     def getUser(self, user):
         return defer.succeed(self.users.get(user, "No such user"))
 
+    def setUser(self, user, status):
+        self.users[user] = status
+
     def getFingerFactory(self):
         f = protocol.ServerFactory()
         f.protocol = FingerProtocol
@@ -44,7 +47,7 @@
     def getFingerSetterFactory(self):
         f = protocol.ServerFactory()
         f.protocol = FingerSetterProtocol
-	f.setUser = self.users.__setitem__ 
+        f.setUser = self.setUser
         return f
 
 application = service.Application('finger', uid=1, gid=1)
Index: doc/core/howto/tutorial/listings/finger/finger14.tac
===================================================================
--- doc/core/howto/tutorial/listings/finger/finger14.tac	(revision 25518)
+++ doc/core/howto/tutorial/listings/finger/finger14.tac	(working copy)
@@ -18,6 +18,7 @@
 
 class FingerService(service.Service):
     def __init__(self, filename):
+        self.users = {} 
         self.filename = filename
         self._read()
 
@@ -51,4 +52,4 @@
 finger = internet.TCPServer(79, f.getFingerFactory())
 
 finger.setServiceParent(service.IServiceCollection(application))
-f.setServiceParent(service.IServiceCollection(application)) 
+f.setServiceParent(service.IServiceCollection(application))
Index: doc/core/howto/tutorial/listings/finger/finger15.tac
===================================================================
--- doc/core/howto/tutorial/listings/finger/finger15.tac	(revision 25518)
+++ doc/core/howto/tutorial/listings/finger/finger15.tac	(working copy)
@@ -19,7 +19,7 @@
         d.addCallback(writeResponse)
 
 class FingerResource(resource.Resource):
-    
+
     def __init__(self, users):
         self.users = users
         resource.Resource.__init__(self)
@@ -27,8 +27,8 @@
     # we treat the path as the username
     def getChild(self, username, request):
         """
-            'path' is a string. 
-            'request' is a 'twisted.protocols.http.Request'.
+        'path' is a string. 
+        'request' is a 'twisted.protocols.http.Request'.
         """
         messagevalue = self.users.get(username)
         username = cgi.escape(username)
@@ -62,11 +62,11 @@
         f.getUser = self.getUser
         f.startService = self.startService
         return f
-    
+
     def getResource(self):
         r = FingerResource(self.users)
         return r
-    
+
 application = service.Application('finger', uid=1, gid=1)
 f = FingerService('/etc/users')
 serviceCollection = service.IServiceCollection(application)
Index: doc/core/howto/tutorial/listings/finger/finger16.tac
===================================================================
--- doc/core/howto/tutorial/listings/finger/finger16.tac	(revision 25518)
+++ doc/core/howto/tutorial/listings/finger/finger16.tac	(working copy)
@@ -1,6 +1,4 @@
-"""
-Read from file, announce on the web, irc
-"""
+# Read from file, announce on the web, irc
 from twisted.application import internet, service
 from twisted.internet import protocol, reactor, defer
 from twisted.words.protocols import irc
@@ -32,7 +30,7 @@
         user = user.split('!')[0]
         if self.nickname.lower() == channel.lower():
             d = self.factory.getUser(msg)
-	    
+
             def onError(err):
                 return 'Internal error in server'
             d.addErrback(onError)
@@ -61,7 +59,7 @@
     def getFingerFactory(self):
         f = protocol.ServerFactory()
         f.protocol = FingerProtocol
-	f.getUser = self.getUser
+        f.getUser = self.getUser
         return f
 
     def getResource(self):
@@ -73,7 +71,7 @@
                       "No such user <p/> usage: site/user")])),
                       'text/html'))
         return r
-    
+
     def getIRCBot(self, nickname):
         f = protocol.ReconnectingClientFactory()
         f.protocol = IRCReplyBot
Index: doc/core/howto/tutorial/listings/finger/finger17.tac
===================================================================
--- doc/core/howto/tutorial/listings/finger/finger17.tac	(revision 25518)
+++ doc/core/howto/tutorial/listings/finger/finger17.tac	(working copy)
@@ -1,5 +1,4 @@
 # Read from file, announce on the web, irc, xml-rpc
-
 from twisted.application import internet, service
 from twisted.internet import protocol, reactor, defer
 from twisted.words.protocols import irc
@@ -29,7 +28,7 @@
         user = user.split('!')[0]
         if self.nickname.lower() == channel.lower():
             d = self.factory.getUser(msg)
-	    
+
             def onError(err):
                 return 'Internal error in server'
             d.addErrback(onError)
@@ -58,7 +57,7 @@
     def getFingerFactory(self):
         f = protocol.ServerFactory()
         f.protocol = FingerProtocol
-	f.getUser = self.getUser
+        f.getUser = self.getUser
         return f
 
     def getResource(self):
@@ -76,8 +75,8 @@
     def getIRCBot(self, nickname):
         f = protocol.ReconnectingClientFactory()
         f.protocol = IRCReplyBot
-	f.nickname = nickname
-	f.getUser = self.getUser
+        f.nickname = nickname
+        f.getUser = self.getUser
         return f
 
 application = service.Application('finger', uid=1, gid=1)
Index: doc/core/howto/tutorial/listings/finger/finger18.tac
===================================================================
--- doc/core/howto/tutorial/listings/finger/finger18.tac	(revision 24967)
+++ doc/core/howto/tutorial/listings/finger/finger18.tac	(working copy)
@@ -20,18 +20,6 @@
         d.addCallback(writeValue)
 
 
-class FingerSetterProtocol(basic.LineReceiver):
-
-    def connectionMade(self):
-        self.lines = []
-
-    def lineReceived(self, line):
-        self.lines.append(line)
-
-    def connectionLost(self, reason):
-        self.factory.setUser(*self.lines[:2])
-
-
 class IRCReplyBot(irc.IRCClient):
 
     def connectionMade(self):
@@ -136,6 +124,7 @@
         f.getUser = self.getUser
         return f
 
+
 application = service.Application('finger', uid=1, gid=1)
 f = FingerService('/etc/users')
 serviceCollection = service.IServiceCollection(application)
