Index: doc/core/howto/tutorial/listings/finger/finger12.tac
===================================================================
--- doc/core/howto/tutorial/listings/finger/finger12.tac	(revision 25137)
+++ doc/core/howto/tutorial/listings/finger/finger12.tac	(working copy)
@@ -33,8 +33,9 @@
         self.lines.append(line)
 
     def connectionLost(self, reason):
-        self.factory.setUser(*self.lines[:2])
-        # first line: user    second line: status
+        user = self.lines[0]
+        status = self.lines[1]
+        self.factory.setUser(user, status)
         
 class FingerSetterFactory(protocol.ServerFactory):
     protocol = FingerSetterProtocol
Index: doc/core/howto/tutorial/listings/finger/finger13.tac
===================================================================
--- doc/core/howto/tutorial/listings/finger/finger13.tac	(revision 25137)
+++ doc/core/howto/tutorial/listings/finger/finger13.tac	(working copy)
@@ -24,8 +24,9 @@
         self.lines.append(line)
 
     def connectionLost(self,reason):
-        self.factory.setUser(*self.lines[:2])
-        # first line: user   second line: status
+        user = self.lines[0]
+        status = self.lines[1]
+        self.factory.setUser(user, status)
 
 class FingerService(service.Service):
     def __init__(self, **kwargs):
Index: doc/core/howto/tutorial/listings/finger/finger14.tac
===================================================================
--- doc/core/howto/tutorial/listings/finger/finger14.tac	(revision 25137)
+++ doc/core/howto/tutorial/listings/finger/finger14.tac	(working copy)
@@ -18,8 +18,8 @@
 
 class FingerService(service.Service):
     def __init__(self, filename):
-        self.users = {}
         self.filename = filename
+        self._read()
 
     def _read(self):
         for line in file(self.filename):
Index: doc/core/howto/tutorial/listings/finger/finger15.tac
===================================================================
--- doc/core/howto/tutorial/listings/finger/finger15.tac	(revision 25137)
+++ doc/core/howto/tutorial/listings/finger/finger15.tac	(working copy)
@@ -18,7 +18,7 @@
             self.transport.loseConnection()
         d.addCallback(writeResponse)
 
-class MotdResource(resource.Resource):
+class FingerResource(resource.Resource):
     
     def __init__(self, users):
         self.users = users
@@ -26,11 +26,15 @@
 
     # we treat the path as the username
     def getChild(self, username, request):
-        motd = self.users.get(username)
+        """
+            'path' is a string. 
+            'request' is a 'twisted.protocols.http.Request'.
+        """
+        messagevalue = self.users.get(username)
         username = cgi.escape(username)
-        if motd is not None:
-            motd = cgi.escape(motd)
-            text = '<h1>%s</h1><p>%s</p>' % (username,motd)
+        if messagevalue is not None:
+            messagevalue = cgi.escape(messagevalue)
+            text = '<h1>%s</h1><p>%s</p>' % (username,messagevalue)
         else:
             text = '<h1>%s</h1><p>No such user</p>' % username
         return static.Data(text, 'text/html')
@@ -60,7 +64,7 @@
         return f
     
     def getResource(self):
-        r = MotdResource(self.users)
+        r = FingerResource(self.users)
         return r
     
 application = service.Application('finger', uid=1, gid=1)
Index: doc/core/howto/tutorial/listings/finger/finger16.tac
===================================================================
--- doc/core/howto/tutorial/listings/finger/finger16.tac	(revision 25137)
+++ doc/core/howto/tutorial/listings/finger/finger16.tac	(working copy)
@@ -22,16 +22,7 @@
             self.transport.loseConnection()
         d.addCallback(writeResponse)
 
-class FingerSetterProtocol(basic.LineReceiver):
-    def connectionMade(self):
-	self.lines = []
 
-    def lineReceived(self, line):
-	self.lines.append(line)
-
-    def connectionLost(self,reason):
-	self.factory.setUser(*self.lines[:2])
-
 class IRCReplyBot(irc.IRCClient):
     def connectionMade(self):
         self.nickname = self.factory.nickname
Index: doc/core/howto/tutorial/listings/finger/finger17.tac
===================================================================
--- doc/core/howto/tutorial/listings/finger/finger17.tac	(revision 25137)
+++ doc/core/howto/tutorial/listings/finger/finger17.tac	(working copy)
@@ -20,16 +20,6 @@
             self.transport.loseConnection()
         d.addCallback(writeResponse)
 
-class FingerSetterProtocol(basic.LineReceiver):
-    def connectionMade(self):
-	self.lines = []
-
-    def lineReceived(self, line):
-	self.lines.append(line)
-
-    def connectionLost(self,reason):
-	self.factory.setUser(*self.lines[:2])
-
 class IRCReplyBot(irc.IRCClient):
     def connectionMade(self):
         self.nickname = self.factory.nickname
