diff --git a/twisted/internet/protocol.py b/twisted/internet/protocol.py
--- a/twisted/internet/protocol.py
+++ b/twisted/internet/protocol.py
@@ -189,7 +189,9 @@ class _InstanceFactory(ClientFactory):
         callable with the value.
         """
         self.pending = None
-        callable(value)
+        # One-args calls to a variable named "callable" are broken by 2to3
+        func = callable
+        func(value)
 
 
 
diff --git a/twisted/persisted/aot.py b/twisted/persisted/aot.py
--- a/twisted/persisted/aot.py
+++ b/twisted/persisted/aot.py
@@ -423,8 +423,8 @@ class AOTUnjellier:
         try:
             l = [None]
             self.unjellyInto(l, 0, ao)
-            for callable, v in self.afterUnjelly:
-                callable(v[0])
+            for func, v in self.afterUnjelly:
+                func(v[0])
             return l[0]
         except:
             log.msg("Error jellying object! Stacktrace follows::")
