diff --git twisted/python/test/test_win32.py twisted/python/test/test_win32.py
index 8494436..99d9b4d 100644
--- twisted/python/test/test_win32.py
+++ twisted/python/test/test_win32.py
@@ -5,6 +5,9 @@
 Tests for L{twisted.python.win32}.
 """
 
+import inspect
+import warnings
+
 from twisted.trial import unittest
 from twisted.python.runtime import platform
 from twisted.python import win32
@@ -68,3 +71,64 @@ class ProgramPathsTests(unittest.TestCase):
         test_getProgramFilesPath.skip = (
             "Cannot figure out the program files path on non-win32 platform")
 
+
+    def test_deprecationWarninggetProgramFilesPath(self):
+        """
+        Tests to ensure that L{getProgramFilesPath} has been deprecated. A
+        call to the deprecated function should cause a deprecation warning
+        to be emitted.
+        """
+        with warnings.catch_warnings(record=True) as emitted_warnings:
+            warnings.simplefilter("always")
+            win32.getProgramFilesPath()
+
+        if not emitted_warnings:
+            self.fail("No warnings emitted")
+
+        self.assertEqual(
+            emitted_warnings[0].message.args[0],
+            "twisted.python.win32.getProgramFilesPath was deprecated in "
+            "Twisted 15.3.0")
+
+    if not platform.isWindows():
+        test_deprecationWarninggetProgramFilesPath.skip = (
+            "Deprecation test is Windows only")
+
+
+    def test_deprecatedDocStringgetProgramsMenuPath(self):
+        """
+        Tests to ensure that L{getProgramFilesPath} has been deprecated.  The
+        last line should always be the deprecation message.
+        """
+        documentation = inspect.getdoc(win32.getProgramsMenuPath)
+        self.assertEqual(
+            documentation.splitlines()[-1], "Deprecated in Twisted 15.3.0.")
+
+
+    def test_deprecatedDocStringgetProgramFilesPath(self):
+        """
+        Tests to ensure that L{getProgramFilesPath} has been deprecated.  The
+        last line should always be the deprecation message.
+        """
+        documentation = inspect.getdoc(win32.getProgramFilesPath)
+        self.assertEqual(
+            documentation.splitlines()[-1], "Deprecated in Twisted 15.3.0.")
+
+
+    def test_deprecationWarninggetProgramsMenuPath(self):
+        """
+        Tests to ensure that L{getProgramsMenuPath} has been deprecated. A
+        call to the deprecated function should cause a deprecation warning
+        to be emitted.
+        """
+        with warnings.catch_warnings(record=True) as emitted_warnings:
+            warnings.simplefilter("always")
+            win32.getProgramsMenuPath()
+
+        if not emitted_warnings:
+            self.fail("No warnings emitted")
+
+        self.assertEqual(
+            emitted_warnings[0].message.args[0],
+            "twisted.python.win32.getProgramsMenuPath was deprecated in "
+            "Twisted 15.3.0")
diff --git twisted/python/win32.py twisted/python/win32.py
index 6e983cd..2068e17 100644
--- twisted/python/win32.py
+++ twisted/python/win32.py
@@ -22,7 +22,9 @@ try:
 except ImportError:
     pass
 
+from twisted.python.deprecate import deprecated
 from twisted.python.runtime import platform
+from twisted.python.versions import Version
 
 # http://msdn.microsoft.com/library/default.asp?url=/library/en-us/debug/base/system_error_codes.asp
 ERROR_FILE_NOT_FOUND = 2
@@ -43,8 +45,8 @@ try:
 except NameError:
     WindowsError = FakeWindowsError
 
-# XXX fix this to use python's builtin _winreg?
 
+@deprecated(Version("Twisted", 15, 3, 0))
 def getProgramsMenuPath():
     """
     Get the path to the Programs menu.
@@ -62,6 +64,7 @@ def getProgramsMenuPath():
     return win32api.RegQueryValueEx(hShellFolders, 'Common Programs')[0]
 
 
+@deprecated(Version("Twisted", 15, 3, 0))
 def getProgramFilesPath():
     """Get the path to the Program Files folder."""
     keyname = 'SOFTWARE\\Microsoft\\Windows\\CurrentVersion'
diff --git twisted/topfiles/7883.removal twisted/topfiles/7883.removal
new file mode 100644
index 0000000..8f498e5
--- /dev/null
+++ twisted/topfiles/7883.removal
@@ -0,0 +1 @@
+twisted.python.win32.getProgramsMenuPath and twisted.python.win32.getProgramFilesPath are now deprecated
\ No newline at end of file
