Index: twisted/python/win32.py
===================================================================
--- twisted/python/win32.py	(revision 44699)
+++ twisted/python/win32.py	(working copy)
@@ -22,7 +22,9 @@
 except ImportError:
     pass
 
+from twisted.python.deprecate import deprecated
 from twisted.python.runtime import platform
+from twisted.python.versions import Version
 
 # http://msdn.microsoft.com/library/default.asp?url=/library/en-us/debug/base/system_error_codes.asp
 ERROR_FILE_NOT_FOUND = 2
@@ -43,8 +45,8 @@
 except NameError:
     WindowsError = FakeWindowsError
 
-# XXX fix this to use python's builtin _winreg?
 
+@deprecated(Version("Twisted", 15, 1, 1))
 def getProgramsMenuPath():
     """
     Get the path to the Programs menu.
@@ -62,6 +64,7 @@
     return win32api.RegQueryValueEx(hShellFolders, 'Common Programs')[0]
 
 
+@deprecated(Version("Twisted", 15, 1, 1))
 def getProgramFilesPath():
     """Get the path to the Program Files folder."""
     keyname = 'SOFTWARE\\Microsoft\\Windows\\CurrentVersion'
Index: twisted/topfiles/7883.removal
===================================================================
--- twisted/topfiles/7883.removal	(revision 0)
+++ twisted/topfiles/7883.removal	(working copy)
@@ -0,0 +1 @@
+twisted.python.win32.getProgramsMenuPath and twisted.python.win32.getProgramFilesPath are now deprecated
\ No newline at end of file
