Index: twisted/topfiles/6514.removal
===================================================================
--- twisted/topfiles/6514.removal	(revision 0)
+++ twisted/topfiles/6514.removal	(revision 0)
@@ -0,0 +1,5 @@
+The following attributes in twisted.web.server have been deprecated
+since v9.0 and have been removed:
+    *Session.loopFactory
+    *lifetime parameter of Session.startCheckingExpiration
+    *Session.checkExpired
Index: twisted/web/server.py
===================================================================
--- twisted/web/server.py	(revision 38477)
+++ twisted/web/server.py	(working copy)
@@ -540,7 +540,6 @@
     @ivar loopFactory: Deprecated in Twisted 9.0.  Does nothing.  Do not use.
     """
     sessionTimeout = 900
-    loopFactory = task.LoopingCall
 
     _expireCall = None
 
@@ -561,7 +560,7 @@
         self.sessionNamespaces = {}
 
 
-    def startCheckingExpiration(self, lifetime=None):
+    def startCheckingExpiration(self):
         """
         Start expiration tracking.
 
@@ -569,11 +568,6 @@
 
         @return: C{None}
         """
-        if lifetime is not None:
-            warnings.warn(
-                "The lifetime parameter to startCheckingExpiration is "
-                "deprecated since Twisted 9.0.  See Session.sessionTimeout "
-                "instead.", DeprecationWarning, stacklevel=2)
         self._expireCall = self._reactor.callLater(
             self.sessionTimeout, self.expire)
 
@@ -608,16 +602,6 @@
             self._expireCall.reset(self.sessionTimeout)
 
 
-    def checkExpired(self):
-        """
-        Deprecated; does nothing.
-        """
-        warnings.warn(
-            "Session.checkExpired is deprecated since Twisted 9.0; sessions "
-            "check themselves now, you don't need to.",
-            stacklevel=2, category=DeprecationWarning)
-
-
 version = networkString("TwistedWeb/%s" % (copyright.version,))
 
 
