Index: twisted/protocols/amp.py
===================================================================
--- twisted/protocols/amp.py    (revision 38323)
+++ twisted/protocols/amp.py    (working copy)
@@ -749,11 +749,11 @@
         tag = self._nextTag()
         if requiresAnswer:
             box[ASK] = tag
-        box._sendTo(self.boxSender)
         if requiresAnswer:
             result = self._outstandingRequests[tag] = Deferred()
         else:
             result = None
+        box._sendTo(self.boxSender)
         return result

