Index: twisted/words/test/test_irc.py
===================================================================
--- twisted/words/test/test_irc.py	(revision 17770)
+++ twisted/words/test/test_irc.py	(working copy)
@@ -370,3 +370,16 @@
         self.assertRaises(ValueError, irc.split, 'foo', 0)
         self.assertEquals([], irc.split('', 1))
         self.assertEquals([], irc.split(''))
+
+from twisted.words.im import ircsupport
+from twisted.internet import error
+
+
+class ClientMsgTests(unittest.TestCase):
+    def setUp(self):
+        self.chatui = object()
+
+    def test_alreadyConnecting(self):
+        account = ircsupport.IRCAccount('la', False, 'la', None, 'localhost', 6667)
+        account._isConnecting = 1
+        self.assertRaises(error.ConnectError, account.logOn, self.chatui)
Index: twisted/words/im/basesupport.py
===================================================================
--- twisted/words/im/basesupport.py	(revision 17770)
+++ twisted/words/im/basesupport.py	(working copy)
@@ -203,7 +203,7 @@
             d.addCallback(chatui.registerAccountClient)
             return d
         else:
-            raise error.ConnectionError("Connection in progress")
+            raise error.ConnectError("Connection in progress")
 
     def getGroup(self, name):
         """Group factory.
