--- Twisted-8.1.0.org/twisted/web/server.py	2007-12-10 15:08:42.000000000 +0100
+++ Twisted-8.1.0/twisted/web/server.py	2008-10-03 15:43:45.000000000 +0200
@@ -485,6 +485,7 @@
     displayTracebacks = True
     sessionFactory = Session
     sessionCheckTime = 1800
+    randomGenerator = None
 
     def __init__(self, resource, logPath=None, timeout=60*60*12):
         """
@@ -507,9 +508,12 @@
         """
         (internal) Generate an opaque, unique ID for a user's session.
         """
-        import md5, random
+        import md5
+        if self.randomGenerator is None:
+            import random
+            self.randomGenerator = random.Random()
         self.counter = self.counter + 1
-        return md5.new("%s_%s" % (str(random.random()) , str(self.counter))).hexdigest()
+        return md5.new("%s_%s" % (str(self.randomGenerator.random()) , str(self.counter))).hexdigest()
 
     def makeSession(self):
         """
