--- twisted/web2/dav/xattrprops.py
+++ twisted/web2/dav/xattrprops.py
@@ -65,7 +65,11 @@
         # attribute name (radar://4202440). We'll quote the strings to get rid
         # of "/" characters for now.
         #
-        result = list("{%s}%s" % name)
+        (ns, attr) = name
+        if (ns is None):
+            ns = ""
+        qname = (ns, attr)
+        result = list("{%s}%s" % qname)
         for i in range(len(result)):
             c = result[i]
             if c in "%/": result[i] = "%%%02X" % (ord(c),)
@@ -79,9 +83,14 @@
     
         if (index is -1 or not len(name) > index or not name[0] == "{"):
             raise ValueError("Invalid encoded name: %r" % (name,))
-    
-        return (name[1:index], name[index+1:])
 
+        ns = name[1:index]
+        attr = name[index+1:]
+        if (not ns):
+            ns = None
+
+        return (ns, attr)
+
     _encode = classmethod(_encode)
     _decode = classmethod(_decode)
 
