Index: b/twisted/python/dist3.py
===================================================================
--- a/twisted/python/dist3.py
+++ b/twisted/python/dist3.py
@@ -47,6 +47,7 @@ modules = [
     "twisted.application.strports",
     "twisted.application.test",
     "twisted.conch",
+    "twisted.conch._version",
     "twisted.conch.ssh",
     "twisted.conch.ssh.common",
     "twisted.conch.ssh.keys",
