diff --git a/twisted/python/lockfile.py b/twisted/python/lockfile.py
--- a/twisted/python/lockfile.py
+++ b/twisted/python/lockfile.py
@@ -9,7 +9,7 @@ Filesystem-based interprocess mutex.
 
 __metaclass__ = type
 
-import errno, os
+import errno, os, sys
 
 from time import time as _uniquefloat
 
@@ -17,13 +17,13 @@ def unique():
     return str(long(_uniquefloat() * 1000))
 
 from os import rename
-try:
+if sys.platform != 'win32':
     from os import kill
     from os import symlink
     from os import readlink
     from os import remove as rmlink
     _windows = False
-except:
+else:
     _windows = True
 
     try:
