Index: twisted/topfiles/4853.misc
===================================================================
Index: twisted/internet/interfaces.py
===================================================================
--- twisted/internet/interfaces.py	(revision 30862)
+++ twisted/internet/interfaces.py	(working copy)
@@ -1114,16 +1114,25 @@
         @return: C{None}
         """
 
+
     def unregisterProducer():
         """
         Stop consuming data from a producer, without disconnecting.
         """
 
+
     def write(data):
         """
         The producer will write data by calling this method.
+
+        The implementation must be non-blocking and perform whatever
+        buffering is necessary.  If the producer has provided enough data
+        for now and it is a L{IPushProducer}, the consumer may call its
+        C{pauseProducing} method.
         """
 
+
+
 class IFinishableConsumer(IConsumer):
     """
     A Consumer for producers that finish.
