diff --git twisted/test/test_ftp.py twisted/test/test_ftp.py
index 03208bb..6dfb9ba 100644
--- twisted/test/test_ftp.py
+++ twisted/test/test_ftp.py
@@ -774,6 +774,15 @@ class FTPServerPasvDataConnectionTestCase(FTPServerTestCase):
 
 
 
+class ConnectionEstablishedNotifyingServerFactory(protocol.ServerFactory):
+    def __init__(self, deferred):
+        self.deferred = deferred
+
+    def buildProtocol(self, addr):
+        p = protocol.ServerFactory.buildProtocol(self, addr)
+        reactor.callLater(0, self.deferred.callback, p)
+        return p
+
 class FTPServerPortDataConnectionTestCase(FTPServerPasvDataConnectionTestCase):
     def setUp(self):
         self.dataPorts = []
@@ -783,13 +792,9 @@ class FTPServerPortDataConnectionTestCase(FTPServerPasvDataConnectionTestCase):
         # Establish an active data connection (i.e. server connecting to
         # client).
         deferred = defer.Deferred()
-        class DataFactory(protocol.ServerFactory):
-            protocol = _BufferingProtocol
-            def buildProtocol(self, addr):
-                p = protocol.ServerFactory.buildProtocol(self, addr)
-                reactor.callLater(0, deferred.callback, p)
-                return p
-        dataPort = reactor.listenTCP(0, DataFactory(), interface='127.0.0.1')
+        df = ConnectionEstablishedNotifyingServerFactory(deferred)
+        df.protocol = _BufferingProtocol
+        dataPort = reactor.listenTCP(0, df, interface='127.0.0.1')
         self.dataPorts.append(dataPort)
         cmd = 'PORT ' + ftp.encodeHostPort('127.0.0.1', dataPort.getHost().port)
         self.client.queueStringCommand(cmd)
