diff --git a/twisted/web/server.py b/twisted/web/server.py
index 2ec4d58..c19ae90 100644
--- a/twisted/web/server.py
+++ b/twisted/web/server.py
@@ -392,7 +392,7 @@ class Request(Copyable, http.Request, components.Componentized):
     def getSession(self, sessionInterface=None):
         # Session management
         if not self.session:
-            cookiename = b"_".join([b'TWISTED_SESSION'] + self.sitepath)
+            cookiename = "_".join(['TWISTED_SESSION'] + self.sitepath)
             sessionCookie = self.getCookie(cookiename)
             if sessionCookie:
                 try:
@@ -402,7 +402,7 @@ class Request(Copyable, http.Request, components.Componentized):
             # if it still hasn't been set, fix it up.
             if not self.session:
                 self.session = self.site.makeSession()
-                self.addCookie(cookiename, self.session.uid, path=b'/')
+                self.addCookie(cookiename, self.session.uid, path='/')
         self.session.touch()
         if sessionInterface:
             return self.session.getComponent(sessionInterface)
diff --git a/twisted/web/test/test_web.py b/twisted/web/test/test_web.py
index ae5a946..60222b9 100644
--- a/twisted/web/test/test_web.py
+++ b/twisted/web/test/test_web.py
@@ -209,6 +209,7 @@ class SessionTests(unittest.TestCase):
 
 
 
+
 # Conditional requests:
 # If-None-Match, If-Modified-Since
 
@@ -551,6 +552,21 @@ class RequestTests(unittest.TestCase):
         # failure
         self.assertEqual(1, len(self.flushLoggedErrors()))
 
+    def test_setDefaultSessionCookie(self):
+        """
+        """
+        d = DummyChannel()
+        request = server.Request(d, 1)
+        request.site = server.Site(resource.Resource())
+        request.sitepath = []
+        request.gotLength(0)
+        session = request.getSession()
+        request.requestReceived(b'GET', b'/', b'HTTP/1.0')
+        cookie_name, _ = request.cookies[0].split("=", 1)
+        # avoid delayed calls lingering after test exit
+        session.expire()
+        self.assertTrue(bool(request.session))
+        self.assertEqual(cookie_name, "TWISTED_SESSION")
 
 
 class GzipEncoderTests(unittest.TestCase):
diff --git a/twisted/web/topfiles/8077.bugfix b/twisted/web/topfiles/8077.bugfix
new file mode 100644
index 0000000..6ba6393
--- /dev/null
+++ b/twisted/web/topfiles/8077.bugfix
@@ -0,0 +1,2 @@
+twisted.web.server.Request.getSession() correctly sets default session cookie in Python3, previously cookie name
+was set to b"TWISTED_SESSION" instead of just "TWISTED_SESSION"
