diff --git twisted/web/test/test_wsgi.py twisted/web/test/test_wsgi.py
index 4e902e5..ef939d5 100644
--- twisted/web/test/test_wsgi.py
+++ twisted/web/test/test_wsgi.py
@@ -4,6 +4,7 @@
 """
 Tests for L{twisted.web.wsgi}.
 """
+import traceback
 
 __metaclass__ = type
 
@@ -1187,7 +1188,15 @@ class StartResponseTests(WSGITestsMixin, TestCase):
                     'HTTP/1.1 200 OK\r\n'))
             self.assertEqual(reraised[0][0], excInfo[0])
             self.assertEqual(reraised[0][1], excInfo[1])
-            self.assertEqual(reraised[0][2].tb_next, excInfo[2])
+
+            # Show that the tracebacks end with the same stack frames.
+            tb1 = reraised[0][2].tb_next
+            tb2 = excInfo[2]
+            self.assertEqual(
+                traceback.extract_tb(tb1)[1],
+                traceback.extract_tb(tb2)[0]
+            )
+
 
         d.addCallback(cbRendered)
 
diff --git twisted/web/wsgi.py twisted/web/wsgi.py
index 0918c4d..7d49af4 100644
--- twisted/web/wsgi.py
+++ twisted/web/wsgi.py
@@ -5,12 +5,13 @@
 An implementation of
 U{Web Resource Gateway Interface<http://www.python.org/dev/peps/pep-0333/>}.
 """
+from twisted.python.compat import iterbytes, reraise
 
 __metaclass__ = type
 
 from sys import exc_info
 
-from zope.interface import implements
+from zope.interface import implementer
 
 from twisted.python.log import msg, err
 from twisted.python.failure import Failure
@@ -183,7 +184,7 @@ class _WSGIResponse:
         else:
             pathInfo = ''
 
-        parts = request.uri.split('?', 1)
+        parts = request.uri.split(b'?', 1)
         if len(parts) == 1:
             queryString = ''
         else:
@@ -254,7 +255,8 @@ class _WSGIResponse:
         This will be called in a non-I/O thread.
         """
         if self.started and excInfo is not None:
-            raise excInfo[0], excInfo[1], excInfo[2]
+            reraise(excInfo[1], excInfo[2])
+            # raise excInfo[0]
         self.status = status
         self.headers = headers
         return self.write
@@ -313,7 +315,7 @@ class _WSGIResponse:
         """
         try:
             appIterator = self.application(self.environ, self.startResponse)
-            for elem in appIterator:
+            for elem in iterbytes(appIterator):
                 if elem:
                     self.write(elem)
                 if self._requestFinished:
@@ -340,7 +342,7 @@ class _WSGIResponse:
         self.started = True
 
 
-
+@implementer(IResource)
 class WSGIResource:
     """
     An L{IResource} implementation which delegates responsibility for all
@@ -354,7 +356,6 @@ class WSGIResource:
 
     @ivar _application: The WSGI application object.
     """
-    implements(IResource)
 
     # Further resource segments are left up to the WSGI application object to
     # handle.
