Index: twisted/test/test_jelly.py
===================================================================
--- twisted/test/test_jelly.py	(revision 31040)
+++ twisted/test/test_jelly.py	(working copy)
@@ -587,7 +587,33 @@
             self._check_newstyle(x, y)
 
 
+    def test_Referenceable(self):
+        """
+        Check that a {flavors.Referenceable} object
+          1. Jellies into a list
+          2. Unjellies correctly
+        """
+        import types
+        from twisted.spread.pb import Broker
+        from twisted.spread.flavors import Referenceable
+        from twisted.spread.jelly import jelly, unjelly
+        from twisted.test.proto_helpers import StringTransport
 
+        ref = Referenceable()
+        jelly_broker = Broker()
+        jelly_broker.makeConnection(StringTransport())
+
+        j = jelly(ref, invoker=jelly_broker)
+        self.assertIsInstance(j, types.ListType)
+
+        unjelly_broker = Broker()
+        unjelly_broker.makeConnection(StringTransport())
+
+        uj = unjelly(j, invoker=unjelly_broker)
+        self.assertIn(uj.luid, jelly_broker.localObjects)
+
+
+
 class ClassA(pb.Copyable, pb.RemoteCopy):
     def __init__(self):
         self.ref = ClassB(self)
Index: twisted/spread/flavors.py
===================================================================
--- twisted/spread/flavors.py	(revision 31040)
+++ twisted/spread/flavors.py	(working copy)
@@ -124,7 +124,7 @@
         serialize this to a peer.
         """
 
-        return "remote", jellier.invoker.registerReference(self)
+        return ["remote", jellier.invoker.registerReference(self)]
 
 
 class Root(Referenceable):
Index: twisted/spread/pb.py
===================================================================
--- twisted/spread/pb.py	(revision 31040)
+++ twisted/spread/pb.py	(working copy)
@@ -452,6 +452,7 @@
         self.pageProducers = []
         self.currentRequestID = 0
         self.currentLocalID = 0
+        self.unserializingPerspective = None
         # Some terms:
         #  PUID: process unique ID; return value of id() function.  type "int".
         #  LUID: locally unique ID; an ID unique to an object mapped over this
