Index: twisted/python/test/test_deprecate.py
===================================================================
--- twisted/python/test/test_deprecate.py	(revision 38305)
+++ twisted/python/test/test_deprecate.py	(working copy)
@@ -620,6 +620,13 @@
         about the deprecation.
         """
 
+        def dummyCallable():
+            """
+            Do nothing.
+
+            This is used to test the deprecation decorators.
+            """
+
         version = Version('Twisted', 8, 0, 0)
         dummy = deprecated(version)(dummyCallable)
 
Index: twisted/topfiles/5197.bugfix
===================================================================
--- twisted/topfiles/5197.bugfix	(revision 0)
+++ twisted/topfiles/5197.bugfix	(revision 0)
@@ -0,0 +1 @@
+twisted.python.test.test_deprecate.TestDeprecationWarnings.test_deprecatedUpdatesDocstring now uses a locally defined function for testing.
\ No newline at end of file
