=== modified file 'twisted/conch/ssh/userauth.py'
--- twisted/conch/ssh/userauth.py	2010-02-10 18:55:24 +0000
+++ twisted/conch/ssh/userauth.py	2010-02-10 20:38:26 +0000
@@ -1,5 +1,5 @@
 # -*- test-case-name: twisted.conch.test.test_userauth -*-
-# Copyright (c) 2001-2009 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2010 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 """
@@ -488,12 +488,28 @@
         We sort the list of methods by their position in self.preferredOrder,
         removing methods that have already succeeded.  We then call
         self.tryAuth with the most preferred method,
+
+        @param packet: the MSG_USERAUTH_FAILURE payload, as described above.
+        @type packet: C{str}
+        @return: a Deferred that will be callbacked with None as soon as all
+            auth methods have been tried, or None if no more auth methods are
+            available.
+        @rtype: C{defer.Deferred} or C{None}
         """
         canContinue, partial = getNS(packet)
         partial = ord(partial)
         if partial:
             self.authenticatedWith.append(self.lastAuth)
         def orderByPreference(meth):
+            """
+            Invoked once per auth method in order to extract a comparison key
+            which is then used for sorting.
+
+            @param meth: the auth method 
+            @type meth: C{str}
+            @return: the comparison key for meth
+            @rtype: C{int}
+            """
             if meth in self.preferredOrder:
                 return self.preferredOrder.index(meth)
             else:

=== modified file 'twisted/conch/test/test_userauth.py'
--- twisted/conch/test/test_userauth.py	2010-02-10 18:55:24 +0000
+++ twisted/conch/test/test_userauth.py	2010-02-10 20:43:53 +0000
@@ -1,5 +1,5 @@
 # -*- test-case-name: twisted.conch.test.test_userauth -*-
-# Copyright (c) 2007-2009 Twisted Matrix Laboratories.
+# Copyright (c) 2007-2010 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 """

=== added file 'twisted/conch/topfiles/4266.bugfix'
--- twisted/conch/topfiles/4266.bugfix	1970-01-01 00:00:00 +0000
+++ twisted/conch/topfiles/4266.bugfix	2010-02-10 20:10:01 +0000
@@ -0,0 +1,2 @@
+twisted.conch.ssh.SSHUserAuthClient now honors preferredOrder when
+authenticating.

