Index: setup.py
===================================================================
--- setup.py	(revision 31546)
+++ setup.py	(working copy)
@@ -15,7 +15,10 @@
 
 import sys, os
 
+# Import bdist_msi for converting version to MSI format
+from distutils.command.bdist_msi import bdist_msi
 
+
 def getExtensions():
     """
     Get all extensions from core and all subprojects.
@@ -35,6 +38,19 @@
     return extensions
 
 
+class bdist_msi_patch_version(bdist_msi):
+    """ MSI builder requires verion to be in the x.x.x format
+
+        http://bugs.python.org/issue6040
+    """
+    def run(self):
+        # strip revision from version (if any), e.g. 11.0.0+r31546
+        saved = self.distribution.metadata.version
+        self.distribution.metadata.version = saved.split('+')[0]
+        bdist_msi.run(self)
+        self.distribution.metadata.version = saved
+
+
 def main(args):
     """
     Invoke twisted.python.dist with the appropriate metadata about the
@@ -78,7 +94,9 @@
                 "Programming Language :: Python :: 2.5",
                 "Programming Language :: Python :: 2.6",
                 "Programming Language :: Python :: 2.7",
-                ])
+                ],
+            cmdclass={'bdist_msi': bdist_msi_patch_version},
+        )
 
     if 'setuptools' in sys.modules:
         from pkg_resources import parse_requirements
