Index: setup.py
===================================================================
--- setup.py	(revision 31546)
+++ setup.py	(working copy)
@@ -15,7 +15,12 @@
 
 import sys, os
 
+# Imports for converting version to MSI format for bdist_msi
+from distutils.command.bdist_msi import bdist_msi
+import inspect
+import types
 
+
 def getExtensions():
     """
     Get all extensions from core and all subprojects.
@@ -35,6 +40,30 @@
     return extensions
 
 
+class bdist_msi_patch_version(bdist_msi):
+    """ MSI builder requires verion to be in the x.x.x format
+
+        @see: http://bugs.python.org/issue6040
+        @since: 11.1
+    """
+    def run(self):
+        def monkey_get_version(self):
+            """ monkey patch replacement for metadata.get_version() that
+                returns MSI compatible version string for bdist_msi
+            """
+            # get filename of the calling function
+            if inspect.stack()[1][1].endswith('bdist_msi.py'):
+                # strip revision from version (if any), e.g. 11.0.0+r31546
+                return self.version.split('-')[0]
+            else:
+                return self.version
+
+        # monkeypatching get_version() call for DistributionMetadata
+        self.distribution.metadata.get_version = \
+            types.MethodType(monkey_get_version, self.distribution.metadata)
+        bdist_msi.run(self)
+
+
 def main(args):
     """
     Invoke twisted.python.dist with the appropriate metadata about the
@@ -78,7 +107,9 @@
                 "Programming Language :: Python :: 2.5",
                 "Programming Language :: Python :: 2.6",
                 "Programming Language :: Python :: 2.7",
-                ])
+                ],
+            cmdclass={'bdist_msi': bdist_msi_patch_version},
+        )
 
     if 'setuptools' in sys.modules:
         from pkg_resources import parse_requirements
