Index: setup.py
===================================================================
--- setup.py	(revision 31541)
+++ setup.py	(working copy)
@@ -42,9 +42,14 @@
     """
     if os.path.exists('twisted'):
         sys.path.insert(0, '.')
-    from twisted import copyright
+    from twisted.copyright import version
     from twisted.python.dist import getDataFiles, getScripts, getPackages, setup
 
+    if 'bdist_msi' in args:
+        # StrictVersion in bdist_msi doesn't allow versions like 11.0.0+r31541
+        print """Stripping revision suffix for bdist_msi"""
+        version = version.split('+')[0]
+
     # "" is included because core scripts are directly in bin/
     projects = [''] + [x for x in os.listdir('bin')
                        if os.path.isdir(os.path.join("bin", x))
@@ -56,7 +61,7 @@
         setup_args = dict(
             # metadata
             name="Twisted",
-            version=copyright.version,
+            version=version,
             description="An asynchronous networking framework written in "
                         "Python",
             author="Twisted Matrix Laboratories",
