=== modified file 'twisted/internet/process.py'
--- twisted/internet/process.py	2009-01-16 17:03:37 +0000
+++ twisted/internet/process.py	2009-10-13 12:29:59 +0000
@@ -55,13 +55,20 @@
     """
     if pid in reapProcessHandlers:
         raise RuntimeError("Try to register an already registered process.")
+    sigchld_occurred = [False]
+    def queue_sigchld(signum, frame):
+        sigchld_occurred[0] = True
+    orig_sigchld_handler = signal.signal(signal.SIGCHLD, queue_sigchld)
     try:
-        auxPID, status = os.waitpid(pid, os.WNOHANG)
+        try:
+            auxPID, status = os.waitpid(pid, os.WNOHANG)
+        finally:
+            signal.signal(signal.SIGCHLD, orig_sigchld_handler)
     except:
         log.msg('Failed to reap %d:' % pid)
         log.err()
         auxPID = None
-    if auxPID:
+    if auxPID or sigchld_occurred[0]:
         process.processEnded(status)
     else:
         # if auxPID is 0, there are children but none have exited

