Index: twisted/mail/test/test_imap.py
===================================================================
--- twisted/mail/test/test_imap.py	(revision 35001)
+++ twisted/mail/test/test_imap.py	(working copy)
@@ -792,6 +792,32 @@
             self.assertEqual(query, expected)
 
 
+    def test_queryKeywordFlagWithQuotes(self):
+        """
+        When passed the C{keyward} argument, L{imap4.Query} should return
+        an unquoted string.
+
+        @see: U{http://tools.ietf.org/html/rfc3501#section-9}
+        @see: U{http://tools.ietf.org/html/rfc3501#section-6.4.4}
+        """
+        query = imap4.Query(keyword='twisted')
+        expected = '(KEYWORD twisted)'
+        self.assertEqual(query, expected)
+
+
+    def test_queryUnkeywordFlagWithQuotes(self):
+        """
+        When passed the C{Unkeyward} argument, L{imap4.Query} should return
+        an unquoted string.
+
+        @see: U{http://tools.ietf.org/html/rfc3501#section-9}
+        @see: U{http://tools.ietf.org/html/rfc3501#section-6.4.4}
+        """
+        query = imap4.Query(unkeyword='twisted')
+        expected = '(UNKEYWORD twisted)'
+        self.assertEqual(query, expected)
+
+
     def test_invalidIdListParser(self):
         """
         Trying to parse an invalid representation of a sequence range raises an
Index: twisted/mail/imap4.py
===================================================================
--- twisted/mail/imap4.py	(revision 35001)
+++ twisted/mail/imap4.py	(working copy)
@@ -3981,7 +3981,7 @@
 )
 
 _NO_QUOTES = (
-    'LARGER', 'SMALLER', 'UID'
+    'LARGER', 'SMALLER', 'UID', 'KEYWORD', 'UNKEYWORD'
 )
 
 def Query(sorted=0, **kwarg):
