Index: twisted/news/nntp.py
===================================================================
--- twisted/news/nntp.py	(revision 33708)
+++ twisted/news/nntp.py	(working copy)
@@ -830,13 +830,6 @@
 
 
     def _gotArticle(self, (index, id, article)):
-        if isinstance(article, types.StringType):
-            import warnings
-            warnings.warn(
-                "Returning the article as a string from `articleRequest' "
-                "is deprecated.  Return a file-like object instead."
-            )
-            article = StringIO.StringIO(article)
         self.currentIndex = index
         self.sendLine('220 %d %s article' % (index, id))
         s = basic.FileSender()
@@ -900,13 +893,6 @@
 
 
     def _gotBody(self, (index, id, body)):
-        if isinstance(body, types.StringType):
-            import warnings
-            warnings.warn(
-                "Returning the article as a string from `articleRequest' "
-                "is deprecated.  Return a file-like object instead."
-            )
-            body = StringIO.StringIO(body)
         self.currentIndex = index
         self.sendLine('221 %d %s article retrieved' % (index, id))
         self.lastsent = ''
