Index: twisted/topfiles/5602.misc
===================================================================
--- twisted/topfiles/5602.misc	(revision 0)
+++ twisted/topfiles/5602.misc	(revision 0)
@@ -0,0 +1 @@
+twisted.web._newclient.Request now shows @type persistent : C{bool} instead of @type persistent :Boolean in its docstring.
Index: twisted/web/_newclient.py
===================================================================
--- twisted/web/_newclient.py	(revision 34099)
+++ twisted/web/_newclient.py	(working copy)
@@ -563,7 +563,7 @@
         produces the content body to send to the remote HTTP server.
 
     @ivar persistent: Set to C{True} when you use HTTP persistent connection.
-    @type persistent: Boolean
+    @type persistent: C{bool}.
     """
     def __init__(self, method, uri, headers, bodyProducer, persistent=False):
         self.method = method
