Index: woven/widgets.py
===================================================================
RCS file: /cvs/Twisted/twisted/web/woven/widgets.py,v
retrieving revision 1.97
diff -u -r1.97 widgets.py
--- woven/widgets.py	26 Jul 2003 09:27:17 -0000	1.97
+++ woven/widgets.py	6 Aug 2003 20:33:58 -0000
@@ -386,7 +386,11 @@
             sm = self.submodel.split('/')[-1]
             slots = domhelpers.locateNodes(self.templateNode, name + 'Of', sm)
             if not slots:
-                slots = domhelpers.locateNodes(self.templateNode, "pattern", name, noNesting=1)
+#                slots = domhelpers.locateNodes(self.templateNode, "pattern", name, noNesting=1)
+                matcher = lambda n, name=name: isinstance(n, Element) and \
+                            n.hasAttribute("pattern") and n.getAttribute("pattern") == name
+                recurseMatcher = lambda n: isinstance(n, Element) and not n.hasAttribute("view")
+                slots = domhelpers.findNodesShallowOnMatch(self.templateNode, matcher, recurseMatcher)
                 if not slots:
                     msg = 'WARNING: No template nodes were found '\
                               '(tagged %s="%s"'\
@@ -525,16 +529,11 @@
     """
     def setUp(self, request, node, data):
         nSplit = data.split('\n')
-        para = request.d.createElement('p')
         for line in nSplit:
             if line.strip():
+                para = request.d.createElement('p')
                 para.appendChild(request.d.createTextNode(line))
-            else:
                 self.add(para)
-                para = request.d.createElement('p')
-        if para.hasChildNodes():
-            self.add(para)
-
 
 class Image(Widget):
     """
