Index: domhelpers.py
===================================================================
RCS file: /cvs/Twisted/twisted/web/domhelpers.py,v
retrieving revision 1.21
diff -u -r1.21 domhelpers.py
--- domhelpers.py	19 Jul 2003 02:09:50 -0000	1.21
+++ domhelpers.py	6 Aug 2003 20:33:30 -0000
@@ -191,6 +191,19 @@
     return accum
 
 
+def findNodesShallowOnMatch(parent, matcher, recurseMatcher, accum=None):
+    if accum is None:
+        accum = []
+    if not parent.hasChildNodes():
+        return accum
+    for child in parent.childNodes:
+        # print child, child.nodeType, child.nodeName
+        if matcher(child):
+            accum.append(child)
+        if recurseMatcher(child):
+            findNodesShallowOnMatch(child, matcher, recurseMatcher, accum)
+    return accum
+
 def findNodesShallow(parent, matcher, accum=None):
     if accum is None:
         accum = []
