--- ../Twisted/twisted/mail/imap4.py	2006-07-31 17:00:51.000000000 -0700
+++ ./imap4.py	2006-07-31 17:37:13.000000000 -0700
@@ -1463,7 +1463,7 @@
                 return False
         else:
             c = q.upper()
-            f = getattr(self, 'search_' + c)
+            f = getattr(self, 'search_' + c, None)
             if f:
                 if not f(query, id, msg):
                     return False
