--- test_imap.py	2006-08-14 16:04:08.000000000 -0700
+++ test_imap.py.1977test	2006-08-14 16:02:43.000000000 -0700
@@ -2229,6 +2229,37 @@
     def testFetchFastUID(self):
         return self.testFetchFast(1)
 
+class DefaultSearchTestCase(IMAP4HelperMixin, unittest.TestCase):
+    def setUp(self):
+        self.server = imap4.IMAP4Server()
+        self.server.state = 'select'
+        self.server.mbox = self
+        self.connected = defer.Deferred()
+        self.client = SimpleClient(self.connected)
+        self.msgObjs = [
+            FakeyMessage({}, (), '', '', 12345, None),
+            FakeyMessage({}, (), '', '', 999, None),
+            FakeyMessage({}, (), '', '', 10101, None),
+        ]
+
+    def fetch(self, messages, uid):
+        self.received_messages = messages
+        self.received_uid = uid
+        return iter(zip(range(len(self.msgObjs)), self.msgObjs))
+
+    def testDefaultSearchUnexpected(self):
+
+        def login():
+            return self.client.login('testuser', 'password-test')
+        def select():
+            return self.client.select('mailbox-test')
+        def search():
+            return self.client.search('1',"UNDELETED","UNSEEN")
+
+        d = self.connected.addCallback(strip(search))
+        d.addCallback(self._cbStopClient)
+        d.addErrback(self._ebGeneral)
+        self.loopback()
 
 class FetchSearchStoreTestCase(unittest.TestCase, IMAP4HelperMixin):
     implements(imap4.ISearchableMailbox)
