Index: twisted/python/versions.py
===================================================================
--- twisted/python/versions.py	(revision 22720)
+++ twisted/python/versions.py	(working copy)
@@ -96,9 +96,16 @@
                     if rev is not None:
                         return rev.encode('ascii')
         except:
-            return "Unknown"
-        
+            entriesFile.seek(0)
+            if entriesFile.readline().strip() == '8':
+                entriesFile.readline()
+                entriesFile.readline()
+                revno = entriesFile.readline().strip()
+                return revno
+            else:
+                return "Unknown"
 
+
     def _getSVNVersion(self):
         """
         Figure out the SVN revision number based on the existance of
