diff --git twisted/conch/test/test_agent.py twisted/conch/test/test_agent.py
index 83115c8..4640531 100644
--- twisted/conch/test/test_agent.py
+++ twisted/conch/test/test_agent.py
@@ -8,14 +8,8 @@ from __future__ import absolute_import, division
 
 import struct
 
-from twisted.internet import reactor
-from twisted.internet.interfaces import ITLSTransport
 from twisted.trial import unittest
-
-if not ITLSTransport.providedBy(reactor):
-    iosim = None
-else:
-    from twisted.test import iosim
+from twisted.test import iosim
 
 try:
     import Crypto.Cipher.DES3
