Index: doc/core/howto/defer-intro.xhtml
===================================================================
--- doc/core/howto/defer-intro.xhtml	(revision 38257)
+++ doc/core/howto/defer-intro.xhtml	(working copy)
@@ -467,7 +467,7 @@
 <p>
   The order of <code class="python">addErrback</code>
   and <code class="python">addCallback</code> matters.  In the next section, we
-  can see what would happen when we them around.
+  can see what would happen when we swap them around.
 </p>
 
 <h3>Handle an error for the entire operation</h3>
