diff --git twisted/internet/interfaces.py twisted/internet/interfaces.py
index 10f9ef3..f07b704 100644
--- twisted/internet/interfaces.py
+++ twisted/internet/interfaces.py
@@ -1838,14 +1838,12 @@ class IStreamClientEndpointStringParser(Interface):
     )
 
 
-    def parseStreamClient(reactor, *args, **kwargs):
+    def parseStreamClient(*args, **kwargs):
         """
         This method is invoked by L{endpoints.clientFromString}, if the type of
         endpoint matches the return value from this
         L{IStreamClientEndpointStringParser}'s C{prefix} method.
 
-        @param reactor: The reactor passed to C{clientFromString}.
-
         @param args: The string arguments, minus the endpoint type, in the
             endpoint description string, parsed according to the rules
             described in L{endpoints.quoteStringArgument}.  For example, if the
diff --git twisted/topfiles/4956.bugfix twisted/topfiles/4956.bugfix
new file mode 100644
index 0000000..c22d943
--- /dev/null
+++ twisted/topfiles/4956.bugfix
@@ -0,0 +1 @@
+IStreamClientEndpointStringParser.parseStreamClient now correctly describes how it will be called by clientFromString
\ No newline at end of file
