Index: twisted/protocols/amp.py
===================================================================
--- twisted/protocols/amp.py	(revision 21530)
+++ twisted/protocols/amp.py	(working copy)
@@ -712,9 +712,9 @@
         Mutate the given box to give it any additional keys (_command, _ask)
         required for the command and request/response machinery, then send it.
 
-        Returns a Deferred which fires with the response C{amp.Box} when it
-        is received, or fails with a C{amp.RemoteAmpError} if an error is
-        received.
+        Unless requiresAnswer is False, returns a Deferred which fires with 
+        the response C{amp.Box} when it is received, or fails with a 
+        C{amp.RemoteAmpError} if an error is received.
 
         If the Deferred fails and the error is not handled by the caller of
         this method, the failure will be logged and the connection dropped.
@@ -748,7 +748,7 @@
 
     def callRemoteString(self, command, requiresAnswer=True, **kw):
         """
-        This is a low-level API, designed only for opitmizing simple messages
+        This is a low-level API, designed only for optimizing simple messages
         for which the overhead of parsing is too great.
 
         @param command: a str naming the command.
