diff --git twisted/python/compat.py twisted/python/compat.py
index a8fc936..123f411 100644
--- twisted/python/compat.py
+++ twisted/python/compat.py
@@ -18,7 +18,7 @@ the latest version of Python directly from your code, if possible.
     string type (bytes in Python 2, unicode in Python 3).
 """
 
-from __future__ import division
+from __future__ import print_function,division,absolute_import
 
 import sys, string, socket, struct
 
@@ -29,6 +29,16 @@ else:
     _PY3 = True
 
 
+# Compatibility for iterations over dictionary:
+if _PY3:
+    iterkeys = lambda d: iter(d.keys())
+    itervalues = lambda d: iter(d.values())
+    iteritems = lambda d: iter(d.items())
+else:
+    iterkeys = lambda d: d.iterkeys()
+    itervalues = lambda d: d.itervalues()
+    iteritems = lambda d: d.iteritems()
+
 
 def inet_pton(af, addr):
     if af == socket.AF_INET:
@@ -404,7 +414,6 @@ interpolation.  For example, this is safe on Python 2 and Python 3:
 @rtype: C{bytes}
 """
 
-
 __all__ = [
     "reraise",
     "execfile",
@@ -418,6 +427,9 @@ __all__ = [
     "networkString",
     "unicode",
     "iterbytes",
+    "iterkeys",
+    "itervalues",
+    "iteritems"
     "intToBytes",
     "lazyByteSlice",
     ]
