--- F:\Grant\Programming\Python\SVN repositories\twistedsvn\trunk\twisted\application\service.py	Sat Jun 18 21:52:08 2005
+++ F:\Grant\Programming\Python\SVN repositories\twistedsvn\grant's trunk\twisted\application\service.py	Fri Jun 24 23:19:36 2005
@@ -288,7 +288,7 @@
     IService(ret).setName(name)
     return ret
 
-def loadApplication(filename, kind, passphrase=None):
+def loadApplication(filename, kind, passphrase=None, zipfile=None):
     """Load Application from file
 
     @type filename: C{str}
@@ -300,7 +300,14 @@
     'xml' or 'python'. If C{passphrase} is given, the application was encrypted
     with the given passphrase.
     """
-    if kind == 'python':
+    if kind == 'python' and zipfile:
+        import zipimport
+        zi = zipimport.zipimporter(zipfile)
+        code = zi.get_code(filename)
+        d = {}
+        exec code in d
+        application = d['application']
+    elif kind == 'python' and not zipfile:
         application = sob.loadValueFromFile(filename, 'application', passphrase)
     else:
         application = sob.load(filename, kind, passphrase)
