--- F:\Grant\Programming\Python\SVN repositories\twistedsvn\trunk\twisted\application\app.py	Fri Jun 24 22:38:17 2005
+++ F:\Grant\Programming\Python\SVN repositories\twistedsvn\grant's trunk\twisted\application\app.py	Sat Jun 25 12:56:49 2005
@@ -154,7 +154,10 @@
     filename, style = s[0], {'file':'pickle'}.get(s[1],s[1])
     try:
         log.msg("Loading %s..." % filename)
-        application = service.loadApplication(filename, style, passphrase)
+        if config['zipfile']:
+            application = service.loadApplication(filename, style, passphrase, config['zipfile'])
+        else:
+            application = service.loadApplication(filename, style, passphrase)
         log.msg("Loaded.")
     except Exception, e:
         s = "Failed to load application: %s" % e
@@ -209,6 +212,8 @@
                       "(Marmalade format)."],
                      ['source', 's', None,
                       "Read an application from a .tas file (AOT format)."],
+                     ['zipfile', 'z', None,
+                      "Read python byte code from a zipfile"],
                      ['rundir','d','.',
                       'Change to a supplied directory before running'],
                      ['reactor', 'r', None,
