diff --git twisted/python/zippath.py twisted/python/zippath.py
index a82f253..2bafac2 100644
--- twisted/python/zippath.py
+++ twisted/python/zippath.py
@@ -8,6 +8,8 @@ This module contains implementations of IFilePath for zip files.
 See the constructor for ZipArchive for use.
 """
 
+from __future__ import print_function,division,absolute_import
+
 __metaclass__ = type
 
 import os
@@ -27,7 +29,7 @@ else:
 
 from twisted.python.filepath import IFilePath, FilePath, AbstractFilePath
 
-from zope.interface import implements
+from zope.interface import implementer
 
 # using FilePath here exclusively rather than os to make sure that we don't do
 # anything OS-path-specific here.
@@ -36,13 +38,12 @@ ZIP_PATH_SEP = '/'              # In zipfiles, "/" is universally used as the
                                 # path separator, regardless of platform.
 
 
+@implementer(IFilePath)
 class ZipPath(AbstractFilePath):
     """
     I represent a file or directory contained within a zip file.
     """
 
-    implements(IFilePath)
-
     sep = ZIP_PATH_SEP
 
     def __init__(self, archive, pathInArchive):
