=== modified file 'twisted/conch/ssh/filetransfer.py'
--- twisted/conch/ssh/filetransfer.py	2008-11-01 14:32:22 +0000
+++ twisted/conch/ssh/filetransfer.py	2009-07-01 14:46:31 +0000
@@ -831,9 +831,20 @@
     def __init__(self, errorCode, errorMessage, lang = ''):
         Exception.__init__(self)
         self.code = errorCode
-        self.message = errorMessage
+        self._message = errorMessage
         self.lang = lang
 
+
+    def message(self):
+        # Python 2.6 deprecates assigning to the 'message' attribute of an
+        # exception. We define this read-only property here in order to
+        # prevent the warning about deprecation while maintaining backwards
+        # compatibility with object clients that rely on the 'message'
+        # attribute being set correctly. See bug #3897.
+        return self._message
+    message = property(message)
+
+
     def __str__(self):
         return 'SFTPError %s: %s' % (self.code, self.message)
 

