Index: twisted/protocols/amp.py
===================================================================
--- twisted/protocols/amp.py	(revision 24223)
+++ twisted/protocols/amp.py	(working copy)
@@ -2052,7 +2052,6 @@
             boxReceiver = self
         if locator is None:
             locator = self
-        boxSender = self
         BoxDispatcher.__init__(self, locator)
         BinaryBoxProtocol.__init__(self, boxReceiver)
 
