diff --git twisted/python/dist3.py twisted/python/dist3.py
index d56173d..eee4f44 100644
--- twisted/python/dist3.py
+++ twisted/python/dist3.py
@@ -102,6 +102,7 @@ modules = [
     "twisted.python.test.modules_helpers",
     "twisted.python.threadable",
     "twisted.python.threadpool",
+    "twisted.python.usage",
     "twisted.python.util",
     "twisted.python.versions",
     "twisted.test",
@@ -203,6 +204,7 @@ testModules = [
     "twisted.test.test_twisted",
     "twisted.test.test_threadpool",
     "twisted.test.test_udp",
+    "twisted.test.test_usage",
     "twisted.trial.test.test_assertions",
     "twisted.trial.test.test_asyncassertions",
     "twisted.trial.test.test_deferred",
diff --git twisted/python/test/test_util.py twisted/python/test/test_util.py
index 241b163..c3e5d28 100644
--- twisted/python/test/test_util.py
+++ twisted/python/test/test_util.py
@@ -1077,6 +1077,68 @@ class FancyStrMixinTests(unittest.TestCase):
         obj = Foo()
         self.assertEqual(str(obj), repr(obj))
 
+
+class PadToTest(unittest.TestCase):
+    """
+    Tests for L{util.padTo}.
+    """
+
+    def test_default(self):
+        """
+        Pad a list with the default value.
+        """
+        l = []
+        res = util.padTo(3, l)
+        self.assertEqual([None] * 3, res)
+
+
+    def test_specificDefaultValue(self):
+        """
+        Pad a list with a specific value.
+        """
+        l = []
+        res = util.padTo(4, l, "x")
+        self.assertEqual(["x"] * 4, res)
+
+
+    def test_padNonEmptyList(self):
+        """
+        Pad a list which already contains elements.
+        """
+        l = [1, 2]
+        res = util.padTo(3, l, "z")
+        self.assertEqual([1, 2, "z"], res)
+
+
+    def test_padToSmallerSize(self):
+        """
+        L{util.padTo} can't pad a list if the size requested is smaller than the
+        size of the list to pad.
+        """
+        l = [1, 2]
+        self.assertRaises(ValueError, util.padTo, 1, l)
+
+
+    def test_alreadyPadded(self):
+        """
+        Padding a list to the exact size of this list just returns a copy.
+        """
+        l = [1, 2]
+        res = util.padTo(len(l), l)
+        self.assertEqual(l, res)
+        self.assertNotEqual(id(res), id(l))
+
+
+    def test_makeCopy(self):
+        """
+        L{util.padTo} doesn't modify the input list but makes a copy.
+        """
+        l = []
+        res = util.padTo(4, l)
+        self.assertEqual([], l)
+        self.assertNotEqual(id(res), id(l))
+
+
 if _PY3:
     del (SwitchUIDTest, SearchUpwardsTest, RunAsEffectiveUserTests,
          OrderedDictTest, IntervalDifferentialTestCase, UtilTestCase,
diff --git twisted/python/usage.py twisted/python/usage.py
index 9280ae2..23250b1 100644
--- twisted/python/usage.py
+++ twisted/python/usage.py
@@ -12,14 +12,18 @@ U{http://twistedmatrix.com/projects/core/documentation/howto/options.html},
 or doc/core/howto/options.xhtml in your Twisted directory.
 """
 
+from __future__ import division, absolute_import
+
 # System Imports
+import inspect
 import os
 import sys
 import getopt
 from os import path
+import textwrap
 
 # Sibling Imports
-from twisted.python import reflect, text, util
+from twisted.python import reflect, util
 
 
 class UsageError(Exception):
@@ -52,7 +56,7 @@ class CoerceParameter(object):
                              % (parameterName,))
         try:
             value = self.coerce(value)
-        except ValueError, e:
+        except ValueError as e:
             raise UsageError("Parameter type enforcement failed: %s" % (e,))
 
         self.options.opts[parameterName] = value
@@ -188,13 +192,13 @@ class Options(dict):
         as dictionary keys.  This is an internal feature used to implement
         the parser.  Do not rely on it in application code.
         """
-        return int(id(self) % sys.maxint)
+        return int(id(self) % sys.maxsize)
 
     def opt_help(self):
         """
         Display this help and exit.
         """
-        print self.__str__()
+        print(self.__str__())
         sys.exit(0)
 
     def opt_version(self):
@@ -202,7 +206,7 @@ class Options(dict):
         Display Twisted version and exit.
         """
         from twisted import copyright
-        print "Twisted version:", copyright.version
+        print("Twisted version:", copyright.version)
         sys.exit(0)
 
     #opt_h = opt_help # this conflicted with existing 'host' options.
@@ -232,7 +236,7 @@ class Options(dict):
         try:
             opts, args = getopt.getopt(options,
                                        self.shortOpt, self.longOpt)
-        except getopt.error, e:
+        except getopt.error as e:
             raise UsageError(str(e))
 
         for opt, arg in opts:
@@ -432,13 +436,11 @@ class Options(dict):
                 reverse_dct[method] = []
             reverse_dct[method].append(name.replace('_', '-'))
 
-        cmpLength = lambda a, b: cmp(len(a), len(b))
-
         for method, names in reverse_dct.items():
             if len(names) < 2:
                 continue
             names_ = names[:]
-            names_.sort(cmpLength)
+            names_.sort(key=len)
             longest = names_.pop()
             for name in names_:
                 synonyms[name] = longest
@@ -535,7 +537,7 @@ class Options(dict):
 
         if longdesc:
             longdesc = ('\n' +
-                        '\n'.join(text.wordWrap(longdesc, width)).strip()
+                        '\n'.join(textwrap.wrap(longdesc, width)).strip()
                         + '\n')
 
         if optDicts:
@@ -935,7 +937,7 @@ def docMakeChunks(optList, width=80):
                 doc = "%s. %s" % (doc, d.doc)
 
         if doc:
-            column2_l = text.wordWrap(doc, colWidth2)
+            column2_l = textwrap.wrap(doc, colWidth2)
         else:
             column2_l = ['']
 
@@ -950,10 +952,10 @@ def docMakeChunks(optList, width=80):
 
 
 def flagFunction(method, name=None):
-    reqArgs = method.im_func.func_code.co_argcount
+    reqArgs = len(inspect.getargspec(method).args)
     if reqArgs > 2:
         raise UsageError('Invalid Option function for %s' %
-                         (name or method.func_name))
+                         (name or method.__name__))
     if reqArgs == 2:
         # argName = method.im_func.func_code.co_varnames[1]
         return 0
diff --git twisted/python/util.py twisted/python/util.py
index 78434f2..7e01957 100644
--- twisted/python/util.py
+++ twisted/python/util.py
@@ -1037,7 +1037,7 @@ __all__ = [
 if _PY3:
     __all3__ = ["FancyEqMixin", "untilConcludes",
                 "runWithWarningsSuppressed", "FancyStrMixin", "nameToLabel",
-                "InsensitiveDict"]
+                "InsensitiveDict", "padTo"]
     for name in __all__[:]:
         if name not in __all3__:
             __all__.remove(name)
diff --git twisted/test/test_usage.py twisted/test/test_usage.py
index b676b7e..82e5023 100644
--- twisted/test/test_usage.py
+++ twisted/test/test_usage.py
@@ -5,6 +5,10 @@
 Tests for L{twisted.python.usage}, a command line option parsing library.
 """
 
+from __future__ import division, absolute_import
+
+import sys
+
 from twisted.trial import unittest
 from twisted.python import usage
 
@@ -393,7 +397,7 @@ class HelpStringTest(unittest.TestCase):
         """
         try:
             self.nice.__str__()
-        except Exception, e:
+        except Exception as e:
             self.fail(e)
 
     def test_whitespaceStripFlagsAndParameters(self):
@@ -610,3 +614,97 @@ class CompleterNotImplementedTestCase(unittest.TestCase):
                 action = cls(None)
             self.assertRaises(NotImplementedError, action._shellCode,
                               None, "bad_shell_type")
+
+
+class FlagFunctionTest(unittest.TestCase):
+    """
+    Tests for L{usage.flagFunction}.
+    """
+
+    class SomeClass:
+        def oneArg(self, a):
+            pass
+
+        def noArg(self):
+            pass
+
+        def manyArgs(self, a, b, c):
+            pass
+
+
+    obj = SomeClass()
+
+
+    def test_hasArg(self):
+        """
+        L{usage.flagFunction} returns a truth value if the method checked allows
+        exactly one argument.
+        """
+        self.assertEqual(0, usage.flagFunction(self.obj.oneArg))
+
+
+    def test_noArg(self):
+        """
+        L{usage.flagFunction} returns a false value if the method checked allows
+        exactly no argument.
+        """
+        self.assertEqual(1, usage.flagFunction(self.obj.noArg))
+
+
+    def test_tooManyArguments(self):
+        """
+        L{usage.flagFunction} raises L{usage.UsageError} if the method checked
+        allows more than one argument.
+        """
+        try:
+            usage.flagFunction(self.obj.manyArgs)
+        except usage.UsageError as e:
+            self.assertEqual("Invalid Option function for manyArgs", str(e))
+        except:
+            self.fail("Didn't raise UsageError")
+
+
+    def test_tooManyArgumentsAndSpecificErrorMessage(self):
+        """
+        L{usage.flagFunction} allows to specify the name of the method in its
+        error message if the method allows too many arguments.
+        """
+        try:
+            usage.flagFunction(self.obj.manyArgs, "flubuduf")
+        except usage.UsageError as e:
+            self.assertEqual("Invalid Option function for flubuduf", str(e))
+        except:
+            self.fail("Didn't raise UsageError")
+
+
+class OptionsInternalTest(unittest.TestCase):
+    """
+    Tests internal behavior of C{usage.Options}.
+    """
+
+    def test_hashValue(self):
+        """
+        The hash of a L{usage.Options} object is limited by the size of the
+        maximum integer of the platform.
+        """
+        option = usage.Options()
+        self.assertLess(hash(option), sys.maxsize)
+
+
+    def test_optionsAliasesOrder(self):
+        """
+        Options which are synonyms to another option are aliases towards the
+        longest option name.
+        """
+        class Opts(usage.Options):
+            def opt_very_very_long(self):
+                pass
+
+            opt_short = opt_very_very_long
+            opt_s = opt_very_very_long
+
+        opts = Opts()
+
+        self.assertEqual("very-very-long", opts.synonyms['s'])
+        self.assertEqual("very-very-long", opts.synonyms['short'])
+        self.assertEqual("very-very-long", opts.synonyms['very-very-long'])
diff --git twisted/topfiles/7038.misc twisted/topfiles/7038.misc
new file mode 100644
index 0000000..e69de29
