--- /Users/davidjb/Downloads/Twisted-14.0.0/twisted/internet/abstract.py	2014-03-16 12:46:59.000000000 +0000
+++ ../internet/abstract.py	2015-05-19 14:24:48.000000000 +0100
@@ -273,7 +273,10 @@
                                               or self.producerPaused):
                 # tell them to supply some more.
                 self.producerPaused = False
-                self.producer.resumeProducing()
+                while (self.producer is not None and
+                       self.offset == len(self.dataBuffer) and
+                       not self._tempDataLen):
+                    self.producer.resumeProducing()
             elif self.disconnecting:
                 # But if I was previously asked to let the connection die, do
                 # so.
