diff -ru ./internet/iocpreactor/proactor.py ..\twisted\./internet/iocpreactor/proactor.py
--- ./internet/iocpreactor/proactor.py	2005-11-09 12:00:40.000000000 -0800
+++ ..\twisted\./internet/iocpreactor/proactor.py	2006-03-20 10:45:05.998235200 -0800
@@ -3,7 +3,7 @@
 
 
 from twisted.internet import defer, base, main
-from twisted.internet.interfaces import IReactorTCP, IReactorUDP, IReactorArbitrary, IReactorProcess
+from twisted.internet.interfaces import IReactorTCP, IReactorUDP, IReactorMulticast, IReactorArbitrary, IReactorProcess
 from twisted.python import threadable, log, reflect
 from zope.interface import implements, implementsOnly
 
@@ -11,10 +11,9 @@
 from _iocp import iocpcore
 
 class Proactor(iocpcore, base.ReactorBase, log.Logger):
-    # TODO: IReactorArbitrary, IReactorUDP, IReactorMulticast,
     # IReactorSSL (or leave it until exarkun finishes TLS)
     # IReactorCore (cleanup)
-    implementsOnly(IReactorTCP, IReactorUDP, IReactorArbitrary, IReactorProcess)
+    implementsOnly(IReactorTCP, IReactorUDP, IReactorMulticast, IReactorArbitrary, IReactorProcess)
     handles = None
     iocp = None
 
@@ -79,6 +78,12 @@
         p.startListening()
         return p
 
+
+    def listenMulticast(self, port, protocol, interface='', maxPacketSize=8192, listenMultiple=False):
+        p = udp.MulticastPort((interface, port), protocol, maxPacketSize)
+        p.startListening()
+        return p
+
     def connectUDPblah(self, remotehost, remoteport, protocol, localport=0,
                   interface='', maxPacketSize=8192):
         p = udp.ConnectedPort((remotehost, remoteport), (interface, localport), protocol, maxPacketSize)
diff -ru ./internet/iocpreactor/udp.py ..\twisted\./internet/iocpreactor/udp.py
--- ./internet/iocpreactor/udp.py	2005-11-22 14:28:52.000000000 -0800
+++ ..\twisted\./internet/iocpreactor/udp.py	2006-03-20 10:45:20.258740800 -0800
@@ -3,8 +3,11 @@
 
 
 import socket
+import struct
+import operator
 
 from twisted.internet import interfaces, defer, error, protocol, address
+from twisted.internet.udp import MulticastMixin
 from twisted.internet.abstract import isIPAddress
 from twisted.persisted import styles
 from twisted.python import log, failure, reflect
@@ -118,7 +121,7 @@
     def handle_listening_readErr(self, ret, bytes):
         log.msg("read failed with err %s" % (ret,))
         # TODO: use Failures or something
-        if ret == 1234: # ERROR_PORT_UNREACHABLE
+        if ret == ERROR_PORT_UNREACHABLE:
             self.protocol.connectionRefused()
         if self.reading:
             self.startReading()
@@ -192,3 +195,20 @@
     def getHost(self):
         return address.IPv4Address('UDP', *(self.socket.getsockname() + ('INET_UDP',)))
 
+
+class MulticastPort(MulticastMixin, Port):
+    """UDP Port that supports multicasting."""
+
+    implements(interfaces.IMulticastTransport)
+
+    def __init__(self, bindAddress, proto, maxPacketSize=8192, listenMultiple=False):
+        Port.__init__(self, bindAddress, proto, maxPacketSize)
+        self.listenMultiple = listenMultiple
+
+    def createInternetSocket(self):
+        skt = Port.createInternetSocket(self)
+        if self.listenMultiple:
+            skt.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+            if hasattr(socket, "SO_REUSEPORT"):
+                skt.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
+        return skt
\ No newline at end of file
diff -ru ./internet/udp.py ..\twisted\./internet/udp.py
--- ./internet/udp.py	2006-01-26 08:18:50.000000000 -0800
+++ ..\twisted\./internet/udp.py	2006-03-20 10:45:36.001377600 -0800
@@ -351,7 +351,10 @@
             cmd = socket.IP_ADD_MEMBERSHIP
         else:
             cmd = socket.IP_DROP_MEMBERSHIP
-        self.socket.setsockopt(socket.IPPROTO_IP, cmd, addr + interface)
+        try:
+            self.socket.setsockopt(socket.IPPROTO_IP, cmd, addr + interface)
+        except socket.error:
+            return 0
         return 1
 
     def leaveGroup(self, addr, interface=""):
Only in .: multicast.diff
