=== modified file 'doc/core/development/policy/svn-dev.xhtml'
--- doc/core/development/policy/svn-dev.xhtml	2010-03-01 20:30:56 +0000
+++ doc/core/development/policy/svn-dev.xhtml	2010-10-11 22:09:15 +0000
@@ -148,10 +148,10 @@
 instead.</p>
 
 <p>To build the HTML form of the howto/ docs, do the following. Note that
-the index file will be placed in <code>doc/howto/index.html</code>.</p>
+the index file will be placed in <code>doc/core/howto/index.html</code>.</p>
 
 <pre class="shell">
-./bin/lore/lore -p --config template=doc/howto/template.tpl doc/howto/*.xhtml
+./bin/lore/lore -p --config template=doc/core/howto/template.tpl doc/core/howto/*.xhtml
 </pre>
 
 <p>To run hlint over a single Lore document, such as
@@ -168,7 +168,7 @@
 <code>doc/development/policy/svn-dev.html</code>:</p>
 
 <pre class="shell">
-./bin/lore/lore -p --config template=doc/howto/template.tpl \
+./bin/lore/lore -p --config template=doc/core/howto/template.tpl \
    doc/development/policy/svn-dev.xhtml
 </pre>
 

=== modified file 'doc/lore/howto/lore.xhtml'
--- doc/lore/howto/lore.xhtml	2010-09-18 12:33:49 +0000
+++ doc/lore/howto/lore.xhtml	2010-10-11 22:32:05 +0000
@@ -253,7 +253,7 @@
 checkout, do:</p>
 
 <pre class="shell">
-% lore --config template=doc/howto/template.tpl doc/howto/*.xhtml
+% lore --config template=doc/core/howto/template.tpl doc/core/howto/*.xhtml
 </pre>
 
 <p>
@@ -261,7 +261,7 @@
 use a different extension:
 </p>
 <pre class="shell">
-% lore --config ext=.html doc/howto/*.xhtml
+% lore --config ext=.html doc/core/howto/*.xhtml
 </pre>
 <h2>Using Lore to Generate LaTex</h2>
 

=== modified file 'emacs/twisted-dev.el'
--- emacs/twisted-dev.el	2008-09-17 01:39:17 +0000
+++ emacs/twisted-dev.el	2010-10-11 23:11:10 +0000
@@ -168,8 +168,8 @@
 
 (defun twisted-dev-gendoc ()
   (interactive)
-  (with-cd (format "%s/doc/howto" twisted-dev-directory)
-    (compile (format "../../bin/lore -p %s" buffer-file-name))))
+  (with-cd (format "%s/doc/core/howto" twisted-dev-directory)
+    (compile (format "../../../bin/lore -p %s" buffer-file-name))))
 
 (defun twisted-dev-grep ()
   (interactive)

=== modified file 'twisted/internet/defer.py'
--- twisted/internet/defer.py	2010-09-18 17:51:09 +0000
+++ twisted/internet/defer.py	2010-10-11 23:07:43 +0000
@@ -190,7 +190,7 @@
     that come from outside packages that are not under our control, we use
     threads (see for example L{twisted.enterprise.adbapi}).
 
-    For more information about Deferreds, see doc/howto/core/defer.html or
+    For more information about Deferreds, see doc/core/howto/defer.html or
     U{http://twistedmatrix.com/documents/current/core/howto/defer.html}
 
     When creating a Deferred, you may provide a canceller function, which

=== modified file 'twisted/python/usage.py'
--- twisted/python/usage.py	2007-05-11 14:10:05 +0000
+++ twisted/python/usage.py	2010-10-11 23:05:45 +0000
@@ -9,7 +9,7 @@
 
 For information on how to use it, see
 U{http://twistedmatrix.com/projects/core/documentation/howto/options.html},
-or doc/howto/options.html in your Twisted directory.
+or doc/core/howto/options.xhtml in your Twisted directory.
 """
 
 # System Imports
@@ -127,7 +127,7 @@
     Advanced functionality is covered in the howto documentation,
     available at
     U{http://twistedmatrix.com/projects/core/documentation/howto/options.html},
-    or doc/howto/options.html in your Twisted directory.
+    or doc/core/howto/options.xhtml in your Twisted directory.
     """
 
     subCommand = None

