Index: soap.py
===================================================================
RCS file: /cvs/Twisted/twisted/web/soap.py,v
retrieving revision 1.4
diff -u -r1.4 soap.py
--- soap.py	14 Jan 2004 01:27:18 -0000	1.4
+++ soap.py	11 Feb 2004 06:11:42 -0000
@@ -71,7 +71,7 @@
 
         p, header, body, attrs = SOAPpy.parseSOAPRPC(data, 1, 1, 1)
 
-        methodName, args, kwargs, ns = p._name, p._aslist, p._asdict, p._ns
+        methodName, args, kwargs, ns = p._name, p._aslist(), p._asdict(), p._ns
         function, useKeywords = self.lookupFunction(methodName)
         
         if not function:
