Index: test_soap.py
===================================================================
RCS file: /cvs/Twisted/twisted/test/test_soap.py,v
retrieving revision 1.1
diff -u -r1.1 test_soap.py
--- test_soap.py	14 Jan 2004 01:27:22 -0000	1.1
+++ test_soap.py	5 Feb 2004 04:17:34 -0000
@@ -96,10 +96,10 @@
         x = self.proxy().callRemote("pair", 'a', 1)
         self.assertEquals(unittest.deferredResult(x), ['a', 1, None])
         x = self.proxy().callRemote("struct")
-        self.assertEquals(unittest.deferredResult(x)._asdict,
+        self.assertEquals(unittest.deferredResult(x)._asdict(),
                           {"a": "c"})
         x = self.proxy().callRemote("complex")
-        self.assertEquals(unittest.deferredResult(x)._asdict,
+        self.assertEquals(unittest.deferredResult(x)._asdict(),
                           {"a": ["b", "c", 12, []], "D": "foo"})
 
     def testErrors(self):
