--- TwistedMail-0.4.0/twisted/mail/test/test_imap.py.orig	2006-12-05 08:35:31.000000000 -0600
+++ TwistedMail-0.4.0/twisted/mail/test/test_imap.py	2007-05-30 13:13:06.000000000 -0500
@@ -1842,6 +1842,110 @@
             ).addCallback(strip(fetch)
             ).addCallback(test)
 
+    def testFlagsChangedInsideFetchSpecificResponse(self):
+
+        class StillSimplerClient(imap4.IMAP4Client):
+            events = []
+            def flagsChanged(self, newFlags):
+                self.events.append(['flagsChanged', newFlags])
+
+        transport = StringTransport()
+        c = StillSimplerClient()
+        c.makeConnection(transport)
+        c.lineReceived('* OK [IMAP4rev1]')
+
+        def login():
+            d = c.login('blah', 'blah')
+            c.dataReceived('0001 OK LOGIN\r\n')
+            return d
+        def select():
+            d = c.select('inbox')
+            c.lineReceived('0002 OK SELECT')
+            return d
+        def fetch():
+            d = c.fetchSpecific('1:*',
+                headerType='HEADER.FIELDS',
+                headerArgs=['SUBJECT'])
+            c.dataReceived('* 1 FETCH (BODY[HEADER.FIELDS ("SUBJECT")] {38}\r\n')
+            c.dataReceived('Subject: Suprise for your woman...\r\n')
+            c.dataReceived('\r\n')
+            c.dataReceived(' FLAGS (\Seen))\r\n')
+            c.dataReceived('* 2 FETCH (FLAGS (\Seen) BODY[HEADER.FIELDS ("SUBJECT")] {75}\r\n')
+            c.dataReceived('Subject: What you been doing. Order your meds here . ,. handcuff madsen\r\n')
+            c.dataReceived('\r\n')
+            c.dataReceived(')\r\n')
+            c.dataReceived('0003 OK FETCH completed\r\n')
+            return d
+        def test(res):
+            self.assertEquals(res, {
+                1: [['BODY', ['HEADER.FIELDS', ['SUBJECT']],
+                    'Subject: Suprise for your woman...\r\n\r\n']],
+                2: [['BODY', ['HEADER.FIELDS', ['SUBJECT']],
+                    'Subject: What you been doing. Order your meds here . ,. handcuff madsen\r\n\r\n']]
+            })
+
+            self.assertEquals(c.events, [
+                ['flagsChanged', {1: ['\\Seen']}],
+                ['flagsChanged', {2: ['\\Seen']}]
+            ])
+
+        return login(
+            ).addCallback(strip(select)
+            ).addCallback(strip(fetch)
+            ).addCallback(test)
+
+    def testFlagsChangedInsideFetchMessageResponse(self):
+
+        class StillSimplerClient(imap4.IMAP4Client):
+            events = []
+            def flagsChanged(self, newFlags):
+                self.events.append(['flagsChanged', newFlags])
+
+        transport = StringTransport()
+        c = StillSimplerClient()
+        c.makeConnection(transport)
+        c.lineReceived('* OK [IMAP4rev1]')
+
+        def login():
+            d = c.login('blah', 'blah')
+            c.dataReceived('0001 OK LOGIN\r\n')
+            return d
+        def select():
+            d = c.select('inbox')
+            c.lineReceived('0002 OK SELECT')
+            return d
+        def fetch():
+            d = c.fetchMessage('1:*')
+            c.dataReceived('* 1 FETCH (RFC822 {38}\r\n')
+            c.dataReceived('Subject: Suprise for your woman...\r\n')
+            c.dataReceived('\r\n')
+            c.dataReceived(' FLAGS (\Seen))\r\n')
+            c.dataReceived('* 2 FETCH (FLAGS (\Recent \Seen) RFC822 {75}\r\n')
+            c.dataReceived('Subject: What you been doing. Order your meds here . ,. handcuff madsen\r\n')
+            c.dataReceived('\r\n')
+            c.dataReceived(')\r\n')
+            c.dataReceived('0003 OK FETCH completed\r\n')
+            return d
+        def test(res):
+            self.assertEquals(res, {
+                1: {'RFC822': 'Subject: Suprise for your woman...\r\n\r\n',
+                    'FLAGS': ['\\Seen']
+                    },
+                2: {'RFC822': 'Subject: What you been doing. Order your meds here . ,. handcuff madsen\r\n\r\n',
+                    'FLAGS': ['\\Recent', '\\Seen']
+                    },
+            })
+
+            self.assertEquals(c.events, [
+                ['flagsChanged', {1: ['\\Seen']}],
+                ['flagsChanged', {2: ['\\Recent', '\\Seen']}]
+            ])
+
+        return login(
+            ).addCallback(strip(select)
+            ).addCallback(strip(fetch)
+            ).addCallback(test)
+
 class FakeyServer(imap4.IMAP4Server):
     state = 'select'
     timeout = None
