--- TwistedMail-0.4.0/twisted/mail/imap4.py.orig	2006-12-05 08:35:31.000000000 -0600
+++ TwistedMail-0.4.0/twisted/mail/imap4.py	2007-05-29 17:45:45.000000000 -0500
@@ -364,15 +364,9 @@
             names = parseNestedParens(L)
             N = len(names)
             if (N >= 1 and names[0] in self._1_RESPONSES or
+                N >= 2 and names[1] in self._2_RESPONSES or
                 N >= 2 and names[0] == 'OK' and isinstance(names[1], types.ListType) and names[1][0] in self._OK_RESPONSES):
                 send.append(L)
-            elif N >= 3 and names[1] in self._2_RESPONSES:
-                if isinstance(names[2], list) and len(names[2]) >= 1 and names[2][0] == 'FLAGS' and 'FLAGS' not in self.args:
-                    unuse.append(L)
-                else:
-                    send.append(L)
-            elif N >= 2 and names[1] in self._2_RESPONSES:
-                send.append(L)
             else:
                 unuse.append(L)
         d, self.defer = self.defer, None
@@ -3336,6 +3330,8 @@
                             if len(data) < 2:
                                 raise IllegalServerResponse("Not enough arguments", data)
                             flags.setdefault(id, {})[data[0]] = data[1]
+                            if data[0] == 'FLAGS':
+                                self.flagsChanged({id: data[1]})
                             del data[:2]
                 else:
                     print '(2)Ignoring ', parts
@@ -3431,7 +3427,15 @@
                     except ValueError:
                         raise IllegalServerResponse, line
                     else:
-                        info[id] = parseNestedParens(parts[2])
+                        data = parseNestedParens(parts[2])[0]
+                        # XXX this will fail if 'FLAGS' is a retrieved part
+                        for i in range(len(data) - 1):
+                            if data[i] == 'FLAGS':
+                                self.flagsChanged({id: data[i+1]})
+                                del data[i:i+2]
+                                break
+                        if data:
+                            info.setdefault(id, []).append(data)
         return info
 
     def _fetch(self, messages, useUID=0, **terms):
