Index: twisted/protocols/amp.py
===================================================================
--- twisted/protocols/amp.py	(revision 23462)
+++ twisted/protocols/amp.py	(working copy)
@@ -119,8 +119,8 @@
 documentation will refer to packets as if they were newline delimited, like
 so::
 
+    C: _ask: ef639e5c892ccb54
     C: _command: sum
-    C: _ask: ef639e5c892ccb54
     C: a: 13
     C: b: 81
 
