Index: twisted/protocols/ftp.py
===================================================================
--- twisted/protocols/ftp.py	(revision 30599)
+++ twisted/protocols/ftp.py	(working copy)
@@ -506,6 +506,8 @@
     # -- configuration variables --
     peerCheck = False
 
+    protocol = DTP
+
     # -- class variables --
     def __init__(self, pi, peerHost=None, reactor=None):
         """Constructor
@@ -530,7 +532,7 @@
         self._state = self._FINISHED
 
         self.cancelTimeout()
-        p = DTP()
+        p = self.protocol()
         p.factory = self
         p.pi = self.pi
         self.pi.dtpInstance = p
