Index: twisted/protocols/ftp.py
===================================================================
--- twisted/protocols/ftp.py	(revision 30599)
+++ twisted/protocols/ftp.py	(working copy)
@@ -662,6 +662,8 @@
     dtpInstance = None
     binary = True
 
+    dtpFactoryClass = DTPFactory
+
     passivePortRange = xrange(0, 1)
 
     listenFactory = reactor.listenTCP
@@ -848,7 +850,7 @@
             # cleanupDTP sets dtpFactory to none.  Later we'll do
             # cleanup here or something.
             self.cleanupDTP()
-        self.dtpFactory = DTPFactory(pi=self)
+        self.dtpFactory = self.dtpFactoryClass(pi=self)
         self.dtpFactory.setTimeout(self.dtpTimeout)
         self.dtpPort = self.getDTPPort(self.dtpFactory)
 
@@ -867,7 +869,7 @@
         if self.dtpFactory is not None:
             self.cleanupDTP()
 
-        self.dtpFactory = DTPFactory(pi=self, peerHost=self.transport.getPeer().host)
+        self.dtpFactory = self.dtpFactoryClass(pi=self, peerHost=self.transport.getPeer().host)
         self.dtpFactory.setTimeout(self.dtpTimeout)
         self.dtpPort = reactor.connectTCP(ip, port, self.dtpFactory)
 
