Index: doc/core/howto/cred.xhtml
===================================================================
--- doc/core/howto/cred.xhtml	(revision 27222)
+++ doc/core/howto/cred.xhtml	(working copy)
@@ -88,15 +88,14 @@
 </li></ul>
 
     <h3>The CredentialChecker</h3>
-<p>This is an object implementing <code class="API"
-base="twisted.cred.checkers">ICredentialsChecker</code> which resolves some
-Credentials to an avatar ID.
-Some examples of CredentialChecker implementations would be:
-InMemoryUsernamePassword, ApacheStyleHTAccessFile,
-UNIXPasswordDatabase, SSHPublicKeyDatabase.  A credential checker
-stipulates some requirements of the credentials it can check by
-specifying a credentialInterfaces attribute, which is a list of
-interfaces.  Credentials passed to its requestAvatarId method must
+<p>This is an object implementing <code class="API" base="twisted.cred.checkers">ICredentialsChecker</code> which resolves some
+Credentials to an avatar ID.  Some examples of CredentialChecker
+implementations could be: InMemoryUsernamePassword, ApacheStyleHTAccessFile,
+UNIXPasswordDatabase, SSHPublicKeyDatabase (please see
+<code class="API" base="twisted.cred.checkers">ICredentialsChecker</code> for known
+implementations).  A credential checker stipulates some requirements of the
+credentials it can check by specifying a credentialInterfaces attribute, which
+is a list of interfaces.  Credentials passed to its requestAvatarId method must
 implement one of those interfaces.</p>
 
 <p>For the most part, these things will just check usernames and passwords
