Index: twisted/python/filepath.py
===================================================================
--- twisted/python/filepath.py	(revision 36417)
+++ twisted/python/filepath.py	(working copy)
@@ -970,7 +970,7 @@
     def getPermissions(self):
         """
         Returns the permissions of the file.  Should also work on Windows,
-        however, those permissions may not what is expected in Windows.
+        however, those permissions may not be what is expected in Windows.
 
         @return: the permissions for the file
         @rtype: L{Permissions}
