diff -ur Twisted/twisted/python/filepath.py Twisted-Typo/twisted/python/filepath.py
--- Twisted/twisted/python/filepath.py	Sat Oct 27 04:03:05 2012
+++ Twisted-Typo/twisted/python/filepath.py	Sat Oct 27 20:45:17 2012
@@ -970,7 +970,7 @@
     def getPermissions(self):
         """
         Returns the permissions of the file.  Should also work on Windows,
-        however, those permissions may not what is expected in Windows.
+        however, those permissions may not be what is expected in Windows.
 
         @return: the permissions for the file
         @rtype: L{Permissions}
