--- interfaces.py	2013-03-26 00:02:44.000000000 +0530
+++ mod_interfaces.py	2013-03-28 23:05:51.030156530 +0530
@@ -1,8 +1,11 @@
+# Copyright (c) Twisted Matrix Laboratories.
+# See LICENSE for details.
+
 """
 Twisted Spread Interfaces.
 
-This module is unused so far. It's also undecided whether this module
-will remain monolithic.
+This module is imported in twisted.spread.jelly
+It undecided whether this module will remain monolithic.
 """
 
 from zope.interface import Interface
