Index: twisted/test/test_ftp.py
===================================================================
--- twisted/test/test_ftp.py	(revision 24139)
+++ twisted/test/test_ftp.py	(working copy)
@@ -10,7 +10,6 @@
 
 import os.path
 from StringIO import StringIO
-import shutil
 import errno
 
 from zope.interface import implements
@@ -107,8 +106,6 @@
 
     def ebTearDown(self, ignore):
         del self.serverProtocol
-        # Clean up temporary directory
-        shutil.rmtree(self.directory)
 
     def assertCommandResponse(self, command, expectedResponseLines,
                               chainDeferred=None):
