diff --git a/doc/core/howto/logging.xhtml b/doc/core/howto/logging.xhtml
index d38634c..2e510ba 100644
--- a/doc/core/howto/logging.xhtml
+++ b/doc/core/howto/logging.xhtml
@@ -54,7 +54,7 @@ log.startLogging(sys.stdout)
         </pre>
       or:
         <pre class="python">
-from twisted.python.logging import DailyLogFile
+from twisted.python.logfile import DailyLogFile
 
 log.startLogging(DailyLogFile.fromFullPath("/var/log/foo.log"))
         </pre>
diff --git a/twisted/topfiles/5083.misc b/twisted/topfiles/5083.misc
new file mode 100644
index 0000000..e69de29
