Index: twisted/python/logfile.py
===================================================================
--- twisted/python/logfile.py	(revision 41845)
+++ twisted/python/logfile.py	(working copy)
@@ -63,22 +63,19 @@
         if os.path.exists(self.path):
             self._file = file(self.path, "r+", 1)
             self._file.seek(0, 2)
-        else:
             if self.defaultMode is not None:
-                # Set the lowest permissions
-                oldUmask = os.umask(0777)
                 try:
-                    self._file = file(self.path, "w+", 1)
-                finally:
-                    os.umask(oldUmask)
+                    os.chmod(self.path, self.defaultMode)
+                except OSError:
+                    # Probably /dev/null or something?
+                    pass
+        else:
+            if self.defaultMode is not None:
+                # Open with the defaultMode
+                self._file = os.fdopen(os.open(self.path, 
+                    os.O_CREAT|os.O_RDWR, self.defaultMode))
             else:
                 self._file = file(self.path, "w+", 1)
-        if self.defaultMode is not None:
-            try:
-                os.chmod(self.path, self.defaultMode)
-            except OSError:
-                # Probably /dev/null or something?
-                pass
 
     def __getstate__(self):
         state = self.__dict__.copy()
