Index: twisted/test/test_logfile.py
===================================================================
--- twisted/test/test_logfile.py    (revision 41933)
+++ twisted/test/test_logfile.py    (working copy)
@@ -253,7 +253,7 @@
         Specifying an invalid directory to L{LogFile} raises C{IOError}.
         """
         e = self.assertRaises(
-            IOError, logfile.LogFile, self.name, 'this_dir_does_not_exist')
+            OSError, logfile.LogFile, self.name, 'this_dir_does_not_exist')
         self.assertEqual(e.errno, errno.ENOENT)
 
 
Index: twisted/python/logfile.py
===================================================================
--- twisted/python/logfile.py   (revision 41933)
+++ twisted/python/logfile.py   (working copy)
@@ -60,26 +60,19 @@
         Open the log file.
         """
         self.closed = False
-        if os.path.exists(self.path):
-            self._file = file(self.path, "r+", 1)
-            self._file.seek(0, 2)
+        if self.defaultMode is None:
+            self._file = self._file = os.fdopen(os.open(
+                self.path, os.O_CREAT|os.O_RDWR), 'r+', 1)
         else:
-            if self.defaultMode is not None:
-                # Set the lowest permissions
-                oldUmask = os.umask(0777)
-                try:
-                    self._file = file(self.path, "w+", 1)
-                finally:
-                    os.umask(oldUmask)
-            else:
-                self._file = file(self.path, "w+", 1)
-        if self.defaultMode is not None:
+            self._file = self._file = os.fdopen(os.open(
+                self.path, os.O_CREAT|os.O_RDWR, 0000), 'r+', 1) 
             try:
                 os.chmod(self.path, self.defaultMode)
             except OSError:
-                # Probably /dev/null or something?
                 pass
+        self._file.seek(0,2)
 
+
     def __getstate__(self):
         state = self.__dict__.copy()
         del state["_file"]
Index: twisted/topfiles/7026.bugfix
===================================================================
--- twisted/topfiles/7026.bugfix    (revision 0)
+++ twisted/topfiles/7026.bugfix    (revision 0)
@@ -0,0 +1 @@
+twisted.python.logfile now retains the correct system umask when concurrently rotating logfiles.
