Index: twisted/scripts/mktap.py
===================================================================
--- twisted/scripts/mktap.py	(revision 13186)
+++ twisted/scripts/mktap.py	(working copy)
@@ -133,7 +133,7 @@
         for (name, module) in tapLookup.iteritems():
             if IServiceMaker.providedBy(module):
                 sc.append((
-                    name, None, lambda m=module: m.options, module.description))
+                    name, None, lambda m=module: m.options(), module.description))
             else:
                 sc.append((
                     name, None, lambda obj=module: obj.load().Options(),
