Index: twisted/words/protocols/irc.py
===================================================================
--- twisted/words/protocols/irc.py	(Revision 28110)
+++ twisted/words/protocols/irc.py	(Arbeitskopie)
@@ -744,7 +744,7 @@
             raise ValueError(
                 'Expecting a maximum of %d channel mode parameters, got %d' % (
                     len(names), len(params)))
-        items = map(lambda key, value: (key, value or ''), names, params)
+        items = zip(names, params.split(','))
         return dict(items)
     _parseChanModesParam = classmethod(_parseChanModesParam)
 
