Index: microdom.py
===================================================================
RCS file: /cvs/Twisted/twisted/web/microdom.py,v
retrieving revision 1.101
diff -u -3 -p -r1.101 microdom.py
--- microdom.py	13 Sep 2003 03:13:33 -0000	1.101
+++ microdom.py	25 Feb 2004 15:00:35 -0000
@@ -683,7 +683,7 @@ class MicroDOMParser(XMLParser):
         self.elementstack.append(el)
         if parent:
             parent.appendChild(el)
-        if (self.beExtremelyLenient and name in self.soonClosers):
+        if (self.beExtremelyLenient and el.tagName in self.soonClosers):
             self.gotTagEnd(name)
 
     def _gotStandalone(self, factory, data):
