Index: basics.xhtml
===================================================================
--- basics.xhtml	(revision 26132)
+++ basics.xhtml	(working copy)
@@ -86,11 +86,6 @@
 Debian.
 </p>
 
-<p>
-Right now, there is a beta Debian archive of a web server available at
-<a href="http://twistedmatrix.com/users/moshez/apt">Moshe's archive</a>.
-</p>
-
 <h2>tap2rpm</h2>
 
 <p><code class="shell">tap2rpm</code> is similar to <code class="shell">tap2deb</code>, except that
