Index: twisted/test/test_failure.py
===================================================================
--- twisted/test/test_failure.py	(revision 28118)
+++ twisted/test/test_failure.py	(working copy)
@@ -196,6 +196,36 @@
         f = failure.Failure(Exception("some error"))
         self.assertEqual(f.getTracebackObject(), None)
 
+    def test_noFrameLocalsStringify(self):
+        class _Decoy(object):
+            def __init__(self, ref):
+                self.ref = ref
+
+            def __repr__(self):
+                msg = "should not have called __repr__ while printing tb"
+                self.ref.fail(msg)
+
+        try:
+            decoy = _Decoy(self)
+            1/0
+        except:
+            f = failure.Failure()
+            f.cleanFailure()
+
+    def test_verboseTracebackDoesWarn(self):
+        out = StringIO.StringIO()
+
+        try:
+            1/0
+        except:
+            f = failure.Failure()
+            self.assertWarns(
+                DeprecationWarning,
+                "Verbose backtraces are being deprecated",
+                failure.__file__,
+                f.printDetailedTraceback,
+                out)
+
 class FindFailureTests(unittest.TestCase):
     """
     Tests for functionality related to L{Failure._findFailure}.
Index: twisted/python/failure.py
===================================================================
--- twisted/python/failure.py	(revision 28118)
+++ twisted/python/failure.py	(working copy)
@@ -50,6 +50,10 @@
             w( '  File "%s", line %s, in %s\n' % (filename, lineno, method))
             w( '    %s\n' % linecache.getline(filename, lineno).strip())
     elif detail == "verbose":
+        import warnings
+        msg = "Verbose backtraces are being deprecated"
+        warnings.warn(msg, DeprecationWarning, stacklevel=2)
+
         for method, filename, lineno, localVars, globalVars in frames:
             w("%s:%d: %s(...)\n" % (filename, lineno, method))
             w(' [ Locals ]\n')
@@ -403,18 +407,18 @@
         return "[Failure instance: %s]" % self.getBriefTraceback()
 
     def __getstate__(self):
-        """Avoid pickling objects in the traceback.
+        """ 
+        When serializing Failures, we strip out all the local/global references
+        from the saved stack frames. This is to avoid chains of circular
+        references in the traceback.
         """
         if self.pickled:
             return self.__dict__
         c = self.__dict__.copy()
 
         c['frames'] = [
-            [
-                v[0], v[1], v[2],
-                [(j[0], reflect.safe_repr(j[1])) for j in v[3]],
-                [(j[0], reflect.safe_repr(j[1])) for j in v[4]]
-            ] for v in self.frames
+            [ f[0], f[1], f[2], None, None ] 
+            for f in self.frames 
         ]
 
         # added 2003-06-23. See comment above in __init__
@@ -424,16 +428,13 @@
             # XXX: This is a band-aid.  I can't figure out where these
             # (failure.stack is None) instances are coming from.
             c['stack'] = [
-                [
-                    v[0], v[1], v[2],
-                    [(j[0], reflect.safe_repr(j[1])) for j in v[3]],
-                    [(j[0], reflect.safe_repr(j[1])) for j in v[4]]
-                ] for v in self.stack
+                [ f[0], f[1], f[2], None, None ]
+                for f in self.stack
             ]
-
+            
         c['pickled'] = 1
         return c
-
+ 
     def cleanFailure(self):
         """Remove references to other objects, replacing them with strings.
         """
