diff --git twisted/internet/defer.py twisted/internet/defer.py
index 76d1b9c..06760ad 100644
--- twisted/internet/defer.py
+++ twisted/internet/defer.py
@@ -1361,6 +1361,9 @@ class DeferredSemaphore(_ConcurrencyPrimitive):
     """
     A semaphore for event driven systems.
 
+    If you are looking into this as a means of limiting parallelism, you might
+    find L{twisted.internet.task.Cooperator} more useful.
+
     @ivar tokens: At most this many users may acquire this semaphore at
         once.
     @type tokens: C{int}
diff --git twisted/internet/task.py twisted/internet/task.py
index 5d8ff4b..0def17c 100644
--- twisted/internet/task.py
+++ twisted/internet/task.py
@@ -499,6 +499,26 @@ class CooperativeTask(object):
 class Cooperator(object):
     """
     Cooperative task scheduler.
+
+    A cooperative task is just an iterator where, as far as C{Cooperator} is
+    concerned, each iteration represents an atomic unit of work.  When the
+    iterator yields, it allows the C{Cooperator} to decide which unit of work
+    to do next.  If the iterator yields a L{defer.Deferred}, then work will
+    only resume after it fires and completes its callback chain.
+
+    There are two ways to add iterators to a C{Cooperator}, L{cooperate} and
+    L{coiterate}.  Both are equivalent, but C{coiterate} returns a
+    L{defer.Deferred} that fires when the task is done.
+
+    When a C{Cooperator} has more than one task, it distribute work between
+    all tasks.
+
+    C{Cooperator} can be used for many things, including but not limited to:
+
+      - run one or more computationally intensive tasks without blocking
+      - execute a particular number of tasks in parellel (i.e. limit
+        parallelism)
+      - do one thing, wait for a Deferred to fire, do the next thing, repeat
     """
 
     def __init__(self,
@@ -536,6 +556,9 @@ class Cooperator(object):
         Add an iterator to the list of iterators this L{Cooperator} is
         currently running.
 
+        Equivalent to L{cooperate}, but returns a L{defer.Deferred} that will
+        be fired when the task is done.
+
         @param doneDeferred: If specified, this will be the Deferred used as
             the completion deferred.  It is suggested that you use the default,
             which creates a new Deferred for you.
diff --git twisted/topfiles/6213.doc twisted/topfiles/6213.doc
new file mode 100644
index 0000000..aa252f5
--- /dev/null
+++ twisted/topfiles/6213.doc
@@ -0,0 +1 @@
+Document how to use Cooperator, and why you might want to.
