Index: twisted/internet/interfaces.py
===================================================================
--- twisted/internet/interfaces.py	(révision 24691)
+++ twisted/internet/interfaces.py	(copie de travail)
@@ -1259,11 +1259,12 @@
 
 
 class IProcessTransport(ITransport):
-    """A process transport.
-
-    @ivar pid: The Process-ID of this process.
     """
+    A process transport.
+    """
 
+    pid = Attribute("The Process-ID of this process.")
+
     def closeStdin():
         """Close stdin after all data has been written out."""
 
