Index: twisted/test/generator_failure_tests.py
===================================================================
--- twisted/test/generator_failure_tests.py	(revision 24958)
+++ twisted/test/generator_failure_tests.py	(working copy)
@@ -60,15 +60,18 @@
         It should be possible to throw the exception that a Failure
         represents into a generator.
         """
-        stuff = []
-        def generator():
+        exec """
+        def generator(stuff):
             try:
                 yield
             except:
                 stuff.append(sys.exc_info())
             else:
                 self.fail("Yield should have yielded exception.")
-        g = generator()
+        """.strip()
+        
+        stuff = []
+        g = generator(stuff)
         f = getDivisionFailure()
         g.next()
         self._throwIntoGenerator(f, g)
