Index: twisted/lore/tree.py
===================================================================
--- twisted/lore/tree.py	(revision 25178)
+++ twisted/lore/tree.py	(working copy)
@@ -183,11 +183,18 @@
         data = '\n'.join(lines[int(node.getAttribute('skipLines', 0)):])
         data = cStringIO.StringIO(text.removeLeadingTrailingBlanks(data))
         htmlizer.filter(data, outfile, writer=htmlizer.SmallerHTMLWriter)
-        val = outfile.getvalue()
+        val = _addLineNumbers(str(outfile.getvalue()))
         _replaceWithListing(node, val, filename, "py-listing")
 
+def _addLineNumbers(val):
+    linecol = []
+    linecol.append('<table><tr><td><pre>')
+    for linenum, line in enumerate(val.splitlines()[1:-1]):
+        linecol.append('%3d' % (linenum + 1))
+    linecol.append('</pre></td><td>')
+    val = '\n'.join(linecol) + val + '</td></tr></table>\n'
+    return val
 
-
 def _replaceWithListing(node, val, filename, class_):
     captionTitle = domhelpers.getNodeText(node)
     if captionTitle == os.path.basename(filename):
