Index: twisted/lore/test/test_lore.py
===================================================================
--- twisted/lore/test/test_lore.py	(revision 24967)
+++ twisted/lore/test/test_lore.py	(working copy)
@@ -328,9 +328,28 @@
                                                  "class",
                                                  "index-link"))
 
+    def test_lineNumbering(self):
+        """
+        L{tree._addLineNumbers} accepts a C{str} and returns a C{str} with
+        a column of line numbers prepended.  The column is enclosed in tags
+        such that it will be rendered as a floating CSS label.  The first six
+        characters of the C{str} are assumed to be '<pre>\n'; these characters
+        are stripped, since the column's tags include another <pre>.  Likewise,
+        the last line is assumed to be '</pre>' and is not numbered.
+        """
+        testcode = u'<pre>\n' \
+                   'Lorem ipsum\n' \
+                   'dolor sit\n' \
+                   '</pre>'
+        self.assertEqual(tree._addLineNumbers(testcode),
+                         u'<pre>'
+                         '<p style="float: left; margin-top: 0px; margin-left:'
+                         ' -15px; background-color: #dddddd">  1\n  2\n</p>'
+                         'Lorem ipsum\n' 
+                         'dolor sit\n'
+                         '</pre>')
 
 
-
 class LatexSpitterTestCase(unittest.TestCase):
     """
     Tests for the Latex output plugin.
