diff --git a/twisted/python/win32.py b/twisted/python/win32.py
--- a/twisted/python/win32.py
+++ b/twisted/python/win32.py
@@ -158,7 +158,7 @@ class _ErrorFormatter(object):
         @rtype: C{str}
         """
         if self.winError is not None:
-            return self.winError(errorcode)[1]
+            return self.winError(errorcode).strerror
         if self.formatMessage is not None:
             return self.formatMessage(errorcode)
         if self.errorTab is not None:
