diff -r f0773f8fb254 twisted/python/win32.py
--- a/twisted/python/win32.py	Fri Nov 04 22:39:52 2011 +0100
+++ b/twisted/python/win32.py	Sun Nov 06 18:45:45 2011 +0100
@@ -156,7 +156,7 @@
         @rtype: C{str}
         """
         if self.winError is not None:
-            return self.winError(errorcode)[1]
+            return self.winError(errorcode).strerror
         if self.formatMessage is not None:
             return self.formatMessage(errorcode)
         if self.errorTab is not None:
diff -r f0773f8fb254 twisted/test/test_strerror.py
--- a/twisted/test/test_strerror.py	Fri Nov 04 22:39:52 2011 +0100
+++ b/twisted/test/test_strerror.py	Sun Nov 06 18:45:45 2011 +0100
@@ -14,6 +14,9 @@
 from twisted.python.runtime import platform
 
 
+class _MyWindowsException(OSError):
+    pass
+
 
 class ErrorFormatingTestCase(TestCase):
     """
@@ -81,7 +84,7 @@
         winCalls = []
         def winError(errorCode):
             winCalls.append(errorCode)
-            return (errorCode, self.probeMessage)
+            return _MyWindowsException(errorCode, self.probeMessage)
         formatter = _ErrorFormatter(
             winError,
             lambda error: 'formatMessage: wrong message',
@@ -101,7 +104,7 @@
             from ctypes import WinError
             self.assertEqual(
                 formatter.formatError(self.probeErrorCode),
-                WinError(self.probeErrorCode)[1])
+                WinError(self.probeErrorCode).strerror)
             formatter.winError = None
 
         if formatter.formatMessage is not None:
@@ -130,7 +133,7 @@
         acceptable = [socket.errorTab[ECONNABORTED]]
         try:
             from ctypes import WinError
-            acceptable.append(WinError(ECONNABORTED)[1])
+            acceptable.append(WinError(ECONNABORTED).strerror)
         except ImportError:
             pass
         try:
