diff --git a/twisted/internet/abstract.py b/twisted/internet/abstract.py
index 6e69f4a..cd313d6 100644
--- a/twisted/internet/abstract.py
+++ b/twisted/internet/abstract.py
@@ -8,6 +8,7 @@ Support for generic select()able objects.
 
 from __future__ import division, absolute_import
 
+import sys
 from socket import AF_INET6, inet_pton, error
 
 from zope.interface import implementer
@@ -17,7 +18,7 @@ from twisted.python.compat import _PY3, unicode, lazyByteSlice
 from twisted.python import _reflectpy3 as reflect, failure
 from twisted.internet import interfaces, main
 
-if _PY3:
+if _PY3 or sys.platform.startswith('java'):
     def _concatenate(bObj, offset, bArray):
         # Python 3 lacks the buffer() builtin and the other primitives don't
         # help in this case.  Just do the copy.  Perhaps later these buffers can
