diff --git a/twisted/python/compat.py b/twisted/python/compat.py
index a8fc936..1d11353 100644
--- a/twisted/python/compat.py
+++ b/twisted/python/compat.py
@@ -357,7 +357,27 @@ else:
         return b"%d" % i
 
 
-    lazyByteSlice = buffer
+    if sys.platform.startswith('java'):
+        def lazyByteSlice(object, offset=0, size=None):
+            """
+            Return a copy of the given bytes-like object.
+
+            If an offset is given, the copy starts at that offset. If a size
+            is given, the copy will only be of that length.
+
+            @param object: C{bytes} to be copied.
+
+            @param offset: C{int}, starting index of copy.
+
+            @param size: Optional, if an C{int} is given limit the length of
+                copy to this size.
+            """
+            if size is None:
+                return object[offset:]
+            else:
+                return object[offset:(offset + size)]
+    else:
+        lazyByteSlice = buffer
 
     def networkString(s):
         if not isinstance(s, str):
