--- jelly.py-old	2005-02-17 09:53:56.548640144 -0600
+++ jelly.py	2005-02-17 10:48:44.187843016 -0600
@@ -795,6 +795,10 @@
                              "str": 1,
                              "int": 1,
                              "float": 1,
+                             "datetime": 1,
+                             "time": 1,
+                             "date": 1,
+                             "timedelta": 1,
                              "NoneType": 1}
         if hasattr(types, 'UnicodeType'):
             self.allowedTypes['unicode'] = 1
