--- trunk/twisted/python/urlpath.py	2006-08-23 11:57:09.000000000 -0600
+++ branches/netloc-2093/twisted/python/urlpath.py	2006-09-16 00:32:50.000000000 -0600
@@ -7,6 +7,69 @@
 import urlparse
 import urllib
 
+class URIAuthority:
+    """
+    An abstraction of the "authority" section of a URI.
+
+    Term taken from RFC2396.
+    """
+    def __init__(self, user=None, password=None, host=None, port=0):
+        self.user = user
+        self.password = password
+        self.host = host
+        self.port = int(port)
+        self.auth = self.unparse()
+
+    def fromString(self, auth):
+        """
+        Return a tuple representing the parts of the authority.
+        """
+        self.auth = auth
+        auths = auth.split('@')
+        if len(auths) == 2:
+            userpass = auths.pop(0)
+            userpass = userpass.split(':')
+            self.user = userpass.pop(0)
+            try:
+                self.password = userpass.pop(0)
+            except IndexError:
+                self.password = ''
+        else:
+            self.user = self.password = ''
+        hostport = auths[0].split(':')
+        self.host = hostport.pop(0) or ''
+        try:
+            self.port = int(hostport.pop(0))
+        except IndexError, ValueError:
+            self.port = None
+        return (self.user, self.password, self.host, self.port)
+
+    fromString = classmethod(fromString)
+
+    def unparse(self):
+        """
+        Return a string representing the URI authority
+        """
+        user = self.user or ''
+        password = host = port = ''
+        if self.password:
+            password = ':%s' % self.password
+        if self.port:
+            port = ':%s' % self.port
+        if self.user or password:
+            host = '@%s' % self.host
+        else:
+            host = self.host
+        return '%s%s%s%s' % (user, password, host, port)
+
+    def __str__(self):
+        return self.unparse()
+
+    def __repr__(self):
+        return ('URIAuthority(user=%r, password=%r, host=%r, port=%r)' 
+            % (self.user, self.password, self.host, self.port))
+            
+
 class URLPath:
     def __init__(self, scheme='', netloc='localhost', path='',
                  query='', fragment=''):
@@ -15,6 +78,8 @@
         self.path = path or '/'
         self.query = query
         self.fragment = fragment
+        auth = URIAuthority.fromString(netloc)
+        self.user, self.password, self.host, self.port = auth
 
     _qpathlist = None
     _uqpathlist = None
