Index: twisted/internet/test/test_inotify.py
===================================================================
--- twisted/internet/test/test_inotify.py	(revision 30608)
+++ twisted/internet/test/test_inotify.py	(working copy)
@@ -378,7 +378,29 @@
         subdir.createDirectory()
         return d
 
+    def test_watchIgnoreWatch(self):
+        """
+        Test to make sure series of ignore/watch
+        on  L{inotify.INotify} same path works.
 
+        This also tests that _isWatch actually returns
+        file descriptor and not bool.
+        """
+        expectedPath = self.dirname.child("foo.bar2")
+        expectedPath.touch()
+        
+        def cbNotified(_ignored1, _ignored2, _ignored3):
+            pass
+
+        self.inotify.watch(expectedPath, callbacks=[cbNotified])
+        self.inotify.ignore(expectedPath)
+        self.inotify.watch(expectedPath, callbacks=[cbNotified])
+        self.inotify.ignore(expectedPath)
+        self.inotify.watch(expectedPath, callbacks=[cbNotified])
+
+        expectedPath.remove()
+
+
     def test_complexSubdirectoryAutoAdd(self):
         """
         L{inotify.INotify} with autoAdd==True for a watched path
