=== modified file 'doc/core/howto/defer.xhtml'
--- doc/core/howto/defer.xhtml	2012-10-21 22:43:03 +0000
+++ doc/core/howto/defer.xhtml	2013-01-21 12:22:11 +0000
@@ -260,8 +260,8 @@
 errbacks: unless you explicitly <code class="python">return</code> a <code
 class="python">Failure</code> or (re-)raise an exception, the error stops
 propagating, and normal callbacks continue executing from that point (using the
-value returned from the errback). If the errback does returns a <code
-class="python">Failure</code> or raise an exception, then that is passed to the
+value returned from the errback). If the errback returns a <code
+class="python">Failure</code> or raises an exception, then that is passed to the
 next errback, and so on.</p>
 
 <p><em>Note:</em> If an errback doesn't return anything, then it effectively

