diff --git a/setup3.py b/setup3.py
index e53d273..e9839b9 100644
--- a/setup3.py
+++ b/setup3.py
@@ -37,13 +37,13 @@ def main():
         sys.path.insert(0, '.')
 
     from twisted.python.dist3 import modulesToInstall
-    from twisted.python.dist import STATIC_PACKAGE_METADATA
+    from twisted.python.dist import STATIC_PACKAGE_METADATA, _EXTRAS_REQUIRE
 
     args = STATIC_PACKAGE_METADATA.copy()
     args['install_requires'] = ["zope.interface >= 4.0.2"]
     args['py_modules'] = modulesToInstall
     args['cmdclass'] = {'sdist': DisabledSdist}
-
+    args['extras_require']=_EXTRAS_REQUIRE
     setup(**args)
 
 
