--- twisted/words/protocols/jabber/xmlstream.py
+++ twisted/words/protocols/jabber/xmlstream.py
@@ -43,7 +43,7 @@ def hashPassword(sid, password):
     Create a SHA1-digest string of a session identifier and password.
     """
     from twisted.python.hashlib import sha1
-    return sha1("%s%s" % (sid, password)).hexdigest()
+    return sha1(("%s%s" % (sid, password)).encode('ascii')).hexdigest()
 
 
 
