Index: words/protocols/irc.py
===================================================================
--- words/protocols/irc.py	(revision 33734)
+++ words/protocols/irc.py	(working copy)
@@ -2684,6 +2684,7 @@
     def buildProtocol(self, addr):
         p = self.protocol(client=self.client, queryData=self.queryData)
         p.factory = self
+        return p
 
     def clientConnectionFailed(self, unused_connector, unused_reason):
         self.client.dcc_sessions.remove(self)
