===================================================================
--- emacs/twisted-dev.el    (revision 32261)
+++ emacs/twisted-dev.el    (working copy)
@@ -58,6 +58,7 @@
   "*Directory to base all twisted-dev utilities out of."
   :group 'twisted-dev
   :type 'string)
+(make-variable-buffer-local 'twisted-dev-directory)
 
 
 (defcustom twisted-dev-tbformat "emacs"
@@ -71,6 +72,7 @@
   from."
   :group 'twisted-dev
   :type 'string)
+(make-variable-buffer-local 'twisted-dev-scratch-directory)
 
 (defcustom twisted-dev-confirm-run-all nil
   "If t, confirm running of all Twisted tests."
@@ -89,6 +91,7 @@
 
 (defvar test-case-name nil "Hello")
 (make-variable-buffer-local 'test-case-name)
+(put 'test-case-name 'safe-local-variable 'stringp)
 
 (defmacro with-cd (dirname &rest code)
   `(let ((old-dirname default-directory)
