--- twisted/cred/checkers.py	2003-07-23 16:54:32.000000000 -0500
+++ ../Twisted.wip/twisted/cred/checkers.py	2003-09-21 19:46:19.000000000 -0500
@@ -64,7 +64,7 @@
                 self.users[credentials.username]).addCallback(
                 self._cbPasswordMatch, credentials.username)
         else:
-            raise error.UnauthorizedLogin()
+            return failure.Failure(error.UnauthorizedLogin())
 
 
 class FilePasswordDB:
