Index: twisted/words/test/test_domish.py
===================================================================
--- twisted/words/test/test_domish.py	(revision 20169)
+++ twisted/words/test/test_domish.py	(working copy)
@@ -5,6 +5,8 @@
 Tests for L{twisted.words.xish.domish}.
 """
 
+import copy
+
 from twisted.trial import unittest
 from twisted.words.xish import domish
 
@@ -75,6 +77,23 @@
         self.assertEquals(e.hasAttribute("attrib2"), 0)
         self.assertEquals(e[("testns2", "attrib2")], "value2")
 
+    def test_deepcopy(self):
+        e = domish.Element(("testns", "foo"))
+        e["attrib1"] = "value1"
+        e[("testns2", "attrib2")] = "value2"
+        e.addElement("bar", content="somecontent")
+        e.addElement(("testns2", "baz")).addElement(('', "quux"))
+        e.addElement(("testns3", "quuux"))
+
+        e2 = copy.deepcopy(e)
+
+        self.assertEqual(e.toXml(), e2.toXml())
+        self.assertNotIdentical(e, e2)
+        self.assertNotIdentical(e.bar, e2.bar)
+        self.assertNotIdentical(e.baz, e2.baz)
+        self.assertNotIdentical(e.baz.quux, e2.baz.quux)
+        self.assertNotIdentical(e.quuux, e2.quuux)
+
 class DomishStreamTests:
     def setUp(self):
         self.doc_started = False
