Index: twisted/persisted/journal/__init__.py
===================================================================
--- twisted/persisted/journal/__init__.py	(revision 28446)
+++ twisted/persisted/journal/__init__.py	(working copy)
@@ -6,5 +6,14 @@
 """
 Command-journalling persistence framework inspired by Prevayler.
 
+This package is deprecated.
+
 Maintainer: Itamar Shtull-Trauring
 """
+
+import warnings
+warnings.warn(
+    "twisted.persisted.journal is deprecated since Twisted 10.0",
+    category=DeprecationWarning,
+    stacklevel=2)
+
