Index: twisted/persisted/test/test_journal.py
===================================================================
--- twisted/persisted/test/test_journal.py	(revision 0)
+++ twisted/persisted/test/test_journal.py	(revision 0)
@@ -0,0 +1,22 @@
+# Copyright (c) 2001-2010 Twisted Matrix Laboratories.
+# See LICENSE for details.
+
+"""
+Tests for deprecation of twisted.persisted.journal
+"""
+
+from twisted.trial.unittest import TestCase
+
+
+class JournalDeprecationTest(TestCase):
+
+    def test_deprecated(self):
+        from twisted.persisted import journal
+        warnings = self.flushWarnings()
+        self.assertEquals(
+            warnings[0]['message'],
+            'twisted.persisted.journal is deprecated since Twisted 10.0')
+        self.assertEquals(warnings[0]['category'], DeprecationWarning)
+        self.assertEquals(len(warnings), 1)
+
+
Index: twisted/persisted/test/__init__.py
===================================================================
Index: twisted/persisted/journal/__init__.py
===================================================================
--- twisted/persisted/journal/__init__.py	(revision 28446)
+++ twisted/persisted/journal/__init__.py	(working copy)
@@ -6,5 +6,14 @@
 """
 Command-journalling persistence framework inspired by Prevayler.
 
+This package is deprecated.
+
 Maintainer: Itamar Shtull-Trauring
 """
+
+import warnings
+warnings.warn(
+    "twisted.persisted.journal is deprecated since Twisted 10.0",
+    category=DeprecationWarning,
+    stacklevel=2)
+
