Index: amp.py
===================================================================
--- amp.py	(revision 26541)
+++ amp.py	(working copy)
@@ -2187,9 +2187,9 @@
         A verbose string representation which gives us information about this
         AMP connection.
         """
-        return '<%s %s at 0x%x>' % (
-            self.__class__.__name__,
-            self.innerProtocol, id(self))
+        inner_repr = self.innerProtocol and ' inner%s' % self.innerProtocol or ''
+        return '<%s%s at 0x%x>' % (
+            self.__class__.__name__, inner_repr, id(self))
 
 
     def makeConnection(self, transport):
