Index: test_amp.py
===================================================================
--- test_amp.py	(revision 26600)
+++ test_amp.py	(working copy)
@@ -1124,6 +1124,30 @@
         self.failUnless("AmpBox" in repr(amp.AmpBox()))
 
 
+    def test_innerProtocolInRepr(self):
+        """
+        Verify that L{AMP} objects output their inner protocol when set.
+        """
+        otherProto = TestProto(None, "outgoing data")
+        a = amp.AMP()
+        a.innerProtocol = otherProto
+        
+        self.assert_("AMP" in repr(a))
+        self.assert_("TestProto" in repr(a))
+        self.assert_("inner" in repr(a))
+
+
+    def test_innerProtocolNotInRepr(self):
+        """
+        Verify that L{AMP} objects do not output 'inner' when no inner is set
+        """
+        a = amp.AMP()
+        
+        self.assert_("AMP" in repr(a))
+        self.assertFalse("inner" in repr(a))
+        self.assertFalse("None" in repr(a))
+
+
     def test_simpleSSLRepr(self):
         """
         L{amp._TLSBox.__repr__} returns a string.
