Index: twisted/protocols/amp.py
===================================================================
--- twisted/protocols/amp.py	(revision 26600)
+++ twisted/protocols/amp.py	(working copy)
@@ -2187,9 +2187,12 @@
         A verbose string representation which gives us information about this
         AMP connection.
         """
-        return '<%s %s at 0x%x>' % (
-            self.__class__.__name__,
-            self.innerProtocol, id(self))
+        if self.innerProtocol:
+            innerRepr = ' inner %r' % (self.innerProtocol,)
+        else:
+            innerRepr = ''
+        return '<%s%s at 0x%x>' % (
+            self.__class__.__name__, innerRepr, id(self))
 
 
     def makeConnection(self, transport):
Index: twisted/test/test_amp.py
===================================================================
--- twisted/test/test_amp.py	(revision 26600)
+++ twisted/test/test_amp.py	(working copy)
@@ -1124,6 +1124,31 @@
         self.failUnless("AmpBox" in repr(amp.AmpBox()))
 
 
+    def test_innerProtocolInRepr(self):
+        """
+        Verify that L{AMP} objects output their innerProtocol when set.
+        """
+        otherProto = TestProto(None, "outgoing data")
+        a = amp.AMP()
+        a.innerProtocol = otherProto
+
+        self.assertIn("AMP", repr(a))
+        self.assertIn("TestProto", repr(a))
+        self.assertIn("inner", repr(a))
+
+
+    def test_innerProtocolNotInRepr(self):
+        """
+        Verify that L{AMP} objects do not output 'inner' when no innerProtocol
+        is set
+        """
+        a = amp.AMP()
+
+        self.assertIn("AMP", repr(a))
+        self.assertNotIn("inner", repr(a))
+        self.assertNotIn("None", repr(a))
+
+
     def test_simpleSSLRepr(self):
         """
         L{amp._TLSBox.__repr__} returns a string.
