diff --git twisted/protocols/ftp.py twisted/protocols/ftp.py
index 0407cf1..daf7f90 100644
--- twisted/protocols/ftp.py
+++ twisted/protocols/ftp.py
@@ -14,7 +14,6 @@ import operator
 import stat
 import errno
 import fnmatch
-import warnings
 
 try:
     import pwd, grp
diff --git twisted/test/test_ftp.py twisted/test/test_ftp.py
index 651b1be..7791ea1 100644
--- twisted/test/test_ftp.py
+++ twisted/test/test_ftp.py
@@ -13,7 +13,7 @@ import getpass
 from zope.interface import implements
 from zope.interface.verify import verifyClass
 
-from twisted.trial import unittest, util
+from twisted.trial import unittest
 from twisted.python.randbytes import insecureRandom
 from twisted.cred.portal import IRealm
 from twisted.protocols import basic
