diff --git twisted/protocols/ftp.py twisted/protocols/ftp.py
index e930ae7..0407cf1 100644
--- twisted/protocols/ftp.py
+++ twisted/protocols/ftp.py
@@ -2871,38 +2871,13 @@ class FTPClient(FTPClientBasic):
 
     def cwd(self, path):
         """
-        Issues the CWD (Change Working Directory) command. It's also
-        available as changeDirectory, which parses the result.
+        Issues the CWD (Change Working Directory) command.
 
         @return: a L{Deferred} that will be called when done.
         """
         return self.queueStringCommand('CWD ' + self.escapePath(path))
 
 
-    def changeDirectory(self, path):
-        """
-        Change the directory on the server and parse the result to determine
-        if it was successful or not.
-
-        @type path: C{str}
-        @param path: The path to which to change.
-
-        @return: a L{Deferred} which will be called back when the directory
-            change has succeeded or errbacked if an error occurrs.
-        """
-        warnings.warn(
-            "FTPClient.changeDirectory is deprecated in Twisted 8.2 and "
-            "newer.  Use FTPClient.cwd instead.",
-            category=DeprecationWarning,
-            stacklevel=2)
-
-        def cbResult(result):
-            if result[-1][:3] != '250':
-                return failure.Failure(CommandFailed(result))
-            return True
-        return self.cwd(path).addCallback(cbResult)
-
-
     def makeDirectory(self, path):
         """
         Make a directory
diff --git twisted/test/test_ftp.py twisted/test/test_ftp.py
index 0f03cd5..651b1be 100644
--- twisted/test/test_ftp.py
+++ twisted/test/test_ftp.py
@@ -27,12 +27,6 @@ from twisted.test import proto_helpers
 from twisted.protocols import ftp, loopback
 
 
-_changeDirectorySuppression = util.suppress(
-    category=DeprecationWarning,
-    message=(
-        r"FTPClient\.changeDirectory is deprecated in Twisted 8\.2 and "
-        r"newer\.  Use FTPClient\.cwd instead\."))
-
 if runtime.platform.isWindows():
     nonPOSIXSkip = "Cannot run on Windows"
 else:
@@ -2122,74 +2116,6 @@ class FTPClientTestCase(unittest.TestCase):
         return d
 
 
-    def test_changeDirectoryDeprecated(self):
-        """
-        L{ftp.FTPClient.changeDirectory} is deprecated and the direct caller of
-        it is warned of this.
-        """
-        self._testLogin()
-        d = self.assertWarns(
-            DeprecationWarning,
-            "FTPClient.changeDirectory is deprecated in Twisted 8.2 and "
-            "newer.  Use FTPClient.cwd instead.",
-            __file__,
-            lambda: self.client.changeDirectory('.'))
-        # This is necessary to make the Deferred fire.  The Deferred needs
-        # to fire so that tearDown doesn't cause it to errback and fail this
-        # or (more likely) a later test.
-        self.client.lineReceived('250 success')
-        return d
-
-
-    def test_changeDirectory(self):
-        """
-        Test the changeDirectory method.
-
-        L{ftp.FTPClient.changeDirectory} should return a Deferred which fires
-        with True if succeeded.
-        """
-        def cbCd(res):
-            self.assertEqual(res, True)
-
-        self._testLogin()
-        d = self.client.changeDirectory("bar/foo").addCallback(cbCd)
-        self.assertEqual(self.transport.value(), 'CWD bar/foo\r\n')
-        self.client.lineReceived('250 Requested File Action Completed OK')
-        return d
-    test_changeDirectory.suppress = [_changeDirectorySuppression]
-
-
-    def test_failedChangeDirectory(self):
-        """
-        Test a failure in the changeDirectory method.
-
-        The behaviour here is the same as a failed CWD.
-        """
-        self._testLogin()
-        d = self.client.changeDirectory("bar/foo")
-        self.assertFailure(d, ftp.CommandFailed)
-        self.assertEqual(self.transport.value(), 'CWD bar/foo\r\n')
-        self.client.lineReceived('550 bar/foo: No such file or directory')
-        return d
-    test_failedChangeDirectory.suppress = [_changeDirectorySuppression]
-
-
-    def test_strangeFailedChangeDirectory(self):
-        """
-        Test a strange failure in changeDirectory method.
-
-        L{ftp.FTPClient.changeDirectory} is stricter than CWD as it checks
-        code 250 for success.
-        """
-        self._testLogin()
-        d = self.client.changeDirectory("bar/foo")
-        self.assertFailure(d, ftp.CommandFailed)
-        self.assertEqual(self.transport.value(), 'CWD bar/foo\r\n')
-        self.client.lineReceived('252 I do what I want !')
-        return d
-    test_strangeFailedChangeDirectory.suppress = [_changeDirectorySuppression]
-
-
     def test_renameFromTo(self):
         """
         L{ftp.FTPClient.rename} issues I{RNTO} and I{RNFR} commands and returns
diff --git twisted/topfiles/6759.removal twisted/topfiles/6759.removal
new file mode 100644
index 0000000..afaeaa3
--- /dev/null
+++ twisted/topfiles/6759.removal
@@ -0,0 +1 @@
+FTPClient.changeDirectory was deprecated in 8.2 and is now removed.
