Index: twisted/mail/topfiles/4796.bugfix
===================================================================
--- twisted/mail/topfiles/4796.bugfix	(revision 0)
+++ twisted/mail/topfiles/4796.bugfix	(revision 0)
@@ -0,0 +1 @@
+Fixes a problem in the twistd mail plugin where a variable isn't initialized unless the --hostname 'option' is specified.
Index: twisted/mail/tap.py
===================================================================
--- twisted/mail/tap.py	(revision 30430)
+++ twisted/mail/tap.py	(working copy)
@@ -1,5 +1,5 @@
 # -*- test-case-name: twisted.mail.test.test_options -*-
-# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
+# Copyright (c) 2001-2011 Twisted Matrix Laboratories.
 # See LICENSE for details.
 
 
@@ -8,6 +8,7 @@
 
 import os
 import sys
+import socket
 
 from twisted.mail import mail
 from twisted.mail import maildir
@@ -32,7 +33,7 @@
         ["relay", "R", None,
             "Relay messages according to their envelope 'To', using the given"
             "path as a queue directory."],
-        ["hostname", "H", None, "The hostname by which to identify this server."],
+        ["hostname", "H", socket.gethostname(), "The hostname by which to identify this server."],
     ]
 
     optFlags = [
@@ -175,9 +176,8 @@
     if config['smtp']:
         f = smtpFactory()
         f.context = ctx
-        if config['hostname']:
-            f.domain = config['hostname']
-            f.fArgs = (f.domain,)
+        f.domain = config['hostname']
+        f.fArgs = (f.domain,)
         if config['esmtp']:
             f.fArgs = (None, None) + f.fArgs
         s = internet.TCPServer(config['smtp'], f)
