=== modified file 'twisted/mail/tap.py'
--- twisted/mail/tap.py	2011-02-14 04:45:15 +0000
+++ twisted/mail/tap.py	2011-03-08 02:21:02 +0000
@@ -133,6 +133,9 @@
                 raise usage.UsageError("Certificate file %r does not exist."
                                        % self['certificate'])
 
+        if self['esmtp'] and self['hostname'] is None:
+            raise usage.UsageError("--esmtp requires --hostname")
+
         if not self['disable-anonymous']:
             self.service.smtpPortal.registerChecker(checkers.AllowAnonymousAccess())
 

=== modified file 'twisted/mail/test/test_options.py'
--- twisted/mail/test/test_options.py	2011-02-14 04:45:15 +0000
+++ twisted/mail/test/test_options.py	2011-03-08 02:28:39 +0000
@@ -59,3 +59,11 @@
         self.assertEquals(warnings[0]['message'], msg)
 
 
+    def testEsmtpRequiresHostname(self):
+        """
+        Test that the --passwordfile option will emit a correct warning.
+        """
+        options = Options()
+        self.assertRaises(UsageError, options.parseOptions, ['--esmtp'])
+
+

