Index: postfix.py
===================================================================
--- postfix.py	(revision 26541)
+++ postfix.py	(working copy)
@@ -113,20 +113,3 @@
 
     def get(self, key):
         return defer.succeed(self.data.get(key))
-
-if __name__ == '__main__':
-    """Test app for PostfixTCPMapServer. Call with parameters
-    KEY1=VAL1 KEY2=VAL2 ..."""
-    from twisted.internet import reactor
-    log.startLogging(sys.stdout)
-    d = {}
-    for arg in sys.argv[1:]:
-        try:
-            k,v = arg.split('=', 1)
-        except ValueError:
-            k = arg
-            v = ''
-        d[k]=v
-    f=PostfixTCPMapDictServerFactory(d)
-    port = reactor.listenTCP(4242, f, interface='127.0.0.1')
-    reactor.run()
