--- http.py~	2007-01-02 00:41:33.000000000 +0100
+++ http.py	2007-02-21 23:01:17.000000000 +0100
@@ -343,7 +343,8 @@
         print "GOT RESPONSE %s: %s" % (num, resp)
         stream_mod.readStream(resp.stream, print_).addCallback(printdone)
     def sendReqs(proto):
-        proto.submitRequest(ClientRequest("GET", "/", {'Host':host}, None)).addCallback(gotResp, 1)
+        proto.setReadPersistent(PERSIST_PIPELINE)
+        proto.submitRequest(ClientRequest("GET", "/", {'Host':host}, None),False).addCallback(gotResp, 1)
         proto.submitRequest(ClientRequest("GET", "/foo", {'Host':host}, None)).addCallback(gotResp, 2)
     d.addCallback(sendReqs)
     del d
