Index: doc/web/howto/web-development.xhtml
===================================================================
--- doc/web/howto/web-development.xhtml	(revision 35287)
+++ doc/web/howto/web-development.xhtml	(working copy)
@@ -15,7 +15,7 @@
 <p>The development of a Twisted Web application should be orthogonal to its
 deployment.  This means is that if you are developing a web application, it
 should be a resource with children, and internal links.  Some of the children
-might use <a href="http://www.divmod.org/projects/nevow">Nevow</a>, some
+might use <a href="https://launchpad.net/nevow">Nevow</a>, some
 might be resources manually using <code>.write</code>, and so on.  Regardless,
 the code should be in a Python module, or package, <em>outside</em> the web
 tree.</p>
