--- ../tcp-ipv6-3014/twisted/test/test_application.py	2009-06-17 10:54:18.000000000 +0200
+++ twisted/test/test_application.py	2009-06-17 10:48:08.000000000 +0200
@@ -367,6 +367,25 @@
         factory.d.addCallback(lambda x : TestEcho.d)
         return factory.d
 
+    def testTCP6(self):
+        s = service.MultiService()
+        s.startService()
+        factory = protocol.ServerFactory()
+        factory.protocol = TestEcho
+        TestEcho.d = defer.Deferred()
+        t = internet.TCP6Server(0, factory)
+        t.setServiceParent(s)
+        num = t._port.getHost().port
+        factory = protocol.ClientFactory()
+        factory.d = defer.Deferred()
+        factory.protocol = Foo
+        factory.line = None
+        internet.TCP6Client('::', num, factory).setServiceParent(s)
+        factory.d.addCallback(self.assertEqual, 'lalala')
+        factory.d.addCallback(lambda x : s.stopService())
+        factory.d.addCallback(lambda x : TestEcho.d)
+        return factory.d
+
 
     def test_UDP(self):
         """
