diff --git twisted/mail/maildir.py twisted/mail/maildir.py
index 343342d..0620110 100644
--- twisted/mail/maildir.py
+++ twisted/mail/maildir.py
@@ -126,7 +126,7 @@ class MaildirMessage(mail.FileMessage):
         @param fp: The file in which to store the message while it is being
             received.
 
-        @type a: 2-L{tuple} of (E{1}) L{bytes}, (E{2}) L{bytes}
+        @type a: 2-L{tuple} of (0) L{bytes}, (1) L{bytes}
         @param a: Positional arguments for L{FileMessage.__init__}.
 
         @type kw: L{dict}
@@ -501,15 +501,15 @@ class MaildirMailbox(pop3.Mailbox):
 
     @ivar path: See L{__init__}.
 
-    @type list: L{list} of L{int} or 2-L{tuple} of (E{1}) file-like object,
-        (E{2}) L{bytes}
+    @type list: L{list} of L{int} or 2-L{tuple} of (0) file-like object,
+        (1) L{bytes}
     @ivar list: Information about the messages in the mailbox. For undeleted
         messages, the file containing the message and the
         full path name of the file are stored.  Deleted messages are indicated
         by 0.
 
-    @type deleted: L{dict} of 2-L{tuple} of (E{1}) file-like object,
-        (E{2}) L{bytes} -> L{bytes}
+    @type deleted: L{dict} of 2-L{tuple} of (0) file-like object,
+        (1) L{bytes} -> L{bytes}
     @type deleted: A mapping of the information about a file before it was
         deleted to the full path name of the deleted file in the .Trash/
         subfolder.
@@ -874,8 +874,8 @@ class MaildirDirdbmDomain(AbstractMaildirDomain):
         @param interfaces: A group of interfaces, one of which the avatar
             must support.
 
-        @rtype: 3-L{tuple} of (E{1}) L{IMailbox <pop3.IMailbox>},
-            (E{2}) L{IMailbox <pop3.IMailbox>} provider, (E{3}) no-argument
+        @rtype: 3-L{tuple} of (0) L{IMailbox <pop3.IMailbox>},
+            (1) L{IMailbox <pop3.IMailbox>} provider, (2) no-argument
             callable
         @return: A tuple of the supported interface, a mailbox, and a
             logout function.
