diff --git a/doc/core/development/policy/writing-standard.xhtml b/doc/core/development/policy/writing-standard.xhtml
index 141842b..03c9779 100644
--- a/doc/core/development/policy/writing-standard.xhtml
+++ b/doc/core/development/policy/writing-standard.xhtml
@@ -51,7 +51,9 @@
     <p>Documentation must use American English spelling, and where possible
     avoid any local variants of either vocabulary or grammar. Grammatically
     complex sentences should ideally be avoided: these make reading
-    unnecessarily difficult, particularly for non-native speakers.</p>
+    unnecessarily difficult, particularly for non-native speakers.
+
+    When using pronouns in reference to a hypothetical person, such as "a user of a website written with twisted.web", gender neutral pronouns (they/their/ them) should be used.</p>
 
     <h2>Evangelism and usage documents</h2>
 
