diff --git a/doc/core/development/policy/writing-standard.xhtml b/doc/core/development/policy/writing-standard.xhtml
index 141842b..60e04db 100644
--- a/doc/core/development/policy/writing-standard.xhtml
+++ b/doc/core/development/policy/writing-standard.xhtml
@@ -53,6 +53,8 @@
     complex sentences should ideally be avoided: these make reading
     unnecessarily difficult, particularly for non-native speakers.</p>
 
+    <p>When referring to a hypothetical person, (such as "a user of a website written with twisted.web"), gender neutral pronouns (they/their/them) should be used.</p>
+
     <h2>Evangelism and usage documents</h2>
 
     <p>The Twisted documentation should maintain a reasonable distinction
