--- common.py	2006-11-26 17:39:59.000000000 +0100
+++ common.py	2006-11-27 11:19:56.000000000 +0100
@@ -88,6 +88,9 @@
     def lookupZone(self, name, timeout = None):
         return self._lookup(name, dns.IN, dns.AXFR, timeout)
 
+    def lookupNamingAuthorityPointer(self,name,timeout=None):
+        return self._lookup(name, dns.IN, dns.NAPTR, timeout)
+
     def lookupAllRecords(self, name, timeout = None):
         return self._lookup(name, dns.IN, dns.ALL_RECORDS, timeout)
 
@@ -152,7 +155,7 @@
     dns.RP:    'lookupResponsibility',
     dns.AFSDB: 'lookupAFSDatabase',
     dns.SRV:   'lookupService',
-
+    dns.NAPTR: 'lookupNamingAuthorityPointer',
     dns.AXFR:         'lookupZone',
     dns.ALL_RECORDS:  'lookupAllRecords',
 }
