--- test/test_names.py.old	2006-11-27 11:12:55.000000000 +0100
+++ test/test_names.py	2006-11-27 11:11:24.000000000 +0100
@@ -79,6 +79,7 @@
             dns.Record_AFSDB(subtype=1, hostname='afsdb.test-domain.com'),
             dns.Record_RP(mbox='whatever.i.dunno', txt='some.more.text'),
             dns.Record_WKS('12.54.78.12', socket.IPPROTO_TCP, '\x12\x01\x16\xfe\xc1\x00\x01'),
+	        dns.Record_NAPTR(100,10,"u","sip+E2U","!^.*$!sip:information@tele2.se!"),
             dns.Record_AAAA('AF43:5634:1294:AFCB:56AC:48EF:34C3:01FF')],
         'http.tcp.test-domain.com': [
             dns.Record_SRV(257, 16383, 43690, 'some.other.place.fool')
@@ -378,6 +379,13 @@
             self.resolver.lookupAddress("anothertest-domain.com"),
             [dns.Record_A('1.2.3.4', ttl=19283784)]
         )
+
+    def testNAPTR(self):
+        """Test DNS 'NAPTR' record queries"""
+        return self.namesTest(
+            self.resolver.lookupNAPTR('test-domain.com'),
+            [dns.Record_NAPTR(100,10,"u","sip+E2U","!^.*$!sip:information@tele2.se!",ttl=19283784)]
+        )
             
 class HelperTestCase(unittest.TestCase):
     def testSerialGenerator(self):
