diff --git a/doc/core/howto/cred.xhtml b/doc/core/howto/cred.xhtml
index c0eedd4..46bedb6 100644
--- a/doc/core/howto/cred.xhtml
+++ b/doc/core/howto/cred.xhtml
@@ -218,7 +218,7 @@ the DELE command can only be used after you are logged in:</p>
    <h3>Application implementation</h3>
 
 <p>The application developer can implement realms and credential checkers. For example,
-she might implement a realm that returns IMailbox implementing avatars, using MySQL
+they might implement a realm that returns IMailbox implementing avatars, using MySQL
 for storage, or perhaps a credential checker that uses LDAP for authentication.
 In the following example, the Realm for a simple remote object service (using
 Twisted's Perspective Broker protocol) is implemented:</p>
diff --git a/doc/core/howto/options.xhtml b/doc/core/howto/options.xhtml
index 3ca8fbb..c34e1c8 100644
--- a/doc/core/howto/options.xhtml
+++ b/doc/core/howto/options.xhtml
@@ -25,12 +25,12 @@
     class="API">twisted.python.usage.Options</code>. It uses
     Python's reflection facilities to provide an easy to use yet
     flexible interface to the command line. While most command line
-    processors either force the application writer to write her own
+    processors either force the application writer to write their own
     loops, or have arbitrary limitations on the command line (the
     most common one being not being able to have more then one
     instance of a specific option, thus rendering the idiom
     <code class="shell">program -v -v -v</code> impossible), Twisted allows the
-    programmer to decide how much control she wants.</p>
+    programmer to decide how much control they want.</p>
 
     <p>The <code>Options</code> class is used by subclassing. Since
     a lot of time it will be used in the <code
diff --git a/doc/core/howto/trial.xhtml b/doc/core/howto/trial.xhtml
index f0a0cc2..6c07a6d 100644
--- a/doc/core/howto/trial.xhtml
+++ b/doc/core/howto/trial.xhtml
@@ -552,7 +552,7 @@ created.</p>
 <p>Currently, if you send an invalid command or invalid arguments to our
 server, it logs an exception and closes the connection. This is a perfectly
 valid behavior, but for the sake of this tutorial, we want to return an error
-to the user if he sends invalid operators, and log any errors on server side.
+to the user if they send invalid operators, and log any errors on server side.
 So we'll want a test like this:</p>
 
 <pre class="python">
diff --git a/doc/core/howto/tutorial/library.xhtml b/doc/core/howto/tutorial/library.xhtml
index 1cd4172..9fed2b9 100644
--- a/doc/core/howto/tutorial/library.xhtml
+++ b/doc/core/howto/tutorial/library.xhtml
@@ -65,9 +65,9 @@ method that will also help build .tap files later:</p>
 </pre>
 
 
-<p>Note: the finger <em>user</em> still has ultimate power: he can use
- <code>makeService</code>, or he can use the lower-level interface if he has
-specific needs (maybe an IRC server on some other port? Maybe we want the
+<p>Note: the finger <em>user</em> still has ultimate power: they can use
+<code>makeService</code>, or they can use the lower-level interface if they
+have specific needs (maybe an IRC server on some other port? Maybe we want the
 non-SSL webserver to listen only locally?  etc. etc.)  This is an important
 design principle: never force a layer of abstraction: allow usage of layers of
 abstractions.</p>
