*** log.py	Sat Jan 28 19:20:33 2006
--- /usr/lib/zope-3.2.0/lib/python/twisted/web2/log.py	Sat Jan 28 20:46:34 2006
***************
*** 148,158 ****
              request.method,
              request.uri,
              '.'.join([str(x) for x in request.clientproto]))
!         
          self.logMessage(
              '%s - %s [%s] "%s" %s %d "%s" "%s"' %(
!                 request.chanRequest.getRemoteHost().host,
                  # XXX: Where to get user from?
                  "-",
                  self.logDateString(
                      response.headers.getHeader('date', 0)),
--- 148,165 ----
              request.method,
              request.uri,
              '.'.join([str(x) for x in request.clientproto]))
!         proxyAddresses = request.headers.getRawHeaders('x-forwarded-for',None)
!         singleHost = request.chanRequest.getRemoteHost().host
!         if proxyAddresses:
!             proxyAddresses.append(singleHost)
!             logAddress =  '>'.join(proxyAddresses)
!         else:
!             logAddress = singleHost
          self.logMessage(
              '%s - %s [%s] "%s" %s %d "%s" "%s"' %(
!                 #request.chanRequest.getRemoteHost().host,
                  # XXX: Where to get user from?
+                 logAddress,
                  "-",
                  self.logDateString(
                      response.headers.getHeader('date', 0)),
