diff --git a/twisted/protocols/ftp.py b/twisted/protocols/ftp.py
index 9f034c0..cbb1621 100644
--- a/twisted/protocols/ftp.py
+++ b/twisted/protocols/ftp.py
@@ -25,7 +25,6 @@ from zope.interface import Interface, implements
 
 # Twisted Imports
 from twisted import copyright
-from twisted.python.deprecate import warnAboutFunction
 from twisted.internet import reactor, interfaces, protocol, error, defer
 from twisted.protocols import basic, policies
 
@@ -933,7 +932,7 @@ class FTP(object, basic.LineReceiver, policies.TimeoutMixin):
         """
         Inspect an element from the list returned by an L{IFTPShell.list}
         implementation to make sure the content is formated to be
-        send on the wire. 
+        send on the wire.
 
         @param name: The name of a file, as returned by L{IFTPShell.list}.
         @type name: L{bytes} or L{unicode}
@@ -1581,14 +1580,14 @@ class IFTPShell(Interface):
         child of the directory.
 
         @param path: The path, as a list of segments, to list
-        @type path: C{list} of C{unicode}
+        @type path: C{list} of C{unicode} or C{bytes}
 
         @param keys: A tuple of keys desired in the resulting
         dictionaries.
 
         @return: A Deferred which fires with a list of (name, list),
-        where the name is the name of the entry as a unicode string
-        and each list contains values corresponding to the requested
+        where the name is the name of the entry as a unicode string or
+        bytes and each list contains values corresponding to the requested
         keys.  The following are possible elements of keys, and the
         values which should be returned for them:
 
diff --git a/twisted/test/test_ftp.py b/twisted/test/test_ftp.py
index 6fb7064..1bea121 100644
--- a/twisted/test/test_ftp.py
+++ b/twisted/test/test_ftp.py
@@ -734,6 +734,18 @@ class FTPServerPasvDataConnectionTestCase(FTPServerTestCase):
             '0 Jan 01  1970 my resum\xc3\xa9\r\n')
 
 
+    def test_LISTNonASCIIBytes(self):
+        """
+        Support for returning byte strings from L{IFTPShell.list} is deprecated
+        and doing so results in a warning, but in the filename being sent as-is.
+        """
+        return self._listTestHelper(
+            "LIST",
+            ('my resum\xc3\xa9', (0, 1, 0777, 0, 0, 'user', 'group')),
+            'drwxrwxrwx   0 user      group                   '
+            '0 Jan 01  1970 my resum\xc3\xa9\r\n')
+
+
     def testManyLargeDownloads(self):
         # Login
         d = self._anonymousLogin()
@@ -829,6 +841,17 @@ class FTPServerPasvDataConnectionTestCase(FTPServerTestCase):
             'my resum\xc3\xa9\r\n')
 
 
+    def test_NLSTNonASCIIBytes(self):
+        """
+        NLST will just pass the non-Unicode data to lower level.
+        """
+        return self._listTestHelper(
+            "NLST",
+            ('my resum\xc3\xa9', (0, 1, 0777, 0, 0, 'user', 'group')),
+            'my resum\xc3\xa9\r\n')
+
+
+
     def test_NLSTOnPathToFile(self):
         """
         NLST on an existent file returns only the path to that file.
