Index: twisted/topfiles/5411.bugfix
===================================================================
--- twisted/topfiles/5411.bugfix	(revision 0)
+++ twisted/topfiles/5411.bugfix	(revision 0)
@@ -0,0 +1 @@
+Fix bug with FTP directory listing trying to write Unicode data (from filenames) to the TCP transport.
Index: twisted/protocols/ftp.py
===================================================================
--- twisted/protocols/ftp.py	(revision 33225)
+++ twisted/protocols/ftp.py	(working copy)
@@ -415,7 +415,7 @@
             'group': group[:8],
             'size': size,
             'date': formatDate(time.gmtime(modified)),
-            'name': name}
+            'name': name.encode("utf-8")}
 
     def sendListResponse(self, name, response):
         self.sendLine(self._formatOneListResponse(name, *response))
