Index: twisted/internet/udp.py
===================================================================
--- twisted/internet/udp.py	(revision 27328)
+++ twisted/internet/udp.py	(working copy)
@@ -93,9 +93,10 @@
         # Make sure that if we listened on port 0, we update that to
         # reflect what the OS actually assigned us.
         self._realPortNumber = skt.getsockname()[1]
+        
+        if self.protocol.noisy:
+            log.msg("%s starting on %s"%(self.protocol.__class__, self._realPortNumber))
 
-        log.msg("%s starting on %s"%(self.protocol.__class__, self._realPortNumber))
-
         self.connected = 1
         self.socket = skt
         self.fileno = self.socket.fileno
@@ -200,7 +201,8 @@
     def connectionLost(self, reason=None):
         """Cleans up my socket.
         """
-        log.msg('(Port %s Closed)' % self._realPortNumber)
+        if self.protocol.noisy:
+            log.msg('(Port %s Closed)' % self._realPortNumber)
         self._realPortNumber = None
         base.BasePort.connectionLost(self, reason)
         if hasattr(self, "protocol"):
Index: twisted/internet/tcp.py
===================================================================
--- twisted/internet/tcp.py	(revision 27328)
+++ twisted/internet/tcp.py	(working copy)
@@ -857,9 +857,9 @@
         # Make sure that if we listened on port 0, we update that to
         # reflect what the OS actually assigned us.
         self._realPortNumber = skt.getsockname()[1]
+        if self.factory.noisy:
+            log.msg("%s starting on %s" % (self.factory.__class__, self._realPortNumber))
 
-        log.msg("%s starting on %s" % (self.factory.__class__, self._realPortNumber))
-
         # The order of the next 6 lines is kind of bizarre.  If no one
         # can explain it, perhaps we should re-arrange them.
         self.factory.doStart()
Index: twisted/internet/protocol.py
===================================================================
--- twisted/internet/protocol.py	(revision 27328)
+++ twisted/internet/protocol.py	(working copy)
@@ -34,7 +34,7 @@
     protocol = None
 
     numPorts = 0
-    noisy = True
+    noisy = False
 
     def doStart(self):
         """Make sure startFactory is called.
@@ -516,7 +516,7 @@
 
     transport = None
     numPorts = 0
-    noisy = True
+    noisy = False
 
     def __getstate__(self):
         d = self.__dict__.copy()
